<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Literature Cited | Applied Biological Data Analysis</title>
  <meta name="description" content="Helping biologists to become more informed users of R and statistical methods." />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Literature Cited | Applied Biological Data Analysis" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://greenquanteco.github.io/index.html" />
  <meta property="og:image" content="https://greenquanteco.github.io/index.html/C:/Users/ngreen62/OneDrive - Kennesaw State University/ksu/website/lab_logo_02.jpg" />
  <meta property="og:description" content="Helping biologists to become more informed users of R and statistical methods." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Literature Cited | Applied Biological Data Analysis" />
  
  <meta name="twitter:description" content="Helping biologists to become more informed users of R and statistical methods." />
  <meta name="twitter:image" content="https://greenquanteco.github.io/index.html/C:/Users/ngreen62/OneDrive - Kennesaw State University/ksu/website/lab_logo_02.jpg" />

<meta name="author" content="Nick Green, Kennesaw State University" />


<meta name="date" content="2022-01-12" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="9.6-conclusions.html"/>

<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a>
<ul>
<li class="chapter" data-level="" data-path="license-and-permissions.html"><a href="license-and-permissions.html"><i class="fa fa-check"></i>License and permissions</a></li>
<li class="chapter" data-level="" data-path="course-description.html"><a href="course-description.html"><i class="fa fa-check"></i>Course description</a></li>
<li class="chapter" data-level="" data-path="course-objectives.html"><a href="course-objectives.html"><i class="fa fa-check"></i>Course objectives</a></li>
<li class="chapter" data-level="" data-path="course-requirements.html"><a href="course-requirements.html"><i class="fa fa-check"></i>Course requirements</a></li>
<li class="chapter" data-level="" data-path="recommended-reading.html"><a href="recommended-reading.html"><i class="fa fa-check"></i>Recommended reading</a></li>
<li class="chapter" data-level="" data-path="course-organization.html"><a href="course-organization.html"><i class="fa fa-check"></i>Course organization</a></li>
<li class="chapter" data-level="" data-path="about-the-author.html"><a href="about-the-author.html"><i class="fa fa-check"></i>About the author</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="1-mod-01.html"><a href="1-mod-01.html"><i class="fa fa-check"></i><b>1</b> Statistics in modern biology</a>
<ul>
<li class="chapter" data-level="1.1" data-path="1.1-overview.html"><a href="1.1-overview.html"><i class="fa fa-check"></i><b>1.1</b> Overview</a></li>
<li class="chapter" data-level="1.2" data-path="1.2-statistics-in-modern-biology.html"><a href="1.2-statistics-in-modern-biology.html"><i class="fa fa-check"></i><b>1.2</b> Statistics in modern biology</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="1.2-statistics-in-modern-biology.html"><a href="1.2-statistics-in-modern-biology.html#the-scientific-method"><i class="fa fa-check"></i><b>1.2.1</b> The scientific method</a></li>
<li class="chapter" data-level="1.2.2" data-path="1.2-statistics-in-modern-biology.html"><a href="1.2-statistics-in-modern-biology.html#example-data-analysis"><i class="fa fa-check"></i><b>1.2.2</b> Example data analysis</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="1.3-misuses-of-statistics.html"><a href="1.3-misuses-of-statistics.html"><i class="fa fa-check"></i><b>1.3</b> Misuses of statistics</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="1.3-misuses-of-statistics.html"><a href="1.3-misuses-of-statistics.html#proving-the-trivial-and-meaningless-hypotheses"><i class="fa fa-check"></i><b>1.3.1</b> “Proving” the trivial and meaningless hypotheses</a></li>
<li class="chapter" data-level="1.3.2" data-path="1.3-misuses-of-statistics.html"><a href="1.3-misuses-of-statistics.html#inappropriate-methods"><i class="fa fa-check"></i><b>1.3.2</b> Inappropriate methods</a></li>
<li class="chapter" data-level="1.3.3" data-path="1.3-misuses-of-statistics.html"><a href="1.3-misuses-of-statistics.html#p-hacking-and-data-dredging"><i class="fa fa-check"></i><b>1.3.3</b> <em>P</em>-hacking and data dredging</a></li>
<li class="chapter" data-level="1.3.4" data-path="1.3-misuses-of-statistics.html"><a href="1.3-misuses-of-statistics.html#inadequate-sample-sizes-and-pseudoreplication"><i class="fa fa-check"></i><b>1.3.4</b> Inadequate sample sizes and pseudoreplication</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="1.4-p-values-and-null-hypothesis-significance-testing-nhst.html"><a href="1.4-p-values-and-null-hypothesis-significance-testing-nhst.html"><i class="fa fa-check"></i><b>1.4</b> <em>P</em>-values and null hypothesis significance testing (NHST)</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="1.4-p-values-and-null-hypothesis-significance-testing-nhst.html"><a href="1.4-p-values-and-null-hypothesis-significance-testing-nhst.html#definition"><i class="fa fa-check"></i><b>1.4.1</b> Definition</a></li>
<li class="chapter" data-level="1.4.2" data-path="1.4-p-values-and-null-hypothesis-significance-testing-nhst.html"><a href="1.4-p-values-and-null-hypothesis-significance-testing-nhst.html#history-and-status-of-p-values"><i class="fa fa-check"></i><b>1.4.2</b> History and status of <em>P</em>-values</a></li>
<li class="chapter" data-level="1.4.3" data-path="1.4-p-values-and-null-hypothesis-significance-testing-nhst.html"><a href="1.4-p-values-and-null-hypothesis-significance-testing-nhst.html#where-p-values-come-from"><i class="fa fa-check"></i><b>1.4.3</b> Where <em>P</em>-values come from</a></li>
<li class="chapter" data-level="1.4.4" data-path="1.4-p-values-and-null-hypothesis-significance-testing-nhst.html"><a href="1.4-p-values-and-null-hypothesis-significance-testing-nhst.html#what-p-values-mean-and-do-not-mean"><i class="fa fa-check"></i><b>1.4.4</b> What <em>P</em>-values mean and do not mean</a></li>
<li class="chapter" data-level="1.4.5" data-path="1.4-p-values-and-null-hypothesis-significance-testing-nhst.html"><a href="1.4-p-values-and-null-hypothesis-significance-testing-nhst.html#do-you-need-a-p-value"><i class="fa fa-check"></i><b>1.4.5</b> Do you need a <em>P</em>-value?</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="1.5-alternatives-to-nhst.html"><a href="1.5-alternatives-to-nhst.html"><i class="fa fa-check"></i><b>1.5</b> Alternatives to NHST</a>
<ul>
<li class="chapter" data-level="1.5.1" data-path="1.5-alternatives-to-nhst.html"><a href="1.5-alternatives-to-nhst.html#bayesian-inference"><i class="fa fa-check"></i><b>1.5.1</b> Bayesian inference</a></li>
<li class="chapter" data-level="1.5.2" data-path="1.5-alternatives-to-nhst.html"><a href="1.5-alternatives-to-nhst.html#information-theoretic-methods"><i class="fa fa-check"></i><b>1.5.2</b> Information-theoretic methods</a></li>
<li class="chapter" data-level="1.5.3" data-path="1.5-alternatives-to-nhst.html"><a href="1.5-alternatives-to-nhst.html#machine-learning"><i class="fa fa-check"></i><b>1.5.3</b> Machine learning</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="2-mod-02.html"><a href="2-mod-02.html"><i class="fa fa-check"></i><b>2</b> Introduction to R</a>
<ul>
<li class="chapter" data-level="2.1" data-path="2.1-getting-started-with-r.html"><a href="2.1-getting-started-with-r.html"><i class="fa fa-check"></i><b>2.1</b> Getting started with R</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="2.1-getting-started-with-r.html"><a href="2.1-getting-started-with-r.html#what-is-r"><i class="fa fa-check"></i><b>2.1.1</b> What is R?</a></li>
<li class="chapter" data-level="2.1.2" data-path="2.1-getting-started-with-r.html"><a href="2.1-getting-started-with-r.html#advantages-of-r"><i class="fa fa-check"></i><b>2.1.2</b> Advantages of R</a></li>
<li class="chapter" data-level="2.1.3" data-path="2.1-getting-started-with-r.html"><a href="2.1-getting-started-with-r.html#disadvantages-of-r"><i class="fa fa-check"></i><b>2.1.3</b> Disadvantages of R</a></li>
<li class="chapter" data-level="2.1.4" data-path="2.1-getting-started-with-r.html"><a href="2.1-getting-started-with-r.html#base-r-and-vs.-tidyverse"><i class="fa fa-check"></i><b>2.1.4</b> Base R and (vs.?) <code>tidyverse</code></a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="2.2-download-and-install-r-and-rstudio.html"><a href="2.2-download-and-install-r-and-rstudio.html"><i class="fa fa-check"></i><b>2.2</b> Download and install R (and RStudio)</a></li>
<li class="chapter" data-level="2.3" data-path="2.3-using-r.html"><a href="2.3-using-r.html"><i class="fa fa-check"></i><b>2.3</b> Using R</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="2.3-using-r.html"><a href="2.3-using-r.html#using-the-base-r-gui"><i class="fa fa-check"></i><b>2.3.1</b> Using the base R GUI</a></li>
<li class="chapter" data-level="2.3.2" data-path="2.3-using-r.html"><a href="2.3-using-r.html#using-r-in-rstudio"><i class="fa fa-check"></i><b>2.3.2</b> Using R in RStudio</a></li>
<li class="chapter" data-level="2.3.3" data-path="2.3-using-r.html"><a href="2.3-using-r.html#using-r-with-other-programs"><i class="fa fa-check"></i><b>2.3.3</b> Using R with other programs</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="2.4-mod-02-first.html"><a href="2.4-mod-02-first.html"><i class="fa fa-check"></i><b>2.4</b> A first R session</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="2.4-mod-02-first.html"><a href="2.4-mod-02-first.html#import-data"><i class="fa fa-check"></i><b>2.4.1</b> Import data</a></li>
<li class="chapter" data-level="2.4.2" data-path="2.4-mod-02-first.html"><a href="2.4-mod-02-first.html#explore-and-visualize-data"><i class="fa fa-check"></i><b>2.4.2</b> Explore and visualize data</a></li>
<li class="chapter" data-level="2.4.3" data-path="2.4-mod-02-first.html"><a href="2.4-mod-02-first.html#transform-data"><i class="fa fa-check"></i><b>2.4.3</b> Transform data</a></li>
<li class="chapter" data-level="2.4.4" data-path="2.4-mod-02-first.html"><a href="2.4-mod-02-first.html#analyze-data"><i class="fa fa-check"></i><b>2.4.4</b> Analyze data</a></li>
<li class="chapter" data-level="2.4.5" data-path="2.4-mod-02-first.html"><a href="2.4-mod-02-first.html#write-out-results"><i class="fa fa-check"></i><b>2.4.5</b> Write out results</a></li>
<li class="chapter" data-level="2.4.6" data-path="2.4-mod-02-first.html"><a href="2.4-mod-02-first.html#save-your-work"><i class="fa fa-check"></i><b>2.4.6</b> Save your work?</a></li>
<li class="chapter" data-level="2.4.7" data-path="2.4-mod-02-first.html"><a href="2.4-mod-02-first.html#whats-next"><i class="fa fa-check"></i><b>2.4.7</b> What’s next?</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="2.5-write-and-execute-commands-in-the-r-console.html"><a href="2.5-write-and-execute-commands-in-the-r-console.html"><i class="fa fa-check"></i><b>2.5</b> Write and execute commands in the R console</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="2.5-write-and-execute-commands-in-the-r-console.html"><a href="2.5-write-and-execute-commands-in-the-r-console.html#r-commandsbasics"><i class="fa fa-check"></i><b>2.5.1</b> R commands–basics</a></li>
<li class="chapter" data-level="2.5.2" data-path="2.5-write-and-execute-commands-in-the-r-console.html"><a href="2.5-write-and-execute-commands-in-the-r-console.html#elements-of-r-code"><i class="fa fa-check"></i><b>2.5.2</b> Elements of R code</a></li>
<li class="chapter" data-level="2.5.3" data-path="2.5-write-and-execute-commands-in-the-r-console.html"><a href="2.5-write-and-execute-commands-in-the-r-console.html#the-r-workspace"><i class="fa fa-check"></i><b>2.5.3</b> The R workspace</a></li>
<li class="chapter" data-level="2.5.4" data-path="2.5-write-and-execute-commands-in-the-r-console.html"><a href="2.5-write-and-execute-commands-in-the-r-console.html#r-code-basics-assignment-and-operators"><i class="fa fa-check"></i><b>2.5.4</b> R code basics: assignment and operators</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="2.6-mod-02-struct.html"><a href="2.6-mod-02-struct.html"><i class="fa fa-check"></i><b>2.6</b> Basic R data structures</a>
<ul>
<li class="chapter" data-level="2.6.1" data-path="2.6-mod-02-struct.html"><a href="2.6-mod-02-struct.html#vectors"><i class="fa fa-check"></i><b>2.6.1</b> Vectors</a></li>
<li class="chapter" data-level="2.6.2" data-path="2.6-mod-02-struct.html"><a href="2.6-mod-02-struct.html#data-frames"><i class="fa fa-check"></i><b>2.6.2</b> Data frames</a></li>
<li class="chapter" data-level="2.6.3" data-path="2.6-mod-02-struct.html"><a href="2.6-mod-02-struct.html#matrices-and-arrays"><i class="fa fa-check"></i><b>2.6.3</b> Matrices and arrays</a></li>
<li class="chapter" data-level="2.6.4" data-path="2.6-mod-02-struct.html"><a href="2.6-mod-02-struct.html#lists"><i class="fa fa-check"></i><b>2.6.4</b> Lists</a></li>
<li class="chapter" data-level="2.6.5" data-path="2.6-mod-02-struct.html"><a href="2.6-mod-02-struct.html#s4-objects"><i class="fa fa-check"></i><b>2.6.5</b> S4 objects</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="2.7-r-data-types.html"><a href="2.7-r-data-types.html"><i class="fa fa-check"></i><b>2.7</b> R data types</a>
<ul>
<li class="chapter" data-level="2.7.1" data-path="2.7-r-data-types.html"><a href="2.7-r-data-types.html#character-type"><i class="fa fa-check"></i><b>2.7.1</b> Character type</a></li>
<li class="chapter" data-level="2.7.2" data-path="2.7-r-data-types.html"><a href="2.7-r-data-types.html#numeric-type"><i class="fa fa-check"></i><b>2.7.2</b> Numeric type</a></li>
<li class="chapter" data-level="2.7.3" data-path="2.7-r-data-types.html"><a href="2.7-r-data-types.html#integer-type"><i class="fa fa-check"></i><b>2.7.3</b> Integer type</a></li>
<li class="chapter" data-level="2.7.4" data-path="2.7-r-data-types.html"><a href="2.7-r-data-types.html#logical-type"><i class="fa fa-check"></i><b>2.7.4</b> Logical type</a></li>
<li class="chapter" data-level="2.7.5" data-path="2.7-r-data-types.html"><a href="2.7-r-data-types.html#special-values"><i class="fa fa-check"></i><b>2.7.5</b> Special values</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="2.8-manage-r-code-as-scripts-.r-files.html"><a href="2.8-manage-r-code-as-scripts-.r-files.html"><i class="fa fa-check"></i><b>2.8</b> Manage R code as scripts (.r files)</a></li>
<li class="chapter" data-level="2.9" data-path="2.9-manage-and-use-r-packages.html"><a href="2.9-manage-and-use-r-packages.html"><i class="fa fa-check"></i><b>2.9</b> Manage and use R packages</a>
<ul>
<li class="chapter" data-level="2.9.1" data-path="2.9-manage-and-use-r-packages.html"><a href="2.9-manage-and-use-r-packages.html#your-r-library"><i class="fa fa-check"></i><b>2.9.1</b> Your R library</a></li>
<li class="chapter" data-level="2.9.2" data-path="2.9-manage-and-use-r-packages.html"><a href="2.9-manage-and-use-r-packages.html#installing-packages-using-the-r-gui"><i class="fa fa-check"></i><b>2.9.2</b> Installing packages using the R GUI</a></li>
<li class="chapter" data-level="2.9.3" data-path="2.9-manage-and-use-r-packages.html"><a href="2.9-manage-and-use-r-packages.html#installing-packages-in-rstudio"><i class="fa fa-check"></i><b>2.9.3</b> Installing packages in RStudio</a></li>
<li class="chapter" data-level="2.9.4" data-path="2.9-manage-and-use-r-packages.html"><a href="2.9-manage-and-use-r-packages.html#installing-packages-using-the-r-console"><i class="fa fa-check"></i><b>2.9.4</b> Installing packages using the R console</a></li>
<li class="chapter" data-level="2.9.5" data-path="2.9-manage-and-use-r-packages.html"><a href="2.9-manage-and-use-r-packages.html#working-with-packages-in-r"><i class="fa fa-check"></i><b>2.9.5</b> Working with packages in R</a></li>
<li class="chapter" data-level="2.9.6" data-path="2.9-manage-and-use-r-packages.html"><a href="2.9-manage-and-use-r-packages.html#package-dependencies"><i class="fa fa-check"></i><b>2.9.6</b> Package dependencies</a></li>
<li class="chapter" data-level="2.9.7" data-path="2.9-manage-and-use-r-packages.html"><a href="2.9-manage-and-use-r-packages.html#citing-packages"><i class="fa fa-check"></i><b>2.9.7</b> Citing packages</a></li>
</ul></li>
<li class="chapter" data-level="2.10" data-path="2.10-r-documentation.html"><a href="2.10-r-documentation.html"><i class="fa fa-check"></i><b>2.10</b> R documentation</a>
<ul>
<li class="chapter" data-level="2.10.1" data-path="2.10-r-documentation.html"><a href="2.10-r-documentation.html#documentation-help-files"><i class="fa fa-check"></i><b>2.10.1</b> Documentation (help) files</a></li>
<li class="chapter" data-level="2.10.2" data-path="2.10-r-documentation.html"><a href="2.10-r-documentation.html#r-vignettes"><i class="fa fa-check"></i><b>2.10.2</b> R vignettes</a></li>
<li class="chapter" data-level="2.10.3" data-path="2.10-r-documentation.html"><a href="2.10-r-documentation.html#official-r-project-resources"><i class="fa fa-check"></i><b>2.10.3</b> Official R Project resources</a></li>
<li class="chapter" data-level="2.10.4" data-path="2.10-r-documentation.html"><a href="2.10-r-documentation.html#unofficial-online-resources"><i class="fa fa-check"></i><b>2.10.4</b> Unofficial online resources</a></li>
<li class="chapter" data-level="2.10.5" data-path="2.10-r-documentation.html"><a href="2.10-r-documentation.html#r-books"><i class="fa fa-check"></i><b>2.10.5</b> R books</a></li>
<li class="chapter" data-level="2.10.6" data-path="2.10-r-documentation.html"><a href="2.10-r-documentation.html#two-reminders"><i class="fa fa-check"></i><b>2.10.6</b> Two reminders</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="3-mod-03.html"><a href="3-mod-03.html"><i class="fa fa-check"></i><b>3</b> Data manipulation with R</a>
<ul>
<li class="chapter" data-level="3.1" data-path="3.1-mod-03-inout.html"><a href="3.1-mod-03-inout.html"><i class="fa fa-check"></i><b>3.1</b> Data import and export</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="3.1-mod-03-inout.html"><a href="3.1-mod-03-inout.html#importing-data-preliminaries"><i class="fa fa-check"></i><b>3.1.1</b> Importing data: preliminaries</a></li>
<li class="chapter" data-level="3.1.2" data-path="3.1-mod-03-inout.html"><a href="3.1-mod-03-inout.html#importing-data-from-text-files-with-read.csv-and-read.table"><i class="fa fa-check"></i><b>3.1.2</b> Importing data from text files with <code>read.csv()</code> and <code>read.table()</code></a></li>
<li class="chapter" data-level="3.1.3" data-path="3.1-mod-03-inout.html"><a href="3.1-mod-03-inout.html#importing-data-from-saved-workspaces"><i class="fa fa-check"></i><b>3.1.3</b> Importing data from saved workspaces</a></li>
<li class="chapter" data-level="3.1.4" data-path="3.1-mod-03-inout.html"><a href="3.1-mod-03-inout.html#importing-data-special-cases"><i class="fa fa-check"></i><b>3.1.4</b> Importing data: special cases:</a></li>
<li class="chapter" data-level="3.1.5" data-path="3.1-mod-03-inout.html"><a href="3.1-mod-03-inout.html#export-data-from-r"><i class="fa fa-check"></i><b>3.1.5</b> Export data from R</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="3.2-making-values-in-r.html"><a href="3.2-making-values-in-r.html"><i class="fa fa-check"></i><b>3.2</b> Making values in R</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="3.2-making-values-in-r.html"><a href="3.2-making-values-in-r.html#producing-arbitrary-values-with-c"><i class="fa fa-check"></i><b>3.2.1</b> Producing arbitrary values with <code>c()</code></a></li>
<li class="chapter" data-level="3.2.2" data-path="3.2-making-values-in-r.html"><a href="3.2-making-values-in-r.html#generating-regular-values"><i class="fa fa-check"></i><b>3.2.2</b> Generating regular values</a></li>
<li class="chapter" data-level="3.2.3" data-path="3.2-making-values-in-r.html"><a href="3.2-making-values-in-r.html#generating-random-values"><i class="fa fa-check"></i><b>3.2.3</b> Generating random values</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="3.3-selecting-data-with.html"><a href="3.3-selecting-data-with.html"><i class="fa fa-check"></i><b>3.3</b> Selecting data with <code>[]</code></a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="3.3-selecting-data-with.html"><a href="3.3-selecting-data-with.html#mod-03-brackets"><i class="fa fa-check"></i><b>3.3.1</b> Basics of brackets</a></li>
<li class="chapter" data-level="3.3.2" data-path="3.3-selecting-data-with.html"><a href="3.3-selecting-data-with.html#extracting-and-selecting-data-with-logical-tests"><i class="fa fa-check"></i><b>3.3.2</b> Extracting and selecting data with logical tests</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="3.4-managing-dates-and-characters.html"><a href="3.4-managing-dates-and-characters.html"><i class="fa fa-check"></i><b>3.4</b> Managing dates and characters</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="3.4-managing-dates-and-characters.html"><a href="3.4-managing-dates-and-characters.html#temporal-data-and-dates"><i class="fa fa-check"></i><b>3.4.1</b> Temporal data and dates</a></li>
<li class="chapter" data-level="3.4.2" data-path="3.4-managing-dates-and-characters.html"><a href="3.4-managing-dates-and-characters.html#character-data-text"><i class="fa fa-check"></i><b>3.4.2</b> Character data (text)</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="3.5-mod-03-dataframe.html"><a href="3.5-mod-03-dataframe.html"><i class="fa fa-check"></i><b>3.5</b> Data frame management</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="3.5-mod-03-dataframe.html"><a href="3.5-mod-03-dataframe.html#data-frame-structure"><i class="fa fa-check"></i><b>3.5.1</b> Data frame structure</a></li>
<li class="chapter" data-level="3.5.2" data-path="3.5-mod-03-dataframe.html"><a href="3.5-mod-03-dataframe.html#common-data-frame-operations"><i class="fa fa-check"></i><b>3.5.2</b> Common data frame operations</a></li>
<li class="chapter" data-level="3.5.3" data-path="3.5-mod-03-dataframe.html"><a href="3.5-mod-03-dataframe.html#other-data-frame-operations"><i class="fa fa-check"></i><b>3.5.3</b> Other data frame operations</a></li>
<li class="chapter" data-level="3.5.4" data-path="3.5-mod-03-dataframe.html"><a href="3.5-mod-03-dataframe.html#mod-03-reshape"><i class="fa fa-check"></i><b>3.5.4</b> Reshaping data frames</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="4-mod-04.html"><a href="4-mod-04.html"><i class="fa fa-check"></i><b>4</b> Exploratory data analysis</a>
<ul>
<li class="chapter" data-level="4.1" data-path="4.1-descriptive-and-summary-statistics.html"><a href="4.1-descriptive-and-summary-statistics.html"><i class="fa fa-check"></i><b>4.1</b> Descriptive and summary statistics</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="4.1-descriptive-and-summary-statistics.html"><a href="4.1-descriptive-and-summary-statistics.html#basic-summary-statistics"><i class="fa fa-check"></i><b>4.1.1</b> Basic summary statistics</a></li>
<li class="chapter" data-level="4.1.2" data-path="4.1-descriptive-and-summary-statistics.html"><a href="4.1-descriptive-and-summary-statistics.html#summarizing-data-with-the-apply-family"><i class="fa fa-check"></i><b>4.1.2</b> Summarizing data with the <code>apply()</code> family</a></li>
<li class="chapter" data-level="4.1.3" data-path="4.1-descriptive-and-summary-statistics.html"><a href="4.1-descriptive-and-summary-statistics.html#mod-04-tabagg"><i class="fa fa-check"></i><b>4.1.3</b> Tabulation and aggregation</a></li>
<li class="chapter" data-level="4.1.4" data-path="4.1-descriptive-and-summary-statistics.html"><a href="4.1-descriptive-and-summary-statistics.html#aggregation-aka-pivot-tables"><i class="fa fa-check"></i><b>4.1.4</b> Aggregation (aka: pivot tables)</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="4.2-mod-04-vis.html"><a href="4.2-mod-04-vis.html"><i class="fa fa-check"></i><b>4.2</b> Visualizing data distributions</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="4.2-mod-04-vis.html"><a href="4.2-mod-04-vis.html#boxplots-aka-box-and-whisker-plots"><i class="fa fa-check"></i><b>4.2.1</b> Boxplots (aka: box-and-whisker plots)</a></li>
<li class="chapter" data-level="4.2.2" data-path="4.2-mod-04-vis.html"><a href="4.2-mod-04-vis.html#histograms"><i class="fa fa-check"></i><b>4.2.2</b> Histograms</a></li>
<li class="chapter" data-level="4.2.3" data-path="4.2-mod-04-vis.html"><a href="4.2-mod-04-vis.html#kernel-density-plots"><i class="fa fa-check"></i><b>4.2.3</b> Kernel density plots</a></li>
<li class="chapter" data-level="4.2.4" data-path="4.2-mod-04-vis.html"><a href="4.2-mod-04-vis.html#empirical-cumulative-distribution-plots-ecdf"><i class="fa fa-check"></i><b>4.2.4</b> Empirical cumulative distribution plots (ECDF)</a></li>
<li class="chapter" data-level="4.2.5" data-path="4.2-mod-04-vis.html"><a href="4.2-mod-04-vis.html#quantile-quantile-qq-plots"><i class="fa fa-check"></i><b>4.2.5</b> Quantile-quantile (QQ) plots</a></li>
<li class="chapter" data-level="4.2.6" data-path="4.2-mod-04-vis.html"><a href="4.2-mod-04-vis.html#how-should-i-plot-my-data"><i class="fa fa-check"></i><b>4.2.6</b> How should I plot my data?</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="4.3-mod-04-dists1.html"><a href="4.3-mod-04-dists1.html"><i class="fa fa-check"></i><b>4.3</b> Statistical distributions</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="4.3-mod-04-dists1.html"><a href="4.3-mod-04-dists1.html#probability-distributions"><i class="fa fa-check"></i><b>4.3.1</b> Probability distributions</a></li>
<li class="chapter" data-level="4.3.2" data-path="4.3-mod-04-dists1.html"><a href="4.3-mod-04-dists1.html#probability-distributions-in-r"><i class="fa fa-check"></i><b>4.3.2</b> Probability distributions in R</a></li>
<li class="chapter" data-level="4.3.3" data-path="4.3-mod-04-dists1.html"><a href="4.3-mod-04-dists1.html#discrete-distributions"><i class="fa fa-check"></i><b>4.3.3</b> Discrete distributions</a></li>
<li class="chapter" data-level="4.3.4" data-path="4.3-mod-04-dists1.html"><a href="4.3-mod-04-dists1.html#continuous-distributions"><i class="fa fa-check"></i><b>4.3.4</b> Continuous distributions</a></li>
<li class="chapter" data-level="4.3.5" data-path="4.3-mod-04-dists1.html"><a href="4.3-mod-04-dists1.html#distributions-summary"><i class="fa fa-check"></i><b>4.3.5</b> Distributions summary</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="4.4-mod-04-dists2.html"><a href="4.4-mod-04-dists2.html"><i class="fa fa-check"></i><b>4.4</b> Fitting and testing distributions</a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="4.4-mod-04-dists2.html"><a href="4.4-mod-04-dists2.html#estimating-distributional-parameters"><i class="fa fa-check"></i><b>4.4.1</b> Estimating distributional parameters</a></li>
<li class="chapter" data-level="4.4.2" data-path="4.4-mod-04-dists2.html"><a href="4.4-mod-04-dists2.html#graphical-methods-for-examining-distributions"><i class="fa fa-check"></i><b>4.4.2</b> Graphical methods for examining distributions</a></li>
<li class="chapter" data-level="4.4.3" data-path="4.4-mod-04-dists2.html"><a href="4.4-mod-04-dists2.html#formal-tests-for-distributions"><i class="fa fa-check"></i><b>4.4.3</b> Formal tests for distributions</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="4.5-mod-04-trans.html"><a href="4.5-mod-04-trans.html"><i class="fa fa-check"></i><b>4.5</b> Data transformations</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="4.5-mod-04-trans.html"><a href="4.5-mod-04-trans.html#why-transform"><i class="fa fa-check"></i><b>4.5.1</b> Why transform?</a></li>
<li class="chapter" data-level="4.5.2" data-path="4.5-mod-04-trans.html"><a href="4.5-mod-04-trans.html#transforms-vs.-link-functions"><i class="fa fa-check"></i><b>4.5.2</b> Transforms vs. link functions</a></li>
<li class="chapter" data-level="4.5.3" data-path="4.5-mod-04-trans.html"><a href="4.5-mod-04-trans.html#log-transformation"><i class="fa fa-check"></i><b>4.5.3</b> Log transformation</a></li>
<li class="chapter" data-level="4.5.4" data-path="4.5-mod-04-trans.html"><a href="4.5-mod-04-trans.html#rank-transformation"><i class="fa fa-check"></i><b>4.5.4</b> Rank transformation</a></li>
<li class="chapter" data-level="4.5.5" data-path="4.5-mod-04-trans.html"><a href="4.5-mod-04-trans.html#other-transforms-less-common"><i class="fa fa-check"></i><b>4.5.5</b> Other transforms (less common)</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="4.6-multivariate-data-exploration.html"><a href="4.6-multivariate-data-exploration.html"><i class="fa fa-check"></i><b>4.6</b> Multivariate data exploration</a>
<ul>
<li class="chapter" data-level="4.6.1" data-path="4.6-multivariate-data-exploration.html"><a href="4.6-multivariate-data-exploration.html#scatterplots-for-two-variables"><i class="fa fa-check"></i><b>4.6.1</b> Scatterplots for two variables</a></li>
<li class="chapter" data-level="4.6.2" data-path="4.6-multivariate-data-exploration.html"><a href="4.6-multivariate-data-exploration.html#mod-04-smat"><i class="fa fa-check"></i><b>4.6.2</b> Scatterplot matrices for many variables</a></li>
<li class="chapter" data-level="4.6.3" data-path="4.6-multivariate-data-exploration.html"><a href="4.6-multivariate-data-exploration.html#lattice-plots-for-hierarchical-data"><i class="fa fa-check"></i><b>4.6.3</b> Lattice plots for hierarchical data</a></li>
</ul></li>
<li class="chapter" data-level="4.7" data-path="4.7-ordination-brief-introduction.html"><a href="4.7-ordination-brief-introduction.html"><i class="fa fa-check"></i><b>4.7</b> Ordination (brief introduction)</a>
<ul>
<li class="chapter" data-level="4.7.1" data-path="4.7-ordination-brief-introduction.html"><a href="4.7-ordination-brief-introduction.html#principal-components-analysis-pca"><i class="fa fa-check"></i><b>4.7.1</b> Principal components analysis (PCA)</a></li>
<li class="chapter" data-level="4.7.2" data-path="4.7-ordination-brief-introduction.html"><a href="4.7-ordination-brief-introduction.html#nonmetric-multidimensional-scaling-nmds"><i class="fa fa-check"></i><b>4.7.2</b> Nonmetric multidimensional scaling (NMDS)</a></li>
<li class="chapter" data-level="4.7.3" data-path="4.7-ordination-brief-introduction.html"><a href="4.7-ordination-brief-introduction.html#plotting-ordinations"><i class="fa fa-check"></i><b>4.7.3</b> Plotting ordinations</a></li>
<li class="chapter" data-level="4.7.4" data-path="4.7-ordination-brief-introduction.html"><a href="4.7-ordination-brief-introduction.html#ordination-wrap-up-for-now"><i class="fa fa-check"></i><b>4.7.4</b> Ordination wrap-up (for now)</a></li>
</ul></li>
<li class="chapter" data-level="4.8" data-path="4.8-mod-04-prob.html"><a href="4.8-mod-04-prob.html"><i class="fa fa-check"></i><b>4.8</b> Common statistical problems</a>
<ul>
<li class="chapter" data-level="4.8.1" data-path="4.8-mod-04-prob.html"><a href="4.8-mod-04-prob.html#outliers-and-erroneous-values"><i class="fa fa-check"></i><b>4.8.1</b> Outliers and erroneous values</a></li>
<li class="chapter" data-level="4.8.2" data-path="4.8-mod-04-prob.html"><a href="4.8-mod-04-prob.html#autocorrelation"><i class="fa fa-check"></i><b>4.8.2</b> Autocorrelation</a></li>
<li class="chapter" data-level="4.8.3" data-path="4.8-mod-04-prob.html"><a href="4.8-mod-04-prob.html#mod-04-multicol"><i class="fa fa-check"></i><b>4.8.3</b> Collinearity</a></li>
<li class="chapter" data-level="4.8.4" data-path="4.8-mod-04-prob.html"><a href="4.8-mod-04-prob.html#missing-data"><i class="fa fa-check"></i><b>4.8.4</b> Missing data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="5-mod-05.html"><a href="5-mod-05.html"><i class="fa fa-check"></i><b>5</b> Generalized linear models (GLM)</a>
<ul>
<li class="chapter" data-level="5.1" data-path="5.1-mod-05-lm.html"><a href="5.1-mod-05-lm.html"><i class="fa fa-check"></i><b>5.1</b> Prelude with linear models</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="5.1-mod-05-lm.html"><a href="5.1-mod-05-lm.html#assumptions-of-linear-models"><i class="fa fa-check"></i><b>5.1.1</b> Assumptions of linear models</a></li>
<li class="chapter" data-level="5.1.2" data-path="5.1-mod-05-lm.html"><a href="5.1-mod-05-lm.html#linear-regression-in-r"><i class="fa fa-check"></i><b>5.1.2</b> Linear regression in R</a></li>
<li class="chapter" data-level="5.1.3" data-path="5.1-mod-05-lm.html"><a href="5.1-mod-05-lm.html#multiple-linear-regression"><i class="fa fa-check"></i><b>5.1.3</b> Multiple linear regression</a></li>
<li class="chapter" data-level="5.1.4" data-path="5.1-mod-05-lm.html"><a href="5.1-mod-05-lm.html#mod-05-anova"><i class="fa fa-check"></i><b>5.1.4</b> ANOVA and ANCOVA with <code>lm()</code></a></li>
<li class="chapter" data-level="5.1.5" data-path="5.1-mod-05-lm.html"><a href="5.1-mod-05-lm.html#variations-on-linear-models"><i class="fa fa-check"></i><b>5.1.5</b> Variations on linear models</a></li>
<li class="chapter" data-level="5.1.6" data-path="5.1-mod-05-lm.html"><a href="5.1-mod-05-lm.html#example-linear-regression-workflow"><i class="fa fa-check"></i><b>5.1.6</b> Example linear regression workflow</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="5.2-mod-05-basic.html"><a href="5.2-mod-05-basic.html"><i class="fa fa-check"></i><b>5.2</b> GLM basics</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="5.2-mod-05-basic.html"><a href="5.2-mod-05-basic.html#example-glms"><i class="fa fa-check"></i><b>5.2.1</b> Example GLMS</a></li>
<li class="chapter" data-level="5.2.2" data-path="5.2-mod-05-basic.html"><a href="5.2-mod-05-basic.html#glm-families"><i class="fa fa-check"></i><b>5.2.2</b> GLM families</a></li>
<li class="chapter" data-level="5.2.3" data-path="5.2-mod-05-basic.html"><a href="5.2-mod-05-basic.html#glm-link-functions"><i class="fa fa-check"></i><b>5.2.3</b> GLM link functions</a></li>
<li class="chapter" data-level="5.2.4" data-path="5.2-mod-05-basic.html"><a href="5.2-mod-05-basic.html#deviance-and-other-glm-diagnostics"><i class="fa fa-check"></i><b>5.2.4</b> Deviance and other GLM diagnostics</a></li>
<li class="chapter" data-level="5.2.5" data-path="5.2-mod-05-basic.html"><a href="5.2-mod-05-basic.html#to-pseudo-r2-or-not-to-pseudo-r2"><i class="fa fa-check"></i><b>5.2.5</b> To pseudo-<em>R</em><sup>2</sup> or not to pseudo-<em>R</em><sup>2</sup>?</a></li>
<li class="chapter" data-level="5.2.6" data-path="5.2-mod-05-basic.html"><a href="5.2-mod-05-basic.html#common-glms"><i class="fa fa-check"></i><b>5.2.6</b> Common GLMs</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="5.3-mod-05-loglin.html"><a href="5.3-mod-05-loglin.html"><i class="fa fa-check"></i><b>5.3</b> Log-linear models</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="5.3-mod-05-loglin.html"><a href="5.3-mod-05-loglin.html#mod-05-loglin-examp"><i class="fa fa-check"></i><b>5.3.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="5.3.2" data-path="5.3-mod-05-loglin.html"><a href="5.3-mod-05-loglin.html#example-with-real-data"><i class="fa fa-check"></i><b>5.3.2</b> Example with real data</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="5.4-mod-05-poisson.html"><a href="5.4-mod-05-poisson.html"><i class="fa fa-check"></i><b>5.4</b> Poisson GLM for counts</a>
<ul>
<li class="chapter" data-level="5.4.1" data-path="5.4-mod-05-poisson.html"><a href="5.4-mod-05-poisson.html#example-with-simulated-data"><i class="fa fa-check"></i><b>5.4.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="5.4.2" data-path="5.4-mod-05-poisson.html"><a href="5.4-mod-05-poisson.html#example-with-real-data-1"><i class="fa fa-check"></i><b>5.4.2</b> Example with real data</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="5.5-mod-05-quasi.html"><a href="5.5-mod-05-quasi.html"><i class="fa fa-check"></i><b>5.5</b> Quasi-Poisson and negative binomial GLM</a>
<ul>
<li class="chapter" data-level="5.5.1" data-path="5.5-mod-05-quasi.html"><a href="5.5-mod-05-quasi.html#example-with-simulated-data-1"><i class="fa fa-check"></i><b>5.5.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="5.5.2" data-path="5.5-mod-05-quasi.html"><a href="5.5-mod-05-quasi.html#example-with-real-data-2"><i class="fa fa-check"></i><b>5.5.2</b> Example with real data</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="5.6-mod-05-logistic.html"><a href="5.6-mod-05-logistic.html"><i class="fa fa-check"></i><b>5.6</b> Logistic regression for binary outcomes</a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="5.6-mod-05-logistic.html"><a href="5.6-mod-05-logistic.html#example-with-simulated-data-2"><i class="fa fa-check"></i><b>5.6.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="5.6.2" data-path="5.6-mod-05-logistic.html"><a href="5.6-mod-05-logistic.html#example-with-real-data-3"><i class="fa fa-check"></i><b>5.6.2</b> Example with real data</a></li>
<li class="chapter" data-level="5.6.3" data-path="5.6-mod-05-logistic.html"><a href="5.6-mod-05-logistic.html#mod-05-auc"><i class="fa fa-check"></i><b>5.6.3</b> Logistic GLM diagnostics: AUC and ROC</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="5.7-mod-05-binom.html"><a href="5.7-mod-05-binom.html"><i class="fa fa-check"></i><b>5.7</b> Binomial GLM for proportional data</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="5.7-mod-05-binom.html"><a href="5.7-mod-05-binom.html#binomial-glm"><i class="fa fa-check"></i><b>5.7.1</b> Binomial GLM</a></li>
<li class="chapter" data-level="5.7.2" data-path="5.7-mod-05-binom.html"><a href="5.7-mod-05-binom.html#example-with-simulated-data-3"><i class="fa fa-check"></i><b>5.7.2</b> Example with simulated data</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="5.8-mod-05-gamma.html"><a href="5.8-mod-05-gamma.html"><i class="fa fa-check"></i><b>5.8</b> Gamma models for overdispersed data</a>
<ul>
<li class="chapter" data-level="5.8.1" data-path="5.8-mod-05-gamma.html"><a href="5.8-mod-05-gamma.html#example-with-simulated-data-4"><i class="fa fa-check"></i><b>5.8.1</b> Example with simulated data</a></li>
</ul></li>
<li class="chapter" data-level="5.9" data-path="5.9-mod-05-beyond.html"><a href="5.9-mod-05-beyond.html"><i class="fa fa-check"></i><b>5.9</b> Beyond GLM: Overview of GAM and GEE</a>
<ul>
<li class="chapter" data-level="5.9.1" data-path="5.9-mod-05-beyond.html"><a href="5.9-mod-05-beyond.html#mod-05-gam"><i class="fa fa-check"></i><b>5.9.1</b> Generalized additive models (GAM)</a></li>
<li class="chapter" data-level="5.9.2" data-path="5.9-mod-05-beyond.html"><a href="5.9-mod-05-beyond.html#mod-05-gee"><i class="fa fa-check"></i><b>5.9.2</b> Generalized estimating equations (GEE)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="6-mod-06.html"><a href="6-mod-06.html"><i class="fa fa-check"></i><b>6</b> Nonlinear models</a>
<ul>
<li class="chapter" data-level="6.1" data-path="6.1-background.html"><a href="6.1-background.html"><i class="fa fa-check"></i><b>6.1</b> Background</a></li>
<li class="chapter" data-level="6.2" data-path="6.2-mod-06-intro.html"><a href="6.2-mod-06-intro.html"><i class="fa fa-check"></i><b>6.2</b> Nonlinear least squares (NLS)</a></li>
<li class="chapter" data-level="6.3" data-path="6.3-mod-06-micmen.html"><a href="6.3-mod-06-micmen.html"><i class="fa fa-check"></i><b>6.3</b> Michaelis-Menten curves</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="6.3-mod-06-micmen.html"><a href="6.3-mod-06-micmen.html#mod-06-micmen-sim"><i class="fa fa-check"></i><b>6.3.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="6.3.2" data-path="6.3-mod-06-micmen.html"><a href="6.3-mod-06-micmen.html#example-with-real-data-4"><i class="fa fa-check"></i><b>6.3.2</b> Example with real data</a></li>
<li class="chapter" data-level="6.3.3" data-path="6.3-mod-06-micmen.html"><a href="6.3-mod-06-micmen.html#alternative-strategies-for-the-analysis"><i class="fa fa-check"></i><b>6.3.3</b> Alternative strategies for the analysis</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="6.4-mod-06-grow.html"><a href="6.4-mod-06-grow.html"><i class="fa fa-check"></i><b>6.4</b> Biological growth curves</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="6.4-mod-06-grow.html"><a href="6.4-mod-06-grow.html#gompertz-and-von-bertalanffy-curves"><i class="fa fa-check"></i><b>6.4.1</b> Gompertz and von Bertalanffy curves</a></li>
<li class="chapter" data-level="6.4.2" data-path="6.4-mod-06-grow.html"><a href="6.4-mod-06-grow.html#example-with-real-data-5"><i class="fa fa-check"></i><b>6.4.2</b> Example with real data</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="6.5-dose-response-curves.html"><a href="6.5-dose-response-curves.html"><i class="fa fa-check"></i><b>6.5</b> Dose response curves</a></li>
<li class="chapter" data-level="6.6" data-path="6.6-alternatives-to-nls.html"><a href="6.6-alternatives-to-nls.html"><i class="fa fa-check"></i><b>6.6</b> Alternatives to NLS</a>
<ul>
<li class="chapter" data-level="6.6.1" data-path="6.6-alternatives-to-nls.html"><a href="6.6-alternatives-to-nls.html#generalized-nonlinear-models"><i class="fa fa-check"></i><b>6.6.1</b> Generalized nonlinear models</a></li>
<li class="chapter" data-level="6.6.2" data-path="6.6-alternatives-to-nls.html"><a href="6.6-alternatives-to-nls.html#quantile-regression"><i class="fa fa-check"></i><b>6.6.2</b> Quantile regression</a></li>
<li class="chapter" data-level="6.6.3" data-path="6.6-alternatives-to-nls.html"><a href="6.6-alternatives-to-nls.html#mod-06-gam"><i class="fa fa-check"></i><b>6.6.3</b> Generalized additive models (GAM)</a></li>
<li class="chapter" data-level="6.6.4" data-path="6.6-alternatives-to-nls.html"><a href="6.6-alternatives-to-nls.html#classification-and-regression-trees-cart"><i class="fa fa-check"></i><b>6.6.4</b> Classification and regression trees (CART)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="7-mod-07.html"><a href="7-mod-07.html"><i class="fa fa-check"></i><b>7</b> Mixed models</a>
<ul>
<li class="chapter" data-level="7.1" data-path="7.1-prelude-glm.html"><a href="7.1-prelude-glm.html"><i class="fa fa-check"></i><b>7.1</b> Prelude (GLM)</a></li>
<li class="chapter" data-level="7.2" data-path="7.2-mod-07-lmm.html"><a href="7.2-mod-07-lmm.html"><i class="fa fa-check"></i><b>7.2</b> Linear mixed models (LMM)</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="7.2-mod-07-lmm.html"><a href="7.2-mod-07-lmm.html#formal-definition-and-example"><i class="fa fa-check"></i><b>7.2.1</b> Formal definition and example</a></li>
<li class="chapter" data-level="7.2.2" data-path="7.2-mod-07-lmm.html"><a href="7.2-mod-07-lmm.html#example-with-simulated-data-5"><i class="fa fa-check"></i><b>7.2.2</b> Example with simulated data</a></li>
<li class="chapter" data-level="7.2.3" data-path="7.2-mod-07-lmm.html"><a href="7.2-mod-07-lmm.html#p-values-in-lmm"><i class="fa fa-check"></i><b>7.2.3</b> <em>P</em>-values in LMM</a></li>
<li class="chapter" data-level="7.2.4" data-path="7.2-mod-07-lmm.html"><a href="7.2-mod-07-lmm.html#specifying-random-effects"><i class="fa fa-check"></i><b>7.2.4</b> Specifying random effects</a></li>
<li class="chapter" data-level="7.2.5" data-path="7.2-mod-07-lmm.html"><a href="7.2-mod-07-lmm.html#lmm-example-with-real-data"><i class="fa fa-check"></i><b>7.2.5</b> LMM example with real data</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="7.3-mod-07-glmm.html"><a href="7.3-mod-07-glmm.html"><i class="fa fa-check"></i><b>7.3</b> Generalized linear mixed models (GLMM)</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="7.3-mod-07-glmm.html"><a href="7.3-mod-07-glmm.html#definition-1"><i class="fa fa-check"></i><b>7.3.1</b> Definition</a></li>
<li class="chapter" data-level="7.3.2" data-path="7.3-mod-07-glmm.html"><a href="7.3-mod-07-glmm.html#glmm-on-simulated-data"><i class="fa fa-check"></i><b>7.3.2</b> GLMM on simulated data</a></li>
<li class="chapter" data-level="7.3.3" data-path="7.3-mod-07-glmm.html"><a href="7.3-mod-07-glmm.html#glmm-on-real-data"><i class="fa fa-check"></i><b>7.3.3</b> GLMM on real data</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="7.4-mod-07-nlme.html"><a href="7.4-mod-07-nlme.html"><i class="fa fa-check"></i><b>7.4</b> Nonlinear mixed models (NLME)</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="7.4-mod-07-nlme.html"><a href="7.4-mod-07-nlme.html#definition-and-background"><i class="fa fa-check"></i><b>7.4.1</b> Definition and background</a></li>
<li class="chapter" data-level="7.4.2" data-path="7.4-mod-07-nlme.html"><a href="7.4-mod-07-nlme.html#nlme-on-simulated-data"><i class="fa fa-check"></i><b>7.4.2</b> NLME on simulated data</a></li>
<li class="chapter" data-level="7.4.3" data-path="7.4-mod-07-nlme.html"><a href="7.4-mod-07-nlme.html#nlme-on-real-data"><i class="fa fa-check"></i><b>7.4.3</b> NLME on real data</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="7.5-mod-07-gamm.html"><a href="7.5-mod-07-gamm.html"><i class="fa fa-check"></i><b>7.5</b> (Generalized) additive mixed models (AMM/GAMM)</a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="7.5-mod-07-gamm.html"><a href="7.5-mod-07-gamm.html#example-gamm-with-real-data"><i class="fa fa-check"></i><b>7.5.1</b> Example GAMM with real data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="8-mod-08.html"><a href="8-mod-08.html"><i class="fa fa-check"></i><b>8</b> Multivariate data analysis</a>
<ul>
<li class="chapter" data-level="8.1" data-path="8.1-multivariate-data.html"><a href="8.1-multivariate-data.html"><i class="fa fa-check"></i><b>8.1</b> Multivariate data</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="8.1-multivariate-data.html"><a href="8.1-multivariate-data.html#univariate-vs.-multivariate-data"><i class="fa fa-check"></i><b>8.1.1</b> Univariate vs. multivariate data</a></li>
<li class="chapter" data-level="8.1.2" data-path="8.1-multivariate-data.html"><a href="8.1-multivariate-data.html#components-of-multivariate-data"><i class="fa fa-check"></i><b>8.1.2</b> Components of multivariate data</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="8.2-mod-08-dist.html"><a href="8.2-mod-08-dist.html"><i class="fa fa-check"></i><b>8.2</b> Distance metrics: biological (dis)similarity</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="8.2-mod-08-dist.html"><a href="8.2-mod-08-dist.html#euclidean-distance"><i class="fa fa-check"></i><b>8.2.1</b> Euclidean distance</a></li>
<li class="chapter" data-level="8.2.2" data-path="8.2-mod-08-dist.html"><a href="8.2-mod-08-dist.html#bray-curtis-and-other-distance-metrics"><i class="fa fa-check"></i><b>8.2.2</b> Bray-Curtis and other distance metrics</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="8.3-clustering.html"><a href="8.3-clustering.html"><i class="fa fa-check"></i><b>8.3</b> Clustering</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="8.3-clustering.html"><a href="8.3-clustering.html#k-means-clustering"><i class="fa fa-check"></i><b>8.3.1</b> <em>K</em>-means clustering</a></li>
<li class="chapter" data-level="8.3.2" data-path="8.3-clustering.html"><a href="8.3-clustering.html#hierarchical-agglomerative-clustering"><i class="fa fa-check"></i><b>8.3.2</b> Hierarchical agglomerative clustering</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="8.4-mod-08-sims.html"><a href="8.4-mod-08-sims.html"><i class="fa fa-check"></i><b>8.4</b> Analyzing dissimilarity</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="8.4-mod-08-sims.html"><a href="8.4-mod-08-sims.html#mantel-tests-distance-vs.-distance"><i class="fa fa-check"></i><b>8.4.1</b> Mantel tests: distance vs. distance</a></li>
<li class="chapter" data-level="8.4.2" data-path="8.4-mod-08-sims.html"><a href="8.4-mod-08-sims.html#comparing-dissimilarity-between-groups"><i class="fa fa-check"></i><b>8.4.2</b> Comparing dissimilarity between groups</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="8.5-mod-08-ord.html"><a href="8.5-mod-08-ord.html"><i class="fa fa-check"></i><b>8.5</b> Ordination</a>
<ul>
<li class="chapter" data-level="8.5.1" data-path="8.5-mod-08-ord.html"><a href="8.5-mod-08-ord.html#principal-components-analysis-pca-1"><i class="fa fa-check"></i><b>8.5.1</b> Principal components analysis (PCA)</a></li>
<li class="chapter" data-level="8.5.2" data-path="8.5-mod-08-ord.html"><a href="8.5-mod-08-ord.html#nmds-and-other-ordination-methods"><i class="fa fa-check"></i><b>8.5.2</b> NMDS and other ordination methods</a></li>
<li class="chapter" data-level="8.5.3" data-path="8.5-mod-08-ord.html"><a href="8.5-mod-08-ord.html#other-ordination-techniques"><i class="fa fa-check"></i><b>8.5.3</b> Other ordination techniques</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="9-mod-09.html"><a href="9-mod-09.html"><i class="fa fa-check"></i><b>9</b> Planning your analysis (what test?)</a>
<ul>
<li class="chapter" data-level="9.1" data-path="9.1-how-to-use-this-guide.html"><a href="9.1-how-to-use-this-guide.html"><i class="fa fa-check"></i><b>9.1</b> How to use this guide</a></li>
<li class="chapter" data-level="9.2" data-path="9.2-what-question-are-you-trying-to-answer.html"><a href="9.2-what-question-are-you-trying-to-answer.html"><i class="fa fa-check"></i><b>9.2</b> What question are you trying to answer?</a></li>
<li class="chapter" data-level="9.3" data-path="9.3-testing-for-a-difference-in-mean-or-location.html"><a href="9.3-testing-for-a-difference-in-mean-or-location.html"><i class="fa fa-check"></i><b>9.3</b> Testing for a difference in mean or location</a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="9.3-testing-for-a-difference-in-mean-or-location.html"><a href="9.3-testing-for-a-difference-in-mean-or-location.html#additional-considerations"><i class="fa fa-check"></i><b>9.3.1</b> Additional considerations</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="9.4-testing-for-a-continuous-relationship-between-two-or-more-variables.html"><a href="9.4-testing-for-a-continuous-relationship-between-two-or-more-variables.html"><i class="fa fa-check"></i><b>9.4</b> Testing for a continuous relationship between two or more variables</a>
<ul>
<li class="chapter" data-level="9.4.1" data-path="9.4-testing-for-a-continuous-relationship-between-two-or-more-variables.html"><a href="9.4-testing-for-a-continuous-relationship-between-two-or-more-variables.html#additional-considerations-1"><i class="fa fa-check"></i><b>9.4.1</b> Additional considerations</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="9.5-classifying-observations.html"><a href="9.5-classifying-observations.html"><i class="fa fa-check"></i><b>9.5</b> Classifying observations</a></li>
<li class="chapter" data-level="9.6" data-path="9.6-conclusions.html"><a href="9.6-conclusions.html"><i class="fa fa-check"></i><b>9.6</b> Conclusions</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="literature-cited.html"><a href="literature-cited.html"><i class="fa fa-check"></i>Literature Cited</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Applied Biological Data Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="literature-cited" class="section level1 unnumbered">
<h1>Literature Cited</h1>

<div id="refs" class="references csl-bib-body hanging-indent" line-spacing="2">
<div class="csl-entry">
Alberti, E., G. Gioia, G. Sironi, S. Zanzani, P. Riccaboni, M. Magrini, and M. Manfredi. 2011. <a href="https://doi.org/10.1017/S0022149X10000647"><em>Elaphostrongylus cervi</em> in a population of red deer (<em><span>C</span>ervus elaphus</em>) and evidence of cerebrospinal nematodiasis in small ruminants in the province of <span>V</span>arese, <span>I</span>taly</a>. Journal of <span>H</span>elminthology 85:313–318.
</div>
<div class="csl-entry">
Anderson, M. J. 2001. <a href="https://doi.org/10.1111/j.1442-9993.2001.01070.pp.x">A new method for non-parametric multivariate analysis of variance</a>. Austral Ecology 26:32–46.
</div>
<div class="csl-entry">
Bodin, N., H. Pethybridge, L. M. Duffy, A. Lorrain, V. Allain, J. M. Logan, F. Ménard, B. Graham, C. A. Choy, C. J. Somes, R. J. Olson, and J. W. Young. 2021. <a href="https://doi.org/10.1002/ecy.3265">Global data set for nitrogen and carbon stable isotopes of tunas</a>. Ecology 102:e03265.
</div>
<div class="csl-entry">
Bolker, B. M. 2008. Ecological models and data in <span>R</span>. Princeton University Press, Princeton, New Jersey, USA.
</div>
<div class="csl-entry">
Bolker, B. M., M. E. Brooks, C. J. Clark, S. W. Geange, J. R. Poulsen, M. H. H. Stevens, and J.-S. S. White. 2009. <a href="https://doi.org/10.1016/j.tree.2008.10.008">Generalized linear mixed models: A practical guide for ecology and evolution</a>. Trends in Ecology &amp; Evolution 24:127–135.
</div>
<div class="csl-entry">
Borcard, D., F. Gillet, and P. Legendre. 2018. Numerical ecology with <span>R</span>. Springer, New York.
</div>
<div class="csl-entry">
Box, G. E. P., and D. R. Cox. 1964. <a href="https://doi.org/10.1111/j.2517-6161.1964.tb00553.x">An analysis of transformations</a>. Journal of the Royal Statistical Society: Series B (Methodological) 26:211–243.
</div>
<div class="csl-entry">
Brieuc, M. S. O., C. D. Waters, D. P. Drinan, and K. A. Naish. 2018. <a href="https://doi.org/10.1111/1755-0998.12773">A practical introduction to random forest for genetic association studies in ecology and evolution</a>. Molecular Ecology Resources 18:755–766.
</div>
<div class="csl-entry">
Broman, K. W., and K. H. Woo. 2018. <a href="https://doi.org/10.1080/00031305.2017.1375989">Data organization in spreadsheets</a>. The American Statistician 72:2–10.
</div>
<div class="csl-entry">
Bubrig, L. T., J. M. Sutton, and J. L. Fierst. 2020. <a href="https://doi.org/10.1002/ece3.6646"><em><span>C</span>aenorhabditis elegans</em> dauers vary recovery in response to bacteria from natural habitat</a>. Ecology and Evolution 10:9886–9895.
</div>
<div class="csl-entry">
Burnham, K. P., and D. R. Anderson. 2004. Multimodel inference: Understanding <span>AIC</span> and <span>BIC</span> in model selection. Sociological methods &amp; research 33:261–304.
</div>
<div class="csl-entry">
Burnham, K., and D. Anderson. 2002. Model selection and multi-model inference: <span class="nocase">a</span> practical information-theoretic approach. Springer-<span>V</span>erlag, New York.
</div>
<div class="csl-entry">
Button, K. S., J. P. Ioannidis, C. Mokrysz, B. A. Nosek, J. Flint, E. S. Robinson, and M. R. Munafò. 2013. <a href="https://doi.org/10.1038/nrn3475">Power failure: <span class="nocase">w</span>hy small sample size undermines the reliability of neuroscience</a>. Nature <span>R</span>eviews <span>N</span>euroscience 14:365–376.
</div>
<div class="csl-entry">
Cade, B. S., and B. R. Noon. 2003. <a href="https://doi.org/10.1890/1540-9295(2003)001[0412:agitqr]2.0.co;2">A gentle introduction to quantile regression for ecologists</a>. Frontiers in Ecology and the Environment 1:412–420.
</div>
<div class="csl-entry">
Carey, V. J. 2019. <a href="https://CRAN.R-project.org/package=gee"><span class="nocase">g</span>ee: Generalized estimation equation solver</a>.
</div>
<div class="csl-entry">
Carnell, R. 2019. <a href="https://CRAN.R-project.org/package=triangle">Triangle: Provides the standard distribution functions for the triangle distribution</a>.
</div>
<div class="csl-entry">
Chivers, D. J., and C. M. Hladik. 1980. <a href="https://doi.org/10.1002/jmor.1051660306">Morphology of the gastrointestinal tract in primates: Comparisons with other mammals in relation to diet</a>. Journal of Morphology 166:337–386.
</div>
<div class="csl-entry">
Clarke, K. R. 1993. <a href="https://doi.org/10.1111/j.1442-9993.1993.tb00438.x">Non-parametric multivariate analyses of changes in community structure</a>. Australian journal of ecology 18:117–143.
</div>
<div class="csl-entry">
Clarke, K. R., P. J. Somerfield, and M. G. Chapman. 2006. <a href="https://doi.org/10.1016/j.jembe.2005.12.017">On resemblance measures for ecological studies, including taxonomic dissimilarities and a zero-adjusted bray-curtis coefficient for denuded assemblages</a>. Journal of Experimental Marine Biology and Ecology 330:55–80.
</div>
<div class="csl-entry">
Cutler, D. R., T. C. Edwards, K. H. Beard, A. Cutler, K. T. Hess, J. Gibson, and J. J. Lawler. 2007. <a href="https://doi.org/10.1890/07-0539.1">Random forests for classification in ecology</a>. Ecology 88:2783–2792.
</div>
<div class="csl-entry">
Dalgaard, P. 2002. Introductory statistics with <span>R</span>. Springer Science+Business Media.
</div>
<div class="csl-entry">
De’ath, G. 2007. <a href="https://doi.org/10.1890/0012-9658(2007)88[243:btfema]2.0.co;2">Boosted trees for ecological modeling and prediction</a>. Ecology 88:243–251.
</div>
<div class="csl-entry">
De’ath, G., and K. E. Fabricius. 2000. <a href="https://doi.org/10.1890/0012-9658(2000)081[3178:CARTAP]2.0.CO;2">Classification and regression trees: A powerful yet simple technique for ecological data analysis</a>. Ecology 81:3178–3192.
</div>
<div class="csl-entry">
De Solla, S. R., P. A. Martin, K. J. Fernie, B. J. Park, and G. Mayne. 2006. <a href="https://doi.org/10.1897/05-165R.1">Effects of environmentally relevant concentrations of atrazine on gonadal development of snapping turtles (chelydra serpentina)</a>. Environmental Toxicology and Chemistry: An International Journal 25:520–526.
</div>
<div class="csl-entry">
Dormann, C. F., J. Elith, S. Bacher, C. Buchmann, G. Carl, G. Carré, J. R. G. Marquéz, B. Gruber, B. Lafourcade, P. J. Leitão, T. Münkemüller, C. McClean, P. E. Osborne, B. Reineking, B. Schröder, A. K. Skidmore, D. Zurell, and S. Lautenbach. 2013. <a href="https://doi.org/10.1111/j.1600-0587.2012.07348.x">Collinearity: A review of methods to deal with it and a simulation study evaluating their performance</a>. Ecography 36:27–46.
</div>
<div class="csl-entry">
Elith, J., and J. Leathwick. 2017. Boosted regression trees for ecological modeling. R Documentation. Available online: https://cran.r-project.org/web/packages/dismo/vignettes/brt.pdf (accessed on 2021-12-23).
</div>
<div class="csl-entry">
Elith, J., J. R. Leathwick, and T. Hastie. 2008. <a href="https://doi.org/10.1111/j.1365-2656.2008.01390.x">A working guide to boosted regression trees</a>. Journal of Animal Ecology 77:802–813.
</div>
<div class="csl-entry">
Faurby, S., M. Davis, R. Ø. Pedersen, S. D. Schowanek, A. Antonelli, and J.-C. Svenning. 2018. <a href="https://doi.org/10.1002/ecy.2443"><span>PHYLACINE</span> 1.2: The phylogenetic atlas of mammal macroecology</a>. Ecology 99:2626.
</div>
<div class="csl-entry">
Foucault, Q., A. Wieser, A.-M. Waldvogel, B. Feldmeyer, and M. Pfenninger. 2018. <a href="https://doi.org/10.1002/ece3.4706">Rapid adaptation to high temperatures in <em><span>C</span>hironomus riparius</em></a>. Ecology and Evolution 8:12780–12789.
</div>
<div class="csl-entry">
Gelfand, A. E., and P. Vounatsou. 2003. <a href="https://doi.org/10.1093/biostatistics/4.1.11">Proper multivariate conditional autoregressive models for spatial data analysis</a>. Biostatistics 4:11–15.
</div>
<div class="csl-entry">
Gould, S. J. 1974. <a href="https://doi.org/10.2307/2407322">The origin and function of ’bizarre’ structures: Antler size and skull size in the ’irish elk,’ <em><span>M</span>egaloceros giganteus</em></a>. Evolution 28:191–220.
</div>
<div class="csl-entry">
Green, N. S., M. L. Wildhaber, and J. L. Albers. 2019. <a href="https://doi.org/10.1111/jai.13701">Potential responses of the <span>L</span>ower <span>M</span>issouri <span>R</span>iver shovelnose sturgeon (<em><span>S</span>caphirhynchus platorynchus</em>) population to a commercial fishing ban</a>. Journal of Applied Ichthyology 35:370–377.
</div>
<div class="csl-entry">
Green, N. S., M. L. Wildhaber, and J. L. Albers. 2021. Effectiveness of a distance sampling from roads program for white-tailed deer in the <span>N</span>ational <span>C</span>apital <span>R</span>egion parks. Natural Resource Report, National Park Service, Fort Collins, Colorado.
</div>
<div class="csl-entry">
Green, N. S., M. L. Wildhaber, J. L. Albers, T. W. Pettit, and M. J. Hooper. 2020. <a href="https://doi.org/10.1002/ieam.4324">Efficient mammal biodiversity surveys for ecological restoration monitoring</a>. Integrated Environmental Assessment and Management.
</div>
<div class="csl-entry">
Grolemund, G., and H. Wickham. 2011. <a href="https://www.jstatsoft.org/v40/i03/">Dates and times made easy with <span class="nocase">lubridate</span></a>. Journal of Statistical Software 40:1–25.
</div>
<div class="csl-entry">
Head, M. L., L. Holman, R. Lanfear, A. T. Kahn, and M. D. Jennions. 2015. <a href="https://doi.org/10.1371/journal.pbio.1002106">The extent and consequences of p-hacking in science</a>. <span>PLoS</span> <span>B</span>iology 13:e1002106.
</div>
<div class="csl-entry">
Hill, M. O. 1979. A <span>FORTRAN</span> program for arranging multivariate data in an ordered two-way table by classification of the individuals and attributes. <span>TWINSPAN</span>.
</div>
<div class="csl-entry">
Hobbs, N. T., and R. Hilborn. 2006. <a href="https://doi.org/10.1890/04-0645">Alternatives to statistical hypothesis testing in ecology: <span class="nocase">a</span> guide to self teaching</a>. Ecological Applications 16:5–19.
</div>
<div class="csl-entry">
Højsgaard, S., U. Halekoh, and J. Yan. 2006. <a href="https://doi.org/10.18637/jss.v015.i02">The <span>R</span> package geepack for generalized estimating equations</a>. Journal of statistical software 15:1–11.
</div>
<div class="csl-entry">
Hurlbert, S. H. 1984. <a href="https://doi.org/10.2307/1942661">Pseudoreplication and the design of ecological field experiments</a>. Ecological <span>M</span>onographs 54:187–211.
</div>
<div class="csl-entry">
Hutcheon, J. M., J. A. W. Kirsch, and T. G. Jr. 2002. <a href="https://doi.org/10.1159/000065938">A comparative analysis of brain size in relation to foraging ecology and phylogeny in the <span>C</span>hiroptera</a>. Brain, behavior and evolution 60:165–180.
</div>
<div class="csl-entry">
Inderjit, J. C. Streibig, and M. Olofsdotter. 2002. <a href="https://doi.org/10.1034/j.1399-3054.2002.1140312.x">Joint action of phenolic acid mixtures and its significance in allelopathy research</a>. Physiologia Plantarum 114:422–428.
</div>
<div class="csl-entry">
James, G., D. Witten, T. Hastie, and R. Tibshirani. 2013. <a href="http://faculty.marshall.usc.edu/gareth-james/ISL/">An introduction to statistical learning</a>. 7th edition. Springer.
</div>
<div class="csl-entry">
Jones, D. T. 2019. <a href="https://doi.org/10.1038/s41580-019-0176-5">Setting the standards for machine learning in biology</a>. Nature <span>R</span>eviews <span>M</span>olecular <span>C</span>ell <span>B</span>iology 20:659–660.
</div>
<div class="csl-entry">
Kan, A. 2017. <a href="https://doi.org/10.1038/icb.2017.16">Machine learning applications in cell image analysis</a>. Immunology and <span>C</span>ell <span>B</span>iology 95:525–530.
</div>
<div class="csl-entry">
Kéry, M. 2002. Inferring the absence of a species: <span class="nocase">a</span> case study of snakes. The Journal of <span>W</span>ildlife <span>M</span>anagement:330–338.
</div>
<div class="csl-entry">
Kéry, M. 2010. Introduction to <span>W</span>in<span>BUGS</span> for ecologists: <span>B</span>ayesian approach to regression, <span>ANOVA</span>, mixed models and related analyses. Academic Press.
</div>
<div class="csl-entry">
Kirchman, D. L., X. A. G. Morán, and H. Ducklow. 2009. Microbial growth in the polar oceans—role of temperature and potential impact of climate change. Nature <span>R</span>eviews <span>M</span>icrobiology 7:451–459.
</div>
<div class="csl-entry">
Kirchner, B. N., N. S. Green, D. A. Sergeant, J. N. Mink, and K. T. Wilkins. 2011. <a href="https://doi.org/10.1674/0003-0031-166.1.112">Responses of small mammals and vegetation to a prescribed burn in a tallgrass blackland prairie</a>. The American Midland Naturalist 166:112–125.
</div>
<div class="csl-entry">
Koenker, R. 2021. <a href="https://CRAN.R-project.org/package=quantreg">Quantreg: Quantile regression</a>.
</div>
<div class="csl-entry">
Kuehn, I., and C. F. Dormann. 2012. <a href="https://doi.org/10.1111/j.1365-2699.2012.02707.x">Less than eight (and a half) misconceptions of spatial analysis</a>. Journal of Biogeography 39:995–998.
</div>
<div class="csl-entry">
Kunz, B. K., J. H. Waddle, and N. S. Green. 2019. <a href="https://doi.org/10.1002/ieam.4202">Amphibian monitoring in hardwood forests: Optimizing methods for contaminant-based compensatory restorations</a>. Integrated environmental assessment and management.
</div>
<div class="csl-entry">
Kuznetsova, A., P. B. Brockhoff, and R. H. B. Christensen. 2017. <a href="https://doi.org/10.18637/jss.v082.i13"><span class="nocase">l</span>mer<span>T</span>est package: Tests in linear mixed effects models</a>. Journal of Statistical Software 82:1–26.
</div>
<div class="csl-entry">
Lazic, S. E., C. J. Clarke-Williams, and M. R. Munafò. 2018. <a href="https://doi.org/10.1371/journal.pbio.2005282">What exactly is ‘<span>N</span>’in cell culture and animal experiments?</a> <span>PLoS</span> <span>B</span>iology 16:e2005282.
</div>
<div class="csl-entry">
Lebreton, J.-D., K. P. Burnham, J. Clobert, and D. R. Anderson. 1992. <a href="https://doi.org/10.2307/2937171">Modeling survival and testing biological hypotheses using marked animals: <span class="nocase">a</span> unified approach with case studies</a>. Ecological <span>M</span>onographs 62:67–118.
</div>
<div class="csl-entry">
Legendre, P., and L. Legendre. 2012. Numerical ecology. Elsevier, Amsterdam.
</div>
<div class="csl-entry">
Leonard, E. M., and C. M. Wood. 2013. <a href="https://doi.org/10.1016/j.cbpc.2013.03.008">Acute toxicity, critical body residues, <span>M</span>ichaelis–<span>M</span>enten analysis of bioaccumulation, and ionoregulatory disturbance in response to waterborne nickel in four invertebrates: <em><span>C</span>hironomus riparius</em>, <em><span>L</span>ymnaea stagnalis</em>, <em><span>L</span>umbriculus variegatus</em> and <em><span>D</span>aphnia pulex</em></a>. Comparative Biochemistry and Physiology Part C: Toxicology &amp; Pharmacology 158:10–21.
</div>
<div class="csl-entry">
Liang, K.-Y., and S. L. Zeger. 1986. Longitudinal data analysis using generalized linear models. Biometrika 73:13–22.
</div>
<div class="csl-entry">
Link, W. A., E. Cam, J. D. Nichols, and E. G. Cooch. 2002. <a href="https://doi.org/10.2307/3803160">Of <span>BUGS</span> and birds: <span>M</span>arkov chain <span>M</span>onte <span>C</span>arlo for hierarchical modeling in wildlife research</a>. The Journal of <span>W</span>ildlife <span>M</span>anagement:277–291.
</div>
<div class="csl-entry">
Ludwig, J. A., J. F. Reynolds, L. QUARTET, and J. Reynolds. 1988. Statistical ecology: A primer in methods and computing. John Wiley &amp; Sons.
</div>
<div class="csl-entry">
Lunn, D., D. Spiegelhalter, A. Thomas, and N. Best. 2009. <a href="https://doi.org/10.1002/sim.3680">The <span>BUGS</span> project: <span class="nocase">e</span>volution, critique and future directions</a>. Statistics in <span>M</span>edicine 28:3049–3067.
</div>
<div class="csl-entry">
Makin, T. R., and J.-J. O. de Xivry. 2019. <a href="https://doi.org/10.7554/eLife.48175">Science forum: Ten common statistical mistakes to watch out for when writing or reviewing a manuscript</a>. Elife 8:e48175.
</div>
<div class="csl-entry">
Mangiafico, S. 2021. <a href="https://CRAN.R-project.org/package=rcompanion"><span class="nocase">r</span>companion: Functions to support extension education program evaluation</a>.
</div>
<div class="csl-entry">
Mantel, N. 1967. <a href="https://cancerres.aacrjournals.org/content/27/2_Part_1/209">The detection of disease clustering and a generalized regression approach</a>. Cancer research 27:209–220.
</div>
<div class="csl-entry">
McCarthy, M. A. 2007. Bayesian methods for ecology. Cambridge University Press, Cambridge, United Kingdom.
</div>
<div class="csl-entry">
McCune, B., J. B. Grace, and D. L. Urban. 2002. Analysis of ecological communities. MjM software design Gleneden Beach, Oregon, USA.
</div>
<div class="csl-entry">
Nelder, J. A., and R. W. M. Wedderburn. 1972. <a href="https://doi.org/10.2307/2344614">Generalized linear models</a>. Journal of the Royal Statistical Society: Series A (General) 135:370.
</div>
<div class="csl-entry">
Penone, C., A. D. Davidson, K. T. Shoemaker, M. Di Marco, C. Rondinini, T. M. Brooks, B. E. Young, C. H. Graham, and G. C. Costa. 2014. <a href="https://doi.org/10.1111/2041-210X.12232">Imputation of missing data in life-history trait datasets: Which approach performs the best?</a> Methods in Ecology and Evolution 5:961–970.
</div>
<div class="csl-entry">
Pethybridge, H., C. A. Choy, J. M. Logan, V. Allain, A. Lorrain, N. Bodin, C. J. Somes, J. Young, F. Ménard, C. Langlais, L. Duffy, A. J. Hobday, P. Kuhnert, B. Fry, C. Menkes, and R. J. Olson. 2018. <a href="https://doi.org/10.1111/geb.12763">A global meta-analysis of marine predator nitrogen stable isotopes: Relationships between trophic structure and environmental conditions</a>. Global Ecology and Biogeography 27:1043–1055.
</div>
<div class="csl-entry">
Plummer, M. 2003. <span>JAGS</span>: A program for analysis of <span>B</span>ayesian graphical models using <span>G</span>ibbs sampling. Pages 1–10 Proceedings of the 3rd international workshop on distributed statistical computing. Vienna, Austria.
</div>
<div class="csl-entry">
Plummer, M. 2021. <a href="https://CRAN.R-project.org/package=rjags"><span class="nocase">r</span>jags: <span>B</span>ayesian graphical models using <span>MCMC</span></a>.
</div>
<div class="csl-entry">
Ritz, C., F. Baty, J. C. Streibig, and D. Gerhard. 2015. <a href="https://doi.org/10.1371/journal.pone.0146021">Dose-response analysis using <span>R</span></a>. PLoS ONE 10:e0146021.
</div>
<div class="csl-entry">
Roleček, J., L. Tichý, D. Zelený, and M. Chytrý. 2009. <a href="https://doi.org/10.1111/j.1654-1103.2009.01062.x">Modified <span>TWINSPAN</span> classification in which the hierarchy respects cluster heterogeneity</a>. Journal of Vegetation Science 20:596–602.
</div>
<div class="csl-entry">
Schabenberger, O., and F. J. Pierce. 2001. Contemporary statistical models for the plant and soil sciences. CRC press, Boca Raton, Florida, USA.
</div>
<div class="csl-entry">
Smith, F. A., and S. K. Lyons. 2013. Animal body size: Linking pattern and process across space, time, and taxonomic group. University of Chicago Press, Chicago.
</div>
<div class="csl-entry">
Spector, P. 2008. Data manipulation with <span>R</span>. Springer Science+Business Media, New York.
</div>
<div class="csl-entry">
Spiess, A.-N. 2018. <a href="https://CRAN.R-project.org/package=propagate"><span class="nocase">p</span>ropagate: Propagation of uncertainty</a>.
</div>
<div class="csl-entry">
Spiess, A.-N., and N. Neumeyer. 2010. <a href="https://doi.org/10.1186/1471-2210-10-6">An evaluation of <span class="math inline">\(R^2\)</span> as an inadequate measure for nonlinear models in pharmacological and biochemical research: A <span>M</span>onte <span>C</span>arlo approach</a>. BMC Pharmacology 10.
</div>
<div class="csl-entry">
Su, Y.-S., and Masanao Yajima. 2021. <a href="https://CRAN.R-project.org/package=R2jags"><span>R2</span>jags: Using <span>R</span> to run ’<span>JAGS</span>’</a>.
</div>
<div class="csl-entry">
Tarca, A. L., V. J. Carey, X. Chen, R. Romero, and S. Drăghici. 2007. <a href="https://doi.org/10.1371/journal.pcbi.0030116">Machine learning and its applications to biology</a>. <span>PLoS</span> <span>C</span>omputational <span>B</span>iology 3:e116.
</div>
<div class="csl-entry">
Therneau, T., and B. Atkinson. 2019. <a href="https://CRAN.R-project.org/package=rpart"><span class="nocase">r</span>part: Recursive partitioning and regression trees</a>.
</div>
<div class="csl-entry">
Tufte, E. 2001. The visual display of quantitative information. Cheshire: Graphic Press.–2001.–213 p.
</div>
<div class="csl-entry">
Turner, H., and D. Firth. 2020. <a href="https://cran.r-project.org/package=gnm">Generalized nonlinear models in <span>R</span>: An overview of the gnm package</a>.
</div>
<div class="csl-entry">
Venables, W. N., and B. D. Ripley. 2002. <a href="https://www.stats.ox.ac.uk/pub/MASS4/">Modern applied statistics with <span>S</span></a>. Fourth. Springer, New York.
</div>
<div class="csl-entry">
Vicente, J., Y. Fierro, and C. Gortazar. 2005. <a href="https://doi.org/10.1007/s00436-004-1255-9">Seasonal dynamics of the fecal excretion of <em><span>E</span>laphostrongylus cervi</em> (<span>N</span>ematoda, <span>M</span>etastrongyloidea) first-stage larvae in <span>I</span>berian red deer (<em><span>C</span>ervus elaphus hispanicus) from southern <span>S</span>pain</em></a>. Parasitology Research 95:60–64.
</div>
<div class="csl-entry">
Voit, E. O. 2019. <a href="https://doi.org/10.1371/journal.pcbi.1007279">Perspective: <span class="nocase">d</span>imensions of the scientific method</a>. <span>PLoS</span> <span>C</span>omputational <span>B</span>iology 15:e1007279.
</div>
<div class="csl-entry">
Ward, J. H. 1963. <a href="https://doi.org/10.1080/01621459.1963.10500845">Hierarchical grouping to optimize an objective function</a>. Journal of the American Statistical Association 58:236–244.
</div>
<div class="csl-entry">
Warton, D. I., and F. K. C. Hui. 2011. <a href="https://doi.org/10.1890/10-0340.1">The arcsine is asinine: The analysis of proportions in ecology</a>. Ecology 92:3–10.
</div>
<div class="csl-entry">
Wasserstein, R. L., and N. A. Lazar. 2016. <a href="https://doi.org/10.1080/00031305.2016.1154108">The ASA statement on p-values: Context, process, and purpose</a>. Taylor &amp; Francis.
</div>
<div class="csl-entry">
Werner, J., and E. M. Griebeler. 2014. <a href="https://doi.org/10.1371/journal.pone.0088834">Allometries of maximum growth rate versus body mass at maximum growth indicate that non-avian dinosaurs had growth rates typical of fast growing ectothermic sauropsids</a>. <span>PloS</span> <span>ONE</span> 9:e88834.
</div>
<div class="csl-entry">
White, J. W., A. Rassweiler, J. F. Samhouri, A. C. Stier, and C. White. 2014. <a href="https://doi.org/10.1111/j.1600-0706.2013.01073.x">Ecologists should not use statistical significance tests to interpret simulation model results</a>. Oikos 123:385–388.
</div>
<div class="csl-entry">
Whytock, R. C., J. Świeżewski, J. A. Zwerts, T. Bara-Słupski, A. F. Koumba Pambo, M. Rogala, L. Bahaa-el-din, K. Boekee, S. Brittain, A. W. Cardoso, and others. 2021. <a href="https://doi.org/10.1111/2041-210X.13576">Robust ecological analysis of camera trap data labelled by a machine learning model</a>. Methods in <span>E</span>cology and <span>E</span>volution.
</div>
<div class="csl-entry">
Wickham, H. 2014. <a href="https://doi.org/10.18637/jss.v059.i10">Tidy data</a>. Journal of Statistical Software 59.
</div>
<div class="csl-entry">
Wickham, H., M. Averick, J. Bryan, W. Chang, L. D. McGowan, R. François, G. Grolemund, A. Hayes, L. Henry, J. Hester, M. Kuhn, T. L. Pedersen, E. Miller, S. M. Bache, K. Müller, J. Ooms, D. Robinson, D. P. Seidel, V. Spinu, K. Takahashi, D. Vaughan, C. Wilke, K. Woo, and H. Yutani. 2019. <a href="https://doi.org/10.21105/joss.01686">Welcome to the <span class="nocase">tidyverse</span></a>. Journal of Open Source Software 4:1686.
</div>
<div class="csl-entry">
Wood, S. N. 2017. Generalized additive models: An introduction with <span>R</span>. 2nd edition. CRC press.
</div>
<div class="csl-entry">
Woodward, H. N., E. A. F. Fowler, J. O. Farlow, and J. R. Horner. 2015. <a href="https://doi.org/10.1017/pab.2015.19"><em><span>M</span>aiasaura</em>, a model organism for extinct vertebrate population biology: A large sample statistical assessment of growth dynamics and survivorship</a>. Paleobiology 41:503–527.
</div>
<div class="csl-entry">
Zuur, A. F., E. N. Ieno, and C. S. Elphick. 2010. <a href="https://doi.org/10.1111/j.2041-210X.2009.00001.x">A protocol for data exploration to avoid common statistical problems</a>. Methods in ecology and evolution 1:3–14.
</div>
<div class="csl-entry">
Zuur, A. F., E. N. Ieno, and G. M. Smith. 2007. Analysing ecological data. Springer Science+Business Media, New York.
</div>
<div class="csl-entry">
Zuur, A. F., E. N. Ieno, N. J. Walker, A. A. Saveliev, G. M. Smith, and others. 2009a. Mixed effects models and extensions in ecology with <span>R</span>. Springer, New York.
</div>
<div class="csl-entry">
Zuur, A., E. N. Ieno, and E. Meesters. 2009b. A beginner’s guide to <span>R</span>. Springer Science+Business Media, New York.
</div>
</div>
</div>
<div class="footnotes footnotes-end-of-document">
<hr />
<ol>
<li id="fn1"><p>Also, gambling.<a href="1.2-statistics-in-modern-biology.html#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p> This imaginary experiment is <strong><em>very</em></strong> loosely inspired by a real study <span class="citation">(<a href="#ref-desolla2006" role="doc-biblioref">De Solla et al. 2006</a>)</span>, with some details changed to facilitate classroom discussion. The results and discussion in this document should <strong><em>not</em></strong> be taken to represent the authors’ work.<a href="1.2-statistics-in-modern-biology.html#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>PBCAK is a very common acronym in IT circles.<a href="1.3-misuses-of-statistics.html#fnref3" class="footnote-back">↩︎</a></p></li>
<li id="fn4"><p>The best bet is probably analysis of variance (ANOVA). Both linear regression and ANOVA are special cases of <strong>linear models</strong>.<a href="1.3-misuses-of-statistics.html#fnref4" class="footnote-back">↩︎</a></p></li>
<li id="fn5"><p>Source: <a href="https://xkcd.com/882/">XKCD</a><a href="1.3-misuses-of-statistics.html#fnref5" class="footnote-back">↩︎</a></p></li>
<li id="fn6"><p>Not to mention among generations of students taking statistics courses.<a href="1.4-p-values-and-null-hypothesis-significance-testing-nhst.html#fnref6" class="footnote-back">↩︎</a></p></li>
<li id="fn7"><p>If this sounds convoluted, that’s because it is.<a href="1.4-p-values-and-null-hypothesis-significance-testing-nhst.html#fnref7" class="footnote-back">↩︎</a></p></li>
<li id="fn8"><p>Source: <a href="https://xkcd.com/2545/">XKCD</a><a href="1.5-alternatives-to-nhst.html#fnref8" class="footnote-back">↩︎</a></p></li>
<li id="fn9"><p><a href="https://xkcd.com/285/">Citation needed</a><a href="1.5-alternatives-to-nhst.html#fnref9" class="footnote-back">↩︎</a></p></li>
<li id="fn10"><p><a href="https://www.nei.nih.gov/learn-about-eye-health/resources-for-health-educators/eye-health-data-and-statistics/nearsightedness-myopia-data-and-statistics/nearsightedness-myopia-tables">Source</a><a href="1.5-alternatives-to-nhst.html#fnref10" class="footnote-back">↩︎</a></p></li>
<li id="fn11"><p><a href="https://www.bls.gov/emp/tables/emp-by-detailed-occupation.htm">Source</a><a href="1.5-alternatives-to-nhst.html#fnref11" class="footnote-back">↩︎</a></p></li>
<li id="fn12"><p>Here’s a <a href="https://www.youtube.com/watch?v=HZGCoVF3YvM">nice video illustrating this concept</a>. I can’t recommend this channel highly enough!<a href="1.5-alternatives-to-nhst.html#fnref12" class="footnote-back">↩︎</a></p></li>
<li id="fn13"><p>As you might suspect, the method is MUCH more complicated than this. <span class="citation">McCarthy (<a href="#ref-mccarthy2007" role="doc-biblioref">2007</a>)</span> has a good introduction to MCMC. A more compact version is in <span class="citation">Link et al. (<a href="#ref-link2002" role="doc-biblioref">2002</a>)</span><a href="1.5-alternatives-to-nhst.html#fnref13" class="footnote-back">↩︎</a></p></li>
<li id="fn14"><p>Source: <a href="https://xkcd.com/1236">XKCD</a><a href="1.5-alternatives-to-nhst.html#fnref14" class="footnote-back">↩︎</a></p></li>
<li id="fn15"><p>Source: <a href="https://xkcd.com/1838/">XKCD</a><a href="1.5-alternatives-to-nhst.html#fnref15" class="footnote-back">↩︎</a></p></li>
<li id="fn16"><p>For example, if a journal editor is adamant about some minute detail of formatting on a figure.<a href="2.1-getting-started-with-r.html#fnref16" class="footnote-back">↩︎</a></p></li>
<li id="fn17"><p>Your RStudio might look different at first. You can change the color scheme by going to Tools - Global Options - Appearance and selecting a theme. Here I’m using the “Classic” theme with size 12 Courier New font and 110% zoom.<a href="2.3-using-r.html#fnref17" class="footnote-back">↩︎</a></p></li>
<li id="fn18"><p>Image: F. Atirador. URL: <a href="https://upload.wikimedia.org/wikipedia/commons/5/5a/Irish_Elk_Side_%28white_background%29.jpg" class="uri">https://upload.wikimedia.org/wikipedia/commons/5/5a/Irish_Elk_Side_%28white_background%29.jpg</a> (accessed 2021-12-20).<a href="2.4-mod-02-first.html#fnref18" class="footnote-back">↩︎</a></p></li>
<li id="fn19"><p>This issue seems to confuse a lot students who are Mac users, so I’m belaboring the point here.<a href="2.4-mod-02-first.html#fnref19" class="footnote-back">↩︎</a></p></li>
<li id="fn20"><p>As we’ll discuss in the <a href="3.1-mod-03-inout.html#mod-03-inout">data manipulation module</a>, naming R objects is a bit of an art. Names should be informative yet brief. I use a standard list of names like “<code>dat</code>”, “<code>met</code>”, and “<code>res</code>”, etc. in all of my projects because they are short but easy to identify (“data”, “metadata”, and “results”, respectively).<a href="2.4-mod-02-first.html#fnref20" class="footnote-back">↩︎</a></p></li>
<li id="fn21"><p>I don’t recommend doing this, as the desktop can become easily cluttered.<a href="2.4-mod-02-first.html#fnref21" class="footnote-back">↩︎</a></p></li>
<li id="fn22"><p>The Mac desktop is <strong>/Users/username/Desktop</strong> (where <strong>username</strong> is your username) or just <strong>~/Desktop</strong>.<a href="2.4-mod-02-first.html#fnref22" class="footnote-back">↩︎</a></p></li>
<li id="fn23"><p>This is true if there are no random elements in your code. If you draw any random numbers, you will need to set the random number seed in your script to ensure exact reproducibility.<a href="2.4-mod-02-first.html#fnref23" class="footnote-back">↩︎</a></p></li>
<li id="fn24"><p>It never even occurred to me to try this until I was prepping these course notes.<a href="2.5-write-and-execute-commands-in-the-r-console.html#fnref24" class="footnote-back">↩︎</a></p></li>
<li id="fn25"><p>This means a <em>very</em> minimal code example that people can use to get the same problem that you are having. Don’t post your entire dataset or include many lines of pre-processing steps or other code that led up to your problem. Use base R functions or built-in datasets to create a small dataset or set of values that will immediately provoke the same error code. Often the process of doing this will help you solve your own problem!<a href="2.10-r-documentation.html#fnref25" class="footnote-back">↩︎</a></p></li>
<li id="fn26"><p>I think this was originally a tweet that went viral, but the joke spread through the internet and now you can buy it on a t-shirt. Regardless, this phenomenon is much older than Twitter.<a href="2.10-r-documentation.html#fnref26" class="footnote-back">↩︎</a></p></li>
<li id="fn27"><p>A binary operator uses two arguments, one before the symbol and one after. For example, <code>+</code> is a binary operator that adds the the argument before the <code>+</code> to the argument after the <code>+</code>.<a href="3.3-selecting-data-with.html#fnref27" class="footnote-back">↩︎</a></p></li>
<li id="fn28"><p>This is slightly longer than a <em>Rent</em> year, which is 525600 minutes (31536000 seconds).<a href="3.4-managing-dates-and-characters.html#fnref28" class="footnote-back">↩︎</a></p></li>
<li id="fn29"><p><code>cbind()</code> also works on matrices.<a href="3.5-mod-03-dataframe.html#fnref29" class="footnote-back">↩︎</a></p></li>
<li id="fn30"><p>R users often use an ellipsis <code>...</code> as a shorthand for unshown arguments to the function, not literally three periods. In this text, the ellipsis is a stand in for other arguments to the function that are irrelevant to the point being made. There are some circumstances where an ellipsis is necessary to control how variables are passed from one function to another<a href="4.1-descriptive-and-summary-statistics.html#fnref30" class="footnote-back">↩︎</a></p></li>
<li id="fn31"><p>I can never remember how this works, so usually have to tinker with the code until I get the table I want :)<a href="4.1-descriptive-and-summary-statistics.html#fnref31" class="footnote-back">↩︎</a></p></li>
<li id="fn32"><p>Distribution <code>x4</code> is a mixture of a normal (mean = 5, SD = 2) with a uniform in [3, 20].<a href="4.2-mod-04-vis.html#fnref32" class="footnote-back">↩︎</a></p></li>
<li id="fn33"><p>adapted from <a href="https://commons.wikimedia.org/wiki/File:Discrete_probability_distribution_illustration.svg" class="uri">https://commons.wikimedia.org/wiki/File:Discrete_probability_distribution_illustration.svg</a> (public domain)<a href="4.2-mod-04-vis.html#fnref33" class="footnote-back">↩︎</a></p></li>
<li id="fn34"><p>After all, how useful is a dataset with a sample size of 1?<a href="4.3-mod-04-dists1.html#fnref34" class="footnote-back">↩︎</a></p></li>
<li id="fn35"><p>If your brain hurts, that’s completely normal. This function gives me headaches too<a href="4.3-mod-04-dists1.html#fnref35" class="footnote-back">↩︎</a></p></li>
<li id="fn36"><p>usually greater than the mean<a href="4.3-mod-04-dists1.html#fnref36" class="footnote-back">↩︎</a></p></li>
<li id="fn37"><p>It is possible, and not uncommon, to analyze data using the natural log scale but plot results on the log<sub>10</sub> scale. This works because the choice of base is arbitrary for most analyses and because it is simple to convert values between scales with different bases (i.e., change of base).<a href="4.3-mod-04-dists1.html#fnref37" class="footnote-back">↩︎</a></p></li>
<li id="fn38"><p>In R, statistics, and in mathematics in general, if you see “log” it usually means “natural log”.<a href="4.3-mod-04-dists1.html#fnref38" class="footnote-back">↩︎</a></p></li>
<li id="fn39"><p>Recall that the logarithm is the inverse of exponentiation: if <span class="math inline">\(a=b^c\)</span>, then <span class="math inline">\(log_b\left(a \right)=c\)</span>. For example, <span class="math inline">\(log_2 \left(8\right)=3\)</span> because <span class="math inline">\(2^3 = 8\)</span>.<a href="4.5-mod-04-trans.html#fnref39" class="footnote-back">↩︎</a></p></li>
<li id="fn40"><p>Whatever you do, don’t call the natural logarithm “lawn” unless you want to look silly.<a href="4.5-mod-04-trans.html#fnref40" class="footnote-back">↩︎</a></p></li>
<li id="fn41"><p><a href="https://www.youtube.com/watch?v=VSi0Z04fWj0">This video</a> illustrates the usefulness of logarithms in statistics <strong><em>and</em></strong> has a nice theme song.<a href="4.5-mod-04-trans.html#fnref41" class="footnote-back">↩︎</a></p></li>
<li id="fn42"><p>Remember that exponential functions relate addition and multiplication. This is how exponentiation is extended to non-integer powers: <span class="math inline">\(X^aX^b=X^{a+b}\)</span>. <a href="4.5-mod-04-trans.html#fnref42" class="footnote-back">↩︎</a></p></li>
<li id="fn43"><p>An <strong>open interval</strong> is an interval (a, b) that contains values &gt; <em>a</em> and &lt; <em>b</em>, but never <em>a</em> or <em>b</em>. Open intervals are denoted by parentheses (). Intervals that contain their limits are called <strong>closed intervals</strong>, denoted <span class="math inline">\(\left[a,b\right]\)</span>.<a href="4.5-mod-04-trans.html#fnref43" class="footnote-back">↩︎</a></p></li>
<li id="fn44"><p>You can also specify colors by their RGB or hex codes. A graphical reference is <a href="https://rstudio-pubs-static.s3.amazonaws.com/3486_79191ad32cf74955b4502b8530aad627.html">here</a> (accessed 2021-12-17). The <a href="https://colorbrewer2.org/">ColorBrewer</a> website (accessed 2021-12-17) and R package can help picking appropriate color schemes.<a href="4.6-multivariate-data-exploration.html#fnref44" class="footnote-back">↩︎</a></p></li>
<li id="fn45"><p>Locally estimated sums of squares, a common smoothing curve algorithm.<a href="4.6-multivariate-data-exploration.html#fnref45" class="footnote-back">↩︎</a></p></li>
<li id="fn46"><p>In the base R PCA functions, these values are obtained from the <code>rotation</code> part of the output in <code>prcomp()</code> or from the <code>loadings</code> component in <code>princomp()</code>.<a href="4.7-ordination-brief-introduction.html#fnref46" class="footnote-back">↩︎</a></p></li>
<li id="fn47"><p>An introduction by the package’s lead author can be found here: <a href="https://cran.r-project.org/web/packages/vegan/vignettes/intro-vegan.pdf" class="uri">https://cran.r-project.org/web/packages/vegan/vignettes/intro-vegan.pdf</a> (accessed 2021-09-17)<a href="4.7-ordination-brief-introduction.html#fnref47" class="footnote-back">↩︎</a></p></li>
<li id="fn48"><p>When calculating a distance matrix using geographic coordinates, be sure to account for the curvature of the Earth. Often using unprojected latitude and longitude will produce a misleading distance matrix, because latitude and longitude are not equivalent.<a href="4.8-mod-04-prob.html#fnref48" class="footnote-back">↩︎</a></p></li>
<li id="fn49"><p>This is mirrored in how different statistical software packages express the variability of a normal distribution. R uses the SD; others use variance or precision (reciprocal of SD or variance). Read the manual!<a href="5.1-mod-05-lm.html#fnref49" class="footnote-back">↩︎</a></p></li>
<li id="fn50"><p>The <span class="math inline">\(\beta\)</span> and <span class="math inline">\(\varepsilon\)</span> are supposed to be bold, but I can’t figure out the right combination of markdown and Latex to get bold Greek letters.<a href="5.1-mod-05-lm.html#fnref50" class="footnote-back">↩︎</a></p></li>
<li id="fn51"><p>This is another version of <span class="math inline">\(\varepsilon\)</span>.<a href="5.1-mod-05-lm.html#fnref51" class="footnote-back">↩︎</a></p></li>
<li id="fn52"><p>The <em>exponential distribution</em> is included in the <em>exponential family</em> of distriutions, which includes many other distributions. Unfortunately the word “family” in a GLM context refers to individual distributions. So, fitting a GLM with an exponential family means using the exponential distribution for residuals.<a href="5.2-mod-05-basic.html#fnref52" class="footnote-back">↩︎</a></p></li>
<li id="fn53"><p>One of these days I might add a section on zero-inflated models.<a href="5.2-mod-05-basic.html#fnref53" class="footnote-back">↩︎</a></p></li>
<li id="fn54"><p>Emphasis on the “kind of”. There are some important ways that link functions are <strong><em>not</em></strong> like transformations, as we shall see later.<a href="5.2-mod-05-basic.html#fnref54" class="footnote-back">↩︎</a></p></li>
<li id="fn55"><p>Some sources say negative inverse, but R and most textbooks say to use (positive) inverse.<a href="5.2-mod-05-basic.html#fnref55" class="footnote-back">↩︎</a></p></li>
<li id="fn56"><p>I highly recommend <a href="https://www.youtube.com/watch?v=9T0wlKdew6I">this video</a> for an introduction to deviance, which also has a nice theme song<a href="5.2-mod-05-basic.html#fnref56" class="footnote-back">↩︎</a></p></li>
<li id="fn57"><p>Obviously, this is an overfit model, but that’s ok because we are using it as a hypothetical “baseline”.<a href="5.2-mod-05-basic.html#fnref57" class="footnote-back">↩︎</a></p></li>
<li id="fn58"><p>By the way, the map in this section and all geographic operations were done using the geographic information systems (GIS) capabilities in R.<a href="5.4-mod-05-poisson.html#fnref58" class="footnote-back">↩︎</a></p></li>
<li id="fn59"><p>As you might suspect, the opposite situation, when <span class="math inline">\(\mu &gt; \sigma^2\)</span>, is called “underdispersion”. Underdispersion does not come up as often as overdispersion and poses less of a problem for analysis.<a href="5.5-mod-05-quasi.html#fnref59" class="footnote-back">↩︎</a></p></li>
<li id="fn60"><p>As you may have guessed as you clicked the footnote, this map was made in R.<a href="5.5-mod-05-quasi.html#fnref60" class="footnote-back">↩︎</a></p></li>
<li id="fn61"><p>I have never really figured out the syntax for how <code>mosaicplot()</code> splits the data (i.e, order of terms in the formula). Try different permutations of the formula until you get the plot you want :)<a href="5.6-mod-05-logistic.html#fnref61" class="footnote-back">↩︎</a></p></li>
<li id="fn62"><p>Infuriatingly, the base function <code>predict.nls()</code> takes <code>se.fit</code> as an argument, but ignores it. This has been the case since I was grad student learning about nonlinear models.<a href="6.3-mod-06-micmen.html#fnref62" class="footnote-back">↩︎</a></p></li>
<li id="fn63"><p>Dummy variables are really a way to manipulate the matrix algebra underlying the model-fitting process. Usually, a factor with <em>n</em> levels will be encoded by <span class="math inline">\(n-1\)</span> dummy variables where each observation has a 1 in at most one of the dummies—the dummy corresponding to the group they belong to. An observation with 0s for all dummies is interpreted as being in the control or baseline group. This example uses a slightly different approach to dummy coding where each level corresponds to one dummy variable and each observation has value 1 for exactly 1 dummy. The first method (<span class="math inline">\(n-1\)</span> dummies) leads to an <strong>effects parameterization</strong> of the model, where effects are presented as changes from the baseline (e.g., as seen in R’s <code>lm()</code> outputs). The second method (<em>n</em> dummies) produces a <strong>means parameterization</strong> of the model, where parameter estimates are presented as group- or level-specific point estimates.<a href="6.3-mod-06-micmen.html#fnref63" class="footnote-back">↩︎</a></p></li>
<li id="fn64"><p>Theoretically, the intercept and slope could each vary by different grouping variables. However, in my opinion, you would have to work very hard to justify that on biological grounds.<a href="7.2-mod-07-lmm.html#fnref64" class="footnote-back">↩︎</a></p></li>
<li id="fn65"><p><a href="https://stats.stackexchange.com/questions/4700/what-is-the-difference-between-fixed-effect-random-effect-and-mixed-effect-mode">Discussion on Stack Overflow</a>, with references. <a href="https://statmodeling.stat.columbia.edu/2005/01/25/why_i_dont_use/">Blog post</a> from Andrew Gelman, prominent statistician. Mixed models are presented and described thoroughly in <span class="citation">Bolker (<a href="#ref-bolker2008ecological" role="doc-biblioref">2008</a>)</span>, <span class="citation">Bolker et al. (<a href="#ref-bolker2009" role="doc-biblioref">2009</a>)</span>, and <span class="citation">Zuur et al. (<a href="#ref-zuur2009mixed" role="doc-biblioref">2009a</a>)</span>.<a href="7.2-mod-07-lmm.html#fnref65" class="footnote-back">↩︎</a></p></li>
<li id="fn66"><p>However, there is no such thing as a free lunch. Just because BLUPs, a.k.a. “conditional modes”, can be estimated for different levels of a factor does not mean that these estimates are equivalent to or as robust as analogous fixed effects.<a href="7.2-mod-07-lmm.html#fnref66" class="footnote-back">↩︎</a></p></li>
<li id="fn67"><p>The source of this example <span class="citation">(<a href="#ref-kery2010" role="doc-biblioref">Kéry 2010</a>)</span> used these data to illustrate fitting Bayesian mixed models with MCMC. I modified the data-generating parameters a bit to make the model more tractable by <code>lme4</code>. <span class="citation">Kéry (<a href="#ref-kery2010" role="doc-biblioref">2010</a>)</span> is one of the books I recommend to students interested in learning Bayesian statistics.<a href="7.2-mod-07-lmm.html#fnref67" class="footnote-back">↩︎</a></p></li>
<li id="fn68"><p> A famous statistician once remarked that BLUPs are like the Holy Roman Empire, because just as that was neither holy, nor Roman, nor an empire, a BLUP is not necessarily “best”, “linear”, or “unbiased”. I can’t find the quote anymore.<a href="7.2-mod-07-lmm.html#fnref68" class="footnote-back">↩︎</a></p></li>
<li id="fn69"><p>This is a bit like asking if humans should be counted as Primates or as Mammals.<a href="7.2-mod-07-lmm.html#fnref69" class="footnote-back">↩︎</a></p></li>
<li id="fn70"><p>This goes to show why <em>reading the manual</em> is important. I can’t count the number of times I’ve spent hours trying to program a custom solution to a problem, only to find that someone had already developed something much more elegant and robust and put it in the documentation.<a href="7.2-mod-07-lmm.html#fnref70" class="footnote-back">↩︎</a></p></li>
<li id="fn71"><p>One of the things I love about R is the degree of control you have when making figures. Every single piece of a figure can be specified <em>exactly</em> the way you want it. This level of detail is very difficult (if not impossible) in Microsoft Excel. It can be approximated with graphics programs like Adobe Illustrator, but you’ll never get the precision as you do when creating figure elements in R. Using R ensures that coordinate systems, coordinate values, colors, alignment, etc., are used precisely. Of course, all of this precision comes at a cost: complex figures can require a lot of code. In my workflow it’s not unusual for really complicated figures to require &gt;100 lines of code. This is partly because there really are a lot of elements to add individually, and partly because I’m a bit obsessive about getting figures just right.<a href="7.2-mod-07-lmm.html#fnref71" class="footnote-back">↩︎</a></p></li>
<li id="fn72"><p>In principle, there is no reason why you couldn’t have random effects following a non-normal distribution…such a model can be easily defined in BUGS or JAGS but I’m not aware of a way to do it in R.<a href="7.3-mod-07-glmm.html#fnref72" class="footnote-back">↩︎</a></p></li>
<li id="fn73"><p>Sometimes model 3 does not converge. If the random number seed is set to 7, only models 1, 2, and 4 will converge. (This was true on my machine. Your results may vary).<a href="7.3-mod-07-glmm.html#fnref73" class="footnote-back">↩︎</a></p></li>
<li id="fn74"><p>Nematode image: <span class="citation">Alberti et al. (<a href="#ref-alberti2011" role="doc-biblioref">2011</a>)</span>. Deer image: O. Deme (<a href="https://upload.wikimedia.org/wikipedia/commons/d/df/Red_deer.jpg" class="uri">https://upload.wikimedia.org/wikipedia/commons/d/df/Red_deer.jpg</a>).<a href="7.3-mod-07-glmm.html#fnref74" class="footnote-back">↩︎</a></p></li>
<li id="fn75"><p>I couldn’t get it to work at all.<a href="7.4-mod-07-nlme.html#fnref75" class="footnote-back">↩︎</a></p></li>
<li id="fn76"><p>OpenBUGS and WinBUGS are two alternative programs that R can call to conduct MCMC sampling. WinBUGS is older and closed-source; OpenBUGS is newer and open-source. The modeling languages of JAGS, WinBUGS, and OpenBUGS are mostly identical. I learned Bayesian modeling with WinBUGS, but since 2020 have switched to JAGS. In my experience, JAGS is faster and more stable, while OpenBUGS is slower and less stable. Both are far preferable to WinBUGS<a href="7.4-mod-07-nlme.html#fnref76" class="footnote-back">↩︎</a></p></li>
<li id="fn77"><p>Going from additive models to generalized additive models is similar to going from linear models to generalized linear models.<a href="7.5-mod-07-gamm.html#fnref77" class="footnote-back">↩︎</a></p></li>
<li id="fn78"><p>Because position is expressed in latitude and longitude, it is implicit that these are positions on a sphere. Euclidean distance between coordinates is assumes that the coordinates are on a plane, and so will be distorted. The distances are also in nonsensical units that combine degrees latitude and longitude, which are not interchangeable. If you ever need to calculate exact distances between geographic locations, transform the coordinates into a projected coordinate system like UTM (which will be in meters, by definition). Or, use the GIS functions in R which can work in spherical coordinates. Despite being incorrect from a GIS perspective, the distances calculated here are close enough for demonstrating NMDS.<a href="8.5-mod-08-ord.html#fnref78" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
<h3>Literature Cited</h3>
<div id="refs" class="references csl-bib-body hanging-indent" line-spacing="2">
<div id="ref-alberti2011" class="csl-entry">
Alberti, E., G. Gioia, G. Sironi, S. Zanzani, P. Riccaboni, M. Magrini, and M. Manfredi. 2011. <a href="https://doi.org/10.1017/S0022149X10000647"><em>Elaphostrongylus cervi</em> in a population of red deer (<em><span>C</span>ervus elaphus</em>) and evidence of cerebrospinal nematodiasis in small ruminants in the province of <span>V</span>arese, <span>I</span>taly</a>. Journal of <span>H</span>elminthology 85:313–318.
</div>
<div id="ref-bolker2008ecological" class="csl-entry">
Bolker, B. M. 2008. Ecological models and data in <span>R</span>. Princeton University Press, Princeton, New Jersey, USA.
</div>
<div id="ref-bolker2009" class="csl-entry">
Bolker, B. M., M. E. Brooks, C. J. Clark, S. W. Geange, J. R. Poulsen, M. H. H. Stevens, and J.-S. S. White. 2009. <a href="https://doi.org/10.1016/j.tree.2008.10.008">Generalized linear mixed models: A practical guide for ecology and evolution</a>. Trends in Ecology &amp; Evolution 24:127–135.
</div>
<div id="ref-desolla2006" class="csl-entry">
De Solla, S. R., P. A. Martin, K. J. Fernie, B. J. Park, and G. Mayne. 2006. <a href="https://doi.org/10.1897/05-165R.1">Effects of environmentally relevant concentrations of atrazine on gonadal development of snapping turtles (chelydra serpentina)</a>. Environmental Toxicology and Chemistry: An International Journal 25:520–526.
</div>
<div id="ref-kery2010" class="csl-entry">
Kéry, M. 2010. Introduction to <span>W</span>in<span>BUGS</span> for ecologists: <span>B</span>ayesian approach to regression, <span>ANOVA</span>, mixed models and related analyses. Academic Press.
</div>
<div id="ref-link2002" class="csl-entry">
Link, W. A., E. Cam, J. D. Nichols, and E. G. Cooch. 2002. <a href="https://doi.org/10.2307/3803160">Of <span>BUGS</span> and birds: <span>M</span>arkov chain <span>M</span>onte <span>C</span>arlo for hierarchical modeling in wildlife research</a>. The Journal of <span>W</span>ildlife <span>M</span>anagement:277–291.
</div>
<div id="ref-mccarthy2007" class="csl-entry">
McCarthy, M. A. 2007. Bayesian methods for ecology. Cambridge University Press, Cambridge, United Kingdom.
</div>
<div id="ref-zuur2009mixed" class="csl-entry">
Zuur, A. F., E. N. Ieno, N. J. Walker, A. A. Saveliev, G. M. Smith, and others. 2009a. Mixed effects models and extensions in ecology with <span>R</span>. Springer, New York.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="9.6-conclusions.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
