<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Literature Cited | Applied Biological Data Analysis: Statistics and R for Biologists</title>
  <meta name="description" content="Helping biologists to become more informed users of R and statistical methods." />
  <meta name="generator" content="bookdown 0.28 and GitBook 2.6.7" />

  <meta property="og:title" content="Literature Cited | Applied Biological Data Analysis: Statistics and R for Biologists" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://greenquanteco.github.io/book/lab_logo_02.jpg" />
  <meta property="og:description" content="Helping biologists to become more informed users of R and statistical methods." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Literature Cited | Applied Biological Data Analysis: Statistics and R for Biologists" />
  
  <meta name="twitter:description" content="Helping biologists to become more informed users of R and statistical methods." />
  <meta name="twitter:image" content="https://greenquanteco.github.io/book/lab_logo_02.jpg" />

<meta name="author" content="Nick Green, Kennesaw State University" />


<meta name="date" content="2022-08-16" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="fav.ico" type="image/x-icon" />
<link rel="prev" href="mod-12.html"/>

<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Applied biological data analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#license-and-permissions"><i class="fa fa-check"></i>License and permissions</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-description"><i class="fa fa-check"></i>Course description</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-objectives"><i class="fa fa-check"></i>Course objectives</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-requirements"><i class="fa fa-check"></i>Course requirements</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#recommended-reading"><i class="fa fa-check"></i>Recommended reading</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-organization"><i class="fa fa-check"></i>Course organization</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#about-the-author"><i class="fa fa-check"></i>About the author</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="mod-01.html"><a href="mod-01.html"><i class="fa fa-check"></i><b>1</b> Statistics in modern biology</a>
<ul>
<li class="chapter" data-level="1.1" data-path="mod-01.html"><a href="mod-01.html#overview"><i class="fa fa-check"></i><b>1.1</b> Overview</a></li>
<li class="chapter" data-level="1.2" data-path="mod-01.html"><a href="mod-01.html#statistics-in-modern-biology"><i class="fa fa-check"></i><b>1.2</b> Statistics in modern biology</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="mod-01.html"><a href="mod-01.html#the-scientific-method"><i class="fa fa-check"></i><b>1.2.1</b> The scientific method</a></li>
<li class="chapter" data-level="1.2.2" data-path="mod-01.html"><a href="mod-01.html#example-data-analysis"><i class="fa fa-check"></i><b>1.2.2</b> Example data analysis</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="mod-01.html"><a href="mod-01.html#misuses-of-statistics"><i class="fa fa-check"></i><b>1.3</b> Misuses of statistics</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="mod-01.html"><a href="mod-01.html#proving-the-trivial-and-meaningless-hypotheses"><i class="fa fa-check"></i><b>1.3.1</b> “Proving” the trivial and meaningless hypotheses</a></li>
<li class="chapter" data-level="1.3.2" data-path="mod-01.html"><a href="mod-01.html#inappropriate-methods"><i class="fa fa-check"></i><b>1.3.2</b> Inappropriate methods</a></li>
<li class="chapter" data-level="1.3.3" data-path="mod-01.html"><a href="mod-01.html#p-hacking-and-data-dredging"><i class="fa fa-check"></i><b>1.3.3</b> <em>P</em>-hacking and data dredging</a></li>
<li class="chapter" data-level="1.3.4" data-path="mod-01.html"><a href="mod-01.html#inadequate-sample-sizes-and-pseudoreplication"><i class="fa fa-check"></i><b>1.3.4</b> Inadequate sample sizes and pseudoreplication</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="mod-01.html"><a href="mod-01.html#mod-01-nhst"><i class="fa fa-check"></i><b>1.4</b> <em>P</em>-values and null hypothesis significance testing (NHST)</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="mod-01.html"><a href="mod-01.html#definition"><i class="fa fa-check"></i><b>1.4.1</b> Definition</a></li>
<li class="chapter" data-level="1.4.2" data-path="mod-01.html"><a href="mod-01.html#aside-another-way-to-frame-null-hypotheses"><i class="fa fa-check"></i><b>1.4.2</b> Aside: Another way to frame null hypotheses</a></li>
<li class="chapter" data-level="1.4.3" data-path="mod-01.html"><a href="mod-01.html#history-and-status-of-p-values"><i class="fa fa-check"></i><b>1.4.3</b> History and status of <em>P</em>-values</a></li>
<li class="chapter" data-level="1.4.4" data-path="mod-01.html"><a href="mod-01.html#where-p-values-come-from"><i class="fa fa-check"></i><b>1.4.4</b> Where <em>P</em>-values come from</a></li>
<li class="chapter" data-level="1.4.5" data-path="mod-01.html"><a href="mod-01.html#what-p-values-mean-and-do-not-mean"><i class="fa fa-check"></i><b>1.4.5</b> What <em>P</em>-values mean and do not mean</a></li>
<li class="chapter" data-level="1.4.6" data-path="mod-01.html"><a href="mod-01.html#do-you-need-a-p-value"><i class="fa fa-check"></i><b>1.4.6</b> Do you need a <em>P</em>-value?</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="mod-01.html"><a href="mod-01.html#alternatives-to-nhst"><i class="fa fa-check"></i><b>1.5</b> Alternatives to NHST</a>
<ul>
<li class="chapter" data-level="1.5.1" data-path="mod-01.html"><a href="mod-01.html#mod-01-bayes"><i class="fa fa-check"></i><b>1.5.1</b> Bayesian inference</a></li>
<li class="chapter" data-level="1.5.2" data-path="mod-01.html"><a href="mod-01.html#information-theoretic-methods"><i class="fa fa-check"></i><b>1.5.2</b> Information-theoretic methods</a></li>
<li class="chapter" data-level="1.5.3" data-path="mod-01.html"><a href="mod-01.html#machine-learning"><i class="fa fa-check"></i><b>1.5.3</b> Machine learning</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="mod-02.html"><a href="mod-02.html"><i class="fa fa-check"></i><b>2</b> Introduction to R</a>
<ul>
<li class="chapter" data-level="2.1" data-path="mod-02.html"><a href="mod-02.html#getting-started-with-r"><i class="fa fa-check"></i><b>2.1</b> Getting started with R</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="mod-02.html"><a href="mod-02.html#what-is-r"><i class="fa fa-check"></i><b>2.1.1</b> What is R?</a></li>
<li class="chapter" data-level="2.1.2" data-path="mod-02.html"><a href="mod-02.html#advantages-of-r"><i class="fa fa-check"></i><b>2.1.2</b> Advantages of R</a></li>
<li class="chapter" data-level="2.1.3" data-path="mod-02.html"><a href="mod-02.html#disadvantages-of-r"><i class="fa fa-check"></i><b>2.1.3</b> Disadvantages of R</a></li>
<li class="chapter" data-level="2.1.4" data-path="mod-02.html"><a href="mod-02.html#base-r-and-vs.-tidyverse"><i class="fa fa-check"></i><b>2.1.4</b> Base R and (vs.?) <code>tidyverse</code></a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="mod-02.html"><a href="mod-02.html#download-and-install-r-and-rstudio"><i class="fa fa-check"></i><b>2.2</b> Download and install R (and RStudio)</a></li>
<li class="chapter" data-level="2.3" data-path="mod-02.html"><a href="mod-02.html#using-r"><i class="fa fa-check"></i><b>2.3</b> Using R</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="mod-02.html"><a href="mod-02.html#using-the-base-r-gui"><i class="fa fa-check"></i><b>2.3.1</b> Using the base R GUI</a></li>
<li class="chapter" data-level="2.3.2" data-path="mod-02.html"><a href="mod-02.html#using-r-in-rstudio"><i class="fa fa-check"></i><b>2.3.2</b> Using R in RStudio</a></li>
<li class="chapter" data-level="2.3.3" data-path="mod-02.html"><a href="mod-02.html#using-r-with-other-programs"><i class="fa fa-check"></i><b>2.3.3</b> Using R with other programs</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="mod-02.html"><a href="mod-02.html#mod-02-first"><i class="fa fa-check"></i><b>2.4</b> A first R session</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="mod-02.html"><a href="mod-02.html#import-data"><i class="fa fa-check"></i><b>2.4.1</b> Import data</a></li>
<li class="chapter" data-level="2.4.2" data-path="mod-02.html"><a href="mod-02.html#explore-and-visualize-data"><i class="fa fa-check"></i><b>2.4.2</b> Explore and visualize data</a></li>
<li class="chapter" data-level="2.4.3" data-path="mod-02.html"><a href="mod-02.html#transform-data"><i class="fa fa-check"></i><b>2.4.3</b> Transform data</a></li>
<li class="chapter" data-level="2.4.4" data-path="mod-02.html"><a href="mod-02.html#analyze-data"><i class="fa fa-check"></i><b>2.4.4</b> Analyze data</a></li>
<li class="chapter" data-level="2.4.5" data-path="mod-02.html"><a href="mod-02.html#mod-01-elkfig"><i class="fa fa-check"></i><b>2.4.5</b> Write out results</a></li>
<li class="chapter" data-level="2.4.6" data-path="mod-02.html"><a href="mod-02.html#save-your-work"><i class="fa fa-check"></i><b>2.4.6</b> Save your work?</a></li>
<li class="chapter" data-level="2.4.7" data-path="mod-02.html"><a href="mod-02.html#whats-next"><i class="fa fa-check"></i><b>2.4.7</b> What’s next?</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="mod-02.html"><a href="mod-02.html#write-and-execute-commands-in-the-r-console"><i class="fa fa-check"></i><b>2.5</b> Write and execute commands in the R console</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="mod-02.html"><a href="mod-02.html#r-commandsbasics"><i class="fa fa-check"></i><b>2.5.1</b> R commands–basics</a></li>
<li class="chapter" data-level="2.5.2" data-path="mod-02.html"><a href="mod-02.html#elements-of-r-code"><i class="fa fa-check"></i><b>2.5.2</b> Elements of R code</a></li>
<li class="chapter" data-level="2.5.3" data-path="mod-02.html"><a href="mod-02.html#the-r-workspace"><i class="fa fa-check"></i><b>2.5.3</b> The R workspace</a></li>
<li class="chapter" data-level="2.5.4" data-path="mod-02.html"><a href="mod-02.html#r-code-basics-assignment-and-operators"><i class="fa fa-check"></i><b>2.5.4</b> R code basics: assignment and operators</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="mod-02.html"><a href="mod-02.html#mod-02-struct"><i class="fa fa-check"></i><b>2.6</b> Basic R data structures</a>
<ul>
<li class="chapter" data-level="2.6.1" data-path="mod-02.html"><a href="mod-02.html#vectors"><i class="fa fa-check"></i><b>2.6.1</b> Vectors</a></li>
<li class="chapter" data-level="2.6.2" data-path="mod-02.html"><a href="mod-02.html#data-frames"><i class="fa fa-check"></i><b>2.6.2</b> Data frames</a></li>
<li class="chapter" data-level="2.6.3" data-path="mod-02.html"><a href="mod-02.html#matrices-and-arrays"><i class="fa fa-check"></i><b>2.6.3</b> Matrices and arrays</a></li>
<li class="chapter" data-level="2.6.4" data-path="mod-02.html"><a href="mod-02.html#lists"><i class="fa fa-check"></i><b>2.6.4</b> Lists</a></li>
<li class="chapter" data-level="2.6.5" data-path="mod-02.html"><a href="mod-02.html#s3-s4-and-other-paradigms"><i class="fa fa-check"></i><b>2.6.5</b> S3, S4, and other paradigms</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="mod-02.html"><a href="mod-02.html#r-data-types"><i class="fa fa-check"></i><b>2.7</b> R data types</a>
<ul>
<li class="chapter" data-level="2.7.1" data-path="mod-02.html"><a href="mod-02.html#character-type"><i class="fa fa-check"></i><b>2.7.1</b> Character type</a></li>
<li class="chapter" data-level="2.7.2" data-path="mod-02.html"><a href="mod-02.html#numeric-type"><i class="fa fa-check"></i><b>2.7.2</b> Numeric type</a></li>
<li class="chapter" data-level="2.7.3" data-path="mod-02.html"><a href="mod-02.html#integer-type"><i class="fa fa-check"></i><b>2.7.3</b> Integer type</a></li>
<li class="chapter" data-level="2.7.4" data-path="mod-02.html"><a href="mod-02.html#logical-type"><i class="fa fa-check"></i><b>2.7.4</b> Logical type</a></li>
<li class="chapter" data-level="2.7.5" data-path="mod-02.html"><a href="mod-02.html#special-values"><i class="fa fa-check"></i><b>2.7.5</b> Special values</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="mod-02.html"><a href="mod-02.html#manage-r-code-as-scripts-.r-files"><i class="fa fa-check"></i><b>2.8</b> Manage R code as scripts (.r files)</a></li>
<li class="chapter" data-level="2.9" data-path="mod-02.html"><a href="mod-02.html#manage-and-use-r-packages"><i class="fa fa-check"></i><b>2.9</b> Manage and use R packages</a>
<ul>
<li class="chapter" data-level="2.9.1" data-path="mod-02.html"><a href="mod-02.html#your-r-library"><i class="fa fa-check"></i><b>2.9.1</b> Your R library</a></li>
<li class="chapter" data-level="2.9.2" data-path="mod-02.html"><a href="mod-02.html#installing-packages-using-the-r-gui"><i class="fa fa-check"></i><b>2.9.2</b> Installing packages using the R GUI</a></li>
<li class="chapter" data-level="2.9.3" data-path="mod-02.html"><a href="mod-02.html#installing-packages-in-rstudio"><i class="fa fa-check"></i><b>2.9.3</b> Installing packages in RStudio</a></li>
<li class="chapter" data-level="2.9.4" data-path="mod-02.html"><a href="mod-02.html#installing-packages-using-the-r-console"><i class="fa fa-check"></i><b>2.9.4</b> Installing packages using the R console</a></li>
<li class="chapter" data-level="2.9.5" data-path="mod-02.html"><a href="mod-02.html#working-with-packages-in-r"><i class="fa fa-check"></i><b>2.9.5</b> Working with packages in R</a></li>
<li class="chapter" data-level="2.9.6" data-path="mod-02.html"><a href="mod-02.html#package-dependencies"><i class="fa fa-check"></i><b>2.9.6</b> Package dependencies</a></li>
<li class="chapter" data-level="2.9.7" data-path="mod-02.html"><a href="mod-02.html#citing-packages"><i class="fa fa-check"></i><b>2.9.7</b> Citing packages</a></li>
</ul></li>
<li class="chapter" data-level="2.10" data-path="mod-02.html"><a href="mod-02.html#r-documentation"><i class="fa fa-check"></i><b>2.10</b> R documentation</a>
<ul>
<li class="chapter" data-level="2.10.1" data-path="mod-02.html"><a href="mod-02.html#documentation-help-files"><i class="fa fa-check"></i><b>2.10.1</b> Documentation (help) files</a></li>
<li class="chapter" data-level="2.10.2" data-path="mod-02.html"><a href="mod-02.html#r-vignettes"><i class="fa fa-check"></i><b>2.10.2</b> R vignettes</a></li>
<li class="chapter" data-level="2.10.3" data-path="mod-02.html"><a href="mod-02.html#official-r-project-resources"><i class="fa fa-check"></i><b>2.10.3</b> Official R Project resources</a></li>
<li class="chapter" data-level="2.10.4" data-path="mod-02.html"><a href="mod-02.html#unofficial-online-resources"><i class="fa fa-check"></i><b>2.10.4</b> Unofficial online resources</a></li>
<li class="chapter" data-level="2.10.5" data-path="mod-02.html"><a href="mod-02.html#r-books"><i class="fa fa-check"></i><b>2.10.5</b> R books</a></li>
<li class="chapter" data-level="2.10.6" data-path="mod-02.html"><a href="mod-02.html#two-reminders"><i class="fa fa-check"></i><b>2.10.6</b> Two reminders</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="mod-03.html"><a href="mod-03.html"><i class="fa fa-check"></i><b>3</b> Data manipulation with R</a>
<ul>
<li class="chapter" data-level="3.1" data-path="mod-03.html"><a href="mod-03.html#mod-03-inout"><i class="fa fa-check"></i><b>3.1</b> Data import and export</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="mod-03.html"><a href="mod-03.html#importing-data-preliminaries"><i class="fa fa-check"></i><b>3.1.1</b> Importing data: preliminaries</a></li>
<li class="chapter" data-level="3.1.2" data-path="mod-03.html"><a href="mod-03.html#importing-data-from-text-files-with-read.csv-and-read.table"><i class="fa fa-check"></i><b>3.1.2</b> Importing data from text files with <code>read.csv()</code> and <code>read.table()</code></a></li>
<li class="chapter" data-level="3.1.3" data-path="mod-03.html"><a href="mod-03.html#importing-data-from-saved-workspaces"><i class="fa fa-check"></i><b>3.1.3</b> Importing data from saved workspaces</a></li>
<li class="chapter" data-level="3.1.4" data-path="mod-03.html"><a href="mod-03.html#importing-data-special-cases"><i class="fa fa-check"></i><b>3.1.4</b> Importing data: special cases:</a></li>
<li class="chapter" data-level="3.1.5" data-path="mod-03.html"><a href="mod-03.html#export-data-from-r"><i class="fa fa-check"></i><b>3.1.5</b> Export data from R</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="mod-03.html"><a href="mod-03.html#making-values-in-r"><i class="fa fa-check"></i><b>3.2</b> Making values in R</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="mod-03.html"><a href="mod-03.html#producing-arbitrary-values-with-c"><i class="fa fa-check"></i><b>3.2.1</b> Producing arbitrary values with <code>c()</code></a></li>
<li class="chapter" data-level="3.2.2" data-path="mod-03.html"><a href="mod-03.html#generating-regular-values"><i class="fa fa-check"></i><b>3.2.2</b> Generating regular values</a></li>
<li class="chapter" data-level="3.2.3" data-path="mod-03.html"><a href="mod-03.html#generating-random-values"><i class="fa fa-check"></i><b>3.2.3</b> Generating random values</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="mod-03.html"><a href="mod-03.html#selecting-data-with"><i class="fa fa-check"></i><b>3.3</b> Selecting data with <code>[]</code></a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="mod-03.html"><a href="mod-03.html#mod-03-brackets"><i class="fa fa-check"></i><b>3.3.1</b> Basics of brackets</a></li>
<li class="chapter" data-level="3.3.2" data-path="mod-03.html"><a href="mod-03.html#extracting-and-selecting-data-with-logical-tests"><i class="fa fa-check"></i><b>3.3.2</b> Extracting and selecting data with logical tests</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="mod-03.html"><a href="mod-03.html#managing-dates-and-characters"><i class="fa fa-check"></i><b>3.4</b> Managing dates and characters</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="mod-03.html"><a href="mod-03.html#temporal-data-and-dates"><i class="fa fa-check"></i><b>3.4.1</b> Temporal data and dates</a></li>
<li class="chapter" data-level="3.4.2" data-path="mod-03.html"><a href="mod-03.html#character-data-text"><i class="fa fa-check"></i><b>3.4.2</b> Character data (text)</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="mod-03.html"><a href="mod-03.html#mod-03-dataframe"><i class="fa fa-check"></i><b>3.5</b> Data frame management</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="mod-03.html"><a href="mod-03.html#data-frame-structure"><i class="fa fa-check"></i><b>3.5.1</b> Data frame structure</a></li>
<li class="chapter" data-level="3.5.2" data-path="mod-03.html"><a href="mod-03.html#common-data-frame-operations"><i class="fa fa-check"></i><b>3.5.2</b> Common data frame operations</a></li>
<li class="chapter" data-level="3.5.3" data-path="mod-03.html"><a href="mod-03.html#other-data-frame-operations"><i class="fa fa-check"></i><b>3.5.3</b> Other data frame operations</a></li>
<li class="chapter" data-level="3.5.4" data-path="mod-03.html"><a href="mod-03.html#mod-03-reshape"><i class="fa fa-check"></i><b>3.5.4</b> Reshaping data frames</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="mod-04.html"><a href="mod-04.html"><i class="fa fa-check"></i><b>4</b> Exploratory data analysis 1: Data description and summarization</a>
<ul>
<li class="chapter" data-level="4.1" data-path="mod-04.html"><a href="mod-04.html#motivation"><i class="fa fa-check"></i><b>4.1</b> Motivation</a></li>
<li class="chapter" data-level="4.2" data-path="mod-04.html"><a href="mod-04.html#descriptive-and-summary-statistics"><i class="fa fa-check"></i><b>4.2</b> Descriptive and summary statistics</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="mod-04.html"><a href="mod-04.html#basic-summary-statistics"><i class="fa fa-check"></i><b>4.2.1</b> Basic summary statistics</a></li>
<li class="chapter" data-level="4.2.2" data-path="mod-04.html"><a href="mod-04.html#mod-04-agg"><i class="fa fa-check"></i><b>4.2.2</b> Summarizing by groups (AKA: pivot tables)</a></li>
<li class="chapter" data-level="4.2.3" data-path="mod-04.html"><a href="mod-04.html#summarizing-data-with-the-apply-family"><i class="fa fa-check"></i><b>4.2.3</b> Summarizing data with the <code>apply()</code> family</a></li>
<li class="chapter" data-level="4.2.4" data-path="mod-04.html"><a href="mod-04.html#mod-04-tab"><i class="fa fa-check"></i><b>4.2.4</b> Tabulation (frequency tables)</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="mod-04.html"><a href="mod-04.html#mod-04-vis"><i class="fa fa-check"></i><b>4.3</b> Visualizing single variables</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="mod-04.html"><a href="mod-04.html#boxplots-aka-box-and-whisker-plots"><i class="fa fa-check"></i><b>4.3.1</b> Boxplots (aka: box-and-whisker plots)</a></li>
<li class="chapter" data-level="4.3.2" data-path="mod-04.html"><a href="mod-04.html#histograms"><i class="fa fa-check"></i><b>4.3.2</b> Histograms</a></li>
<li class="chapter" data-level="4.3.3" data-path="mod-04.html"><a href="mod-04.html#mod04-pdfplots"><i class="fa fa-check"></i><b>4.3.3</b> Kernel density plots</a></li>
<li class="chapter" data-level="4.3.4" data-path="mod-04.html"><a href="mod-04.html#empirical-cumulative-density-function-ecdf-plots"><i class="fa fa-check"></i><b>4.3.4</b> Empirical cumulative density function (ECDF) plots</a></li>
<li class="chapter" data-level="4.3.5" data-path="mod-04.html"><a href="mod-04.html#quantile-quantile-qq-plots"><i class="fa fa-check"></i><b>4.3.5</b> Quantile-quantile (QQ) plots</a></li>
<li class="chapter" data-level="4.3.6" data-path="mod-04.html"><a href="mod-04.html#how-should-i-plot-my-data"><i class="fa fa-check"></i><b>4.3.6</b> How should I plot my data?</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="mod-05.html"><a href="mod-05.html"><i class="fa fa-check"></i><b>5</b> Exploratory data analysis 2: Probability distributions for biology</a>
<ul>
<li class="chapter" data-level="5.1" data-path="mod-05.html"><a href="mod-05.html#mod-05-dists1"><i class="fa fa-check"></i><b>5.1</b> Statistical distributions</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="mod-05.html"><a href="mod-05.html#probability-distributions-in-r"><i class="fa fa-check"></i><b>5.1.1</b> Probability distributions in R</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="mod-05.html"><a href="mod-05.html#mod-05-discrete"><i class="fa fa-check"></i><b>5.2</b> Discrete distributions</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="mod-05.html"><a href="mod-05.html#bernoulli-distribution"><i class="fa fa-check"></i><b>5.2.1</b> Bernoulli distribution</a></li>
<li class="chapter" data-level="5.2.2" data-path="mod-05.html"><a href="mod-05.html#binomial-distribution"><i class="fa fa-check"></i><b>5.2.2</b> Binomial distribution</a></li>
<li class="chapter" data-level="5.2.3" data-path="mod-05.html"><a href="mod-05.html#poisson-distribution"><i class="fa fa-check"></i><b>5.2.3</b> Poisson distribution</a></li>
<li class="chapter" data-level="5.2.4" data-path="mod-05.html"><a href="mod-05.html#negative-binomial-distribution"><i class="fa fa-check"></i><b>5.2.4</b> Negative binomial distribution</a></li>
<li class="chapter" data-level="5.2.5" data-path="mod-05.html"><a href="mod-05.html#geometric-distribution"><i class="fa fa-check"></i><b>5.2.5</b> Geometric distribution</a></li>
<li class="chapter" data-level="5.2.6" data-path="mod-05.html"><a href="mod-05.html#beta-binomial-distribution"><i class="fa fa-check"></i><b>5.2.6</b> Beta-binomial distribution</a></li>
<li class="chapter" data-level="5.2.7" data-path="mod-05.html"><a href="mod-05.html#multinomial-distribution"><i class="fa fa-check"></i><b>5.2.7</b> Multinomial distribution</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="mod-05.html"><a href="mod-05.html#mod-05-condists"><i class="fa fa-check"></i><b>5.3</b> Continuous distributions</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="mod-05.html"><a href="mod-05.html#uniform-distribution"><i class="fa fa-check"></i><b>5.3.1</b> Uniform distribution</a></li>
<li class="chapter" data-level="5.3.2" data-path="mod-05.html"><a href="mod-05.html#mod-05-norm"><i class="fa fa-check"></i><b>5.3.2</b> Normal distribution</a></li>
<li class="chapter" data-level="5.3.3" data-path="mod-05.html"><a href="mod-05.html#lognormal-distribution"><i class="fa fa-check"></i><b>5.3.3</b> Lognormal distribution</a></li>
<li class="chapter" data-level="5.3.4" data-path="mod-05.html"><a href="mod-05.html#gamma-distribution"><i class="fa fa-check"></i><b>5.3.4</b> Gamma distribution</a></li>
<li class="chapter" data-level="5.3.5" data-path="mod-05.html"><a href="mod-05.html#beta-distribution"><i class="fa fa-check"></i><b>5.3.5</b> Beta distribution</a></li>
<li class="chapter" data-level="5.3.6" data-path="mod-05.html"><a href="mod-05.html#dirichlet-distribution"><i class="fa fa-check"></i><b>5.3.6</b> Dirichlet distribution</a></li>
<li class="chapter" data-level="5.3.7" data-path="mod-05.html"><a href="mod-05.html#exponential-distribution"><i class="fa fa-check"></i><b>5.3.7</b> Exponential distribution</a></li>
<li class="chapter" data-level="5.3.8" data-path="mod-05.html"><a href="mod-05.html#triangular-distribution"><i class="fa fa-check"></i><b>5.3.8</b> Triangular distribution</a></li>
<li class="chapter" data-level="5.3.9" data-path="mod-05.html"><a href="mod-05.html#distributions-summary"><i class="fa fa-check"></i><b>5.3.9</b> Distributions summary</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="mod-05.html"><a href="mod-05.html#mod-05-dists2"><i class="fa fa-check"></i><b>5.4</b> Fitting and testing distributions</a>
<ul>
<li class="chapter" data-level="5.4.1" data-path="mod-05.html"><a href="mod-05.html#estimating-distributional-parameters"><i class="fa fa-check"></i><b>5.4.1</b> Estimating distributional parameters</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="mod-05.html"><a href="mod-05.html#graphical-methods-for-examining-distributions"><i class="fa fa-check"></i><b>5.5</b> Graphical methods for examining distributions</a>
<ul>
<li class="chapter" data-level="5.5.1" data-path="mod-05.html"><a href="mod-05.html#comparing-distributions-with-pdf-or-pmf-plots"><i class="fa fa-check"></i><b>5.5.1</b> Comparing distributions with PDF or PMF plots</a></li>
<li class="chapter" data-level="5.5.2" data-path="mod-05.html"><a href="mod-05.html#comparing-distributions-using-cdf-and-ecdf"><i class="fa fa-check"></i><b>5.5.2</b> Comparing distributions using CDF and ECDF</a></li>
<li class="chapter" data-level="5.5.3" data-path="mod-05.html"><a href="mod-05.html#comparing-distributions-using-qq-plots"><i class="fa fa-check"></i><b>5.5.3</b> Comparing distributions using QQ plots</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="mod-05.html"><a href="mod-05.html#formal-tests-for-distributions"><i class="fa fa-check"></i><b>5.6</b> Formal tests for distributions</a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="mod-05.html"><a href="mod-05.html#tests-for-normality"><i class="fa fa-check"></i><b>5.6.1</b> Tests for normality</a></li>
<li class="chapter" data-level="5.6.2" data-path="mod-05.html"><a href="mod-05.html#kolmogorov-smirnov-tests"><i class="fa fa-check"></i><b>5.6.2</b> Kolmogorov-Smirnov tests</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="mod-05.html"><a href="mod-05.html#mod-05-trans"><i class="fa fa-check"></i><b>5.7</b> Data transformations</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="mod-05.html"><a href="mod-05.html#why-transform"><i class="fa fa-check"></i><b>5.7.1</b> Why transform?</a></li>
<li class="chapter" data-level="5.7.2" data-path="mod-05.html"><a href="mod-05.html#transforms-vs.-link-functions"><i class="fa fa-check"></i><b>5.7.2</b> Transforms vs. link functions</a></li>
<li class="chapter" data-level="5.7.3" data-path="mod-05.html"><a href="mod-05.html#common-transformations"><i class="fa fa-check"></i><b>5.7.3</b> Common transformations</a></li>
<li class="chapter" data-level="5.7.4" data-path="mod-05.html"><a href="mod-05.html#uncommon-transformations"><i class="fa fa-check"></i><b>5.7.4</b> Uncommon transformations</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="mod-06.html"><a href="mod-06.html"><i class="fa fa-check"></i><b>6</b> Exploratory data analysis 3: Multivariate data exploration</a>
<ul>
<li class="chapter" data-level="6.1" data-path="mod-06.html"><a href="mod-06.html#scatterplots-for-2-continuous-variables"><i class="fa fa-check"></i><b>6.1</b> Scatterplots for 2 continuous variables</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="mod-06.html"><a href="mod-06.html#r-plotting-basics"><i class="fa fa-check"></i><b>6.1.1</b> R plotting basics</a></li>
<li class="chapter" data-level="6.1.2" data-path="mod-06.html"><a href="mod-06.html#graphical-parameters-and-advanced-plots-with-par"><i class="fa fa-check"></i><b>6.1.2</b> Graphical parameters and advanced plots with <code>par()</code></a></li>
<li class="chapter" data-level="6.1.3" data-path="mod-06.html"><a href="mod-06.html#really-advanced-plotting"><i class="fa fa-check"></i><b>6.1.3</b> Really advanced plotting</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="mod-06.html"><a href="mod-06.html#mod-06-smat"><i class="fa fa-check"></i><b>6.2</b> Scatterplot matrices for many variables</a></li>
<li class="chapter" data-level="6.3" data-path="mod-06.html"><a href="mod-06.html#lattice-plots-for-hierarchical-data"><i class="fa fa-check"></i><b>6.3</b> Lattice plots for hierarchical data</a></li>
<li class="chapter" data-level="6.4" data-path="mod-06.html"><a href="mod-06.html#continuous-variables-by-group"><i class="fa fa-check"></i><b>6.4</b> Continuous variables by group</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="mod-06.html"><a href="mod-06.html#boxplots"><i class="fa fa-check"></i><b>6.4.1</b> Boxplots</a></li>
<li class="chapter" data-level="6.4.2" data-path="mod-06.html"><a href="mod-06.html#barplots"><i class="fa fa-check"></i><b>6.4.2</b> Barplots</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="mod-06.html"><a href="mod-06.html#ordination-brief-introduction"><i class="fa fa-check"></i><b>6.5</b> Ordination (brief introduction)</a>
<ul>
<li class="chapter" data-level="6.5.1" data-path="mod-06.html"><a href="mod-06.html#principal-components-analysis-pca"><i class="fa fa-check"></i><b>6.5.1</b> Principal components analysis (PCA)</a></li>
<li class="chapter" data-level="6.5.2" data-path="mod-06.html"><a href="mod-06.html#nonmetric-multidimensional-scaling-nmds"><i class="fa fa-check"></i><b>6.5.2</b> Nonmetric multidimensional scaling (NMDS)</a></li>
<li class="chapter" data-level="6.5.3" data-path="mod-06.html"><a href="mod-06.html#plotting-ordinations"><i class="fa fa-check"></i><b>6.5.3</b> Plotting ordinations</a></li>
<li class="chapter" data-level="6.5.4" data-path="mod-06.html"><a href="mod-06.html#ordination-wrap-up-for-now"><i class="fa fa-check"></i><b>6.5.4</b> Ordination wrap-up (for now)</a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="mod-06.html"><a href="mod-06.html#principles-for-good-data-graphics"><i class="fa fa-check"></i><b>6.6</b> Principles for good data graphics</a>
<ul>
<li class="chapter" data-level="6.6.1" data-path="mod-06.html"><a href="mod-06.html#maximize-the-data-to-ink-ratio"><i class="fa fa-check"></i><b>6.6.1</b> Maximize the data-to-ink ratio</a></li>
<li class="chapter" data-level="6.6.2" data-path="mod-06.html"><a href="mod-06.html#show-variability"><i class="fa fa-check"></i><b>6.6.2</b> Show variability</a></li>
<li class="chapter" data-level="6.6.3" data-path="mod-06.html"><a href="mod-06.html#use-labels-not-symbols"><i class="fa fa-check"></i><b>6.6.3</b> Use labels, not symbols</a></li>
<li class="chapter" data-level="6.6.4" data-path="mod-06.html"><a href="mod-06.html#be-consistent"><i class="fa fa-check"></i><b>6.6.4</b> Be consistent</a></li>
<li class="chapter" data-level="6.6.5" data-path="mod-06.html"><a href="mod-06.html#set-yourself-up-for-success"><i class="fa fa-check"></i><b>6.6.5</b> Set yourself up for success</a></li>
<li class="chapter" data-level="6.6.6" data-path="mod-06.html"><a href="mod-06.html#consider-legibility"><i class="fa fa-check"></i><b>6.6.6</b> Consider legibility</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="mod-07.html"><a href="mod-07.html"><i class="fa fa-check"></i><b>7</b> Common statistical problems</a>
<ul>
<li class="chapter" data-level="7.1" data-path="mod-07.html"><a href="mod-07.html#outliers-and-erroneous-values"><i class="fa fa-check"></i><b>7.1</b> Outliers and erroneous values</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="mod-07.html"><a href="mod-07.html#detecting-outliers"><i class="fa fa-check"></i><b>7.1.1</b> Detecting outliers</a></li>
<li class="chapter" data-level="7.1.2" data-path="mod-07.html"><a href="mod-07.html#dealing-with-outliers"><i class="fa fa-check"></i><b>7.1.2</b> Dealing with outliers</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="mod-07.html"><a href="mod-07.html#autocorrelation"><i class="fa fa-check"></i><b>7.2</b> Autocorrelation</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="mod-07.html"><a href="mod-07.html#graphical-methods-to-identify-autocorrelation"><i class="fa fa-check"></i><b>7.2.1</b> Graphical methods to identify autocorrelation</a></li>
<li class="chapter" data-level="7.2.2" data-path="mod-07.html"><a href="mod-07.html#mantel-tests-for-autocorrelation"><i class="fa fa-check"></i><b>7.2.2</b> Mantel tests for autocorrelation</a></li>
<li class="chapter" data-level="7.2.3" data-path="mod-07.html"><a href="mod-07.html#dealing-with-autocorrelation"><i class="fa fa-check"></i><b>7.2.3</b> Dealing with autocorrelation</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="mod-07.html"><a href="mod-07.html#mod-07-multicol"><i class="fa fa-check"></i><b>7.3</b> Collinearity</a></li>
<li class="chapter" data-level="7.4" data-path="mod-07.html"><a href="mod-07.html#missing-data"><i class="fa fa-check"></i><b>7.4</b> Missing data</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="mod-07.html"><a href="mod-07.html#option-1-ignore-missing-data"><i class="fa fa-check"></i><b>7.4.1</b> Option 1: Ignore missing data</a></li>
<li class="chapter" data-level="7.4.2" data-path="mod-07.html"><a href="mod-07.html#option-2-interpolate-or-impute-missing-data"><i class="fa fa-check"></i><b>7.4.2</b> Option 2: Interpolate or impute missing data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="mod-08.html"><a href="mod-08.html"><i class="fa fa-check"></i><b>8</b> Planning your analysis (what test?)</a>
<ul>
<li class="chapter" data-level="8.1" data-path="mod-08.html"><a href="mod-08.html#how-to-use-this-guide"><i class="fa fa-check"></i><b>8.1</b> How to use this guide</a></li>
<li class="chapter" data-level="8.2" data-path="mod-08.html"><a href="mod-08.html#what-question-are-you-trying-to-answer"><i class="fa fa-check"></i><b>8.2</b> What question are you trying to answer?</a></li>
<li class="chapter" data-level="8.3" data-path="mod-08.html"><a href="mod-08.html#testing-for-a-difference-in-mean-or-location"><i class="fa fa-check"></i><b>8.3</b> Testing for a difference in mean or location</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="mod-08.html"><a href="mod-08.html#additional-considerations"><i class="fa fa-check"></i><b>8.3.1</b> Additional considerations</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="mod-08.html"><a href="mod-08.html#testing-for-a-continuous-relationship-between-two-or-more-variables"><i class="fa fa-check"></i><b>8.4</b> Testing for a continuous relationship between two or more variables</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="mod-08.html"><a href="mod-08.html#additional-considerations-1"><i class="fa fa-check"></i><b>8.4.1</b> Additional considerations</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="mod-08.html"><a href="mod-08.html#independence-between-categories"><i class="fa fa-check"></i><b>8.5</b> Independence between categories</a></li>
<li class="chapter" data-level="8.6" data-path="mod-08.html"><a href="mod-08.html#classifying-observations"><i class="fa fa-check"></i><b>8.6</b> Classifying observations</a></li>
<li class="chapter" data-level="8.7" data-path="mod-08.html"><a href="mod-08.html#conclusions"><i class="fa fa-check"></i><b>8.7</b> Conclusions</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="mod-09.html"><a href="mod-09.html"><i class="fa fa-check"></i><b>9</b> Generalized linear models (GLM)</a>
<ul>
<li class="chapter" data-level="9.1" data-path="mod-09.html"><a href="mod-09.html#mod-09-lm"><i class="fa fa-check"></i><b>9.1</b> Prelude with linear models</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="mod-09.html"><a href="mod-09.html#assumptions-of-linear-models"><i class="fa fa-check"></i><b>9.1.1</b> Assumptions of linear models</a></li>
<li class="chapter" data-level="9.1.2" data-path="mod-09.html"><a href="mod-09.html#linear-regression-in-r"><i class="fa fa-check"></i><b>9.1.2</b> Linear regression in R</a></li>
<li class="chapter" data-level="9.1.3" data-path="mod-09.html"><a href="mod-09.html#multiple-linear-regression"><i class="fa fa-check"></i><b>9.1.3</b> Multiple linear regression</a></li>
<li class="chapter" data-level="9.1.4" data-path="mod-09.html"><a href="mod-09.html#mod-09-anova"><i class="fa fa-check"></i><b>9.1.4</b> ANOVA and ANCOVA with <code>lm()</code></a></li>
<li class="chapter" data-level="9.1.5" data-path="mod-09.html"><a href="mod-09.html#variations-on-linear-models"><i class="fa fa-check"></i><b>9.1.5</b> Variations on linear models</a></li>
<li class="chapter" data-level="9.1.6" data-path="mod-09.html"><a href="mod-09.html#example-linear-regression-workflow"><i class="fa fa-check"></i><b>9.1.6</b> Example linear regression workflow</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="mod-09.html"><a href="mod-09.html#mod-09-basic"><i class="fa fa-check"></i><b>9.2</b> GLM basics</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="mod-09.html"><a href="mod-09.html#example-glms"><i class="fa fa-check"></i><b>9.2.1</b> Example GLMS</a></li>
<li class="chapter" data-level="9.2.2" data-path="mod-09.html"><a href="mod-09.html#glm-families"><i class="fa fa-check"></i><b>9.2.2</b> GLM families</a></li>
<li class="chapter" data-level="9.2.3" data-path="mod-09.html"><a href="mod-09.html#glm-link-functions"><i class="fa fa-check"></i><b>9.2.3</b> GLM link functions</a></li>
<li class="chapter" data-level="9.2.4" data-path="mod-09.html"><a href="mod-09.html#deviance-and-other-glm-diagnostics"><i class="fa fa-check"></i><b>9.2.4</b> Deviance and other GLM diagnostics</a></li>
<li class="chapter" data-level="9.2.5" data-path="mod-09.html"><a href="mod-09.html#to-pseudo-r2-or-not-to-pseudo-r2"><i class="fa fa-check"></i><b>9.2.5</b> To pseudo-<em>R</em><sup>2</sup> or not to pseudo-<em>R</em><sup>2</sup>?</a></li>
<li class="chapter" data-level="9.2.6" data-path="mod-09.html"><a href="mod-09.html#common-glms"><i class="fa fa-check"></i><b>9.2.6</b> Common GLMs</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="mod-09.html"><a href="mod-09.html#mod-09-loglin"><i class="fa fa-check"></i><b>9.3</b> Log-linear models</a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="mod-09.html"><a href="mod-09.html#mod-09-loglin-examp"><i class="fa fa-check"></i><b>9.3.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="9.3.2" data-path="mod-09.html"><a href="mod-09.html#example-with-real-data"><i class="fa fa-check"></i><b>9.3.2</b> Example with real data</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="mod-09.html"><a href="mod-09.html#mod-09-poisson"><i class="fa fa-check"></i><b>9.4</b> Poisson GLM for counts</a>
<ul>
<li class="chapter" data-level="9.4.1" data-path="mod-09.html"><a href="mod-09.html#example-with-simulated-data"><i class="fa fa-check"></i><b>9.4.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="9.4.2" data-path="mod-09.html"><a href="mod-09.html#example-with-real-data-1"><i class="fa fa-check"></i><b>9.4.2</b> Example with real data</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="mod-09.html"><a href="mod-09.html#mod-09-quasi"><i class="fa fa-check"></i><b>9.5</b> Quasi-Poisson and negative binomial GLM</a>
<ul>
<li class="chapter" data-level="9.5.1" data-path="mod-09.html"><a href="mod-09.html#example-with-simulated-data-1"><i class="fa fa-check"></i><b>9.5.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="9.5.2" data-path="mod-09.html"><a href="mod-09.html#example-with-real-data-2"><i class="fa fa-check"></i><b>9.5.2</b> Example with real data</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="mod-09.html"><a href="mod-09.html#mod-09-logistic"><i class="fa fa-check"></i><b>9.6</b> Logistic regression for binary outcomes</a>
<ul>
<li class="chapter" data-level="9.6.1" data-path="mod-09.html"><a href="mod-09.html#example-with-simulated-data-2"><i class="fa fa-check"></i><b>9.6.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="9.6.2" data-path="mod-09.html"><a href="mod-09.html#example-with-real-data-3"><i class="fa fa-check"></i><b>9.6.2</b> Example with real data</a></li>
<li class="chapter" data-level="9.6.3" data-path="mod-09.html"><a href="mod-09.html#mod-09-auc"><i class="fa fa-check"></i><b>9.6.3</b> Logistic GLM diagnostics: AUC and ROC</a></li>
</ul></li>
<li class="chapter" data-level="9.7" data-path="mod-09.html"><a href="mod-09.html#mod-09-binom"><i class="fa fa-check"></i><b>9.7</b> Binomial GLM for proportional data</a>
<ul>
<li class="chapter" data-level="9.7.1" data-path="mod-09.html"><a href="mod-09.html#binomial-glm"><i class="fa fa-check"></i><b>9.7.1</b> Binomial GLM</a></li>
<li class="chapter" data-level="9.7.2" data-path="mod-09.html"><a href="mod-09.html#example-with-simulated-data-3"><i class="fa fa-check"></i><b>9.7.2</b> Example with simulated data</a></li>
</ul></li>
<li class="chapter" data-level="9.8" data-path="mod-09.html"><a href="mod-09.html#mod-09-gamma"><i class="fa fa-check"></i><b>9.8</b> Gamma models for overdispersed data</a>
<ul>
<li class="chapter" data-level="9.8.1" data-path="mod-09.html"><a href="mod-09.html#example-with-simulated-data-4"><i class="fa fa-check"></i><b>9.8.1</b> Example with simulated data</a></li>
</ul></li>
<li class="chapter" data-level="9.9" data-path="mod-09.html"><a href="mod-09.html#mod-09-beyond"><i class="fa fa-check"></i><b>9.9</b> Beyond GLM: Overview of GAM and GEE</a>
<ul>
<li class="chapter" data-level="9.9.1" data-path="mod-09.html"><a href="mod-09.html#mod-09-gam"><i class="fa fa-check"></i><b>9.9.1</b> Generalized additive models (GAM)</a></li>
<li class="chapter" data-level="9.9.2" data-path="mod-09.html"><a href="mod-09.html#mod-09-gee"><i class="fa fa-check"></i><b>9.9.2</b> Generalized estimating equations (GEE)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="mod-10.html"><a href="mod-10.html"><i class="fa fa-check"></i><b>10</b> Nonlinear models</a>
<ul>
<li class="chapter" data-level="10.1" data-path="mod-10.html"><a href="mod-10.html#background"><i class="fa fa-check"></i><b>10.1</b> Background</a></li>
<li class="chapter" data-level="10.2" data-path="mod-10.html"><a href="mod-10.html#mod-10-intro"><i class="fa fa-check"></i><b>10.2</b> Nonlinear least squares (NLS)</a></li>
<li class="chapter" data-level="10.3" data-path="mod-10.html"><a href="mod-10.html#mod-10-micmen"><i class="fa fa-check"></i><b>10.3</b> Michaelis-Menten curves</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="mod-10.html"><a href="mod-10.html#mod-10-micmen-sim"><i class="fa fa-check"></i><b>10.3.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="10.3.2" data-path="mod-10.html"><a href="mod-10.html#example-with-real-data-4"><i class="fa fa-check"></i><b>10.3.2</b> Example with real data</a></li>
<li class="chapter" data-level="10.3.3" data-path="mod-10.html"><a href="mod-10.html#alternative-strategies-for-the-analysis"><i class="fa fa-check"></i><b>10.3.3</b> Alternative strategies for the analysis</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="mod-10.html"><a href="mod-10.html#mod-10-grow"><i class="fa fa-check"></i><b>10.4</b> Biological growth curves</a>
<ul>
<li class="chapter" data-level="10.4.1" data-path="mod-10.html"><a href="mod-10.html#gompertz-and-von-bertalanffy-curves"><i class="fa fa-check"></i><b>10.4.1</b> Gompertz and von Bertalanffy curves</a></li>
<li class="chapter" data-level="10.4.2" data-path="mod-10.html"><a href="mod-10.html#example-with-real-data-5"><i class="fa fa-check"></i><b>10.4.2</b> Example with real data</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="mod-10.html"><a href="mod-10.html#dose-response-curves"><i class="fa fa-check"></i><b>10.5</b> Dose response curves</a></li>
<li class="chapter" data-level="10.6" data-path="mod-10.html"><a href="mod-10.html#alternatives-to-nls"><i class="fa fa-check"></i><b>10.6</b> Alternatives to NLS</a>
<ul>
<li class="chapter" data-level="10.6.1" data-path="mod-10.html"><a href="mod-10.html#generalized-nonlinear-models"><i class="fa fa-check"></i><b>10.6.1</b> Generalized nonlinear models</a></li>
<li class="chapter" data-level="10.6.2" data-path="mod-10.html"><a href="mod-10.html#quantile-regression"><i class="fa fa-check"></i><b>10.6.2</b> Quantile regression</a></li>
<li class="chapter" data-level="10.6.3" data-path="mod-10.html"><a href="mod-10.html#mod-10-gam"><i class="fa fa-check"></i><b>10.6.3</b> Generalized additive models (GAM)</a></li>
<li class="chapter" data-level="10.6.4" data-path="mod-10.html"><a href="mod-10.html#classification-and-regression-trees-cart"><i class="fa fa-check"></i><b>10.6.4</b> Classification and regression trees (CART)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="mod-11.html"><a href="mod-11.html"><i class="fa fa-check"></i><b>11</b> Mixed models</a>
<ul>
<li class="chapter" data-level="11.1" data-path="mod-11.html"><a href="mod-11.html#prelude-glm"><i class="fa fa-check"></i><b>11.1</b> Prelude (GLM)</a></li>
<li class="chapter" data-level="11.2" data-path="mod-11.html"><a href="mod-11.html#mod-11-lmm"><i class="fa fa-check"></i><b>11.2</b> Linear mixed models (LMM)</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="mod-11.html"><a href="mod-11.html#formal-definition-and-example"><i class="fa fa-check"></i><b>11.2.1</b> Formal definition and example</a></li>
<li class="chapter" data-level="11.2.2" data-path="mod-11.html"><a href="mod-11.html#example-with-simulated-data-5"><i class="fa fa-check"></i><b>11.2.2</b> Example with simulated data</a></li>
<li class="chapter" data-level="11.2.3" data-path="mod-11.html"><a href="mod-11.html#p-values-in-lmm"><i class="fa fa-check"></i><b>11.2.3</b> <em>P</em>-values in LMM</a></li>
<li class="chapter" data-level="11.2.4" data-path="mod-11.html"><a href="mod-11.html#specifying-random-effects"><i class="fa fa-check"></i><b>11.2.4</b> Specifying random effects</a></li>
<li class="chapter" data-level="11.2.5" data-path="mod-11.html"><a href="mod-11.html#lmm-example-with-real-data"><i class="fa fa-check"></i><b>11.2.5</b> LMM example with real data</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="mod-11.html"><a href="mod-11.html#mod-11-glmm"><i class="fa fa-check"></i><b>11.3</b> Generalized linear mixed models (GLMM)</a>
<ul>
<li class="chapter" data-level="11.3.1" data-path="mod-11.html"><a href="mod-11.html#definition-1"><i class="fa fa-check"></i><b>11.3.1</b> Definition</a></li>
<li class="chapter" data-level="11.3.2" data-path="mod-11.html"><a href="mod-11.html#glmm-on-simulated-data"><i class="fa fa-check"></i><b>11.3.2</b> GLMM on simulated data</a></li>
<li class="chapter" data-level="11.3.3" data-path="mod-11.html"><a href="mod-11.html#glmm-on-real-data"><i class="fa fa-check"></i><b>11.3.3</b> GLMM on real data</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="mod-11.html"><a href="mod-11.html#mod-11-nlme"><i class="fa fa-check"></i><b>11.4</b> Nonlinear mixed models (NLME)</a>
<ul>
<li class="chapter" data-level="11.4.1" data-path="mod-11.html"><a href="mod-11.html#definition-and-background"><i class="fa fa-check"></i><b>11.4.1</b> Definition and background</a></li>
<li class="chapter" data-level="11.4.2" data-path="mod-11.html"><a href="mod-11.html#nlme-on-simulated-data"><i class="fa fa-check"></i><b>11.4.2</b> NLME on simulated data</a></li>
<li class="chapter" data-level="11.4.3" data-path="mod-11.html"><a href="mod-11.html#nlme-on-real-data"><i class="fa fa-check"></i><b>11.4.3</b> NLME on real data</a></li>
</ul></li>
<li class="chapter" data-level="11.5" data-path="mod-11.html"><a href="mod-11.html#mod-11-gamm"><i class="fa fa-check"></i><b>11.5</b> (Generalized) additive mixed models (AMM/GAMM)</a>
<ul>
<li class="chapter" data-level="11.5.1" data-path="mod-11.html"><a href="mod-11.html#example-gamm-with-real-data"><i class="fa fa-check"></i><b>11.5.1</b> Example GAMM with real data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="12" data-path="mod-12.html"><a href="mod-12.html"><i class="fa fa-check"></i><b>12</b> Multivariate data analysis</a>
<ul>
<li class="chapter" data-level="12.1" data-path="mod-12.html"><a href="mod-12.html#multivariate-data"><i class="fa fa-check"></i><b>12.1</b> Multivariate data</a>
<ul>
<li class="chapter" data-level="12.1.1" data-path="mod-12.html"><a href="mod-12.html#univariate-vs.-multivariate-data"><i class="fa fa-check"></i><b>12.1.1</b> Univariate vs. multivariate data</a></li>
<li class="chapter" data-level="12.1.2" data-path="mod-12.html"><a href="mod-12.html#components-of-multivariate-data"><i class="fa fa-check"></i><b>12.1.2</b> Components of multivariate data</a></li>
</ul></li>
<li class="chapter" data-level="12.2" data-path="mod-12.html"><a href="mod-12.html#mod-12-dist"><i class="fa fa-check"></i><b>12.2</b> Distance metrics: biological (dis)similarity</a>
<ul>
<li class="chapter" data-level="12.2.1" data-path="mod-12.html"><a href="mod-12.html#euclidean-distance"><i class="fa fa-check"></i><b>12.2.1</b> Euclidean distance</a></li>
<li class="chapter" data-level="12.2.2" data-path="mod-12.html"><a href="mod-12.html#bray-curtis-and-other-distance-metrics"><i class="fa fa-check"></i><b>12.2.2</b> Bray-Curtis and other distance metrics</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="mod-12.html"><a href="mod-12.html#clustering"><i class="fa fa-check"></i><b>12.3</b> Clustering</a>
<ul>
<li class="chapter" data-level="12.3.1" data-path="mod-12.html"><a href="mod-12.html#k-means-clustering"><i class="fa fa-check"></i><b>12.3.1</b> <em>K</em>-means clustering</a></li>
<li class="chapter" data-level="12.3.2" data-path="mod-12.html"><a href="mod-12.html#hierarchical-agglomerative-clustering"><i class="fa fa-check"></i><b>12.3.2</b> Hierarchical agglomerative clustering</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="mod-12.html"><a href="mod-12.html#mod-12-sims"><i class="fa fa-check"></i><b>12.4</b> Analyzing dissimilarity</a>
<ul>
<li class="chapter" data-level="12.4.1" data-path="mod-12.html"><a href="mod-12.html#mantel-tests-distance-vs.-distance"><i class="fa fa-check"></i><b>12.4.1</b> Mantel tests: distance vs. distance</a></li>
<li class="chapter" data-level="12.4.2" data-path="mod-12.html"><a href="mod-12.html#comparing-dissimilarity-between-groups"><i class="fa fa-check"></i><b>12.4.2</b> Comparing dissimilarity between groups</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="mod-12.html"><a href="mod-12.html#mod-12-ord"><i class="fa fa-check"></i><b>12.5</b> Ordination</a>
<ul>
<li class="chapter" data-level="12.5.1" data-path="mod-12.html"><a href="mod-12.html#principal-components-analysis-pca-1"><i class="fa fa-check"></i><b>12.5.1</b> Principal components analysis (PCA)</a></li>
<li class="chapter" data-level="12.5.2" data-path="mod-12.html"><a href="mod-12.html#nmds-and-other-ordination-methods"><i class="fa fa-check"></i><b>12.5.2</b> NMDS and other ordination methods</a></li>
<li class="chapter" data-level="12.5.3" data-path="mod-12.html"><a href="mod-12.html#other-ordination-techniques"><i class="fa fa-check"></i><b>12.5.3</b> Other ordination techniques</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="literature-cited.html"><a href="literature-cited.html"><i class="fa fa-check"></i>Literature Cited</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Applied Biological Data Analysis: Statistics and R for Biologists</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="literature-cited" class="section level1 unnumbered hasAnchor">
<h1>Literature Cited<a href="literature-cited.html#literature-cited" class="anchor-section" aria-label="Anchor link to header"></a></h1>

<div id="refs" class="references csl-bib-body hanging-indent" line-spacing="2">
<div class="csl-entry">
Alberti, E., G. Gioia, G. Sironi, S. Zanzani, P. Riccaboni, M. Magrini, and M. Manfredi. 2011. <a href="https://doi.org/10.1017/S0022149X10000647"><em>Elaphostrongylus cervi</em> in a population of red deer (<em><span>C</span>ervus elaphus</em>) and evidence of cerebrospinal nematodiasis in small ruminants in the province of <span>V</span>arese, <span>I</span>taly</a>. Journal of <span>H</span>elminthology 85:313–318.
</div>
<div class="csl-entry">
Anderson, M. J. 2001. <a href="https://doi.org/10.1111/j.1442-9993.2001.01070.pp.x">A new method for non-parametric multivariate analysis of variance</a>. Austral Ecology 26:32–46.
</div>
<div class="csl-entry">
Bodin, N., H. Pethybridge, L. M. Duffy, A. Lorrain, V. Allain, J. M. Logan, F. Ménard, B. Graham, C. A. Choy, C. J. Somes, R. J. Olson, and J. W. Young. 2021. <a href="https://doi.org/10.1002/ecy.3265">Global data set for nitrogen and carbon stable isotopes of tunas</a>. Ecology 102:e03265.
</div>
<div class="csl-entry">
Bolker, B. M. 2008. Ecological models and data in <span>R</span>. Princeton University Press, Princeton, New Jersey, USA.
</div>
<div class="csl-entry">
Bolker, B. M., M. E. Brooks, C. J. Clark, S. W. Geange, J. R. Poulsen, M. H. H. Stevens, and J.-S. S. White. 2009. <a href="https://doi.org/10.1016/j.tree.2008.10.008">Generalized linear mixed models: A practical guide for ecology and evolution</a>. Trends in Ecology &amp; Evolution 24:127–135.
</div>
<div class="csl-entry">
Borcard, D., F. Gillet, and P. Legendre. 2018. Numerical ecology with <span>R</span>. Springer, New York.
</div>
<div class="csl-entry">
Box, G. E. P., and D. R. Cox. 1964. <a href="https://doi.org/10.1111/j.2517-6161.1964.tb00553.x">An analysis of transformations</a>. Journal of the Royal Statistical Society: Series B (Methodological) 26:211–243.
</div>
<div class="csl-entry">
Brieuc, M. S. O., C. D. Waters, D. P. Drinan, and K. A. Naish. 2018. <a href="https://doi.org/10.1111/1755-0998.12773">A practical introduction to random forest for genetic association studies in ecology and evolution</a>. Molecular Ecology Resources 18:755–766.
</div>
<div class="csl-entry">
Broman, K. W., and K. H. Woo. 2018. <a href="https://doi.org/10.1080/00031305.2017.1375989">Data organization in spreadsheets</a>. The American Statistician 72:2–10.
</div>
<div class="csl-entry">
Bubrig, L. T., J. M. Sutton, and J. L. Fierst. 2020. <a href="https://doi.org/10.1002/ece3.6646"><em><span>C</span>aenorhabditis elegans</em> dauers vary recovery in response to bacteria from natural habitat</a>. Ecology and Evolution 10:9886–9895.
</div>
<div class="csl-entry">
Burnham, K. P., and D. R. Anderson. 2004. Multimodel inference: Understanding <span>AIC</span> and <span>BIC</span> in model selection. Sociological methods &amp; research 33:261–304.
</div>
<div class="csl-entry">
Burnham, K., and D. Anderson. 2002. Model selection and multi-model inference: <span class="nocase">a</span> practical information-theoretic approach. Springer-<span>V</span>erlag, New York.
</div>
<div class="csl-entry">
Button, K. S., J. P. Ioannidis, C. Mokrysz, B. A. Nosek, J. Flint, E. S. Robinson, and M. R. Munafò. 2013. <a href="https://doi.org/10.1038/nrn3475">Power failure: <span class="nocase">w</span>hy small sample size undermines the reliability of neuroscience</a>. Nature <span>R</span>eviews <span>N</span>euroscience 14:365–376.
</div>
<div class="csl-entry">
Cade, B. S., and B. R. Noon. 2003. <a href="https://doi.org/10.1890/1540-9295(2003)001[0412:agitqr]2.0.co;2">A gentle introduction to quantile regression for ecologists</a>. Frontiers in Ecology and the Environment 1:412–420.
</div>
<div class="csl-entry">
Carey, V. J. 2019. <a href="https://CRAN.R-project.org/package=gee"><span class="nocase">g</span>ee: Generalized estimation equation solver</a>.
</div>
<div class="csl-entry">
Carnell, R. 2019. <a href="https://CRAN.R-project.org/package=triangle">Triangle: Provides the standard distribution functions for the triangle distribution</a>.
</div>
<div class="csl-entry">
Chivers, D. J., and C. M. Hladik. 1980. <a href="https://doi.org/10.1002/jmor.1051660306">Morphology of the gastrointestinal tract in primates: Comparisons with other mammals in relation to diet</a>. Journal of Morphology 166:337–386.
</div>
<div class="csl-entry">
Clarke, K. R. 1993. <a href="https://doi.org/10.1111/j.1442-9993.1993.tb00438.x">Non-parametric multivariate analyses of changes in community structure</a>. Australian journal of ecology 18:117–143.
</div>
<div class="csl-entry">
Clarke, K. R., P. J. Somerfield, and M. G. Chapman. 2006. <a href="https://doi.org/10.1016/j.jembe.2005.12.017">On resemblance measures for ecological studies, including taxonomic dissimilarities and a zero-adjusted <span>B</span>ray-<span>C</span>urtis coefficient for denuded assemblages</a>. Journal of Experimental Marine Biology and Ecology 330:55–80.
</div>
<div class="csl-entry">
Cutler, D. R., T. C. Edwards, K. H. Beard, A. Cutler, K. T. Hess, J. Gibson, and J. J. Lawler. 2007. <a href="https://doi.org/10.1890/07-0539.1">Random forests for classification in ecology</a>. Ecology 88:2783–2792.
</div>
<div class="csl-entry">
Dalgaard, P. 2002. Introductory statistics with <span>R</span>. Springer Science+Business Media.
</div>
<div class="csl-entry">
De’ath, G. 2007. <a href="https://doi.org/10.1890/0012-9658(2007)88[243:btfema]2.0.co;2">Boosted trees for ecological modeling and prediction</a>. Ecology 88:243–251.
</div>
<div class="csl-entry">
De’ath, G., and K. E. Fabricius. 2000. <a href="https://doi.org/10.1890/0012-9658(2000)081[3178:CARTAP]2.0.CO;2">Classification and regression trees: A powerful yet simple technique for ecological data analysis</a>. Ecology 81:3178–3192.
</div>
<div class="csl-entry">
De Solla, S. R., P. A. Martin, K. J. Fernie, B. J. Park, and G. Mayne. 2006. <a href="https://doi.org/10.1897/05-165R.1">Effects of environmentally relevant concentrations of atrazine on gonadal development of snapping turtles (<em><span>C</span>helydra serpentina</em>)</a>. Environmental Toxicology and Chemistry: An International Journal 25:520–526.
</div>
<div class="csl-entry">
Dormann, C. F., J. Elith, S. Bacher, C. Buchmann, G. Carl, G. Carré, J. R. G. Marquéz, B. Gruber, B. Lafourcade, P. J. Leitão, T. Münkemüller, C. McClean, P. E. Osborne, B. Reineking, B. Schröder, A. K. Skidmore, D. Zurell, and S. Lautenbach. 2013. <a href="https://doi.org/10.1111/j.1600-0587.2012.07348.x">Collinearity: A review of methods to deal with it and a simulation study evaluating their performance</a>. Ecography 36:27–46.
</div>
<div class="csl-entry">
Elith, J., and J. Leathwick. 2017. Boosted regression trees for ecological modeling. R Documentation. Available online: https://cran.r-project.org/web/packages/dismo/vignettes/brt.pdf (accessed on 2021-12-23).
</div>
<div class="csl-entry">
Elith, J., J. R. Leathwick, and T. Hastie. 2008. <a href="https://doi.org/10.1111/j.1365-2656.2008.01390.x">A working guide to boosted regression trees</a>. Journal of Animal Ecology 77:802–813.
</div>
<div class="csl-entry">
Faurby, S., M. Davis, R. Ø. Pedersen, S. D. Schowanek, A. Antonelli, and J.-C. Svenning. 2018. <a href="https://doi.org/10.1002/ecy.2443"><span>PHYLACINE</span> 1.2: The phylogenetic atlas of mammal macroecology</a>. Ecology 99:2626.
</div>
<div class="csl-entry">
Foucault, Q., A. Wieser, A.-M. Waldvogel, B. Feldmeyer, and M. Pfenninger. 2018. <a href="https://doi.org/10.1002/ece3.4706">Rapid adaptation to high temperatures in <em><span>C</span>hironomus riparius</em></a>. Ecology and Evolution 8:12780–12789.
</div>
<div class="csl-entry">
Gelfand, A. E., and P. Vounatsou. 2003. <a href="https://doi.org/10.1093/biostatistics/4.1.11">Proper multivariate conditional autoregressive models for spatial data analysis</a>. Biostatistics 4:11–15.
</div>
<div class="csl-entry">
Gould, S. J. 1974. <a href="https://doi.org/10.2307/2407322">The origin and function of ’bizarre’ structures: Antler size and skull size in the ’irish elk,’ <em><span>M</span>egaloceros giganteus</em></a>. Evolution 28:191–220.
</div>
<div class="csl-entry">
Green, N. S., S. Li, J. D. Maul, and J. P. Overmyer. 2022. <a href="https://doi.org/10.1016/j.scitotenv.2022.155344">Natural and anthropogenic factors and their interactions drive stream community integrity in a <span>N</span>orth <span>A</span>merican river basin at a large spatial scale</a>. Science of the Total Environment 835:155344.
</div>
<div class="csl-entry">
Green, N. S., M. L. Wildhaber, and J. L. Albers. 2019. <a href="https://doi.org/10.1111/jai.13701">Potential responses of the <span>L</span>ower <span>M</span>issouri <span>R</span>iver shovelnose sturgeon (<em><span>S</span>caphirhynchus platorynchus</em>) population to a commercial fishing ban</a>. Journal of Applied Ichthyology 35:370–377.
</div>
<div class="csl-entry">
Green, N. S., M. L. Wildhaber, and J. L. Albers. 2021. Effectiveness of a distance sampling from roads program for white-tailed deer in the <span>N</span>ational <span>C</span>apital <span>R</span>egion parks. Natural Resource Report, National Park Service, Fort Collins, Colorado.
</div>
<div class="csl-entry">
Green, N. S., M. L. Wildhaber, J. L. Albers, T. W. Pettit, and M. J. Hooper. 2020. <a href="https://doi.org/10.1002/ieam.4324">Efficient mammal biodiversity surveys for ecological restoration monitoring</a>. Integrated Environmental Assessment and Management.
</div>
<div class="csl-entry">
Grolemund, G., and H. Wickham. 2011. <a href="https://www.jstatsoft.org/v40/i03/">Dates and times made easy with <span class="nocase">lubridate</span></a>. Journal of Statistical Software 40:1–25.
</div>
<div class="csl-entry">
Head, M. L., L. Holman, R. Lanfear, A. T. Kahn, and M. D. Jennions. 2015. <a href="https://doi.org/10.1371/journal.pbio.1002106">The extent and consequences of p-hacking in science</a>. <span>PLoS</span> <span>B</span>iology 13:e1002106.
</div>
<div class="csl-entry">
Hill, M. O. 1979. A <span>FORTRAN</span> program for arranging multivariate data in an ordered two-way table by classification of the individuals and attributes. <span>TWINSPAN</span>.
</div>
<div class="csl-entry">
Hobbs, N. T., and R. Hilborn. 2006. <a href="https://doi.org/10.1890/04-0645">Alternatives to statistical hypothesis testing in ecology: <span class="nocase">a</span> guide to self teaching</a>. Ecological Applications 16:5–19.
</div>
<div class="csl-entry">
Højsgaard, S., U. Halekoh, and J. Yan. 2006. <a href="https://doi.org/10.18637/jss.v015.i02">The <span>R</span> package geepack for generalized estimating equations</a>. Journal of statistical software 15:1–11.
</div>
<div class="csl-entry">
Hurlbert, S. H. 1984. <a href="https://doi.org/10.2307/1942661">Pseudoreplication and the design of ecological field experiments</a>. Ecological <span>M</span>onographs 54:187–211.
</div>
<div class="csl-entry">
Hutcheon, J. M., J. A. W. Kirsch, and T. G. Jr. 2002. <a href="https://doi.org/10.1159/000065938">A comparative analysis of brain size in relation to foraging ecology and phylogeny in the <span>C</span>hiroptera</a>. Brain, behavior and evolution 60:165–180.
</div>
<div class="csl-entry">
Inderjit, J. C. Streibig, and M. Olofsdotter. 2002. <a href="https://doi.org/10.1034/j.1399-3054.2002.1140312.x">Joint action of phenolic acid mixtures and its significance in allelopathy research</a>. Physiologia Plantarum 114:422–428.
</div>
<div class="csl-entry">
James, G., D. Witten, T. Hastie, and R. Tibshirani. 2013. <a href="http://faculty.marshall.usc.edu/gareth-james/ISL/">An introduction to statistical learning</a>. 7th edition. Springer.
</div>
<div class="csl-entry">
Jones, D. T. 2019. <a href="https://doi.org/10.1038/s41580-019-0176-5">Setting the standards for machine learning in biology</a>. Nature <span>R</span>eviews <span>M</span>olecular <span>C</span>ell <span>B</span>iology 20:659–660.
</div>
<div class="csl-entry">
Kan, A. 2017. <a href="https://doi.org/10.1038/icb.2017.16">Machine learning applications in cell image analysis</a>. Immunology and <span>C</span>ell <span>B</span>iology 95:525–530.
</div>
<div class="csl-entry">
Kéry, M. 2002. Inferring the absence of a species: <span class="nocase">a</span> case study of snakes. The Journal of <span>W</span>ildlife <span>M</span>anagement:330–338.
</div>
<div class="csl-entry">
Kéry, M. 2010. Introduction to <span>W</span>in<span>BUGS</span> for ecologists: <span>B</span>ayesian approach to regression, <span>ANOVA</span>, mixed models and related analyses. Academic Press.
</div>
<div class="csl-entry">
Kirchman, D. L., X. A. G. Morán, and H. Ducklow. 2009. Microbial growth in the polar oceans—role of temperature and potential impact of climate change. Nature <span>R</span>eviews <span>M</span>icrobiology 7:451–459.
</div>
<div class="csl-entry">
Kirchner, B. N., N. S. Green, D. A. Sergeant, J. N. Mink, and K. T. Wilkins. 2011. <a href="https://doi.org/10.1674/0003-0031-166.1.112">Responses of small mammals and vegetation to a prescribed burn in a tallgrass blackland prairie</a>. The American Midland Naturalist 166:112–125.
</div>
<div class="csl-entry">
Koenker, R. 2021. <a href="https://CRAN.R-project.org/package=quantreg">Quantreg: Quantile regression</a>.
</div>
<div class="csl-entry">
Kuehn, I., and C. F. Dormann. 2012. <a href="https://doi.org/10.1111/j.1365-2699.2012.02707.x">Less than eight (and a half) misconceptions of spatial analysis</a>. Journal of Biogeography 39:995–998.
</div>
<div class="csl-entry">
Kunz, B. K., J. H. Waddle, and N. S. Green. 2019. <a href="https://doi.org/10.1002/ieam.4202">Amphibian monitoring in hardwood forests: Optimizing methods for contaminant-based compensatory restorations</a>. Integrated environmental assessment and management.
</div>
<div class="csl-entry">
Kuznetsova, A., P. B. Brockhoff, and R. H. B. Christensen. 2017. <a href="https://doi.org/10.18637/jss.v082.i13"><span class="nocase">l</span>mer<span>T</span>est package: Tests in linear mixed effects models</a>. Journal of Statistical Software 82:1–26.
</div>
<div class="csl-entry">
Lazic, S. E., C. J. Clarke-Williams, and M. R. Munafò. 2018. <a href="https://doi.org/10.1371/journal.pbio.2005282">What exactly is ‘<span>N</span>’in cell culture and animal experiments?</a> <span>PLoS</span> <span>B</span>iology 16:e2005282.
</div>
<div class="csl-entry">
Lebreton, J.-D., K. P. Burnham, J. Clobert, and D. R. Anderson. 1992. <a href="https://doi.org/10.2307/2937171">Modeling survival and testing biological hypotheses using marked animals: <span class="nocase">a</span> unified approach with case studies</a>. Ecological <span>M</span>onographs 62:67–118.
</div>
<div class="csl-entry">
Legendre, P., and L. Legendre. 2012. Numerical ecology. Elsevier, Amsterdam.
</div>
<div class="csl-entry">
Leonard, E. M., and C. M. Wood. 2013. <a href="https://doi.org/10.1016/j.cbpc.2013.03.008">Acute toxicity, critical body residues, <span>M</span>ichaelis–<span>M</span>enten analysis of bioaccumulation, and ionoregulatory disturbance in response to waterborne nickel in four invertebrates: <em><span>C</span>hironomus riparius</em>, <em><span>L</span>ymnaea stagnalis</em>, <em><span>L</span>umbriculus variegatus</em> and <em><span>D</span>aphnia pulex</em></a>. Comparative Biochemistry and Physiology Part C: Toxicology &amp; Pharmacology 158:10–21.
</div>
<div class="csl-entry">
Liang, K.-Y., and S. L. Zeger. 1986. Longitudinal data analysis using generalized linear models. Biometrika 73:13–22.
</div>
<div class="csl-entry">
Link, W. A., E. Cam, J. D. Nichols, and E. G. Cooch. 2002. <a href="https://doi.org/10.2307/3803160">Of <span>BUGS</span> and birds: <span>M</span>arkov chain <span>M</span>onte <span>C</span>arlo for hierarchical modeling in wildlife research</a>. The Journal of <span>W</span>ildlife <span>M</span>anagement:277–291.
</div>
<div class="csl-entry">
Ludwig, J. A., J. F. Reynolds, L. QUARTET, and J. Reynolds. 1988. Statistical ecology: <span class="nocase">a</span> primer in methods and computing. John Wiley &amp; Sons.
</div>
<div class="csl-entry">
Lunn, D., D. Spiegelhalter, A. Thomas, and N. Best. 2009. <a href="https://doi.org/10.1002/sim.3680">The <span>BUGS</span> project: <span class="nocase">e</span>volution, critique and future directions</a>. Statistics in <span>M</span>edicine 28:3049–3067.
</div>
<div class="csl-entry">
Makin, T. R., and J.-J. O. de Xivry. 2019. <a href="https://doi.org/10.7554/eLife.48175">Science forum: Ten common statistical mistakes to watch out for when writing or reviewing a manuscript</a>. Elife 8:e48175.
</div>
<div class="csl-entry">
Mangiafico, S. 2021. <a href="https://CRAN.R-project.org/package=rcompanion"><span class="nocase">r</span>companion: Functions to support extension education program evaluation</a>.
</div>
<div class="csl-entry">
Mantel, N. 1967. <a href="https://cancerres.aacrjournals.org/content/27/2_Part_1/209">The detection of disease clustering and a generalized regression approach</a>. Cancer research 27:209–220.
</div>
<div class="csl-entry">
McCarthy, M. A. 2007. Bayesian methods for ecology. Cambridge University Press, Cambridge, United Kingdom.
</div>
<div class="csl-entry">
McCune, B., J. B. Grace, and D. L. Urban. 2002. Analysis of ecological communities. MjM software design, Gleneden Beach, Oregon, USA.
</div>
<div class="csl-entry">
Nelder, J. A., and R. W. M. Wedderburn. 1972. <a href="https://doi.org/10.2307/2344614">Generalized linear models</a>. Journal of the Royal Statistical Society: Series A (General) 135:370.
</div>
<div class="csl-entry">
Penone, C., A. D. Davidson, K. T. Shoemaker, M. Di Marco, C. Rondinini, T. M. Brooks, B. E. Young, C. H. Graham, and G. C. Costa. 2014. <a href="https://doi.org/10.1111/2041-210X.12232">Imputation of missing data in life-history trait datasets: Which approach performs the best?</a> Methods in Ecology and Evolution 5:961–970.
</div>
<div class="csl-entry">
Pethybridge, H., C. A. Choy, J. M. Logan, V. Allain, A. Lorrain, N. Bodin, C. J. Somes, J. Young, F. Ménard, C. Langlais, L. Duffy, A. J. Hobday, P. Kuhnert, B. Fry, C. Menkes, and R. J. Olson. 2018. <a href="https://doi.org/10.1111/geb.12763">A global meta-analysis of marine predator nitrogen stable isotopes: Relationships between trophic structure and environmental conditions</a>. Global Ecology and Biogeography 27:1043–1055.
</div>
<div class="csl-entry">
Plummer, M. 2003. <span>JAGS</span>: A program for analysis of <span>B</span>ayesian graphical models using <span>G</span>ibbs sampling. Pages 1–10 Proceedings of the 3rd international workshop on distributed statistical computing. Vienna, Austria.
</div>
<div class="csl-entry">
Plummer, M. 2021. <a href="https://CRAN.R-project.org/package=rjags"><span class="nocase">r</span>jags: <span>B</span>ayesian graphical models using <span>MCMC</span></a>.
</div>
<div class="csl-entry">
Ritz, C., F. Baty, J. C. Streibig, and D. Gerhard. 2015. <a href="https://doi.org/10.1371/journal.pone.0146021">Dose-response analysis using <span>R</span></a>. PLoS ONE 10:e0146021.
</div>
<div class="csl-entry">
Roleček, J., L. Tichý, D. Zelený, and M. Chytrý. 2009. <a href="https://doi.org/10.1111/j.1654-1103.2009.01062.x">Modified <span>TWINSPAN</span> classification in which the hierarchy respects cluster heterogeneity</a>. Journal of Vegetation Science 20:596–602.
</div>
<div class="csl-entry">
Schabenberger, O., and F. J. Pierce. 2001. Contemporary statistical models for the plant and soil sciences. CRC press, Boca Raton, Florida, USA.
</div>
<div class="csl-entry">
Smith, F. A., and S. K. Lyons. 2013. Animal body size: Linking pattern and process across space, time, and taxonomic group. University of Chicago Press, Chicago.
</div>
<div class="csl-entry">
Spector, P. 2008. Data manipulation with <span>R</span>. Springer Science+Business Media, New York.
</div>
<div class="csl-entry">
Spiess, A.-N. 2018. <a href="https://CRAN.R-project.org/package=propagate"><span class="nocase">p</span>ropagate: Propagation of uncertainty</a>.
</div>
<div class="csl-entry">
Spiess, A.-N., and N. Neumeyer. 2010. <a href="https://doi.org/10.1186/1471-2210-10-6">An evaluation of <span class="math inline">\(R^2\)</span> as an inadequate measure for nonlinear models in pharmacological and biochemical research: A <span>M</span>onte <span>C</span>arlo approach</a>. BMC Pharmacology 10.
</div>
<div class="csl-entry">
Su, Y.-S., and Masanao Yajima. 2021. <a href="https://CRAN.R-project.org/package=R2jags"><span>R2</span>jags: Using <span>R</span> to run ’<span>JAGS</span>’</a>.
</div>
<div class="csl-entry">
Tarca, A. L., V. J. Carey, X. Chen, R. Romero, and S. Drăghici. 2007. <a href="https://doi.org/10.1371/journal.pcbi.0030116">Machine learning and its applications to biology</a>. <span>PLoS</span> <span>C</span>omputational <span>B</span>iology 3:e116.
</div>
<div class="csl-entry">
Therneau, T., and B. Atkinson. 2019. <a href="https://CRAN.R-project.org/package=rpart"><span class="nocase">r</span>part: Recursive partitioning and regression trees</a>.
</div>
<div class="csl-entry">
Tufte, E. 2001. The visual display of quantitative information. Cheshire: Graphic Press.–2001.–213 p.
</div>
<div class="csl-entry">
Turner, H., and D. Firth. 2020. <a href="https://cran.r-project.org/package=gnm">Generalized nonlinear models in <span>R</span>: An overview of the gnm package</a>.
</div>
<div class="csl-entry">
Venables, W. N., and B. D. Ripley. 2002. <a href="https://www.stats.ox.ac.uk/pub/MASS4/">Modern applied statistics with <span>S</span></a>. Fourth. Springer, New York.
</div>
<div class="csl-entry">
Vicente, J., Y. Fierro, and C. Gortazar. 2005. <a href="https://doi.org/10.1007/s00436-004-1255-9">Seasonal dynamics of the fecal excretion of <em><span>E</span>laphostrongylus cervi</em> (<span>N</span>ematoda, <span>M</span>etastrongyloidea) first-stage larvae in <span>I</span>berian red deer (<em><span>C</span>ervus elaphus hispanicus) from southern <span>S</span>pain</em></a>. Parasitology Research 95:60–64.
</div>
<div class="csl-entry">
Ward, J. H. 1963. <a href="https://doi.org/10.1080/01621459.1963.10500845">Hierarchical grouping to optimize an objective function</a>. Journal of the American Statistical Association 58:236–244.
</div>
<div class="csl-entry">
Warton, D. I., and F. K. C. Hui. 2011. <a href="https://doi.org/10.1890/10-0340.1">The arcsine is asinine: The analysis of proportions in ecology</a>. Ecology 92:3–10.
</div>
<div class="csl-entry">
Wasserstein, R. L., and N. A. Lazar. 2016. <a href="https://doi.org/10.1080/00031305.2016.1154108">The ASA statement on p-values: Context, process, and purpose</a>. Taylor &amp; Francis.
</div>
<div class="csl-entry">
Werner, J., and E. M. Griebeler. 2014. <a href="https://doi.org/10.1371/journal.pone.0088834">Allometries of maximum growth rate versus body mass at maximum growth indicate that non-avian dinosaurs had growth rates typical of fast growing ectothermic sauropsids</a>. <span>PloS</span> <span>ONE</span> 9:e88834.
</div>
<div class="csl-entry">
White, J. W., A. Rassweiler, J. F. Samhouri, A. C. Stier, and C. White. 2014. <a href="https://doi.org/10.1111/j.1600-0706.2013.01073.x">Ecologists should not use statistical significance tests to interpret simulation model results</a>. Oikos 123:385–388.
</div>
<div class="csl-entry">
Whytock, R. C., J. Świeżewski, J. A. Zwerts, T. Bara-Słupski, A. F. Koumba Pambo, M. Rogala, L. Bahaa-el-din, K. Boekee, S. Brittain, A. W. Cardoso, and others. 2021. <a href="https://doi.org/10.1111/2041-210X.13576">Robust ecological analysis of camera trap data labelled by a machine learning model</a>. Methods in <span>E</span>cology and <span>E</span>volution.
</div>
<div class="csl-entry">
Wickham, H. 2014. <a href="https://doi.org/10.18637/jss.v059.i10">Tidy data</a>. Journal of Statistical Software 59.
</div>
<div class="csl-entry">
Wickham, H., M. Averick, J. Bryan, W. Chang, L. D. McGowan, R. François, G. Grolemund, A. Hayes, L. Henry, J. Hester, M. Kuhn, T. L. Pedersen, E. Miller, S. M. Bache, K. Müller, J. Ooms, D. Robinson, D. P. Seidel, V. Spinu, K. Takahashi, D. Vaughan, C. Wilke, K. Woo, and H. Yutani. 2019. <a href="https://doi.org/10.21105/joss.01686">Welcome to the <span class="nocase">tidyverse</span></a>. Journal of Open Source Software 4:1686.
</div>
<div class="csl-entry">
Wood, S. N. 2017. Generalized additive models: An introduction with <span>R</span>. 2nd edition. CRC press.
</div>
<div class="csl-entry">
Woodward, H. N., E. A. F. Fowler, J. O. Farlow, and J. R. Horner. 2015. <a href="https://doi.org/10.1017/pab.2015.19"><em><span>M</span>aiasaura</em>, a model organism for extinct vertebrate population biology: A large sample statistical assessment of growth dynamics and survivorship</a>. Paleobiology 41:503–527.
</div>
<div class="csl-entry">
Zuur, A. F., E. N. Ieno, and C. S. Elphick. 2010. <a href="https://doi.org/10.1111/j.2041-210X.2009.00001.x">A protocol for data exploration to avoid common statistical problems</a>. Methods in ecology and evolution 1:3–14.
</div>
<div class="csl-entry">
Zuur, A. F., E. N. Ieno, and G. M. Smith. 2007. Analysing ecological data. Springer Science+Business Media, New York.
</div>
<div class="csl-entry">
Zuur, A. F., E. N. Ieno, N. J. Walker, A. A. Saveliev, G. M. Smith, and others. 2009a. Mixed effects models and extensions in ecology with <span>R</span>. Springer, New York.
</div>
<div class="csl-entry">
Zuur, A., E. N. Ieno, and E. Meesters. 2009b. A beginner’s guide to <span>R</span>. Springer Science+Business Media, New York.
</div>
</div>
</div>






































































































































            </section>

          </div>
        </div>
      </div>
<a href="mod-12.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/13-references.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["applied_biol_data_analysis_20220816.pdf", "applied_biol_data_analysis_20220816.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
