<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Module 11 Nonlinear models | Applied Biological Data Analysis: Statistics and R for Biologists</title>
  <meta name="description" content="Helping biologists to become more informed users of R and statistical methods." />
  <meta name="generator" content="bookdown 0.37 and GitBook 2.6.7" />

  <meta property="og:title" content="Module 11 Nonlinear models | Applied Biological Data Analysis: Statistics and R for Biologists" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://greenquanteco.github.io/book/lab_logo_02.jpg" />
  <meta property="og:description" content="Helping biologists to become more informed users of R and statistical methods." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Module 11 Nonlinear models | Applied Biological Data Analysis: Statistics and R for Biologists" />
  
  <meta name="twitter:description" content="Helping biologists to become more informed users of R and statistical methods." />
  <meta name="twitter:image" content="https://greenquanteco.github.io/book/lab_logo_02.jpg" />

<meta name="author" content="Dr. Nick Green, Kennesaw State University" />


<meta name="date" content="2024-08-20" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="fav.ico" type="image/x-icon" />
<link rel="prev" href="mod-10.html"/>
<link rel="next" href="mod-12.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Applied biological data analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#work-in-progress-warning"><i class="fa fa-check"></i>Work in progress warning!</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#license-and-permissions"><i class="fa fa-check"></i>License and permissions</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-description"><i class="fa fa-check"></i>Course description</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-objectives"><i class="fa fa-check"></i>Course objectives</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-requirements"><i class="fa fa-check"></i>Course requirements</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#recommended-reading"><i class="fa fa-check"></i>Recommended reading</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-organization"><i class="fa fa-check"></i>Course organization</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#about-the-author"><i class="fa fa-check"></i>About the author</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="mod-01.html"><a href="mod-01.html"><i class="fa fa-check"></i><b>1</b> Statistics in modern biology</a>
<ul>
<li class="chapter" data-level="1.1" data-path="mod-01.html"><a href="mod-01.html#overview"><i class="fa fa-check"></i><b>1.1</b> Overview</a></li>
<li class="chapter" data-level="1.2" data-path="mod-01.html"><a href="mod-01.html#statistics-in-modern-biology"><i class="fa fa-check"></i><b>1.2</b> Statistics in modern biology</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="mod-01.html"><a href="mod-01.html#the-scientific-method"><i class="fa fa-check"></i><b>1.2.1</b> The scientific method</a></li>
<li class="chapter" data-level="1.2.2" data-path="mod-01.html"><a href="mod-01.html#example-data-analysis"><i class="fa fa-check"></i><b>1.2.2</b> Example data analysis</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="mod-01.html"><a href="mod-01.html#misuses-of-statistics"><i class="fa fa-check"></i><b>1.3</b> Misuses of statistics</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="mod-01.html"><a href="mod-01.html#proving-the-trivial-and-meaningless-hypotheses"><i class="fa fa-check"></i><b>1.3.1</b> “Proving” the trivial and meaningless hypotheses</a></li>
<li class="chapter" data-level="1.3.2" data-path="mod-01.html"><a href="mod-01.html#inappropriate-methods"><i class="fa fa-check"></i><b>1.3.2</b> Inappropriate methods</a></li>
<li class="chapter" data-level="1.3.3" data-path="mod-01.html"><a href="mod-01.html#p-hacking-and-data-dredging"><i class="fa fa-check"></i><b>1.3.3</b> <em>P</em>-hacking and data dredging</a></li>
<li class="chapter" data-level="1.3.4" data-path="mod-01.html"><a href="mod-01.html#inadequate-sample-sizes-and-pseudoreplication"><i class="fa fa-check"></i><b>1.3.4</b> Inadequate sample sizes and pseudoreplication</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="mod-01.html"><a href="mod-01.html#mod-01-nhst"><i class="fa fa-check"></i><b>1.4</b> <em>P</em>-values and null hypothesis significance testing (NHST)</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="mod-01.html"><a href="mod-01.html#definition"><i class="fa fa-check"></i><b>1.4.1</b> Definition</a></li>
<li class="chapter" data-level="1.4.2" data-path="mod-01.html"><a href="mod-01.html#aside-another-way-to-frame-null-hypotheses"><i class="fa fa-check"></i><b>1.4.2</b> Aside: Another way to frame null hypotheses</a></li>
<li class="chapter" data-level="1.4.3" data-path="mod-01.html"><a href="mod-01.html#history-and-status-of-p-values"><i class="fa fa-check"></i><b>1.4.3</b> History and status of <em>P</em>-values</a></li>
<li class="chapter" data-level="1.4.4" data-path="mod-01.html"><a href="mod-01.html#where-p-values-come-from"><i class="fa fa-check"></i><b>1.4.4</b> Where <em>P</em>-values come from</a></li>
<li class="chapter" data-level="1.4.5" data-path="mod-01.html"><a href="mod-01.html#what-p-values-mean-and-do-not-mean"><i class="fa fa-check"></i><b>1.4.5</b> What <em>P</em>-values mean and do not mean</a></li>
<li class="chapter" data-level="1.4.6" data-path="mod-01.html"><a href="mod-01.html#do-you-need-a-p-value"><i class="fa fa-check"></i><b>1.4.6</b> Do you need a <em>P</em>-value?</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="mod-01.html"><a href="mod-01.html#alternatives-to-nhst"><i class="fa fa-check"></i><b>1.5</b> Alternatives to NHST</a>
<ul>
<li class="chapter" data-level="1.5.1" data-path="mod-01.html"><a href="mod-01.html#mod-01-bayes"><i class="fa fa-check"></i><b>1.5.1</b> Bayesian inference</a></li>
<li class="chapter" data-level="1.5.2" data-path="mod-01.html"><a href="mod-01.html#information-theoretic-methods"><i class="fa fa-check"></i><b>1.5.2</b> Information-theoretic methods</a></li>
<li class="chapter" data-level="1.5.3" data-path="mod-01.html"><a href="mod-01.html#machine-learning"><i class="fa fa-check"></i><b>1.5.3</b> Machine learning</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="mod-02.html"><a href="mod-02.html"><i class="fa fa-check"></i><b>2</b> Introduction to R</a>
<ul>
<li class="chapter" data-level="2.1" data-path="mod-02.html"><a href="mod-02.html#getting-started-with-r"><i class="fa fa-check"></i><b>2.1</b> Getting started with R</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="mod-02.html"><a href="mod-02.html#what-is-r"><i class="fa fa-check"></i><b>2.1.1</b> What is R?</a></li>
<li class="chapter" data-level="2.1.2" data-path="mod-02.html"><a href="mod-02.html#advantages-of-r"><i class="fa fa-check"></i><b>2.1.2</b> Advantages of R</a></li>
<li class="chapter" data-level="2.1.3" data-path="mod-02.html"><a href="mod-02.html#disadvantages-of-r"><i class="fa fa-check"></i><b>2.1.3</b> Disadvantages of R</a></li>
<li class="chapter" data-level="2.1.4" data-path="mod-02.html"><a href="mod-02.html#base-r-and-vs.-tidyverse"><i class="fa fa-check"></i><b>2.1.4</b> Base R and (vs.?) <code>tidyverse</code></a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="mod-02.html"><a href="mod-02.html#download-and-install-r-and-rstudio"><i class="fa fa-check"></i><b>2.2</b> Download and install R (and RStudio)</a></li>
<li class="chapter" data-level="2.3" data-path="mod-02.html"><a href="mod-02.html#using-r"><i class="fa fa-check"></i><b>2.3</b> Using R</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="mod-02.html"><a href="mod-02.html#using-the-base-r-gui"><i class="fa fa-check"></i><b>2.3.1</b> Using the base R GUI</a></li>
<li class="chapter" data-level="2.3.2" data-path="mod-02.html"><a href="mod-02.html#using-r-in-rstudio"><i class="fa fa-check"></i><b>2.3.2</b> Using R in RStudio</a></li>
<li class="chapter" data-level="2.3.3" data-path="mod-02.html"><a href="mod-02.html#using-r-with-other-programs"><i class="fa fa-check"></i><b>2.3.3</b> Using R with other programs</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="mod-02.html"><a href="mod-02.html#mod-02-first"><i class="fa fa-check"></i><b>2.4</b> A first R session</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="mod-02.html"><a href="mod-02.html#import-data"><i class="fa fa-check"></i><b>2.4.1</b> Import data</a></li>
<li class="chapter" data-level="2.4.2" data-path="mod-02.html"><a href="mod-02.html#explore-and-visualize-data"><i class="fa fa-check"></i><b>2.4.2</b> Explore and visualize data</a></li>
<li class="chapter" data-level="2.4.3" data-path="mod-02.html"><a href="mod-02.html#transform-data"><i class="fa fa-check"></i><b>2.4.3</b> Transform data</a></li>
<li class="chapter" data-level="2.4.4" data-path="mod-02.html"><a href="mod-02.html#analyze-data"><i class="fa fa-check"></i><b>2.4.4</b> Analyze data</a></li>
<li class="chapter" data-level="2.4.5" data-path="mod-02.html"><a href="mod-02.html#mod-01-elkfig"><i class="fa fa-check"></i><b>2.4.5</b> Write out results</a></li>
<li class="chapter" data-level="2.4.6" data-path="mod-02.html"><a href="mod-02.html#save-your-work"><i class="fa fa-check"></i><b>2.4.6</b> Save your work?</a></li>
<li class="chapter" data-level="2.4.7" data-path="mod-02.html"><a href="mod-02.html#whats-next"><i class="fa fa-check"></i><b>2.4.7</b> What’s next?</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="mod-02.html"><a href="mod-02.html#write-and-execute-commands-in-the-r-console"><i class="fa fa-check"></i><b>2.5</b> Write and execute commands in the R console</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="mod-02.html"><a href="mod-02.html#r-commandsbasics"><i class="fa fa-check"></i><b>2.5.1</b> R commands–basics</a></li>
<li class="chapter" data-level="2.5.2" data-path="mod-02.html"><a href="mod-02.html#elements-of-r-code"><i class="fa fa-check"></i><b>2.5.2</b> Elements of R code</a></li>
<li class="chapter" data-level="2.5.3" data-path="mod-02.html"><a href="mod-02.html#the-r-workspace"><i class="fa fa-check"></i><b>2.5.3</b> The R workspace</a></li>
<li class="chapter" data-level="2.5.4" data-path="mod-02.html"><a href="mod-02.html#r-code-basics-assignment-and-operators"><i class="fa fa-check"></i><b>2.5.4</b> R code basics: assignment and operators</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="mod-02.html"><a href="mod-02.html#mod-02-struct"><i class="fa fa-check"></i><b>2.6</b> Basic R data structures</a>
<ul>
<li class="chapter" data-level="2.6.1" data-path="mod-02.html"><a href="mod-02.html#vectors"><i class="fa fa-check"></i><b>2.6.1</b> Vectors</a></li>
<li class="chapter" data-level="2.6.2" data-path="mod-02.html"><a href="mod-02.html#data-frames"><i class="fa fa-check"></i><b>2.6.2</b> Data frames</a></li>
<li class="chapter" data-level="2.6.3" data-path="mod-02.html"><a href="mod-02.html#matrices-and-arrays"><i class="fa fa-check"></i><b>2.6.3</b> Matrices and arrays</a></li>
<li class="chapter" data-level="2.6.4" data-path="mod-02.html"><a href="mod-02.html#lists"><i class="fa fa-check"></i><b>2.6.4</b> Lists</a></li>
<li class="chapter" data-level="2.6.5" data-path="mod-02.html"><a href="mod-02.html#s3-s4-and-other-paradigms"><i class="fa fa-check"></i><b>2.6.5</b> S3, S4, and other paradigms</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="mod-02.html"><a href="mod-02.html#r-data-types"><i class="fa fa-check"></i><b>2.7</b> R data types</a>
<ul>
<li class="chapter" data-level="2.7.1" data-path="mod-02.html"><a href="mod-02.html#character-type"><i class="fa fa-check"></i><b>2.7.1</b> Character type</a></li>
<li class="chapter" data-level="2.7.2" data-path="mod-02.html"><a href="mod-02.html#numeric-type"><i class="fa fa-check"></i><b>2.7.2</b> Numeric type</a></li>
<li class="chapter" data-level="2.7.3" data-path="mod-02.html"><a href="mod-02.html#integer-type"><i class="fa fa-check"></i><b>2.7.3</b> Integer type</a></li>
<li class="chapter" data-level="2.7.4" data-path="mod-02.html"><a href="mod-02.html#logical-type"><i class="fa fa-check"></i><b>2.7.4</b> Logical type</a></li>
<li class="chapter" data-level="2.7.5" data-path="mod-02.html"><a href="mod-02.html#special-values"><i class="fa fa-check"></i><b>2.7.5</b> Special values</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="mod-02.html"><a href="mod-02.html#manage-r-code-as-scripts-.r-files"><i class="fa fa-check"></i><b>2.8</b> Manage R code as scripts (.r files)</a></li>
<li class="chapter" data-level="2.9" data-path="mod-02.html"><a href="mod-02.html#manage-and-use-r-packages"><i class="fa fa-check"></i><b>2.9</b> Manage and use R packages</a>
<ul>
<li class="chapter" data-level="2.9.1" data-path="mod-02.html"><a href="mod-02.html#your-r-library"><i class="fa fa-check"></i><b>2.9.1</b> Your R library</a></li>
<li class="chapter" data-level="2.9.2" data-path="mod-02.html"><a href="mod-02.html#installing-packages-using-the-r-gui"><i class="fa fa-check"></i><b>2.9.2</b> Installing packages using the R GUI</a></li>
<li class="chapter" data-level="2.9.3" data-path="mod-02.html"><a href="mod-02.html#installing-packages-in-rstudio"><i class="fa fa-check"></i><b>2.9.3</b> Installing packages in RStudio</a></li>
<li class="chapter" data-level="2.9.4" data-path="mod-02.html"><a href="mod-02.html#installing-packages-using-the-r-console"><i class="fa fa-check"></i><b>2.9.4</b> Installing packages using the R console</a></li>
<li class="chapter" data-level="2.9.5" data-path="mod-02.html"><a href="mod-02.html#working-with-packages-in-r"><i class="fa fa-check"></i><b>2.9.5</b> Working with packages in R</a></li>
<li class="chapter" data-level="2.9.6" data-path="mod-02.html"><a href="mod-02.html#package-dependencies"><i class="fa fa-check"></i><b>2.9.6</b> Package dependencies</a></li>
<li class="chapter" data-level="2.9.7" data-path="mod-02.html"><a href="mod-02.html#citing-packages"><i class="fa fa-check"></i><b>2.9.7</b> Citing packages</a></li>
</ul></li>
<li class="chapter" data-level="2.10" data-path="mod-02.html"><a href="mod-02.html#r-documentation"><i class="fa fa-check"></i><b>2.10</b> R documentation</a>
<ul>
<li class="chapter" data-level="2.10.1" data-path="mod-02.html"><a href="mod-02.html#documentation-help-files"><i class="fa fa-check"></i><b>2.10.1</b> Documentation (help) files</a></li>
<li class="chapter" data-level="2.10.2" data-path="mod-02.html"><a href="mod-02.html#r-vignettes"><i class="fa fa-check"></i><b>2.10.2</b> R vignettes</a></li>
<li class="chapter" data-level="2.10.3" data-path="mod-02.html"><a href="mod-02.html#official-r-project-resources"><i class="fa fa-check"></i><b>2.10.3</b> Official R Project resources</a></li>
<li class="chapter" data-level="2.10.4" data-path="mod-02.html"><a href="mod-02.html#unofficial-online-resources"><i class="fa fa-check"></i><b>2.10.4</b> Unofficial online resources</a></li>
<li class="chapter" data-level="2.10.5" data-path="mod-02.html"><a href="mod-02.html#r-books"><i class="fa fa-check"></i><b>2.10.5</b> R books</a></li>
<li class="chapter" data-level="2.10.6" data-path="mod-02.html"><a href="mod-02.html#two-reminders"><i class="fa fa-check"></i><b>2.10.6</b> Two reminders</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="mod-03.html"><a href="mod-03.html"><i class="fa fa-check"></i><b>3</b> Data manipulation with R</a>
<ul>
<li class="chapter" data-level="3.1" data-path="mod-03.html"><a href="mod-03.html#mod-03-inout"><i class="fa fa-check"></i><b>3.1</b> Data import and export</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="mod-03.html"><a href="mod-03.html#importing-data-preliminaries"><i class="fa fa-check"></i><b>3.1.1</b> Importing data: preliminaries</a></li>
<li class="chapter" data-level="3.1.2" data-path="mod-03.html"><a href="mod-03.html#importing-data-from-text-files-with-read.csv-and-read.table"><i class="fa fa-check"></i><b>3.1.2</b> Importing data from text files with <code>read.csv()</code> and <code>read.table()</code></a></li>
<li class="chapter" data-level="3.1.3" data-path="mod-03.html"><a href="mod-03.html#importing-data-from-saved-workspaces"><i class="fa fa-check"></i><b>3.1.3</b> Importing data from saved workspaces</a></li>
<li class="chapter" data-level="3.1.4" data-path="mod-03.html"><a href="mod-03.html#importing-data-special-cases"><i class="fa fa-check"></i><b>3.1.4</b> Importing data: special cases:</a></li>
<li class="chapter" data-level="3.1.5" data-path="mod-03.html"><a href="mod-03.html#export-data-from-r"><i class="fa fa-check"></i><b>3.1.5</b> Export data from R</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="mod-03.html"><a href="mod-03.html#making-values-in-r"><i class="fa fa-check"></i><b>3.2</b> Making values in R</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="mod-03.html"><a href="mod-03.html#producing-arbitrary-values-with-c"><i class="fa fa-check"></i><b>3.2.1</b> Producing arbitrary values with <code>c()</code></a></li>
<li class="chapter" data-level="3.2.2" data-path="mod-03.html"><a href="mod-03.html#generating-regular-values"><i class="fa fa-check"></i><b>3.2.2</b> Generating regular values</a></li>
<li class="chapter" data-level="3.2.3" data-path="mod-03.html"><a href="mod-03.html#generating-random-values"><i class="fa fa-check"></i><b>3.2.3</b> Generating random values</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="mod-03.html"><a href="mod-03.html#selecting-data-with"><i class="fa fa-check"></i><b>3.3</b> Selecting data with <code>[]</code></a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="mod-03.html"><a href="mod-03.html#mod-03-brackets"><i class="fa fa-check"></i><b>3.3.1</b> Basics of brackets</a></li>
<li class="chapter" data-level="3.3.2" data-path="mod-03.html"><a href="mod-03.html#extracting-and-selecting-data-with-logical-tests"><i class="fa fa-check"></i><b>3.3.2</b> Extracting and selecting data with logical tests</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="mod-03.html"><a href="mod-03.html#managing-dates-and-characters"><i class="fa fa-check"></i><b>3.4</b> Managing dates and characters</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="mod-03.html"><a href="mod-03.html#temporal-data-and-dates"><i class="fa fa-check"></i><b>3.4.1</b> Temporal data and dates</a></li>
<li class="chapter" data-level="3.4.2" data-path="mod-03.html"><a href="mod-03.html#character-data-text"><i class="fa fa-check"></i><b>3.4.2</b> Character data (text)</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="mod-03.html"><a href="mod-03.html#mod-03-dataframe"><i class="fa fa-check"></i><b>3.5</b> Data frame management</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="mod-03.html"><a href="mod-03.html#data-frame-structure"><i class="fa fa-check"></i><b>3.5.1</b> Data frame structure</a></li>
<li class="chapter" data-level="3.5.2" data-path="mod-03.html"><a href="mod-03.html#common-data-frame-operations"><i class="fa fa-check"></i><b>3.5.2</b> Common data frame operations</a></li>
<li class="chapter" data-level="3.5.3" data-path="mod-03.html"><a href="mod-03.html#other-data-frame-operations"><i class="fa fa-check"></i><b>3.5.3</b> Other data frame operations</a></li>
<li class="chapter" data-level="3.5.4" data-path="mod-03.html"><a href="mod-03.html#mod-03-reshape"><i class="fa fa-check"></i><b>3.5.4</b> Reshaping data frames</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="mod-04.html"><a href="mod-04.html"><i class="fa fa-check"></i><b>4</b> Exploratory data analysis 1: Data description and summarization</a>
<ul>
<li class="chapter" data-level="4.1" data-path="mod-04.html"><a href="mod-04.html#motivation"><i class="fa fa-check"></i><b>4.1</b> Motivation</a></li>
<li class="chapter" data-level="4.2" data-path="mod-04.html"><a href="mod-04.html#descriptive-and-summary-statistics"><i class="fa fa-check"></i><b>4.2</b> Descriptive and summary statistics</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="mod-04.html"><a href="mod-04.html#basic-summary-statistics"><i class="fa fa-check"></i><b>4.2.1</b> Basic summary statistics</a></li>
<li class="chapter" data-level="4.2.2" data-path="mod-04.html"><a href="mod-04.html#mod-04-agg"><i class="fa fa-check"></i><b>4.2.2</b> Summarizing by groups (AKA: pivot tables)</a></li>
<li class="chapter" data-level="4.2.3" data-path="mod-04.html"><a href="mod-04.html#summarizing-data-with-the-apply-family"><i class="fa fa-check"></i><b>4.2.3</b> Summarizing data with the <code>apply()</code> family</a></li>
<li class="chapter" data-level="4.2.4" data-path="mod-04.html"><a href="mod-04.html#mod-04-tab"><i class="fa fa-check"></i><b>4.2.4</b> Tabulation (frequency tables)</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="mod-04.html"><a href="mod-04.html#mod-04-vis"><i class="fa fa-check"></i><b>4.3</b> Visualizing single variables</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="mod-04.html"><a href="mod-04.html#boxplots-aka-box-and-whisker-plots"><i class="fa fa-check"></i><b>4.3.1</b> Boxplots (aka: box-and-whisker plots)</a></li>
<li class="chapter" data-level="4.3.2" data-path="mod-04.html"><a href="mod-04.html#histograms"><i class="fa fa-check"></i><b>4.3.2</b> Histograms</a></li>
<li class="chapter" data-level="4.3.3" data-path="mod-04.html"><a href="mod-04.html#mod04-pdfplots"><i class="fa fa-check"></i><b>4.3.3</b> Kernel density plots</a></li>
<li class="chapter" data-level="4.3.4" data-path="mod-04.html"><a href="mod-04.html#empirical-cumulative-density-function-ecdf-plots"><i class="fa fa-check"></i><b>4.3.4</b> Empirical cumulative density function (ECDF) plots</a></li>
<li class="chapter" data-level="4.3.5" data-path="mod-04.html"><a href="mod-04.html#quantile-quantile-qq-plots"><i class="fa fa-check"></i><b>4.3.5</b> Quantile-quantile (QQ) plots</a></li>
<li class="chapter" data-level="4.3.6" data-path="mod-04.html"><a href="mod-04.html#how-should-i-plot-my-data"><i class="fa fa-check"></i><b>4.3.6</b> How should I plot my data?</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="mod-05.html"><a href="mod-05.html"><i class="fa fa-check"></i><b>5</b> Exploratory data analysis 2: Probability distributions for biology</a>
<ul>
<li class="chapter" data-level="5.1" data-path="mod-05.html"><a href="mod-05.html#mod-05-dists1"><i class="fa fa-check"></i><b>5.1</b> Statistical distributions</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="mod-05.html"><a href="mod-05.html#probability-distributions-in-r"><i class="fa fa-check"></i><b>5.1.1</b> Probability distributions in R</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="mod-05.html"><a href="mod-05.html#mod-05-discrete"><i class="fa fa-check"></i><b>5.2</b> Discrete distributions</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="mod-05.html"><a href="mod-05.html#bernoulli-distribution"><i class="fa fa-check"></i><b>5.2.1</b> Bernoulli distribution</a></li>
<li class="chapter" data-level="5.2.2" data-path="mod-05.html"><a href="mod-05.html#mod-05-binom"><i class="fa fa-check"></i><b>5.2.2</b> Binomial distribution</a></li>
<li class="chapter" data-level="5.2.3" data-path="mod-05.html"><a href="mod-05.html#poisson-distribution"><i class="fa fa-check"></i><b>5.2.3</b> Poisson distribution</a></li>
<li class="chapter" data-level="5.2.4" data-path="mod-05.html"><a href="mod-05.html#mod-05-negbin"><i class="fa fa-check"></i><b>5.2.4</b> Negative binomial distribution</a></li>
<li class="chapter" data-level="5.2.5" data-path="mod-05.html"><a href="mod-05.html#mod-05-geom"><i class="fa fa-check"></i><b>5.2.5</b> Geometric distribution</a></li>
<li class="chapter" data-level="5.2.6" data-path="mod-05.html"><a href="mod-05.html#beta-binomial-distribution"><i class="fa fa-check"></i><b>5.2.6</b> Beta-binomial distribution</a></li>
<li class="chapter" data-level="5.2.7" data-path="mod-05.html"><a href="mod-05.html#multinomial-distribution"><i class="fa fa-check"></i><b>5.2.7</b> Multinomial distribution</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="mod-05.html"><a href="mod-05.html#mod-05-condists"><i class="fa fa-check"></i><b>5.3</b> Continuous distributions</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="mod-05.html"><a href="mod-05.html#uniform-distribution"><i class="fa fa-check"></i><b>5.3.1</b> Uniform distribution</a></li>
<li class="chapter" data-level="5.3.2" data-path="mod-05.html"><a href="mod-05.html#mod-05-norm"><i class="fa fa-check"></i><b>5.3.2</b> Normal distribution</a></li>
<li class="chapter" data-level="5.3.3" data-path="mod-05.html"><a href="mod-05.html#lognormal-distribution"><i class="fa fa-check"></i><b>5.3.3</b> Lognormal distribution</a></li>
<li class="chapter" data-level="5.3.4" data-path="mod-05.html"><a href="mod-05.html#mod-05-gamma"><i class="fa fa-check"></i><b>5.3.4</b> Gamma distribution</a></li>
<li class="chapter" data-level="5.3.5" data-path="mod-05.html"><a href="mod-05.html#beta-distribution"><i class="fa fa-check"></i><b>5.3.5</b> Beta distribution</a></li>
<li class="chapter" data-level="5.3.6" data-path="mod-05.html"><a href="mod-05.html#dirichlet-distribution"><i class="fa fa-check"></i><b>5.3.6</b> Dirichlet distribution</a></li>
<li class="chapter" data-level="5.3.7" data-path="mod-05.html"><a href="mod-05.html#mod-05-expo"><i class="fa fa-check"></i><b>5.3.7</b> Exponential distribution</a></li>
<li class="chapter" data-level="5.3.8" data-path="mod-05.html"><a href="mod-05.html#triangular-distribution"><i class="fa fa-check"></i><b>5.3.8</b> Triangular distribution</a></li>
<li class="chapter" data-level="5.3.9" data-path="mod-05.html"><a href="mod-05.html#distributions-summary"><i class="fa fa-check"></i><b>5.3.9</b> Distributions summary</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="mod-05.html"><a href="mod-05.html#mod-05-dists2"><i class="fa fa-check"></i><b>5.4</b> Fitting and testing distributions</a>
<ul>
<li class="chapter" data-level="5.4.1" data-path="mod-05.html"><a href="mod-05.html#estimating-distributional-parameters"><i class="fa fa-check"></i><b>5.4.1</b> Estimating distributional parameters</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="mod-05.html"><a href="mod-05.html#graphical-methods-for-examining-distributions"><i class="fa fa-check"></i><b>5.5</b> Graphical methods for examining distributions</a>
<ul>
<li class="chapter" data-level="5.5.1" data-path="mod-05.html"><a href="mod-05.html#comparing-distributions-with-pdf-or-pmf-plots"><i class="fa fa-check"></i><b>5.5.1</b> Comparing distributions with PDF or PMF plots</a></li>
<li class="chapter" data-level="5.5.2" data-path="mod-05.html"><a href="mod-05.html#comparing-distributions-using-cdf-and-ecdf"><i class="fa fa-check"></i><b>5.5.2</b> Comparing distributions using CDF and ECDF</a></li>
<li class="chapter" data-level="5.5.3" data-path="mod-05.html"><a href="mod-05.html#comparing-distributions-using-qq-plots"><i class="fa fa-check"></i><b>5.5.3</b> Comparing distributions using QQ plots</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="mod-05.html"><a href="mod-05.html#formal-tests-for-distributions"><i class="fa fa-check"></i><b>5.6</b> Formal tests for distributions</a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="mod-05.html"><a href="mod-05.html#tests-for-normality"><i class="fa fa-check"></i><b>5.6.1</b> Tests for normality</a></li>
<li class="chapter" data-level="5.6.2" data-path="mod-05.html"><a href="mod-05.html#kolmogorov-smirnov-tests"><i class="fa fa-check"></i><b>5.6.2</b> Kolmogorov-Smirnov tests</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="mod-05.html"><a href="mod-05.html#mod-05-trans"><i class="fa fa-check"></i><b>5.7</b> Data transformations</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="mod-05.html"><a href="mod-05.html#why-transform"><i class="fa fa-check"></i><b>5.7.1</b> Why transform?</a></li>
<li class="chapter" data-level="5.7.2" data-path="mod-05.html"><a href="mod-05.html#transforms-vs.-link-functions"><i class="fa fa-check"></i><b>5.7.2</b> Transforms vs. link functions</a></li>
<li class="chapter" data-level="5.7.3" data-path="mod-05.html"><a href="mod-05.html#common-transformations"><i class="fa fa-check"></i><b>5.7.3</b> Common transformations</a></li>
<li class="chapter" data-level="5.7.4" data-path="mod-05.html"><a href="mod-05.html#uncommon-transformations"><i class="fa fa-check"></i><b>5.7.4</b> Uncommon transformations</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="mod-06.html"><a href="mod-06.html"><i class="fa fa-check"></i><b>6</b> Exploratory data analysis 3: Multivariate data exploration</a>
<ul>
<li class="chapter" data-level="6.1" data-path="mod-06.html"><a href="mod-06.html#scatterplots-for-2-continuous-variables"><i class="fa fa-check"></i><b>6.1</b> Scatterplots for 2 continuous variables</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="mod-06.html"><a href="mod-06.html#r-plotting-basics"><i class="fa fa-check"></i><b>6.1.1</b> R plotting basics</a></li>
<li class="chapter" data-level="6.1.2" data-path="mod-06.html"><a href="mod-06.html#graphical-parameters-and-advanced-plots-with-par"><i class="fa fa-check"></i><b>6.1.2</b> Graphical parameters and advanced plots with <code>par()</code></a></li>
<li class="chapter" data-level="6.1.3" data-path="mod-06.html"><a href="mod-06.html#really-advanced-plotting"><i class="fa fa-check"></i><b>6.1.3</b> Really advanced plotting</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="mod-06.html"><a href="mod-06.html#mod-06-smat"><i class="fa fa-check"></i><b>6.2</b> Scatterplot matrices for many variables</a></li>
<li class="chapter" data-level="6.3" data-path="mod-06.html"><a href="mod-06.html#lattice-plots-for-hierarchical-data"><i class="fa fa-check"></i><b>6.3</b> Lattice plots for hierarchical data</a></li>
<li class="chapter" data-level="6.4" data-path="mod-06.html"><a href="mod-06.html#continuous-variables-by-group"><i class="fa fa-check"></i><b>6.4</b> Continuous variables by group</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="mod-06.html"><a href="mod-06.html#boxplots"><i class="fa fa-check"></i><b>6.4.1</b> Boxplots</a></li>
<li class="chapter" data-level="6.4.2" data-path="mod-06.html"><a href="mod-06.html#barplots"><i class="fa fa-check"></i><b>6.4.2</b> Barplots</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="mod-06.html"><a href="mod-06.html#ordination-brief-introduction"><i class="fa fa-check"></i><b>6.5</b> Ordination (brief introduction)</a>
<ul>
<li class="chapter" data-level="6.5.1" data-path="mod-06.html"><a href="mod-06.html#principal-components-analysis-pca"><i class="fa fa-check"></i><b>6.5.1</b> Principal components analysis (PCA)</a></li>
<li class="chapter" data-level="6.5.2" data-path="mod-06.html"><a href="mod-06.html#nonmetric-multidimensional-scaling-nmds"><i class="fa fa-check"></i><b>6.5.2</b> Nonmetric multidimensional scaling (NMDS)</a></li>
<li class="chapter" data-level="6.5.3" data-path="mod-06.html"><a href="mod-06.html#plotting-ordinations"><i class="fa fa-check"></i><b>6.5.3</b> Plotting ordinations</a></li>
<li class="chapter" data-level="6.5.4" data-path="mod-06.html"><a href="mod-06.html#ordination-wrap-up-for-now"><i class="fa fa-check"></i><b>6.5.4</b> Ordination wrap-up (for now)</a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="mod-06.html"><a href="mod-06.html#principles-for-good-data-graphics"><i class="fa fa-check"></i><b>6.6</b> Principles for good data graphics</a>
<ul>
<li class="chapter" data-level="6.6.1" data-path="mod-06.html"><a href="mod-06.html#maximize-the-data-to-ink-ratio"><i class="fa fa-check"></i><b>6.6.1</b> Maximize the data-to-ink ratio</a></li>
<li class="chapter" data-level="6.6.2" data-path="mod-06.html"><a href="mod-06.html#show-variability"><i class="fa fa-check"></i><b>6.6.2</b> Show variability</a></li>
<li class="chapter" data-level="6.6.3" data-path="mod-06.html"><a href="mod-06.html#use-labels-not-symbols"><i class="fa fa-check"></i><b>6.6.3</b> Use labels, not symbols</a></li>
<li class="chapter" data-level="6.6.4" data-path="mod-06.html"><a href="mod-06.html#be-consistent"><i class="fa fa-check"></i><b>6.6.4</b> Be consistent</a></li>
<li class="chapter" data-level="6.6.5" data-path="mod-06.html"><a href="mod-06.html#set-yourself-up-for-success"><i class="fa fa-check"></i><b>6.6.5</b> Set yourself up for success</a></li>
<li class="chapter" data-level="6.6.6" data-path="mod-06.html"><a href="mod-06.html#consider-legibility"><i class="fa fa-check"></i><b>6.6.6</b> Consider legibility</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="mod-07.html"><a href="mod-07.html"><i class="fa fa-check"></i><b>7</b> Common statistical problems</a>
<ul>
<li class="chapter" data-level="7.1" data-path="mod-07.html"><a href="mod-07.html#outliers-and-erroneous-values"><i class="fa fa-check"></i><b>7.1</b> Outliers and erroneous values</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="mod-07.html"><a href="mod-07.html#detecting-outliers"><i class="fa fa-check"></i><b>7.1.1</b> Detecting outliers</a></li>
<li class="chapter" data-level="7.1.2" data-path="mod-07.html"><a href="mod-07.html#dealing-with-outliers"><i class="fa fa-check"></i><b>7.1.2</b> Dealing with outliers</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="mod-07.html"><a href="mod-07.html#autocorrelation"><i class="fa fa-check"></i><b>7.2</b> Autocorrelation</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="mod-07.html"><a href="mod-07.html#graphical-methods-to-identify-autocorrelation"><i class="fa fa-check"></i><b>7.2.1</b> Graphical methods to identify autocorrelation</a></li>
<li class="chapter" data-level="7.2.2" data-path="mod-07.html"><a href="mod-07.html#mantel-tests-for-autocorrelation"><i class="fa fa-check"></i><b>7.2.2</b> Mantel tests for autocorrelation</a></li>
<li class="chapter" data-level="7.2.3" data-path="mod-07.html"><a href="mod-07.html#dealing-with-autocorrelation"><i class="fa fa-check"></i><b>7.2.3</b> Dealing with autocorrelation</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="mod-07.html"><a href="mod-07.html#mod-07-multicol"><i class="fa fa-check"></i><b>7.3</b> Collinearity</a></li>
<li class="chapter" data-level="7.4" data-path="mod-07.html"><a href="mod-07.html#missing-data"><i class="fa fa-check"></i><b>7.4</b> Missing data</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="mod-07.html"><a href="mod-07.html#option-1-ignore-missing-data"><i class="fa fa-check"></i><b>7.4.1</b> Option 1: Ignore missing data</a></li>
<li class="chapter" data-level="7.4.2" data-path="mod-07.html"><a href="mod-07.html#option-2-interpolate-or-impute-missing-data"><i class="fa fa-check"></i><b>7.4.2</b> Option 2: Interpolate or impute missing data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="mod-08.html"><a href="mod-08.html"><i class="fa fa-check"></i><b>8</b> Planning your analysis (what test?)</a>
<ul>
<li class="chapter" data-level="8.1" data-path="mod-08.html"><a href="mod-08.html#how-to-use-this-guide"><i class="fa fa-check"></i><b>8.1</b> How to use this guide</a></li>
<li class="chapter" data-level="8.2" data-path="mod-08.html"><a href="mod-08.html#what-question-are-you-trying-to-answer"><i class="fa fa-check"></i><b>8.2</b> What question are you trying to answer?</a></li>
<li class="chapter" data-level="8.3" data-path="mod-08.html"><a href="mod-08.html#testing-for-a-difference-in-mean-or-location"><i class="fa fa-check"></i><b>8.3</b> Testing for a difference in mean or location</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="mod-08.html"><a href="mod-08.html#additional-considerations"><i class="fa fa-check"></i><b>8.3.1</b> Additional considerations</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="mod-08.html"><a href="mod-08.html#testing-for-a-continuous-relationship-between-two-or-more-variables"><i class="fa fa-check"></i><b>8.4</b> Testing for a continuous relationship between two or more variables</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="mod-08.html"><a href="mod-08.html#additional-considerations-1"><i class="fa fa-check"></i><b>8.4.1</b> Additional considerations</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="mod-08.html"><a href="mod-08.html#independence-between-categories"><i class="fa fa-check"></i><b>8.5</b> Independence between categories</a></li>
<li class="chapter" data-level="8.6" data-path="mod-08.html"><a href="mod-08.html#classifying-observations"><i class="fa fa-check"></i><b>8.6</b> Classifying observations</a></li>
<li class="chapter" data-level="8.7" data-path="mod-08.html"><a href="mod-08.html#conclusions"><i class="fa fa-check"></i><b>8.7</b> Conclusions</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="mod-09.html"><a href="mod-09.html"><i class="fa fa-check"></i><b>9</b> Basic statistics with R</a>
<ul>
<li class="chapter" data-level="9.1" data-path="mod-09.html"><a href="mod-09.html#mod-09-datasets"><i class="fa fa-check"></i><b>9.1</b> Datasets and variables</a></li>
<li class="chapter" data-level="9.2" data-path="mod-09.html"><a href="mod-09.html#mod-09-frequentist"><i class="fa fa-check"></i><b>9.2</b> Frequentist statistics</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="mod-09.html"><a href="mod-09.html#mod-09-pvalues"><i class="fa fa-check"></i><b>9.2.1</b> <em>P</em>-values and NHST</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="mod-09.html"><a href="mod-09.html#mod-09-whattest"><i class="fa fa-check"></i><b>9.3</b> What kind of test do you need?</a></li>
<li class="chapter" data-level="9.4" data-path="mod-09.html"><a href="mod-09.html#mod-09-tests-for-loc"><i class="fa fa-check"></i><b>9.4</b> Tests for differences in mean or location</a>
<ul>
<li class="chapter" data-level="9.4.1" data-path="mod-09.html"><a href="mod-09.html#tests-for-1-group"><i class="fa fa-check"></i><b>9.4.1</b> Tests for 1 group</a></li>
<li class="chapter" data-level="9.4.2" data-path="mod-09.html"><a href="mod-09.html#tests-comparing-two-groups"><i class="fa fa-check"></i><b>9.4.2</b> Tests comparing two groups</a></li>
<li class="chapter" data-level="9.4.3" data-path="mod-09.html"><a href="mod-09.html#aside-whats-with-all-these-t-tests"><i class="fa fa-check"></i><b>9.4.3</b> Aside: what’s with all these <em>t</em>-tests?</a></li>
<li class="chapter" data-level="9.4.4" data-path="mod-09.html"><a href="mod-09.html#comparing-3-or-more-groups"><i class="fa fa-check"></i><b>9.4.4</b> Comparing 3 or more groups</a></li>
<li class="chapter" data-level="9.4.5" data-path="mod-09.html"><a href="mod-09.html#polynomial-patterns-in-grouped-data"><i class="fa fa-check"></i><b>9.4.5</b> Polynomial patterns in grouped data</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="mod-09.html"><a href="mod-09.html#tests-for-patterns-in-continuous-data"><i class="fa fa-check"></i><b>9.5</b> Tests for patterns in continuous data</a>
<ul>
<li class="chapter" data-level="9.5.1" data-path="mod-09.html"><a href="mod-09.html#mod-09-correlation"><i class="fa fa-check"></i><b>9.5.1</b> Correlation</a></li>
<li class="chapter" data-level="9.5.2" data-path="mod-09.html"><a href="mod-09.html#mod-09-linear-mods"><i class="fa fa-check"></i><b>9.5.2</b> Linear models</a></li>
<li class="chapter" data-level="9.5.3" data-path="mod-09.html"><a href="mod-09.html#linear-models-wrap-up-for-now"><i class="fa fa-check"></i><b>9.5.3</b> Linear models wrap-up (for now)</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="mod-09.html"><a href="mod-09.html#mod-09-contingency"><i class="fa fa-check"></i><b>9.6</b> Tests for proportions and contingency</a>
<ul>
<li class="chapter" data-level="9.6.1" data-path="mod-09.html"><a href="mod-09.html#mod-09-ctables"><i class="fa fa-check"></i><b>9.6.1</b> Contingency tables</a></li>
<li class="chapter" data-level="9.6.2" data-path="mod-09.html"><a href="mod-09.html#tests-on-contingency-tables"><i class="fa fa-check"></i><b>9.6.2</b> Tests on contingency tables</a></li>
<li class="chapter" data-level="9.6.3" data-path="mod-09.html"><a href="mod-09.html#mod-09-oddsratios"><i class="fa fa-check"></i><b>9.6.3</b> Contingency tests and odds ratios</a></li>
</ul></li>
<li class="chapter" data-level="9.7" data-path="mod-09.html"><a href="mod-09.html#mod-09-classification"><i class="fa fa-check"></i><b>9.7</b> Classification</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="mod-10.html"><a href="mod-10.html"><i class="fa fa-check"></i><b>10</b> Generalized linear models (GLM)</a>
<ul>
<li class="chapter" data-level="10.1" data-path="mod-10.html"><a href="mod-10.html#mod-10-lm"><i class="fa fa-check"></i><b>10.1</b> Prelude with linear models</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="mod-10.html"><a href="mod-10.html#assumptions-of-linear-models"><i class="fa fa-check"></i><b>10.1.1</b> Assumptions of linear models</a></li>
<li class="chapter" data-level="10.1.2" data-path="mod-10.html"><a href="mod-10.html#linear-regression-in-r"><i class="fa fa-check"></i><b>10.1.2</b> Linear regression in R</a></li>
<li class="chapter" data-level="10.1.3" data-path="mod-10.html"><a href="mod-10.html#multiple-linear-regression"><i class="fa fa-check"></i><b>10.1.3</b> Multiple linear regression</a></li>
<li class="chapter" data-level="10.1.4" data-path="mod-10.html"><a href="mod-10.html#mod-10-anova"><i class="fa fa-check"></i><b>10.1.4</b> ANOVA and ANCOVA with <code>lm()</code></a></li>
<li class="chapter" data-level="10.1.5" data-path="mod-10.html"><a href="mod-10.html#variations-on-linear-models"><i class="fa fa-check"></i><b>10.1.5</b> Variations on linear models</a></li>
<li class="chapter" data-level="10.1.6" data-path="mod-10.html"><a href="mod-10.html#example-linear-regression-workflow"><i class="fa fa-check"></i><b>10.1.6</b> Example linear regression workflow</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="mod-10.html"><a href="mod-10.html#mod-10-basic"><i class="fa fa-check"></i><b>10.2</b> GLM basics</a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="mod-10.html"><a href="mod-10.html#example-glms"><i class="fa fa-check"></i><b>10.2.1</b> Example GLMS</a></li>
<li class="chapter" data-level="10.2.2" data-path="mod-10.html"><a href="mod-10.html#glm-families"><i class="fa fa-check"></i><b>10.2.2</b> GLM families</a></li>
<li class="chapter" data-level="10.2.3" data-path="mod-10.html"><a href="mod-10.html#glm-link-functions"><i class="fa fa-check"></i><b>10.2.3</b> GLM link functions</a></li>
<li class="chapter" data-level="10.2.4" data-path="mod-10.html"><a href="mod-10.html#deviance-and-other-glm-diagnostics"><i class="fa fa-check"></i><b>10.2.4</b> Deviance and other GLM diagnostics</a></li>
<li class="chapter" data-level="10.2.5" data-path="mod-10.html"><a href="mod-10.html#to-pseudo-r2-or-not-to-pseudo-r2"><i class="fa fa-check"></i><b>10.2.5</b> To pseudo-<em>R</em><sup>2</sup> or not to pseudo-<em>R</em><sup>2</sup>?</a></li>
<li class="chapter" data-level="10.2.6" data-path="mod-10.html"><a href="mod-10.html#common-glms"><i class="fa fa-check"></i><b>10.2.6</b> Common GLMs</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="mod-10.html"><a href="mod-10.html#mod-10-loglin"><i class="fa fa-check"></i><b>10.3</b> Log-linear models</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="mod-10.html"><a href="mod-10.html#mod-10-loglin-examp"><i class="fa fa-check"></i><b>10.3.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="10.3.2" data-path="mod-10.html"><a href="mod-10.html#example-with-real-data"><i class="fa fa-check"></i><b>10.3.2</b> Example with real data</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="mod-10.html"><a href="mod-10.html#mod-10-poisson"><i class="fa fa-check"></i><b>10.4</b> Poisson GLM for counts</a>
<ul>
<li class="chapter" data-level="10.4.1" data-path="mod-10.html"><a href="mod-10.html#example-with-simulated-data"><i class="fa fa-check"></i><b>10.4.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="10.4.2" data-path="mod-10.html"><a href="mod-10.html#example-with-real-data-1"><i class="fa fa-check"></i><b>10.4.2</b> Example with real data</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="mod-10.html"><a href="mod-10.html#mod-10-quasi"><i class="fa fa-check"></i><b>10.5</b> Quasi-Poisson and negative binomial GLM</a>
<ul>
<li class="chapter" data-level="10.5.1" data-path="mod-10.html"><a href="mod-10.html#example-with-simulated-data-1"><i class="fa fa-check"></i><b>10.5.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="10.5.2" data-path="mod-10.html"><a href="mod-10.html#example-with-real-data-2"><i class="fa fa-check"></i><b>10.5.2</b> Example with real data</a></li>
</ul></li>
<li class="chapter" data-level="10.6" data-path="mod-10.html"><a href="mod-10.html#mod-10-logistic"><i class="fa fa-check"></i><b>10.6</b> Logistic regression for binary outcomes</a>
<ul>
<li class="chapter" data-level="10.6.1" data-path="mod-10.html"><a href="mod-10.html#example-with-simulated-data-2"><i class="fa fa-check"></i><b>10.6.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="10.6.2" data-path="mod-10.html"><a href="mod-10.html#example-with-real-data-3"><i class="fa fa-check"></i><b>10.6.2</b> Example with real data</a></li>
<li class="chapter" data-level="10.6.3" data-path="mod-10.html"><a href="mod-10.html#mod-10-auc"><i class="fa fa-check"></i><b>10.6.3</b> Logistic GLM diagnostics: AUC and ROC</a></li>
</ul></li>
<li class="chapter" data-level="10.7" data-path="mod-10.html"><a href="mod-10.html#mod-10-binom"><i class="fa fa-check"></i><b>10.7</b> Binomial GLM for proportional data</a>
<ul>
<li class="chapter" data-level="10.7.1" data-path="mod-10.html"><a href="mod-10.html#binomial-glm"><i class="fa fa-check"></i><b>10.7.1</b> Binomial GLM</a></li>
<li class="chapter" data-level="10.7.2" data-path="mod-10.html"><a href="mod-10.html#example-with-simulated-data-3"><i class="fa fa-check"></i><b>10.7.2</b> Example with simulated data</a></li>
</ul></li>
<li class="chapter" data-level="10.8" data-path="mod-10.html"><a href="mod-10.html#mod-10-gamma"><i class="fa fa-check"></i><b>10.8</b> Gamma models for overdispersed data</a>
<ul>
<li class="chapter" data-level="10.8.1" data-path="mod-10.html"><a href="mod-10.html#example-with-simulated-data-4"><i class="fa fa-check"></i><b>10.8.1</b> Example with simulated data</a></li>
</ul></li>
<li class="chapter" data-level="10.9" data-path="mod-10.html"><a href="mod-10.html#mod-10-beyond"><i class="fa fa-check"></i><b>10.9</b> Beyond GLM: Overview of GAM and GEE</a>
<ul>
<li class="chapter" data-level="10.9.1" data-path="mod-10.html"><a href="mod-10.html#mod-10-gam"><i class="fa fa-check"></i><b>10.9.1</b> Generalized additive models (GAM)</a></li>
<li class="chapter" data-level="10.9.2" data-path="mod-10.html"><a href="mod-10.html#mod-10-gee"><i class="fa fa-check"></i><b>10.9.2</b> Generalized estimating equations (GEE)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="mod-11.html"><a href="mod-11.html"><i class="fa fa-check"></i><b>11</b> Nonlinear models</a>
<ul>
<li class="chapter" data-level="11.1" data-path="mod-11.html"><a href="mod-11.html#background"><i class="fa fa-check"></i><b>11.1</b> Background</a></li>
<li class="chapter" data-level="11.2" data-path="mod-11.html"><a href="mod-11.html#mod-11-intro"><i class="fa fa-check"></i><b>11.2</b> Nonlinear least squares (NLS)</a></li>
<li class="chapter" data-level="11.3" data-path="mod-11.html"><a href="mod-11.html#mod-11-micmen"><i class="fa fa-check"></i><b>11.3</b> Michaelis-Menten curves</a>
<ul>
<li class="chapter" data-level="11.3.1" data-path="mod-11.html"><a href="mod-11.html#mod-11-micmen-sim"><i class="fa fa-check"></i><b>11.3.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="11.3.2" data-path="mod-11.html"><a href="mod-11.html#example-with-real-data-4"><i class="fa fa-check"></i><b>11.3.2</b> Example with real data</a></li>
<li class="chapter" data-level="11.3.3" data-path="mod-11.html"><a href="mod-11.html#alternative-strategies-for-the-analysis"><i class="fa fa-check"></i><b>11.3.3</b> Alternative strategies for the analysis</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="mod-11.html"><a href="mod-11.html#mod-11-grow"><i class="fa fa-check"></i><b>11.4</b> Biological growth curves</a>
<ul>
<li class="chapter" data-level="11.4.1" data-path="mod-11.html"><a href="mod-11.html#gompertz-and-von-bertalanffy-curves"><i class="fa fa-check"></i><b>11.4.1</b> Gompertz and von Bertalanffy curves</a></li>
<li class="chapter" data-level="11.4.2" data-path="mod-11.html"><a href="mod-11.html#example-with-real-data-5"><i class="fa fa-check"></i><b>11.4.2</b> Example with real data</a></li>
</ul></li>
<li class="chapter" data-level="11.5" data-path="mod-11.html"><a href="mod-11.html#dose-response-curves"><i class="fa fa-check"></i><b>11.5</b> Dose response curves</a></li>
<li class="chapter" data-level="11.6" data-path="mod-11.html"><a href="mod-11.html#alternatives-to-nls"><i class="fa fa-check"></i><b>11.6</b> Alternatives to NLS</a>
<ul>
<li class="chapter" data-level="11.6.1" data-path="mod-11.html"><a href="mod-11.html#generalized-nonlinear-models"><i class="fa fa-check"></i><b>11.6.1</b> Generalized nonlinear models</a></li>
<li class="chapter" data-level="11.6.2" data-path="mod-11.html"><a href="mod-11.html#quantile-regression"><i class="fa fa-check"></i><b>11.6.2</b> Quantile regression</a></li>
<li class="chapter" data-level="11.6.3" data-path="mod-11.html"><a href="mod-11.html#mod-11-gam"><i class="fa fa-check"></i><b>11.6.3</b> Generalized additive models (GAM)</a></li>
<li class="chapter" data-level="11.6.4" data-path="mod-11.html"><a href="mod-11.html#classification-and-regression-trees-cart"><i class="fa fa-check"></i><b>11.6.4</b> Classification and regression trees (CART)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="12" data-path="mod-12.html"><a href="mod-12.html"><i class="fa fa-check"></i><b>12</b> Mixed models</a>
<ul>
<li class="chapter" data-level="12.1" data-path="mod-12.html"><a href="mod-12.html#prelude-glm"><i class="fa fa-check"></i><b>12.1</b> Prelude (GLM)</a></li>
<li class="chapter" data-level="12.2" data-path="mod-12.html"><a href="mod-12.html#mod-12-lmm"><i class="fa fa-check"></i><b>12.2</b> Linear mixed models (LMM)</a>
<ul>
<li class="chapter" data-level="12.2.1" data-path="mod-12.html"><a href="mod-12.html#formal-definition-and-example"><i class="fa fa-check"></i><b>12.2.1</b> Formal definition and example</a></li>
<li class="chapter" data-level="12.2.2" data-path="mod-12.html"><a href="mod-12.html#example-with-simulated-data-5"><i class="fa fa-check"></i><b>12.2.2</b> Example with simulated data</a></li>
<li class="chapter" data-level="12.2.3" data-path="mod-12.html"><a href="mod-12.html#p-values-in-lmm"><i class="fa fa-check"></i><b>12.2.3</b> <em>P</em>-values in LMM</a></li>
<li class="chapter" data-level="12.2.4" data-path="mod-12.html"><a href="mod-12.html#specifying-random-effects"><i class="fa fa-check"></i><b>12.2.4</b> Specifying random effects</a></li>
<li class="chapter" data-level="12.2.5" data-path="mod-12.html"><a href="mod-12.html#lmm-example-with-real-data"><i class="fa fa-check"></i><b>12.2.5</b> LMM example with real data</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="mod-12.html"><a href="mod-12.html#mod-12-glmm"><i class="fa fa-check"></i><b>12.3</b> Generalized linear mixed models (GLMM)</a>
<ul>
<li class="chapter" data-level="12.3.1" data-path="mod-12.html"><a href="mod-12.html#definition-1"><i class="fa fa-check"></i><b>12.3.1</b> Definition</a></li>
<li class="chapter" data-level="12.3.2" data-path="mod-12.html"><a href="mod-12.html#glmm-on-simulated-data"><i class="fa fa-check"></i><b>12.3.2</b> GLMM on simulated data</a></li>
<li class="chapter" data-level="12.3.3" data-path="mod-12.html"><a href="mod-12.html#glmm-on-real-data"><i class="fa fa-check"></i><b>12.3.3</b> GLMM on real data</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="mod-12.html"><a href="mod-12.html#mod-12-nlme"><i class="fa fa-check"></i><b>12.4</b> Nonlinear mixed models (NLME)</a>
<ul>
<li class="chapter" data-level="12.4.1" data-path="mod-12.html"><a href="mod-12.html#definition-and-background"><i class="fa fa-check"></i><b>12.4.1</b> Definition and background</a></li>
<li class="chapter" data-level="12.4.2" data-path="mod-12.html"><a href="mod-12.html#nlme-on-simulated-data"><i class="fa fa-check"></i><b>12.4.2</b> NLME on simulated data</a></li>
<li class="chapter" data-level="12.4.3" data-path="mod-12.html"><a href="mod-12.html#nlme-on-real-data"><i class="fa fa-check"></i><b>12.4.3</b> NLME on real data</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="mod-12.html"><a href="mod-12.html#mod-12-gamm"><i class="fa fa-check"></i><b>12.5</b> (Generalized) additive mixed models (AMM/GAMM)</a>
<ul>
<li class="chapter" data-level="12.5.1" data-path="mod-12.html"><a href="mod-12.html#example-gamm-with-real-data"><i class="fa fa-check"></i><b>12.5.1</b> Example GAMM with real data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="13" data-path="mod-13.html"><a href="mod-13.html"><i class="fa fa-check"></i><b>13</b> Multivariate data analysis</a>
<ul>
<li class="chapter" data-level="13.1" data-path="mod-13.html"><a href="mod-13.html#multivariate-data"><i class="fa fa-check"></i><b>13.1</b> Multivariate data</a>
<ul>
<li class="chapter" data-level="13.1.1" data-path="mod-13.html"><a href="mod-13.html#univariate-vs.-multivariate-data"><i class="fa fa-check"></i><b>13.1.1</b> Univariate vs. multivariate data</a></li>
<li class="chapter" data-level="13.1.2" data-path="mod-13.html"><a href="mod-13.html#components-of-multivariate-data"><i class="fa fa-check"></i><b>13.1.2</b> Components of multivariate data</a></li>
</ul></li>
<li class="chapter" data-level="13.2" data-path="mod-13.html"><a href="mod-13.html#mod-13-dist"><i class="fa fa-check"></i><b>13.2</b> Distance metrics: biological (dis)similarity</a>
<ul>
<li class="chapter" data-level="13.2.1" data-path="mod-13.html"><a href="mod-13.html#euclidean-distance"><i class="fa fa-check"></i><b>13.2.1</b> Euclidean distance</a></li>
<li class="chapter" data-level="13.2.2" data-path="mod-13.html"><a href="mod-13.html#bray-curtis-and-other-distance-metrics"><i class="fa fa-check"></i><b>13.2.2</b> Bray-Curtis and other distance metrics</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="mod-13.html"><a href="mod-13.html#clustering"><i class="fa fa-check"></i><b>13.3</b> Clustering</a>
<ul>
<li class="chapter" data-level="13.3.1" data-path="mod-13.html"><a href="mod-13.html#k-means-clustering"><i class="fa fa-check"></i><b>13.3.1</b> <em>K</em>-means clustering</a></li>
<li class="chapter" data-level="13.3.2" data-path="mod-13.html"><a href="mod-13.html#hierarchical-agglomerative-clustering"><i class="fa fa-check"></i><b>13.3.2</b> Hierarchical agglomerative clustering</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="mod-13.html"><a href="mod-13.html#mod-13-sims"><i class="fa fa-check"></i><b>13.4</b> Analyzing dissimilarity</a>
<ul>
<li class="chapter" data-level="13.4.1" data-path="mod-13.html"><a href="mod-13.html#mantel-tests-distance-vs.-distance"><i class="fa fa-check"></i><b>13.4.1</b> Mantel tests: distance vs. distance</a></li>
<li class="chapter" data-level="13.4.2" data-path="mod-13.html"><a href="mod-13.html#comparing-dissimilarity-between-groups"><i class="fa fa-check"></i><b>13.4.2</b> Comparing dissimilarity between groups</a></li>
</ul></li>
<li class="chapter" data-level="13.5" data-path="mod-13.html"><a href="mod-13.html#mod-13-ord"><i class="fa fa-check"></i><b>13.5</b> Ordination</a>
<ul>
<li class="chapter" data-level="13.5.1" data-path="mod-13.html"><a href="mod-13.html#mod-13-pca"><i class="fa fa-check"></i><b>13.5.1</b> Principal components analysis (PCA)</a></li>
<li class="chapter" data-level="13.5.2" data-path="mod-13.html"><a href="mod-13.html#nmds-and-other-ordination-methods"><i class="fa fa-check"></i><b>13.5.2</b> NMDS and other ordination methods</a></li>
<li class="chapter" data-level="13.5.3" data-path="mod-13.html"><a href="mod-13.html#other-ordination-techniques"><i class="fa fa-check"></i><b>13.5.3</b> Other ordination techniques</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="literature-cited.html"><a href="literature-cited.html"><i class="fa fa-check"></i>Literature Cited</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Applied Biological Data Analysis: Statistics and R for Biologists</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="mod-11" class="section level1 hasAnchor" number="11">
<h1><span class="header-section-number">Module 11</span> Nonlinear models<a href="mod-11.html#mod-11" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Many biological processes and phenomena are related in ways that do not form straight lines. What this really means is that rates of change are not constant. While linear models and the assumption of linearity are powerful ways to think about and analyze biological phenomena, they are not always appropriate. This module will introduce you to some ways of exploring situations where one quantity does not scale as a constant proportion of another.</p>
<div id="background" class="section level2 hasAnchor" number="11.1">
<h2><span class="header-section-number">11.1</span> Background<a href="mod-11.html#background" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>So far we have learned about a wide variety of <strong>linear models</strong>, generalized and otherwise. Linear models predict some response variable <em>Y</em> as a linear function of one or more predictor variables <em>X</em>. A linear function is one that can be written in the form <span class="math inline">\(Y=mX+b\)</span>, where <em>m</em> and <em>b</em> are constants. The scalar <em>m</em> defines how much <em>Y</em> changes for every unit increase in <em>X</em> (i.e., if <em>X</em> increases by 1, then <em>Y</em> increases by <em>m</em>). In order for the relationship to be a straight line, the change in <em>Y</em> resulting from a change in <em>X</em> does not vary. In other words, the slope of the line (<span class="math inline">\(dY/dX\)</span>) is a constant. This is what “linear” means.</p>
<p><img src="11_01.jpg" width="100%" style="display: block; margin: auto;" /></p>
<p>Some relationships do not follow such a rule. Consider the relationship below. Is the slope constant?</p>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-705-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>As shown, the relationship between <em>Y</em> and <em>X</em> is nonlinear because the slope of the line is different at different <em>X</em>. However, we can linearize the relationship by taking the logarithm of <em>Y</em>. Compare the left panel below to the right panel. All that has changed is the <em>y</em>-axis scaling.</p>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-706-1.png" width="864" style="display: block; margin: auto;" /></p>
<p>Other relationships are not amenable to this kind of transformation. For example, the figure below shows <em>Y</em> varying as a <strong>Gompertz function</strong> of <em>X</em>. Unlike the <a href="#mod-09-loglin">log-linear model</a> shown above, there is no transform that will turn the Gompertz curve into a straight line. So, if we wanted to analyze data that followed such a pattern, we would need a different set of tools.</p>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-707-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<div id="mod-11-intro" class="section level2 hasAnchor" number="11.2">
<h2><span class="header-section-number">11.2</span> Nonlinear least squares (NLS)<a href="mod-11.html#mod-11-intro" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The basic <strong>nonlinear least squares (NLS)</strong> model has some parts in common with the basic linear regression model (LM). In the state-space GLM notation, we can see that the relationship between the observed <em>Y</em> variables and the expected value <span class="math inline">\(\mu\)</span> is the same as in the LM. However, the relationship between the expected value and the predictor variables is not the same.</p>
<table>
<colgroup>
<col width="28%" />
<col width="35%" />
<col width="35%" />
</colgroup>
<thead>
<tr class="header">
<th>Component</th>
<th align="center">Linear model (LM)</th>
<th align="center">Nonlinear model (NLS)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Stochastic part</td>
<td align="center"><span class="math inline">\(Y~Normal\left(\mu,\ \sigma^2\right)\)</span></td>
<td align="center"><span class="math inline">\(Y~Normal\left(\mu,\ \sigma^2\right)\)</span></td>
</tr>
<tr class="even">
<td>Link function</td>
<td align="center"><span class="math inline">\(\mu=\eta\)</span></td>
<td align="center"><span class="math inline">\(\mu=\eta\)</span></td>
</tr>
<tr class="odd">
<td>Expected value</td>
<td align="center"><span class="math inline">\(\eta=\beta_0+\beta_{1}X\)</span></td>
<td align="center"><span class="math inline">\(\mu=f(X,\theta)\)</span></td>
</tr>
</tbody>
</table>
<p>In the NLS model, as with LM, the response variable <em>Y</em> is drawn from a normal distribution with mean <span class="math inline">\(\mu\)</span> and variance <span class="math inline">\(\sigma^2\)</span>. Also like LM, the mean <span class="math inline">\(\mu\)</span> is identical to the predictor <span class="math inline">\(\eta\)</span>. Unlike LM, in NLS the expected value <span class="math inline">\(\eta\)</span> is not a linear function but instead a function <span class="math inline">\(f\)</span>* of the explanatory variable <em>X</em> and some vector of parameters <span class="math inline">\(\theta\)</span>. The function <span class="math inline">\(f\left(X,\ \theta\right)\)</span> can be any number of functions. In the case of a linear model, the parameters <span class="math inline">\(\theta\)</span> are the intercept and slope. If <span class="math inline">\(f\left(X,\ \theta\right)\)</span> is a linear function, then the NLS model simplifies to a linear model.</p>
<p>The equations above show an important assumption of NLS: the errors are normally distributed with constant variance. In some cases, the function <span class="math inline">\(f\left(X,\ \theta\right)\)</span> can be transformed to make it linear. However, doing so will likely make the errors non-normal or heteroscedastic, which means that the LM on transformed data wouldn’t be appropriate anyway.</p>
<p>There are several methods of fitting NLS models. Unlike LM, where it is relatively straightforward to solve some linear algebra problems and get stable parameter estimates, in NLS the parameters must be approximated by searching the parameter space (i.e., set of possible values) for <span class="math inline">\(\theta\)</span> that minimize the sum of squared residuals. The methods for computing these estimates are extremely complicated and sensitive to starting inputs, and we won’t go into them here. <span class="citation">Bolker (<a href="#ref-bolker2008ecological">2008</a>)</span> gives an overview of some of these methods.</p>
</div>
<div id="mod-11-micmen" class="section level2 hasAnchor" number="11.3">
<h2><span class="header-section-number">11.3</span> Michaelis-Menten curves<a href="mod-11.html#mod-11-micmen" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The <strong>Michaelis-Menten model</strong> is a nonlinear model that is usually used to describe processes that grow and approach an asymptote. One of the classic applications is modeling the relationship between biochemical reaction rates and substrate concentrations. The Michaelis-Menten equation takes the general form</p>
<p><span class="math display">\[Y=\frac{aX}{b+X}\]</span></p>
<p>where <em>a</em> and <em>b</em> are constants.</p>
<ul>
<li><strong><em>a</em></strong> defines the <strong>asymptote</strong>: the maximum <em>Y</em> value that is approached but never reached.</li>
<li><strong><em>b</em></strong> is the <strong>Michaelis constant</strong>. It represents the <em>X</em> value at which <span class="math inline">\(Y = a/2\)</span>. Interestingly, for any proportion of the form <span class="math inline">\(c/(c+1)\)</span>, where <em>c</em> is a positive integer, the <em>X</em> value where <span class="math inline">\(Y = (c/(c+1))a\)</span> is <span class="math inline">\(cb\)</span>. For example, the <em>X</em> value when <span class="math inline">\(Y = (3/4)a\)</span> is <span class="math inline">\(3b\)</span>.</li>
</ul>
<p>The Michaelis-Menten curve goes by many names. In enzyme kinetics and biochemistry, it is called the Michaelis-Menten curve. Biologists also sometimes call it the <strong>Monod function</strong>, the <strong>Holling type II</strong> functional response, or the <strong>Beverton-Holt model</strong> (the latter name is probably the most common alternative to Michaelis-Menten). This diversity of application speaks to the generality of this function. The original definition predicted a reaction rate. However, Michaelis-Menten curves can also be used to model other phenomena. For example, <span class="citation">Kunz et al. (<a href="#ref-kunz2019">2019</a>)</span> and <span class="citation">Green et al. (<a href="#ref-green2020">2020</a>)</span> used Michaelis-Menten curves to model the effectiveness of biodiversity monitoring programs.</p>
<div id="mod-11-micmen-sim" class="section level3 hasAnchor" number="11.3.1">
<h3><span class="header-section-number">11.3.1</span> Example with simulated data<a href="mod-11.html#mod-11-micmen-sim" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>For our simulated example, we will use parameters from the R help page for <code>SSmicmen()</code>, one of the utility functions for working with Michaelis-Menten models.</p>
<div class="sourceCode" id="cb1223"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1223-1"><a href="mod-11.html#cb1223-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb1223-2"><a href="mod-11.html#cb1223-2" tabindex="-1"></a></span>
<span id="cb1223-3"><a href="mod-11.html#cb1223-3" tabindex="-1"></a><span class="co"># x values: concentrations in [0.02, 1.28] ppm</span></span>
<span id="cb1223-4"><a href="mod-11.html#cb1223-4" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">2</span><span class="sc">^</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>)<span class="sc">/</span><span class="dv">100</span>, <span class="at">each=</span><span class="dv">3</span>)</span>
<span id="cb1223-5"><a href="mod-11.html#cb1223-5" tabindex="-1"></a></span>
<span id="cb1223-6"><a href="mod-11.html#cb1223-6" tabindex="-1"></a><span class="co"># define model parameters</span></span>
<span id="cb1223-7"><a href="mod-11.html#cb1223-7" tabindex="-1"></a><span class="do">## Vm = a = asymptote</span></span>
<span id="cb1223-8"><a href="mod-11.html#cb1223-8" tabindex="-1"></a><span class="do">## K  = b = Michaelis constant</span></span>
<span id="cb1223-9"><a href="mod-11.html#cb1223-9" tabindex="-1"></a>Vm <span class="ot">&lt;-</span> <span class="dv">210</span></span>
<span id="cb1223-10"><a href="mod-11.html#cb1223-10" tabindex="-1"></a>K <span class="ot">&lt;-</span> <span class="fl">0.06</span></span>
<span id="cb1223-11"><a href="mod-11.html#cb1223-11" tabindex="-1"></a></span>
<span id="cb1223-12"><a href="mod-11.html#cb1223-12" tabindex="-1"></a><span class="co"># expected values</span></span>
<span id="cb1223-13"><a href="mod-11.html#cb1223-13" tabindex="-1"></a>y <span class="ot">&lt;-</span> (Vm<span class="sc">*</span>x)<span class="sc">/</span>(K<span class="sc">+</span>x)</span>
<span id="cb1223-14"><a href="mod-11.html#cb1223-14" tabindex="-1"></a></span>
<span id="cb1223-15"><a href="mod-11.html#cb1223-15" tabindex="-1"></a><span class="co"># add residual variation</span></span>
<span id="cb1223-16"><a href="mod-11.html#cb1223-16" tabindex="-1"></a>y <span class="ot">&lt;-</span> y <span class="sc">+</span> <span class="fu">rnorm</span>(<span class="fu">length</span>(x), <span class="dv">0</span>, <span class="dv">10</span>)</span>
<span id="cb1223-17"><a href="mod-11.html#cb1223-17" tabindex="-1"></a></span>
<span id="cb1223-18"><a href="mod-11.html#cb1223-18" tabindex="-1"></a><span class="co"># plot data</span></span>
<span id="cb1223-19"><a href="mod-11.html#cb1223-19" tabindex="-1"></a><span class="fu">plot</span>(x,y, <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="fu">max</span>(y)),</span>
<span id="cb1223-20"><a href="mod-11.html#cb1223-20" tabindex="-1"></a>     <span class="at">xlab=</span><span class="st">&quot;Substrate conc. (ppm)&quot;</span>, </span>
<span id="cb1223-21"><a href="mod-11.html#cb1223-21" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">&quot;Reaction rate&quot;</span>,</span>
<span id="cb1223-22"><a href="mod-11.html#cb1223-22" tabindex="-1"></a>     <span class="at">pch=</span><span class="dv">16</span>, <span class="at">cex=</span><span class="fl">1.2</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-708-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>The R function that fits nonlinear least squares model is <code>nls()</code>. Like <code>lm()</code> and <code>glm()</code>, <code>nls()</code> has a formula interface to specifying models and produces an object that contains many of the diagnostics you need to evaluate your model. Unlike <code>lm()</code> and <code>glm()</code>, <code>nls()</code> usually needs to be supplied starting values for model parameters. These values are used as starting points for the numerical optimization algorithm. There are two basic ways to supply starting values. For some models you can use a “self-starting” model that will guess at good starting values. The other way is for you to guess.</p>
<p>The command below shows how to use the “self-start” Michaelis-Menten model. The righthand side of the formula is the function <code>SSmicmen()</code> (“self-starting Michaelis-Menten”) instead of the model equation. The arguments to a self-start function are usually the predictor variable followed by the names you want to use for the terms in the model. The order of those names matters, because R will use them in the order you provide them. Below we use the default names <code>Vm</code> and <code>K</code>, which R interprets as the asymptote (“Vmax” or “V-max”) and Michaelis constant, respectively. When using the Michaelis-Menten model for something other than enzyme kinetics, I prefer to use <code>a</code> and <code>b</code> instead of <code>Vm</code> and <code>K</code><a href="#fn82" class="footnote-ref" id="fnref82"><sup>82</sup></a>.</p>
<div class="sourceCode" id="cb1224"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1224-1"><a href="mod-11.html#cb1224-1" tabindex="-1"></a><span class="co"># method 1: fit with self-start models</span></span>
<span id="cb1224-2"><a href="mod-11.html#cb1224-2" tabindex="-1"></a>mod1 <span class="ot">&lt;-</span> <span class="fu">nls</span>(y<span class="sc">~</span><span class="fu">SSmicmen</span>(x, Vm, K))</span>
<span id="cb1224-3"><a href="mod-11.html#cb1224-3" tabindex="-1"></a></span>
<span id="cb1224-4"><a href="mod-11.html#cb1224-4" tabindex="-1"></a><span class="co"># same but with different parameter names:</span></span>
<span id="cb1224-5"><a href="mod-11.html#cb1224-5" tabindex="-1"></a>mod1.v2 <span class="ot">&lt;-</span> <span class="fu">nls</span>(y<span class="sc">~</span><span class="fu">SSmicmen</span>(x, a, b))</span>
<span id="cb1224-6"><a href="mod-11.html#cb1224-6" tabindex="-1"></a><span class="fu">summary</span>(mod1.v2)</span></code></pre></div>
<pre><code>## 
## Formula: y ~ SSmicmen(x, a, b)
## 
## Parameters:
##    Estimate Std. Error t value Pr(&gt;|t|)    
## a 2.080e+02  4.499e+00   46.23  &lt; 2e-16 ***
## b 5.640e-02  4.995e-03   11.29 7.21e-10 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 10 on 19 degrees of freedom
## 
## Number of iterations to convergence: 0 
## Achieved convergence tolerance: 2.333e-07</code></pre>
<p>If you are fitting a curve for which there is no self-start method, or the self-start method doesn’t work, or if you enjoy reading error messages, you will need to supply starting values. Coming up with good starting values is part art, part science, and part random guessing. For many models you can often estimate some values by thinking about what they represent. For our Michaelis-Menten example, we can see that the curve stops increasing once it gets up to about <em>Y</em> = 200 or so. Thus, 200 might be a good guess for the asymptote (<code>Vm</code> in our code). Likewise, we can see that <em>Y</em> reaches half of its asymptotic value at around <em>X</em> = 0.05. So, 0.05 is probably a decent guess for the Michaelis constant <em>K</em>.</p>
<p>For models with more parameters, you can often guess at 1 or 2 parameters based on the scatterplot, then pick a few representative points and solve for other parameters. For example, if we had guessed that <span class="math inline">\(V_{max}\)</span> might be 200, we could pick some points from our data and solve for <em>K</em>, conditional on <span class="math inline">\(V_{max}\)</span> = 200. Below we calculate <code>test.k</code>, which contains estimates of <code>K</code> conditional on <code>Vm</code> = 200. The mean or median of that vector is a good guess for a starting value of <code>K</code>.</p>
<div class="sourceCode" id="cb1226"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1226-1"><a href="mod-11.html#cb1226-1" tabindex="-1"></a>test.vm <span class="ot">&lt;-</span> <span class="dv">200</span></span>
<span id="cb1226-2"><a href="mod-11.html#cb1226-2" tabindex="-1"></a>test.k <span class="ot">&lt;-</span> (test.vm<span class="sc">*</span>x)<span class="sc">/</span>y<span class="sc">-</span>x</span>
<span id="cb1226-3"><a href="mod-11.html#cb1226-3" tabindex="-1"></a><span class="fu">mean</span>(test.k)</span></code></pre></div>
<pre><code>## [1] 0.042005</code></pre>
<p>Use these estimates as starting values to <code>nls()</code>. Notice that because we are not using a self-starting model, we must define the model equation manually. The variable names that you use in the formula must match the variable names in the list of starting values <strong><em>exactly</em></strong>. This includes capitalization!</p>
<p>The second example is just to “eyeball” some starting values. In the case of the Michaelis-Menten model this is easy. <span class="math inline">\(V_{max}\)</span> can be estimated as the maximum value or just above it, while <em>K</em> is the <em>X</em> value at which <em>Y</em> is about half of <span class="math inline">\(V_{max}\)</span>. As long as your guesses are in the right ballpark, <code>nls()</code> should be able to find good estimates.</p>
<div class="sourceCode" id="cb1228"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1228-1"><a href="mod-11.html#cb1228-1" tabindex="-1"></a><span class="co"># method 2: eyeball some starting values</span></span>
<span id="cb1228-2"><a href="mod-11.html#cb1228-2" tabindex="-1"></a>start.list <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">Vm=</span><span class="dv">200</span>, <span class="at">K=</span><span class="fl">0.042</span>)</span>
<span id="cb1228-3"><a href="mod-11.html#cb1228-3" tabindex="-1"></a>mod2 <span class="ot">&lt;-</span> <span class="fu">nls</span>(y<span class="sc">~</span>(Vm<span class="sc">*</span>x)<span class="sc">/</span>(K<span class="sc">+</span>x),</span>
<span id="cb1228-4"><a href="mod-11.html#cb1228-4" tabindex="-1"></a>            <span class="at">start=</span>start.list)</span>
<span id="cb1228-5"><a href="mod-11.html#cb1228-5" tabindex="-1"></a><span class="fu">summary</span>(mod2)</span></code></pre></div>
<pre><code>## 
## Formula: y ~ (Vm * x)/(K + x)
## 
## Parameters:
##     Estimate Std. Error t value Pr(&gt;|t|)    
## Vm 2.080e+02  4.499e+00   46.23  &lt; 2e-16 ***
## K  5.640e-02  4.995e-03   11.29 7.21e-10 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 10 on 19 degrees of freedom
## 
## Number of iterations to convergence: 4 
## Achieved convergence tolerance: 2.98e-07</code></pre>
<p>Both methods estimated parameters that were close to the true values. The coefficients table can be presented in much the same way that the coefficients table for a linear regression or GLM is presented.
Unlike linear models, there is no straightforward way to calculate an <em>R</em><sup>2</sup> value for nonlinear models. This is because of two key differences between nonlinear models and linear models:</p>
<ul>
<li>Nonlinear models do not have an intercept term, which is implicit in the <em>R</em><sup>2</sup> calculation in linear models.</li>
<li>In linear models, the equation <span class="math inline">\({SS}_{total} = {SS}_{model} + {SS}_{residuals}\)</span> is always true, which implies that <span class="math inline">\({SS}_{model} \le {SS}_{total}\)</span>. For nonlinear models, it is not. This means that the usual <em>R</em><sup>2</sup> equation, <span class="math inline">\({SS}_{model} / {SS}_{total}\)</span>, is not guaranteed to be between 0 and 1.</li>
</ul>
<p>Nevertheless, there are measures of model performance that we can use to approximate an <em>R</em><sup>2</sup> value. One way is to calculate a pseudo-<em>R</em><sup>2</sup> similar to that we calculated for GLM:</p>
<p><span class="math display">\[pseudo-R^{2}=1-\frac{{SS}_{residual}}{{SS}_{total}}\]</span></p>
<p>This method is recommended by <span class="citation">Schabenberger and Pierce (<a href="#ref-schabenberger2001contemporary">2001</a>)</span> and popular on Stack Overflow, but it can be misleading when model fit is poor <span class="citation">(<a href="#ref-spiess2010">Spiess and Neumeyer 2010</a>)</span>. The pseudo-<em>R</em><sup>2</sup> should only ever be considered a rough guide to model fit. Unlike a true <em>R</em><sup>2</sup>, which expresses the total variance explained by a model because of the way it is defined to be in the interval [0, 1], the pseudo-<em>R</em><sup>2</sup> is not restricted to [0, 1] and thus is not the same as variance explained. Better fitting models will usually have greater pseudo-<em>R</em><sup>2</sup> than worse fitting models, but this statistic does not represent “variance explained” or anything similar.</p>
<p>There is no base function for pseudo-<em>R</em><sup>2</sup>, but it is easy to calculate. The function below takes in the model object <code>mod</code> and the original response variable <code>Y</code>, and returns the pseudo-<em>R</em><sup>2</sup>.</p>
<div class="sourceCode" id="cb1230"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1230-1"><a href="mod-11.html#cb1230-1" tabindex="-1"></a>pseudoR2 <span class="ot">&lt;-</span> <span class="cf">function</span>(mod, Y){</span>
<span id="cb1230-2"><a href="mod-11.html#cb1230-2" tabindex="-1"></a>    ss.tot <span class="ot">&lt;-</span> <span class="fu">sum</span>((<span class="fu">fitted</span>(mod)<span class="sc">-</span><span class="fu">mean</span>(Y))<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb1230-3"><a href="mod-11.html#cb1230-3" tabindex="-1"></a>    ss.res <span class="ot">&lt;-</span> <span class="fu">sum</span>(<span class="fu">residuals</span>(mod)<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb1230-4"><a href="mod-11.html#cb1230-4" tabindex="-1"></a>    <span class="dv">1</span><span class="sc">-</span>(ss.res<span class="sc">/</span>ss.tot)</span>
<span id="cb1230-5"><a href="mod-11.html#cb1230-5" tabindex="-1"></a>}</span>
<span id="cb1230-6"><a href="mod-11.html#cb1230-6" tabindex="-1"></a></span>
<span id="cb1230-7"><a href="mod-11.html#cb1230-7" tabindex="-1"></a><span class="co"># test the function:</span></span>
<span id="cb1230-8"><a href="mod-11.html#cb1230-8" tabindex="-1"></a><span class="fu">pseudoR2</span>(mod2, y)</span></code></pre></div>
<pre><code>## [1] 0.9651684</code></pre>
<p>The package <code>rcompanion</code> <span class="citation">(<a href="#ref-mangiafico2021">Mangiafico 2021</a>)</span> has a function called <code>nagelkerke()</code> that will calculate and present several alternative pseudo-<em>R</em><sup>2</sup> values, including the version above. To use it, you need to specify a <strong>null model</strong> that includes no predictor variables. Null models usually just use the mean value of <em>Y</em> as the predictor of <em>Y</em>.</p>
<div class="sourceCode" id="cb1232"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1232-1"><a href="mod-11.html#cb1232-1" tabindex="-1"></a><span class="co"># not run:</span></span>
<span id="cb1232-2"><a href="mod-11.html#cb1232-2" tabindex="-1"></a><span class="co">#library(rcompanion)</span></span>
<span id="cb1232-3"><a href="mod-11.html#cb1232-3" tabindex="-1"></a></span>
<span id="cb1232-4"><a href="mod-11.html#cb1232-4" tabindex="-1"></a><span class="co"># function that uses mean (m) to predict y</span></span>
<span id="cb1232-5"><a href="mod-11.html#cb1232-5" tabindex="-1"></a><span class="co"># x is included because it is the predictor in mod2</span></span>
<span id="cb1232-6"><a href="mod-11.html#cb1232-6" tabindex="-1"></a>null.fun <span class="ot">&lt;-</span> <span class="cf">function</span>(x, m){m}</span>
<span id="cb1232-7"><a href="mod-11.html#cb1232-7" tabindex="-1"></a></span>
<span id="cb1232-8"><a href="mod-11.html#cb1232-8" tabindex="-1"></a><span class="co"># fit null model using nls().</span></span>
<span id="cb1232-9"><a href="mod-11.html#cb1232-9" tabindex="-1"></a><span class="co"># notice that m is initialized</span></span>
<span id="cb1232-10"><a href="mod-11.html#cb1232-10" tabindex="-1"></a><span class="co"># as the mean of y</span></span>
<span id="cb1232-11"><a href="mod-11.html#cb1232-11" tabindex="-1"></a>null.mod <span class="ot">&lt;-</span> <span class="fu">nls</span>(y<span class="sc">~</span><span class="fu">null.fun</span>(x, m), </span>
<span id="cb1232-12"><a href="mod-11.html#cb1232-12" tabindex="-1"></a>                <span class="at">start=</span><span class="fu">list</span>(<span class="at">m=</span><span class="fu">mean</span>(y)))</span>
<span id="cb1232-13"><a href="mod-11.html#cb1232-13" tabindex="-1"></a></span>
<span id="cb1232-14"><a href="mod-11.html#cb1232-14" tabindex="-1"></a><span class="co"># get the pseudo R-squared values (and other stuff)</span></span>
<span id="cb1232-15"><a href="mod-11.html#cb1232-15" tabindex="-1"></a><span class="co"># uncomment; had issues with package versions when rendering</span></span>
<span id="cb1232-16"><a href="mod-11.html#cb1232-16" tabindex="-1"></a><span class="co">#nagelkerke(mod2, null=null.mod)</span></span></code></pre></div>
<p>Once you have your fitted model, it’s time to present the model predictions alongside the original data. This process is a little more complicated for <code>nls()</code> outputs than for <code>lm()</code> or <code>glm()</code>. Part of the problem is that calculating confidence intervals for NLS predictions is not straightforward: just like the parameters cannot be calculated directly, the uncertainty also cannot be calculated directly. Instead, we have to <strong>approximate</strong> the CI. The base R <code>predict()</code> method for <code>nls()</code> outputs does not do this<a href="#fn83" class="footnote-ref" id="fnref83"><sup>83</sup></a>. The package <code>propagate</code> <span class="citation">(<a href="#ref-spiess2018">Spiess 2018</a>)</span> has function <code>predictNLS()</code> to approximate CIs for NLS models. This function uses Monte Carlo sampling (i.e., random simulations) to estimate the CIs, so it can take a while to run.</p>
<div class="sourceCode" id="cb1233"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1233-1"><a href="mod-11.html#cb1233-1" tabindex="-1"></a><span class="fu">library</span>(propagate)</span>
<span id="cb1233-2"><a href="mod-11.html#cb1233-2" tabindex="-1"></a>L <span class="ot">&lt;-</span> <span class="dv">20</span></span>
<span id="cb1233-3"><a href="mod-11.html#cb1233-3" tabindex="-1"></a>px <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fu">min</span>(x), <span class="fu">max</span>(x), <span class="at">length=</span>L)</span>
<span id="cb1233-4"><a href="mod-11.html#cb1233-4" tabindex="-1"></a>pred <span class="ot">&lt;-</span> <span class="fu">predictNLS</span>(mod2,</span>
<span id="cb1233-5"><a href="mod-11.html#cb1233-5" tabindex="-1"></a>                   <span class="at">newdata=</span><span class="fu">data.frame</span>(<span class="at">x=</span>px),</span>
<span id="cb1233-6"><a href="mod-11.html#cb1233-6" tabindex="-1"></a>                   <span class="at">interval=</span><span class="st">&quot;confidence&quot;</span>)</span>
<span id="cb1233-7"><a href="mod-11.html#cb1233-7" tabindex="-1"></a><span class="co"># we just need this part:</span></span>
<span id="cb1233-8"><a href="mod-11.html#cb1233-8" tabindex="-1"></a>pred <span class="ot">&lt;-</span> pred<span class="sc">$</span>summary</span>
<span id="cb1233-9"><a href="mod-11.html#cb1233-9" tabindex="-1"></a></span>
<span id="cb1233-10"><a href="mod-11.html#cb1233-10" tabindex="-1"></a><span class="co"># look at result</span></span>
<span id="cb1233-11"><a href="mod-11.html#cb1233-11" tabindex="-1"></a><span class="co"># (rounded to fit better on the screen)</span></span>
<span id="cb1233-12"><a href="mod-11.html#cb1233-12" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">round</span>(pred, <span class="dv">2</span>))</span></code></pre></div>
<pre><code>##   Prop.Mean.1 Prop.Mean.2 Prop.sd.1 Prop.sd.2 Prop.2.5% Prop.97.5% Sim.Mean
## 1       54.46       54.63      2.81      2.82     48.73      60.53    54.66
## 2      125.81      125.90      3.02      3.02    119.57     132.22   125.91
## 3      151.89      151.92      2.52      2.53    146.63     157.21   151.93
## 4      165.41      165.41      2.44      2.44    160.30     170.53   165.41
## 5      173.68      173.67      2.55      2.55    168.34     179.00   173.67
## 6      179.26      179.25      2.70      2.70    173.60     184.90   179.25
##   Sim.sd Sim.Median Sim.MAD Sim.2.5% Sim.97.5%
## 1   3.03      54.46    2.86    49.27     61.21
## 2   3.21     125.82    3.08   119.83    132.54
## 3   2.68     151.89    2.57   146.71    157.32
## 4   2.59     165.41    2.49   160.32    170.56
## 5   2.69     173.68    2.60   168.33    179.00
## 6   2.86     179.26    2.75   173.55    184.86</code></pre>
<p>The result contains several predicted means, SD, and quantiles for the response variable at every value of the input <em>X</em> values. The different versions are calculated using first- or second-order Taylor series (<code>.1</code> and <code>.2</code>), or Monte Carlo simulation (<code>Sim.</code>). The first-order Taylor series approximation is also called the <strong>Delta method</strong>. For each method you can use the mean and SD, or the mean and 95% confidence limits, or use the mean and SD to calculate your own confidence limits. For most models there isn’t much difference between the Taylor series approximations and the Monte Carlo estimates, so use whichever you prefer. In the example below we’ll plot the Taylor series and Monte Carlo CI to compare; in your work you only need to present one (and identify which one you used!).</p>
<div class="sourceCode" id="cb1235"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1235-1"><a href="mod-11.html#cb1235-1" tabindex="-1"></a><span class="co"># 1st order Taylor series</span></span>
<span id="cb1235-2"><a href="mod-11.html#cb1235-2" tabindex="-1"></a>mn1 <span class="ot">&lt;-</span> pred<span class="sc">$</span>Prop.Mean<span class="fl">.1</span></span>
<span id="cb1235-3"><a href="mod-11.html#cb1235-3" tabindex="-1"></a>lo1 <span class="ot">&lt;-</span> pred<span class="sc">$</span><span class="st">&#39;Prop.2.5%&#39;</span></span>
<span id="cb1235-4"><a href="mod-11.html#cb1235-4" tabindex="-1"></a>up1 <span class="ot">&lt;-</span> pred<span class="sc">$</span><span class="st">&#39;Prop.97.5%&#39;</span></span>
<span id="cb1235-5"><a href="mod-11.html#cb1235-5" tabindex="-1"></a></span>
<span id="cb1235-6"><a href="mod-11.html#cb1235-6" tabindex="-1"></a><span class="co"># Monte Carlo</span></span>
<span id="cb1235-7"><a href="mod-11.html#cb1235-7" tabindex="-1"></a>mn2 <span class="ot">&lt;-</span> pred<span class="sc">$</span>Sim.Mean</span>
<span id="cb1235-8"><a href="mod-11.html#cb1235-8" tabindex="-1"></a>lo2 <span class="ot">&lt;-</span> pred<span class="sc">$</span><span class="st">&#39;Sim.2.5%&#39;</span></span>
<span id="cb1235-9"><a href="mod-11.html#cb1235-9" tabindex="-1"></a>up2 <span class="ot">&lt;-</span> pred<span class="sc">$</span><span class="st">&#39;Sim.97.5%&#39;</span></span>
<span id="cb1235-10"><a href="mod-11.html#cb1235-10" tabindex="-1"></a></span>
<span id="cb1235-11"><a href="mod-11.html#cb1235-11" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="at">bty=</span><span class="st">&quot;n&quot;</span>, <span class="at">lend=</span><span class="dv">1</span>, <span class="at">las=</span><span class="dv">1</span>,</span>
<span id="cb1235-12"><a href="mod-11.html#cb1235-12" tabindex="-1"></a>    <span class="at">cex.axis=</span><span class="fl">1.3</span>, <span class="at">cex.lab=</span><span class="fl">1.3</span>)</span>
<span id="cb1235-13"><a href="mod-11.html#cb1235-13" tabindex="-1"></a><span class="fu">plot</span>(x,y, <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="fu">max</span>(y)),</span>
<span id="cb1235-14"><a href="mod-11.html#cb1235-14" tabindex="-1"></a>     <span class="at">xlab=</span><span class="st">&quot;Substrate conc. (ppm)&quot;</span>, </span>
<span id="cb1235-15"><a href="mod-11.html#cb1235-15" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">&quot;Reaction rate&quot;</span>,</span>
<span id="cb1235-16"><a href="mod-11.html#cb1235-16" tabindex="-1"></a>     <span class="at">pch=</span><span class="dv">16</span>, <span class="at">cex=</span><span class="fl">1.2</span>,</span>
<span id="cb1235-17"><a href="mod-11.html#cb1235-17" tabindex="-1"></a>     <span class="at">main=</span><span class="st">&quot;Taylor series CI&quot;</span>)</span>
<span id="cb1235-18"><a href="mod-11.html#cb1235-18" tabindex="-1"></a><span class="fu">points</span>(px, lo1, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1235-19"><a href="mod-11.html#cb1235-19" tabindex="-1"></a><span class="fu">points</span>(px, up1, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1235-20"><a href="mod-11.html#cb1235-20" tabindex="-1"></a><span class="fu">points</span>(px, mn1, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>)</span>
<span id="cb1235-21"><a href="mod-11.html#cb1235-21" tabindex="-1"></a></span>
<span id="cb1235-22"><a href="mod-11.html#cb1235-22" tabindex="-1"></a><span class="fu">plot</span>(x,y, <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="fu">max</span>(y)),</span>
<span id="cb1235-23"><a href="mod-11.html#cb1235-23" tabindex="-1"></a>     <span class="at">xlab=</span><span class="st">&quot;Substrate conc. (ppm)&quot;</span>, </span>
<span id="cb1235-24"><a href="mod-11.html#cb1235-24" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">&quot;Reaction rate&quot;</span>,</span>
<span id="cb1235-25"><a href="mod-11.html#cb1235-25" tabindex="-1"></a>     <span class="at">pch=</span><span class="dv">16</span>, <span class="at">cex=</span><span class="fl">1.2</span>,</span>
<span id="cb1235-26"><a href="mod-11.html#cb1235-26" tabindex="-1"></a>     <span class="at">main=</span><span class="st">&quot;Monte Carlo CI&quot;</span>)</span>
<span id="cb1235-27"><a href="mod-11.html#cb1235-27" tabindex="-1"></a><span class="fu">points</span>(px, lo2, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1235-28"><a href="mod-11.html#cb1235-28" tabindex="-1"></a><span class="fu">points</span>(px, up2, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1235-29"><a href="mod-11.html#cb1235-29" tabindex="-1"></a><span class="fu">points</span>(px, mn2, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-713-1.png" width="864" style="display: block; margin: auto;" /></p>
<p>In our simulated example, the Taylor series approximation and Monte Carlo simulations produced nearly identical predictions and CI.</p>
</div>
<div id="example-with-real-data-4" class="section level3 hasAnchor" number="11.3.2">
<h3><span class="header-section-number">11.3.2</span> Example with real data<a href="mod-11.html#example-with-real-data-4" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><span class="citation">Leonard and Wood (<a href="#ref-leonard2013acute">2013</a>)</span> studied the bioaccumulation of nickel (Ni) by four species of invertebrates in different environmental conditions. Invertebrates were held for 14 days in aquaria at varying aqueous Ni concentrations in either hard water (140 mg/L CaCO<sub>3</sub>) or soft water (40 mg/L CaCO<sub>3</sub>), after which whole-body Ni residues were measured.</p>
<p><img src="11_02.jpg" width="50%" style="display: block; margin: auto;" /></p>
<p>The authors used Michaelis-Menten models to measure the uptake of Ni from the water as a function of Ni concentration. We will focus on part of their dataset, accumulation of Ni by midge <em>Chironomus riparius</em> (above<a href="#fn84" class="footnote-ref" id="fnref84"><sup>84</sup></a>). Data were extracted from the paper based on parameter estimates in their Table 4 and data in Figure 6. The variables in this dataset are:</p>
<table>
<thead>
<tr class="header">
<th>Variable</th>
<th>Units</th>
<th>Meaning</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>ni</td>
<td><span class="math inline">\(\mu\)</span> mol/L</td>
<td>Concentration of Ni in water</td>
</tr>
<tr class="even">
<td>res</td>
<td><span class="math inline">\(\mu\)</span> mol/kg</td>
<td>Concentration of Ni in body tissues</td>
</tr>
<tr class="odd">
<td>water</td>
<td>N/A</td>
<td>Water treatment: hard or soft (CaCO<sub>3</sub> conc.)</td>
</tr>
</tbody>
</table>
<p>Download the dataset <a href="https://greenquanteco.github.io/leonard_2013_data.csv">leonard_2013_data.csv</a> and put it in your R home directory.</p>
<div class="sourceCode" id="cb1236"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1236-1"><a href="mod-11.html#cb1236-1" tabindex="-1"></a>in.name <span class="ot">&lt;-</span> <span class="st">&quot;leonard_2013_data.csv&quot;</span></span>
<span id="cb1236-2"><a href="mod-11.html#cb1236-2" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(in.name, <span class="at">header=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p>Part of the original experiment was to see whether water hardness affected uptake of Ni. This is recorded in the factor <code>water</code> in the dataset. Plot the data and color the points by water treatment to see if there are any patterns.</p>
<div class="sourceCode" id="cb1237"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1237-1"><a href="mod-11.html#cb1237-1" tabindex="-1"></a><span class="fu">plot</span>(dat<span class="sc">$</span>ni, dat<span class="sc">$</span>res,</span>
<span id="cb1237-2"><a href="mod-11.html#cb1237-2" tabindex="-1"></a>     <span class="at">col=</span><span class="fu">ifelse</span>(dat<span class="sc">$</span>water<span class="sc">==</span><span class="st">&quot;soft&quot;</span>, <span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>))</span>
<span id="cb1237-3"><a href="mod-11.html#cb1237-3" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;topleft&quot;</span>,</span>
<span id="cb1237-4"><a href="mod-11.html#cb1237-4" tabindex="-1"></a>       <span class="at">legend=</span><span class="fu">c</span>(<span class="st">&quot;Soft&quot;</span>, <span class="st">&quot;Hard&quot;</span>),</span>
<span id="cb1237-5"><a href="mod-11.html#cb1237-5" tabindex="-1"></a>       <span class="at">col=</span><span class="fu">c</span>(<span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>),</span>
<span id="cb1237-6"><a href="mod-11.html#cb1237-6" tabindex="-1"></a>       <span class="at">pch=</span><span class="dv">1</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-716-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>The figure suggests that Ni accumulates more slowly in hard water than in soft water, and that less Ni is taken up in hard water. The data seem to follow a curve that approaches an asymptote, around 60000 <span class="math inline">\(\mu\)</span> mol/kg for soft water and 20000 <span class="math inline">\(\mu\)</span> mol/kg for hard water. The biological interpretation is that organisms can accumulate Ni from their environment, but eventually the rate of uptake will slow as the organism’s cells and tissues become saturated (or, because they reach an equilibrium with their environment). High levels of Ni accumulation can directly cause mortality, further limiting uptake. We might suspect that a Michaelis-Menten model is appropriate because it (1) has an asymptote term to capture the maximum accumulation; and (2) describes processes where a rate or first derivative slows as the system saturates. Because the dynamics appear so different for the hard water and soft water treatments, we will fit a separate model for each group.</p>
<div class="sourceCode" id="cb1238"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1238-1"><a href="mod-11.html#cb1238-1" tabindex="-1"></a><span class="co"># fit models using nls() and SSmicmen()</span></span>
<span id="cb1238-2"><a href="mod-11.html#cb1238-2" tabindex="-1"></a>flag1 <span class="ot">&lt;-</span> <span class="fu">which</span>(dat<span class="sc">$</span>water <span class="sc">==</span> <span class="st">&quot;soft&quot;</span>)</span>
<span id="cb1238-3"><a href="mod-11.html#cb1238-3" tabindex="-1"></a>flag2 <span class="ot">&lt;-</span> <span class="fu">which</span>(dat<span class="sc">$</span>water <span class="sc">==</span> <span class="st">&quot;hard&quot;</span>)</span>
<span id="cb1238-4"><a href="mod-11.html#cb1238-4" tabindex="-1"></a></span>
<span id="cb1238-5"><a href="mod-11.html#cb1238-5" tabindex="-1"></a>mod1 <span class="ot">&lt;-</span> <span class="fu">nls</span>(res<span class="sc">~</span><span class="fu">SSmicmen</span>(ni, a, b),</span>
<span id="cb1238-6"><a href="mod-11.html#cb1238-6" tabindex="-1"></a>            <span class="at">data=</span>dat[flag1,])</span>
<span id="cb1238-7"><a href="mod-11.html#cb1238-7" tabindex="-1"></a>mod2 <span class="ot">&lt;-</span> <span class="fu">nls</span>(res<span class="sc">~</span><span class="fu">SSmicmen</span>(ni, a, b),</span>
<span id="cb1238-8"><a href="mod-11.html#cb1238-8" tabindex="-1"></a>            <span class="at">data=</span>dat[flag2,])</span></code></pre></div>
<p>How good are the models? We can calculate the pseudo-<em>R</em><sup>2</sup> values using the function <code>pseudoR2()</code> defined above. Both models look pretty good:</p>
<div class="sourceCode" id="cb1239"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1239-1"><a href="mod-11.html#cb1239-1" tabindex="-1"></a><span class="fu">pseudoR2</span>(mod1, dat<span class="sc">$</span>res[flag1])</span></code></pre></div>
<pre><code>## [1] 0.9754908</code></pre>
<div class="sourceCode" id="cb1241"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1241-1"><a href="mod-11.html#cb1241-1" tabindex="-1"></a><span class="fu">pseudoR2</span>(mod2, dat<span class="sc">$</span>res[flag2])</span></code></pre></div>
<pre><code>## [1] 0.9125796</code></pre>
<p>Finally, let’s produce the usual predicted values and 95% CI for each group.</p>
<div class="sourceCode" id="cb1243"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1243-1"><a href="mod-11.html#cb1243-1" tabindex="-1"></a><span class="fu">library</span>(propagate)</span>
<span id="cb1243-2"><a href="mod-11.html#cb1243-2" tabindex="-1"></a></span>
<span id="cb1243-3"><a href="mod-11.html#cb1243-3" tabindex="-1"></a><span class="co"># values for prediction</span></span>
<span id="cb1243-4"><a href="mod-11.html#cb1243-4" tabindex="-1"></a>L <span class="ot">&lt;-</span> <span class="dv">20</span></span>
<span id="cb1243-5"><a href="mod-11.html#cb1243-5" tabindex="-1"></a>px <span class="ot">&lt;-</span> dat<span class="sc">$</span>ni[<span class="fu">which</span>(dat<span class="sc">$</span>ni <span class="sc">&gt;</span> <span class="dv">0</span>)]</span>
<span id="cb1243-6"><a href="mod-11.html#cb1243-6" tabindex="-1"></a>px <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fu">min</span>(px), <span class="fu">max</span>(px), <span class="at">length=</span>L)</span>
<span id="cb1243-7"><a href="mod-11.html#cb1243-7" tabindex="-1"></a></span>
<span id="cb1243-8"><a href="mod-11.html#cb1243-8" tabindex="-1"></a><span class="co"># calculate predictions and CI</span></span>
<span id="cb1243-9"><a href="mod-11.html#cb1243-9" tabindex="-1"></a>pred1 <span class="ot">&lt;-</span> <span class="fu">predictNLS</span>(mod1,</span>
<span id="cb1243-10"><a href="mod-11.html#cb1243-10" tabindex="-1"></a>                    <span class="at">newdata=</span><span class="fu">data.frame</span>(<span class="at">ni=</span>px),</span>
<span id="cb1243-11"><a href="mod-11.html#cb1243-11" tabindex="-1"></a>                    <span class="at">interval=</span><span class="st">&quot;confidence&quot;</span>)</span>
<span id="cb1243-12"><a href="mod-11.html#cb1243-12" tabindex="-1"></a>pred2 <span class="ot">&lt;-</span> <span class="fu">predictNLS</span>(mod2,</span>
<span id="cb1243-13"><a href="mod-11.html#cb1243-13" tabindex="-1"></a>                    <span class="at">newdata=</span><span class="fu">data.frame</span>(<span class="at">ni=</span>px),</span>
<span id="cb1243-14"><a href="mod-11.html#cb1243-14" tabindex="-1"></a>                    <span class="at">interval=</span><span class="st">&quot;confidence&quot;</span>)</span>
<span id="cb1243-15"><a href="mod-11.html#cb1243-15" tabindex="-1"></a></span>
<span id="cb1243-16"><a href="mod-11.html#cb1243-16" tabindex="-1"></a><span class="co"># just need this part</span></span>
<span id="cb1243-17"><a href="mod-11.html#cb1243-17" tabindex="-1"></a>pred1 <span class="ot">&lt;-</span> pred1<span class="sc">$</span>summary</span>
<span id="cb1243-18"><a href="mod-11.html#cb1243-18" tabindex="-1"></a>pred2 <span class="ot">&lt;-</span> pred2<span class="sc">$</span>summary</span></code></pre></div>
<p>Which predictions should we use? It usually doesn’t matter whether you use the 1st order Taylor series, 2nd order Taylor series, or Monte Carlo values, but sometimes it does. We can examine histograms of the differences between the predictions to get a sense of which methods agree with each other.</p>
<div class="sourceCode" id="cb1244"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1244-1"><a href="mod-11.html#cb1244-1" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>))</span>
<span id="cb1244-2"><a href="mod-11.html#cb1244-2" tabindex="-1"></a><span class="fu">hist</span>(pred1<span class="sc">$</span>Prop.Mean<span class="fl">.1</span> <span class="sc">-</span> pred1<span class="sc">$</span>Prop.Mean<span class="fl">.2</span>,<span class="at">main=</span><span class="st">&quot;1st - 2nd Taylor&quot;</span>)</span>
<span id="cb1244-3"><a href="mod-11.html#cb1244-3" tabindex="-1"></a><span class="fu">hist</span>(pred1<span class="sc">$</span>Prop.Mean<span class="fl">.1</span> <span class="sc">-</span> pred1<span class="sc">$</span>Sim.Mean,<span class="at">main=</span><span class="st">&quot;1st Taylor - Monte Carlo&quot;</span>)</span>
<span id="cb1244-4"><a href="mod-11.html#cb1244-4" tabindex="-1"></a><span class="fu">hist</span>(pred1<span class="sc">$</span>Prop.Mean<span class="fl">.2</span> <span class="sc">-</span> pred1<span class="sc">$</span>Sim.Mean,<span class="at">main=</span><span class="st">&quot;2nd Taylor - Monte Carlo&quot;</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-719-1.png" width="864" style="display: block; margin: auto;" /></p>
<p>Take a look at the <em>x</em>-axis scales: the differences between the 1st order Taylor series and Monte Carlo predictions are quite small relative to some of the differences between either of those and the 2nd order Taylor series estimates. This suggests that we should use one of the two methods that closely agree with each other. Another way to compare the predictions is to plot them against each other, with a line that shows where predictions are equal.</p>
<div class="sourceCode" id="cb1245"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1245-1"><a href="mod-11.html#cb1245-1" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>))</span>
<span id="cb1245-2"><a href="mod-11.html#cb1245-2" tabindex="-1"></a><span class="fu">plot</span>(pred1<span class="sc">$</span>Prop.Mean<span class="fl">.1</span>, pred1<span class="sc">$</span>Prop.Mean<span class="fl">.2</span>)</span>
<span id="cb1245-3"><a href="mod-11.html#cb1245-3" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">a=</span><span class="dv">0</span>, <span class="at">b=</span><span class="dv">1</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>)</span>
<span id="cb1245-4"><a href="mod-11.html#cb1245-4" tabindex="-1"></a><span class="fu">plot</span>(pred1<span class="sc">$</span>Prop.Mean<span class="fl">.1</span>, pred1<span class="sc">$</span>Sim.Mean)</span>
<span id="cb1245-5"><a href="mod-11.html#cb1245-5" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">a=</span><span class="dv">0</span>, <span class="at">b=</span><span class="dv">1</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>)</span>
<span id="cb1245-6"><a href="mod-11.html#cb1245-6" tabindex="-1"></a><span class="fu">plot</span>(pred1<span class="sc">$</span>Prop.Mean<span class="fl">.2</span>, pred1<span class="sc">$</span>Sim.Mean)</span>
<span id="cb1245-7"><a href="mod-11.html#cb1245-7" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">a=</span><span class="dv">0</span>, <span class="at">b=</span><span class="dv">1</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-720-1.png" width="864" style="display: block; margin: auto;" /></p>
<p>The second set of figures shows a serious problem with the 2nd order approximation: it predicts negative concentrations! Because of this, let’s ignore the 2nd order approximation and stick with the Monte Carlo predictions.</p>
<div class="sourceCode" id="cb1246"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1246-1"><a href="mod-11.html#cb1246-1" tabindex="-1"></a><span class="co"># extract predictions and CI</span></span>
<span id="cb1246-2"><a href="mod-11.html#cb1246-2" tabindex="-1"></a>mn1 <span class="ot">&lt;-</span> pred1<span class="sc">$</span>Sim.Mean</span>
<span id="cb1246-3"><a href="mod-11.html#cb1246-3" tabindex="-1"></a>lo1 <span class="ot">&lt;-</span> pred1<span class="sc">$</span><span class="st">&#39;Sim.2.5%&#39;</span></span>
<span id="cb1246-4"><a href="mod-11.html#cb1246-4" tabindex="-1"></a>up1 <span class="ot">&lt;-</span> pred1<span class="sc">$</span><span class="st">&#39;Sim.97.5%&#39;</span></span>
<span id="cb1246-5"><a href="mod-11.html#cb1246-5" tabindex="-1"></a>mn2 <span class="ot">&lt;-</span> pred2<span class="sc">$</span>Sim.Mean</span>
<span id="cb1246-6"><a href="mod-11.html#cb1246-6" tabindex="-1"></a>lo2 <span class="ot">&lt;-</span> pred2<span class="sc">$</span><span class="st">&#39;Sim.2.5%&#39;</span></span>
<span id="cb1246-7"><a href="mod-11.html#cb1246-7" tabindex="-1"></a>up2 <span class="ot">&lt;-</span> pred2<span class="sc">$</span><span class="st">&#39;Sim.97.5%&#39;</span></span>
<span id="cb1246-8"><a href="mod-11.html#cb1246-8" tabindex="-1"></a></span>
<span id="cb1246-9"><a href="mod-11.html#cb1246-9" tabindex="-1"></a><span class="co"># fancy plot options</span></span>
<span id="cb1246-10"><a href="mod-11.html#cb1246-10" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>), <span class="at">mar=</span><span class="fu">c</span>(<span class="fl">5.1</span>, <span class="fl">7.1</span>, <span class="fl">1.1</span>, <span class="fl">1.1</span>),</span>
<span id="cb1246-11"><a href="mod-11.html#cb1246-11" tabindex="-1"></a>    <span class="at">las=</span><span class="dv">1</span>, <span class="at">lend=</span><span class="dv">1</span>, <span class="at">bty=</span><span class="st">&quot;n&quot;</span>, </span>
<span id="cb1246-12"><a href="mod-11.html#cb1246-12" tabindex="-1"></a>    <span class="at">cex.axis=</span><span class="fl">1.3</span>, <span class="at">cex.lab=</span><span class="fl">1.3</span>)</span>
<span id="cb1246-13"><a href="mod-11.html#cb1246-13" tabindex="-1"></a><span class="co"># make the plot</span></span>
<span id="cb1246-14"><a href="mod-11.html#cb1246-14" tabindex="-1"></a><span class="fu">plot</span>(dat<span class="sc">$</span>ni, dat<span class="sc">$</span>res,</span>
<span id="cb1246-15"><a href="mod-11.html#cb1246-15" tabindex="-1"></a>     <span class="at">col=</span><span class="fu">ifelse</span>(dat<span class="sc">$</span>water<span class="sc">==</span><span class="st">&quot;soft&quot;</span>, <span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>),</span>
<span id="cb1246-16"><a href="mod-11.html#cb1246-16" tabindex="-1"></a>     <span class="at">pch=</span><span class="dv">16</span>, <span class="at">cex=</span><span class="fl">1.2</span>,</span>
<span id="cb1246-17"><a href="mod-11.html#cb1246-17" tabindex="-1"></a>     <span class="at">xlab=</span><span class="fu">expression</span>(Ni<span class="sc">~</span>concentration<span class="sc">~</span>(mu<span class="sc">*</span>g<span class="sc">/</span>L)),</span>
<span id="cb1246-18"><a href="mod-11.html#cb1246-18" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb1246-19"><a href="mod-11.html#cb1246-19" tabindex="-1"></a><span class="co"># custom y axis (need to move outwards)</span></span>
<span id="cb1246-20"><a href="mod-11.html#cb1246-20" tabindex="-1"></a><span class="fu">title</span>(<span class="at">ylab=</span><span class="fu">expression</span>(Ni<span class="sc">~</span>body<span class="sc">~</span>residue<span class="sc">~</span>(mu<span class="sc">*</span>g<span class="sc">/</span>kg)), <span class="at">line=</span><span class="dv">5</span>)</span>
<span id="cb1246-21"><a href="mod-11.html#cb1246-21" tabindex="-1"></a><span class="fu">points</span>(px, lo1, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1246-22"><a href="mod-11.html#cb1246-22" tabindex="-1"></a><span class="fu">points</span>(px, up1, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1246-23"><a href="mod-11.html#cb1246-23" tabindex="-1"></a><span class="fu">points</span>(px, mn1, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>)</span>
<span id="cb1246-24"><a href="mod-11.html#cb1246-24" tabindex="-1"></a><span class="fu">points</span>(px, lo2, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1246-25"><a href="mod-11.html#cb1246-25" tabindex="-1"></a><span class="fu">points</span>(px, up2, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1246-26"><a href="mod-11.html#cb1246-26" tabindex="-1"></a><span class="fu">points</span>(px, mn2, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>)</span>
<span id="cb1246-27"><a href="mod-11.html#cb1246-27" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;topleft&quot;</span>,</span>
<span id="cb1246-28"><a href="mod-11.html#cb1246-28" tabindex="-1"></a>       <span class="at">legend=</span><span class="fu">c</span>(<span class="st">&quot;Soft water&quot;</span>, <span class="st">&quot;Hard water&quot;</span>),</span>
<span id="cb1246-29"><a href="mod-11.html#cb1246-29" tabindex="-1"></a>       <span class="at">col=</span><span class="fu">c</span>(<span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>),</span>
<span id="cb1246-30"><a href="mod-11.html#cb1246-30" tabindex="-1"></a>       <span class="at">pch=</span><span class="dv">16</span>, <span class="at">bty=</span><span class="st">&quot;n&quot;</span>, <span class="at">cex=</span><span class="fl">1.2</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-721-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<div id="alternative-strategies-for-the-analysis" class="section level3 hasAnchor" number="11.3.3">
<h3><span class="header-section-number">11.3.3</span> Alternative strategies for the analysis<a href="mod-11.html#alternative-strategies-for-the-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In the example above, we fit two separate models to understand the effect of a continuous predictor in two groups: hard water and soft water. It is also possible to estimate the effects in a single model in nls(). This produces a NLS model analogous to ANCOVA: the response variable can vary between groups, and/or the response to the continuous predictor can vary between groups. The models are little more complicated than the analogous LM. We can define models where either the asymptote (<em>a</em>), the Michaelis constant (<em>b</em>), or both <em>a</em> and <em>b</em> vary by group. The deterministic parts of these models can be written compactly as:</p>
<table>
<colgroup>
<col width="44%" />
<col width="55%" />
</colgroup>
<thead>
<tr class="header">
<th>Model</th>
<th align="center">Deterministic part</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Asymptote by group</td>
<td align="center"><span class="math inline">\(\eta=\frac{a_{water}X}{b+X}\)</span></td>
</tr>
<tr class="even">
<td>Michaelis constant by group</td>
<td align="center"><span class="math inline">\(\eta=\frac{aX}{b_{water}+X}\)</span></td>
</tr>
<tr class="odd">
<td>Asymptote and Michaelis constant by group</td>
<td align="center"><span class="math inline">\(\eta=\frac{a_{water}X}{b_{water}+X}\)</span></td>
</tr>
</tbody>
</table>
<p>These models can’t be coded directly in <code>nls()</code>. Instead, we need to rewrite the model first to include some vectors of 1s and 0s to isolate the observations in each treatment. The equations below show what that looks like.</p>
<table>
<colgroup>
<col width="44%" />
<col width="55%" />
</colgroup>
<thead>
<tr class="header">
<th>Model</th>
<th align="center">Deterministic part</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Asymptote by group</td>
<td align="center"><span class="math inline">\(\eta=\frac{{\psi_{hard}a}_{hard}X}{b+X}+\frac{{\psi_{soft}a}_{soft}X}{b+X}\)</span></td>
</tr>
<tr class="even">
<td>Michaelis constant by group</td>
<td align="center"><span class="math inline">\(\eta=\frac{\psi_{hard}aX}{b_{hard}+X}+\frac{\psi_{soft}aX}{b_{soft}+X}\)</span></td>
</tr>
<tr class="odd">
<td>Asymptote and Michaelis constant by group</td>
<td align="center"><span class="math inline">\(\eta=\frac{{\psi_{hard}a}_{hard}X}{b_{hard}+X}+\frac{{\psi_{soft}a}_{soft}X}{b_{soft}+X}\)</span></td>
</tr>
</tbody>
</table>
<p>In these equations the coefficient <span class="math inline">\(\psi_{hard}\)</span> takes value 1 for observations from the hard water group and value 0 for observations in the soft water group (<span class="math inline">\(\psi\)</span> is the Greek letter “psi” and is often used for probabilities or binary variables). The coefficient <span class="math inline">\(\psi_{soft}\)</span> does exactly the opposite: takes value 1 for observations from the soft water group and value 0 for observations in the hard water group. This form means that the parameters specific to hard water cannot affect observations from soft water and vice versa. These sorts of parameters are called <strong>dummy variables</strong><a href="#fn85" class="footnote-ref" id="fnref85"><sup>85</sup></a>.</p>
<p>We can code each of these models in <code>nls()</code>. For starting values we can use the parameter estimates from the models fitted to each group separately (<code>mod1</code> and <code>mod2</code>).</p>
<div class="sourceCode" id="cb1247"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1247-1"><a href="mod-11.html#cb1247-1" tabindex="-1"></a>flag3 <span class="ot">&lt;-</span> <span class="fu">which</span>(dat<span class="sc">$</span>ni <span class="sc">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb1247-2"><a href="mod-11.html#cb1247-2" tabindex="-1"></a>dat2 <span class="ot">&lt;-</span> dat[flag3,]</span>
<span id="cb1247-3"><a href="mod-11.html#cb1247-3" tabindex="-1"></a></span>
<span id="cb1247-4"><a href="mod-11.html#cb1247-4" tabindex="-1"></a>dat2<span class="sc">$</span>psi.hard <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(dat2<span class="sc">$</span>water <span class="sc">==</span> <span class="st">&quot;hard&quot;</span>)</span>
<span id="cb1247-5"><a href="mod-11.html#cb1247-5" tabindex="-1"></a>dat2<span class="sc">$</span>psi.soft <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(dat2<span class="sc">$</span>water <span class="sc">==</span> <span class="st">&quot;soft&quot;</span>)</span>
<span id="cb1247-6"><a href="mod-11.html#cb1247-6" tabindex="-1"></a></span>
<span id="cb1247-7"><a href="mod-11.html#cb1247-7" tabindex="-1"></a><span class="co"># asymptote by group:</span></span>
<span id="cb1247-8"><a href="mod-11.html#cb1247-8" tabindex="-1"></a>mod3 <span class="ot">&lt;-</span> <span class="fu">nls</span>(res<span class="sc">~</span></span>
<span id="cb1247-9"><a href="mod-11.html#cb1247-9" tabindex="-1"></a>            (psi.hard)<span class="sc">*</span>((a.hard<span class="sc">*</span>ni)<span class="sc">/</span>(b<span class="sc">+</span>ni))<span class="sc">+</span></span>
<span id="cb1247-10"><a href="mod-11.html#cb1247-10" tabindex="-1"></a>            (psi.soft)<span class="sc">*</span>((a.soft<span class="sc">*</span>ni)<span class="sc">/</span>(b<span class="sc">+</span>ni)),</span>
<span id="cb1247-11"><a href="mod-11.html#cb1247-11" tabindex="-1"></a>            <span class="at">data=</span>dat2, </span>
<span id="cb1247-12"><a href="mod-11.html#cb1247-12" tabindex="-1"></a>            <span class="at">start=</span><span class="fu">list</span>(<span class="at">a.hard=</span><span class="dv">25038</span>, <span class="at">a.soft=</span><span class="dv">71997</span>, <span class="at">b=</span><span class="dv">2000</span>))</span>
<span id="cb1247-13"><a href="mod-11.html#cb1247-13" tabindex="-1"></a><span class="co"># Michaelis constant by group:</span></span>
<span id="cb1247-14"><a href="mod-11.html#cb1247-14" tabindex="-1"></a>mod4 <span class="ot">&lt;-</span> <span class="fu">nls</span>(res<span class="sc">~</span></span>
<span id="cb1247-15"><a href="mod-11.html#cb1247-15" tabindex="-1"></a>            (psi.hard)<span class="sc">*</span>((a<span class="sc">*</span>ni)<span class="sc">/</span>(b.hard<span class="sc">+</span>ni))<span class="sc">+</span></span>
<span id="cb1247-16"><a href="mod-11.html#cb1247-16" tabindex="-1"></a>            (psi.soft)<span class="sc">*</span>((a<span class="sc">*</span>ni)<span class="sc">/</span>(b.soft<span class="sc">+</span>ni)),</span>
<span id="cb1247-17"><a href="mod-11.html#cb1247-17" tabindex="-1"></a>            <span class="at">data=</span>dat2, </span>
<span id="cb1247-18"><a href="mod-11.html#cb1247-18" tabindex="-1"></a>            <span class="at">start=</span><span class="fu">list</span>(<span class="at">a=</span><span class="dv">48500</span>, <span class="at">b.hard=</span><span class="dv">1928</span>, <span class="at">b.soft=</span><span class="dv">4310</span>))</span>
<span id="cb1247-19"><a href="mod-11.html#cb1247-19" tabindex="-1"></a><span class="co"># both parameters by group</span></span>
<span id="cb1247-20"><a href="mod-11.html#cb1247-20" tabindex="-1"></a>mod5 <span class="ot">&lt;-</span> <span class="fu">nls</span>(res<span class="sc">~</span></span>
<span id="cb1247-21"><a href="mod-11.html#cb1247-21" tabindex="-1"></a>            (psi.hard)<span class="sc">*</span>((a.hard<span class="sc">*</span>ni)<span class="sc">/</span>(b.hard<span class="sc">+</span>ni))<span class="sc">+</span></span>
<span id="cb1247-22"><a href="mod-11.html#cb1247-22" tabindex="-1"></a>            (psi.soft)<span class="sc">*</span>((a.soft<span class="sc">*</span>ni)<span class="sc">/</span>(b.soft<span class="sc">+</span>ni)),</span>
<span id="cb1247-23"><a href="mod-11.html#cb1247-23" tabindex="-1"></a>            <span class="at">data=</span>dat2, </span>
<span id="cb1247-24"><a href="mod-11.html#cb1247-24" tabindex="-1"></a>            <span class="at">start=</span><span class="fu">list</span>(<span class="at">a.hard=</span><span class="dv">20000</span>, <span class="at">a.soft=</span><span class="dv">50000</span>,</span>
<span id="cb1247-25"><a href="mod-11.html#cb1247-25" tabindex="-1"></a>                       <span class="at">b.hard=</span><span class="dv">1000</span>, <span class="at">b.soft=</span><span class="dv">4300</span>))</span></code></pre></div>
<p>For complex models like model 5, you may get the dreaded <code>singular gradient</code> error in <code>nls()</code>, which means that R can’t find a solution from the starting values you provided. You might be able to get the model to work by changing the starting values, but probably not. We’ll come back to model 5 later and explore an alternative way to fit it.</p>
<p>We can use AIC to compare the fits of the models that did work:</p>
<div class="sourceCode" id="cb1248"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1248-1"><a href="mod-11.html#cb1248-1" tabindex="-1"></a><span class="fu">AIC</span>(mod3, mod4, mod5)</span></code></pre></div>
<pre><code>##      df      AIC
## mod3  4 1256.312
## mod4  4 1293.576
## mod5  5 1250.118</code></pre>
<p>We can also use package <code>propagate</code> to calculate fitted values and 95% CI.</p>
<div class="sourceCode" id="cb1250"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1250-1"><a href="mod-11.html#cb1250-1" tabindex="-1"></a>px <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fu">min</span>(dat2<span class="sc">$</span>ni), <span class="fu">max</span>(dat2<span class="sc">$</span>ni), <span class="at">length=</span><span class="dv">20</span>)</span>
<span id="cb1250-2"><a href="mod-11.html#cb1250-2" tabindex="-1"></a>dx <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="at">ni=</span>px, <span class="at">psi.hard=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">0</span>))</span>
<span id="cb1250-3"><a href="mod-11.html#cb1250-3" tabindex="-1"></a>dx<span class="sc">$</span>psi.soft <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(dx<span class="sc">$</span>psi.hard<span class="sc">==</span><span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>)</span>
<span id="cb1250-4"><a href="mod-11.html#cb1250-4" tabindex="-1"></a></span>
<span id="cb1250-5"><a href="mod-11.html#cb1250-5" tabindex="-1"></a>pred <span class="ot">&lt;-</span> <span class="fu">predictNLS</span>(mod5,</span>
<span id="cb1250-6"><a href="mod-11.html#cb1250-6" tabindex="-1"></a>                   <span class="at">newdata=</span><span class="fu">data.frame</span>(dx),</span>
<span id="cb1250-7"><a href="mod-11.html#cb1250-7" tabindex="-1"></a>                   <span class="at">interval=</span><span class="st">&quot;confidence&quot;</span>)</span>
<span id="cb1250-8"><a href="mod-11.html#cb1250-8" tabindex="-1"></a>pred <span class="ot">&lt;-</span> pred<span class="sc">$</span>summary</span></code></pre></div>
<p>Extract the values you need for a plot:</p>
<div class="sourceCode" id="cb1251"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1251-1"><a href="mod-11.html#cb1251-1" tabindex="-1"></a>flag.soft <span class="ot">&lt;-</span> <span class="fu">which</span>(dx<span class="sc">$</span>psi.soft <span class="sc">==</span> <span class="dv">1</span>)</span>
<span id="cb1251-2"><a href="mod-11.html#cb1251-2" tabindex="-1"></a>flag.hard <span class="ot">&lt;-</span> <span class="fu">which</span>(dx<span class="sc">$</span>psi.hard <span class="sc">==</span> <span class="dv">1</span>)</span>
<span id="cb1251-3"><a href="mod-11.html#cb1251-3" tabindex="-1"></a></span>
<span id="cb1251-4"><a href="mod-11.html#cb1251-4" tabindex="-1"></a>mn.soft <span class="ot">&lt;-</span> pred<span class="sc">$</span>Sim.Mean[flag.soft]</span>
<span id="cb1251-5"><a href="mod-11.html#cb1251-5" tabindex="-1"></a>lo.soft <span class="ot">&lt;-</span> pred<span class="sc">$</span><span class="st">&#39;Sim.2.5%&#39;</span>[flag.soft]</span>
<span id="cb1251-6"><a href="mod-11.html#cb1251-6" tabindex="-1"></a>up.soft <span class="ot">&lt;-</span> pred<span class="sc">$</span><span class="st">&#39;Sim.97.5%&#39;</span>[flag.soft]</span>
<span id="cb1251-7"><a href="mod-11.html#cb1251-7" tabindex="-1"></a></span>
<span id="cb1251-8"><a href="mod-11.html#cb1251-8" tabindex="-1"></a>mn.hard <span class="ot">&lt;-</span> pred<span class="sc">$</span>Sim.Mean[flag.hard]</span>
<span id="cb1251-9"><a href="mod-11.html#cb1251-9" tabindex="-1"></a>lo.hard <span class="ot">&lt;-</span> pred<span class="sc">$</span><span class="st">&#39;Sim.2.5%&#39;</span>[flag.hard]</span>
<span id="cb1251-10"><a href="mod-11.html#cb1251-10" tabindex="-1"></a>up.hard <span class="ot">&lt;-</span> pred<span class="sc">$</span><span class="st">&#39;Sim.97.5%&#39;</span>[flag.hard]</span></code></pre></div>
<p>Now make a fancy plot:</p>
<div class="sourceCode" id="cb1252"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1252-1"><a href="mod-11.html#cb1252-1" tabindex="-1"></a><span class="co"># fancy plot options</span></span>
<span id="cb1252-2"><a href="mod-11.html#cb1252-2" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>), <span class="at">mar=</span><span class="fu">c</span>(<span class="fl">5.1</span>, <span class="fl">7.1</span>, <span class="fl">1.1</span>, <span class="fl">1.1</span>),</span>
<span id="cb1252-3"><a href="mod-11.html#cb1252-3" tabindex="-1"></a>    <span class="at">las=</span><span class="dv">1</span>, <span class="at">lend=</span><span class="dv">1</span>, <span class="at">bty=</span><span class="st">&quot;n&quot;</span>, </span>
<span id="cb1252-4"><a href="mod-11.html#cb1252-4" tabindex="-1"></a>    <span class="at">cex.axis=</span><span class="fl">1.3</span>, <span class="at">cex.lab=</span><span class="fl">1.3</span>)</span>
<span id="cb1252-5"><a href="mod-11.html#cb1252-5" tabindex="-1"></a><span class="co"># make the plot</span></span>
<span id="cb1252-6"><a href="mod-11.html#cb1252-6" tabindex="-1"></a><span class="fu">plot</span>(dat2<span class="sc">$</span>ni, dat2<span class="sc">$</span>res,</span>
<span id="cb1252-7"><a href="mod-11.html#cb1252-7" tabindex="-1"></a>     <span class="at">col=</span><span class="fu">ifelse</span>(dat2<span class="sc">$</span>water<span class="sc">==</span><span class="st">&quot;soft&quot;</span>, <span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>),</span>
<span id="cb1252-8"><a href="mod-11.html#cb1252-8" tabindex="-1"></a>     <span class="at">pch=</span><span class="dv">16</span>, <span class="at">cex=</span><span class="fl">1.2</span>,</span>
<span id="cb1252-9"><a href="mod-11.html#cb1252-9" tabindex="-1"></a>     <span class="at">xlab=</span><span class="fu">expression</span>(Ni<span class="sc">~</span>concentration<span class="sc">~</span>(mu<span class="sc">*</span>g<span class="sc">/</span>L)),</span>
<span id="cb1252-10"><a href="mod-11.html#cb1252-10" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb1252-11"><a href="mod-11.html#cb1252-11" tabindex="-1"></a><span class="co"># custom y axis (need to move outwards)</span></span>
<span id="cb1252-12"><a href="mod-11.html#cb1252-12" tabindex="-1"></a><span class="fu">title</span>(<span class="at">ylab=</span><span class="fu">expression</span>(Ni<span class="sc">~</span>body<span class="sc">~</span>residue<span class="sc">~</span>(mu<span class="sc">*</span>g<span class="sc">/</span>kg)), <span class="at">line=</span><span class="dv">5</span>)</span>
<span id="cb1252-13"><a href="mod-11.html#cb1252-13" tabindex="-1"></a><span class="fu">points</span>(px, lo.soft, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1252-14"><a href="mod-11.html#cb1252-14" tabindex="-1"></a><span class="fu">points</span>(px, up.soft, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1252-15"><a href="mod-11.html#cb1252-15" tabindex="-1"></a><span class="fu">points</span>(px, mn.soft, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>)</span>
<span id="cb1252-16"><a href="mod-11.html#cb1252-16" tabindex="-1"></a><span class="fu">points</span>(px, lo.hard, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1252-17"><a href="mod-11.html#cb1252-17" tabindex="-1"></a><span class="fu">points</span>(px, up.hard, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1252-18"><a href="mod-11.html#cb1252-18" tabindex="-1"></a><span class="fu">points</span>(px, mn.hard, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>)</span>
<span id="cb1252-19"><a href="mod-11.html#cb1252-19" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;topleft&quot;</span>,</span>
<span id="cb1252-20"><a href="mod-11.html#cb1252-20" tabindex="-1"></a>       <span class="at">legend=</span><span class="fu">c</span>(<span class="st">&quot;Soft water&quot;</span>, <span class="st">&quot;Hard water&quot;</span>),</span>
<span id="cb1252-21"><a href="mod-11.html#cb1252-21" tabindex="-1"></a>       <span class="at">col=</span><span class="fu">c</span>(<span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>),</span>
<span id="cb1252-22"><a href="mod-11.html#cb1252-22" tabindex="-1"></a>       <span class="at">pch=</span><span class="dv">16</span>, <span class="at">bty=</span><span class="st">&quot;n&quot;</span>, <span class="at">cex=</span><span class="fl">1.2</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-726-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>Complicated nonlinear models like model 5 can be difficult to fit using the optimization algorithms in R. There is an alternative set of techniques called <strong>Markov chain Monte Carlo (MCMC)</strong> that can be used for fitting complex models. MCMC uses long chains of random samples to try lots of different possible values for each parameter, and adjusts parameters at each iteration (step in the chain) depending on how well or poorly the model fits with the current parameters. Using MCMC turns your analysis into a Bayesian analysis. A more in-depth explanation of Bayesian inference was provided in <a href="mod-01.html#mod-01-bayes">Module 1</a>.</p>
<p>For MCMC we will need the program JAGS as well as the R packages <code>rjags</code> and <code>R2jags</code>.</p>
<div class="sourceCode" id="cb1253"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1253-1"><a href="mod-11.html#cb1253-1" tabindex="-1"></a>mod.name <span class="ot">&lt;-</span> <span class="st">&quot;mod05.txt&quot;</span></span>
<span id="cb1253-2"><a href="mod-11.html#cb1253-2" tabindex="-1"></a><span class="fu">sink</span>(mod.name)</span>
<span id="cb1253-3"><a href="mod-11.html#cb1253-3" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">&quot;</span></span>
<span id="cb1253-4"><a href="mod-11.html#cb1253-4" tabindex="-1"></a><span class="st">    model{</span></span>
<span id="cb1253-5"><a href="mod-11.html#cb1253-5" tabindex="-1"></a><span class="st">        # priors</span></span>
<span id="cb1253-6"><a href="mod-11.html#cb1253-6" tabindex="-1"></a><span class="st">        ## element 1 = soft</span></span>
<span id="cb1253-7"><a href="mod-11.html#cb1253-7" tabindex="-1"></a><span class="st">        ## element 2 = hard</span></span>
<span id="cb1253-8"><a href="mod-11.html#cb1253-8" tabindex="-1"></a><span class="st">        ### asymptote         </span></span>
<span id="cb1253-9"><a href="mod-11.html#cb1253-9" tabindex="-1"></a><span class="st">        for(i in 1:2){a.water[i]~dunif(1e3, 1e5)}</span></span>
<span id="cb1253-10"><a href="mod-11.html#cb1253-10" tabindex="-1"></a><span class="st">        ### Michaelis constant</span></span>
<span id="cb1253-11"><a href="mod-11.html#cb1253-11" tabindex="-1"></a><span class="st">        for(i in 1:2){b.water[i]~dunif(1e2, 1e4)}</span></span>
<span id="cb1253-12"><a href="mod-11.html#cb1253-12" tabindex="-1"></a><span class="st">        ## residual SD -&gt; precision (tau)</span></span>
<span id="cb1253-13"><a href="mod-11.html#cb1253-13" tabindex="-1"></a><span class="st">        sigma ~ dunif(0, 1e4)</span></span>
<span id="cb1253-14"><a href="mod-11.html#cb1253-14" tabindex="-1"></a><span class="st">        tau.y &lt;- 1 / (sigma * sigma)</span></span>
<span id="cb1253-15"><a href="mod-11.html#cb1253-15" tabindex="-1"></a><span class="st">        # likelihood</span></span>
<span id="cb1253-16"><a href="mod-11.html#cb1253-16" tabindex="-1"></a><span class="st">        for(i in 1:N){</span></span>
<span id="cb1253-17"><a href="mod-11.html#cb1253-17" tabindex="-1"></a><span class="st">            y[i] ~ dnorm(eta[i], tau.y)</span></span>
<span id="cb1253-18"><a href="mod-11.html#cb1253-18" tabindex="-1"></a><span class="st">            eta[i] &lt;- (a.water[water[i]] *</span></span>
<span id="cb1253-19"><a href="mod-11.html#cb1253-19" tabindex="-1"></a><span class="st">                x[i])/(b.water[water[i]]+x[i])</span></span>
<span id="cb1253-20"><a href="mod-11.html#cb1253-20" tabindex="-1"></a><span class="st">        }# i for N</span></span>
<span id="cb1253-21"><a href="mod-11.html#cb1253-21" tabindex="-1"></a><span class="st">    }#model</span></span>
<span id="cb1253-22"><a href="mod-11.html#cb1253-22" tabindex="-1"></a><span class="st">&quot;</span>, <span class="at">fill=</span><span class="cn">TRUE</span>)</span>
<span id="cb1253-23"><a href="mod-11.html#cb1253-23" tabindex="-1"></a><span class="fu">sink</span>()</span></code></pre></div>
<div class="sourceCode" id="cb1254"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1254-1"><a href="mod-11.html#cb1254-1" tabindex="-1"></a><span class="fu">library</span>(R2jags)</span>
<span id="cb1254-2"><a href="mod-11.html#cb1254-2" tabindex="-1"></a><span class="fu">library</span>(rjags)</span>
<span id="cb1254-3"><a href="mod-11.html#cb1254-3" tabindex="-1"></a><span class="co"># define initial values for MCMC chains</span></span>
<span id="cb1254-4"><a href="mod-11.html#cb1254-4" tabindex="-1"></a>init.fun <span class="ot">&lt;-</span> <span class="cf">function</span>(nc){</span>
<span id="cb1254-5"><a href="mod-11.html#cb1254-5" tabindex="-1"></a>    res <span class="ot">&lt;-</span> <span class="fu">vector</span>(<span class="st">&quot;list&quot;</span>, <span class="at">length=</span>nc)</span>
<span id="cb1254-6"><a href="mod-11.html#cb1254-6" tabindex="-1"></a>    <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>nc){</span>
<span id="cb1254-7"><a href="mod-11.html#cb1254-7" tabindex="-1"></a>        res[[i]] <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb1254-8"><a href="mod-11.html#cb1254-8" tabindex="-1"></a>            <span class="at">a.water=</span><span class="fu">runif</span>(<span class="dv">2</span>, <span class="fl">1e3</span>, <span class="fl">1e5</span>),</span>
<span id="cb1254-9"><a href="mod-11.html#cb1254-9" tabindex="-1"></a>            <span class="at">b.water=</span><span class="fu">runif</span>(<span class="dv">2</span>, <span class="fl">1e2</span>, <span class="fl">1e4</span>),</span>
<span id="cb1254-10"><a href="mod-11.html#cb1254-10" tabindex="-1"></a>            <span class="at">sigma=</span><span class="fu">runif</span>(<span class="dv">1</span>, <span class="fl">0.1</span>, <span class="fl">1e4</span>))</span>
<span id="cb1254-11"><a href="mod-11.html#cb1254-11" tabindex="-1"></a>    }<span class="co">#i</span></span>
<span id="cb1254-12"><a href="mod-11.html#cb1254-12" tabindex="-1"></a>    <span class="fu">return</span>(res)</span>
<span id="cb1254-13"><a href="mod-11.html#cb1254-13" tabindex="-1"></a>}</span>
<span id="cb1254-14"><a href="mod-11.html#cb1254-14" tabindex="-1"></a></span>
<span id="cb1254-15"><a href="mod-11.html#cb1254-15" tabindex="-1"></a>nchains <span class="ot">&lt;-</span> <span class="dv">3</span></span>
<span id="cb1254-16"><a href="mod-11.html#cb1254-16" tabindex="-1"></a>inits <span class="ot">&lt;-</span> <span class="fu">init.fun</span>(nchains)</span>
<span id="cb1254-17"><a href="mod-11.html#cb1254-17" tabindex="-1"></a></span>
<span id="cb1254-18"><a href="mod-11.html#cb1254-18" tabindex="-1"></a><span class="co"># parameters to monitor</span></span>
<span id="cb1254-19"><a href="mod-11.html#cb1254-19" tabindex="-1"></a>params <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;a.water&quot;</span>, <span class="st">&quot;b.water&quot;</span>, <span class="st">&quot;sigma&quot;</span>)</span>
<span id="cb1254-20"><a href="mod-11.html#cb1254-20" tabindex="-1"></a></span>
<span id="cb1254-21"><a href="mod-11.html#cb1254-21" tabindex="-1"></a><span class="co"># MCMC parameters</span></span>
<span id="cb1254-22"><a href="mod-11.html#cb1254-22" tabindex="-1"></a>n.iter <span class="ot">&lt;-</span> <span class="fl">2e4</span></span>
<span id="cb1254-23"><a href="mod-11.html#cb1254-23" tabindex="-1"></a>n.burnin <span class="ot">&lt;-</span> <span class="fl">4e3</span></span>
<span id="cb1254-24"><a href="mod-11.html#cb1254-24" tabindex="-1"></a>n.thin <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb1254-25"><a href="mod-11.html#cb1254-25" tabindex="-1"></a></span>
<span id="cb1254-26"><a href="mod-11.html#cb1254-26" tabindex="-1"></a><span class="co"># package data for JAGS</span></span>
<span id="cb1254-27"><a href="mod-11.html#cb1254-27" tabindex="-1"></a>in.data <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">y=</span>dat2<span class="sc">$</span>res,</span>
<span id="cb1254-28"><a href="mod-11.html#cb1254-28" tabindex="-1"></a>                <span class="at">x=</span>dat2<span class="sc">$</span>ni,</span>
<span id="cb1254-29"><a href="mod-11.html#cb1254-29" tabindex="-1"></a>                <span class="at">N=</span><span class="fu">nrow</span>(dat2),</span>
<span id="cb1254-30"><a href="mod-11.html#cb1254-30" tabindex="-1"></a>                <span class="at">water=</span>dat2<span class="sc">$</span>psi.hard<span class="sc">+</span><span class="dv">1</span>)</span>
<span id="cb1254-31"><a href="mod-11.html#cb1254-31" tabindex="-1"></a></span>
<span id="cb1254-32"><a href="mod-11.html#cb1254-32" tabindex="-1"></a>model05 <span class="ot">&lt;-</span> <span class="fu">jags</span>(<span class="at">data=</span>in.data, <span class="at">inits=</span>inits,</span>
<span id="cb1254-33"><a href="mod-11.html#cb1254-33" tabindex="-1"></a>                <span class="at">parameters.to.save=</span>params,</span>
<span id="cb1254-34"><a href="mod-11.html#cb1254-34" tabindex="-1"></a>                <span class="at">model.file=</span>mod.name,</span>
<span id="cb1254-35"><a href="mod-11.html#cb1254-35" tabindex="-1"></a>                <span class="at">n.chains=</span>nchains, <span class="at">n.iter=</span>n.iter,</span>
<span id="cb1254-36"><a href="mod-11.html#cb1254-36" tabindex="-1"></a>                <span class="at">n.burnin=</span>n.burnin, <span class="at">n.thin=</span>n.thin)<span class="co">#jags</span></span></code></pre></div>
<pre><code>## Compiling model graph
##    Resolving undeclared variables
##    Allocating nodes
## Graph information:
##    Observed stochastic nodes: 66
##    Unobserved stochastic nodes: 5
##    Total graph size: 278
## 
## Initializing model</code></pre>
<p>The most important part of the output is here:</p>
<div class="sourceCode" id="cb1256"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1256-1"><a href="mod-11.html#cb1256-1" tabindex="-1"></a>out <span class="ot">&lt;-</span> model05<span class="sc">$</span>BUGSoutput<span class="sc">$</span>summary</span>
<span id="cb1256-2"><a href="mod-11.html#cb1256-2" tabindex="-1"></a></span>
<span id="cb1256-3"><a href="mod-11.html#cb1256-3" tabindex="-1"></a><span class="co"># rounded to fit better on screen:</span></span>
<span id="cb1256-4"><a href="mod-11.html#cb1256-4" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">round</span>(out, <span class="dv">1</span>))</span></code></pre></div>
<pre><code>##               mean     sd    2.5%     25%     50%     75%   97.5% Rhat n.eff
## a.water[1] 72849.1 3257.2 66711.7 70672.3 72833.6 74848.2 79183.4    1   470
## a.water[2] 25761.0 1894.1 22570.3 24323.8 25674.6 27011.9 29616.5    1   480
## b.water[1]  4480.2  543.7  3489.4  4098.9  4469.4  4823.9  5608.4    1   480
## b.water[2]  2208.8  602.6  1242.1  1783.3  2147.1  2528.5  3644.3    1   480
## deviance    1245.9    3.5  1241.0  1243.4  1245.1  1247.7  1254.0    1   480
## sigma       3066.0  283.0  2546.9  2860.5  3046.3  3244.2  3676.0    1   480</code></pre>
<p>The parameter estimates (mean) are close to their true values. The next few columns show SD and various quantiles of the posterior distribution of each parameter. The Gelman-Rubin statistic, <span class="math inline">\(\hat{R}\)</span> (“R-hat”) are &lt;1.01 for each parameter, suggesting that the MCMC chains converged. This means that the 3 chains reached about the same distribution for a parameter.</p>
<p>Using JAGS can be overkill sometimes, but might be the only way to get a model to fit (and there is no guarantee that JAGS can fit a model!). This method has the advantage of fitting the model directly, without adding in dummy variables for each treatment.</p>
<p>We can use the MCMC samples saved in the model object to generate model predictions. The intervals produced for this plot are not true confidence intervals (CI), but their Bayesian analogue, <strong>credible intervals (CRI)</strong>.</p>
<div class="sourceCode" id="cb1258"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1258-1"><a href="mod-11.html#cb1258-1" tabindex="-1"></a><span class="co"># MCMC samples</span></span>
<span id="cb1258-2"><a href="mod-11.html#cb1258-2" tabindex="-1"></a>z <span class="ot">&lt;-</span> model05<span class="sc">$</span>BUGSoutput<span class="sc">$</span>sims.list</span>
<span id="cb1258-3"><a href="mod-11.html#cb1258-3" tabindex="-1"></a></span>
<span id="cb1258-4"><a href="mod-11.html#cb1258-4" tabindex="-1"></a><span class="co"># make data frame for prediction:</span></span>
<span id="cb1258-5"><a href="mod-11.html#cb1258-5" tabindex="-1"></a>px <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fu">min</span>(dat2<span class="sc">$</span>ni), <span class="fu">max</span>(dat2<span class="sc">$</span>ni), <span class="at">length=</span><span class="dv">20</span>)</span>
<span id="cb1258-6"><a href="mod-11.html#cb1258-6" tabindex="-1"></a></span>
<span id="cb1258-7"><a href="mod-11.html#cb1258-7" tabindex="-1"></a><span class="co"># water: 1 = soft, 2 = hard</span></span>
<span id="cb1258-8"><a href="mod-11.html#cb1258-8" tabindex="-1"></a>dx <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="at">ni=</span>px, <span class="at">water=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>)</span>
<span id="cb1258-9"><a href="mod-11.html#cb1258-9" tabindex="-1"></a>dx<span class="sc">$</span>lo <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb1258-10"><a href="mod-11.html#cb1258-10" tabindex="-1"></a>dx<span class="sc">$</span>up <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb1258-11"><a href="mod-11.html#cb1258-11" tabindex="-1"></a>dx<span class="sc">$</span>mn <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb1258-12"><a href="mod-11.html#cb1258-12" tabindex="-1"></a></span>
<span id="cb1258-13"><a href="mod-11.html#cb1258-13" tabindex="-1"></a>mx.mn <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>, <span class="at">nrow=</span><span class="fu">nrow</span>(dx), <span class="at">ncol=</span><span class="fu">nrow</span>(z[[<span class="dv">1</span>]]))</span>
<span id="cb1258-14"><a href="mod-11.html#cb1258-14" tabindex="-1"></a>mx.lo <span class="ot">&lt;-</span> mx.mn</span>
<span id="cb1258-15"><a href="mod-11.html#cb1258-15" tabindex="-1"></a>mx.up <span class="ot">&lt;-</span> mx.mn</span>
<span id="cb1258-16"><a href="mod-11.html#cb1258-16" tabindex="-1"></a></span>
<span id="cb1258-17"><a href="mod-11.html#cb1258-17" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(mx.mn)){</span>
<span id="cb1258-18"><a href="mod-11.html#cb1258-18" tabindex="-1"></a>    eta <span class="ot">&lt;-</span> (z<span class="sc">$</span>a.water[i,dx<span class="sc">$</span>water]<span class="sc">*</span>dx<span class="sc">$</span>ni)<span class="sc">/</span>(</span>
<span id="cb1258-19"><a href="mod-11.html#cb1258-19" tabindex="-1"></a>            z<span class="sc">$</span>b.water[i,dx<span class="sc">$</span>water]<span class="sc">+</span>dx<span class="sc">$</span>ni)</span>
<span id="cb1258-20"><a href="mod-11.html#cb1258-20" tabindex="-1"></a>    mx.mn[,i] <span class="ot">&lt;-</span> eta</span>
<span id="cb1258-21"><a href="mod-11.html#cb1258-21" tabindex="-1"></a>    mx.lo[,i] <span class="ot">&lt;-</span> <span class="fu">pmax</span>(<span class="dv">0</span>, eta<span class="sc">-</span><span class="fu">mean</span>(z<span class="sc">$</span>sigma[,<span class="dv">1</span>]))</span>
<span id="cb1258-22"><a href="mod-11.html#cb1258-22" tabindex="-1"></a>    mx.up[,i] <span class="ot">&lt;-</span> eta<span class="sc">+</span><span class="fu">mean</span>(z<span class="sc">$</span>sigma[,<span class="dv">1</span>])</span>
<span id="cb1258-23"><a href="mod-11.html#cb1258-23" tabindex="-1"></a>}</span>
<span id="cb1258-24"><a href="mod-11.html#cb1258-24" tabindex="-1"></a>dx<span class="sc">$</span>lo <span class="ot">&lt;-</span> <span class="fu">apply</span>(mx.lo, <span class="dv">1</span>, median)</span>
<span id="cb1258-25"><a href="mod-11.html#cb1258-25" tabindex="-1"></a>dx<span class="sc">$</span>up <span class="ot">&lt;-</span> <span class="fu">apply</span>(mx.up, <span class="dv">1</span>, median)</span>
<span id="cb1258-26"><a href="mod-11.html#cb1258-26" tabindex="-1"></a>dx<span class="sc">$</span>mn <span class="ot">&lt;-</span> <span class="fu">apply</span>(mx.mn, <span class="dv">1</span>, median)</span>
<span id="cb1258-27"><a href="mod-11.html#cb1258-27" tabindex="-1"></a></span>
<span id="cb1258-28"><a href="mod-11.html#cb1258-28" tabindex="-1"></a><span class="co"># make the plot</span></span>
<span id="cb1258-29"><a href="mod-11.html#cb1258-29" tabindex="-1"></a>flag.soft <span class="ot">&lt;-</span> <span class="fu">which</span>(dx<span class="sc">$</span>water <span class="sc">==</span> <span class="dv">1</span>)</span>
<span id="cb1258-30"><a href="mod-11.html#cb1258-30" tabindex="-1"></a>flag.hard <span class="ot">&lt;-</span> <span class="fu">which</span>(dx<span class="sc">$</span>water <span class="sc">==</span> <span class="dv">2</span>)</span>
<span id="cb1258-31"><a href="mod-11.html#cb1258-31" tabindex="-1"></a></span>
<span id="cb1258-32"><a href="mod-11.html#cb1258-32" tabindex="-1"></a><span class="co"># fancy plot options</span></span>
<span id="cb1258-33"><a href="mod-11.html#cb1258-33" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>), <span class="at">mar=</span><span class="fu">c</span>(<span class="fl">5.1</span>, <span class="fl">7.1</span>, <span class="fl">1.1</span>, <span class="fl">1.1</span>),</span>
<span id="cb1258-34"><a href="mod-11.html#cb1258-34" tabindex="-1"></a>    <span class="at">las=</span><span class="dv">1</span>, <span class="at">lend=</span><span class="dv">1</span>, <span class="at">bty=</span><span class="st">&quot;n&quot;</span>, </span>
<span id="cb1258-35"><a href="mod-11.html#cb1258-35" tabindex="-1"></a>    <span class="at">cex.axis=</span><span class="fl">1.3</span>, <span class="at">cex.lab=</span><span class="fl">1.3</span>)</span>
<span id="cb1258-36"><a href="mod-11.html#cb1258-36" tabindex="-1"></a><span class="co"># make the plot</span></span>
<span id="cb1258-37"><a href="mod-11.html#cb1258-37" tabindex="-1"></a><span class="fu">plot</span>(dat2<span class="sc">$</span>ni, dat2<span class="sc">$</span>res,</span>
<span id="cb1258-38"><a href="mod-11.html#cb1258-38" tabindex="-1"></a>     <span class="at">col=</span><span class="fu">ifelse</span>(dat2<span class="sc">$</span>water<span class="sc">==</span><span class="st">&quot;soft&quot;</span>, <span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>),</span>
<span id="cb1258-39"><a href="mod-11.html#cb1258-39" tabindex="-1"></a>     <span class="at">pch=</span><span class="dv">16</span>, <span class="at">cex=</span><span class="fl">1.2</span>,</span>
<span id="cb1258-40"><a href="mod-11.html#cb1258-40" tabindex="-1"></a>     <span class="at">xlab=</span><span class="fu">expression</span>(Ni<span class="sc">~</span>concentration<span class="sc">~</span>(mu<span class="sc">*</span>g<span class="sc">/</span>L)),</span>
<span id="cb1258-41"><a href="mod-11.html#cb1258-41" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb1258-42"><a href="mod-11.html#cb1258-42" tabindex="-1"></a><span class="co"># custom y axis (need to move outwards)</span></span>
<span id="cb1258-43"><a href="mod-11.html#cb1258-43" tabindex="-1"></a><span class="fu">title</span>(<span class="at">ylab=</span><span class="fu">expression</span>(Ni<span class="sc">~</span>body<span class="sc">~</span>residue<span class="sc">~</span>(mu<span class="sc">*</span>g<span class="sc">/</span>kg)), <span class="at">line=</span><span class="dv">5</span>)</span>
<span id="cb1258-44"><a href="mod-11.html#cb1258-44" tabindex="-1"></a><span class="fu">points</span>(px, dx<span class="sc">$</span>lo[flag.soft], <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1258-45"><a href="mod-11.html#cb1258-45" tabindex="-1"></a><span class="fu">points</span>(px, dx<span class="sc">$</span>up[flag.soft], <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1258-46"><a href="mod-11.html#cb1258-46" tabindex="-1"></a><span class="fu">points</span>(px, dx<span class="sc">$</span>mn[flag.soft], <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>)</span>
<span id="cb1258-47"><a href="mod-11.html#cb1258-47" tabindex="-1"></a><span class="fu">points</span>(px, dx<span class="sc">$</span>lo[flag.hard], <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1258-48"><a href="mod-11.html#cb1258-48" tabindex="-1"></a><span class="fu">points</span>(px, dx<span class="sc">$</span>up[flag.hard], <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1258-49"><a href="mod-11.html#cb1258-49" tabindex="-1"></a><span class="fu">points</span>(px, dx<span class="sc">$</span>mn[flag.hard], <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>)</span>
<span id="cb1258-50"><a href="mod-11.html#cb1258-50" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;topleft&quot;</span>,</span>
<span id="cb1258-51"><a href="mod-11.html#cb1258-51" tabindex="-1"></a>       <span class="at">legend=</span><span class="fu">c</span>(<span class="st">&quot;Soft water&quot;</span>, <span class="st">&quot;Hard water&quot;</span>),</span>
<span id="cb1258-52"><a href="mod-11.html#cb1258-52" tabindex="-1"></a>       <span class="at">col=</span><span class="fu">c</span>(<span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>),</span>
<span id="cb1258-53"><a href="mod-11.html#cb1258-53" tabindex="-1"></a>       <span class="at">pch=</span><span class="dv">16</span>, <span class="at">bty=</span><span class="st">&quot;n&quot;</span>, <span class="at">cex=</span><span class="fl">1.2</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-730-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>In this case, we got a model fit similar to what <code>nls()</code> produced, and similar predicted values. For this case, JAGS and MCMC are probably more trouble than they are worth. If you are having trouble getting <code>nls()</code> to fit a model, MCMC may be a more effective alternative…if the model can be fit and if the pattern is really there.</p>
</div>
</div>
<div id="mod-11-grow" class="section level2 hasAnchor" number="11.4">
<h2><span class="header-section-number">11.4</span> Biological growth curves<a href="mod-11.html#mod-11-grow" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This section explores several common biological <strong>growth curves</strong>. Living organisms often grow in nonlinear ways. For example, humans grow much faster during their first two years of life than during the rest of their lives. Biologists have developed many mathematical models to describe these patterns of growth. What is common to many of these models is that the rate of growth at any given time is proportional to the organism’s size at that time. Consequently, biological growth models often include at least one <strong>exponential</strong> term. For mathematical convenience, that exponential term is usually of the form <span class="math inline">\(e^x\)</span>, where <em>e</em> is Euler’s constant (<span class="math inline">\(\approx\)</span> 2.718282), also known as the base of the natural logarithm.</p>
<div id="gompertz-and-von-bertalanffy-curves" class="section level3 hasAnchor" number="11.4.1">
<h3><span class="header-section-number">11.4.1</span> Gompertz and von Bertalanffy curves<a href="mod-11.html#gompertz-and-von-bertalanffy-curves" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We will practice fitting two common growth curves, the <strong>Gompertz curve</strong> and the <strong>von Bertalanffy curve</strong>. Both are very commonly used in biology to model the growth of individual organisms. The Gompertz curve is also used to model population growth. Both models are special cases of the <a href="https://en.wikipedia.org/wiki/Generalised_logistic_function">generalized logistic function</a>.</p>
<p>The Gompertz curve is written in several ways. You can use whichever form you want, but you need to be explicit about what you did. Most people use the second of the two forms presented below.</p>
<p><strong>Gompertz model form 1 (<span class="citation">Bolker (<a href="#ref-bolker2008ecological">2008</a>)</span>; page 96)</strong></p>
<p><span class="math display">\[f\left(x\right)=e^{-ae^{-bx}}\]</span></p>
<ul>
<li><strong><em>a</em></strong>: Displacement along <em>x</em>-axis (equivalent to <em>b</em> in form 2)</li>
<li><strong><em>b</em></strong>: Growth rate (similar to <em>c</em> in form 2)</li>
</ul>
<p>The figure below shows the effects of varying the parameters in form 1 of the Gompertz model. Notice that the asymptote is fixed at 1. Because of this, form 2 is more commonly used.</p>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-731-1.png" width="864" style="display: block; margin: auto;" /></p>
<p><strong>Gompertz model form 2 (more common)</strong></p>
<p><span class="math display">\[f\left(x\right)={ae}^{-e^{b-ct}}\]</span></p>
<ul>
<li><strong><em>a</em></strong>: Asymptote (maximum value)</li>
<li><strong><em>b</em></strong>: Displacement along <em>x</em>-axis (shifts left or right)</li>
<li><strong><em>c</em></strong>: Growth rate (similar to <em>b</em> in form 1)</li>
</ul>
<p>The figure below shows the effects of varying the parameters in the second form of the Gompertz model:</p>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-732-1.png" width="864" style="display: block; margin: auto;" /></p>
<p><strong>von Betalanffy model</strong></p>
<p>Like the Gompertz curve, the von Bertalanffy growth curve describes size increasing and approaching an asymptote. The von Bertalanffy model is commonly written as:</p>
<p><span class="math display">\[f\left(x\right)=L_\infty\left(1-e^{-k\left(x-t_0\right)}\right)\]</span></p>
<ul>
<li><strong><em><span class="math inline">\(L_{\infty}\)</span></em></strong>: Asymptotic length as <em>x</em> approaches <span class="math inline">\(\infty\)</span></li>
<li><strong><em>k</em></strong>: Growth rate</li>
<li><strong><em><span class="math inline">\(t_0\)</span></em></strong>: Theoretical time at which <span class="math inline">\(f(x)=0\)</span>. I.e., displacement along <em>x</em>-axis. Should be negative.</li>
</ul>
<p>The von Bertalanffy model is commonly used in fisheries to describe fish growth, but in principle it could be used for any situation where its parameters make sense. The same function is also sometimes called the <strong>monomolecular curve</strong>. Like the Gompertz curve, the von Bertalanffy model has several forms. The form given above is the most common. The figure below shows the effects of varying the parameters in the von Bertalanffy model:</p>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-733-1.png" width="864" style="display: block; margin: auto;" /></p>
<p>Both the Gompertz and von Bertalanffy curves can also be used to model decreasing phenomena, but this is not commonly done.</p>
</div>
<div id="example-with-real-data-5" class="section level3 hasAnchor" number="11.4.2">
<h3><span class="header-section-number">11.4.2</span> Example with real data<a href="mod-11.html#example-with-real-data-5" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><span class="citation">Woodward et al. (<a href="#ref-woodward2015">2015</a>)</span> studied the life history dynamics of the Cretaceous hadrosaurid dinosaur <em>Maiasaura peeblesorum</em> (below<a href="#fn86" class="footnote-ref" id="fnref86"><sup>86</sup></a>). They used microscopic growth rings in the tibias of museum specimens to estimate age at death and the sizes of several bones to estimate total body length and mass. We will use their age and mass data to fit growth curves for this species.</p>
<p><img src="11_03.jpg" width="90%" style="display: block; margin: auto;" /></p>
<p>The data are stored in the file <a href="https://greenquanteco.github.io/woodward2015data.csv">woodward2015data.csv</a>. Download the dataset and put it in your R home directory.</p>
<div class="sourceCode" id="cb1259"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1259-1"><a href="mod-11.html#cb1259-1" tabindex="-1"></a>in.name <span class="ot">&lt;-</span> <span class="st">&quot;woodward2015data.csv&quot;</span></span>
<span id="cb1259-2"><a href="mod-11.html#cb1259-2" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(in.name, <span class="at">header=</span><span class="cn">TRUE</span>)</span>
<span id="cb1259-3"><a href="mod-11.html#cb1259-3" tabindex="-1"></a><span class="fu">plot</span>(dat<span class="sc">$</span>age, dat<span class="sc">$</span>mass)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-735-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>Use <code>nls()</code> to fit the two growth curves. There is a self-starting Gompertz model available, so we’ll use that. There is no self-start von Bertalanffy model, so we’ll have to guess at the parameters. Based on the plot above, we might guess that the asymptote <span class="math inline">\(L_\infty\)</span> is about 4000, and that <span class="math inline">\(t_0\)</span> is about -1. Using those rough estimates, we can rearrange the von Bertalanffy model and solve for the missing term <em>k</em> as:</p>
<p><span class="math display">\[\hat{k}=\frac{-\log{\left(1-\frac{f\left(x\right)}{L_\infty}\right)}}{x-t_0}\]</span></p>
<p>Or in R:</p>
<div class="sourceCode" id="cb1260"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1260-1"><a href="mod-11.html#cb1260-1" tabindex="-1"></a>est.k <span class="ot">&lt;-</span> <span class="sc">-</span><span class="dv">1</span><span class="sc">*</span>(<span class="fu">log</span>(<span class="dv">1</span><span class="sc">-</span>dat<span class="sc">$</span>mass<span class="sc">/</span><span class="dv">4000</span>))<span class="sc">/</span>(dat<span class="sc">$</span>age<span class="sc">-</span> <span class="sc">-</span><span class="dv">1</span>)</span>
<span id="cb1260-2"><a href="mod-11.html#cb1260-2" tabindex="-1"></a><span class="fu">mean</span>(est.k)</span></code></pre></div>
<pre><code>## [1] 0.09268849</code></pre>
<p>The mean of the estimated <em>k</em> (<span class="math inline">\(\hat{k}\)</span>, or “k-hat”) is about 0.09. So, we’ll use that for our starting guess. Now we can fit both the Gompertz and von Bertalanffy models:</p>
<div class="sourceCode" id="cb1262"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1262-1"><a href="mod-11.html#cb1262-1" tabindex="-1"></a>mod1 <span class="ot">&lt;-</span> <span class="fu">nls</span>(mass<span class="sc">~</span><span class="fu">SSgompertz</span>(age, A, B, C), <span class="at">data=</span>dat)</span>
<span id="cb1262-2"><a href="mod-11.html#cb1262-2" tabindex="-1"></a>mod2 <span class="ot">&lt;-</span> <span class="fu">nls</span>(mass<span class="sc">~</span>Linf<span class="sc">*</span>(<span class="dv">1</span><span class="sc">-</span><span class="fu">exp</span>(<span class="sc">-</span>k<span class="sc">*</span>(age<span class="sc">-</span>t0))),</span>
<span id="cb1262-3"><a href="mod-11.html#cb1262-3" tabindex="-1"></a>            <span class="at">data=</span>dat,</span>
<span id="cb1262-4"><a href="mod-11.html#cb1262-4" tabindex="-1"></a>            <span class="at">start=</span><span class="fu">list</span>(<span class="at">Linf=</span><span class="fl">4e3</span>, <span class="at">k=</span><span class="fl">0.09</span>, <span class="at">t0=</span><span class="sc">-</span><span class="dv">1</span>))</span>
<span id="cb1262-5"><a href="mod-11.html#cb1262-5" tabindex="-1"></a><span class="fu">summary</span>(mod1)</span></code></pre></div>
<pre><code>## 
## Formula: mass ~ SSgompertz(age, A, B, C)
## 
## Parameters:
##    Estimate Std. Error t value Pr(&gt;|t|)    
## A 3.145e+03  1.718e+02   18.31  &lt; 2e-16 ***
## B 3.076e+00  2.858e-01   10.76 3.64e-12 ***
## C 6.514e-01  3.081e-02   21.14  &lt; 2e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 208.4 on 32 degrees of freedom
## 
## Number of iterations to convergence: 5 
## Achieved convergence tolerance: 1.862e-06</code></pre>
<div class="sourceCode" id="cb1264"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1264-1"><a href="mod-11.html#cb1264-1" tabindex="-1"></a><span class="fu">summary</span>(mod2)</span></code></pre></div>
<pre><code>## 
## Formula: mass ~ Linf * (1 - exp(-k * (age - t0)))
## 
## Parameters:
##       Estimate Std. Error t value Pr(&gt;|t|)    
## Linf 3.966e+03  3.902e+02  10.165 1.52e-11 ***
## k    1.572e-01  2.656e-02   5.917 1.38e-06 ***
## t0   2.315e-01  1.288e-01   1.797   0.0817 .  
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 212.7 on 32 degrees of freedom
## 
## Number of iterations to convergence: 7 
## Achieved convergence tolerance: 1.724e-06</code></pre>
<p>Both models were able to fit all of their parameters, although the <span class="math inline">\(t_0\)</span> term in the von Bertalanffy model was nonsignificant (i.e., not different from 0). This might indicate that we should go with the Gompertz model. The pseudo-<em>R</em><sup>2</sup> values are similar and don’t strongly suggest choosing one model over another:</p>
<div class="sourceCode" id="cb1266"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1266-1"><a href="mod-11.html#cb1266-1" tabindex="-1"></a>pseudoR2 <span class="ot">&lt;-</span> <span class="cf">function</span>(mod, Y){</span>
<span id="cb1266-2"><a href="mod-11.html#cb1266-2" tabindex="-1"></a>    ss.tot <span class="ot">&lt;-</span> <span class="fu">sum</span>((<span class="fu">fitted</span>(mod)<span class="sc">-</span><span class="fu">mean</span>(Y))<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb1266-3"><a href="mod-11.html#cb1266-3" tabindex="-1"></a>    ss.res <span class="ot">&lt;-</span> <span class="fu">sum</span>(<span class="fu">residuals</span>(mod)<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb1266-4"><a href="mod-11.html#cb1266-4" tabindex="-1"></a>    <span class="dv">1</span><span class="sc">-</span>(ss.res<span class="sc">/</span>ss.tot)</span>
<span id="cb1266-5"><a href="mod-11.html#cb1266-5" tabindex="-1"></a>}</span>
<span id="cb1266-6"><a href="mod-11.html#cb1266-6" tabindex="-1"></a><span class="fu">pseudoR2</span>(mod1, dat<span class="sc">$</span>mass)</span></code></pre></div>
<pre><code>## [1] 0.946444</code></pre>
<div class="sourceCode" id="cb1268"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1268-1"><a href="mod-11.html#cb1268-1" tabindex="-1"></a><span class="co"># [1] 0.946444</span></span>
<span id="cb1268-2"><a href="mod-11.html#cb1268-2" tabindex="-1"></a><span class="fu">pseudoR2</span>(mod2, dat<span class="sc">$</span>mass)</span></code></pre></div>
<pre><code>## [1] 0.9459103</code></pre>
<div class="sourceCode" id="cb1270"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1270-1"><a href="mod-11.html#cb1270-1" tabindex="-1"></a><span class="co"># [1] 0.9459103</span></span></code></pre></div>
<p>The AICs of the models suggest that the Gompertz model is slightly better, but not significantly so. The difference in AIC is &lt;2, suggesting that the models are equivalent in their support. This means that we would be justified choosing either one, or not choosing and presenting both.</p>
<div class="sourceCode" id="cb1271"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1271-1"><a href="mod-11.html#cb1271-1" tabindex="-1"></a><span class="fu">AIC</span>(mod1, mod2)</span></code></pre></div>
<pre><code>##      df      AIC
## mod1  4 477.9660
## mod2  4 479.3922</code></pre>
<div class="sourceCode" id="cb1273"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1273-1"><a href="mod-11.html#cb1273-1" tabindex="-1"></a><span class="co">#      df      AIC</span></span>
<span id="cb1273-2"><a href="mod-11.html#cb1273-2" tabindex="-1"></a><span class="co"># mod1  4 477.9660</span></span>
<span id="cb1273-3"><a href="mod-11.html#cb1273-3" tabindex="-1"></a><span class="co"># mod2  4 479.3922</span></span>
<span id="cb1273-4"><a href="mod-11.html#cb1273-4" tabindex="-1"></a><span class="fu">AIC</span>(mod2) <span class="sc">-</span> <span class="fu">AIC</span>(mod1)</span></code></pre></div>
<pre><code>## [1] 1.426267</code></pre>
<p>Another way to determine which model fits better is to compare their ability to predict out-of-sample data. That is, data that were not used to fit the model. The most common strategy for this is called “cross-validation”. There are many kinds of cross-validation. What they all have in common is that the model is fit to some of the data, and then its predictive accuracy is checked using the held-out data. One popular strategy is <em>k</em>-fold cross-validation, the data are divided into <em>k</em> subsets, or “folds”. Then, the model is fit <em>k</em> times: each time to all of the data except those in one fold. The predictive accuracy of these models is then calculated for the data in the fold that was left out. The better the model is at predicting out-of-sample data, the smaller the average error should be.</p>
<p>Let’s try another method called “leave-one-out” cross-validation. In this strategy, the model is fit to all observations except 1 at a time. Then, the root mean squared error (RMSE) is calculated for both models.</p>
<div class="sourceCode" id="cb1275"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1275-1"><a href="mod-11.html#cb1275-1" tabindex="-1"></a>N <span class="ot">&lt;-</span> <span class="fu">nrow</span>(dat)</span>
<span id="cb1275-2"><a href="mod-11.html#cb1275-2" tabindex="-1"></a>err1 <span class="ot">&lt;-</span> <span class="fu">numeric</span>(N)</span>
<span id="cb1275-3"><a href="mod-11.html#cb1275-3" tabindex="-1"></a>err2 <span class="ot">&lt;-</span> err1</span>
<span id="cb1275-4"><a href="mod-11.html#cb1275-4" tabindex="-1"></a></span>
<span id="cb1275-5"><a href="mod-11.html#cb1275-5" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>N){</span>
<span id="cb1275-6"><a href="mod-11.html#cb1275-6" tabindex="-1"></a>    mod1i <span class="ot">&lt;-</span> <span class="fu">nls</span>(mass<span class="sc">~</span><span class="fu">SSgompertz</span>(age, A, B, C), <span class="at">data=</span>dat[<span class="sc">-</span>i,])</span>
<span id="cb1275-7"><a href="mod-11.html#cb1275-7" tabindex="-1"></a>    mod2i <span class="ot">&lt;-</span> <span class="fu">nls</span>(mass<span class="sc">~</span>Linf<span class="sc">*</span>(<span class="dv">1</span><span class="sc">-</span><span class="fu">exp</span>(<span class="sc">-</span>k<span class="sc">*</span>(age<span class="sc">-</span>t0))),</span>
<span id="cb1275-8"><a href="mod-11.html#cb1275-8" tabindex="-1"></a>        <span class="at">data=</span>dat[<span class="sc">-</span>i,],</span>
<span id="cb1275-9"><a href="mod-11.html#cb1275-9" tabindex="-1"></a>        <span class="at">start=</span><span class="fu">list</span>(<span class="at">Linf=</span><span class="fl">4e3</span>, <span class="at">k=</span><span class="fl">0.09</span>, <span class="at">t0=</span><span class="sc">-</span><span class="dv">1</span>))</span>
<span id="cb1275-10"><a href="mod-11.html#cb1275-10" tabindex="-1"></a>    px <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">age=</span>dat<span class="sc">$</span>age[i])</span>
<span id="cb1275-11"><a href="mod-11.html#cb1275-11" tabindex="-1"></a>    err1[i] <span class="ot">&lt;-</span> <span class="fu">predict</span>(mod1i,</span>
<span id="cb1275-12"><a href="mod-11.html#cb1275-12" tabindex="-1"></a>        <span class="at">newdata=</span><span class="fu">data.frame</span>(px)) <span class="sc">-</span> dat<span class="sc">$</span>age[i]</span>
<span id="cb1275-13"><a href="mod-11.html#cb1275-13" tabindex="-1"></a>    err2[i] <span class="ot">&lt;-</span> <span class="fu">predict</span>(mod2i,</span>
<span id="cb1275-14"><a href="mod-11.html#cb1275-14" tabindex="-1"></a>        <span class="at">newdata=</span><span class="fu">data.frame</span>(px)) <span class="sc">-</span> dat<span class="sc">$</span>age[i]</span>
<span id="cb1275-15"><a href="mod-11.html#cb1275-15" tabindex="-1"></a>}</span>
<span id="cb1275-16"><a href="mod-11.html#cb1275-16" tabindex="-1"></a><span class="fu">sqrt</span>(<span class="fu">mean</span>(err1<span class="sc">^</span><span class="dv">2</span>))</span></code></pre></div>
<pre><code>## [1] 1599.034</code></pre>
<div class="sourceCode" id="cb1277"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1277-1"><a href="mod-11.html#cb1277-1" tabindex="-1"></a><span class="fu">sqrt</span>(<span class="fu">mean</span>(err2<span class="sc">^</span><span class="dv">2</span>))</span></code></pre></div>
<pre><code>## [1] 1634.066</code></pre>
<p>Again, the Gompertz curve looks slightly better than the von Bertalanffy, but not by much. For our example let’s go ahead and plot both models.</p>
<div class="sourceCode" id="cb1279"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1279-1"><a href="mod-11.html#cb1279-1" tabindex="-1"></a><span class="fu">library</span>(propagate)</span>
<span id="cb1279-2"><a href="mod-11.html#cb1279-2" tabindex="-1"></a>px <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fu">min</span>(dat<span class="sc">$</span>age), <span class="fu">max</span>(dat<span class="sc">$</span>age), <span class="at">length=</span><span class="dv">25</span>)</span>
<span id="cb1279-3"><a href="mod-11.html#cb1279-3" tabindex="-1"></a>pred <span class="ot">&lt;-</span> <span class="fu">predictNLS</span>(mod1,</span>
<span id="cb1279-4"><a href="mod-11.html#cb1279-4" tabindex="-1"></a>                   <span class="at">newdata=</span><span class="fu">data.frame</span>(<span class="at">age=</span>px),</span>
<span id="cb1279-5"><a href="mod-11.html#cb1279-5" tabindex="-1"></a>                   <span class="at">interval=</span><span class="st">&quot;confidence&quot;</span>)</span>
<span id="cb1279-6"><a href="mod-11.html#cb1279-6" tabindex="-1"></a>pred <span class="ot">&lt;-</span> pred<span class="sc">$</span>summary</span>
<span id="cb1279-7"><a href="mod-11.html#cb1279-7" tabindex="-1"></a></span>
<span id="cb1279-8"><a href="mod-11.html#cb1279-8" tabindex="-1"></a>mn1 <span class="ot">&lt;-</span> pred<span class="sc">$</span>Sim.Mean</span>
<span id="cb1279-9"><a href="mod-11.html#cb1279-9" tabindex="-1"></a>lo1 <span class="ot">&lt;-</span> pred<span class="sc">$</span><span class="st">&#39;Sim.2.5%&#39;</span></span>
<span id="cb1279-10"><a href="mod-11.html#cb1279-10" tabindex="-1"></a>up1 <span class="ot">&lt;-</span> pred<span class="sc">$</span><span class="st">&#39;Sim.97.5%&#39;</span></span>
<span id="cb1279-11"><a href="mod-11.html#cb1279-11" tabindex="-1"></a></span>
<span id="cb1279-12"><a href="mod-11.html#cb1279-12" tabindex="-1"></a>pred2 <span class="ot">&lt;-</span> <span class="fu">predictNLS</span>(mod2,</span>
<span id="cb1279-13"><a href="mod-11.html#cb1279-13" tabindex="-1"></a>                   <span class="at">newdata=</span><span class="fu">data.frame</span>(<span class="at">age=</span>px),</span>
<span id="cb1279-14"><a href="mod-11.html#cb1279-14" tabindex="-1"></a>                   <span class="at">interval=</span><span class="st">&quot;confidence&quot;</span>)</span>
<span id="cb1279-15"><a href="mod-11.html#cb1279-15" tabindex="-1"></a>pred2 <span class="ot">&lt;-</span> pred2<span class="sc">$</span>summary</span>
<span id="cb1279-16"><a href="mod-11.html#cb1279-16" tabindex="-1"></a></span>
<span id="cb1279-17"><a href="mod-11.html#cb1279-17" tabindex="-1"></a>mn2 <span class="ot">&lt;-</span> pred2<span class="sc">$</span>Sim.Mean</span>
<span id="cb1279-18"><a href="mod-11.html#cb1279-18" tabindex="-1"></a>lo2 <span class="ot">&lt;-</span> pred2<span class="sc">$</span><span class="st">&#39;Sim.2.5%&#39;</span></span>
<span id="cb1279-19"><a href="mod-11.html#cb1279-19" tabindex="-1"></a>up2 <span class="ot">&lt;-</span> pred2<span class="sc">$</span><span class="st">&#39;Sim.97.5%&#39;</span></span></code></pre></div>
<p>Now make the plot:</p>
<div class="sourceCode" id="cb1280"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1280-1"><a href="mod-11.html#cb1280-1" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>), <span class="at">mar=</span><span class="fu">c</span>(<span class="fl">5.1</span>, <span class="fl">7.1</span>, <span class="fl">1.1</span>, <span class="fl">1.1</span>),</span>
<span id="cb1280-2"><a href="mod-11.html#cb1280-2" tabindex="-1"></a>    <span class="at">las=</span><span class="dv">1</span>, <span class="at">lend=</span><span class="dv">1</span>, <span class="at">bty=</span><span class="st">&quot;n&quot;</span>, </span>
<span id="cb1280-3"><a href="mod-11.html#cb1280-3" tabindex="-1"></a>    <span class="at">cex.axis=</span><span class="fl">1.3</span>, <span class="at">cex.lab=</span><span class="fl">1.3</span>)</span>
<span id="cb1280-4"><a href="mod-11.html#cb1280-4" tabindex="-1"></a><span class="co"># make the plot</span></span>
<span id="cb1280-5"><a href="mod-11.html#cb1280-5" tabindex="-1"></a><span class="fu">plot</span>(dat<span class="sc">$</span>age, dat<span class="sc">$</span>mass,</span>
<span id="cb1280-6"><a href="mod-11.html#cb1280-6" tabindex="-1"></a>     <span class="at">pch=</span><span class="dv">16</span>, <span class="at">cex=</span><span class="fl">1.2</span>,</span>
<span id="cb1280-7"><a href="mod-11.html#cb1280-7" tabindex="-1"></a>     <span class="at">xlab=</span><span class="st">&quot;Age (years)&quot;</span>,</span>
<span id="cb1280-8"><a href="mod-11.html#cb1280-8" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">&quot;&quot;</span>, <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">3400</span>))</span>
<span id="cb1280-9"><a href="mod-11.html#cb1280-9" tabindex="-1"></a><span class="fu">title</span>(<span class="at">ylab=</span><span class="st">&quot;Mass (kg)&quot;</span>, <span class="at">line=</span><span class="dv">5</span>)</span>
<span id="cb1280-10"><a href="mod-11.html#cb1280-10" tabindex="-1"></a><span class="fu">polygon</span>(<span class="at">x=</span><span class="fu">c</span>(px, <span class="fu">rev</span>(px)), <span class="at">y=</span><span class="fu">c</span>(lo1, <span class="fu">rev</span>(up1)),</span>
<span id="cb1280-11"><a href="mod-11.html#cb1280-11" tabindex="-1"></a>    <span class="at">border=</span><span class="cn">NA</span>, <span class="at">col=</span><span class="st">&quot;#FF000030&quot;</span>)</span>
<span id="cb1280-12"><a href="mod-11.html#cb1280-12" tabindex="-1"></a><span class="fu">polygon</span>(<span class="at">x=</span><span class="fu">c</span>(px, <span class="fu">rev</span>(px)), <span class="at">y=</span><span class="fu">c</span>(lo2, <span class="fu">rev</span>(up2)),</span>
<span id="cb1280-13"><a href="mod-11.html#cb1280-13" tabindex="-1"></a>    <span class="at">border=</span><span class="cn">NA</span>, <span class="at">col=</span><span class="st">&quot;#0000FF30&quot;</span>)</span>
<span id="cb1280-14"><a href="mod-11.html#cb1280-14" tabindex="-1"></a><span class="fu">points</span>(px, mn1, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>)</span>
<span id="cb1280-15"><a href="mod-11.html#cb1280-15" tabindex="-1"></a><span class="fu">points</span>(px, mn2, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">col=</span><span class="st">&quot;blue&quot;</span>)</span>
<span id="cb1280-16"><a href="mod-11.html#cb1280-16" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;bottomright&quot;</span>,</span>
<span id="cb1280-17"><a href="mod-11.html#cb1280-17" tabindex="-1"></a>    <span class="at">legend=</span><span class="fu">c</span>(<span class="st">&quot;Gompertz&quot;</span>, <span class="st">&quot;von Bertalanffy&quot;</span>),</span>
<span id="cb1280-18"><a href="mod-11.html#cb1280-18" tabindex="-1"></a>    <span class="at">lwd=</span><span class="dv">4</span>, <span class="at">col=</span><span class="fu">c</span>(<span class="st">&quot;red&quot;</span>, <span class="st">&quot;blue&quot;</span>),</span>
<span id="cb1280-19"><a href="mod-11.html#cb1280-19" tabindex="-1"></a>    <span class="at">bty=</span><span class="st">&quot;n&quot;</span>, <span class="at">cex=</span><span class="fl">1.3</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-741-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>The plot reveals a problem with the von Bertalanffy: it predicts negative masses! This is probably because the model had a non-significant <span class="math inline">\(t_0\)</span> with a point estimate that was non-negative. The <span class="math inline">\(t_0\)</span> term is the age when the response variable is 0, and must be positive to ensure that only positive sizes are predicted.</p>
<p>It might be worth asking how fast the animals grew, and how that growth rate changed over the course of their lives. Technically, the growth rate would be the first derivative of the fitted function. However, getting a first derivative might be difficult or impossible for many NLS models. Instead, we can calculate the values empirically. Over a period of 12 years (the domain of our fitted function), daily time steps offer a reasonable approximation of the instantaneous rate of change (daily time steps are more biologically interpretable than a literally instantaneous rate, as well).</p>
<div class="sourceCode" id="cb1281"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1281-1"><a href="mod-11.html#cb1281-1" tabindex="-1"></a><span class="co"># daily x values as fractional years</span></span>
<span id="cb1281-2"><a href="mod-11.html#cb1281-2" tabindex="-1"></a>xday <span class="ot">&lt;-</span> <span class="dv">0</span><span class="sc">:</span>(<span class="dv">12</span><span class="sc">*</span><span class="fl">365.25</span>) <span class="co"># Julian year length</span></span>
<span id="cb1281-3"><a href="mod-11.html#cb1281-3" tabindex="-1"></a>xday <span class="ot">&lt;-</span> xday <span class="sc">/</span> <span class="fl">365.25</span></span>
<span id="cb1281-4"><a href="mod-11.html#cb1281-4" tabindex="-1"></a></span>
<span id="cb1281-5"><a href="mod-11.html#cb1281-5" tabindex="-1"></a><span class="co"># predicted daily mass</span></span>
<span id="cb1281-6"><a href="mod-11.html#cb1281-6" tabindex="-1"></a>yday <span class="ot">&lt;-</span> <span class="fu">predict</span>(mod1, <span class="at">newdata=</span><span class="fu">data.frame</span>(<span class="at">age=</span>xday))</span>
<span id="cb1281-7"><a href="mod-11.html#cb1281-7" tabindex="-1"></a></span>
<span id="cb1281-8"><a href="mod-11.html#cb1281-8" tabindex="-1"></a>ydiff <span class="ot">&lt;-</span> <span class="fu">diff</span>(yday)</span>
<span id="cb1281-9"><a href="mod-11.html#cb1281-9" tabindex="-1"></a>xdiff <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(ydiff)<span class="sc">/</span><span class="fl">365.25</span></span>
<span id="cb1281-10"><a href="mod-11.html#cb1281-10" tabindex="-1"></a></span>
<span id="cb1281-11"><a href="mod-11.html#cb1281-11" tabindex="-1"></a><span class="fu">plot</span>(xdiff, ydiff, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>,</span>
<span id="cb1281-12"><a href="mod-11.html#cb1281-12" tabindex="-1"></a>     <span class="at">xlab=</span><span class="st">&quot;Age (years)&quot;</span>,</span>
<span id="cb1281-13"><a href="mod-11.html#cb1281-13" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">&quot;Daily mass gain (kg)&quot;</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-742-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>The figure shows that daily growth peaked during the 3rd year of life (age 2 to age 3), when animals gained an average of 1.5 kg per day. Interestingly, this the age at which the authors of the original study hypothesized the animals reached sexual maturity <span class="citation">(<a href="#ref-woodward2015">Woodward et al. 2015</a>)</span>.</p>
</div>
</div>
<div id="dose-response-curves" class="section level2 hasAnchor" number="11.5">
<h2><span class="header-section-number">11.5</span> Dose response curves<a href="mod-11.html#dose-response-curves" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>A <strong>dose response curve</strong> describes the change in a biological system following a quantitative perturbation. Most dose response models are nonlinear because of the dynamics of the systems they describe. Mechanistically, there is often a rate that scales with the current concentration of a compound, or some other aspect of the system. In other words, the dynamics of the system are described by a kind of <strong>differential equation</strong>. The state of the system is thus described by solutions to those equations.</p>
<p>There are many dose-response models out there. We will fit two common ones using the R package <code>drc</code> <span class="citation">(<a href="#ref-ritz2015">Ritz et al. 2015</a>)</span>. Many dose response models could also be fit by <code>nls()</code>, but we will use the <code>drc</code> package because of its convenience and so that we can learn an alternative approach.</p>
<p>We will use an example from the <a href="https://doi.org/10.1371/journal.pone.0146021.s001">Supplemental Information</a> provided by the authors of <code>drc</code> <span class="citation">(<a href="#ref-ritz2015">Ritz et al. 2015</a>)</span>. <span class="citation">Inderjit et al. (<a href="#ref-inderjit2002">2002</a>)</span> studied the effect of an herbicide mixture on perennial ryegrass (<em>Lolium perenne</em>)<a href="#fn87" class="footnote-ref" id="fnref87"><sup>87</sup></a>. Rye grass seeds were placed in Petri dishes and treated with a solution of ferulic acid in concentrations ranging from 0 to 30 mM. The seeds were allowed to germinate and develop for 8 days. After 8 days the root length of each individual seed was measured.</p>
<p><img src="11_04.jpg" width="50%" style="display: block; margin: auto;" /></p>
<p>Load the <code>drc</code> package and access the dataset <code>ryegrass</code>. Then, plot root length against ferulic acid concentration.</p>
<div class="sourceCode" id="cb1282"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1282-1"><a href="mod-11.html#cb1282-1" tabindex="-1"></a><span class="fu">library</span>(drc)</span></code></pre></div>
<pre><code>## Warning: package &#39;drc&#39; was built under R version 4.3.3</code></pre>
<div class="sourceCode" id="cb1284"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1284-1"><a href="mod-11.html#cb1284-1" tabindex="-1"></a>dat <span class="ot">&lt;-</span> ryegrass</span>
<span id="cb1284-2"><a href="mod-11.html#cb1284-2" tabindex="-1"></a><span class="fu">plot</span>(dat<span class="sc">$</span>conc, dat<span class="sc">$</span>rootl)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-744-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>The plot suggests that increasing ferulic acid concentration decreases mean root length. The relationship also appears nonlinear: root length decreases very quickly at low ferulic acid concentration, and then levels out as concentration increases beyond about 5 mM.</p>
<p>We will try two nonlinear models found in the <code>drc</code> package. The first is the <strong>log-logistic model</strong>. The log-logistic model comes in several flavors, defined by the number of parameters used to specify the curve. Two common varieties are shown below. By convention, one of the parameters is labeled “<em>e</em>”, but it is not the same as Euler’s constant <em>e</em> (<span class="math inline">\(\approx\)</span> 2.71828). The parameter labeled <em>e</em> shows up only in the exponent as <span class="math inline">\(log(e)\)</span>, but this does not mean <span class="math inline">\(log(exp(1))\)</span> (which would be 1, by definition). The bases of the exponents in the models below are Euler’s <em>e</em>. For clarity, the parameter <em>e</em> is sometimes labeled with another letter, or the exponential function <span class="math inline">\(e^x\)</span> is written as <span class="math inline">\(exp(x)\)</span>.</p>
<p><strong>Three-parameter log-logistic:</strong></p>
<p><span class="math display">\[f\left(x,\left(b,d,e\right)\right)=\frac{d}{1+e^{b\left(log{\left(x\right)}-log{\left(e\right)}\right)}}=\frac{d}{1+exp\left(log{\left(x\right)}-log{\left(e\right)}\right)}\]</span></p>
<p><strong>Four-parameter log-logistic:</strong></p>
<p><span class="math display">\[f\left(x,\left(b,c,d,e\right)\right)=\frac{d-c}{1+e^{b\left(log{\left(x\right)}-log{\left(e\right)}\right)}}=\frac{d-c}{1+exp\left(b\left(log{\left(x\right)}-log{\left(e\right)}\right)\right)}\]</span></p>
<p>Just for fun we can also try a <strong>Weibull type I</strong> function:
<span class="math display">\[\begin{matrix}f\left(x,\ \left(b,c,d,e\right)\right)=&amp;c+\left(d-c\right)e^{-e^{b\left(log{\left(x\right)}-log{\left(e\right)}\right)}}\\f\left(x,\ \left(b,c,d,e\right)\right)=&amp;c+\left(d-c\right)exp\left(-exp\left(b\left(log{\left(x\right)}-log{\left(e\right)}\right)\right)\right)\\\end{matrix}\]</span></p>
<p>Fitting the models in <code>drc</code> is straightforward. First define the response and predictor variables in a formula (as in <code>lm()</code> and other functions), specify the dataset to be used, and finally the curve (<code>fct</code>) to be fitted. The full list of functions that <code>drm()</code> can fit can be found with the command <code>getMeanFunctions()</code>.</p>
<div class="sourceCode" id="cb1285"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1285-1"><a href="mod-11.html#cb1285-1" tabindex="-1"></a>mod1 <span class="ot">&lt;-</span> <span class="fu">drm</span>(rootl<span class="sc">~</span>conc, <span class="at">data=</span>ryegrass, <span class="at">fct=</span><span class="fu">LL.3</span>())</span>
<span id="cb1285-2"><a href="mod-11.html#cb1285-2" tabindex="-1"></a>mod2 <span class="ot">&lt;-</span> <span class="fu">drm</span>(rootl<span class="sc">~</span>conc, <span class="at">data=</span>ryegrass, <span class="at">fct=</span><span class="fu">LL.4</span>())</span>
<span id="cb1285-3"><a href="mod-11.html#cb1285-3" tabindex="-1"></a>mod3 <span class="ot">&lt;-</span> <span class="fu">drm</span>(rootl<span class="sc">~</span>conc, <span class="at">data=</span>ryegrass, <span class="at">fct=</span><span class="fu">weibull1</span>())</span>
<span id="cb1285-4"><a href="mod-11.html#cb1285-4" tabindex="-1"></a></span>
<span id="cb1285-5"><a href="mod-11.html#cb1285-5" tabindex="-1"></a><span class="fu">summary</span>(mod1)</span></code></pre></div>
<pre><code>## 
## Model fitted: Log-logistic (ED50 as parameter) with lower limit at 0 (3 parms)
## 
## Parameter estimates:
## 
##               Estimate Std. Error t-value   p-value    
## b:(Intercept)  2.47033    0.34168  7.2299 4.011e-07 ***
## d:(Intercept)  7.85543    0.20438 38.4352 &lt; 2.2e-16 ***
## e:(Intercept)  3.26336    0.19641 16.6154 1.474e-13 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error:
## 
##  0.5615802 (21 degrees of freedom)</code></pre>
<div class="sourceCode" id="cb1287"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1287-1"><a href="mod-11.html#cb1287-1" tabindex="-1"></a><span class="fu">summary</span>(mod2)</span></code></pre></div>
<pre><code>## 
## Model fitted: Log-logistic (ED50 as parameter) (4 parms)
## 
## Parameter estimates:
## 
##               Estimate Std. Error t-value   p-value    
## b:(Intercept)  2.98222    0.46506  6.4125 2.960e-06 ***
## c:(Intercept)  0.48141    0.21219  2.2688   0.03451 *  
## d:(Intercept)  7.79296    0.18857 41.3272 &lt; 2.2e-16 ***
## e:(Intercept)  3.05795    0.18573 16.4644 4.268e-13 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error:
## 
##  0.5196256 (20 degrees of freedom)</code></pre>
<div class="sourceCode" id="cb1289"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1289-1"><a href="mod-11.html#cb1289-1" tabindex="-1"></a><span class="fu">summary</span>(mod3)</span></code></pre></div>
<pre><code>## 
## Model fitted: Weibull (type 1) (4 parms)
## 
## Parameter estimates:
## 
##               Estimate Std. Error t-value   p-value    
## b:(Intercept)  2.39341    0.47832  5.0038 6.813e-05 ***
## c:(Intercept)  0.66045    0.18857  3.5023  0.002243 ** 
## d:(Intercept)  7.80586    0.20852 37.4348 &lt; 2.2e-16 ***
## e:(Intercept)  3.60013    0.20311 17.7250 1.068e-13 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error:
## 
##  0.5488238 (20 degrees of freedom)</code></pre>
<p>All three models fit, had statistically-significant parameters, and had similar residual standard errors (RSE; a measure of model fit similar to RMSE). Model 2 (4 parameter log-logistic) had the smallest RSE and the smallest AIC:</p>
<div class="sourceCode" id="cb1291"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1291-1"><a href="mod-11.html#cb1291-1" tabindex="-1"></a><span class="fu">AIC</span>(mod1, mod2, mod3)</span></code></pre></div>
<pre><code>##      df      AIC
## mod1  4 45.20827
## mod2  5 42.31029
## mod3  5 44.93439</code></pre>
<p>This is moderately convincing evidence that model 2 is the best-fitting of the three models that we fit. Let’s calculate predictions and 95% CI for a range of concentrations and present them in a graph. Unlike <code>nls()</code> objects, models fitted using <code>drm()</code> have a predict method that can generate confidence intervals. The code below may return warnings, but these messages can probably be ignored.</p>
<div class="sourceCode" id="cb1293"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1293-1"><a href="mod-11.html#cb1293-1" tabindex="-1"></a>L <span class="ot">&lt;-</span> <span class="dv">50</span></span>
<span id="cb1293-2"><a href="mod-11.html#cb1293-2" tabindex="-1"></a>px <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fu">min</span>(dat<span class="sc">$</span>conc), <span class="fu">max</span>(dat<span class="sc">$</span>conc), <span class="at">length=</span>L)</span>
<span id="cb1293-3"><a href="mod-11.html#cb1293-3" tabindex="-1"></a>pred <span class="ot">&lt;-</span> <span class="fu">predict</span>(mod2,</span>
<span id="cb1293-4"><a href="mod-11.html#cb1293-4" tabindex="-1"></a>                <span class="at">newdata=</span><span class="fu">data.frame</span>(<span class="at">conc=</span>px),</span>
<span id="cb1293-5"><a href="mod-11.html#cb1293-5" tabindex="-1"></a>                <span class="at">interval=</span><span class="st">&quot;confidence&quot;</span>)</span>
<span id="cb1293-6"><a href="mod-11.html#cb1293-6" tabindex="-1"></a>mn <span class="ot">&lt;-</span> pred[,<span class="dv">1</span>]</span>
<span id="cb1293-7"><a href="mod-11.html#cb1293-7" tabindex="-1"></a>lo <span class="ot">&lt;-</span> pred[,<span class="dv">2</span>]</span>
<span id="cb1293-8"><a href="mod-11.html#cb1293-8" tabindex="-1"></a>up <span class="ot">&lt;-</span> pred[,<span class="dv">3</span>]</span>
<span id="cb1293-9"><a href="mod-11.html#cb1293-9" tabindex="-1"></a></span>
<span id="cb1293-10"><a href="mod-11.html#cb1293-10" tabindex="-1"></a><span class="co"># fancy plot options</span></span>
<span id="cb1293-11"><a href="mod-11.html#cb1293-11" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>), <span class="at">mar=</span><span class="fu">c</span>(<span class="fl">5.1</span>, <span class="fl">5.1</span>, <span class="fl">1.1</span>, <span class="fl">1.1</span>),</span>
<span id="cb1293-12"><a href="mod-11.html#cb1293-12" tabindex="-1"></a>    <span class="at">las=</span><span class="dv">1</span>, <span class="at">lend=</span><span class="dv">1</span>, <span class="at">bty=</span><span class="st">&quot;n&quot;</span>, </span>
<span id="cb1293-13"><a href="mod-11.html#cb1293-13" tabindex="-1"></a>    <span class="at">cex.axis=</span><span class="fl">1.3</span>, <span class="at">cex.lab=</span><span class="fl">1.3</span>)</span>
<span id="cb1293-14"><a href="mod-11.html#cb1293-14" tabindex="-1"></a></span>
<span id="cb1293-15"><a href="mod-11.html#cb1293-15" tabindex="-1"></a><span class="co"># make the plot</span></span>
<span id="cb1293-16"><a href="mod-11.html#cb1293-16" tabindex="-1"></a><span class="fu">plot</span>(dat<span class="sc">$</span>conc, dat<span class="sc">$</span>rootl,</span>
<span id="cb1293-17"><a href="mod-11.html#cb1293-17" tabindex="-1"></a>     <span class="at">pch=</span><span class="dv">16</span>, <span class="at">cex=</span><span class="fl">1.2</span>,</span>
<span id="cb1293-18"><a href="mod-11.html#cb1293-18" tabindex="-1"></a>     <span class="at">xlab=</span><span class="st">&quot;Ferulic acid (mM)&quot;</span>,</span>
<span id="cb1293-19"><a href="mod-11.html#cb1293-19" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">&quot;Root length (cm)&quot;</span>)</span>
<span id="cb1293-20"><a href="mod-11.html#cb1293-20" tabindex="-1"></a><span class="fu">polygon</span>(<span class="fu">c</span>(px, <span class="fu">rev</span>(px)), <span class="fu">c</span>(lo, <span class="fu">rev</span>(up)),</span>
<span id="cb1293-21"><a href="mod-11.html#cb1293-21" tabindex="-1"></a>        <span class="at">border=</span><span class="cn">NA</span>, <span class="at">col=</span><span class="st">&quot;lightgrey&quot;</span>)</span>
<span id="cb1293-22"><a href="mod-11.html#cb1293-22" tabindex="-1"></a><span class="fu">points</span>(px, mn, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>)</span>
<span id="cb1293-23"><a href="mod-11.html#cb1293-23" tabindex="-1"></a><span class="fu">points</span>(dat<span class="sc">$</span>conc, dat<span class="sc">$</span>rootl, <span class="at">pch=</span><span class="dv">16</span>, <span class="at">cex=</span><span class="fl">1.2</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/chunkx720-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>Regardless of whether or not you get warning messages, you should carefully inspect the data, the residuals, and the model predictions to be sure that the estimates are biologically reasonable. The figure above shows no obvious red flags, but your data might not be so well-behaved.</p>
</div>
<div id="alternatives-to-nls" class="section level2 hasAnchor" number="11.6">
<h2><span class="header-section-number">11.6</span> Alternatives to NLS<a href="mod-11.html#alternatives-to-nls" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>If your goal is to model some response variable that varies nonlinearly with predictor variables, there are several good alternatives to NLS. The examples below introduce generalized nonlinear models (GNM), quantile regression, generalized additive models (GAM), and classification and regression trees (CART). This page briefly describes and demonstrates each method. More in-depth tutorials might be developed in the future.</p>
<div id="generalized-nonlinear-models" class="section level3 hasAnchor" number="11.6.1">
<h3><span class="header-section-number">11.6.1</span> Generalized nonlinear models<a href="mod-11.html#generalized-nonlinear-models" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Just as generalized linear models (GLM) extend the linear model by allowing for non-normal response distributions, the generalized nonlinear model (GNM) extends nonlinear models with non-normal response distributions. GNMs are available in the R package gnm <span class="citation">(<a href="#ref-turner2020">Turner and Firth 2020</a>)</span>. This package is fairly new and tricky to fit models with. I prefer to fit GNM using MCMC in JAGS.</p>
</div>
<div id="quantile-regression" class="section level3 hasAnchor" number="11.6.2">
<h3><span class="header-section-number">11.6.2</span> Quantile regression<a href="mod-11.html#quantile-regression" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>Quantile regression</strong> is a technique to model specific quantiles of a response variable as a function of predictor variables. In linear regression, GLM, NLS, and GNM, the fitted model predicts the mean value of the response variable. With quantile regression, you can model other quantiles of the distribution. Quantile regression for the 0.5 quantile (50th percentile) is very similar to regression on the mean. Regression for other quantiles can allow you to describe relationships where the mean and predictor variables are not linearly related. A good reference for quantile regression is <span class="citation">Cade and Noon (<a href="#ref-cade2003">2003</a>)</span>. Quantile regression models are available in the R package <code>quantreg</code> <span class="citation">(<a href="#ref-koenker2021">Koenker 2021</a>)</span>. Like least-squares models, quantile regression can be used to model both linear and nonlinear relationships.</p>
<div id="linear-quantile-regression-example" class="section level4 hasAnchor" number="11.6.2.1">
<h4><span class="header-section-number">11.6.2.1</span> Linear quantile regression example<a href="mod-11.html#linear-quantile-regression-example" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>For this example and the nonlinear quantile regression example we will use the “Mammals” dataset found in package <code>quantreg</code>. This dataset contains the body weights and maximum running speeds of 107 species of mammals.</p>
<div class="sourceCode" id="cb1294"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1294-1"><a href="mod-11.html#cb1294-1" tabindex="-1"></a><span class="fu">library</span>(quantreg)</span></code></pre></div>
<pre><code>## Warning: package &#39;quantreg&#39; was built under R version 4.3.3</code></pre>
<pre><code>## Loading required package: SparseM</code></pre>
<pre><code>## Warning: package &#39;SparseM&#39; was built under R version 4.3.3</code></pre>
<pre><code>## 
## Attaching package: &#39;SparseM&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:Matrix&#39;:
## 
##     det</code></pre>
<div class="sourceCode" id="cb1300"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1300-1"><a href="mod-11.html#cb1300-1" tabindex="-1"></a><span class="fu">data</span>(Mammals)</span>
<span id="cb1300-2"><a href="mod-11.html#cb1300-2" tabindex="-1"></a>dat <span class="ot">&lt;-</span> Mammals</span>
<span id="cb1300-3"><a href="mod-11.html#cb1300-3" tabindex="-1"></a>dat<span class="sc">$</span>x <span class="ot">&lt;-</span> <span class="fu">log</span>(dat<span class="sc">$</span>weight)</span>
<span id="cb1300-4"><a href="mod-11.html#cb1300-4" tabindex="-1"></a>dat<span class="sc">$</span>y <span class="ot">&lt;-</span> <span class="fu">log</span>(dat<span class="sc">$</span>speed)</span>
<span id="cb1300-5"><a href="mod-11.html#cb1300-5" tabindex="-1"></a><span class="fu">plot</span>(dat<span class="sc">$</span>x, dat<span class="sc">$</span>y)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-748-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>As you might expect, larger animals tend to run faster than smaller animals. The relationship might be linear or nonlinear, so we will try both linear and nonlinear quantile regression models. First we will define a series of quantiles in [0, 1] at which to fit models. In my experience, quantiles <span class="math inline">\(\le\)</span> 0.05 and <span class="math inline">\(\ge\)</span> 0.95 tend to have numerical problems, especially for small datasets. By convention, the quantiles in <code>quantreg</code> are identified by the Greek letter <span class="math inline">\(\tau\)</span> (“tau”). The fastest way to fit quantile regressions is in a <code>for()</code> loop.</p>
<div class="sourceCode" id="cb1301"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1301-1"><a href="mod-11.html#cb1301-1" tabindex="-1"></a>taus <span class="ot">&lt;-</span> <span class="dv">5</span><span class="sc">:</span><span class="dv">95</span><span class="sc">/</span><span class="dv">100</span></span>
<span id="cb1301-2"><a href="mod-11.html#cb1301-2" tabindex="-1"></a>ntau <span class="ot">&lt;-</span> <span class="fu">length</span>(taus)</span>
<span id="cb1301-3"><a href="mod-11.html#cb1301-3" tabindex="-1"></a>rq.list <span class="ot">&lt;-</span> <span class="fu">vector</span>(<span class="st">&quot;list&quot;</span>, ntau)</span>
<span id="cb1301-4"><a href="mod-11.html#cb1301-4" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>ntau){</span>
<span id="cb1301-5"><a href="mod-11.html#cb1301-5" tabindex="-1"></a>    rq.list[[i]] <span class="ot">&lt;-</span> <span class="fu">rq</span>(y<span class="sc">~</span>x, <span class="at">data=</span>dat, <span class="at">tau=</span>taus[i])</span>
<span id="cb1301-6"><a href="mod-11.html#cb1301-6" tabindex="-1"></a>}</span></code></pre></div>
<p>The list <code>rq.list</code> contains a fitted model for each quantile <span class="math inline">\(\tau\)</span> in [0.05, 0.95]. We will extract some results from these model objects to present as our model outputs.</p>
<div class="sourceCode" id="cb1302"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1302-1"><a href="mod-11.html#cb1302-1" tabindex="-1"></a><span class="co"># data frame of parameter estimates and CI</span></span>
<span id="cb1302-2"><a href="mod-11.html#cb1302-2" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">tau=</span>taus)</span>
<span id="cb1302-3"><a href="mod-11.html#cb1302-3" tabindex="-1"></a>res<span class="sc">$</span>b0.mn <span class="ot">&lt;-</span> <span class="fu">sapply</span>(rq.list, <span class="cf">function</span>(x){<span class="fu">summary</span>(x)<span class="sc">$</span>coefficients[<span class="dv">1</span>,<span class="dv">1</span>]})</span>
<span id="cb1302-4"><a href="mod-11.html#cb1302-4" tabindex="-1"></a>res<span class="sc">$</span>b0.lo <span class="ot">&lt;-</span> <span class="fu">sapply</span>(rq.list, <span class="cf">function</span>(x){<span class="fu">summary</span>(x)<span class="sc">$</span>coefficients[<span class="dv">1</span>,<span class="dv">2</span>]})</span>
<span id="cb1302-5"><a href="mod-11.html#cb1302-5" tabindex="-1"></a>res<span class="sc">$</span>b0.up <span class="ot">&lt;-</span> <span class="fu">sapply</span>(rq.list, <span class="cf">function</span>(x){<span class="fu">summary</span>(x)<span class="sc">$</span>coefficients[<span class="dv">1</span>,<span class="dv">3</span>]})</span>
<span id="cb1302-6"><a href="mod-11.html#cb1302-6" tabindex="-1"></a>res<span class="sc">$</span>b1.mn <span class="ot">&lt;-</span> <span class="fu">sapply</span>(rq.list, <span class="cf">function</span>(x){<span class="fu">summary</span>(x)<span class="sc">$</span>coefficients[<span class="dv">2</span>,<span class="dv">1</span>]})</span>
<span id="cb1302-7"><a href="mod-11.html#cb1302-7" tabindex="-1"></a>res<span class="sc">$</span>b1.lo <span class="ot">&lt;-</span> <span class="fu">sapply</span>(rq.list, <span class="cf">function</span>(x){<span class="fu">summary</span>(x)<span class="sc">$</span>coefficients[<span class="dv">2</span>,<span class="dv">2</span>]})</span>
<span id="cb1302-8"><a href="mod-11.html#cb1302-8" tabindex="-1"></a>res<span class="sc">$</span>b1.up <span class="ot">&lt;-</span> <span class="fu">sapply</span>(rq.list, <span class="cf">function</span>(x){<span class="fu">summary</span>(x)<span class="sc">$</span>coefficients[<span class="dv">2</span>,<span class="dv">3</span>]})</span></code></pre></div>
<p>Rather than present model predictions for all 90 quantiles, let’s just present a subset. Here we present the predicted 0.1, 0.3, 0.5, 0.7, and 0.9 quantiles of running speed (i.e., 10th, 30th, 50th, 70th, and 90th percentiles).</p>
<div class="sourceCode" id="cb1303"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1303-1"><a href="mod-11.html#cb1303-1" tabindex="-1"></a><span class="co"># predictions for selected quantiles</span></span>
<span id="cb1303-2"><a href="mod-11.html#cb1303-2" tabindex="-1"></a>ptau <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">5</span>, <span class="dv">7</span>, <span class="dv">9</span>)<span class="sc">/</span><span class="dv">10</span></span>
<span id="cb1303-3"><a href="mod-11.html#cb1303-3" tabindex="-1"></a>px <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fu">min</span>(dat<span class="sc">$</span>x), <span class="fu">max</span>(dat<span class="sc">$</span>x), <span class="at">length=</span><span class="dv">50</span>)</span>
<span id="cb1303-4"><a href="mod-11.html#cb1303-4" tabindex="-1"></a>pmat <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>, <span class="at">nrow=</span><span class="fu">length</span>(px), <span class="at">ncol=</span><span class="fu">length</span>(ptau))</span>
<span id="cb1303-5"><a href="mod-11.html#cb1303-5" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(ptau)){</span>
<span id="cb1303-6"><a href="mod-11.html#cb1303-6" tabindex="-1"></a>    curr.mod <span class="ot">&lt;-</span> rq.list[[<span class="fu">which</span>(taus <span class="sc">==</span> ptau[i])]]</span>
<span id="cb1303-7"><a href="mod-11.html#cb1303-7" tabindex="-1"></a>    pmat[,i] <span class="ot">&lt;-</span> <span class="fu">predict</span>(curr.mod, <span class="at">newdata=</span><span class="fu">data.frame</span>(<span class="at">x=</span>px))</span>
<span id="cb1303-8"><a href="mod-11.html#cb1303-8" tabindex="-1"></a>}</span></code></pre></div>
<p>Finally, we can produce a figure that shows (1) the data and predicted quantiles; (2) the estimate and 95% CI of the model intercept for each <span class="math inline">\(\tau\)</span>; and (3) the estimate and 95% CI of the model slope for each <span class="math inline">\(\tau\)</span>. The code for the figure has a lot of components, but each of them should be something familiar by now.</p>
<div class="sourceCode" id="cb1304"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1304-1"><a href="mod-11.html#cb1304-1" tabindex="-1"></a><span class="co"># plot data and parameter estimates vs. tau</span></span>
<span id="cb1304-2"><a href="mod-11.html#cb1304-2" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>), <span class="at">bty=</span><span class="st">&quot;n&quot;</span>, <span class="at">mar=</span><span class="fu">c</span>(<span class="fl">5.1</span>, <span class="fl">5.1</span>, <span class="fl">1.1</span>, <span class="fl">1.1</span>),</span>
<span id="cb1304-3"><a href="mod-11.html#cb1304-3" tabindex="-1"></a>    <span class="at">lend=</span><span class="dv">1</span>, <span class="at">las=</span><span class="dv">1</span>, <span class="at">cex.axis=</span><span class="fl">1.3</span>, <span class="at">cex.lab=</span><span class="fl">1.3</span>)</span>
<span id="cb1304-4"><a href="mod-11.html#cb1304-4" tabindex="-1"></a><span class="fu">plot</span>(dat<span class="sc">$</span>x, dat<span class="sc">$</span>y, <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">5.5</span>),</span>
<span id="cb1304-5"><a href="mod-11.html#cb1304-5" tabindex="-1"></a>    <span class="at">xlab=</span><span class="st">&quot;Body mass (kg)&quot;</span>,</span>
<span id="cb1304-6"><a href="mod-11.html#cb1304-6" tabindex="-1"></a>    <span class="at">ylab=</span><span class="st">&quot;Running speed (km/h)&quot;</span>,</span>
<span id="cb1304-7"><a href="mod-11.html#cb1304-7" tabindex="-1"></a>    <span class="at">xaxt=</span><span class="st">&quot;n&quot;</span>, <span class="at">yaxt=</span><span class="st">&quot;n&quot;</span>)</span>
<span id="cb1304-8"><a href="mod-11.html#cb1304-8" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(ptau)){</span>
<span id="cb1304-9"><a href="mod-11.html#cb1304-9" tabindex="-1"></a>    <span class="fu">points</span>(px, pmat[,i], <span class="at">type=</span><span class="st">&quot;l&quot;</span>)</span>
<span id="cb1304-10"><a href="mod-11.html#cb1304-10" tabindex="-1"></a>    <span class="fu">text</span>(px[<span class="fu">length</span>(px)], pmat[<span class="fu">nrow</span>(pmat),i], ptau[i], <span class="at">adj=</span><span class="dv">0</span>)</span>
<span id="cb1304-11"><a href="mod-11.html#cb1304-11" tabindex="-1"></a>}</span>
<span id="cb1304-12"><a href="mod-11.html#cb1304-12" tabindex="-1"></a>xax <span class="ot">&lt;-</span> <span class="dv">10</span><span class="sc">^</span><span class="fu">seq</span>(<span class="sc">-</span><span class="dv">2</span>, <span class="dv">4</span>, <span class="at">by=</span><span class="dv">1</span>)</span>
<span id="cb1304-13"><a href="mod-11.html#cb1304-13" tabindex="-1"></a><span class="fu">axis</span>(<span class="at">side=</span><span class="dv">1</span>, <span class="at">at=</span><span class="fu">log</span>(xax), <span class="at">labels=</span>xax)</span>
<span id="cb1304-14"><a href="mod-11.html#cb1304-14" tabindex="-1"></a>yax1 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">10</span>, <span class="dv">100</span>)</span>
<span id="cb1304-15"><a href="mod-11.html#cb1304-15" tabindex="-1"></a>yax2 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">9</span>, <span class="dv">2</span><span class="sc">:</span><span class="dv">9</span><span class="sc">*</span><span class="dv">10</span>, <span class="dv">110</span>)</span>
<span id="cb1304-16"><a href="mod-11.html#cb1304-16" tabindex="-1"></a><span class="fu">axis</span>(<span class="at">side=</span><span class="dv">2</span>, <span class="at">at=</span><span class="fu">log</span>(yax1), <span class="at">labels=</span>yax1)</span>
<span id="cb1304-17"><a href="mod-11.html#cb1304-17" tabindex="-1"></a><span class="fu">axis</span>(<span class="at">side=</span><span class="dv">2</span>, <span class="at">at=</span><span class="fu">log</span>(yax2), <span class="at">labels=</span><span class="cn">NA</span>, <span class="at">tcl=</span><span class="sc">-</span><span class="fl">0.3</span>)</span>
<span id="cb1304-18"><a href="mod-11.html#cb1304-18" tabindex="-1"></a></span>
<span id="cb1304-19"><a href="mod-11.html#cb1304-19" tabindex="-1"></a><span class="fu">plot</span>(taus, res<span class="sc">$</span>b0.mn, <span class="at">type=</span><span class="st">&quot;s&quot;</span>,</span>
<span id="cb1304-20"><a href="mod-11.html#cb1304-20" tabindex="-1"></a>    <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">ylim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="dv">5</span>),</span>
<span id="cb1304-21"><a href="mod-11.html#cb1304-21" tabindex="-1"></a>    <span class="at">xlab=</span><span class="fu">expression</span>(Quantile<span class="sc">~</span>(tau)),</span>
<span id="cb1304-22"><a href="mod-11.html#cb1304-22" tabindex="-1"></a>    <span class="at">ylab=</span><span class="fu">expression</span>(Model<span class="sc">~</span>intercept<span class="sc">~</span>(beta[<span class="dv">0</span>])))</span>
<span id="cb1304-23"><a href="mod-11.html#cb1304-23" tabindex="-1"></a><span class="fu">points</span>(taus, res<span class="sc">$</span>b0.lo, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">lwd=</span><span class="dv">2</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1304-24"><a href="mod-11.html#cb1304-24" tabindex="-1"></a><span class="fu">points</span>(taus, res<span class="sc">$</span>b0.up, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">lwd=</span><span class="dv">2</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1304-25"><a href="mod-11.html#cb1304-25" tabindex="-1"></a><span class="fu">segments</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="at">lty=</span><span class="dv">2</span>, <span class="at">lwd=</span><span class="dv">1</span>)</span>
<span id="cb1304-26"><a href="mod-11.html#cb1304-26" tabindex="-1"></a></span>
<span id="cb1304-27"><a href="mod-11.html#cb1304-27" tabindex="-1"></a><span class="fu">plot</span>(taus, res<span class="sc">$</span>b1.mn, <span class="at">type=</span><span class="st">&quot;s&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>,</span>
<span id="cb1304-28"><a href="mod-11.html#cb1304-28" tabindex="-1"></a>    <span class="at">ylim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.5</span>, <span class="fl">0.5</span>),</span>
<span id="cb1304-29"><a href="mod-11.html#cb1304-29" tabindex="-1"></a>    <span class="at">xlab=</span><span class="fu">expression</span>(Quantile<span class="sc">~</span>(tau)),</span>
<span id="cb1304-30"><a href="mod-11.html#cb1304-30" tabindex="-1"></a>    <span class="at">ylab=</span><span class="fu">expression</span>(Model<span class="sc">~</span>slope<span class="sc">~</span>(beta[<span class="dv">1</span>])))</span>
<span id="cb1304-31"><a href="mod-11.html#cb1304-31" tabindex="-1"></a><span class="fu">points</span>(taus, res<span class="sc">$</span>b1.lo, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">lwd=</span><span class="dv">2</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1304-32"><a href="mod-11.html#cb1304-32" tabindex="-1"></a><span class="fu">points</span>(taus, res<span class="sc">$</span>b1.up, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">lwd=</span><span class="dv">2</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1304-33"><a href="mod-11.html#cb1304-33" tabindex="-1"></a><span class="fu">segments</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="at">lty=</span><span class="dv">2</span>, <span class="at">lwd=</span><span class="dv">1</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-752-1.png" width="864" style="display: block; margin: auto;" /></p>
<p>How do we interpret this result? One result is that the model intercept increases steadily with quantile. In other words, the quantile regression line has a greater intercept for greater response quantiles. This is very common. The rightmost plot shows that the quantile regression slope doesn’t vary much between quantiles. This suggests that the data would have worked well with ordinary linear regression. If there is a trend in slope (<span class="math inline">\(\beta_1\)</span>) vs. quantile (<span class="math inline">\(\tau\)</span>), this often indicates that the predictor variables affect not only the mean response, but may limit the response (for significant <span class="math inline">\(\beta_1\)</span> at greater <span class="math inline">\(\tau\)</span>) or facilitate the response (for significant <span class="math inline">\(\beta_1\)</span> at smaller <span class="math inline">\(\tau\)</span>).</p>
</div>
</div>
<div id="mod-11-gam" class="section level3 hasAnchor" number="11.6.3">
<h3><span class="header-section-number">11.6.3</span> Generalized additive models (GAM)<a href="mod-11.html#mod-11-gam" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Models such as the LM and GLM, when fit with &gt;1 predictor variable, have another property that we have not discussed: the effects of the different predictors are additive. In other words, the effect of one predictor is added to the effect of the other predictors to define the expected value of the response. This means that the effects of the predictors are independent of each other.</p>
<p>Generalized additive models (GAM) go one step further than GLMs by relaxing an additional assumption: the assumption of a linear relationship between the response variable on the link scale and the predictors. GAMs fit a curve to data that can vary in complexity. These curves are commonly called “smoothers”. However, the parameters of the curve are not tested individually in the way that the parameters in something like a quadratic model or non-linear model would be. Instead, the curve itself is treated as a regression coefficient. Compare the equation for the linear predictor of a GLM with <em>k</em> predictors</p>
<p><span class="math display">\[g\left(E\left(Y\right)\right)=\beta_0+\beta_1X_1+\beta_1X_1+\ldots+\beta_kX_k\]</span></p>
<p>to the equation of the predictor of a GAM with <em>k</em> predictors:</p>
<p><span class="math display">\[g\left(E\left(Y\right)\right)=\beta_0+f_1\left(X_1\right)+f_2\left(X_2\right)+\ldots+f_k\left(X_k\right)\]</span></p>
<p>Instead of each predictor being multiplied by a coefficient, a smoothing function <span class="math inline">\(f_k\left(x\right)\)</span> of each predictor is estimated.</p>
<p>When should you use GAM? The most common application is situations where the response is not linearly related to the predictors, or if the response curve takes on a complicated shape. As you might suspect, the smoothers fitted in a GAM can be vulnerable to overfitting just as can be polynomial models or multiple regression models. Researchers usually control for this overfitting of the smoothers by limiting the complexity of the curves; this parameter is usually called the number of “knots”. Compared to NLS, GAMs have the significant advantage that the curve does not have to be defined ahead of time, and is thus much more flexible. The disadvantage is related to the advantage: unlike a LM, GLM, or NLS fit, a GAM fit will not result in model parameters (e.g., slopes) that can be interpreted.</p>
<p>Two good references to get you started are <span class="citation">Wood (<a href="#ref-wood2017generalized">2017</a>)</span> and <span class="citation">Zuur et al. (<a href="#ref-zuur2007analysing">2007</a>)</span>. The best-known R package for fitting GAMs is mgcv (<a href="https://cran.r-project.org/web/packages/mgcv/index.html">CRAN page</a>, accessed 2021-09-17).</p>
<div id="gam-example-with-real-data" class="section level4 hasAnchor" number="11.6.3.1">
<h4><span class="header-section-number">11.6.3.1</span> GAM example with real data<a href="mod-11.html#gam-example-with-real-data" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Let’s explore GAMs by re-analyzing one of the NLS models we saw earlier. Download the <a href="https://greenquanteco.github.io/woodward2015data.csv">dataset</a> here and put it in your R working directory.</p>
<div class="sourceCode" id="cb1305"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1305-1"><a href="mod-11.html#cb1305-1" tabindex="-1"></a>in.name <span class="ot">&lt;-</span> <span class="st">&quot;woodward2015data.csv&quot;</span></span>
<span id="cb1305-2"><a href="mod-11.html#cb1305-2" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(in.name, <span class="at">header=</span><span class="cn">TRUE</span>)</span>
<span id="cb1305-3"><a href="mod-11.html#cb1305-3" tabindex="-1"></a><span class="fu">plot</span>(dat<span class="sc">$</span>age, dat<span class="sc">$</span>mass)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-753-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>In a <a href="mod-11.html#mod-11-grow">previous section</a> we fit Gompertz and von Bertalanffy growth curves to the data. The Gompertz curve fit slightly better than the von Bertalanffy curve. The final Gompertz curve and the data are shown below:</p>
<p><img src="11_05.png" width="70%" style="display: block; margin: auto;" /></p>
<p>If we didn’t know about the Gompertz growth model, or had no idea what curve to start with, we could start by fitting GAMs to get a sense of the kind of curve we need. This is sometimes better than just plotting the data and eyeballing a curve. Fit the GAMs with the function <code>gam()</code>.</p>
<div class="sourceCode" id="cb1306"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1306-1"><a href="mod-11.html#cb1306-1" tabindex="-1"></a><span class="fu">library</span>(mgcv)</span></code></pre></div>
<pre><code>## Loading required package: nlme</code></pre>
<pre><code>## This is mgcv 1.9-0. For overview type &#39;help(&quot;mgcv-package&quot;)&#39;.</code></pre>
<div class="sourceCode" id="cb1309"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1309-1"><a href="mod-11.html#cb1309-1" tabindex="-1"></a>mod1 <span class="ot">&lt;-</span> <span class="fu">gam</span>(mass<span class="sc">~</span><span class="fu">s</span>(age), <span class="at">data=</span>dat)</span>
<span id="cb1309-2"><a href="mod-11.html#cb1309-2" tabindex="-1"></a><span class="fu">summary</span>(mod1)</span></code></pre></div>
<pre><code>## 
## Family: gaussian 
## Link function: identity 
## 
## Formula:
## mass ~ s(age)
## 
## Parametric coefficients:
##             Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept)  1351.14      34.37   39.31   &lt;2e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Approximate significance of smooth terms:
##          edf Ref.df     F p-value    
## s(age) 2.887  3.535 182.7  &lt;2e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## R-sq.(adj) =   0.95   Deviance explained = 95.4%
## GCV =  46522  Scale est. = 41356     n = 35</code></pre>
<p>Notice that the usual coefficients table is split into two parts: one for the intercept, and another for the smoother terms. The <code>s(age)</code> term in the model formula is how we requested a smoother on age. Without the <code>s()</code> function, we would get a linear term. The model summary also returns a pseudo-<em>R</em><sup>2</sup> based on deviance.</p>
<p>Next, we get the model predictions in the usual way:</p>
<div class="sourceCode" id="cb1311"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1311-1"><a href="mod-11.html#cb1311-1" tabindex="-1"></a>px <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fu">min</span>(dat<span class="sc">$</span>age), <span class="fu">max</span>(dat<span class="sc">$</span>age), <span class="at">length=</span><span class="dv">100</span>)</span>
<span id="cb1311-2"><a href="mod-11.html#cb1311-2" tabindex="-1"></a>pred <span class="ot">&lt;-</span> <span class="fu">predict</span>(mod1, <span class="at">newdata=</span><span class="fu">data.frame</span>(<span class="at">age=</span>px),</span>
<span id="cb1311-3"><a href="mod-11.html#cb1311-3" tabindex="-1"></a>    <span class="at">type=</span><span class="st">&quot;response&quot;</span>, <span class="at">se.fit=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p>The values in <code>pred</code> reveal a problem: the model predicts negative values! We probably should have used a log-link function to ensure positivity in the response.</p>
<div class="sourceCode" id="cb1312"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1312-1"><a href="mod-11.html#cb1312-1" tabindex="-1"></a>mod2 <span class="ot">&lt;-</span> <span class="fu">gam</span>(mass<span class="sc">~</span><span class="fu">s</span>(age), <span class="at">data=</span>dat, <span class="at">family=</span><span class="fu">gaussian</span>(<span class="at">link=</span><span class="st">&quot;log&quot;</span>))</span>
<span id="cb1312-2"><a href="mod-11.html#cb1312-2" tabindex="-1"></a><span class="fu">summary</span>(mod2)</span></code></pre></div>
<pre><code>## 
## Family: gaussian 
## Link function: log 
## 
## Formula:
## mass ~ s(age)
## 
## Parametric coefficients:
##             Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept)  6.94435    0.05751   120.7   &lt;2e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Approximate significance of smooth terms:
##          edf Ref.df     F p-value    
## s(age) 5.492  6.532 48.21  &lt;2e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## R-sq.(adj) =  0.949   Deviance explained = 95.7%
## GCV =  52451  Scale est. = 42723     n = 35</code></pre>
<p>Now make the predictions but on the link scale, which we can back-transform.</p>
<div class="sourceCode" id="cb1314"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1314-1"><a href="mod-11.html#cb1314-1" tabindex="-1"></a>px <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fu">min</span>(dat<span class="sc">$</span>age), <span class="fu">max</span>(dat<span class="sc">$</span>age), <span class="at">length=</span><span class="dv">100</span>)</span>
<span id="cb1314-2"><a href="mod-11.html#cb1314-2" tabindex="-1"></a>pred <span class="ot">&lt;-</span> <span class="fu">predict</span>(mod2, <span class="at">newdata=</span><span class="fu">data.frame</span>(<span class="at">age=</span>px),</span>
<span id="cb1314-3"><a href="mod-11.html#cb1314-3" tabindex="-1"></a>    <span class="at">type=</span><span class="st">&quot;link&quot;</span>, <span class="at">se.fit=</span><span class="cn">TRUE</span>)</span>
<span id="cb1314-4"><a href="mod-11.html#cb1314-4" tabindex="-1"></a>mn <span class="ot">&lt;-</span> pred<span class="sc">$</span>fit</span>
<span id="cb1314-5"><a href="mod-11.html#cb1314-5" tabindex="-1"></a>lo <span class="ot">&lt;-</span> <span class="fu">exp</span>(<span class="fu">qnorm</span>(<span class="fl">0.025</span>, mn, pred<span class="sc">$</span>se.fit))</span>
<span id="cb1314-6"><a href="mod-11.html#cb1314-6" tabindex="-1"></a>up <span class="ot">&lt;-</span> <span class="fu">exp</span>(<span class="fu">qnorm</span>(<span class="fl">0.975</span>, mn, pred<span class="sc">$</span>se.fit))</span>
<span id="cb1314-7"><a href="mod-11.html#cb1314-7" tabindex="-1"></a>mn <span class="ot">&lt;-</span> <span class="fu">exp</span>(mn)</span></code></pre></div>
<p>Now make a plot. Compare this to the equivalent plot from the Gompertz model fit above.</p>
<div class="sourceCode" id="cb1315"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1315-1"><a href="mod-11.html#cb1315-1" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>), <span class="at">mar=</span><span class="fu">c</span>(<span class="fl">5.1</span>, <span class="fl">7.1</span>, <span class="fl">1.1</span>, <span class="fl">1.1</span>),</span>
<span id="cb1315-2"><a href="mod-11.html#cb1315-2" tabindex="-1"></a>    <span class="at">las=</span><span class="dv">1</span>, <span class="at">lend=</span><span class="dv">1</span>, <span class="at">bty=</span><span class="st">&quot;n&quot;</span>, </span>
<span id="cb1315-3"><a href="mod-11.html#cb1315-3" tabindex="-1"></a>    <span class="at">cex.axis=</span><span class="fl">1.3</span>, <span class="at">cex.lab=</span><span class="fl">1.3</span>)</span>
<span id="cb1315-4"><a href="mod-11.html#cb1315-4" tabindex="-1"></a><span class="co"># make the plot</span></span>
<span id="cb1315-5"><a href="mod-11.html#cb1315-5" tabindex="-1"></a><span class="fu">plot</span>(dat<span class="sc">$</span>age, dat<span class="sc">$</span>mass,</span>
<span id="cb1315-6"><a href="mod-11.html#cb1315-6" tabindex="-1"></a>     <span class="at">pch=</span><span class="dv">16</span>, <span class="at">cex=</span><span class="fl">1.2</span>,</span>
<span id="cb1315-7"><a href="mod-11.html#cb1315-7" tabindex="-1"></a>     <span class="at">xlab=</span><span class="st">&quot;Age (years)&quot;</span>,</span>
<span id="cb1315-8"><a href="mod-11.html#cb1315-8" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">&quot;&quot;</span>, <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">4000</span>))</span>
<span id="cb1315-9"><a href="mod-11.html#cb1315-9" tabindex="-1"></a><span class="fu">title</span>(<span class="at">ylab=</span><span class="st">&quot;Mass (kg)&quot;</span>, <span class="at">line=</span><span class="dv">5</span>)</span>
<span id="cb1315-10"><a href="mod-11.html#cb1315-10" tabindex="-1"></a><span class="fu">points</span>(px, lo, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1315-11"><a href="mod-11.html#cb1315-11" tabindex="-1"></a><span class="fu">points</span>(px, up, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1315-12"><a href="mod-11.html#cb1315-12" tabindex="-1"></a><span class="fu">points</span>(px, mn, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-758-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Up to about 8 years of age, the GAM fit and the Gompertz fit look similar. However, above 8 years the GAM CI starts behaving strangely. This is because there are few data in that region with which to train the model. We could attempt to restrain the predictions by tweaking the model. One way is to reduce the number of “knots”, or bends in the smoother. This is done by adding an argument <em>k</em> to the <code>s()</code> function. Usually <em>k</em> should be <span class="math inline">\(\ge\)</span> 2 and &lt;<em>n</em>, where <em>n</em> is the number of observations. Try changing <em>k</em> in the commands below and see how the predictions change.</p>
<div class="sourceCode" id="cb1316"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1316-1"><a href="mod-11.html#cb1316-1" tabindex="-1"></a>mod3 <span class="ot">&lt;-</span> <span class="fu">gam</span>(mass<span class="sc">~</span><span class="fu">s</span>(age, <span class="at">k=</span><span class="dv">4</span>), <span class="at">data=</span>dat, <span class="at">family=</span><span class="fu">gaussian</span>(<span class="at">link=</span><span class="st">&quot;log&quot;</span>))</span>
<span id="cb1316-2"><a href="mod-11.html#cb1316-2" tabindex="-1"></a><span class="fu">summary</span>(mod3)</span></code></pre></div>
<pre><code>## 
## Family: gaussian 
## Link function: log 
## 
## Formula:
## mass ~ s(age, k = 4)
## 
## Parametric coefficients:
##             Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept)  6.99435    0.05263   132.9   &lt;2e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Approximate significance of smooth terms:
##          edf Ref.df    F p-value    
## s(age) 2.937  2.997 86.4  &lt;2e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## R-sq.(adj) =  0.933   Deviance explained = 93.8%
## GCV =  63276  Scale est. = 56159     n = 35</code></pre>
<div class="sourceCode" id="cb1318"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1318-1"><a href="mod-11.html#cb1318-1" tabindex="-1"></a>px <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fu">min</span>(dat<span class="sc">$</span>age), <span class="fu">max</span>(dat<span class="sc">$</span>age), <span class="at">length=</span><span class="dv">100</span>)</span>
<span id="cb1318-2"><a href="mod-11.html#cb1318-2" tabindex="-1"></a>pred <span class="ot">&lt;-</span> <span class="fu">predict</span>(mod3, <span class="at">newdata=</span><span class="fu">data.frame</span>(<span class="at">age=</span>px),</span>
<span id="cb1318-3"><a href="mod-11.html#cb1318-3" tabindex="-1"></a>    <span class="at">type=</span><span class="st">&quot;link&quot;</span>, <span class="at">se.fit=</span><span class="cn">TRUE</span>)</span>
<span id="cb1318-4"><a href="mod-11.html#cb1318-4" tabindex="-1"></a>mn <span class="ot">&lt;-</span> pred<span class="sc">$</span>fit</span>
<span id="cb1318-5"><a href="mod-11.html#cb1318-5" tabindex="-1"></a>lo <span class="ot">&lt;-</span> <span class="fu">exp</span>(<span class="fu">qnorm</span>(<span class="fl">0.025</span>, mn, pred<span class="sc">$</span>se.fit))</span>
<span id="cb1318-6"><a href="mod-11.html#cb1318-6" tabindex="-1"></a>up <span class="ot">&lt;-</span> <span class="fu">exp</span>(<span class="fu">qnorm</span>(<span class="fl">0.975</span>, mn, pred<span class="sc">$</span>se.fit))</span>
<span id="cb1318-7"><a href="mod-11.html#cb1318-7" tabindex="-1"></a>mn <span class="ot">&lt;-</span> <span class="fu">exp</span>(mn)</span>
<span id="cb1318-8"><a href="mod-11.html#cb1318-8" tabindex="-1"></a></span>
<span id="cb1318-9"><a href="mod-11.html#cb1318-9" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>), <span class="at">mar=</span><span class="fu">c</span>(<span class="fl">5.1</span>, <span class="fl">7.1</span>, <span class="fl">1.1</span>, <span class="fl">1.1</span>),</span>
<span id="cb1318-10"><a href="mod-11.html#cb1318-10" tabindex="-1"></a>    <span class="at">las=</span><span class="dv">1</span>, <span class="at">lend=</span><span class="dv">1</span>, <span class="at">bty=</span><span class="st">&quot;n&quot;</span>, </span>
<span id="cb1318-11"><a href="mod-11.html#cb1318-11" tabindex="-1"></a>    <span class="at">cex.axis=</span><span class="fl">1.3</span>, <span class="at">cex.lab=</span><span class="fl">1.3</span>)</span>
<span id="cb1318-12"><a href="mod-11.html#cb1318-12" tabindex="-1"></a><span class="co"># make the plot</span></span>
<span id="cb1318-13"><a href="mod-11.html#cb1318-13" tabindex="-1"></a><span class="fu">plot</span>(dat<span class="sc">$</span>age, dat<span class="sc">$</span>mass,</span>
<span id="cb1318-14"><a href="mod-11.html#cb1318-14" tabindex="-1"></a>     <span class="at">pch=</span><span class="dv">16</span>, <span class="at">cex=</span><span class="fl">1.2</span>,</span>
<span id="cb1318-15"><a href="mod-11.html#cb1318-15" tabindex="-1"></a>     <span class="at">xlab=</span><span class="st">&quot;Age (years)&quot;</span>,</span>
<span id="cb1318-16"><a href="mod-11.html#cb1318-16" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">&quot;&quot;</span>, <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">4000</span>))</span>
<span id="cb1318-17"><a href="mod-11.html#cb1318-17" tabindex="-1"></a><span class="fu">title</span>(<span class="at">ylab=</span><span class="st">&quot;Mass (kg)&quot;</span>, <span class="at">line=</span><span class="dv">5</span>)</span>
<span id="cb1318-18"><a href="mod-11.html#cb1318-18" tabindex="-1"></a><span class="fu">points</span>(px, lo, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1318-19"><a href="mod-11.html#cb1318-19" tabindex="-1"></a><span class="fu">points</span>(px, up, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb1318-20"><a href="mod-11.html#cb1318-20" tabindex="-1"></a><span class="fu">points</span>(px, mn, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-759-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Selecting the optimal number of knots is bit of an art as well as a science. Using too many knots places your model at risk of overfitting. Using too few knots risks not capturing a biologically relevant pattern. The best option is probably to fit the model using different numbers of knots and comparing them using AIC or predictive performance. For example:</p>
<div class="sourceCode" id="cb1319"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1319-1"><a href="mod-11.html#cb1319-1" tabindex="-1"></a>mod<span class="fl">.1</span> <span class="ot">&lt;-</span> <span class="fu">gam</span>(mass<span class="sc">~</span><span class="fu">s</span>(age), <span class="at">data=</span>dat, <span class="at">family=</span><span class="fu">gaussian</span>(<span class="at">link=</span><span class="st">&quot;log&quot;</span>))</span>
<span id="cb1319-2"><a href="mod-11.html#cb1319-2" tabindex="-1"></a>mod<span class="fl">.2</span> <span class="ot">&lt;-</span> <span class="fu">gam</span>(mass<span class="sc">~</span><span class="fu">s</span>(age, <span class="at">k=</span><span class="dv">2</span>), <span class="at">data=</span>dat, <span class="at">family=</span><span class="fu">gaussian</span>(<span class="at">link=</span><span class="st">&quot;log&quot;</span>))</span></code></pre></div>
<pre><code>## Warning in smooth.construct.tp.smooth.spec(object, dk$data, dk$knots): basis dimension, k, increased to minimum possible</code></pre>
<div class="sourceCode" id="cb1321"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1321-1"><a href="mod-11.html#cb1321-1" tabindex="-1"></a>mod<span class="fl">.3</span> <span class="ot">&lt;-</span> <span class="fu">gam</span>(mass<span class="sc">~</span><span class="fu">s</span>(age, <span class="at">k=</span><span class="dv">3</span>), <span class="at">data=</span>dat, <span class="at">family=</span><span class="fu">gaussian</span>(<span class="at">link=</span><span class="st">&quot;log&quot;</span>))</span>
<span id="cb1321-2"><a href="mod-11.html#cb1321-2" tabindex="-1"></a>mod<span class="fl">.4</span> <span class="ot">&lt;-</span> <span class="fu">gam</span>(mass<span class="sc">~</span><span class="fu">s</span>(age, <span class="at">k=</span><span class="dv">4</span>), <span class="at">data=</span>dat, <span class="at">family=</span><span class="fu">gaussian</span>(<span class="at">link=</span><span class="st">&quot;log&quot;</span>))</span>
<span id="cb1321-3"><a href="mod-11.html#cb1321-3" tabindex="-1"></a>mod<span class="fl">.5</span> <span class="ot">&lt;-</span> <span class="fu">gam</span>(mass<span class="sc">~</span><span class="fu">s</span>(age, <span class="at">k=</span><span class="dv">5</span>), <span class="at">data=</span>dat, <span class="at">family=</span><span class="fu">gaussian</span>(<span class="at">link=</span><span class="st">&quot;log&quot;</span>))</span>
<span id="cb1321-4"><a href="mod-11.html#cb1321-4" tabindex="-1"></a>mod<span class="fl">.6</span> <span class="ot">&lt;-</span> <span class="fu">gam</span>(mass<span class="sc">~</span><span class="fu">s</span>(age, <span class="at">k=</span><span class="dv">6</span>), <span class="at">data=</span>dat, <span class="at">family=</span><span class="fu">gaussian</span>(<span class="at">link=</span><span class="st">&quot;log&quot;</span>))</span>
<span id="cb1321-5"><a href="mod-11.html#cb1321-5" tabindex="-1"></a>mod<span class="fl">.7</span> <span class="ot">&lt;-</span> <span class="fu">gam</span>(mass<span class="sc">~</span><span class="fu">s</span>(age, <span class="at">k=</span><span class="dv">7</span>), <span class="at">data=</span>dat, <span class="at">family=</span><span class="fu">gaussian</span>(<span class="at">link=</span><span class="st">&quot;log&quot;</span>))</span>
<span id="cb1321-6"><a href="mod-11.html#cb1321-6" tabindex="-1"></a><span class="fu">AIC</span>(mod<span class="fl">.1</span>, mod<span class="fl">.2</span>, mod<span class="fl">.3</span>, mod<span class="fl">.4</span>, mod<span class="fl">.5</span>, mod<span class="fl">.6</span>, mod<span class="fl">.7</span>)</span></code></pre></div>
<pre><code>##             df      AIC
## mod.1 7.491541 480.3155
## mod.2 3.984267 501.6689
## mod.3 3.984267 501.6689
## mod.4 4.936648 487.7807
## mod.5 5.896483 480.4869
## mod.6 6.851747 477.4232
## mod.7 7.161195 478.5131</code></pre>
</div>
</div>
<div id="classification-and-regression-trees-cart" class="section level3 hasAnchor" number="11.6.4">
<h3><span class="header-section-number">11.6.4</span> Classification and regression trees (CART)<a href="mod-11.html#classification-and-regression-trees-cart" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The final alternative to NLS that we will explore is <strong>classification and regression trees (CART)</strong>. CART is really a family of methods that use nonparametric partitioning to either place observations into categories (classification trees) or predict their numeric values (regression trees). A typical regression tree is shown below <span class="citation">(<a href="#ref-de2000">De’ath and Fabricius 2000</a>)</span>:</p>
<p><img src="11_06.jpg" width="50%" style="display: block; margin: auto;" /></p>
<p>The tree predicts sediment thickness around coral reefs (as an ordinal variable) using two factors (shelf and reef type) and a numerical predictor (depth). The expected value for each observation is determined by going down the tree, selecting the correct statement about that observation. For example, an observation on the outer shelf (shelf = O), at 12 m depth, would have expected sediment thickness 0.46 (so values 0 or 1 would be most common). Each decision point is called a “node”; the predicted values are at the terminal nodes (or “leaves”). In this tree, the distribution of response values at each leaf is shown.</p>
<p>Tree-based models can describe patterns where there is no simple data model (e.g., curve shape), or where there are complex interactions between predictor variables (interactions manifest as effects being nested within each other). Tree-based models have the disadvantage that they require large amounts of data and can be difficult to present and interpret. Worse, it can be difficult to distinguish between a tree-based model that explains a high proportion of variation in the data, and a model that is badly overfit (i.e., fitting random noise). Cross-validation can help with this problem but is no guarantee of a robust fit.</p>
<p>Another name for these kinds of tree-based models is “recursive partitioning”. This means that the models partition the data over and over until homogenous groups are found. The main R package for CART is <code>rpart</code> <span class="citation">(<a href="#ref-therneau2019">Therneau and Atkinson 2019</a>)</span>.</p>
<div id="mod-11-cart1" class="section level4 hasAnchor" number="11.6.4.1">
<h4><span class="header-section-number">11.6.4.1</span> Classification tree example with real data<a href="mod-11.html#mod-11-cart1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><span class="citation">Hutcheon et al. (<a href="#ref-hutcheon2002">2002</a>)</span> reported data on brain morphology and lifestyle from 63 species of bats. Their dataset contains the following variables:</p>
<table>
<thead>
<tr class="header">
<th>Variable</th>
<th>Meaning</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>species</code></td>
<td>Species</td>
</tr>
<tr class="even">
<td><code>family</code></td>
<td>Taxonomic family</td>
</tr>
<tr class="odd">
<td><code>diet</code></td>
<td>Herbivore, gleaner, hawker, or vampire</td>
</tr>
<tr class="even">
<td><code>bow</code></td>
<td>Body weight (g)</td>
</tr>
<tr class="odd">
<td><code>brw</code></td>
<td>Brain weight (<span class="math inline">\(\mu\)</span>g)</td>
</tr>
<tr class="even">
<td><code>aud</code></td>
<td>Auditory nuclei volume (mm<sup>3</sup>)</td>
</tr>
<tr class="odd">
<td><code>mob</code></td>
<td>Main olfactory bulb volume (mm<sup>3</sup>)</td>
</tr>
<tr class="even">
<td><code>hip</code></td>
<td>Hippcampus volume (mm<sup>3</sup>)</td>
</tr>
</tbody>
</table>
<p>Import the dataset <code>bat_data_example.csv</code>. You can download it <a href="https://greenquanteco.github.io/bat_data_example.csv">here</a>. The code below requires that you have the file in your R working directory.</p>
<div class="sourceCode" id="cb1323"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1323-1"><a href="mod-11.html#cb1323-1" tabindex="-1"></a>in.name <span class="ot">&lt;-</span> <span class="st">&quot;bat_data_example.csv&quot;</span></span>
<span id="cb1323-2"><a href="mod-11.html#cb1323-2" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(in.name, <span class="at">header=</span><span class="cn">TRUE</span>)</span>
<span id="cb1323-3"><a href="mod-11.html#cb1323-3" tabindex="-1"></a>meas.cols <span class="ot">&lt;-</span> <span class="dv">4</span><span class="sc">:</span><span class="fu">ncol</span>(dat)</span></code></pre></div>
<p>We are interested in whether the relative sizes of different brain components are related to diet. Statistically, we are testing whether any combination of brain component sizes can correctly classify diet type.</p>
<p>Let’s briefly explore the data. First, we can see how the variables are related to each other. We can do that with a scatterplot matrix.</p>
<div class="sourceCode" id="cb1324"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1324-1"><a href="mod-11.html#cb1324-1" tabindex="-1"></a>panel.cor <span class="ot">&lt;-</span> <span class="cf">function</span>(x, y, <span class="at">digits =</span> <span class="dv">2</span>,</span>
<span id="cb1324-2"><a href="mod-11.html#cb1324-2" tabindex="-1"></a>    <span class="at">prefix =</span> <span class="st">&quot;&quot;</span>, cex.cor, ...){</span>
<span id="cb1324-3"><a href="mod-11.html#cb1324-3" tabindex="-1"></a>    r <span class="ot">&lt;-</span> <span class="fu">cor</span>(x, y, <span class="at">use=</span><span class="st">&quot;complete.obs&quot;</span>)</span>
<span id="cb1324-4"><a href="mod-11.html#cb1324-4" tabindex="-1"></a>    txt <span class="ot">&lt;-</span> <span class="fu">format</span>(<span class="fu">c</span>(r, <span class="fl">0.123456789</span>), <span class="at">digits =</span> digits)[<span class="dv">1</span>]</span>
<span id="cb1324-5"><a href="mod-11.html#cb1324-5" tabindex="-1"></a>    txt <span class="ot">&lt;-</span> <span class="fu">paste0</span>(prefix, txt)</span>
<span id="cb1324-6"><a href="mod-11.html#cb1324-6" tabindex="-1"></a>    <span class="cf">if</span>(<span class="fu">missing</span>(cex.cor)) cex.cor <span class="ot">&lt;-</span> <span class="fl">0.8</span><span class="sc">/</span><span class="fu">strwidth</span>(txt)</span>
<span id="cb1324-7"><a href="mod-11.html#cb1324-7" tabindex="-1"></a>    <span class="fu">text</span>(<span class="fl">0.5</span>, <span class="fl">0.5</span>, txt, <span class="at">cex =</span> cex.cor <span class="sc">*</span> <span class="fu">abs</span>(r))</span>
<span id="cb1324-8"><a href="mod-11.html#cb1324-8" tabindex="-1"></a>}</span>
<span id="cb1324-9"><a href="mod-11.html#cb1324-9" tabindex="-1"></a></span>
<span id="cb1324-10"><a href="mod-11.html#cb1324-10" tabindex="-1"></a><span class="fu">pairs</span>(dat[,meas.cols], </span>
<span id="cb1324-11"><a href="mod-11.html#cb1324-11" tabindex="-1"></a>    <span class="at">lower.panel =</span> panel.smooth, </span>
<span id="cb1324-12"><a href="mod-11.html#cb1324-12" tabindex="-1"></a>    <span class="at">upper.panel =</span> panel.cor,</span>
<span id="cb1324-13"><a href="mod-11.html#cb1324-13" tabindex="-1"></a>    <span class="at">gap=</span><span class="dv">0</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-763-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>What these correlations mean is that the variables are redundant with each other, and with overall size (body mass, <code>bow</code>). This makes a lot of sense for morphometric data. We need to keep this in mind as we analyze the data.</p>
<p>It might make sense to log-transform the data because we are working with measurements (which cannot be negative). Log-scale data will also be more normally distributed, which is not necessary for CART but can be more convenient.</p>
<div class="sourceCode" id="cb1325"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1325-1"><a href="mod-11.html#cb1325-1" tabindex="-1"></a><span class="co"># log transform</span></span>
<span id="cb1325-2"><a href="mod-11.html#cb1325-2" tabindex="-1"></a>dat[,meas.cols] <span class="ot">&lt;-</span> <span class="fu">apply</span>(dat[,meas.cols], <span class="dv">2</span>, log)</span></code></pre></div>
<p>In addition, we should we can factor out the effect of overall brain size. This will make it so that any patterns we see are not just a result of brain size. Below we use the geometric mean of the three brain volumes as the “total brain size” because the variable brw contains many other components (and is a mass, not a volume). Note that the mean of the log-transformed values is really the geometric mean, not the arithmetic mean.</p>
<div class="sourceCode" id="cb1326"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1326-1"><a href="mod-11.html#cb1326-1" tabindex="-1"></a>vols <span class="ot">&lt;-</span> meas.cols[<span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>)]</span>
<span id="cb1326-2"><a href="mod-11.html#cb1326-2" tabindex="-1"></a>dat<span class="sc">$</span>geom <span class="ot">&lt;-</span> <span class="fu">apply</span>(dat[,vols], <span class="dv">1</span>, mean)</span>
<span id="cb1326-3"><a href="mod-11.html#cb1326-3" tabindex="-1"></a>dat[,vols] <span class="ot">&lt;-</span> <span class="fu">apply</span>(dat[,vols], <span class="dv">2</span>, <span class="cf">function</span>(x){x<span class="sc">/</span>dat<span class="sc">$</span>geom})</span></code></pre></div>
<p>Remake the scatterplot matrix and observe that the three brain volumes are no longer so strongly correlated.</p>
<div class="sourceCode" id="cb1327"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1327-1"><a href="mod-11.html#cb1327-1" tabindex="-1"></a><span class="fu">pairs</span>(dat[,meas.cols], </span>
<span id="cb1327-2"><a href="mod-11.html#cb1327-2" tabindex="-1"></a>    <span class="at">lower.panel =</span> panel.smooth, </span>
<span id="cb1327-3"><a href="mod-11.html#cb1327-3" tabindex="-1"></a>    <span class="at">upper.panel =</span> panel.cor,</span>
<span id="cb1327-4"><a href="mod-11.html#cb1327-4" tabindex="-1"></a>    <span class="at">gap=</span><span class="dv">0</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/chunkx740-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Now we should be ready to fit a classification tree. The syntax is simple (and similar to a GLM):</p>
<div class="sourceCode" id="cb1328"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1328-1"><a href="mod-11.html#cb1328-1" tabindex="-1"></a><span class="fu">library</span>(rpart)</span>
<span id="cb1328-2"><a href="mod-11.html#cb1328-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb1328-3"><a href="mod-11.html#cb1328-3" tabindex="-1"></a>mod1 <span class="ot">&lt;-</span> <span class="fu">rpart</span>(diet<span class="sc">~</span>bow<span class="sc">+</span>aud<span class="sc">+</span>mob<span class="sc">+</span>hip, <span class="at">data=</span>dat)</span>
<span id="cb1328-4"><a href="mod-11.html#cb1328-4" tabindex="-1"></a></span>
<span id="cb1328-5"><a href="mod-11.html#cb1328-5" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>), <span class="at">mar=</span><span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">0.5</span>, <span class="fl">0.5</span>, <span class="fl">0.5</span>), <span class="at">xpd=</span><span class="cn">NA</span>)</span>
<span id="cb1328-6"><a href="mod-11.html#cb1328-6" tabindex="-1"></a><span class="fu">plot</span>(mod1)</span>
<span id="cb1328-7"><a href="mod-11.html#cb1328-7" tabindex="-1"></a><span class="fu">text</span>(mod1, <span class="at">use.n=</span><span class="cn">TRUE</span>, <span class="at">all=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-766-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>The default plot isn’t very attractive, but it does the job. A nicer version is hard to make in R, but can be assembled in Powerpoint or your favorite drawing program very easily. The tree summarizes graphically how the algorithm partitioned the data:</p>
<ul>
<li>If mob <span class="math inline">\(\ge\)</span> 0.9589, classify as herbivore.</li>
<li>If mob &lt; 0.9589 AND aud &lt; 0.7689, classify as gleaner.</li>
<li>If mob &lt; 0.9589 AND aud <span class="math inline">\(\ge\)</span> 0.7689, classify as hawker (aerial predator).</li>
</ul>
<p>For each observation, start at the top and go LEFT if the statement at a node is true, and right if the statement is false. The numbers at each terminal node show the number of observations from each category that wound up at that node. For example, 32 observations were classified as herbivores, and 28 of them were correct (0/2/28/2). We can also see that 22 out of 24 observations classified as hawkers by the model were actually hawkers. The confusion matrix for the model is obtained by making a frequency table of the predicted classifications and the actual classifications. The table below puts the predictions on the columns and the actual diets on the rows.</p>
<div class="sourceCode" id="cb1329"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1329-1"><a href="mod-11.html#cb1329-1" tabindex="-1"></a>pred <span class="ot">&lt;-</span> <span class="fu">predict</span>(mod1, <span class="at">type=</span><span class="st">&quot;class&quot;</span>)</span>
<span id="cb1329-2"><a href="mod-11.html#cb1329-2" tabindex="-1"></a><span class="fu">ftable</span>(pred<span class="sc">~</span>dat<span class="sc">$</span>diet)</span></code></pre></div>
<pre><code>##           pred gleaner hawker herbivore vampire
## dat$diet                                       
## gleaner              4      1         0       0
## hawker               3     22         2       0
## herbivore            0      1        28       0
## vampire              0      0         2       0</code></pre>
</div>
<div id="mod-11-cart2" class="section level4 hasAnchor" number="11.6.4.2">
<h4><span class="header-section-number">11.6.4.2</span> Regression tree example with real data<a href="mod-11.html#mod-11-cart2" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>We’ll use the bat data again to construct a simple regression tree. Can we predict main olfactory bulb volume from diet and other facts about a species? Load the bat data from above and go through the steps of log-transforming and relativizing to overall mass.</p>
<p>Now fit the model and plot it:</p>
<div class="sourceCode" id="cb1331"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1331-1"><a href="mod-11.html#cb1331-1" tabindex="-1"></a>mod2 <span class="ot">&lt;-</span> <span class="fu">rpart</span>(mob<span class="sc">~</span>diet<span class="sc">+</span>bow, <span class="at">data=</span>dat)</span>
<span id="cb1331-2"><a href="mod-11.html#cb1331-2" tabindex="-1"></a></span>
<span id="cb1331-3"><a href="mod-11.html#cb1331-3" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>), <span class="at">mar=</span><span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">0.5</span>, <span class="fl">0.5</span>, <span class="fl">0.5</span>), <span class="at">xpd=</span><span class="cn">NA</span>)</span>
<span id="cb1331-4"><a href="mod-11.html#cb1331-4" tabindex="-1"></a><span class="fu">plot</span>(mod2)</span>
<span id="cb1331-5"><a href="mod-11.html#cb1331-5" tabindex="-1"></a><span class="fu">text</span>(mod2, <span class="at">use.n=</span><span class="cn">TRUE</span>, <span class="at">all=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-768-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>The regression tree predicts the mean response variable at each terminal node. The logic of how to follow the tree to classify each observation is similar to before. At each node, go left when the statement is true and right when it is false. Factor levels are identified by letters corresponding to their order (usually alphabetical; check <code>levels(factor(dat$diet))</code>). Keep in mind that the values are all on the all on the log scale, because we transformed the values. Another option would have been to standardize the values.</p>
<p>The interpretation of this tree is as follows:</p>
<ul>
<li>Overall mean = 0.8695, <em>n</em> = 63 (top node)</li>
<li><code>bow</code> &lt; 1.793 <span class="math inline">\(\rightarrow\)</span> 8 observations with mean = 0.08476</li>
<li><code>bow</code> <span class="math inline">\(\ge\)</span> 1.793 and diet = gleaner (a) or hawker (b) <span class="math inline">\(\rightarrow\)</span> 31 observations with mean = 1.115</li>
<li><code>bow</code> <span class="math inline">\(\ge\)</span> 1.793 and diet <span class="math inline">\(\neq\)</span> gleaner or hawker<span class="math inline">\(\rightarrow\)</span> and bow &lt; 2.451 <span class="math inline">\(\rightarrow\)</span> 7 observations with mean = 0.6348</li>
<li><code>bow</code> <span class="math inline">\(\ge\)</span> 1.793 and diet <span class="math inline">\(\neq\)</span> gleaner or hawker<span class="math inline">\(\rightarrow\)</span> and bow <span class="math inline">\(\ge\)</span> 2.451 <span class="math inline">\(\rightarrow\)</span> 17 observations with mean = 0.8885</li>
</ul>
<p>We can visually inspect the effectiveness of the tree by plotting the predicted vs. observed values:</p>
<div class="sourceCode" id="cb1332"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1332-1"><a href="mod-11.html#cb1332-1" tabindex="-1"></a>pred <span class="ot">&lt;-</span> <span class="fu">predict</span>(mod2, <span class="at">type=</span><span class="st">&quot;vector&quot;</span>)</span>
<span id="cb1332-2"><a href="mod-11.html#cb1332-2" tabindex="-1"></a></span>
<span id="cb1332-3"><a href="mod-11.html#cb1332-3" tabindex="-1"></a>pchs <span class="ot">&lt;-</span> <span class="dv">15</span><span class="sc">:</span><span class="dv">18</span></span>
<span id="cb1332-4"><a href="mod-11.html#cb1332-4" tabindex="-1"></a>diets <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">unique</span>(dat<span class="sc">$</span>diet))</span>
<span id="cb1332-5"><a href="mod-11.html#cb1332-5" tabindex="-1"></a>use.pch <span class="ot">&lt;-</span> pchs[<span class="fu">match</span>(dat<span class="sc">$</span>diet, diets)]</span>
<span id="cb1332-6"><a href="mod-11.html#cb1332-6" tabindex="-1"></a><span class="fu">plot</span>(dat<span class="sc">$</span>mob, <span class="fu">jitter</span>(pred), <span class="at">pch=</span>use.pch)</span>
<span id="cb1332-7"><a href="mod-11.html#cb1332-7" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">a=</span><span class="dv">0</span>, <span class="at">b=</span><span class="dv">1</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>)</span>
<span id="cb1332-8"><a href="mod-11.html#cb1332-8" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;topleft&quot;</span>, <span class="at">legend=</span>diets, <span class="at">pch=</span>pchs)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-769-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>Remember that regression trees only predict the mean within each group, not individual values. A good regression tree will have a relatively small spread within each group; the tree we just fit is not very good.</p>
<p>Better results can be obtained with random forests <span class="citation">(<a href="#ref-cutler2007">Cutler et al. 2007</a>, <a href="#ref-brieuc2018">Brieuc et al. 2018</a>)</span> and boosted regression trees (BRT) <span class="citation">(<a href="#ref-death2007">De’ath 2007</a>, <a href="#ref-elith2008">Elith et al. 2008</a>)</span>. BRT use not one but many trees to obtain a more robust fit. Both of those techniques are powerful and underutilized in biology…but that is changing. At some point I’ll produce a page demonstrating BRT. <span class="citation">James et al. (<a href="#ref-james2013">2013</a>)</span> contains primers on these and other machine learning techniques.</p>

</div>
</div>
</div>
</div>
<h3>Literature Cited<a href="literature-cited.html#literature-cited" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0" line-spacing="2">
<div id="ref-bolker2008ecological" class="csl-entry">
Bolker, B. M. 2008. Ecological models and data in <span>R</span>. Princeton University Press, Princeton, New Jersey, USA.
</div>
<div id="ref-brieuc2018" class="csl-entry">
Brieuc, M. S. O., C. D. Waters, D. P. Drinan, and K. A. Naish. 2018. <a href="https://doi.org/10.1111/1755-0998.12773">A practical introduction to random forest for genetic association studies in ecology and evolution</a>. Molecular Ecology Resources 18:755–766.
</div>
<div id="ref-cade2003" class="csl-entry">
Cade, B. S., and B. R. Noon. 2003. <a href="https://doi.org/10.1890/1540-9295(2003)001[0412:agitqr]2.0.co;2">A gentle introduction to quantile regression for ecologists</a>. Frontiers in Ecology and the Environment 1:412–420.
</div>
<div id="ref-cutler2007" class="csl-entry">
Cutler, D. R., T. C. Edwards, K. H. Beard, A. Cutler, K. T. Hess, J. Gibson, and J. J. Lawler. 2007. <a href="https://doi.org/10.1890/07-0539.1">Random forests for classification in ecology</a>. Ecology 88:2783–2792.
</div>
<div id="ref-death2007" class="csl-entry">
De’ath, G. 2007. <a href="https://doi.org/10.1890/0012-9658(2007)88[243:btfema]2.0.co;2">Boosted trees for ecological modeling and prediction</a>. Ecology 88:243–251.
</div>
<div id="ref-de2000" class="csl-entry">
De’ath, G., and K. E. Fabricius. 2000. <a href="https://doi.org/10.1890/0012-9658(2000)081[3178:CARTAP]2.0.CO;2">Classification and regression trees: A powerful yet simple technique for ecological data analysis</a>. Ecology 81:3178–3192.
</div>
<div id="ref-elith2008" class="csl-entry">
Elith, J., J. R. Leathwick, and T. Hastie. 2008. <a href="https://doi.org/10.1111/j.1365-2656.2008.01390.x">A working guide to boosted regression trees</a>. Journal of Animal Ecology 77:802–813.
</div>
<div id="ref-foucault2018" class="csl-entry">
Foucault, Q., A. Wieser, A.-M. Waldvogel, B. Feldmeyer, and M. Pfenninger. 2018. <a href="https://doi.org/10.1002/ece3.4706">Rapid adaptation to high temperatures in <em><span>C</span>hironomus riparius</em></a>. Ecology and Evolution 8:12780–12789.
</div>
<div id="ref-green2020" class="csl-entry">
Green, N. S., M. L. Wildhaber, J. L. Albers, T. W. Pettit, and M. J. Hooper. 2020. <a href="https://doi.org/10.1002/ieam.4324">Efficient mammal biodiversity surveys for ecological restoration monitoring</a>. Integrated Environmental Assessment and Management.
</div>
<div id="ref-hutcheon2002" class="csl-entry">
Hutcheon, J. M., J. A. W. Kirsch, and T. G. Jr. 2002. <a href="https://doi.org/10.1159/000065938">A comparative analysis of brain size in relation to foraging ecology and phylogeny in the <span>C</span>hiroptera</a>. Brain, behavior and evolution 60:165–180.
</div>
<div id="ref-inderjit2002" class="csl-entry">
Inderjit, J. C. Streibig, and M. Olofsdotter. 2002. <a href="https://doi.org/10.1034/j.1399-3054.2002.1140312.x">Joint action of phenolic acid mixtures and its significance in allelopathy research</a>. Physiologia Plantarum 114:422–428.
</div>
<div id="ref-james2013" class="csl-entry">
James, G., D. Witten, T. Hastie, and R. Tibshirani. 2013. <a href="http://faculty.marshall.usc.edu/gareth-james/ISL/">An introduction to statistical learning</a>. 7th edition. Springer.
</div>
<div id="ref-koenker2021" class="csl-entry">
Koenker, R. 2021. <a href="https://CRAN.R-project.org/package=quantreg">Quantreg: Quantile regression</a>.
</div>
<div id="ref-kunz2019" class="csl-entry">
Kunz, B. K., J. H. Waddle, and N. S. Green. 2019. <a href="https://doi.org/10.1002/ieam.4202">Amphibian monitoring in hardwood forests: Optimizing methods for contaminant-based compensatory restorations</a>. Integrated environmental assessment and management.
</div>
<div id="ref-leonard2013acute" class="csl-entry">
Leonard, E. M., and C. M. Wood. 2013. <a href="https://doi.org/10.1016/j.cbpc.2013.03.008">Acute toxicity, critical body residues, <span>M</span>ichaelis–<span>M</span>enten analysis of bioaccumulation, and ionoregulatory disturbance in response to waterborne nickel in four invertebrates: <em><span>C</span>hironomus riparius</em>, <em><span>L</span>ymnaea stagnalis</em>, <em><span>L</span>umbriculus variegatus</em> and <em><span>D</span>aphnia pulex</em></a>. Comparative Biochemistry and Physiology Part C: Toxicology &amp; Pharmacology 158:10–21.
</div>
<div id="ref-mangiafico2021" class="csl-entry">
Mangiafico, S. 2021. <a href="https://CRAN.R-project.org/package=rcompanion"><span class="nocase">r</span>companion: Functions to support extension education program evaluation</a>.
</div>
<div id="ref-ritz2015" class="csl-entry">
Ritz, C., F. Baty, J. C. Streibig, and D. Gerhard. 2015. <a href="https://doi.org/10.1371/journal.pone.0146021">Dose-response analysis using <span>R</span></a>. PLoS ONE 10:e0146021.
</div>
<div id="ref-schabenberger2001contemporary" class="csl-entry">
Schabenberger, O., and F. J. Pierce. 2001. Contemporary statistical models for the plant and soil sciences. CRC press, Boca Raton, Florida, USA.
</div>
<div id="ref-spiess2018" class="csl-entry">
Spiess, A.-N. 2018. <a href="https://CRAN.R-project.org/package=propagate"><span class="nocase">p</span>ropagate: Propagation of uncertainty</a>.
</div>
<div id="ref-spiess2010" class="csl-entry">
Spiess, A.-N., and N. Neumeyer. 2010. <a href="https://doi.org/10.1186/1471-2210-10-6">An evaluation of <span class="math inline">\(R^2\)</span> as an inadequate measure for nonlinear models in pharmacological and biochemical research: A <span>M</span>onte <span>C</span>arlo approach</a>. BMC Pharmacology 10.
</div>
<div id="ref-therneau2019" class="csl-entry">
Therneau, T., and B. Atkinson. 2019. <a href="https://CRAN.R-project.org/package=rpart"><span class="nocase">r</span>part: Recursive partitioning and regression trees</a>.
</div>
<div id="ref-turner2020" class="csl-entry">
Turner, H., and D. Firth. 2020. <a href="https://cran.r-project.org/package=gnm">Generalized nonlinear models in <span>R</span>: An overview of the gnm package</a>.
</div>
<div id="ref-wood2017generalized" class="csl-entry">
Wood, S. N. 2017. Generalized additive models: An introduction with <span>R</span>. 2nd edition. CRC press.
</div>
<div id="ref-woodward2015" class="csl-entry">
Woodward, H. N., E. A. F. Fowler, J. O. Farlow, and J. R. Horner. 2015. <a href="https://doi.org/10.1017/pab.2015.19"><em><span>M</span>aiasaura</em>, a model organism for extinct vertebrate population biology: A large sample statistical assessment of growth dynamics and survivorship</a>. Paleobiology 41:503–527.
</div>
<div id="ref-zuur2007analysing" class="csl-entry">
Zuur, A. F., E. N. Ieno, and G. M. Smith. 2007. Analysing ecological data. Springer Science+Business Media, New York.
</div>
</div>
<div class="footnotes">
<hr />
<ol start="82">
<li id="fn82"><p>As an ecologist, I have never fit a Michaelis-Menten model for enzyme kinetics :)<a href="mod-11.html#fnref82" class="footnote-back">↩︎</a></p></li>
<li id="fn83"><p>Infuriatingly, the base function <code>predict.nls()</code> takes <code>se.fit</code> as an argument, but ignores it. This has been the case since I was grad student learning about nonlinear models.<a href="mod-11.html#fnref83" class="footnote-back">↩︎</a></p></li>
<li id="fn84"><p>Image: <span class="citation">Foucault et al. (<a href="#ref-foucault2018">2018</a>)</span><a href="mod-11.html#fnref84" class="footnote-back">↩︎</a></p></li>
<li id="fn85"><p>Dummy variables are really a way to manipulate the matrix algebra underlying the model-fitting process. Usually, a factor with <em>n</em> levels will be encoded by <span class="math inline">\(n-1\)</span> dummy variables where each observation has a 1 in at most one of the dummies—the dummy corresponding to the group they belong to. An observation with 0s for all dummies is interpreted as being in the control or baseline group. This example uses a slightly different approach to dummy coding where each level corresponds to one dummy variable and each observation has value 1 for exactly 1 dummy. The first method (<span class="math inline">\(n-1\)</span> dummies) leads to an <strong>effects parameterization</strong> of the model, where effects are presented as changes from the baseline (e.g., as seen in R’s <code>lm()</code> outputs). The second method (<em>n</em> dummies) produces a <strong>means parameterization</strong> of the model, where parameter estimates are presented as group- or level-specific point estimates.<a href="mod-11.html#fnref85" class="footnote-back">↩︎</a></p></li>
<li id="fn86"><p>Photo: V. Socha. <a href="https://upload.wikimedia.org/wikipedia/commons/1/16/Maiasaura_skeleton.jpg">Wikimedia</a> (accessed 2021-09-24)<a href="mod-11.html#fnref86" class="footnote-back">↩︎</a></p></li>
<li id="fn87"><p>Image: <a href="https://www.flower-db.com/en/flowers/lolium-perenne" class="uri">https://www.flower-db.com/en/flowers/lolium-perenne</a>) (accessed 2021-09-24)<a href="mod-11.html#fnref87" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="mod-10.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="mod-12.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/11-non.rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["applied_biol_data_analysis_20220816.pdf", "applied_biol_data_analysis_20220816.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
