<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Module 2 Introduction to R | Applied Biological Data Analysis: Statistics and R for Biologists</title>
  <meta name="description" content="Helping biologists to become more informed users of R and statistical methods." />
  <meta name="generator" content="bookdown 0.37 and GitBook 2.6.7" />

  <meta property="og:title" content="Module 2 Introduction to R | Applied Biological Data Analysis: Statistics and R for Biologists" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://greenquanteco.github.io/book/lab_logo_02.jpg" />
  <meta property="og:description" content="Helping biologists to become more informed users of R and statistical methods." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Module 2 Introduction to R | Applied Biological Data Analysis: Statistics and R for Biologists" />
  
  <meta name="twitter:description" content="Helping biologists to become more informed users of R and statistical methods." />
  <meta name="twitter:image" content="https://greenquanteco.github.io/book/lab_logo_02.jpg" />

<meta name="author" content="Dr. Nick Green, Kennesaw State University" />


<meta name="date" content="2024-08-20" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="fav.ico" type="image/x-icon" />
<link rel="prev" href="mod-01.html"/>
<link rel="next" href="mod-03.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Applied biological data analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#work-in-progress-warning"><i class="fa fa-check"></i>Work in progress warning!</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#license-and-permissions"><i class="fa fa-check"></i>License and permissions</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-description"><i class="fa fa-check"></i>Course description</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-objectives"><i class="fa fa-check"></i>Course objectives</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-requirements"><i class="fa fa-check"></i>Course requirements</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#recommended-reading"><i class="fa fa-check"></i>Recommended reading</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-organization"><i class="fa fa-check"></i>Course organization</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#about-the-author"><i class="fa fa-check"></i>About the author</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="mod-01.html"><a href="mod-01.html"><i class="fa fa-check"></i><b>1</b> Statistics in modern biology</a>
<ul>
<li class="chapter" data-level="1.1" data-path="mod-01.html"><a href="mod-01.html#overview"><i class="fa fa-check"></i><b>1.1</b> Overview</a></li>
<li class="chapter" data-level="1.2" data-path="mod-01.html"><a href="mod-01.html#statistics-in-modern-biology"><i class="fa fa-check"></i><b>1.2</b> Statistics in modern biology</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="mod-01.html"><a href="mod-01.html#the-scientific-method"><i class="fa fa-check"></i><b>1.2.1</b> The scientific method</a></li>
<li class="chapter" data-level="1.2.2" data-path="mod-01.html"><a href="mod-01.html#example-data-analysis"><i class="fa fa-check"></i><b>1.2.2</b> Example data analysis</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="mod-01.html"><a href="mod-01.html#misuses-of-statistics"><i class="fa fa-check"></i><b>1.3</b> Misuses of statistics</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="mod-01.html"><a href="mod-01.html#proving-the-trivial-and-meaningless-hypotheses"><i class="fa fa-check"></i><b>1.3.1</b> “Proving” the trivial and meaningless hypotheses</a></li>
<li class="chapter" data-level="1.3.2" data-path="mod-01.html"><a href="mod-01.html#inappropriate-methods"><i class="fa fa-check"></i><b>1.3.2</b> Inappropriate methods</a></li>
<li class="chapter" data-level="1.3.3" data-path="mod-01.html"><a href="mod-01.html#p-hacking-and-data-dredging"><i class="fa fa-check"></i><b>1.3.3</b> <em>P</em>-hacking and data dredging</a></li>
<li class="chapter" data-level="1.3.4" data-path="mod-01.html"><a href="mod-01.html#inadequate-sample-sizes-and-pseudoreplication"><i class="fa fa-check"></i><b>1.3.4</b> Inadequate sample sizes and pseudoreplication</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="mod-01.html"><a href="mod-01.html#mod-01-nhst"><i class="fa fa-check"></i><b>1.4</b> <em>P</em>-values and null hypothesis significance testing (NHST)</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="mod-01.html"><a href="mod-01.html#definition"><i class="fa fa-check"></i><b>1.4.1</b> Definition</a></li>
<li class="chapter" data-level="1.4.2" data-path="mod-01.html"><a href="mod-01.html#aside-another-way-to-frame-null-hypotheses"><i class="fa fa-check"></i><b>1.4.2</b> Aside: Another way to frame null hypotheses</a></li>
<li class="chapter" data-level="1.4.3" data-path="mod-01.html"><a href="mod-01.html#history-and-status-of-p-values"><i class="fa fa-check"></i><b>1.4.3</b> History and status of <em>P</em>-values</a></li>
<li class="chapter" data-level="1.4.4" data-path="mod-01.html"><a href="mod-01.html#where-p-values-come-from"><i class="fa fa-check"></i><b>1.4.4</b> Where <em>P</em>-values come from</a></li>
<li class="chapter" data-level="1.4.5" data-path="mod-01.html"><a href="mod-01.html#what-p-values-mean-and-do-not-mean"><i class="fa fa-check"></i><b>1.4.5</b> What <em>P</em>-values mean and do not mean</a></li>
<li class="chapter" data-level="1.4.6" data-path="mod-01.html"><a href="mod-01.html#do-you-need-a-p-value"><i class="fa fa-check"></i><b>1.4.6</b> Do you need a <em>P</em>-value?</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="mod-01.html"><a href="mod-01.html#alternatives-to-nhst"><i class="fa fa-check"></i><b>1.5</b> Alternatives to NHST</a>
<ul>
<li class="chapter" data-level="1.5.1" data-path="mod-01.html"><a href="mod-01.html#mod-01-bayes"><i class="fa fa-check"></i><b>1.5.1</b> Bayesian inference</a></li>
<li class="chapter" data-level="1.5.2" data-path="mod-01.html"><a href="mod-01.html#information-theoretic-methods"><i class="fa fa-check"></i><b>1.5.2</b> Information-theoretic methods</a></li>
<li class="chapter" data-level="1.5.3" data-path="mod-01.html"><a href="mod-01.html#machine-learning"><i class="fa fa-check"></i><b>1.5.3</b> Machine learning</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="mod-02.html"><a href="mod-02.html"><i class="fa fa-check"></i><b>2</b> Introduction to R</a>
<ul>
<li class="chapter" data-level="2.1" data-path="mod-02.html"><a href="mod-02.html#getting-started-with-r"><i class="fa fa-check"></i><b>2.1</b> Getting started with R</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="mod-02.html"><a href="mod-02.html#what-is-r"><i class="fa fa-check"></i><b>2.1.1</b> What is R?</a></li>
<li class="chapter" data-level="2.1.2" data-path="mod-02.html"><a href="mod-02.html#advantages-of-r"><i class="fa fa-check"></i><b>2.1.2</b> Advantages of R</a></li>
<li class="chapter" data-level="2.1.3" data-path="mod-02.html"><a href="mod-02.html#disadvantages-of-r"><i class="fa fa-check"></i><b>2.1.3</b> Disadvantages of R</a></li>
<li class="chapter" data-level="2.1.4" data-path="mod-02.html"><a href="mod-02.html#base-r-and-vs.-tidyverse"><i class="fa fa-check"></i><b>2.1.4</b> Base R and (vs.?) <code>tidyverse</code></a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="mod-02.html"><a href="mod-02.html#download-and-install-r-and-rstudio"><i class="fa fa-check"></i><b>2.2</b> Download and install R (and RStudio)</a></li>
<li class="chapter" data-level="2.3" data-path="mod-02.html"><a href="mod-02.html#using-r"><i class="fa fa-check"></i><b>2.3</b> Using R</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="mod-02.html"><a href="mod-02.html#using-the-base-r-gui"><i class="fa fa-check"></i><b>2.3.1</b> Using the base R GUI</a></li>
<li class="chapter" data-level="2.3.2" data-path="mod-02.html"><a href="mod-02.html#using-r-in-rstudio"><i class="fa fa-check"></i><b>2.3.2</b> Using R in RStudio</a></li>
<li class="chapter" data-level="2.3.3" data-path="mod-02.html"><a href="mod-02.html#using-r-with-other-programs"><i class="fa fa-check"></i><b>2.3.3</b> Using R with other programs</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="mod-02.html"><a href="mod-02.html#mod-02-first"><i class="fa fa-check"></i><b>2.4</b> A first R session</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="mod-02.html"><a href="mod-02.html#import-data"><i class="fa fa-check"></i><b>2.4.1</b> Import data</a></li>
<li class="chapter" data-level="2.4.2" data-path="mod-02.html"><a href="mod-02.html#explore-and-visualize-data"><i class="fa fa-check"></i><b>2.4.2</b> Explore and visualize data</a></li>
<li class="chapter" data-level="2.4.3" data-path="mod-02.html"><a href="mod-02.html#transform-data"><i class="fa fa-check"></i><b>2.4.3</b> Transform data</a></li>
<li class="chapter" data-level="2.4.4" data-path="mod-02.html"><a href="mod-02.html#analyze-data"><i class="fa fa-check"></i><b>2.4.4</b> Analyze data</a></li>
<li class="chapter" data-level="2.4.5" data-path="mod-02.html"><a href="mod-02.html#mod-01-elkfig"><i class="fa fa-check"></i><b>2.4.5</b> Write out results</a></li>
<li class="chapter" data-level="2.4.6" data-path="mod-02.html"><a href="mod-02.html#save-your-work"><i class="fa fa-check"></i><b>2.4.6</b> Save your work?</a></li>
<li class="chapter" data-level="2.4.7" data-path="mod-02.html"><a href="mod-02.html#whats-next"><i class="fa fa-check"></i><b>2.4.7</b> What’s next?</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="mod-02.html"><a href="mod-02.html#write-and-execute-commands-in-the-r-console"><i class="fa fa-check"></i><b>2.5</b> Write and execute commands in the R console</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="mod-02.html"><a href="mod-02.html#r-commandsbasics"><i class="fa fa-check"></i><b>2.5.1</b> R commands–basics</a></li>
<li class="chapter" data-level="2.5.2" data-path="mod-02.html"><a href="mod-02.html#elements-of-r-code"><i class="fa fa-check"></i><b>2.5.2</b> Elements of R code</a></li>
<li class="chapter" data-level="2.5.3" data-path="mod-02.html"><a href="mod-02.html#the-r-workspace"><i class="fa fa-check"></i><b>2.5.3</b> The R workspace</a></li>
<li class="chapter" data-level="2.5.4" data-path="mod-02.html"><a href="mod-02.html#r-code-basics-assignment-and-operators"><i class="fa fa-check"></i><b>2.5.4</b> R code basics: assignment and operators</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="mod-02.html"><a href="mod-02.html#mod-02-struct"><i class="fa fa-check"></i><b>2.6</b> Basic R data structures</a>
<ul>
<li class="chapter" data-level="2.6.1" data-path="mod-02.html"><a href="mod-02.html#vectors"><i class="fa fa-check"></i><b>2.6.1</b> Vectors</a></li>
<li class="chapter" data-level="2.6.2" data-path="mod-02.html"><a href="mod-02.html#data-frames"><i class="fa fa-check"></i><b>2.6.2</b> Data frames</a></li>
<li class="chapter" data-level="2.6.3" data-path="mod-02.html"><a href="mod-02.html#matrices-and-arrays"><i class="fa fa-check"></i><b>2.6.3</b> Matrices and arrays</a></li>
<li class="chapter" data-level="2.6.4" data-path="mod-02.html"><a href="mod-02.html#lists"><i class="fa fa-check"></i><b>2.6.4</b> Lists</a></li>
<li class="chapter" data-level="2.6.5" data-path="mod-02.html"><a href="mod-02.html#s3-s4-and-other-paradigms"><i class="fa fa-check"></i><b>2.6.5</b> S3, S4, and other paradigms</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="mod-02.html"><a href="mod-02.html#r-data-types"><i class="fa fa-check"></i><b>2.7</b> R data types</a>
<ul>
<li class="chapter" data-level="2.7.1" data-path="mod-02.html"><a href="mod-02.html#character-type"><i class="fa fa-check"></i><b>2.7.1</b> Character type</a></li>
<li class="chapter" data-level="2.7.2" data-path="mod-02.html"><a href="mod-02.html#numeric-type"><i class="fa fa-check"></i><b>2.7.2</b> Numeric type</a></li>
<li class="chapter" data-level="2.7.3" data-path="mod-02.html"><a href="mod-02.html#integer-type"><i class="fa fa-check"></i><b>2.7.3</b> Integer type</a></li>
<li class="chapter" data-level="2.7.4" data-path="mod-02.html"><a href="mod-02.html#logical-type"><i class="fa fa-check"></i><b>2.7.4</b> Logical type</a></li>
<li class="chapter" data-level="2.7.5" data-path="mod-02.html"><a href="mod-02.html#special-values"><i class="fa fa-check"></i><b>2.7.5</b> Special values</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="mod-02.html"><a href="mod-02.html#manage-r-code-as-scripts-.r-files"><i class="fa fa-check"></i><b>2.8</b> Manage R code as scripts (.r files)</a></li>
<li class="chapter" data-level="2.9" data-path="mod-02.html"><a href="mod-02.html#manage-and-use-r-packages"><i class="fa fa-check"></i><b>2.9</b> Manage and use R packages</a>
<ul>
<li class="chapter" data-level="2.9.1" data-path="mod-02.html"><a href="mod-02.html#your-r-library"><i class="fa fa-check"></i><b>2.9.1</b> Your R library</a></li>
<li class="chapter" data-level="2.9.2" data-path="mod-02.html"><a href="mod-02.html#installing-packages-using-the-r-gui"><i class="fa fa-check"></i><b>2.9.2</b> Installing packages using the R GUI</a></li>
<li class="chapter" data-level="2.9.3" data-path="mod-02.html"><a href="mod-02.html#installing-packages-in-rstudio"><i class="fa fa-check"></i><b>2.9.3</b> Installing packages in RStudio</a></li>
<li class="chapter" data-level="2.9.4" data-path="mod-02.html"><a href="mod-02.html#installing-packages-using-the-r-console"><i class="fa fa-check"></i><b>2.9.4</b> Installing packages using the R console</a></li>
<li class="chapter" data-level="2.9.5" data-path="mod-02.html"><a href="mod-02.html#working-with-packages-in-r"><i class="fa fa-check"></i><b>2.9.5</b> Working with packages in R</a></li>
<li class="chapter" data-level="2.9.6" data-path="mod-02.html"><a href="mod-02.html#package-dependencies"><i class="fa fa-check"></i><b>2.9.6</b> Package dependencies</a></li>
<li class="chapter" data-level="2.9.7" data-path="mod-02.html"><a href="mod-02.html#citing-packages"><i class="fa fa-check"></i><b>2.9.7</b> Citing packages</a></li>
</ul></li>
<li class="chapter" data-level="2.10" data-path="mod-02.html"><a href="mod-02.html#r-documentation"><i class="fa fa-check"></i><b>2.10</b> R documentation</a>
<ul>
<li class="chapter" data-level="2.10.1" data-path="mod-02.html"><a href="mod-02.html#documentation-help-files"><i class="fa fa-check"></i><b>2.10.1</b> Documentation (help) files</a></li>
<li class="chapter" data-level="2.10.2" data-path="mod-02.html"><a href="mod-02.html#r-vignettes"><i class="fa fa-check"></i><b>2.10.2</b> R vignettes</a></li>
<li class="chapter" data-level="2.10.3" data-path="mod-02.html"><a href="mod-02.html#official-r-project-resources"><i class="fa fa-check"></i><b>2.10.3</b> Official R Project resources</a></li>
<li class="chapter" data-level="2.10.4" data-path="mod-02.html"><a href="mod-02.html#unofficial-online-resources"><i class="fa fa-check"></i><b>2.10.4</b> Unofficial online resources</a></li>
<li class="chapter" data-level="2.10.5" data-path="mod-02.html"><a href="mod-02.html#r-books"><i class="fa fa-check"></i><b>2.10.5</b> R books</a></li>
<li class="chapter" data-level="2.10.6" data-path="mod-02.html"><a href="mod-02.html#two-reminders"><i class="fa fa-check"></i><b>2.10.6</b> Two reminders</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="mod-03.html"><a href="mod-03.html"><i class="fa fa-check"></i><b>3</b> Data manipulation with R</a>
<ul>
<li class="chapter" data-level="3.1" data-path="mod-03.html"><a href="mod-03.html#mod-03-inout"><i class="fa fa-check"></i><b>3.1</b> Data import and export</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="mod-03.html"><a href="mod-03.html#importing-data-preliminaries"><i class="fa fa-check"></i><b>3.1.1</b> Importing data: preliminaries</a></li>
<li class="chapter" data-level="3.1.2" data-path="mod-03.html"><a href="mod-03.html#importing-data-from-text-files-with-read.csv-and-read.table"><i class="fa fa-check"></i><b>3.1.2</b> Importing data from text files with <code>read.csv()</code> and <code>read.table()</code></a></li>
<li class="chapter" data-level="3.1.3" data-path="mod-03.html"><a href="mod-03.html#importing-data-from-saved-workspaces"><i class="fa fa-check"></i><b>3.1.3</b> Importing data from saved workspaces</a></li>
<li class="chapter" data-level="3.1.4" data-path="mod-03.html"><a href="mod-03.html#importing-data-special-cases"><i class="fa fa-check"></i><b>3.1.4</b> Importing data: special cases:</a></li>
<li class="chapter" data-level="3.1.5" data-path="mod-03.html"><a href="mod-03.html#export-data-from-r"><i class="fa fa-check"></i><b>3.1.5</b> Export data from R</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="mod-03.html"><a href="mod-03.html#making-values-in-r"><i class="fa fa-check"></i><b>3.2</b> Making values in R</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="mod-03.html"><a href="mod-03.html#producing-arbitrary-values-with-c"><i class="fa fa-check"></i><b>3.2.1</b> Producing arbitrary values with <code>c()</code></a></li>
<li class="chapter" data-level="3.2.2" data-path="mod-03.html"><a href="mod-03.html#generating-regular-values"><i class="fa fa-check"></i><b>3.2.2</b> Generating regular values</a></li>
<li class="chapter" data-level="3.2.3" data-path="mod-03.html"><a href="mod-03.html#generating-random-values"><i class="fa fa-check"></i><b>3.2.3</b> Generating random values</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="mod-03.html"><a href="mod-03.html#selecting-data-with"><i class="fa fa-check"></i><b>3.3</b> Selecting data with <code>[]</code></a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="mod-03.html"><a href="mod-03.html#mod-03-brackets"><i class="fa fa-check"></i><b>3.3.1</b> Basics of brackets</a></li>
<li class="chapter" data-level="3.3.2" data-path="mod-03.html"><a href="mod-03.html#extracting-and-selecting-data-with-logical-tests"><i class="fa fa-check"></i><b>3.3.2</b> Extracting and selecting data with logical tests</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="mod-03.html"><a href="mod-03.html#managing-dates-and-characters"><i class="fa fa-check"></i><b>3.4</b> Managing dates and characters</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="mod-03.html"><a href="mod-03.html#temporal-data-and-dates"><i class="fa fa-check"></i><b>3.4.1</b> Temporal data and dates</a></li>
<li class="chapter" data-level="3.4.2" data-path="mod-03.html"><a href="mod-03.html#character-data-text"><i class="fa fa-check"></i><b>3.4.2</b> Character data (text)</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="mod-03.html"><a href="mod-03.html#mod-03-dataframe"><i class="fa fa-check"></i><b>3.5</b> Data frame management</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="mod-03.html"><a href="mod-03.html#data-frame-structure"><i class="fa fa-check"></i><b>3.5.1</b> Data frame structure</a></li>
<li class="chapter" data-level="3.5.2" data-path="mod-03.html"><a href="mod-03.html#common-data-frame-operations"><i class="fa fa-check"></i><b>3.5.2</b> Common data frame operations</a></li>
<li class="chapter" data-level="3.5.3" data-path="mod-03.html"><a href="mod-03.html#other-data-frame-operations"><i class="fa fa-check"></i><b>3.5.3</b> Other data frame operations</a></li>
<li class="chapter" data-level="3.5.4" data-path="mod-03.html"><a href="mod-03.html#mod-03-reshape"><i class="fa fa-check"></i><b>3.5.4</b> Reshaping data frames</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="mod-04.html"><a href="mod-04.html"><i class="fa fa-check"></i><b>4</b> Exploratory data analysis 1: Data description and summarization</a>
<ul>
<li class="chapter" data-level="4.1" data-path="mod-04.html"><a href="mod-04.html#motivation"><i class="fa fa-check"></i><b>4.1</b> Motivation</a></li>
<li class="chapter" data-level="4.2" data-path="mod-04.html"><a href="mod-04.html#descriptive-and-summary-statistics"><i class="fa fa-check"></i><b>4.2</b> Descriptive and summary statistics</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="mod-04.html"><a href="mod-04.html#basic-summary-statistics"><i class="fa fa-check"></i><b>4.2.1</b> Basic summary statistics</a></li>
<li class="chapter" data-level="4.2.2" data-path="mod-04.html"><a href="mod-04.html#mod-04-agg"><i class="fa fa-check"></i><b>4.2.2</b> Summarizing by groups (AKA: pivot tables)</a></li>
<li class="chapter" data-level="4.2.3" data-path="mod-04.html"><a href="mod-04.html#summarizing-data-with-the-apply-family"><i class="fa fa-check"></i><b>4.2.3</b> Summarizing data with the <code>apply()</code> family</a></li>
<li class="chapter" data-level="4.2.4" data-path="mod-04.html"><a href="mod-04.html#mod-04-tab"><i class="fa fa-check"></i><b>4.2.4</b> Tabulation (frequency tables)</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="mod-04.html"><a href="mod-04.html#mod-04-vis"><i class="fa fa-check"></i><b>4.3</b> Visualizing single variables</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="mod-04.html"><a href="mod-04.html#boxplots-aka-box-and-whisker-plots"><i class="fa fa-check"></i><b>4.3.1</b> Boxplots (aka: box-and-whisker plots)</a></li>
<li class="chapter" data-level="4.3.2" data-path="mod-04.html"><a href="mod-04.html#histograms"><i class="fa fa-check"></i><b>4.3.2</b> Histograms</a></li>
<li class="chapter" data-level="4.3.3" data-path="mod-04.html"><a href="mod-04.html#mod04-pdfplots"><i class="fa fa-check"></i><b>4.3.3</b> Kernel density plots</a></li>
<li class="chapter" data-level="4.3.4" data-path="mod-04.html"><a href="mod-04.html#empirical-cumulative-density-function-ecdf-plots"><i class="fa fa-check"></i><b>4.3.4</b> Empirical cumulative density function (ECDF) plots</a></li>
<li class="chapter" data-level="4.3.5" data-path="mod-04.html"><a href="mod-04.html#quantile-quantile-qq-plots"><i class="fa fa-check"></i><b>4.3.5</b> Quantile-quantile (QQ) plots</a></li>
<li class="chapter" data-level="4.3.6" data-path="mod-04.html"><a href="mod-04.html#how-should-i-plot-my-data"><i class="fa fa-check"></i><b>4.3.6</b> How should I plot my data?</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="mod-05.html"><a href="mod-05.html"><i class="fa fa-check"></i><b>5</b> Exploratory data analysis 2: Probability distributions for biology</a>
<ul>
<li class="chapter" data-level="5.1" data-path="mod-05.html"><a href="mod-05.html#mod-05-dists1"><i class="fa fa-check"></i><b>5.1</b> Statistical distributions</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="mod-05.html"><a href="mod-05.html#probability-distributions-in-r"><i class="fa fa-check"></i><b>5.1.1</b> Probability distributions in R</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="mod-05.html"><a href="mod-05.html#mod-05-discrete"><i class="fa fa-check"></i><b>5.2</b> Discrete distributions</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="mod-05.html"><a href="mod-05.html#bernoulli-distribution"><i class="fa fa-check"></i><b>5.2.1</b> Bernoulli distribution</a></li>
<li class="chapter" data-level="5.2.2" data-path="mod-05.html"><a href="mod-05.html#mod-05-binom"><i class="fa fa-check"></i><b>5.2.2</b> Binomial distribution</a></li>
<li class="chapter" data-level="5.2.3" data-path="mod-05.html"><a href="mod-05.html#poisson-distribution"><i class="fa fa-check"></i><b>5.2.3</b> Poisson distribution</a></li>
<li class="chapter" data-level="5.2.4" data-path="mod-05.html"><a href="mod-05.html#mod-05-negbin"><i class="fa fa-check"></i><b>5.2.4</b> Negative binomial distribution</a></li>
<li class="chapter" data-level="5.2.5" data-path="mod-05.html"><a href="mod-05.html#mod-05-geom"><i class="fa fa-check"></i><b>5.2.5</b> Geometric distribution</a></li>
<li class="chapter" data-level="5.2.6" data-path="mod-05.html"><a href="mod-05.html#beta-binomial-distribution"><i class="fa fa-check"></i><b>5.2.6</b> Beta-binomial distribution</a></li>
<li class="chapter" data-level="5.2.7" data-path="mod-05.html"><a href="mod-05.html#multinomial-distribution"><i class="fa fa-check"></i><b>5.2.7</b> Multinomial distribution</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="mod-05.html"><a href="mod-05.html#mod-05-condists"><i class="fa fa-check"></i><b>5.3</b> Continuous distributions</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="mod-05.html"><a href="mod-05.html#uniform-distribution"><i class="fa fa-check"></i><b>5.3.1</b> Uniform distribution</a></li>
<li class="chapter" data-level="5.3.2" data-path="mod-05.html"><a href="mod-05.html#mod-05-norm"><i class="fa fa-check"></i><b>5.3.2</b> Normal distribution</a></li>
<li class="chapter" data-level="5.3.3" data-path="mod-05.html"><a href="mod-05.html#lognormal-distribution"><i class="fa fa-check"></i><b>5.3.3</b> Lognormal distribution</a></li>
<li class="chapter" data-level="5.3.4" data-path="mod-05.html"><a href="mod-05.html#mod-05-gamma"><i class="fa fa-check"></i><b>5.3.4</b> Gamma distribution</a></li>
<li class="chapter" data-level="5.3.5" data-path="mod-05.html"><a href="mod-05.html#beta-distribution"><i class="fa fa-check"></i><b>5.3.5</b> Beta distribution</a></li>
<li class="chapter" data-level="5.3.6" data-path="mod-05.html"><a href="mod-05.html#dirichlet-distribution"><i class="fa fa-check"></i><b>5.3.6</b> Dirichlet distribution</a></li>
<li class="chapter" data-level="5.3.7" data-path="mod-05.html"><a href="mod-05.html#mod-05-expo"><i class="fa fa-check"></i><b>5.3.7</b> Exponential distribution</a></li>
<li class="chapter" data-level="5.3.8" data-path="mod-05.html"><a href="mod-05.html#triangular-distribution"><i class="fa fa-check"></i><b>5.3.8</b> Triangular distribution</a></li>
<li class="chapter" data-level="5.3.9" data-path="mod-05.html"><a href="mod-05.html#distributions-summary"><i class="fa fa-check"></i><b>5.3.9</b> Distributions summary</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="mod-05.html"><a href="mod-05.html#mod-05-dists2"><i class="fa fa-check"></i><b>5.4</b> Fitting and testing distributions</a>
<ul>
<li class="chapter" data-level="5.4.1" data-path="mod-05.html"><a href="mod-05.html#estimating-distributional-parameters"><i class="fa fa-check"></i><b>5.4.1</b> Estimating distributional parameters</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="mod-05.html"><a href="mod-05.html#graphical-methods-for-examining-distributions"><i class="fa fa-check"></i><b>5.5</b> Graphical methods for examining distributions</a>
<ul>
<li class="chapter" data-level="5.5.1" data-path="mod-05.html"><a href="mod-05.html#comparing-distributions-with-pdf-or-pmf-plots"><i class="fa fa-check"></i><b>5.5.1</b> Comparing distributions with PDF or PMF plots</a></li>
<li class="chapter" data-level="5.5.2" data-path="mod-05.html"><a href="mod-05.html#comparing-distributions-using-cdf-and-ecdf"><i class="fa fa-check"></i><b>5.5.2</b> Comparing distributions using CDF and ECDF</a></li>
<li class="chapter" data-level="5.5.3" data-path="mod-05.html"><a href="mod-05.html#comparing-distributions-using-qq-plots"><i class="fa fa-check"></i><b>5.5.3</b> Comparing distributions using QQ plots</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="mod-05.html"><a href="mod-05.html#formal-tests-for-distributions"><i class="fa fa-check"></i><b>5.6</b> Formal tests for distributions</a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="mod-05.html"><a href="mod-05.html#tests-for-normality"><i class="fa fa-check"></i><b>5.6.1</b> Tests for normality</a></li>
<li class="chapter" data-level="5.6.2" data-path="mod-05.html"><a href="mod-05.html#kolmogorov-smirnov-tests"><i class="fa fa-check"></i><b>5.6.2</b> Kolmogorov-Smirnov tests</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="mod-05.html"><a href="mod-05.html#mod-05-trans"><i class="fa fa-check"></i><b>5.7</b> Data transformations</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="mod-05.html"><a href="mod-05.html#why-transform"><i class="fa fa-check"></i><b>5.7.1</b> Why transform?</a></li>
<li class="chapter" data-level="5.7.2" data-path="mod-05.html"><a href="mod-05.html#transforms-vs.-link-functions"><i class="fa fa-check"></i><b>5.7.2</b> Transforms vs. link functions</a></li>
<li class="chapter" data-level="5.7.3" data-path="mod-05.html"><a href="mod-05.html#common-transformations"><i class="fa fa-check"></i><b>5.7.3</b> Common transformations</a></li>
<li class="chapter" data-level="5.7.4" data-path="mod-05.html"><a href="mod-05.html#uncommon-transformations"><i class="fa fa-check"></i><b>5.7.4</b> Uncommon transformations</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="mod-06.html"><a href="mod-06.html"><i class="fa fa-check"></i><b>6</b> Exploratory data analysis 3: Multivariate data exploration</a>
<ul>
<li class="chapter" data-level="6.1" data-path="mod-06.html"><a href="mod-06.html#scatterplots-for-2-continuous-variables"><i class="fa fa-check"></i><b>6.1</b> Scatterplots for 2 continuous variables</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="mod-06.html"><a href="mod-06.html#r-plotting-basics"><i class="fa fa-check"></i><b>6.1.1</b> R plotting basics</a></li>
<li class="chapter" data-level="6.1.2" data-path="mod-06.html"><a href="mod-06.html#graphical-parameters-and-advanced-plots-with-par"><i class="fa fa-check"></i><b>6.1.2</b> Graphical parameters and advanced plots with <code>par()</code></a></li>
<li class="chapter" data-level="6.1.3" data-path="mod-06.html"><a href="mod-06.html#really-advanced-plotting"><i class="fa fa-check"></i><b>6.1.3</b> Really advanced plotting</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="mod-06.html"><a href="mod-06.html#mod-06-smat"><i class="fa fa-check"></i><b>6.2</b> Scatterplot matrices for many variables</a></li>
<li class="chapter" data-level="6.3" data-path="mod-06.html"><a href="mod-06.html#lattice-plots-for-hierarchical-data"><i class="fa fa-check"></i><b>6.3</b> Lattice plots for hierarchical data</a></li>
<li class="chapter" data-level="6.4" data-path="mod-06.html"><a href="mod-06.html#continuous-variables-by-group"><i class="fa fa-check"></i><b>6.4</b> Continuous variables by group</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="mod-06.html"><a href="mod-06.html#boxplots"><i class="fa fa-check"></i><b>6.4.1</b> Boxplots</a></li>
<li class="chapter" data-level="6.4.2" data-path="mod-06.html"><a href="mod-06.html#barplots"><i class="fa fa-check"></i><b>6.4.2</b> Barplots</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="mod-06.html"><a href="mod-06.html#ordination-brief-introduction"><i class="fa fa-check"></i><b>6.5</b> Ordination (brief introduction)</a>
<ul>
<li class="chapter" data-level="6.5.1" data-path="mod-06.html"><a href="mod-06.html#principal-components-analysis-pca"><i class="fa fa-check"></i><b>6.5.1</b> Principal components analysis (PCA)</a></li>
<li class="chapter" data-level="6.5.2" data-path="mod-06.html"><a href="mod-06.html#nonmetric-multidimensional-scaling-nmds"><i class="fa fa-check"></i><b>6.5.2</b> Nonmetric multidimensional scaling (NMDS)</a></li>
<li class="chapter" data-level="6.5.3" data-path="mod-06.html"><a href="mod-06.html#plotting-ordinations"><i class="fa fa-check"></i><b>6.5.3</b> Plotting ordinations</a></li>
<li class="chapter" data-level="6.5.4" data-path="mod-06.html"><a href="mod-06.html#ordination-wrap-up-for-now"><i class="fa fa-check"></i><b>6.5.4</b> Ordination wrap-up (for now)</a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="mod-06.html"><a href="mod-06.html#principles-for-good-data-graphics"><i class="fa fa-check"></i><b>6.6</b> Principles for good data graphics</a>
<ul>
<li class="chapter" data-level="6.6.1" data-path="mod-06.html"><a href="mod-06.html#maximize-the-data-to-ink-ratio"><i class="fa fa-check"></i><b>6.6.1</b> Maximize the data-to-ink ratio</a></li>
<li class="chapter" data-level="6.6.2" data-path="mod-06.html"><a href="mod-06.html#show-variability"><i class="fa fa-check"></i><b>6.6.2</b> Show variability</a></li>
<li class="chapter" data-level="6.6.3" data-path="mod-06.html"><a href="mod-06.html#use-labels-not-symbols"><i class="fa fa-check"></i><b>6.6.3</b> Use labels, not symbols</a></li>
<li class="chapter" data-level="6.6.4" data-path="mod-06.html"><a href="mod-06.html#be-consistent"><i class="fa fa-check"></i><b>6.6.4</b> Be consistent</a></li>
<li class="chapter" data-level="6.6.5" data-path="mod-06.html"><a href="mod-06.html#set-yourself-up-for-success"><i class="fa fa-check"></i><b>6.6.5</b> Set yourself up for success</a></li>
<li class="chapter" data-level="6.6.6" data-path="mod-06.html"><a href="mod-06.html#consider-legibility"><i class="fa fa-check"></i><b>6.6.6</b> Consider legibility</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="mod-07.html"><a href="mod-07.html"><i class="fa fa-check"></i><b>7</b> Common statistical problems</a>
<ul>
<li class="chapter" data-level="7.1" data-path="mod-07.html"><a href="mod-07.html#outliers-and-erroneous-values"><i class="fa fa-check"></i><b>7.1</b> Outliers and erroneous values</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="mod-07.html"><a href="mod-07.html#detecting-outliers"><i class="fa fa-check"></i><b>7.1.1</b> Detecting outliers</a></li>
<li class="chapter" data-level="7.1.2" data-path="mod-07.html"><a href="mod-07.html#dealing-with-outliers"><i class="fa fa-check"></i><b>7.1.2</b> Dealing with outliers</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="mod-07.html"><a href="mod-07.html#autocorrelation"><i class="fa fa-check"></i><b>7.2</b> Autocorrelation</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="mod-07.html"><a href="mod-07.html#graphical-methods-to-identify-autocorrelation"><i class="fa fa-check"></i><b>7.2.1</b> Graphical methods to identify autocorrelation</a></li>
<li class="chapter" data-level="7.2.2" data-path="mod-07.html"><a href="mod-07.html#mantel-tests-for-autocorrelation"><i class="fa fa-check"></i><b>7.2.2</b> Mantel tests for autocorrelation</a></li>
<li class="chapter" data-level="7.2.3" data-path="mod-07.html"><a href="mod-07.html#dealing-with-autocorrelation"><i class="fa fa-check"></i><b>7.2.3</b> Dealing with autocorrelation</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="mod-07.html"><a href="mod-07.html#mod-07-multicol"><i class="fa fa-check"></i><b>7.3</b> Collinearity</a></li>
<li class="chapter" data-level="7.4" data-path="mod-07.html"><a href="mod-07.html#missing-data"><i class="fa fa-check"></i><b>7.4</b> Missing data</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="mod-07.html"><a href="mod-07.html#option-1-ignore-missing-data"><i class="fa fa-check"></i><b>7.4.1</b> Option 1: Ignore missing data</a></li>
<li class="chapter" data-level="7.4.2" data-path="mod-07.html"><a href="mod-07.html#option-2-interpolate-or-impute-missing-data"><i class="fa fa-check"></i><b>7.4.2</b> Option 2: Interpolate or impute missing data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="mod-08.html"><a href="mod-08.html"><i class="fa fa-check"></i><b>8</b> Planning your analysis (what test?)</a>
<ul>
<li class="chapter" data-level="8.1" data-path="mod-08.html"><a href="mod-08.html#how-to-use-this-guide"><i class="fa fa-check"></i><b>8.1</b> How to use this guide</a></li>
<li class="chapter" data-level="8.2" data-path="mod-08.html"><a href="mod-08.html#what-question-are-you-trying-to-answer"><i class="fa fa-check"></i><b>8.2</b> What question are you trying to answer?</a></li>
<li class="chapter" data-level="8.3" data-path="mod-08.html"><a href="mod-08.html#testing-for-a-difference-in-mean-or-location"><i class="fa fa-check"></i><b>8.3</b> Testing for a difference in mean or location</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="mod-08.html"><a href="mod-08.html#additional-considerations"><i class="fa fa-check"></i><b>8.3.1</b> Additional considerations</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="mod-08.html"><a href="mod-08.html#testing-for-a-continuous-relationship-between-two-or-more-variables"><i class="fa fa-check"></i><b>8.4</b> Testing for a continuous relationship between two or more variables</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="mod-08.html"><a href="mod-08.html#additional-considerations-1"><i class="fa fa-check"></i><b>8.4.1</b> Additional considerations</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="mod-08.html"><a href="mod-08.html#independence-between-categories"><i class="fa fa-check"></i><b>8.5</b> Independence between categories</a></li>
<li class="chapter" data-level="8.6" data-path="mod-08.html"><a href="mod-08.html#classifying-observations"><i class="fa fa-check"></i><b>8.6</b> Classifying observations</a></li>
<li class="chapter" data-level="8.7" data-path="mod-08.html"><a href="mod-08.html#conclusions"><i class="fa fa-check"></i><b>8.7</b> Conclusions</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="mod-09.html"><a href="mod-09.html"><i class="fa fa-check"></i><b>9</b> Basic statistics with R</a>
<ul>
<li class="chapter" data-level="9.1" data-path="mod-09.html"><a href="mod-09.html#mod-09-datasets"><i class="fa fa-check"></i><b>9.1</b> Datasets and variables</a></li>
<li class="chapter" data-level="9.2" data-path="mod-09.html"><a href="mod-09.html#mod-09-frequentist"><i class="fa fa-check"></i><b>9.2</b> Frequentist statistics</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="mod-09.html"><a href="mod-09.html#mod-09-pvalues"><i class="fa fa-check"></i><b>9.2.1</b> <em>P</em>-values and NHST</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="mod-09.html"><a href="mod-09.html#mod-09-whattest"><i class="fa fa-check"></i><b>9.3</b> What kind of test do you need?</a></li>
<li class="chapter" data-level="9.4" data-path="mod-09.html"><a href="mod-09.html#mod-09-tests-for-loc"><i class="fa fa-check"></i><b>9.4</b> Tests for differences in mean or location</a>
<ul>
<li class="chapter" data-level="9.4.1" data-path="mod-09.html"><a href="mod-09.html#tests-for-1-group"><i class="fa fa-check"></i><b>9.4.1</b> Tests for 1 group</a></li>
<li class="chapter" data-level="9.4.2" data-path="mod-09.html"><a href="mod-09.html#tests-comparing-two-groups"><i class="fa fa-check"></i><b>9.4.2</b> Tests comparing two groups</a></li>
<li class="chapter" data-level="9.4.3" data-path="mod-09.html"><a href="mod-09.html#aside-whats-with-all-these-t-tests"><i class="fa fa-check"></i><b>9.4.3</b> Aside: what’s with all these <em>t</em>-tests?</a></li>
<li class="chapter" data-level="9.4.4" data-path="mod-09.html"><a href="mod-09.html#comparing-3-or-more-groups"><i class="fa fa-check"></i><b>9.4.4</b> Comparing 3 or more groups</a></li>
<li class="chapter" data-level="9.4.5" data-path="mod-09.html"><a href="mod-09.html#ordered-factors-and-polynomial-patterns-in-grouped-data"><i class="fa fa-check"></i><b>9.4.5</b> Ordered factors and polynomial patterns in grouped data</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="mod-09.html"><a href="mod-09.html#tests-for-patterns-in-continuous-data"><i class="fa fa-check"></i><b>9.5</b> Tests for patterns in continuous data</a>
<ul>
<li class="chapter" data-level="9.5.1" data-path="mod-09.html"><a href="mod-09.html#mod-09-correlation"><i class="fa fa-check"></i><b>9.5.1</b> Correlation</a></li>
<li class="chapter" data-level="9.5.2" data-path="mod-09.html"><a href="mod-09.html#mod-09-linear-mods"><i class="fa fa-check"></i><b>9.5.2</b> Linear models</a></li>
<li class="chapter" data-level="9.5.3" data-path="mod-09.html"><a href="mod-09.html#linear-models-wrap-up-for-now"><i class="fa fa-check"></i><b>9.5.3</b> Linear models wrap-up (for now)</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="mod-09.html"><a href="mod-09.html#mod-09-contingency"><i class="fa fa-check"></i><b>9.6</b> Tests for proportions and contingency</a>
<ul>
<li class="chapter" data-level="9.6.1" data-path="mod-09.html"><a href="mod-09.html#mod-09-ctables"><i class="fa fa-check"></i><b>9.6.1</b> Contingency tables</a></li>
<li class="chapter" data-level="9.6.2" data-path="mod-09.html"><a href="mod-09.html#tests-on-contingency-tables"><i class="fa fa-check"></i><b>9.6.2</b> Tests on contingency tables</a></li>
<li class="chapter" data-level="9.6.3" data-path="mod-09.html"><a href="mod-09.html#mod-09-oddsratios"><i class="fa fa-check"></i><b>9.6.3</b> Contingency tests and odds ratios</a></li>
</ul></li>
<li class="chapter" data-level="9.7" data-path="mod-09.html"><a href="mod-09.html#mod-09-classification"><i class="fa fa-check"></i><b>9.7</b> Classification</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="mod-10.html"><a href="mod-10.html"><i class="fa fa-check"></i><b>10</b> Generalized linear models (GLM)</a>
<ul>
<li class="chapter" data-level="10.1" data-path="mod-10.html"><a href="mod-10.html#mod-10-lm"><i class="fa fa-check"></i><b>10.1</b> Prelude with linear models</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="mod-10.html"><a href="mod-10.html#assumptions-of-linear-models"><i class="fa fa-check"></i><b>10.1.1</b> Assumptions of linear models</a></li>
<li class="chapter" data-level="10.1.2" data-path="mod-10.html"><a href="mod-10.html#linear-regression-in-r"><i class="fa fa-check"></i><b>10.1.2</b> Linear regression in R</a></li>
<li class="chapter" data-level="10.1.3" data-path="mod-10.html"><a href="mod-10.html#multiple-linear-regression"><i class="fa fa-check"></i><b>10.1.3</b> Multiple linear regression</a></li>
<li class="chapter" data-level="10.1.4" data-path="mod-10.html"><a href="mod-10.html#mod-10-anova"><i class="fa fa-check"></i><b>10.1.4</b> ANOVA and ANCOVA with <code>lm()</code></a></li>
<li class="chapter" data-level="10.1.5" data-path="mod-10.html"><a href="mod-10.html#variations-on-linear-models"><i class="fa fa-check"></i><b>10.1.5</b> Variations on linear models</a></li>
<li class="chapter" data-level="10.1.6" data-path="mod-10.html"><a href="mod-10.html#example-linear-regression-workflow"><i class="fa fa-check"></i><b>10.1.6</b> Example linear regression workflow</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="mod-10.html"><a href="mod-10.html#mod-10-basic"><i class="fa fa-check"></i><b>10.2</b> GLM basics</a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="mod-10.html"><a href="mod-10.html#example-glms"><i class="fa fa-check"></i><b>10.2.1</b> Example GLMS</a></li>
<li class="chapter" data-level="10.2.2" data-path="mod-10.html"><a href="mod-10.html#glm-families"><i class="fa fa-check"></i><b>10.2.2</b> GLM families</a></li>
<li class="chapter" data-level="10.2.3" data-path="mod-10.html"><a href="mod-10.html#glm-link-functions"><i class="fa fa-check"></i><b>10.2.3</b> GLM link functions</a></li>
<li class="chapter" data-level="10.2.4" data-path="mod-10.html"><a href="mod-10.html#deviance-and-other-glm-diagnostics"><i class="fa fa-check"></i><b>10.2.4</b> Deviance and other GLM diagnostics</a></li>
<li class="chapter" data-level="10.2.5" data-path="mod-10.html"><a href="mod-10.html#to-pseudo-r2-or-not-to-pseudo-r2"><i class="fa fa-check"></i><b>10.2.5</b> To pseudo-<em>R</em><sup>2</sup> or not to pseudo-<em>R</em><sup>2</sup>?</a></li>
<li class="chapter" data-level="10.2.6" data-path="mod-10.html"><a href="mod-10.html#common-glms"><i class="fa fa-check"></i><b>10.2.6</b> Common GLMs</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="mod-10.html"><a href="mod-10.html#mod-10-loglin"><i class="fa fa-check"></i><b>10.3</b> Log-linear models</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="mod-10.html"><a href="mod-10.html#mod-10-loglin-examp"><i class="fa fa-check"></i><b>10.3.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="10.3.2" data-path="mod-10.html"><a href="mod-10.html#example-with-real-data"><i class="fa fa-check"></i><b>10.3.2</b> Example with real data</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="mod-10.html"><a href="mod-10.html#mod-10-poisson"><i class="fa fa-check"></i><b>10.4</b> Poisson GLM for counts</a>
<ul>
<li class="chapter" data-level="10.4.1" data-path="mod-10.html"><a href="mod-10.html#example-with-simulated-data"><i class="fa fa-check"></i><b>10.4.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="10.4.2" data-path="mod-10.html"><a href="mod-10.html#example-with-real-data-1"><i class="fa fa-check"></i><b>10.4.2</b> Example with real data</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="mod-10.html"><a href="mod-10.html#mod-10-quasi"><i class="fa fa-check"></i><b>10.5</b> Quasi-Poisson and negative binomial GLM</a>
<ul>
<li class="chapter" data-level="10.5.1" data-path="mod-10.html"><a href="mod-10.html#example-with-simulated-data-1"><i class="fa fa-check"></i><b>10.5.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="10.5.2" data-path="mod-10.html"><a href="mod-10.html#example-with-real-data-2"><i class="fa fa-check"></i><b>10.5.2</b> Example with real data</a></li>
</ul></li>
<li class="chapter" data-level="10.6" data-path="mod-10.html"><a href="mod-10.html#mod-10-logistic"><i class="fa fa-check"></i><b>10.6</b> Logistic regression for binary outcomes</a>
<ul>
<li class="chapter" data-level="10.6.1" data-path="mod-10.html"><a href="mod-10.html#example-with-simulated-data-2"><i class="fa fa-check"></i><b>10.6.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="10.6.2" data-path="mod-10.html"><a href="mod-10.html#example-with-real-data-3"><i class="fa fa-check"></i><b>10.6.2</b> Example with real data</a></li>
<li class="chapter" data-level="10.6.3" data-path="mod-10.html"><a href="mod-10.html#mod-10-auc"><i class="fa fa-check"></i><b>10.6.3</b> Logistic GLM diagnostics: AUC and ROC</a></li>
</ul></li>
<li class="chapter" data-level="10.7" data-path="mod-10.html"><a href="mod-10.html#mod-10-binom"><i class="fa fa-check"></i><b>10.7</b> Binomial GLM for proportional data</a>
<ul>
<li class="chapter" data-level="10.7.1" data-path="mod-10.html"><a href="mod-10.html#binomial-glm"><i class="fa fa-check"></i><b>10.7.1</b> Binomial GLM</a></li>
<li class="chapter" data-level="10.7.2" data-path="mod-10.html"><a href="mod-10.html#example-with-simulated-data-3"><i class="fa fa-check"></i><b>10.7.2</b> Example with simulated data</a></li>
</ul></li>
<li class="chapter" data-level="10.8" data-path="mod-10.html"><a href="mod-10.html#mod-10-gamma"><i class="fa fa-check"></i><b>10.8</b> Gamma models for overdispersed data</a>
<ul>
<li class="chapter" data-level="10.8.1" data-path="mod-10.html"><a href="mod-10.html#example-with-simulated-data-4"><i class="fa fa-check"></i><b>10.8.1</b> Example with simulated data</a></li>
</ul></li>
<li class="chapter" data-level="10.9" data-path="mod-10.html"><a href="mod-10.html#mod-10-beyond"><i class="fa fa-check"></i><b>10.9</b> Beyond GLM: Overview of GAM and GEE</a>
<ul>
<li class="chapter" data-level="10.9.1" data-path="mod-10.html"><a href="mod-10.html#mod-10-gam"><i class="fa fa-check"></i><b>10.9.1</b> Generalized additive models (GAM)</a></li>
<li class="chapter" data-level="10.9.2" data-path="mod-10.html"><a href="mod-10.html#mod-10-gee"><i class="fa fa-check"></i><b>10.9.2</b> Generalized estimating equations (GEE)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="mod-11.html"><a href="mod-11.html"><i class="fa fa-check"></i><b>11</b> Nonlinear models</a>
<ul>
<li class="chapter" data-level="11.1" data-path="mod-11.html"><a href="mod-11.html#background"><i class="fa fa-check"></i><b>11.1</b> Background</a></li>
<li class="chapter" data-level="11.2" data-path="mod-11.html"><a href="mod-11.html#mod-11-intro"><i class="fa fa-check"></i><b>11.2</b> Nonlinear least squares (NLS)</a></li>
<li class="chapter" data-level="11.3" data-path="mod-11.html"><a href="mod-11.html#mod-11-micmen"><i class="fa fa-check"></i><b>11.3</b> Michaelis-Menten curves</a>
<ul>
<li class="chapter" data-level="11.3.1" data-path="mod-11.html"><a href="mod-11.html#mod-11-micmen-sim"><i class="fa fa-check"></i><b>11.3.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="11.3.2" data-path="mod-11.html"><a href="mod-11.html#example-with-real-data-4"><i class="fa fa-check"></i><b>11.3.2</b> Example with real data</a></li>
<li class="chapter" data-level="11.3.3" data-path="mod-11.html"><a href="mod-11.html#alternative-strategies-for-the-analysis"><i class="fa fa-check"></i><b>11.3.3</b> Alternative strategies for the analysis</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="mod-11.html"><a href="mod-11.html#mod-11-grow"><i class="fa fa-check"></i><b>11.4</b> Biological growth curves</a>
<ul>
<li class="chapter" data-level="11.4.1" data-path="mod-11.html"><a href="mod-11.html#gompertz-and-von-bertalanffy-curves"><i class="fa fa-check"></i><b>11.4.1</b> Gompertz and von Bertalanffy curves</a></li>
<li class="chapter" data-level="11.4.2" data-path="mod-11.html"><a href="mod-11.html#example-with-real-data-5"><i class="fa fa-check"></i><b>11.4.2</b> Example with real data</a></li>
</ul></li>
<li class="chapter" data-level="11.5" data-path="mod-11.html"><a href="mod-11.html#dose-response-curves"><i class="fa fa-check"></i><b>11.5</b> Dose response curves</a></li>
<li class="chapter" data-level="11.6" data-path="mod-11.html"><a href="mod-11.html#alternatives-to-nls"><i class="fa fa-check"></i><b>11.6</b> Alternatives to NLS</a>
<ul>
<li class="chapter" data-level="11.6.1" data-path="mod-11.html"><a href="mod-11.html#generalized-nonlinear-models"><i class="fa fa-check"></i><b>11.6.1</b> Generalized nonlinear models</a></li>
<li class="chapter" data-level="11.6.2" data-path="mod-11.html"><a href="mod-11.html#quantile-regression"><i class="fa fa-check"></i><b>11.6.2</b> Quantile regression</a></li>
<li class="chapter" data-level="11.6.3" data-path="mod-11.html"><a href="mod-11.html#mod-11-gam"><i class="fa fa-check"></i><b>11.6.3</b> Generalized additive models (GAM)</a></li>
<li class="chapter" data-level="11.6.4" data-path="mod-11.html"><a href="mod-11.html#classification-and-regression-trees-cart"><i class="fa fa-check"></i><b>11.6.4</b> Classification and regression trees (CART)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="12" data-path="mod-12.html"><a href="mod-12.html"><i class="fa fa-check"></i><b>12</b> Mixed models</a>
<ul>
<li class="chapter" data-level="12.1" data-path="mod-12.html"><a href="mod-12.html#prelude-glm"><i class="fa fa-check"></i><b>12.1</b> Prelude (GLM)</a></li>
<li class="chapter" data-level="12.2" data-path="mod-12.html"><a href="mod-12.html#mod-12-lmm"><i class="fa fa-check"></i><b>12.2</b> Linear mixed models (LMM)</a>
<ul>
<li class="chapter" data-level="12.2.1" data-path="mod-12.html"><a href="mod-12.html#formal-definition-and-example"><i class="fa fa-check"></i><b>12.2.1</b> Formal definition and example</a></li>
<li class="chapter" data-level="12.2.2" data-path="mod-12.html"><a href="mod-12.html#example-with-simulated-data-5"><i class="fa fa-check"></i><b>12.2.2</b> Example with simulated data</a></li>
<li class="chapter" data-level="12.2.3" data-path="mod-12.html"><a href="mod-12.html#p-values-in-lmm"><i class="fa fa-check"></i><b>12.2.3</b> <em>P</em>-values in LMM</a></li>
<li class="chapter" data-level="12.2.4" data-path="mod-12.html"><a href="mod-12.html#specifying-random-effects"><i class="fa fa-check"></i><b>12.2.4</b> Specifying random effects</a></li>
<li class="chapter" data-level="12.2.5" data-path="mod-12.html"><a href="mod-12.html#lmm-example-with-real-data"><i class="fa fa-check"></i><b>12.2.5</b> LMM example with real data</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="mod-12.html"><a href="mod-12.html#mod-12-glmm"><i class="fa fa-check"></i><b>12.3</b> Generalized linear mixed models (GLMM)</a>
<ul>
<li class="chapter" data-level="12.3.1" data-path="mod-12.html"><a href="mod-12.html#definition-1"><i class="fa fa-check"></i><b>12.3.1</b> Definition</a></li>
<li class="chapter" data-level="12.3.2" data-path="mod-12.html"><a href="mod-12.html#glmm-on-simulated-data"><i class="fa fa-check"></i><b>12.3.2</b> GLMM on simulated data</a></li>
<li class="chapter" data-level="12.3.3" data-path="mod-12.html"><a href="mod-12.html#glmm-on-real-data"><i class="fa fa-check"></i><b>12.3.3</b> GLMM on real data</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="mod-12.html"><a href="mod-12.html#mod-12-nlme"><i class="fa fa-check"></i><b>12.4</b> Nonlinear mixed models (NLME)</a>
<ul>
<li class="chapter" data-level="12.4.1" data-path="mod-12.html"><a href="mod-12.html#definition-and-background"><i class="fa fa-check"></i><b>12.4.1</b> Definition and background</a></li>
<li class="chapter" data-level="12.4.2" data-path="mod-12.html"><a href="mod-12.html#nlme-on-simulated-data"><i class="fa fa-check"></i><b>12.4.2</b> NLME on simulated data</a></li>
<li class="chapter" data-level="12.4.3" data-path="mod-12.html"><a href="mod-12.html#nlme-on-real-data"><i class="fa fa-check"></i><b>12.4.3</b> NLME on real data</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="mod-12.html"><a href="mod-12.html#mod-12-gamm"><i class="fa fa-check"></i><b>12.5</b> (Generalized) additive mixed models (AMM/GAMM)</a>
<ul>
<li class="chapter" data-level="12.5.1" data-path="mod-12.html"><a href="mod-12.html#example-gamm-with-real-data"><i class="fa fa-check"></i><b>12.5.1</b> Example GAMM with real data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="13" data-path="mod-13.html"><a href="mod-13.html"><i class="fa fa-check"></i><b>13</b> Multivariate data analysis</a>
<ul>
<li class="chapter" data-level="13.1" data-path="mod-13.html"><a href="mod-13.html#multivariate-data"><i class="fa fa-check"></i><b>13.1</b> Multivariate data</a>
<ul>
<li class="chapter" data-level="13.1.1" data-path="mod-13.html"><a href="mod-13.html#univariate-vs.-multivariate-data"><i class="fa fa-check"></i><b>13.1.1</b> Univariate vs. multivariate data</a></li>
<li class="chapter" data-level="13.1.2" data-path="mod-13.html"><a href="mod-13.html#components-of-multivariate-data"><i class="fa fa-check"></i><b>13.1.2</b> Components of multivariate data</a></li>
</ul></li>
<li class="chapter" data-level="13.2" data-path="mod-13.html"><a href="mod-13.html#mod-13-dist"><i class="fa fa-check"></i><b>13.2</b> Distance metrics: biological (dis)similarity</a>
<ul>
<li class="chapter" data-level="13.2.1" data-path="mod-13.html"><a href="mod-13.html#euclidean-distance"><i class="fa fa-check"></i><b>13.2.1</b> Euclidean distance</a></li>
<li class="chapter" data-level="13.2.2" data-path="mod-13.html"><a href="mod-13.html#bray-curtis-and-other-distance-metrics"><i class="fa fa-check"></i><b>13.2.2</b> Bray-Curtis and other distance metrics</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="mod-13.html"><a href="mod-13.html#clustering"><i class="fa fa-check"></i><b>13.3</b> Clustering</a>
<ul>
<li class="chapter" data-level="13.3.1" data-path="mod-13.html"><a href="mod-13.html#k-means-clustering"><i class="fa fa-check"></i><b>13.3.1</b> <em>K</em>-means clustering</a></li>
<li class="chapter" data-level="13.3.2" data-path="mod-13.html"><a href="mod-13.html#hierarchical-agglomerative-clustering"><i class="fa fa-check"></i><b>13.3.2</b> Hierarchical agglomerative clustering</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="mod-13.html"><a href="mod-13.html#mod-13-sims"><i class="fa fa-check"></i><b>13.4</b> Analyzing dissimilarity</a>
<ul>
<li class="chapter" data-level="13.4.1" data-path="mod-13.html"><a href="mod-13.html#mantel-tests-distance-vs.-distance"><i class="fa fa-check"></i><b>13.4.1</b> Mantel tests: distance vs. distance</a></li>
<li class="chapter" data-level="13.4.2" data-path="mod-13.html"><a href="mod-13.html#comparing-dissimilarity-between-groups"><i class="fa fa-check"></i><b>13.4.2</b> Comparing dissimilarity between groups</a></li>
</ul></li>
<li class="chapter" data-level="13.5" data-path="mod-13.html"><a href="mod-13.html#mod-13-ord"><i class="fa fa-check"></i><b>13.5</b> Ordination</a>
<ul>
<li class="chapter" data-level="13.5.1" data-path="mod-13.html"><a href="mod-13.html#mod-13-pca"><i class="fa fa-check"></i><b>13.5.1</b> Principal components analysis (PCA)</a></li>
<li class="chapter" data-level="13.5.2" data-path="mod-13.html"><a href="mod-13.html#nmds-and-other-ordination-methods"><i class="fa fa-check"></i><b>13.5.2</b> NMDS and other ordination methods</a></li>
<li class="chapter" data-level="13.5.3" data-path="mod-13.html"><a href="mod-13.html#other-ordination-techniques"><i class="fa fa-check"></i><b>13.5.3</b> Other ordination techniques</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="literature-cited.html"><a href="literature-cited.html"><i class="fa fa-check"></i>Literature Cited</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Applied Biological Data Analysis: Statistics and R for Biologists</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="mod-02" class="section level1 hasAnchor" number="2">
<h1><span class="header-section-number">Module 2</span> Introduction to R<a href="mod-02.html#mod-02" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>This module is an introduction to the R program and language. We will begin with a brief discussion of R itself, then run through a typical R work session, and then take a deeper dive into some R basics–including how to get help. Most of the topics in this module are covered in greater detail in <a href="mod-03.html#mod-03">the module on data manipulation</a>. The current module is meant to get you up and running before moving on to more advanced material.</p>
<p>By the end of this module you should be able to:</p>
<ul>
<li>Explain what R is and what it is used for</li>
<li>Explain the object-oriented programming paradigm used by R</li>
<li>Download and install R and RStudio</li>
<li>Write and execute commands in the R console</li>
<li>Describe basic R data types and structures</li>
<li>Manage R code as scripts (.r files)</li>
<li>Manage and use R packages</li>
<li>Identify sources of help for R programming</li>
</ul>
<div id="getting-started-with-r" class="section level2 hasAnchor" number="2.1">
<h2><span class="header-section-number">2.1</span> Getting started with R<a href="mod-02.html#getting-started-with-r" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="what-is-r" class="section level3 hasAnchor" number="2.1.1">
<h3><span class="header-section-number">2.1.1</span> What is R?<a href="mod-02.html#what-is-r" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>R is an open-source language and environment for data analysis, statistics, and data visualization. It is descended from an older language called S, which is the basis for the commercial statistics software S-Plus. R is part of part of the <a href="https://www.gnu.org/software/">GNU free software project</a>. The GNU project is a free and open-source software ecosystem including an operating system and many programs that can replace paid proprietary software. GNU is a recursive acronym that stands for “GNU’s Not Unix”.</p>
<p>R is designed for statistics and data analysis, but it is also a general-purpose programming language. One of the most important features of the R language is that it is <a href="https://en.wikipedia.org/wiki/Object-oriented_programming"><strong>object-oriented</strong></a>. R commands and programs focus on manipulating <strong>objects</strong> in the computer memory. An object is a construct that has a <strong>class</strong>, and can be acted upon by <strong>methods</strong>.</p>
<ul>
<li>A <strong>class</strong> is like a blueprint for an object. The class defines the data format in an object, and how other objects can interact with that object.</li>
<li>An <strong>object</strong> is an instance of a class.</li>
<li>A <strong>method</strong> is a function that works with objects according to their class.</li>
</ul>
<p>For example, a graphics program might have classes such as “circle”, “square”, and “triangle”. Every circle, square, or triangle that the user draws is an instance of one of those classes. This way the user can create different examples of common shape types without having to define the shapes anew every time one is created. The graphics program might have a function for scaling shapes that acts differently depending on whether its input is of the circle or square class.</p>
<p><strong><em>Everything in R is an object of one kind or another!</em></strong> Understanding this idea is the key to R success. We will learn about some of the most important R classes <a href="mod-02.html#mod-02-struct">later</a>.</p>
</div>
<div id="advantages-of-r" class="section level3 hasAnchor" number="2.1.2">
<h3><span class="header-section-number">2.1.2</span> Advantages of R<a href="mod-02.html#advantages-of-r" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>If you are going to choose a tool for statistics in your research, you need to have a good reason. R offers several very good reasons to choose it for your statistical computing needs. The key advantages of R are:</p>
<ul>
<li>R is free!</li>
<li>R is widely used for statistics and data analysis.</li>
<li>R has a large and active community of users and developers. This means that new statistical techniques are often implemented in R before they are implemented in other software.</li>
<li>R is a highly marketable professional skill because it is widely used in academia, government, and industry.</li>
<li>R is open source: you can modify it to suit your needs, and see exactly how it does what it does. The R source code is <a href="https://github.com/wch/r-source">hosted on github</a> (accessed 2021-12-20).</li>
</ul>
</div>
<div id="disadvantages-of-r" class="section level3 hasAnchor" number="2.1.3">
<h3><span class="header-section-number">2.1.3</span> Disadvantages of R<a href="mod-02.html#disadvantages-of-r" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>For all of its advantages, R has some drawbacks. Chief among them:</p>
<ul>
<li>R is free. Unlike some paid and proprietary software, there is no official help line to call if you have issues.</li>
<li>Along the same lines, R is distributed with absolutely no warranty (it says so every time you open R). The accuracy of your analysis depends on the competence of the (unpaid) contributors and the user.</li>
<li>R often requires more coding than SAS or other tools to get a similar amount of output or perform a similar amount of work. The <code>tidyverse</code> ecosystem of packages addresses some of these issues, but not without some costs (see next section).</li>
<li>R is slower than SAS and other tools at handling large datasets because it holds all data and objects in memory (RAM) at once. This limits the performance of your machine if you are doing other tasks.</li>
<li>R is single-threaded by default. This means that R does not handle parallel computation very easily, which could potentially speed up many tasks.</li>
<li>R has a difficult learning curve because it is so programming-focused. There are methods to acheive parallel computing in R, but they can be more trouble than they are worth for most users without some programming background.</li>
<li>Error messages that R gives are often vague and infuriating.</li>
</ul>
</div>
<div id="base-r-and-vs.-tidyverse" class="section level3 hasAnchor" number="2.1.4">
<h3><span class="header-section-number">2.1.4</span> Base R and (vs.?) <code>tidyverse</code><a href="mod-02.html#base-r-and-vs.-tidyverse" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>A colleague of mine who moved from using mostly SAS to using mostly R says that the <code>tidyverse</code> makes R feel more like SAS. The difference he was referring to is the way that SAS functions (“PROCs”, or procedures) tend to automate or abstract away a lot of the low-level functionality and options associated with their methods. Those functions and options are there, but they can be inconvenient to access. The equivalent base R functions, on the other hand, require the user to specify options and manipulate output to a far greater extent. Or, depending on your perspective, they <em>allow</em> the user to specify options and manipulate output. Some people like that, and some people don’t. The amount of interaction and coding in R require to get outputs that other programs like SAS produce automatically is, in my experience, one of the biggest stumbling blocks for new R users.</p>
<p>Because of the amount of work required to get base R to do anything, the degree of automation and abstraction available in the <code>tidyverse</code> packages <span class="citation">(<a href="#ref-wickham2019">Wickham et al. 2019</a>)</span> is very attractive to many R users: they can focus on high-level decisions without having to get into every picayune detail of routine workflows and standard analyses. Tidyverse packages allow for very streamlined and (once you learn it) easy to understand code. These advantages come at a cost: because the functions abstract the finer details away, those details are harder to see. This can be problematic if you need to work more directly with function inputs and outputs, or manipulate options or data in non-standard ways. For example, if a journal editor is adamant about some minute detail of formatting on a figure, it might be much easier to made the changes in base graphics rather than Tidyverse graphics. The abstraction of the low-level workings of an analysis or program can save a lot of time and headache, but can also make it harder to track down or even to notice errors (although to be fair, systematic approaches to data manipulation such as in package <code>dplyr</code> can also <strong>prevent</strong> many errors as well).</p>
<p>Recently Norm Matloff of the University of California, Davis posted an <a href="https://github.com/matloff/TidyverseSkeptic">essay</a> about the advantages of base R compared to the tidyverse<a href="#fn22" class="footnote-ref" id="fnref22"><sup>22</sup></a>. Matloff is the author of several books on R, former editor of the R journal, and a longtime teacher of statistics and computer science. He argues that new R users should focus on learning base R rather than tidyverse. His main points can be summarized as follows:</p>
<ul>
<li>Tidyverse packages and code tend to utilize <a href="https://en.wikipedia.org/wiki/Functional_programming">functional programming</a>, rather than object-oriented programming. Functional programming is more abstract than object-oriented programming, which makes it harder to new users with little to no programming background.</li>
<li>Tidyverse functions eschew use of the <code>$</code> operator, square brackets <code>[]</code>, and base R data structures in general<a href="#fn23" class="footnote-ref" id="fnref23"><sup>23</sup></a>. This means that people who learn only <code>tidyverse</code> are not equipped to take advantage of some of the most fundamental and powerful features of R–including many statistical functions!</li>
<li>Tidyverse code avoids loops, a fundamental programming construct. Loops are, for beginning programmers, easy to “step through” and debug. In <code>tidyverse</code>, operations that would be in loops in base R are “hidden” inside abstract functions. This makes it very hard to track down errors. There are base R functions that can be used to avoid loops as well (e.g., <code>lapply()</code> and its cousins), and they are equally hard for beginners to use.</li>
<li>Tidyverse is intentionally designed to equip users with a small set of R tools. This means that <code>tidyverse</code>-only users are less equipped to deal with new situations.</li>
</ul>
<p>In my own workflow, and in this course, I tend to use base R methods instead of <code>tidyverse</code> for reasons above, as well as a few other resaons that are entirely personal and subjective.</p>
<ol style="list-style-type: decimal">
<li>I learned R before <code>tidyverse</code> became a thing. This isn’t an advantage or disadvantage to either paradigm; it just is. For me, the time and effort costs of switching to <code>tidyverse</code> would likely outweigh the benefits.</li>
<li>I prefer the syntax of base R to the “grammar” of <code>tidyverse</code>. In particular, <code>tidyverse</code> makes extensive use of the pipe operator <code>%&gt;%</code>, which tends to make for obfuscated code. In base R, multi-step operations are laid out on multiple lines, with intermediate objects being available in the workspace. This is not the case in <code>tidyverse</code>.</li>
<li>I prefer to write code with as few dependencies as possible. This makes it less likely that some random update will break my code.</li>
</ol>
<p>The other two reasons are more functional and less subjective:</p>
<ol start="4" style="list-style-type: decimal">
<li>Debugging and error correction is easier in base R because of the <em>lack</em> of abstraction. If every step is coded manually and out in the open, errors are easier to find, isolate, and fix.</li>
<li>Debugging and fixing <code>tidy</code> code often requires knowing base R, but the reverse is never true.</li>
</ol>
<p>The takehome message is that while both base R and <code>tidyverse</code> offer powerful tools for working with and analyzing data, they represent two overlapping but different programming paradigms. Which framework you use depends on what level of abstraction vs. explicitness you are comfortable with in your own code. Base R makes it easier to work flexibly with the different kinds of inputs and outputs, but usually at the cost of a steeper learning curve. <code>tidyverse</code> makes routine tasks more streamlined with a unified syntax and grammar for data manipulation, but at the cost of flexibility. My advice is to learn base R first, then move on to <code>tidyverse</code> if and only if it has a specific tool you need that is not available in base R<a href="#fn24" class="footnote-ref" id="fnref24"><sup>24</sup></a>.</p>
</div>
</div>
<div id="download-and-install-r-and-rstudio" class="section level2 hasAnchor" number="2.2">
<h2><span class="header-section-number">2.2</span> Download and install R (and RStudio)<a href="mod-02.html#download-and-install-r-and-rstudio" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>If you are on a university campus, many on-campus computers may already have R installed on them. This is the bare minimum you need to work with R. Many people also prefer to use another program for writing R code. The most popular is probably <a href="https://www.rstudio.com/products/rstudio/download/#download">RStudio</a>, which is a more user-friendly interface for R<a href="#fn25" class="footnote-ref" id="fnref25"><sup>25</sup></a>. If you are on a PC or Mac, you will need to download and install both R and RStudio. If you are using a Linux machine, you probably already have R because it is included in most distros. You can still install RStudio to make your life easier.</p>
<p>First, install the latest version of R from the <a href="https://cran.r-project.org/mirrors.html">R-project website</a> (accessed 2021-12-20). Select a mirror (download server) from the list. Once you select a mirror, you will be taken to a page where you will select the version of R that corresponds to your operating system. Click on the appropriate link for your operating system and follow the instructions from there to install R. Like many programs you download from the internet, R will come with an installer that will largely take care of everything automatically. Just let the installer run and accept the default options.</p>
<p>Next, install any code editing or IDE you want. RStudio can be installed from <a href="https://www.rstudio.com/products/rstudio/download/#download">here</a> (accessed 2021-12-20). As with R, select the version that corresponds to your operating system. As you did when you installed R above, just let the installer run and accept the default options.</p>
</div>
<div id="using-r" class="section level2 hasAnchor" number="2.3">
<h2><span class="header-section-number">2.3</span> Using R<a href="mod-02.html#using-r" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="using-the-base-r-gui" class="section level3 hasAnchor" number="2.3.1">
<h3><span class="header-section-number">2.3.1</span> Using the base R GUI<a href="mod-02.html#using-the-base-r-gui" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>R has a very basic front end, or <strong>graphical user interface (GUI)</strong>. Most users find it easier to write their R code in another program. Some of these programs, such as RStudio, are <strong>integrated development environments (IDE)</strong>. Others are text editors with programming capabilities. We’ll explore the default R GUI first and then meet some editor options. The basic R GUI is shown below<a href="#fn26" class="footnote-ref" id="fnref26"><sup>26</sup></a>.</p>
<p><img src="02_01.jpg" width="90%" style="display: block; margin: auto;" /></p>
<p>The main window with R is the <strong>R console</strong>. This is the command line interface to R. Some important features of the console are:</p>
<ol style="list-style-type: decimal">
<li>Command prompt: The <code>&gt;</code> symbol signifies the beginning of a command. You can enter commands here.</li>
<li>Continue prompt: <code>+</code> signifies the continuation of a previous command. If you have a <code>+</code> at the beginning of your line, you can get back to a new line using ESCAPE. Note that this will delete the currently incomplete line or command.</li>
<li>The ENTER key executes the current command. This could mean the current line, or multiple lines going back to the most recent <code>&gt;</code>.</li>
<li><code>[1]</code> at the beginning of results: because your result is an object (with length <span class="math inline">\(\ge\)</span> 1). The very first value (“element”) in the result is element 1.</li>
<li>Up arrow key: cycles through previously entered commands. You can see many previous commands at once by using the command <code>history()</code>.</li>
</ol>
<p>Other important features of the R GUI are found in the menu bar:</p>
<ul>
<li>File
<ul>
<li>New script: opens a new script window within the GUI.</li>
<li>Ability to open saved scripts or load previous workspaces</li>
</ul></li>
<li>Edit
<ul>
<li>Copy, paste, etc.: very useful!</li>
<li>Data editor: opens a rudimentary spreadsheet-like viewer for datasets that can be used to make edits. Not recommended for routine use.</li>
<li>GUI preferences: options to change colors, fonts, font sizes, etc., in the GUI. To make changes permanent you may need to save a new <code>.RConsole</code> file in your home directory.</li>
</ul></li>
<li>View</li>
<li>Misc
<ul>
<li>List objects: lists all objects in workspace. Equivalent to command <code>ls()</code>.</li>
<li>Remove all objects: removes all objects form workspace. Equivalent to command <code>rm(list=ls(all=TRUE))</code>.</li>
<li>List search path: lists currently attached packages. Equivalent to command <code>search()</code>.</li>
</ul></li>
<li>Packages: functions to download and install packages.</li>
<li>Windows: options to change how windows are displayed within GUI.</li>
<li>Help: options and resources for getting help, including several free R manuals in PDF format.</li>
</ul>
</div>
<div id="using-r-in-rstudio" class="section level3 hasAnchor" number="2.3.2">
<h3><span class="header-section-number">2.3.2</span> Using R in RStudio<a href="mod-02.html#using-r-in-rstudio" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Although you can type R code directly into the console, or type in script windows within the base R GUI, it is usually easiest to use a separate program for writing code. The two best, in my experience, are RStudio and Notepad++ (see below). <strong>RStudio</strong> is an IDE designed to manage R projects, and has many useful features. Unlike some other options, RStudio has R built in so you can write code and execute it within the same program. RStudio can be used on Windows, Mac, and Linux operating systems. RStudio also has built-in support for utilities like RMarkdown.</p>
<p>Once you have RStudio installed, open it and you will see a screen like this<a href="#fn27" class="footnote-ref" id="fnref27"><sup>27</sup></a></p>
<p><img src="02_02.jpg" width="90%" style="display: block; margin: auto;" /></p>
<p>RStudio has an entire environment for working with R in one window. Each of the panels performs different functions. Starting at the bottom left and going clockwise, the panels are:</p>
<ul>
<li>Console: command line interface for R. You can type commands here and execute them by pressing ENTER.</li>
<li>Script: a space where you can type code and later execute it by either clicking “Run” or pressing CRTL-ENTER (CMD-ENTER on a Mac). Usually people write code in the script window rather than typing it directly into the console.</li>
<li>Environment: this window lists all of the data objects that you have loaded.</li>
<li>Files: this window lists all of the files available in your R home directory. You can also navigate to other folders on your machine and select files to open.</li>
</ul>
<p>The bottom right panel also has a tab called “Plots”. This is where plots that you make can be seen. Sometimes you will need to resize this panel to see the plots clearly. All of the panels can be resized by clicking and dragging the borders between them.</p>
</div>
<div id="using-r-with-other-programs" class="section level3 hasAnchor" number="2.3.3">
<h3><span class="header-section-number">2.3.3</span> Using R with other programs<a href="mod-02.html#using-r-with-other-programs" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>Notepad++</strong> is an open-source code editor that is <a href="https://notepad-plus-plus.org/downloads/">available on Windows</a> (accessed 2021-12-20). Notepad++ is a more generic source code editor with built-in R support and many programming features. It includes programming support for R as well as other languages used by biologists, such as Python, C++, Visual Basic, and LaTeX. The key disadvantage of Notepad++ is that it does not connect to R, so you must copy and paste your code to an R console. Another disadvantage is that it is only available for Windows. The image below shows a script in Notepad++,</p>
<p><img src="02_03.jpg" width="90%" style="display: block; margin: auto;" /></p>
<p><strong>emacs</strong> is a GNU program that is an extremely powerful text editor, code editor, and pretty much everything else. With its extension ess or “emacs speaks statistics”, it can also be an R IDE. Be forewarned, getting emacs and ess to work on Windows can be a major pain.</p>
<p>Other good options include VS Code, Tinn-R, Jupyter notebooks, and many more. All of these options are subject to irrationally strong personal and organizational preferences. Some people even use different environments for different purposes. For example, I use Notepad++ in my research for writing and editing R code. But, I use RStudio when working with markdown because RStudio has a lot of facilities that make that easier (e.g., I used RStudio to build this website). Try a few R interfaces and use what works best for you. Or, use the interface that your manager tells you to use.</p>
<p>Whatever you do, you should not use a word processor (e.g., Microsoft Word) to write and edit R code. The reason that you should not use Word for code editing is because Word documents include lots of invisible formatting that can cause issues in R. Worse, Word autocorrects some characters in counter-productive ways. The most problematic are automatic capitalization (because R is case sensitive) and quotation marks. Word autocorrects <code>""</code> to “” as you type, which makes for nice looking text. However, the curved quotation marks (““) are not read as valid quotation marks by R. Only”straight quotes” (<code>""</code>) are recognized by R.</p>
</div>
</div>
<div id="mod-02-first" class="section level2 hasAnchor" number="2.4">
<h2><span class="header-section-number">2.4</span> A first R session<a href="mod-02.html#mod-02-first" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>One of the best ways to get a feel for R is to step through a typical R workflow. We are going to analyze a classic dataset from evolutionary biology that contains data on the body size and antler size of many species of deer (family Cervidae). The data were collected by Stephen Jay Gould <span class="citation">(<a href="#ref-gould1974">Gould 1974</a>)</span> to investigate the reason for the massive antler size in an extinct mammal, the Irish elk (<em>Megaloceros giganteus</em>), seen below<a href="#fn28" class="footnote-ref" id="fnref28"><sup>28</sup></a>. Gould was interested in whether the massive antlers might have resulted from runaway sexual selection or were simply the result of allometric growth. The Irish elk lived across northern Eurasia during the Pleistocene epoch, until about 7700 years ago. Adults stood around 2.1 m high at the shoulder and males bore antlers that could reach up to 3.65 m (<span class="math inline">\(\approx\)</span> 12 ft) across. Body mass of adult males is estimated to have ranged from 450 to 700 kg (990 to 1540 lb). Irish elk were on average larger than the largest extant cervid, the moose (<em>Alces alces</em>).</p>
<p><img src="02_04.jpg" width="70%" style="display: block; margin: auto;" /></p>
<div id="import-data" class="section level3 hasAnchor" number="2.4.1">
<h3><span class="header-section-number">2.4.1</span> Import data<a href="mod-02.html#import-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>First, load the data into R. The data are contained in the text file <a href="https://greenquanteco.github.io/elk_data_2021-01-21.csv">elk_data_2021-01-21.csv</a>. Download this file and save it on your computer. If you want to save some time later, save the data file in your <strong>R home directory</strong>. This is the default folder where R and RStudio will look for files. The R home directory depends on your operating system:</p>
<ul>
<li><strong>Windows users:</strong> the home directory is your “Documents” folder. The address of this folder is “<strong>C:/Users/username/Documents</strong>”, where “<strong>username</strong>” is your Windows logon name. You can get to your Documents folder by opening Windows Explorer (<strong>Win-E</strong>) and double-clicking “<strong>Documents</strong>”.
<ul>
<li>If you use OneDrive, make sure that you go to the local Documents folder, which might be different than the Documents folder in your OneDrive (R can access OneDrive, but by default it goes to the local Documents folder on your C drive).</li>
</ul></li>
<li><strong>Mac users:</strong> the home directory is your <strong>home folder</strong>. The address of this folder is “<strong>/Users/username</strong>” (where username is your username) or “<strong>~/</strong>”. You can open your home folder by pressing <strong>CMD-SHIFT-H</strong>, or by using the <strong>Go</strong> pulldown menu from the menu bar.
<ul>
<li>Note for Mac users: When you download a .csv file in MacOS, your computer might open it in a program called “Numbers”. This is a Mac-specific spreadsheet program. When you close the program it is possible to inadvertently save the data file as a new file in .numbers format, which is not readable by R. If you do this, your datafile will become “<strong>elk_data_2021-01-21.numbers</strong>”, and you will not be able to open it in R or RStudio. <strong><em>.numbers is a different file format than .csv</em></strong>. A .numbers file has the same file name, but it has a different file extension and so it is not the same as a .csv<a href="#fn29" class="footnote-ref" id="fnref29"><sup>29</sup></a>.</li>
</ul></li>
</ul>
<p>Once you save the data file on your computer, you can import it to RStudio in several ways. Notice that in both methods shown here, nothing is printed to the console after a successful import. Instead, the function reads the CSV file and saves the data to an object that we are calling <code>dat</code>. This short for “data”, but we don’t want to call our dataset “data” because <code>data()</code> is already the name of a function in R…and we don’t want to overwrite that function!</p>
<div id="method-1-import-in-rstudio-using-fileimport-dataset" class="section level4 hasAnchor" number="2.4.1.1">
<h4><span class="header-section-number">2.4.1.1</span> Method 1: Import in RStudio using File–Import Dataset<a href="mod-02.html#method-1-import-in-rstudio-using-fileimport-dataset" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>In RStudio, click on <strong>File–Import Dataset–From Text (base)</strong>.</p>
<p><img src="02_05.jpg" width="90%" style="display: block; margin: auto;" /></p>
<p>Navigate to the data file just as you would if opening a file in any other program. Click Open.</p>
<p><img src="02_06.jpg" width="90%" style="display: block; margin: auto;" /></p>
<p>In the dialog box that pops up, change the name of the incoming dataset to “dat”. This is what we will call the dataset when working with it in R. Make sure that the “Strings as factors” box is NOT checked.</p>
<p><img src="02_07.jpg" width="90%" style="display: block; margin: auto;" /></p>
<p>Click Import. RStudio will import the data for you and save it in the workspace as an object called <code>dat</code>. Notice that a command like <code>dat &lt;- read.csv("~/elk_data_2021-01-21.csv")</code> was run in the console. This command is what RStudio used internally to import the data. If you want to know more, look at method 2 below.</p>
<p>After you click Import, RStudio will import the data and display it for you. The image below shows what that looks like. Notice that RStudio automatically generated the code to import the data and executed it (red arrow); that the dataset is listed in the “Environment” box in the top right (blue arrow), and the dataset is displayed in the top left (green arrow).</p>
<p><img src="02_08.jpg" width="90%" style="display: block; margin: auto;" /></p>
<p>Once the dataset is imported, you can view it in the console by typing its name and pressing ENTER. Each method above saved the dataset to an object named <code>dat</code>. Obviously, this is short for “data”<a href="#fn30" class="footnote-ref" id="fnref30"><sup>30</sup></a>. However, we didn’t call our dataset “data” because <code>data()</code> is already the name of an R function and we don’t want to accidentally overwrite that function.</p>
</div>
<div id="method-2-import-from-command-line" class="section level4 hasAnchor" number="2.4.1.2">
<h4><span class="header-section-number">2.4.1.2</span> Method 2: Import from command line<a href="mod-02.html#method-2-import-from-command-line" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>You can also import files from the command line directly. This option is preferred by users who need to import files automatically, or who aren’t using RStudio. To use this method, you need to know the name of the folder where the data file is stored.</p>
<p>If the data file is in your home directory (see previous section), the command is simple:</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="mod-02.html#cb35-1" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">&quot;elk_data_2021-01-21.csv&quot;</span>)</span></code></pre></div>
<p>The command above will look in your home directory for a file with the filename you specify. The filename must match <strong><em>exactly</em></strong>, including capitalization. For convenience I usually split the command up into two, with one specifying the name and the other importing the file:</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="mod-02.html#cb36-1" tabindex="-1"></a>dat.name <span class="ot">&lt;-</span> <span class="st">&quot;elk_data_2021-01-21.csv&quot;</span></span>
<span id="cb36-2"><a href="mod-02.html#cb36-2" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(dat.name)</span></code></pre></div>
<p>If the data file is NOT in your home directory, you need to tell R which folder to look in. In the example below, I put the file on my Windows desktop<a href="#fn31" class="footnote-ref" id="fnref31"><sup>31</sup></a> (<strong>C:/Users/ngreen62/Desktop</strong>). The Mac desktop is <strong>/Users/username/Desktop</strong>, where <strong>username</strong> is your username, or just <strong>~/Desktop</strong>. This folder address is provided to R as a character string, and that string is combined with the file name using the <code>paste()</code> command. Again, this command is case-sensitive.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="mod-02.html#cb37-1" tabindex="-1"></a>use.fold <span class="ot">&lt;-</span> <span class="st">&quot;C:/Users/ngreen62/Desktop&quot;</span></span>
<span id="cb37-2"><a href="mod-02.html#cb37-2" tabindex="-1"></a>dat.name <span class="ot">&lt;-</span> <span class="st">&quot;elk_data_2021-01-21.csv&quot;</span></span>
<span id="cb37-3"><a href="mod-02.html#cb37-3" tabindex="-1"></a>dat <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="fu">paste</span>(use.fold, dat.name, <span class="at">sep=</span><span class="st">&quot;/&quot;</span>))</span></code></pre></div>
<p>Notice that the folder address above uses front-slashes (<code>/</code>). Windows uses back-slashes (“\”) in folder addresses, while R requires forward slashes (“/”). Mac and Linux already use front-slashes. If you are a Windows user, you will need to change the back-slashes to front-slashes in the folder address.</p>
<p>Once you have imported the data, you can type the name of the dataset into the console and press ENTER to view it:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="mod-02.html#cb38-1" tabindex="-1"></a>dat</span></code></pre></div>
<pre><code>##    height antler     species
## 1    69.3   24.8     unknown
## 2    69.3   34.5     unknown
## 3    87.7   53.5     unknown
## 4    99.8   58.3     unknown
## 5    69.3   61.2     unknown
## 6    84.1   67.3     unknown
## 7    89.4   72.2     unknown
## 8    91.9   81.5     unknown
## 9   116.6   86.6     unknown
## 10  104.7   91.4     unknown
## 11  129.9   99.3     unknown
## 12   94.6  101.5     unknown
## 13  115.5  107.8     unknown
## 14  138.4  115.7     unknown
## 15  138.4  123.3     unknown
## 16  127.5  128.5     unknown
## 17  175.1  131.4       alces
## 18  164.3  162.6     unknown
## 19  204.5  197.5       alces
## 20  183.6  239.1 megaloceros</code></pre>
</div>
</div>
<div id="explore-and-visualize-data" class="section level3 hasAnchor" number="2.4.2">
<h3><span class="header-section-number">2.4.2</span> Explore and visualize data<a href="mod-02.html#explore-and-visualize-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Before we do anything, we should look at the dataset. We have two continuous variables, <code>height</code> and <code>antler</code>, so the natural way to examine their relationship is with a scatterplot. Scatterplots are made by the <code>plot()</code> function. By default, the first argument to <code>plot()</code> contains the <em>x</em> values, and the second contains the <em>y</em> values. The values are accessed from the data frame <code>dat</code> using the <code>$</code> operator. The command below also uses the arguments <code>xlab</code> and <code>ylab</code> to set <em>x</em>-axis and <em>y</em>-axis labels. Notice that R will set the limits of the graph according to the range of each variable; we’ll go over how to change those limits later on.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="mod-02.html#cb40-1" tabindex="-1"></a><span class="fu">plot</span>(dat<span class="sc">$</span>height, dat<span class="sc">$</span>antler,</span>
<span id="cb40-2"><a href="mod-02.html#cb40-2" tabindex="-1"></a>    <span class="at">xlab=</span><span class="st">&quot;Shoulder height (cm)&quot;</span>,</span>
<span id="cb40-3"><a href="mod-02.html#cb40-3" tabindex="-1"></a>    <span class="at">ylab=</span><span class="st">&quot;Antler length (cm)&quot;</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-59-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>We might also be interested in how the data are distributed. This will be important later because the data need to meet particular assumptions for our statistical analysis. The simplest way to view a data distribution is with a histogram, <code>hist()</code>.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="mod-02.html#cb41-1" tabindex="-1"></a><span class="fu">hist</span>(dat<span class="sc">$</span>antler, <span class="at">xlab=</span><span class="st">&quot;Antler length (cm)&quot;</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-60-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>We could also try a <strong>probability density function (PDF)</strong> plot, or plotting the <strong>empirical cumulative distribution function (ECDF)</strong>. Histograms, PDF plots, and ECDF plots are just different ways of <a href="mod-04.html#mod-04-vis">displaying how the data are spread out</a>. In the figure below, we use the function <code>par()</code> to set some graphical parameters (including a 1 row <span class="math inline">\(\times\)</span> 3 column layout using argument <code>mfrow</code>). We also saved the text string “Antler length (cm)” as an object (<code>use.xlab</code>), and then used it in each of the plotting commands so we didn’t have to type the same thing over and over.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="mod-02.html#cb42-1" tabindex="-1"></a>use.xlab <span class="ot">&lt;-</span> <span class="st">&quot;Antler length (cm)&quot;</span></span>
<span id="cb42-2"><a href="mod-02.html#cb42-2" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>))</span>
<span id="cb42-3"><a href="mod-02.html#cb42-3" tabindex="-1"></a><span class="fu">hist</span>(dat<span class="sc">$</span>antler, <span class="at">xlab=</span>use.xlab)</span>
<span id="cb42-4"><a href="mod-02.html#cb42-4" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(dat<span class="sc">$</span>antler, <span class="at">from=</span><span class="dv">0</span>), <span class="at">xlab=</span>use.xlab)</span>
<span id="cb42-5"><a href="mod-02.html#cb42-5" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">ecdf</span>(dat<span class="sc">$</span>antler), <span class="at">xlab=</span>use.xlab)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-61-1.png" width="864" style="display: block; margin: auto;" /></p>
<p>All three plots above show that the antler lengths probably follow a <a href="mod-05.html#mod-05-norm"><strong>normal distribution</strong></a>. Notice how the PDF plot (center) looks like a very smoothed version of the histogram. This is because it basically is. A histogram shows how many observations fall into discrete ranges, or bins. The PDF of a distribution shows how likely every possible value is relative to other values. That is, the PDF is basically the heights of a histogram with bin widths = 0. The rightmost plot, the ECDF, shows what percentage of observations in a distribution are equal to or less than each value (this is the “cumulative” part of ECDF). If we wanted, we could estimate the actual CDF of the distribution. The PDF and CDF are intimately related because the PDF is the derivative of the CDF (or, the CDF is the integral of the PDF). This is why the <em>y</em>-axis values on a PDF plot look so weird: they are the instantaneous rate of change in the CDF at each value.</p>
<p>Don’t worry if that doesn’t make a lot of sense right now. We will explore data distributions some more <a href="mod-05.html#mod-05">later in the course</a>, both in practical terms and in the language of PDFs and CDFs](#mod-05-condists).</p>
</div>
<div id="transform-data" class="section level3 hasAnchor" number="2.4.3">
<h3><span class="header-section-number">2.4.3</span> Transform data<a href="mod-02.html#transform-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The plots above show that the antler lengths probably follow a normal distribution, which is convenient for statistical analysis. In fact, many statistical methods were developed to work only with normally distributed data! But, there is a problem. The normal distribution can take on any real value. In statistical jargon, we say that the normal distribution is supported on the interval [<span class="math inline">\(-\infty\)</span>, <span class="math inline">\(+\infty\)</span>]. A normally distributed variable can thus take on negative values. Does this make sense for antler lengths? Or for any kind of length? Of course not.</p>
<p>One way to avoid the awkwardness of a statistical model that predicts negative lengths is to log-transform the variables. This ensures that any value predicted by the model must be positive. We are actually going to log-transform both the response and predictor variable for reasons that will become clear in the next step. In R this can be done with the <code>log()</code> command:</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="mod-02.html#cb43-1" tabindex="-1"></a>dat<span class="sc">$</span>logy <span class="ot">&lt;-</span> <span class="fu">log</span>(dat<span class="sc">$</span>antler)</span>
<span id="cb43-2"><a href="mod-02.html#cb43-2" tabindex="-1"></a>dat<span class="sc">$</span>logx <span class="ot">&lt;-</span> <span class="fu">log</span>(dat<span class="sc">$</span>height)</span></code></pre></div>
<p>Notice that we used the natural logarithm (base <em>e</em>) rather than log<sub>10</sub>. This will be important later when we plot the model and its predictions.</p>
</div>
<div id="analyze-data" class="section level3 hasAnchor" number="2.4.4">
<h3><span class="header-section-number">2.4.4</span> Analyze data<a href="mod-02.html#analyze-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now that we have log-transformed the data, we should plot them to help us decide on an appropriate statistical model. We can make a new version of the scatterplot from earlier. Notice that the axes are now on a different scale, the logarithm of the original values. So, the <em>x</em>-coordinates around 4.9 on the log scale correspond to shoulder heights of <span class="math inline">\(e^{4.9}\)</span>, or about 134.3 cm.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="mod-02.html#cb44-1" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>)) <span class="co"># reset plot layout</span></span>
<span id="cb44-2"><a href="mod-02.html#cb44-2" tabindex="-1"></a><span class="fu">plot</span>(dat<span class="sc">$</span>logx, dat<span class="sc">$</span>logy,</span>
<span id="cb44-3"><a href="mod-02.html#cb44-3" tabindex="-1"></a>    <span class="at">xlab=</span><span class="st">&quot;Log Shoulder height (cm)&quot;</span>,</span>
<span id="cb44-4"><a href="mod-02.html#cb44-4" tabindex="-1"></a>    <span class="at">ylab=</span><span class="st">&quot;Log Antler length (cm)&quot;</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-63-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>We could also have acheived this with by setting the <code>log</code> argument in <code>plot()</code>. Sometimes this method is more convenient than plotting the transformed values because the axes are on the log scale, but show the original values.</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="mod-02.html#cb45-1" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>)) <span class="co"># reset plot layout</span></span>
<span id="cb45-2"><a href="mod-02.html#cb45-2" tabindex="-1"></a><span class="fu">plot</span>(dat<span class="sc">$</span>height, dat<span class="sc">$</span>antler,</span>
<span id="cb45-3"><a href="mod-02.html#cb45-3" tabindex="-1"></a>    <span class="at">xlab=</span><span class="st">&quot;Log Shoulder height (cm)&quot;</span>,</span>
<span id="cb45-4"><a href="mod-02.html#cb45-4" tabindex="-1"></a>    <span class="at">ylab=</span><span class="st">&quot;Log Antler length (cm)&quot;</span>,</span>
<span id="cb45-5"><a href="mod-02.html#cb45-5" tabindex="-1"></a>    <span class="at">log=</span><span class="st">&quot;xy&quot;</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-64-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>It looks like antler length and shoulder height have a linear relationship on the log scale. That is exactly what we are going to fit, but we need to understand what this represents. The linear model we can fit with the log-transformed data is:</p>
<p><span class="math display">\[\log{\left(Y\right)=\beta_0+\beta_1\log{\left(X\right)}+\varepsilon}\]</span></p>
<p>In this equation:</p>
<ul>
<li><em>Y</em> is the response or dependent variable (antler length)</li>
<li><em>X</em> is the explanatory or independent variable (shoulder height)</li>
<li><span class="math inline">\(\beta_0\)</span> is the <em>y</em>-intercept (i.e., the value of <span class="math inline">\(log(Y)\)</span> when <span class="math inline">\(log(X) = 0\)</span>). Called “beta zero” or “beta naught”.</li>
<li><span class="math inline">\(\beta_1\)</span> is the slope or regression coefficient (i.e., the change in <span class="math inline">\(log(Y)\)</span> per unit change in <span class="math inline">\(log(X)\)</span>). If <span class="math inline">\(log(X)\)</span> increases by 1, then <span class="math inline">\(log(Y)\)</span> increases by <span class="math inline">\(\beta_1\)</span>. Called “beta one”.</li>
<li><span class="math inline">\(\varepsilon\)</span> is a random error term that describes residual variation not explained by the model. Called “epsilon”. In a linear regression model, residuals are identically and independently distributed (<em>i.i.d.</em>) according to a normal distribution with mean 0 and variance <span class="math inline">\(\sigma^2\)</span>. Almost all statistical models assume <em>i.i.d.</em> residuals, which means that errors all come from the same distribution, and are completely independent of each other.</li>
</ul>
<p>If we wanted to estimate actual antler lengths rather than the logarithm of antler lengths, we have to exponentiate both sides. Ignoring the residuals term for the moment, this gives us:</p>
<p><span class="math display">\[Y=e^{\beta_0}X^{\beta_1}\]</span></p>
<p>Which is more commonly written as a <strong>power law</strong>. Power laws are very common in anatomy and morphology. The equation below is the usual form of a power law, where the coefficient <em>a</em> = <span class="math inline">\(e^{\beta_0}\)</span>.</p>
<p><span class="math display">\[Y=aX^b\]</span></p>
<p>A power law relates <em>proportional changes in y</em> to <em>proportional changes in x</em>. This is useful in biology because many things are proportional to each other.</p>
<p>The linear model is fit using the <code>lm()</code> function. Notice that the model terms are specified as a “formula”. The response variable is on the left, then a <code>~</code>, then the predictor variable (or variables) on the right. The model is assigned (<code>&lt;-</code>) to an object called <code>mod1</code>, short for “model 1”.</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="mod-02.html#cb46-1" tabindex="-1"></a>mod1 <span class="ot">&lt;-</span> <span class="fu">lm</span>(logy<span class="sc">~</span>logx, <span class="at">data=</span>dat)</span></code></pre></div>
<p>We can view the results using the <code>summary()</code> command on the output:</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="mod-02.html#cb47-1" tabindex="-1"></a><span class="fu">summary</span>(mod1)</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = logy ~ logx, data = dat)
## 
## Residuals:
##      Min       1Q   Median       3Q      Max 
## -0.55752 -0.12729  0.00769  0.15201  0.38060 
## 
## Coefficients:
##             Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept)  -2.6476     0.7918  -3.344  0.00361 ** 
## logx          1.5138     0.1675   9.037 4.14e-08 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 0.24 on 18 degrees of freedom
## Multiple R-squared:  0.8194, Adjusted R-squared:  0.8094 
## F-statistic: 81.67 on 1 and 18 DF,  p-value: 4.141e-08</code></pre>
<p>This tells us that the intercept and slope are statistically significant (<em>P</em> &lt; 0.05 for both terms), and that the model explains about 81% of variation in antler length (adjusted <em>R</em><sup>2</sup> = 0.8094). That’s pretty good.</p>
<p>What if we had not transformed the data? Then we would have had to fit the power law directly. This is possible in R using the <code>nls()</code> function (nonlinear least squares), but this route is usually harder than using <code>lm()</code> on transformed data. Here is the equivalent model fit using the nonlinear model function <code>nls()</code>:</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="mod-02.html#cb49-1" tabindex="-1"></a>mod2 <span class="ot">&lt;-</span> <span class="fu">nls</span>(antler<span class="sc">~</span>a<span class="sc">*</span>height<span class="sc">^</span>b, <span class="at">data=</span>dat,</span>
<span id="cb49-2"><a href="mod-02.html#cb49-2" tabindex="-1"></a>            <span class="at">start=</span><span class="fu">list</span>(<span class="at">a=</span><span class="fu">exp</span>(<span class="sc">-</span><span class="fl">2.6</span>), <span class="at">b=</span><span class="fl">1.51</span>))</span>
<span id="cb49-3"><a href="mod-02.html#cb49-3" tabindex="-1"></a><span class="fu">summary</span>(mod2)</span></code></pre></div>
<pre><code>## 
## Formula: antler ~ a * height^b
## 
## Parameters:
##   Estimate Std. Error t value Pr(&gt;|t|)    
## a  0.13219    0.09861   1.341    0.197    
## b  1.38863    0.14929   9.302 2.69e-08 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 21.49 on 18 degrees of freedom
## 
## Number of iterations to convergence: 5 
## Achieved convergence tolerance: 8.23e-07</code></pre>
<p><a href="mod-10.html#mod-10">Fitting nonlinear models in R</a> can be tricky because you usually need to supply starting values for the algorithm to try. In the example above I used the estimates from the linearized version of the model as starting values. If the starting values are too far away from the “correct” values then the model may not converge, leaving you without a model fit. Even if the model does converge, there is no guarantee that the fitted parameters are correct because there might be multiple stable solutions. Furthermore, some of the common postprocessing done on fitted models such as multiple inference, model comparison, prediction, and error propagation are harder to do in R with nonlinear models (<code>nls()</code> outputs) than with linear models (<code>lm()</code> outputs). We’ll explore how to do this <a href="#mod-10-intro">later in the course</a>, but for now we will stick with the linearized fit.</p>
</div>
<div id="mod-01-elkfig" class="section level3 hasAnchor" number="2.4.5">
<h3><span class="header-section-number">2.4.5</span> Write out results<a href="mod-02.html#mod-01-elkfig" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now that we have successfully fit a statistical model, we need to report our findings. Reporting model coefficients is easy enough: just put them into a table.</p>
<p><strong>Table 1. Log-transformed antler length (cm) varied as a linear function of log-transformed shoulder height (cm). Model <em>R</em><sup>2</sup> = 0.8094. Parameters shown are estimate</strong> <span class="math inline">\(\pm\)</span> standard error (SE); <em>t</em> is <em>t</em>-statistic, and <em>P</em> is <em>P</em>-value.</p>
<table>
<thead>
<tr class="header">
<th>Parameter</th>
<th align="right">Estimate <span class="math inline">\(\pm\)</span> SE</th>
<th align="right"><em>t</em></th>
<th align="right"><em>P</em></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><span class="math inline">\(Log\left(\beta_0\right)\)</span></td>
<td align="right">-2.65 <span class="math inline">\(\pm\)</span> 0.79</td>
<td align="right">-3.34</td>
<td align="right">0.0036</td>
</tr>
<tr class="even">
<td><span class="math inline">\(\beta_1\)</span></td>
<td align="right">1.51 <span class="math inline">\(\pm\)</span> 0.17</td>
<td align="right">9.04</td>
<td align="right">&lt;0.0001</td>
</tr>
</tbody>
</table>
<p>We might also want to present the fitted model in context with the original data. In R this is done using <strong>predicted values</strong> and their associated uncertainties. Instead of presenting predicted values for the original data, it is common practice to present predicted values for a smooth set of predictor values within the domain of the original data. We can see the minimum and maximum values of the predictor with <code>range()</code>. Then, we use <code>seq()</code> to generate a regular sequence between those limits. Finally, calculate the predictions using <code>predict()</code>, with argument <code>se.fit</code> to get the uncertainty associated with the predictions.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="mod-02.html#cb51-1" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb51-2"><a href="mod-02.html#cb51-2" tabindex="-1"></a>new.x <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="at">from=</span><span class="fu">min</span>(dat<span class="sc">$</span>logx), <span class="at">to=</span><span class="fu">max</span>(dat<span class="sc">$</span>logx), <span class="at">length=</span>n)</span>
<span id="cb51-3"><a href="mod-02.html#cb51-3" tabindex="-1"></a>new.x2 <span class="ot">&lt;-</span> <span class="fu">exp</span>(new.x) <span class="co"># needed for plot later</span></span>
<span id="cb51-4"><a href="mod-02.html#cb51-4" tabindex="-1"></a>pred <span class="ot">&lt;-</span> <span class="fu">predict</span>(mod1,</span>
<span id="cb51-5"><a href="mod-02.html#cb51-5" tabindex="-1"></a>                <span class="at">newdata=</span><span class="fu">data.frame</span>(<span class="at">logx=</span>new.x),</span>
<span id="cb51-6"><a href="mod-02.html#cb51-6" tabindex="-1"></a>                <span class="at">se.fit=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p>The last piece we will need is the 95% <strong>confidence limits</strong> of the predictions, also referred to as the <strong>confidence interval (CI)</strong>. We can calculate these from the predicted means and SE in object <code>pred</code>. Because the predictions are on the logarithmic scale, we will back-transform the predictions. We can get the back-transformed predictions by exponentiating the predicted mean, or by requesting quantiles from the lognormal distribution with function <code>qlnorm()</code>. To get the 95% CI, we request the 2.5% and 97.5% percentiles (i.e., 0.025 and 0.975 quantiles) because the interval between them covers 95% of the distribution.</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="mod-02.html#cb52-1" tabindex="-1"></a>mn <span class="ot">&lt;-</span> <span class="fu">exp</span>(pred<span class="sc">$</span>fit)</span>
<span id="cb52-2"><a href="mod-02.html#cb52-2" tabindex="-1"></a>lo <span class="ot">&lt;-</span> <span class="fu">qlnorm</span>(<span class="fl">0.025</span>, pred<span class="sc">$</span>fit, pred<span class="sc">$</span>se.fit)</span>
<span id="cb52-3"><a href="mod-02.html#cb52-3" tabindex="-1"></a>hi <span class="ot">&lt;-</span> <span class="fu">qlnorm</span>(<span class="fl">0.975</span>, pred<span class="sc">$</span>fit, pred<span class="sc">$</span>se.fit)</span>
<span id="cb52-4"><a href="mod-02.html#cb52-4" tabindex="-1"></a></span>
<span id="cb52-5"><a href="mod-02.html#cb52-5" tabindex="-1"></a><span class="co"># not run: alternative (sometimes more reliable)</span></span>
<span id="cb52-6"><a href="mod-02.html#cb52-6" tabindex="-1"></a><span class="co">#          way to get predicted mean/median</span></span>
<span id="cb52-7"><a href="mod-02.html#cb52-7" tabindex="-1"></a><span class="co"># mn &lt;- qlnorm(0.5, pred$fit, pred$se.fit)</span></span></code></pre></div>
<p>Now we have all the pieces we need to put together a polished, manuscript-quality figure with the original data and the model predictions. We will use the <code>par()</code> command again to set some additional graphics options.</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="mod-02.html#cb53-1" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>),              <span class="co"># 1 x 1 layout</span></span>
<span id="cb53-2"><a href="mod-02.html#cb53-2" tabindex="-1"></a>    <span class="at">mar=</span><span class="fu">c</span>(<span class="fl">4.1</span>, <span class="fl">4.1</span>, <span class="fl">1.1</span>, <span class="fl">1.1</span>), <span class="co"># margin sizes (clockwise from bottom)</span></span>
<span id="cb53-3"><a href="mod-02.html#cb53-3" tabindex="-1"></a>    <span class="at">bty=</span><span class="st">&quot;n&quot;</span>,                   <span class="co"># no box around plot</span></span>
<span id="cb53-4"><a href="mod-02.html#cb53-4" tabindex="-1"></a>    <span class="at">las=</span><span class="dv">1</span>,                     <span class="co"># axis labels in reading direction</span></span>
<span id="cb53-5"><a href="mod-02.html#cb53-5" tabindex="-1"></a>    <span class="at">lend=</span><span class="dv">1</span>,                    <span class="co"># flat line ends</span></span>
<span id="cb53-6"><a href="mod-02.html#cb53-6" tabindex="-1"></a>    <span class="at">cex.lab=</span><span class="fl">1.3</span>,               <span class="co"># bigger axis title (default is 1)</span></span>
<span id="cb53-7"><a href="mod-02.html#cb53-7" tabindex="-1"></a>    <span class="at">cex.axis=</span><span class="fl">1.2</span>)              <span class="co"># bigger axis label (default is 1)</span></span>
<span id="cb53-8"><a href="mod-02.html#cb53-8" tabindex="-1"></a><span class="fu">plot</span>(dat<span class="sc">$</span>height, dat<span class="sc">$</span>antler, <span class="at">type=</span><span class="st">&quot;n&quot;</span>,</span>
<span id="cb53-9"><a href="mod-02.html#cb53-9" tabindex="-1"></a>     <span class="at">xlab=</span><span class="st">&quot;Shoulder height (cm)&quot;</span>,</span>
<span id="cb53-10"><a href="mod-02.html#cb53-10" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">&quot;Antler length (cm)&quot;</span>,</span>
<span id="cb53-11"><a href="mod-02.html#cb53-11" tabindex="-1"></a>     <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">300</span>))</span>
<span id="cb53-12"><a href="mod-02.html#cb53-12" tabindex="-1"></a><span class="fu">points</span>(new.x2, lo, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb53-13"><a href="mod-02.html#cb53-13" tabindex="-1"></a><span class="fu">points</span>(new.x2, hi, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb53-14"><a href="mod-02.html#cb53-14" tabindex="-1"></a><span class="fu">points</span>(new.x2, mn, <span class="at">type=</span><span class="st">&quot;l&quot;</span>, <span class="at">lwd=</span><span class="dv">3</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>)</span>
<span id="cb53-15"><a href="mod-02.html#cb53-15" tabindex="-1"></a><span class="fu">points</span>(dat<span class="sc">$</span>height, dat<span class="sc">$</span>antler, <span class="at">cex=</span><span class="fl">1.3</span>, <span class="at">pch=</span><span class="dv">16</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-69-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>There are some other graphics options we could set, but we’ll save those for later in the course. This figure is clean and spare, with a very high data to ink ratio…just the sort of figure that we want to represent our work <span class="citation">(<a href="#ref-tufte2001visual">Tufte 2001</a>)</span>.</p>
<p>Just for fun, let’s label the original object of our investigation, the Irish elk. We can add a text label and a pointer with the commands <code>text()</code> and <code>segments()</code>. This can require some fiddling to get the coordinates just right. First we’ll use function <code>locator()</code> to find some appropriate coordinates. To use <code>locator()</code>, first run that command in the console. Then, click the points on the graph where you want coordinates. Then, right-click and select “Stop”. The coordinates you clicked will be printed to the console.</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="mod-02.html#cb54-1" tabindex="-1"></a><span class="co"># not run:</span></span>
<span id="cb54-2"><a href="mod-02.html#cb54-2" tabindex="-1"></a><span class="co">#locator()</span></span>
<span id="cb54-3"><a href="mod-02.html#cb54-3" tabindex="-1"></a><span class="co"># click on the figure where you want arrow ends to be</span></span>
<span id="cb54-4"><a href="mod-02.html#cb54-4" tabindex="-1"></a><span class="co"># $x</span></span>
<span id="cb54-5"><a href="mod-02.html#cb54-5" tabindex="-1"></a><span class="co"># [1] 161.7056 177.5281</span></span>
<span id="cb54-6"><a href="mod-02.html#cb54-6" tabindex="-1"></a></span>
<span id="cb54-7"><a href="mod-02.html#cb54-7" tabindex="-1"></a><span class="co"># $y</span></span>
<span id="cb54-8"><a href="mod-02.html#cb54-8" tabindex="-1"></a><span class="co"># [1] 257.0432 244.5633</span></span></code></pre></div>
<p>Then, use the coordinates you found and add the pointer and text. Of course you could do this in another program like Powerpoint, but keeping everything in R can streamline your workflow.</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="mod-02.html#cb55-1" tabindex="-1"></a><span class="co"># add pointer and genus name:</span></span>
<span id="cb55-2"><a href="mod-02.html#cb55-2" tabindex="-1"></a><span class="fu">segments</span>(<span class="fl">161.7</span>, <span class="fl">257.0</span>, <span class="dv">180</span>, <span class="dv">242</span>)</span>
<span id="cb55-3"><a href="mod-02.html#cb55-3" tabindex="-1"></a><span class="fu">text</span>(<span class="dv">160</span>, <span class="dv">260</span>, </span>
<span id="cb55-4"><a href="mod-02.html#cb55-4" tabindex="-1"></a>     <span class="fu">expression</span>(<span class="fu">italic</span>(Megaloceros)), </span>
<span id="cb55-5"><a href="mod-02.html#cb55-5" tabindex="-1"></a>     <span class="at">adj=</span><span class="dv">1</span>,   <span class="co"># right-aligned at point</span></span>
<span id="cb55-6"><a href="mod-02.html#cb55-6" tabindex="-1"></a>     <span class="at">cex=</span><span class="fl">1.3</span>) <span class="co"># text size</span></span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-72-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="save-your-work" class="section level3 hasAnchor" number="2.4.6">
<h3><span class="header-section-number">2.4.6</span> Save your work?<a href="mod-02.html#save-your-work" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now that we are done with the analysis, we can save our work. If you have the code in an R script, you can just save the script and rerun it whenever you want. The results and outputs will be exactly the same next time you open a new R console and run the script<a href="#fn32" class="footnote-ref" id="fnref32"><sup>32</sup></a>. R scripts have the extension <strong>.r</strong> or <strong>.R</strong>.</p>
<p>If you want to save the actual R workspace, you can do this in the console with the <code>save.image()</code> command. By default, this command saves the workspace in your home directory. R workspaces have the extension <strong>.RData</strong>.</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="mod-02.html#cb56-1" tabindex="-1"></a><span class="fu">save.image</span>(<span class="st">&quot;elk_analysis_2021-06-14.RData&quot;</span>)</span></code></pre></div>
<p>The resulting file is the R workspace created by running the script. If you open it, a new R instance will open that contains the same objects and command history as the original. This is more useful in situations involving calculations that take a long time to run, or when the outputs are more complicated than simple text files or images. Notice that the workspace was saved with an informative name and a date. This is important because without the user-supplied name, R will save the workspace with a generic name…potentially overwriting a previously saved workspace!</p>
</div>
<div id="whats-next" class="section level3 hasAnchor" number="2.4.7">
<h3><span class="header-section-number">2.4.7</span> What’s next?<a href="mod-02.html#whats-next" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>This section was designed to demonstrate a typical R workflow with straightforward examples and good coding practices. In the next section, we’ll take a closer look at how R works. Some the material in the upcoming sections was already demonstrated, but not explained, in our first R sesssion.</p>
</div>
</div>
<div id="write-and-execute-commands-in-the-r-console" class="section level2 hasAnchor" number="2.5">
<h2><span class="header-section-number">2.5</span> Write and execute commands in the R console<a href="mod-02.html#write-and-execute-commands-in-the-r-console" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Now that you’ve survived your <a href="mod-02.html#mod-02-first">first R adventure</a>, it’s time to take a closer look at how R works. In this section we’ll focus on getting familiar with the basic syntax of the R language and how to enter R commands.</p>
<div id="r-commandsbasics" class="section level3 hasAnchor" number="2.5.1">
<h3><span class="header-section-number">2.5.1</span> R commands–basics<a href="mod-02.html#r-commandsbasics" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Open up a new R instance or RStudio session and take a look at the <strong>console</strong>. Here commands can be entered at the command prompt, <code>&gt;</code>. The ENTER or RETURN key will execute the current command and print the results (if any) to the console. Try a simple command like <code>2+2</code>. Notice how the output prints to the console, with a <code>[1]</code> at the start of the line. The <code>[1]</code> signifies that that line starts with the first element of the output. This is to help read outputs with lots of elements…try the command <code>rnorm(100)</code> to see why.</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="mod-02.html#cb57-1" tabindex="-1"></a><span class="dv">2</span><span class="sc">+</span><span class="dv">2</span></span></code></pre></div>
<pre><code>## [1] 4</code></pre>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="mod-02.html#cb59-1" tabindex="-1"></a><span class="fu">rnorm</span>(<span class="dv">100</span>)</span></code></pre></div>
<pre><code>##   [1]  0.1868067548  1.2953417219  1.3496341764 -1.8199058851 -1.0129597934
##   [6]  0.7758079984 -0.2446729873 -0.7628311610  1.1871287322  0.5687341349
##  [11]  0.2515907763 -1.3367505458 -1.1523874739  2.4543256127  0.1869963139
##  [16] -0.4393163702 -0.0591246386  2.3218882814  0.6452875006  0.3217042164
##  [21] -1.0585980833 -1.8577768116  0.2816177047 -0.2696880316  0.6292340639
##  [26] -2.0491752819  0.3865173066  0.9388559144  0.9614268935 -1.2941535699
##  [31]  0.1987585317  1.3948937976 -0.4443539013 -0.3467818080 -1.3394891028
##  [36]  0.3248352500  0.3453442934  1.0882126165 -1.0850969947  1.1509566751
##  [41] -1.2319974844  1.3428871601 -0.9701891498  1.9068274828  0.3370097559
##  [46] -1.3898744374 -0.4298016089 -0.0442773529 -0.2836812472 -0.0520081219
##  [51] -1.0893354258 -0.5703620158  1.1049930829  1.2552194034 -1.4890854621
##  [56]  1.3473576941  0.0748210623 -1.6124316377  0.4919105726  0.1515581094
##  [61]  0.5180248518 -0.4768916424  0.1350865322  0.1408681220  0.6483650912
##  [66]  0.2824644386 -0.0001100443 -0.8806680807 -2.3269645388  1.1982580677
##  [71] -0.4832922566  0.5087733866  0.2542475936 -1.2960109532  0.5023409534
##  [76]  1.0496202326 -0.2426459452 -1.4389497661 -0.9029431446  0.8612725843
##  [81] -0.8210071585 -1.4689588806  0.1794126511  0.1425198141 -1.0560004007
##  [86] -1.2274682135 -0.6413455301  0.2250168317  0.3148611203 -0.2958914087
##  [91] -1.6643713237  1.8103941493  0.9552156749 -0.8964807728 -1.2136214600
##  [96]  0.5029205644  0.8911613760 -0.4542943772  0.2878110215 -0.2596923043</code></pre>
<p>Commands can be split over multiple lines. When that happens, a <code>+</code> will be seen at the left side of the window instead of a <code>&gt;</code>. A split command will be executed once a complete expression is input. If you accidentally start a command and aren’t sure how to finish it, you can exit and cancel the command with the ESCAPE key. R executes commands, not lines. You can have more than one command on a line separated by a semi-colon <code>;</code>, but this generally frowned upon.</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="mod-02.html#cb61-1" tabindex="-1"></a><span class="co"># command split across lines:</span></span>
<span id="cb61-2"><a href="mod-02.html#cb61-2" tabindex="-1"></a><span class="dv">2</span><span class="sc">+</span></span>
<span id="cb61-3"><a href="mod-02.html#cb61-3" tabindex="-1"></a><span class="dv">1</span></span>
<span id="cb61-4"><a href="mod-02.html#cb61-4" tabindex="-1"></a><span class="do">## [1] 3</span></span>
<span id="cb61-5"><a href="mod-02.html#cb61-5" tabindex="-1"></a></span>
<span id="cb61-6"><a href="mod-02.html#cb61-6" tabindex="-1"></a><span class="co"># two commands on a line</span></span>
<span id="cb61-7"><a href="mod-02.html#cb61-7" tabindex="-1"></a><span class="dv">2</span><span class="sc">+</span><span class="dv">2</span>; <span class="dv">3</span><span class="sc">+</span><span class="dv">3</span></span>
<span id="cb61-8"><a href="mod-02.html#cb61-8" tabindex="-1"></a><span class="do">## [1] 4</span></span>
<span id="cb61-9"><a href="mod-02.html#cb61-9" tabindex="-1"></a><span class="do">## [1] 6</span></span></code></pre></div>
<p>For this course you should be able to copy and paste most commands directly from the page to your R console. You can also copy and paste commands to the script window and run them from there. This is useful for running multiple commands at once, or for writing more complicated programs. The image below shows this method in action:</p>
<p><img src="02_10.jpg" width="90%" style="display: block; margin: auto;" /></p>
</div>
<div id="elements-of-r-code" class="section level3 hasAnchor" number="2.5.2">
<h3><span class="header-section-number">2.5.2</span> Elements of R code<a href="mod-02.html#elements-of-r-code" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>As you start to learn R, it is important that you save the code you write for later. This way you can refer back to how you solved problems in the past, and have a record of what you did. Sections of code, or whole files filled with code, that accomplish a specific task are referred to as scripts. Get in the habit of scripting everything. This not only documents your workflow, but makes your work reproducible, modifiable, and eventually more efficient. R scripts can even be called from R and executed, which has obvious implications for reproducibility and automation.</p>
<div id="comments" class="section level4 hasAnchor" number="2.5.2.1">
<h4><span class="header-section-number">2.5.2.1</span> Comments<a href="mod-02.html#comments" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>One of the most important elements of your scripts are <strong>comments</strong>. Comments are notes to yourself (or a future reader) that are not read by R. It is difficult to put too many comments in code. Comments are a mechanism for explaining what your code is doing and why. Some comments are just place markers (e.g., <code># Analysis begins here</code>) to help you find important sections. Other comments might be more lengthy, holding explanations of <strong><em>why</em></strong> you chose to code a task the way that you did (e.g., <code># condense with sapply() (maintains order of subunits</code>). The latter kind of comment can be extremely valuable to your future self<a href="#fn33" class="footnote-ref" id="fnref33"><sup>33</sup></a>.</p>
<p>The comment symbol in R is <code>#</code>. Everything on a line after <code>#</code> is not interpreted by R.</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="mod-02.html#cb62-1" tabindex="-1"></a><span class="co"># Here is a comment.</span></span>
<span id="cb62-2"><a href="mod-02.html#cb62-2" tabindex="-1"></a><span class="dv">2</span><span class="sc">+</span><span class="dv">2</span>     <span class="co"># Here is a comment on the end of a line.</span></span></code></pre></div>
<p>Below is an example where the a command is broken across lines, with each line getting an explanatory comment. This is possible because R ignores <strong>whitespace</strong> (spaces, tabs, returns, etc.). Very complex functions or loops might benefit from this kind of commenting. Some programmers do not like this kind of comment…but you comment code the way that works for you.</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="mod-02.html#cb63-1" tabindex="-1"></a><span class="fu">rnorm</span>(</span>
<span id="cb63-2"><a href="mod-02.html#cb63-2" tabindex="-1"></a>    <span class="dv">5</span>,  <span class="co"># sample size</span></span>
<span id="cb63-3"><a href="mod-02.html#cb63-3" tabindex="-1"></a>    <span class="dv">10</span>, <span class="co"># mean</span></span>
<span id="cb63-4"><a href="mod-02.html#cb63-4" tabindex="-1"></a>    <span class="dv">2</span>   <span class="co"># sd</span></span>
<span id="cb63-5"><a href="mod-02.html#cb63-5" tabindex="-1"></a>)<span class="co">#rnorm</span></span></code></pre></div>
<p>Notice in the above example that the closing parenthesis for <code>rnorm()</code> is marked with a comment. This is a good practice if the open and close parentheses are separated by many lines of code. Most IDEs will highlight matching parentheses, but marking them yourself can potentially save some headache.</p>
<p>When writing comments you should keep in mind that most code editors don’t do line breaks very well. Or, they will wrap text displayed in the editor, but this might not display the same on every device. As a rule I try to keep lines of code to <span class="math inline">\(\le\)</span> 80 characters, so I don’t have to scroll over to see the end of a command. You can do this too, by breaking up commands and long comments into multiple lines.</p>
</div>
<div id="objects" class="section level4 hasAnchor" number="2.5.2.2">
<h4><span class="header-section-number">2.5.2.2</span> Objects<a href="mod-02.html#objects" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Everything in R is an <strong>object</strong> of some kind. When you import data, run an analysis, or even just when you do basic calculations, you are creating objects that contain those values. Most of R programming is just calling and manipulating objects in your workspace. That’s sort of like saying that language is just calling and manipulating sounds.</p>
<p>The commands below show how to create an object by assignment (<code>&lt;-</code>), and print the value assigned to an object to the console. The third command shows how the name of an object can be used in place of the value it holds. The fourth command demonstrates that even simple calculations return objects as results, even if the object is just printed to the console instead of saved.</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="mod-02.html#cb64-1" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="dv">3</span></span>
<span id="cb64-2"><a href="mod-02.html#cb64-2" tabindex="-1"></a>a</span>
<span id="cb64-3"><a href="mod-02.html#cb64-3" tabindex="-1"></a><span class="do">## [1] 3</span></span>
<span id="cb64-4"><a href="mod-02.html#cb64-4" tabindex="-1"></a>a<span class="sc">+</span><span class="dv">4</span></span>
<span id="cb64-5"><a href="mod-02.html#cb64-5" tabindex="-1"></a><span class="do">## [1] 7</span></span>
<span id="cb64-6"><a href="mod-02.html#cb64-6" tabindex="-1"></a><span class="fu">class</span>(a<span class="sc">+</span><span class="dv">4</span>)</span>
<span id="cb64-7"><a href="mod-02.html#cb64-7" tabindex="-1"></a><span class="do">## [1] &quot;numeric&quot;</span></span></code></pre></div>
<p>All objects have at least one <strong>class</strong>, and their values have a <strong>type</strong>. We’ll explore some of the most important classes and types <a href="mod-02.html#mod-02-struct">later</a>. For now, the class of an object is what kind of object it is; the type of a value is the kind of value. The class of an object or the type of a value determines how those things interact with R functions.</p>
</div>
<div id="functions" class="section level4 hasAnchor" number="2.5.2.3">
<h4><span class="header-section-number">2.5.2.3</span> Functions<a href="mod-02.html#functions" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>R code is built around using <strong>functions</strong> to manipulate objects (and functions are themselves a kind of object). Functions are called by their names, with the arguments to the function specified within parentheses. This is referred to as a function <strong>call</strong> or a <strong>command</strong>. The arguments of an R function can be named or not. Below are some examples using <code>rnorm()</code>, the function that draws random values from the normal distribution.</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="mod-02.html#cb65-1" tabindex="-1"></a><span class="co"># commands with named arguments:</span></span>
<span id="cb65-2"><a href="mod-02.html#cb65-2" tabindex="-1"></a><span class="fu">rnorm</span>(<span class="at">n=</span><span class="dv">10</span>, <span class="at">mean=</span><span class="dv">5</span>, <span class="at">sd=</span><span class="dv">2</span>)</span>
<span id="cb65-3"><a href="mod-02.html#cb65-3" tabindex="-1"></a><span class="do">##  [1] 5.951771 4.566881 5.216120 2.718580 5.581812 5.251499 2.496277 3.445045</span></span>
<span id="cb65-4"><a href="mod-02.html#cb65-4" tabindex="-1"></a><span class="do">##  [9] 1.361433 2.033040</span></span>
<span id="cb65-5"><a href="mod-02.html#cb65-5" tabindex="-1"></a><span class="fu">rnorm</span>(<span class="at">mean=</span><span class="dv">5</span>, <span class="at">sd=</span><span class="dv">2</span>, <span class="at">n=</span><span class="dv">10</span>)</span>
<span id="cb65-6"><a href="mod-02.html#cb65-6" tabindex="-1"></a><span class="do">##  [1] 3.177625 2.191203 3.969389 6.378298 4.807159 5.320653 4.335875 5.363562</span></span>
<span id="cb65-7"><a href="mod-02.html#cb65-7" tabindex="-1"></a><span class="do">##  [9] 8.644716 3.552593</span></span></code></pre></div>
<p>When the arguments of a function are not named, then R will assume which value to use for which argument. This decision is based on a default order of arguments. The default order can be determined from the help page for the function (accessed in this case by <code>?rnorm</code>). Below are two examples that call <code>rnorm()</code> with unnamed arguments. Notice how different the results are. This is because R expects unnamed arguments to be the number of values, the mean, and the standard deviation (SD), <em>in that order</em>.</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="mod-02.html#cb66-1" tabindex="-1"></a><span class="fu">rnorm</span>(<span class="dv">10</span>,<span class="dv">5</span>,<span class="dv">2</span>)</span>
<span id="cb66-2"><a href="mod-02.html#cb66-2" tabindex="-1"></a><span class="do">##  [1] 6.725276 8.163568 7.725576 4.101334 7.069661 3.312772 5.516937 6.931724</span></span>
<span id="cb66-3"><a href="mod-02.html#cb66-3" tabindex="-1"></a><span class="do">##  [9] 1.116817 5.648863</span></span>
<span id="cb66-4"><a href="mod-02.html#cb66-4" tabindex="-1"></a><span class="fu">rnorm</span>(<span class="dv">2</span>,<span class="dv">5</span>,<span class="dv">10</span>)</span>
<span id="cb66-5"><a href="mod-02.html#cb66-5" tabindex="-1"></a><span class="do">## [1]  8.180806 14.923327</span></span></code></pre></div>
<p>Most R functions require their arguments to be in the correct format (or type). For example, a function might require a number and not a character string. The example below illustrates this by providing the SD as a character string and not as a number. This line should return an error because it cannot run with an incorrectly formatted SD argument.</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="mod-02.html#cb67-1" tabindex="-1"></a><span class="fu">rnorm</span>(<span class="dv">10</span>,<span class="dv">5</span>,<span class="st">&quot;2&quot;</span>)</span></code></pre></div>
<pre><code>## Error in rnorm(10, 5, &quot;2&quot;): invalid arguments</code></pre>
<p>To recap, argument <strong>order</strong> is important when using functions. You can supply arguments to a function without naming them if you put them in the proper order. See the help files to see what the order is. <em>When in doubt, name your arguments</em>.</p>
<p>Functions expect arguments in a particular format. If a function is not working, you might have an argument formatted incorrectly. Sometimes it’s helpful to look at each argument separately in the R console, outside of the function. To do this, copy/paste each argument to the R console separately and execute by hitting ENTER.</p>
<p>Some function arguments have default values, so you don’t have to supply them. For example, <code>runif()</code>, which draws random numbers from the uniform distribution, defaults to the interval between 0 and 1. So, the two <code>runif()</code> commands below are equivalent because 0 and 1 are the defaults for the limits of the interval. The <code>set.seed()</code> command before each <code>runif()</code> command is to set the random number seed, ensuring that the results are identical. If you run the <code>runif()</code> commands without resetting the random number seed, the commands will return different results because of the random nature of the function.</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="mod-02.html#cb69-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb69-2"><a href="mod-02.html#cb69-2" tabindex="-1"></a><span class="fu">runif</span>(<span class="dv">10</span>,<span class="dv">0</span>,<span class="dv">1</span>)</span></code></pre></div>
<pre><code>##  [1] 0.9148060 0.9370754 0.2861395 0.8304476 0.6417455 0.5190959 0.7365883
##  [8] 0.1346666 0.6569923 0.7050648</code></pre>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="mod-02.html#cb71-1" tabindex="-1"></a><span class="co"># again, but with default arguments implied</span></span>
<span id="cb71-2"><a href="mod-02.html#cb71-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb71-3"><a href="mod-02.html#cb71-3" tabindex="-1"></a><span class="fu">runif</span>(<span class="dv">10</span>)</span></code></pre></div>
<pre><code>##  [1] 0.9148060 0.9370754 0.2861395 0.8304476 0.6417455 0.5190959 0.7365883
##  [8] 0.1346666 0.6569923 0.7050648</code></pre>
</div>
</div>
<div id="the-r-workspace" class="section level3 hasAnchor" number="2.5.3">
<h3><span class="header-section-number">2.5.3</span> The R workspace<a href="mod-02.html#the-r-workspace" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>When you use R, all objects and data that you use exist in the <strong>R workspace</strong>. Objects in the workspace can be called from the console. You can think of the workspace like a kind of sandbox or workbench.</p>
<p>Most of the time, when you start R this creates a new workspace with no objects. If you open a previously saved workspace, you will see the message <code>[Previously saved workspace restored]</code>. If you saved your R workspace without a unique name, it can be hard to figure out what workspace you are in. You can use the <code>ls()</code> command to see what is in the workspace, which might help you figure out what workspace has been loaded. If there are no objects in the workspace, the result of <code>ls()</code> will be <code>character(0)</code>. This output means that the <code>ls()</code> command returned a vector of text values (<code>character</code>) with 0 elements. In other words, the workspace contains no objects.</p>
<p>As you work, your workspace will fill up with objects. This isn’t a big deal until you start to either have too many names to keep track of, or start to use too much memory. The first problem can be dealt with by commenting your code. The second problem can be solved by deleting unneeded objects to save space. You can check to see how much memory your R objects are using with the following command:</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="mod-02.html#cb73-1" tabindex="-1"></a><span class="co"># not run, but try on your machine</span></span>
<span id="cb73-2"><a href="mod-02.html#cb73-2" tabindex="-1"></a><span class="fu">sort</span>(<span class="fu">sapply</span>(<span class="fu">mget</span>(<span class="fu">ls</span>()),object.size), <span class="at">decreasing=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p>This will print the sizes of objects in memory in the current workspace, in descending order. The default units are bytes. If you want the results in Mb, divide the output by 1.024e6.</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="mod-02.html#cb74-1" tabindex="-1"></a><span class="co"># not run, but try on your machine</span></span>
<span id="cb74-2"><a href="mod-02.html#cb74-2" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">sapply</span>(<span class="fu">mget</span>(<span class="fu">ls</span>()),object.size), <span class="at">decreasing=</span><span class="cn">TRUE</span>)</span>
<span id="cb74-3"><a href="mod-02.html#cb74-3" tabindex="-1"></a>a<span class="sc">/</span><span class="fl">1.024e6</span></span></code></pre></div>
<p>R objects can be removed using the <code>rm()</code> function. You can remove objects one at a time or by supplying a set of names. Note that the set of names is supplied to an argument called <code>list</code>, which is confusing because <code>list</code> is also the name of a special type of object in R.</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="mod-02.html#cb75-1" tabindex="-1"></a><span class="co"># not run, but try on your machine</span></span>
<span id="cb75-2"><a href="mod-02.html#cb75-2" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb75-3"><a href="mod-02.html#cb75-3" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="dv">2</span></span>
<span id="cb75-4"><a href="mod-02.html#cb75-4" tabindex="-1"></a>z <span class="ot">&lt;-</span> <span class="dv">3</span></span>
<span id="cb75-5"><a href="mod-02.html#cb75-5" tabindex="-1"></a><span class="fu">ls</span>() <span class="co"># shows x, y, and z in workspace</span></span>
<span id="cb75-6"><a href="mod-02.html#cb75-6" tabindex="-1"></a><span class="fu">rm</span>(x)</span>
<span id="cb75-7"><a href="mod-02.html#cb75-7" tabindex="-1"></a><span class="fu">ls</span>() <span class="co"># x is gone</span></span>
<span id="cb75-8"><a href="mod-02.html#cb75-8" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list=</span><span class="fu">c</span>(<span class="st">&quot;y&quot;</span>, <span class="st">&quot;z&quot;</span>))</span>
<span id="cb75-9"><a href="mod-02.html#cb75-9" tabindex="-1"></a><span class="fu">ls</span>() <span class="co"># x, y, and z are gone</span></span></code></pre></div>
</div>
<div id="r-code-basics-assignment-and-operators" class="section level3 hasAnchor" number="2.5.4">
<h3><span class="header-section-number">2.5.4</span> R code basics: assignment and operators<a href="mod-02.html#r-code-basics-assignment-and-operators" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="assignment" class="section level4 hasAnchor" number="2.5.4.1">
<h4><span class="header-section-number">2.5.4.1</span> Assignment<a href="mod-02.html#assignment" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The most important R function is <strong>assignment</strong>. When you <strong>assign</strong> a value to a variable, you can then use that variable in place of the value. Assigning a value to variable name will automatically create a new object in the workspace with that name. If an object with that name already exists, assigning a new value to that name will overwrite the old object.</p>
<p>The left arrow <code>&lt;-</code> is the assignment operator. You can use <code>=</code> for assignment but really shouldn’t because <code>=</code> is easily confused with the equality test operator <code>==</code>. Also, <code>=</code> is a mathematical symbol, but <code>&lt;-</code> is unambiguously an R operation. You can also use <code>-&gt;</code>, but doing so usually makes your code harder to read rather than easier.</p>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="mod-02.html#cb76-1" tabindex="-1"></a>aa <span class="ot">&lt;-</span> <span class="dv">4</span></span>
<span id="cb76-2"><a href="mod-02.html#cb76-2" tabindex="-1"></a>bb <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span></span></code></pre></div>
<p>Don’t do this:</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="mod-02.html#cb77-1" tabindex="-1"></a>cc <span class="ot">=</span> <span class="dv">5</span></span></code></pre></div>
<p>Don’t do this either.</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="mod-02.html#cb78-1" tabindex="-1"></a><span class="dv">5</span> <span class="ot">-&gt;</span> cc</span></code></pre></div>
<p>And definitely don’t do this<a href="#fn34" class="footnote-ref" id="fnref34"><sup>34</sup></a>:</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="mod-02.html#cb79-1" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="dv">5</span> <span class="ot">-&gt;</span> b</span></code></pre></div>
<p>You can chain multiple assignments together in a single command. This is useful if you need to quickly create several objects with the same structure:</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="mod-02.html#cb80-1" tabindex="-1"></a>list3 <span class="ot">&lt;-</span> list2 <span class="ot">&lt;-</span> list1 <span class="ot">&lt;-</span> <span class="fu">vector</span>(<span class="st">&quot;list&quot;</span>, <span class="dv">5</span>)</span></code></pre></div>
<p>Assignment can also be used to <em>change or overwrite parts of pre-existing objects</em>. Usually this involves the <strong>bracket notation</strong>. In the example below, the third element of object a (<code>a[3]</code>) is replaced with the number 42. We’ll practice with the bracket notation in the <a href="mod-03.html#mod-03-brackets">data manipulation module</a>.</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="mod-02.html#cb81-1" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span></span>
<span id="cb81-2"><a href="mod-02.html#cb81-2" tabindex="-1"></a>a</span></code></pre></div>
<pre><code>## [1] 1 2 3 4 5</code></pre>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="mod-02.html#cb83-1" tabindex="-1"></a>a[<span class="dv">3</span>] <span class="ot">&lt;-</span> <span class="dv">42</span></span>
<span id="cb83-2"><a href="mod-02.html#cb83-2" tabindex="-1"></a>a</span></code></pre></div>
<pre><code>## [1]  1  2 42  4  5</code></pre>
<p>One little-known feature of R assignment is that it can <strong>resize</strong> objects. For example, if you assign a value to the <em>n</em>-th element of a vector that has fewer than <em>n</em> elements, R will increase the length of the object to the size needed. It should be noted that this can be done accidentally as well as intentionally, so you need to be very careful if you resize objects using assignment. Personally, I don’t do this because it is often clearer to change the size of objects in a separate command.</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="mod-02.html#cb85-1" tabindex="-1"></a><span class="co"># resize by assignment (not recommended)</span></span>
<span id="cb85-2"><a href="mod-02.html#cb85-2" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span></span>
<span id="cb85-3"><a href="mod-02.html#cb85-3" tabindex="-1"></a>a[<span class="dv">5</span>] <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb85-4"><a href="mod-02.html#cb85-4" tabindex="-1"></a>a</span></code></pre></div>
<pre><code>## [1]  1  2  3 NA  5</code></pre>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="mod-02.html#cb87-1" tabindex="-1"></a><span class="co"># resize in separate command (preferred)</span></span>
<span id="cb87-2"><a href="mod-02.html#cb87-2" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span></span>
<span id="cb87-3"><a href="mod-02.html#cb87-3" tabindex="-1"></a><span class="fu">length</span>(a) <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb87-4"><a href="mod-02.html#cb87-4" tabindex="-1"></a>a</span></code></pre></div>
<pre><code>## [1]  1  2  3 NA NA</code></pre>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="mod-02.html#cb89-1" tabindex="-1"></a>a[<span class="dv">5</span>] <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb89-2"><a href="mod-02.html#cb89-2" tabindex="-1"></a>a</span></code></pre></div>
<pre><code>## [1]  1  2  3 NA  5</code></pre>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="mod-02.html#cb91-1" tabindex="-1"></a><span class="co"># resize by concatenation (preferred)</span></span>
<span id="cb91-2"><a href="mod-02.html#cb91-2" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span></span>
<span id="cb91-3"><a href="mod-02.html#cb91-3" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">c</span>(a, <span class="dv">4</span><span class="sc">:</span><span class="dv">5</span>)</span>
<span id="cb91-4"><a href="mod-02.html#cb91-4" tabindex="-1"></a>a</span></code></pre></div>
<pre><code>## [1] 1 2 3 4 5</code></pre>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="mod-02.html#cb93-1" tabindex="-1"></a><span class="co"># alternatively:</span></span>
<span id="cb93-2"><a href="mod-02.html#cb93-2" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span></span>
<span id="cb93-3"><a href="mod-02.html#cb93-3" tabindex="-1"></a>b <span class="ot">&lt;-</span> <span class="dv">4</span><span class="sc">:</span><span class="dv">5</span></span>
<span id="cb93-4"><a href="mod-02.html#cb93-4" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">c</span>(a, b)</span>
<span id="cb93-5"><a href="mod-02.html#cb93-5" tabindex="-1"></a>a</span></code></pre></div>
<pre><code>## [1] 1 2 3 4 5</code></pre>
</div>
<div id="r-operators" class="section level4 hasAnchor" number="2.5.4.2">
<h4><span class="header-section-number">2.5.4.2</span> R operators<a href="mod-02.html#r-operators" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Most math operators in R are similar to those in other languages/programs. For math expressions, the standard order of operations (PEMDAS) applies, but you can use parentheses if you want to be sure. The integer sequence operator <code>:</code> comes first, and comparison tests (<code>==</code>, <code>&lt;</code>, <code>&lt;=</code>, <code>&gt;</code>, and <code>&gt;=</code>) come last. The examples below show how many common R operators function, taking advantage of the fact that you can use an object’s name in place of its value in most situations.</p>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="mod-02.html#cb95-1" tabindex="-1"></a><span class="co"># make some values to work with</span></span>
<span id="cb95-2"><a href="mod-02.html#cb95-2" tabindex="-1"></a>aa <span class="ot">&lt;-</span> <span class="dv">5</span></span>
<span id="cb95-3"><a href="mod-02.html#cb95-3" tabindex="-1"></a>bb <span class="ot">&lt;-</span> <span class="dv">8</span></span>
<span id="cb95-4"><a href="mod-02.html#cb95-4" tabindex="-1"></a></span>
<span id="cb95-5"><a href="mod-02.html#cb95-5" tabindex="-1"></a><span class="co"># multiply</span></span>
<span id="cb95-6"><a href="mod-02.html#cb95-6" tabindex="-1"></a>aa<span class="sc">*</span><span class="dv">3</span>            </span>
<span id="cb95-7"><a href="mod-02.html#cb95-7" tabindex="-1"></a><span class="do">## [1] 15</span></span>
<span id="cb95-8"><a href="mod-02.html#cb95-8" tabindex="-1"></a></span>
<span id="cb95-9"><a href="mod-02.html#cb95-9" tabindex="-1"></a><span class="co"># add</span></span>
<span id="cb95-10"><a href="mod-02.html#cb95-10" tabindex="-1"></a>bb<span class="sc">+</span><span class="dv">3</span>            </span>
<span id="cb95-11"><a href="mod-02.html#cb95-11" tabindex="-1"></a><span class="do">## [1] 11</span></span>
<span id="cb95-12"><a href="mod-02.html#cb95-12" tabindex="-1"></a></span>
<span id="cb95-13"><a href="mod-02.html#cb95-13" tabindex="-1"></a><span class="co"># square root</span></span>
<span id="cb95-14"><a href="mod-02.html#cb95-14" tabindex="-1"></a><span class="fu">sqrt</span>(aa)        </span>
<span id="cb95-15"><a href="mod-02.html#cb95-15" tabindex="-1"></a><span class="do">## [1] 2.236068</span></span>
<span id="cb95-16"><a href="mod-02.html#cb95-16" tabindex="-1"></a></span>
<span id="cb95-17"><a href="mod-02.html#cb95-17" tabindex="-1"></a><span class="co"># natural logarithm (base e)</span></span>
<span id="cb95-18"><a href="mod-02.html#cb95-18" tabindex="-1"></a><span class="fu">log</span>(aa)         </span>
<span id="cb95-19"><a href="mod-02.html#cb95-19" tabindex="-1"></a><span class="do">## [1] 1.609438</span></span>
<span id="cb95-20"><a href="mod-02.html#cb95-20" tabindex="-1"></a></span>
<span id="cb95-21"><a href="mod-02.html#cb95-21" tabindex="-1"></a><span class="co"># logarithm (arbitrary base)</span></span>
<span id="cb95-22"><a href="mod-02.html#cb95-22" tabindex="-1"></a><span class="fu">log</span>(aa, <span class="at">base=</span><span class="dv">3</span>) </span>
<span id="cb95-23"><a href="mod-02.html#cb95-23" tabindex="-1"></a><span class="do">## [1] 1.464974</span></span>
<span id="cb95-24"><a href="mod-02.html#cb95-24" tabindex="-1"></a><span class="fu">log</span>(aa, pi)     </span>
<span id="cb95-25"><a href="mod-02.html#cb95-25" tabindex="-1"></a><span class="do">## [1] 1.405954</span></span>
<span id="cb95-26"><a href="mod-02.html#cb95-26" tabindex="-1"></a></span>
<span id="cb95-27"><a href="mod-02.html#cb95-27" tabindex="-1"></a><span class="co"># common logarithm (base 10)</span></span>
<span id="cb95-28"><a href="mod-02.html#cb95-28" tabindex="-1"></a><span class="fu">log10</span>(aa)       </span>
<span id="cb95-29"><a href="mod-02.html#cb95-29" tabindex="-1"></a><span class="do">## [1] 0.69897</span></span>
<span id="cb95-30"><a href="mod-02.html#cb95-30" tabindex="-1"></a></span>
<span id="cb95-31"><a href="mod-02.html#cb95-31" tabindex="-1"></a><span class="co"># exponentiation with base e (antilog)</span></span>
<span id="cb95-32"><a href="mod-02.html#cb95-32" tabindex="-1"></a><span class="fu">exp</span>(aa)         </span>
<span id="cb95-33"><a href="mod-02.html#cb95-33" tabindex="-1"></a><span class="do">## [1] 148.4132</span></span>
<span id="cb95-34"><a href="mod-02.html#cb95-34" tabindex="-1"></a></span>
<span id="cb95-35"><a href="mod-02.html#cb95-35" tabindex="-1"></a><span class="co"># Vectorized (element-wise) multiplication</span></span>
<span id="cb95-36"><a href="mod-02.html#cb95-36" tabindex="-1"></a>bb<span class="sc">*</span><span class="dv">2</span><span class="sc">:</span><span class="dv">5</span>          </span>
<span id="cb95-37"><a href="mod-02.html#cb95-37" tabindex="-1"></a><span class="do">## [1] 16 24 32 40</span></span>
<span id="cb95-38"><a href="mod-02.html#cb95-38" tabindex="-1"></a></span>
<span id="cb95-39"><a href="mod-02.html#cb95-39" tabindex="-1"></a><span class="co"># Exponentiation</span></span>
<span id="cb95-40"><a href="mod-02.html#cb95-40" tabindex="-1"></a>bb<span class="sc">^</span><span class="dv">2</span>            </span>
<span id="cb95-41"><a href="mod-02.html#cb95-41" tabindex="-1"></a><span class="do">## [1] 64</span></span>
<span id="cb95-42"><a href="mod-02.html#cb95-42" tabindex="-1"></a></span>
<span id="cb95-43"><a href="mod-02.html#cb95-43" tabindex="-1"></a><span class="co"># Alternate symbol for exponentiation (rare)</span></span>
<span id="cb95-44"><a href="mod-02.html#cb95-44" tabindex="-1"></a>bb<span class="sc">**</span><span class="dv">2</span>           </span>
<span id="cb95-45"><a href="mod-02.html#cb95-45" tabindex="-1"></a><span class="do">## [1] 64</span></span>
<span id="cb95-46"><a href="mod-02.html#cb95-46" tabindex="-1"></a></span>
<span id="cb95-47"><a href="mod-02.html#cb95-47" tabindex="-1"></a><span class="co"># Remainder (modulus)</span></span>
<span id="cb95-48"><a href="mod-02.html#cb95-48" tabindex="-1"></a>bb <span class="sc">%%</span> <span class="dv">3</span>         </span>
<span id="cb95-49"><a href="mod-02.html#cb95-49" tabindex="-1"></a><span class="do">## [1] 2</span></span>
<span id="cb95-50"><a href="mod-02.html#cb95-50" tabindex="-1"></a></span>
<span id="cb95-51"><a href="mod-02.html#cb95-51" tabindex="-1"></a><span class="co"># Quotient without remainder</span></span>
<span id="cb95-52"><a href="mod-02.html#cb95-52" tabindex="-1"></a>bb <span class="sc">%/%</span> <span class="dv">3</span>        </span>
<span id="cb95-53"><a href="mod-02.html#cb95-53" tabindex="-1"></a><span class="do">## [1] 2</span></span>
<span id="cb95-54"><a href="mod-02.html#cb95-54" tabindex="-1"></a></span>
<span id="cb95-55"><a href="mod-02.html#cb95-55" tabindex="-1"></a><span class="co"># Order of operations</span></span>
<span id="cb95-56"><a href="mod-02.html#cb95-56" tabindex="-1"></a>aa<span class="sc">+</span><span class="dv">2</span><span class="sc">*</span><span class="dv">3</span>          </span>
<span id="cb95-57"><a href="mod-02.html#cb95-57" tabindex="-1"></a><span class="do">## [1] 11</span></span>
<span id="cb95-58"><a href="mod-02.html#cb95-58" tabindex="-1"></a></span>
<span id="cb95-59"><a href="mod-02.html#cb95-59" tabindex="-1"></a><span class="co"># Order of operations</span></span>
<span id="cb95-60"><a href="mod-02.html#cb95-60" tabindex="-1"></a>aa<span class="sc">+</span>(<span class="dv">2</span><span class="sc">*</span><span class="dv">3</span>)        </span>
<span id="cb95-61"><a href="mod-02.html#cb95-61" tabindex="-1"></a><span class="do">## [1] 11</span></span>
<span id="cb95-62"><a href="mod-02.html#cb95-62" tabindex="-1"></a></span>
<span id="cb95-63"><a href="mod-02.html#cb95-63" tabindex="-1"></a><span class="co"># Order of operations</span></span>
<span id="cb95-64"><a href="mod-02.html#cb95-64" tabindex="-1"></a>(aa<span class="sc">+</span><span class="dv">2</span>)<span class="sc">*</span><span class="dv">3</span>        </span>
<span id="cb95-65"><a href="mod-02.html#cb95-65" tabindex="-1"></a><span class="do">## [1] 21</span></span>
<span id="cb95-66"><a href="mod-02.html#cb95-66" tabindex="-1"></a></span>
<span id="cb95-67"><a href="mod-02.html#cb95-67" tabindex="-1"></a><span class="co"># Sequence of integers</span></span>
<span id="cb95-68"><a href="mod-02.html#cb95-68" tabindex="-1"></a><span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>            </span>
<span id="cb95-69"><a href="mod-02.html#cb95-69" tabindex="-1"></a><span class="do">##  [1]  1  2  3  4  5  6  7  8  9 10</span></span>
<span id="cb95-70"><a href="mod-02.html#cb95-70" tabindex="-1"></a></span>
<span id="cb95-71"><a href="mod-02.html#cb95-71" tabindex="-1"></a><span class="co"># Note that &quot;:&quot; comes first!</span></span>
<span id="cb95-72"><a href="mod-02.html#cb95-72" tabindex="-1"></a><span class="dv">1</span><span class="sc">:</span><span class="dv">10</span><span class="sc">*</span><span class="dv">10</span>         </span>
<span id="cb95-73"><a href="mod-02.html#cb95-73" tabindex="-1"></a><span class="do">##  [1]  10  20  30  40  50  60  70  80  90 100</span></span>
<span id="cb95-74"><a href="mod-02.html#cb95-74" tabindex="-1"></a></span>
<span id="cb95-75"><a href="mod-02.html#cb95-75" tabindex="-1"></a><span class="co"># Comparisons are last</span></span>
<span id="cb95-76"><a href="mod-02.html#cb95-76" tabindex="-1"></a><span class="dv">2</span><span class="sc">*</span><span class="dv">2</span> <span class="sc">&gt;</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span><span class="sc">*</span><span class="dv">2</span>     </span>
<span id="cb95-77"><a href="mod-02.html#cb95-77" tabindex="-1"></a><span class="do">## [1]  TRUE FALSE FALSE FALSE FALSE</span></span></code></pre></div>
<p>Some of these examples illustrate a key concept in R programming: <strong>vectorization</strong>. Many R functions can operate on sets of numbers (vectors) just as well as on single values (scalars). The function will operate on every element of the vector <em>in order</em>. Most R functions are vectorized to some extent. This is very handy because many operations in statistics deal with vectors of values.</p>
<p>Vectorization can cause headaches if the vectors are not of the same length. If the length of one vector is a multiple of the length of the other, then the shorter vector gets recycled. This is sometimes called the “recycling rule”. The recycling rule can be very convenient if you know what you’re doing and are paying attention. If you don’t pay attention, or aren’t aware that recycling is happening, it can create a headache because values are recycled without telling you.</p>
<p>If the length of the longer vector is not a multiple of the length of the shorter vector, the shorter one will be recycled a non-integer number of times and R will return a warning (<code>longer object length is not a multiple of shorter object length</code>).</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="mod-02.html#cb96-1" tabindex="-1"></a><span class="co"># lengths compatible:</span></span>
<span id="cb96-2"><a href="mod-02.html#cb96-2" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span></span>
<span id="cb96-3"><a href="mod-02.html#cb96-3" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="dv">3</span><span class="sc">:</span><span class="dv">6</span></span>
<span id="cb96-4"><a href="mod-02.html#cb96-4" tabindex="-1"></a>x<span class="sc">+</span>y</span></code></pre></div>
<pre><code>## [1] 4 6 6 8</code></pre>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="mod-02.html#cb98-1" tabindex="-1"></a><span class="co"># lengths not compatible:</span></span>
<span id="cb98-2"><a href="mod-02.html#cb98-2" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span></span>
<span id="cb98-3"><a href="mod-02.html#cb98-3" tabindex="-1"></a>b <span class="ot">&lt;-</span> <span class="dv">4</span><span class="sc">:</span><span class="dv">5</span></span>
<span id="cb98-4"><a href="mod-02.html#cb98-4" tabindex="-1"></a>a <span class="sc">+</span> b</span></code></pre></div>
<pre><code>## Warning in a + b: longer object length is not a multiple of shorter object
## length</code></pre>
<pre><code>## [1] 5 7 7</code></pre>
</div>
<div id="selecting-with-brackets" class="section level4 hasAnchor" number="2.5.4.3">
<h4><span class="header-section-number">2.5.4.3</span> Selecting with brackets <code>[]</code><a href="mod-02.html#selecting-with-brackets" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The most common way to select parts of R objects is with the <strong>bracket notation</strong>. The square bracket symbols <code>[]</code> are used to define subsets of an object. Within the brackets, pieces of an object are selected by the indices of the elements to be selected. Unlike many programming languages, indices in R start at 1 (instead of 0). The example below shows how brackets work on a vector of numeric values.</p>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="mod-02.html#cb101-1" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="dv">3</span><span class="sc">:</span><span class="dv">8</span></span>
<span id="cb101-2"><a href="mod-02.html#cb101-2" tabindex="-1"></a>a[<span class="dv">1</span>]    <span class="co"># first element of a</span></span>
<span id="cb101-3"><a href="mod-02.html#cb101-3" tabindex="-1"></a><span class="do">## [1] 3</span></span>
<span id="cb101-4"><a href="mod-02.html#cb101-4" tabindex="-1"></a>a[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]  <span class="co"># first through third elements</span></span>
<span id="cb101-5"><a href="mod-02.html#cb101-5" tabindex="-1"></a><span class="do">## [1] 3 4 5</span></span>
<span id="cb101-6"><a href="mod-02.html#cb101-6" tabindex="-1"></a>a[<span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">4</span>)]   <span class="co"># second and fourth elements</span></span>
<span id="cb101-7"><a href="mod-02.html#cb101-7" tabindex="-1"></a><span class="do">## [1] 4 6</span></span></code></pre></div>
<p>Remember, everything in R is an object. You can save the output from one of the commands above as an object:</p>
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb102-1"><a href="mod-02.html#cb102-1" tabindex="-1"></a>b <span class="ot">&lt;-</span> a[<span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">4</span>)]</span>
<span id="cb102-2"><a href="mod-02.html#cb102-2" tabindex="-1"></a>b</span></code></pre></div>
<pre><code>## [1] 4 6</code></pre>
<p>Brackets work on more complicated objects as well. The figure below shows how brackets work on a matrix, data frame, or array, which have 2, 2, and <span class="math inline">\(\ge\)</span> 3 dimensions, respectively.</p>
<p><img src="02_11.jpg" width="90%" style="display: block; margin: auto;" /></p>
<p>The examples below select parts of the <code>iris</code> dataset. The <code>iris</code> dataset is a <strong>data frame</strong>, one of the most important types of objects in R. Unlike the vector <code>a</code> above, <code>iris</code> has two dimensions: <strong>rows</strong> and <strong>columns</strong>. In the brackets, the dimensions are given separated by a comma. You <em>must</em> include the comma which separates the dimensions, even if you don’t supply a value for one of the dimensions. Supplying the wrong number of dimensions will either cause an error (good) or something unexpected (bad). A missing value for rows or columns will select all rows or columns.</p>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb104-1"><a href="mod-02.html#cb104-1" tabindex="-1"></a><span class="co"># not run because output very long; try on your machine!</span></span>
<span id="cb104-2"><a href="mod-02.html#cb104-2" tabindex="-1"></a></span>
<span id="cb104-3"><a href="mod-02.html#cb104-3" tabindex="-1"></a><span class="co"># first 4 rows, all columns of iris</span></span>
<span id="cb104-4"><a href="mod-02.html#cb104-4" tabindex="-1"></a>iris[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>,]  </span>
<span id="cb104-5"><a href="mod-02.html#cb104-5" tabindex="-1"></a></span>
<span id="cb104-6"><a href="mod-02.html#cb104-6" tabindex="-1"></a><span class="co"># first 3 columns, all rows of iris</span></span>
<span id="cb104-7"><a href="mod-02.html#cb104-7" tabindex="-1"></a>iris[,<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]  </span>
<span id="cb104-8"><a href="mod-02.html#cb104-8" tabindex="-1"></a></span>
<span id="cb104-9"><a href="mod-02.html#cb104-9" tabindex="-1"></a><span class="co"># rows 12-16 and columns 1-4</span></span>
<span id="cb104-10"><a href="mod-02.html#cb104-10" tabindex="-1"></a>iris[<span class="dv">12</span><span class="sc">:</span><span class="dv">16</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]    </span></code></pre></div>
<p>The bracket notation can be extremely powerful when logical tests are used instead of supplying numbers directly. Any R expression or command that evaluates to numeric values can be supplied inside the brackets. This is how I usually subset my datasets.</p>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb105-1"><a href="mod-02.html#cb105-1" tabindex="-1"></a><span class="co"># not run because output very long; try on your machine!</span></span>
<span id="cb105-2"><a href="mod-02.html#cb105-2" tabindex="-1"></a></span>
<span id="cb105-3"><a href="mod-02.html#cb105-3" tabindex="-1"></a><span class="co"># find rows by logical tests</span></span>
<span id="cb105-4"><a href="mod-02.html#cb105-4" tabindex="-1"></a>flag1 <span class="ot">&lt;-</span> <span class="fu">which</span>(iris<span class="sc">$</span>Species <span class="sc">==</span> <span class="st">&quot;setosa&quot;</span>)</span>
<span id="cb105-5"><a href="mod-02.html#cb105-5" tabindex="-1"></a>flag2 <span class="ot">&lt;-</span> <span class="fu">which</span>(iris<span class="sc">$</span>Sepal.Length <span class="sc">&gt;</span> <span class="fl">5.4</span>)</span>
<span id="cb105-6"><a href="mod-02.html#cb105-6" tabindex="-1"></a>iris[flag1,]                       <span class="co"># meets first condition</span></span>
<span id="cb105-7"><a href="mod-02.html#cb105-7" tabindex="-1"></a>iris[<span class="fu">intersect</span>(flag1, flag2),]     <span class="co"># meets both conditions</span></span>
<span id="cb105-8"><a href="mod-02.html#cb105-8" tabindex="-1"></a>iris[<span class="fu">union</span>(flag1, flag2),]         <span class="co"># meets either condition</span></span>
<span id="cb105-9"><a href="mod-02.html#cb105-9" tabindex="-1"></a>iris[<span class="fu">setdiff</span>(flag1, flag2),]       <span class="co"># meets neither condition</span></span></code></pre></div>
<p>When using brackets you need to be careful to only supply indices that actually exist. R’s behavior when invalid indices are requested depends on the type of object you are trying to extract from. Requesting invalid indices from an array (vectors, matrices, or arrays) will return <code>NA</code> or an error.</p>
<div class="sourceCode" id="cb106"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb106-1"><a href="mod-02.html#cb106-1" tabindex="-1"></a><span class="co"># NA:</span></span>
<span id="cb106-2"><a href="mod-02.html#cb106-2" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span></span>
<span id="cb106-3"><a href="mod-02.html#cb106-3" tabindex="-1"></a>x[<span class="dv">6</span>]</span></code></pre></div>
<pre><code>## [1] NA</code></pre>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb108-1"><a href="mod-02.html#cb108-1" tabindex="-1"></a><span class="co"># error:</span></span>
<span id="cb108-2"><a href="mod-02.html#cb108-2" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">12</span>, <span class="at">nrow=</span><span class="dv">3</span>)</span>
<span id="cb108-3"><a href="mod-02.html#cb108-3" tabindex="-1"></a>x[<span class="dv">7</span>,]</span></code></pre></div>
<pre><code>## Error in x[7, ]: subscript out of bounds</code></pre>
<p>Requesting a row from a data frame that does not exist will return a data frame with the same columns as the data frame, but with row full of <code>NA</code>. This feels like an error, but can be useful if you want to make a different version of a data frame with a similar structure. The example below makes new versions of data frame <code>x</code> with 1 and 5 empty rows.</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="mod-02.html#cb110-1" tabindex="-1"></a>x <span class="ot">&lt;-</span> iris[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,]</span>
<span id="cb110-2"><a href="mod-02.html#cb110-2" tabindex="-1"></a>x[<span class="dv">6</span>,]</span></code></pre></div>
<pre><code>##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## NA           NA          NA           NA          NA    &lt;NA&gt;</code></pre>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="mod-02.html#cb112-1" tabindex="-1"></a>x[<span class="dv">11</span><span class="sc">:</span><span class="dv">15</span>,]</span></code></pre></div>
<pre><code>##      Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## NA             NA          NA           NA          NA    &lt;NA&gt;
## NA.1           NA          NA           NA          NA    &lt;NA&gt;
## NA.2           NA          NA           NA          NA    &lt;NA&gt;
## NA.3           NA          NA           NA          NA    &lt;NA&gt;
## NA.4           NA          NA           NA          NA    &lt;NA&gt;</code></pre>
<p>Requesting a column that does not exist from a data frame will return an error:</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="mod-02.html#cb114-1" tabindex="-1"></a>x[,<span class="dv">7</span>]</span></code></pre></div>
<pre><code>## Error in `[.data.frame`(x, , 7): undefined columns selected</code></pre>
<p>We will go over brackets and their usage in much more detail in a <a href="mod-03.html#mod-03-brackets">later section</a>.</p>
</div>
</div>
</div>
<div id="mod-02-struct" class="section level2 hasAnchor" number="2.6">
<h2><span class="header-section-number">2.6</span> Basic R data structures<a href="mod-02.html#mod-02-struct" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In R, data are stored in <strong>objects</strong> with a well-defined structure. The structures of these objects can facilitate manipulating and retrieving your data. The structure of an R object is determined by its class. The format of data (i.e., the kinds of values contained in an object) is referred to as the data type.</p>
<p>For a solid introduction to R classes and data types, refer to <a href="https://cran.r-project.org/doc/manuals/r-release/R-intro.html">An introduction to R</a>, one of the official R help manuals. The book <em>Data manipulation with R</em> <span class="citation">(<a href="#ref-spector2008">Spector 2008</a>)</span> is also good. If you want to explore the odder and less benign aspects of R data structures, I recommend <a href="https://www.burns-stat.com/pages/Tutor/R_inferno.pdf"><em>The R Inferno</em></a> (accessed 2021-12-20).</p>
<div id="vectors" class="section level3 hasAnchor" number="2.6.1">
<h3><span class="header-section-number">2.6.1</span> Vectors<a href="mod-02.html#vectors" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In R, as in mathematics, a <strong>vector</strong> is an ordered set of values. Vectors are a key concept in statistics and linear algebra. For example, the set [1, 2, 3, 4, 5] is a vector. This vector is different than the set [2, 3, 1, 5, 4]. These sets are different vectors because even though they contain the same values, those values are in different orders. In R we could call either of those sets a “vector”. Note that we don’t call them “lists” because a list is a special kind of object in R that is not synonymous with vector.</p>
<p>As a language designed primarily for statistics, R has a strong emphasis on vectors and their manipulation. Vectors are probably the simplest type of data object in R. Understanding them is the key to understanding other data objects in R.</p>
<p>Another way to think of vectors is as an <strong>array</strong> with one dimension, length. In computing, an “array” is a systematic arrangement of similar elements. An array with 1 dimension is a vector. A two-dimensional array is a <strong>matrix</strong>, and an array with <span class="math inline">\(\ge\)</span> 3 dimensions is just called an array. Vectors, matrices, and arrays in R can be a convenient way to store structured data that are all of the same type (numeric, logical, character, etc.).</p>
<p>Vectors can be created by the function <code>vector()</code> as well as by functions specific to the type of value to be stored. The most common types are <strong>numeric</strong>, <strong>logical</strong>, and <strong>character</strong>. These hold numbers, logical values, and text, respectively. Notice that each type has a default value (<code>0</code>, <code>FALSE</code>, and <code>""</code> (blank), respectively). Notice also that in each command below R prints the results to the console because the resulting vectors are not assigned to an object.</p>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="mod-02.html#cb116-1" tabindex="-1"></a><span class="fu">vector</span>(<span class="st">&quot;numeric&quot;</span>, <span class="dv">5</span>)</span>
<span id="cb116-2"><a href="mod-02.html#cb116-2" tabindex="-1"></a><span class="do">## [1] 0 0 0 0 0</span></span>
<span id="cb116-3"><a href="mod-02.html#cb116-3" tabindex="-1"></a><span class="fu">vector</span>(<span class="st">&quot;logical&quot;</span>, <span class="dv">5</span>)</span>
<span id="cb116-4"><a href="mod-02.html#cb116-4" tabindex="-1"></a><span class="do">## [1] FALSE FALSE FALSE FALSE FALSE</span></span>
<span id="cb116-5"><a href="mod-02.html#cb116-5" tabindex="-1"></a><span class="fu">vector</span>(<span class="st">&quot;character&quot;</span>,<span class="dv">5</span>)</span>
<span id="cb116-6"><a href="mod-02.html#cb116-6" tabindex="-1"></a><span class="do">## [1] &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot;</span></span>
<span id="cb116-7"><a href="mod-02.html#cb116-7" tabindex="-1"></a></span>
<span id="cb116-8"><a href="mod-02.html#cb116-8" tabindex="-1"></a><span class="co"># equivalent to above commands:</span></span>
<span id="cb116-9"><a href="mod-02.html#cb116-9" tabindex="-1"></a><span class="fu">numeric</span>(<span class="dv">5</span>)</span>
<span id="cb116-10"><a href="mod-02.html#cb116-10" tabindex="-1"></a><span class="do">## [1] 0 0 0 0 0</span></span>
<span id="cb116-11"><a href="mod-02.html#cb116-11" tabindex="-1"></a><span class="fu">logical</span>(<span class="dv">5</span>)</span>
<span id="cb116-12"><a href="mod-02.html#cb116-12" tabindex="-1"></a><span class="do">## [1] FALSE FALSE FALSE FALSE FALSE</span></span>
<span id="cb116-13"><a href="mod-02.html#cb116-13" tabindex="-1"></a><span class="fu">character</span>(<span class="dv">5</span>)</span>
<span id="cb116-14"><a href="mod-02.html#cb116-14" tabindex="-1"></a><span class="do">## [1] &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot;</span></span></code></pre></div>
<p>Creating vectors this way is cumbersome, but useful when programming complex tasks. It is usually more efficient, from a memory usage standpoint, to create an object and fill it later than to create it and repeatedly change its size.</p>
<p>You can also create vectors by defining their contents. When you do this, R will guess at the type of vector based on the values you give it. These guesses are usually okay.</p>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="mod-02.html#cb117-1" tabindex="-1"></a><span class="co"># numeric:</span></span>
<span id="cb117-2"><a href="mod-02.html#cb117-2" tabindex="-1"></a><span class="fu">class</span>(<span class="fu">runif</span>(<span class="dv">10</span>))</span></code></pre></div>
<pre><code>## [1] &quot;numeric&quot;</code></pre>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb119-1"><a href="mod-02.html#cb119-1" tabindex="-1"></a><span class="co"># integer, which is stored in numeric vectors</span></span>
<span id="cb119-2"><a href="mod-02.html#cb119-2" tabindex="-1"></a><span class="co"># despite being a different type:</span></span>
<span id="cb119-3"><a href="mod-02.html#cb119-3" tabindex="-1"></a><span class="fu">class</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)</span></code></pre></div>
<pre><code>## [1] &quot;integer&quot;</code></pre>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb121-1"><a href="mod-02.html#cb121-1" tabindex="-1"></a><span class="fu">is.numeric</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>Vectors can also be made by putting values together using the function <code>c()</code> (short for “combine” or “concatenate”). Note that if you try to combine values of different types, R will convert all values to character strings.</p>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb123-1"><a href="mod-02.html#cb123-1" tabindex="-1"></a>my.vec <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">4</span><span class="sc">:</span><span class="dv">8</span>,<span class="dv">10</span>)</span>
<span id="cb123-2"><a href="mod-02.html#cb123-2" tabindex="-1"></a>my.vec</span>
<span id="cb123-3"><a href="mod-02.html#cb123-3" tabindex="-1"></a><span class="do">## [1]  1  2  4  5  6  7  8 10</span></span>
<span id="cb123-4"><a href="mod-02.html#cb123-4" tabindex="-1"></a>my.vec2 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;zebra&quot;</span>,<span class="st">&quot;hippo&quot;</span>,<span class="st">&quot;lion&quot;</span>,<span class="st">&quot;rhino&quot;</span>,<span class="st">&quot;rhino&quot;</span>)</span>
<span id="cb123-5"><a href="mod-02.html#cb123-5" tabindex="-1"></a>my.vec2</span>
<span id="cb123-6"><a href="mod-02.html#cb123-6" tabindex="-1"></a><span class="do">## [1] &quot;zebra&quot; &quot;hippo&quot; &quot;lion&quot;  &quot;rhino&quot; &quot;rhino&quot;</span></span>
<span id="cb123-7"><a href="mod-02.html#cb123-7" tabindex="-1"></a>my.vec3 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="st">&quot;hippo&quot;</span>,<span class="dv">1</span><span class="sc">+</span><span class="dv">1</span><span class="sc">==</span><span class="dv">3</span>)</span>
<span id="cb123-8"><a href="mod-02.html#cb123-8" tabindex="-1"></a>my.vec3</span>
<span id="cb123-9"><a href="mod-02.html#cb123-9" tabindex="-1"></a><span class="do">## [1] &quot;1&quot;     &quot;hippo&quot; &quot;FALSE&quot;</span></span></code></pre></div>
<p>Each value within a vector is called an <strong>element</strong>. Elements of a vector are accessed by the bracket notation.</p>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="mod-02.html#cb124-1" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">runif</span>(<span class="dv">5</span>)</span>
<span id="cb124-2"><a href="mod-02.html#cb124-2" tabindex="-1"></a></span>
<span id="cb124-3"><a href="mod-02.html#cb124-3" tabindex="-1"></a><span class="co"># print to console</span></span>
<span id="cb124-4"><a href="mod-02.html#cb124-4" tabindex="-1"></a>a          </span>
<span id="cb124-5"><a href="mod-02.html#cb124-5" tabindex="-1"></a><span class="do">## [1] 0.90403139 0.13871017 0.98889173 0.94666823 0.08243756</span></span>
<span id="cb124-6"><a href="mod-02.html#cb124-6" tabindex="-1"></a></span>
<span id="cb124-7"><a href="mod-02.html#cb124-7" tabindex="-1"></a><span class="co"># 3rd element</span></span>
<span id="cb124-8"><a href="mod-02.html#cb124-8" tabindex="-1"></a>a[<span class="dv">3</span>]       </span>
<span id="cb124-9"><a href="mod-02.html#cb124-9" tabindex="-1"></a><span class="do">## [1] 0.9888917</span></span>
<span id="cb124-10"><a href="mod-02.html#cb124-10" tabindex="-1"></a></span>
<span id="cb124-11"><a href="mod-02.html#cb124-11" tabindex="-1"></a><span class="co"># elements 2-4</span></span>
<span id="cb124-12"><a href="mod-02.html#cb124-12" tabindex="-1"></a>a[<span class="dv">2</span><span class="sc">:</span><span class="dv">4</span>]     </span>
<span id="cb124-13"><a href="mod-02.html#cb124-13" tabindex="-1"></a><span class="do">## [1] 0.1387102 0.9888917 0.9466682</span></span>
<span id="cb124-14"><a href="mod-02.html#cb124-14" tabindex="-1"></a></span>
<span id="cb124-15"><a href="mod-02.html#cb124-15" tabindex="-1"></a><span class="co"># elements other than 1</span></span>
<span id="cb124-16"><a href="mod-02.html#cb124-16" tabindex="-1"></a>a[<span class="sc">-</span><span class="dv">1</span>]      </span>
<span id="cb124-17"><a href="mod-02.html#cb124-17" tabindex="-1"></a><span class="do">## [1] 0.13871017 0.98889173 0.94666823 0.08243756</span></span>
<span id="cb124-18"><a href="mod-02.html#cb124-18" tabindex="-1"></a></span>
<span id="cb124-19"><a href="mod-02.html#cb124-19" tabindex="-1"></a><span class="co"># elements other than 1 and 2</span></span>
<span id="cb124-20"><a href="mod-02.html#cb124-20" tabindex="-1"></a>a[<span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>)] </span>
<span id="cb124-21"><a href="mod-02.html#cb124-21" tabindex="-1"></a><span class="do">## [1] 0.98889173 0.94666823 0.08243756</span></span></code></pre></div>
<p>You can inspect a vector (or any object) by printing it to the console. Do this by typing the object’s name into the console at the command prompt <code>&gt;</code> and pressing ENTER.</p>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb125-1"><a href="mod-02.html#cb125-1" tabindex="-1"></a>my.vec</span></code></pre></div>
<pre><code>## [1]  1  2  4  5  6  7  8 10</code></pre>
<p>The <code>[1]</code> at the start of the line with your results indicates that that line starts with the first element of the object. If an object runs over to multiple lines, each line will begin with the index of the first value printed on that line. Make your console window narrower and try this command:</p>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb127-1"><a href="mod-02.html#cb127-1" tabindex="-1"></a><span class="fu">runif</span>(<span class="dv">20</span>)</span></code></pre></div>
<pre><code>##  [1] 0.514211784 0.390203467 0.905738131 0.446969628 0.836004260 0.737595618
##  [7] 0.811055141 0.388108283 0.685169729 0.003948339 0.832916080 0.007334147
## [13] 0.207658973 0.906601408 0.611778643 0.379559241 0.435771585 0.037431033
## [19] 0.973539914 0.431751249</code></pre>
<p>Each element of a vector can have a name, set or accessed using the function <code>names()</code>. Element names can even be used to access elements of a vector.</p>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb129-1"><a href="mod-02.html#cb129-1" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">runif</span>(<span class="dv">3</span>)</span>
<span id="cb129-2"><a href="mod-02.html#cb129-2" tabindex="-1"></a><span class="fu">names</span>(a) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;value1&quot;</span>, <span class="st">&quot;value2&quot;</span>, <span class="st">&quot;value3&quot;</span>)</span>
<span id="cb129-3"><a href="mod-02.html#cb129-3" tabindex="-1"></a><span class="fu">names</span>(a)</span>
<span id="cb129-4"><a href="mod-02.html#cb129-4" tabindex="-1"></a><span class="do">## [1] &quot;value1&quot; &quot;value2&quot; &quot;value3&quot;</span></span>
<span id="cb129-5"><a href="mod-02.html#cb129-5" tabindex="-1"></a>a</span>
<span id="cb129-6"><a href="mod-02.html#cb129-6" tabindex="-1"></a><span class="do">##    value1    value2    value3 </span></span>
<span id="cb129-7"><a href="mod-02.html#cb129-7" tabindex="-1"></a><span class="do">## 0.9575766 0.8877549 0.6399788</span></span>
<span id="cb129-8"><a href="mod-02.html#cb129-8" tabindex="-1"></a>a[<span class="st">&quot;value3&quot;</span>]</span>
<span id="cb129-9"><a href="mod-02.html#cb129-9" tabindex="-1"></a><span class="do">##    value3 </span></span>
<span id="cb129-10"><a href="mod-02.html#cb129-10" tabindex="-1"></a><span class="do">## 0.6399788</span></span></code></pre></div>
<p>Vectors can be extended using <code>c()</code> or by changing their length. Changing the length of a vector can also shorten it. If a vector is made longer than the number of values it has, the extra elements will be filled in with <code>NA</code>, which is the label R uses for blanks.</p>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb130-1"><a href="mod-02.html#cb130-1" tabindex="-1"></a>b <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span></span>
<span id="cb130-2"><a href="mod-02.html#cb130-2" tabindex="-1"></a>b <span class="ot">&lt;-</span> <span class="fu">c</span>(b, <span class="dv">10</span>)</span>
<span id="cb130-3"><a href="mod-02.html#cb130-3" tabindex="-1"></a>b</span>
<span id="cb130-4"><a href="mod-02.html#cb130-4" tabindex="-1"></a><span class="do">## [1]  1  2  3  4  5  6  7  8 10</span></span>
<span id="cb130-5"><a href="mod-02.html#cb130-5" tabindex="-1"></a>b <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span></span>
<span id="cb130-6"><a href="mod-02.html#cb130-6" tabindex="-1"></a><span class="fu">length</span>(b) <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb130-7"><a href="mod-02.html#cb130-7" tabindex="-1"></a>b</span>
<span id="cb130-8"><a href="mod-02.html#cb130-8" tabindex="-1"></a><span class="do">##  [1]  1  2  3  4  5  6  7  8 NA NA</span></span>
<span id="cb130-9"><a href="mod-02.html#cb130-9" tabindex="-1"></a>b <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span></span>
<span id="cb130-10"><a href="mod-02.html#cb130-10" tabindex="-1"></a><span class="fu">length</span>(b) <span class="ot">&lt;-</span> <span class="dv">6</span></span>
<span id="cb130-11"><a href="mod-02.html#cb130-11" tabindex="-1"></a>b</span>
<span id="cb130-12"><a href="mod-02.html#cb130-12" tabindex="-1"></a><span class="do">## [1] 1 2 3 4 5 6</span></span></code></pre></div>
<p>As mentioned elsewhere, many R functions take vectors as input and use them one element at a time. This is very handy when working with data, because you will often need to perform the same operation on every observation. Like any R data object, you can use a vector in a function simply by using the name in place of the values:</p>
<div class="sourceCode" id="cb131"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb131-1"><a href="mod-02.html#cb131-1" tabindex="-1"></a>my.vec <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span></span>
<span id="cb131-2"><a href="mod-02.html#cb131-2" tabindex="-1"></a>my.vec<span class="sc">*</span><span class="dv">2</span></span></code></pre></div>
<pre><code>## [1]  2  4  6  8 10 12 14 16</code></pre>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb133-1"><a href="mod-02.html#cb133-1" tabindex="-1"></a><span class="fu">mean</span>(my.vec)</span></code></pre></div>
<pre><code>## [1] 4.5</code></pre>
<div id="atomic-vectors" class="section level4 hasAnchor" number="2.6.1.1">
<h4><span class="header-section-number">2.6.1.1</span> Atomic vectors<a href="mod-02.html#atomic-vectors" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>A final note about vectors: some R functions, help pages, and error messages refer to “atomic vectors”. These are any object type whose elements cannot be split (i.e., “atomic” in the <del>pedantic</del>original sense). For example, the scalar <code>2</code> is considered atomic, being a numeric vector of length 1 whose sole element cannot be split up into something smaller. Likewise, the vector <code>2:3</code> is also atomic, because neither of its elements (<code>2</code> and <code>3</code>) can be split up into small elements. This definition extends to arrays with <span class="math inline">\(\geq3\)</span> dimensions.</p>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb135-1"><a href="mod-02.html#cb135-1" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="dv">4</span></span>
<span id="cb135-2"><a href="mod-02.html#cb135-2" tabindex="-1"></a>a</span></code></pre></div>
<pre><code>## [1] 4</code></pre>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb137-1"><a href="mod-02.html#cb137-1" tabindex="-1"></a><span class="fu">is.atomic</span>(a)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb139-1"><a href="mod-02.html#cb139-1" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">runif</span>(<span class="dv">10</span>)</span>
<span id="cb139-2"><a href="mod-02.html#cb139-2" tabindex="-1"></a>a</span></code></pre></div>
<pre><code>##  [1] 0.97096661 0.61883821 0.33342721 0.34674825 0.39848541 0.78469278
##  [7] 0.03893649 0.74879539 0.67727683 0.17126433</code></pre>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb141-1"><a href="mod-02.html#cb141-1" tabindex="-1"></a><span class="fu">is.atomic</span>(a)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb143"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb143-1"><a href="mod-02.html#cb143-1" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>, <span class="at">ncol=</span><span class="dv">3</span>)</span>
<span id="cb143-2"><a href="mod-02.html#cb143-2" tabindex="-1"></a>a</span></code></pre></div>
<pre><code>##      [,1] [,2] [,3]
## [1,]    1    3    5
## [2,]    2    4    6</code></pre>
<div class="sourceCode" id="cb145"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb145-1"><a href="mod-02.html#cb145-1" tabindex="-1"></a><span class="fu">is.atomic</span>(a)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb147-1"><a href="mod-02.html#cb147-1" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">array</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">12</span>, <span class="at">dim=</span><span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">2</span>,<span class="dv">2</span>))</span>
<span id="cb147-2"><a href="mod-02.html#cb147-2" tabindex="-1"></a>a</span></code></pre></div>
<pre><code>## , , 1
## 
##      [,1] [,2]
## [1,]    1    4
## [2,]    2    5
## [3,]    3    6
## 
## , , 2
## 
##      [,1] [,2]
## [1,]    7   10
## [2,]    8   11
## [3,]    9   12</code></pre>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb149-1"><a href="mod-02.html#cb149-1" tabindex="-1"></a><span class="fu">is.atomic</span>(a)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>On the other hand, a <strong>list</strong> is not atomic but rather <strong>recursive</strong>. This means that any of its elements can be another list. Consider the following:</p>
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb151-1"><a href="mod-02.html#cb151-1" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">x=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">12</span>, <span class="at">y=</span><span class="st">&quot;a&quot;</span>, <span class="at">z=</span><span class="fu">list</span>(<span class="at">m=</span><span class="fu">t.test</span>(<span class="fu">rnorm</span>(<span class="dv">20</span>)), <span class="at">n=</span><span class="st">&quot;a&quot;</span>))</span>
<span id="cb151-2"><a href="mod-02.html#cb151-2" tabindex="-1"></a>a</span></code></pre></div>
<pre><code>## $x
##  [1]  1  2  3  4  5  6  7  8  9 10 11 12
## 
## $y
## [1] &quot;a&quot;
## 
## $z
## $z$m
## 
##  One Sample t-test
## 
## data:  rnorm(20)
## t = -1.1843, df = 19, p-value = 0.2509
## alternative hypothesis: true mean is not equal to 0
## 95 percent confidence interval:
##  -0.7107776  0.1970939
## sample estimates:
##  mean of x 
## -0.2568418 
## 
## 
## $z$n
## [1] &quot;a&quot;</code></pre>
<div class="sourceCode" id="cb153"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb153-1"><a href="mod-02.html#cb153-1" tabindex="-1"></a><span class="fu">is.atomic</span>(a)</span></code></pre></div>
<pre><code>## [1] FALSE</code></pre>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb155-1"><a href="mod-02.html#cb155-1" tabindex="-1"></a><span class="fu">is.recursive</span>(a)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>The list <code>a</code> is recursive because any of its elements could be another list. However, consider the list <code>b</code>:</p>
<div class="sourceCode" id="cb157"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb157-1"><a href="mod-02.html#cb157-1" tabindex="-1"></a>b <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">x=</span><span class="dv">1</span>, <span class="at">y=</span><span class="dv">2</span>, <span class="at">z=</span><span class="dv">3</span>)</span>
<span id="cb157-2"><a href="mod-02.html#cb157-2" tabindex="-1"></a>b</span></code></pre></div>
<pre><code>## $x
## [1] 1
## 
## $y
## [1] 2
## 
## $z
## [1] 3</code></pre>
<div class="sourceCode" id="cb159"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb159-1"><a href="mod-02.html#cb159-1" tabindex="-1"></a><span class="fu">is.atomic</span>(b)</span></code></pre></div>
<pre><code>## [1] FALSE</code></pre>
<div class="sourceCode" id="cb161"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb161-1"><a href="mod-02.html#cb161-1" tabindex="-1"></a><span class="fu">is.recursive</span>(b)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>Note that while every element of <code>b</code> is atomic in and of itself (satisfying the definition of an atomic object above), <code>b</code> is in fact recursive. This is because any element of <code>b</code>, <strong><em>could</em></strong> be a list like <code>b</code>. This is a bit paradoxical, but not enough to cause any trouble.</p>
<p>Why do I bring this up? Because it is very easy to provoke the dreaded, <code>operator invalid for atomic vectors</code> error. This usually means you are trying to treat a non-list as a list.</p>
<div class="sourceCode" id="cb163"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb163-1"><a href="mod-02.html#cb163-1" tabindex="-1"></a><span class="co"># make a simple matrix</span></span>
<span id="cb163-2"><a href="mod-02.html#cb163-2" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="at">nrow=</span><span class="dv">2</span>)</span>
<span id="cb163-3"><a href="mod-02.html#cb163-3" tabindex="-1"></a><span class="fu">colnames</span>(a) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;a&quot;</span>, <span class="st">&quot;b&quot;</span>)</span>
<span id="cb163-4"><a href="mod-02.html#cb163-4" tabindex="-1"></a>a</span></code></pre></div>
<pre><code>##      a b
## [1,] 1 3
## [2,] 2 4</code></pre>
<div class="sourceCode" id="cb165"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb165-1"><a href="mod-02.html#cb165-1" tabindex="-1"></a><span class="co"># incorrect: using $ to extract from an atomic vector</span></span>
<span id="cb165-2"><a href="mod-02.html#cb165-2" tabindex="-1"></a>a<span class="sc">$</span>a</span></code></pre></div>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb166-1"><a href="mod-02.html#cb166-1" tabindex="-1"></a><span class="co"># correct: using column names and []</span></span>
<span id="cb166-2"><a href="mod-02.html#cb166-2" tabindex="-1"></a>a[,<span class="st">&quot;a&quot;</span>]</span></code></pre></div>
<pre><code>## [1] 1 2</code></pre>
<p>Atomic vectors can contain 1 and only one type of data: logical, integer, double, character, complex, and raw.</p>
</div>
</div>
<div id="data-frames" class="section level3 hasAnchor" number="2.6.2">
<h3><span class="header-section-number">2.6.2</span> Data frames<a href="mod-02.html#data-frames" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Most of the time when you work with data in R, you will work with objects of the <strong>data frame</strong> class. A data frame looks and acts a lot like a spreadsheet. One key aspect of data frames is that data frames can store data of more than one type, while vectors, matrices, and arrays cannot. The examples below use the built-in data frame <code>iris</code>.</p>
<p>Each <strong>row</strong> of a data frame usually corresponds to one <strong>observation</strong>. Each <strong>column</strong> contains the values of one <strong>variable</strong>. You can access data frame columns by name with the <code>$</code> operator:</p>
<div class="sourceCode" id="cb168"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb168-1"><a href="mod-02.html#cb168-1" tabindex="-1"></a>iris<span class="sc">$</span>Sepal.Length</span></code></pre></div>
<pre><code>##   [1] 5.1 4.9 4.7 4.6 5.0 5.4 4.6 5.0 4.4 4.9 5.4 4.8 4.8 4.3 5.8 5.7 5.4 5.1
##  [19] 5.7 5.1 5.4 5.1 4.6 5.1 4.8 5.0 5.0 5.2 5.2 4.7 4.8 5.4 5.2 5.5 4.9 5.0
##  [37] 5.5 4.9 4.4 5.1 5.0 4.5 4.4 5.0 5.1 4.8 5.1 4.6 5.3 5.0 7.0 6.4 6.9 5.5
##  [55] 6.5 5.7 6.3 4.9 6.6 5.2 5.0 5.9 6.0 6.1 5.6 6.7 5.6 5.8 6.2 5.6 5.9 6.1
##  [73] 6.3 6.1 6.4 6.6 6.8 6.7 6.0 5.7 5.5 5.5 5.8 6.0 5.4 6.0 6.7 6.3 5.6 5.5
##  [91] 5.5 6.1 5.8 5.0 5.6 5.7 5.7 6.2 5.1 5.7 6.3 5.8 7.1 6.3 6.5 7.6 4.9 7.3
## [109] 6.7 7.2 6.5 6.4 6.8 5.7 5.8 6.4 6.5 7.7 7.7 6.0 6.9 5.6 7.7 6.3 6.7 7.2
## [127] 6.2 6.1 6.4 7.2 7.4 7.9 6.4 6.3 6.1 7.7 6.3 6.4 6.0 6.9 6.7 6.9 5.8 6.8
## [145] 6.7 6.7 6.3 6.5 6.2 5.9</code></pre>
<p>You can also access data frame columns by number or name with bracket notation:</p>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb170-1"><a href="mod-02.html#cb170-1" tabindex="-1"></a>iris[,<span class="dv">1</span>]</span></code></pre></div>
<pre><code>##   [1] 5.1 4.9 4.7 4.6 5.0 5.4 4.6 5.0 4.4 4.9 5.4 4.8 4.8 4.3 5.8 5.7 5.4 5.1
##  [19] 5.7 5.1 5.4 5.1 4.6 5.1 4.8 5.0 5.0 5.2 5.2 4.7 4.8 5.4 5.2 5.5 4.9 5.0
##  [37] 5.5 4.9 4.4 5.1 5.0 4.5 4.4 5.0 5.1 4.8 5.1 4.6 5.3 5.0 7.0 6.4 6.9 5.5
##  [55] 6.5 5.7 6.3 4.9 6.6 5.2 5.0 5.9 6.0 6.1 5.6 6.7 5.6 5.8 6.2 5.6 5.9 6.1
##  [73] 6.3 6.1 6.4 6.6 6.8 6.7 6.0 5.7 5.5 5.5 5.8 6.0 5.4 6.0 6.7 6.3 5.6 5.5
##  [91] 5.5 6.1 5.8 5.0 5.6 5.7 5.7 6.2 5.1 5.7 6.3 5.8 7.1 6.3 6.5 7.6 4.9 7.3
## [109] 6.7 7.2 6.5 6.4 6.8 5.7 5.8 6.4 6.5 7.7 7.7 6.0 6.9 5.6 7.7 6.3 6.7 7.2
## [127] 6.2 6.1 6.4 7.2 7.4 7.9 6.4 6.3 6.1 7.7 6.3 6.4 6.0 6.9 6.7 6.9 5.8 6.8
## [145] 6.7 6.7 6.3 6.5 6.2 5.9</code></pre>
<div class="sourceCode" id="cb172"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb172-1"><a href="mod-02.html#cb172-1" tabindex="-1"></a>iris[,<span class="st">&quot;Petal.Width&quot;</span>]</span></code></pre></div>
<pre><code>##   [1] 0.2 0.2 0.2 0.2 0.2 0.4 0.3 0.2 0.2 0.1 0.2 0.2 0.1 0.1 0.2 0.4 0.4 0.3
##  [19] 0.3 0.3 0.2 0.4 0.2 0.5 0.2 0.2 0.4 0.2 0.2 0.2 0.2 0.4 0.1 0.2 0.2 0.2
##  [37] 0.2 0.1 0.2 0.2 0.3 0.3 0.2 0.6 0.4 0.3 0.2 0.2 0.2 0.2 1.4 1.5 1.5 1.3
##  [55] 1.5 1.3 1.6 1.0 1.3 1.4 1.0 1.5 1.0 1.4 1.3 1.4 1.5 1.0 1.5 1.1 1.8 1.3
##  [73] 1.5 1.2 1.3 1.4 1.4 1.7 1.5 1.0 1.1 1.0 1.2 1.6 1.5 1.6 1.5 1.3 1.3 1.3
##  [91] 1.2 1.4 1.2 1.0 1.3 1.2 1.3 1.3 1.1 1.3 2.5 1.9 2.1 1.8 2.2 2.1 1.7 1.8
## [109] 1.8 2.5 2.0 1.9 2.1 2.0 2.4 2.3 1.8 2.2 2.3 1.5 2.3 2.0 2.0 1.8 2.1 1.8
## [127] 1.8 1.8 2.1 1.6 1.9 2.0 2.2 1.5 1.4 2.3 2.4 1.8 1.8 2.1 2.4 2.3 1.9 2.3
## [145] 2.5 2.3 1.9 2.0 2.3 1.8</code></pre>
<p>Rows of a data frame are accessed by number with bracket notation. As seen above, you can also select using logical tests.</p>
<div class="sourceCode" id="cb174"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb174-1"><a href="mod-02.html#cb174-1" tabindex="-1"></a>iris[<span class="dv">1</span>,]</span>
<span id="cb174-2"><a href="mod-02.html#cb174-2" tabindex="-1"></a><span class="do">##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species</span></span>
<span id="cb174-3"><a href="mod-02.html#cb174-3" tabindex="-1"></a><span class="do">## 1          5.1         3.5          1.4         0.2  setosa</span></span></code></pre></div>
<div class="sourceCode" id="cb175"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb175-1"><a href="mod-02.html#cb175-1" tabindex="-1"></a><span class="co"># not run because output very long; try it on your machine!</span></span>
<span id="cb175-2"><a href="mod-02.html#cb175-2" tabindex="-1"></a>iris[<span class="fu">which</span>(iris<span class="sc">$</span>Species <span class="sc">==</span> <span class="st">&quot;setosa&quot;</span>),]</span></code></pre></div>
<p>Unlike matrices, rows and columns of data frames are something different. This is because under the surface, a data frame is really a list, and what looks like a column of a data frame is really an element of the list. So, a row of a data frame is really a new data frame with one row. When you work with data frames in R, the data frame class and its associated methods offer a convenient way to interact with the underlying list structure.</p>
<div class="sourceCode" id="cb176"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb176-1"><a href="mod-02.html#cb176-1" tabindex="-1"></a><span class="fu">class</span>(iris[<span class="dv">1</span>,])</span></code></pre></div>
<pre><code>## [1] &quot;data.frame&quot;</code></pre>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb178-1"><a href="mod-02.html#cb178-1" tabindex="-1"></a><span class="fu">class</span>(iris[,<span class="dv">1</span>])</span></code></pre></div>
<pre><code>## [1] &quot;numeric&quot;</code></pre>
<p>Practically, this means that if you select a column from a data frame you get a vector, and if you select a row (or rows) you get another data frame. So, if you want to operate on values within a row, you might need to use an <code>apply()</code> function.</p>
<div class="sourceCode" id="cb180"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb180-1"><a href="mod-02.html#cb180-1" tabindex="-1"></a><span class="fu">mean</span>(iris[<span class="dv">1</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>])           <span class="co"># apply function needed</span></span>
<span id="cb180-2"><a href="mod-02.html#cb180-2" tabindex="-1"></a><span class="do">## Warning in mean.default(iris[1, 1:4]): argument is not numeric or logical:</span></span>
<span id="cb180-3"><a href="mod-02.html#cb180-3" tabindex="-1"></a><span class="do">## returning NA</span></span>
<span id="cb180-4"><a href="mod-02.html#cb180-4" tabindex="-1"></a><span class="do">## [1] NA</span></span>
<span id="cb180-5"><a href="mod-02.html#cb180-5" tabindex="-1"></a><span class="fu">apply</span>(iris[<span class="dv">1</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>], <span class="dv">1</span>, mean) <span class="co"># works!</span></span>
<span id="cb180-6"><a href="mod-02.html#cb180-6" tabindex="-1"></a><span class="do">##    1 </span></span>
<span id="cb180-7"><a href="mod-02.html#cb180-7" tabindex="-1"></a><span class="do">## 2.55</span></span>
<span id="cb180-8"><a href="mod-02.html#cb180-8" tabindex="-1"></a><span class="fu">sum</span>(iris[<span class="dv">1</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>])            <span class="co"># works!</span></span>
<span id="cb180-9"><a href="mod-02.html#cb180-9" tabindex="-1"></a><span class="do">## [1] 10.2</span></span></code></pre></div>
<p>Columns can be added to data frames by assigning them a value:</p>
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb181-1"><a href="mod-02.html#cb181-1" tabindex="-1"></a>iris2 <span class="ot">&lt;-</span> iris   <span class="co"># Make a spare copy</span></span>
<span id="cb181-2"><a href="mod-02.html#cb181-2" tabindex="-1"></a>iris2<span class="sc">$</span>petal.area <span class="ot">&lt;-</span> iris2<span class="sc">$</span>Petal.Length<span class="sc">*</span>iris2<span class="sc">$</span>Petal.Width</span></code></pre></div>
<p>You can view the first few rows of a data frame with the command <code>head()</code>, and the last few rows with the command <code>tail()</code>. These are incredibly useful!</p>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="mod-02.html#cb182-1" tabindex="-1"></a><span class="fu">head</span>(iris2)</span></code></pre></div>
<pre><code>##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species petal.area
## 1          5.1         3.5          1.4         0.2  setosa       0.28
## 2          4.9         3.0          1.4         0.2  setosa       0.28
## 3          4.7         3.2          1.3         0.2  setosa       0.26
## 4          4.6         3.1          1.5         0.2  setosa       0.30
## 5          5.0         3.6          1.4         0.2  setosa       0.28
## 6          5.4         3.9          1.7         0.4  setosa       0.68</code></pre>
<div class="sourceCode" id="cb184"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb184-1"><a href="mod-02.html#cb184-1" tabindex="-1"></a><span class="fu">tail</span>(iris2)</span></code></pre></div>
<pre><code>##     Sepal.Length Sepal.Width Petal.Length Petal.Width   Species petal.area
## 145          6.7         3.3          5.7         2.5 virginica      14.25
## 146          6.7         3.0          5.2         2.3 virginica      11.96
## 147          6.3         2.5          5.0         1.9 virginica       9.50
## 148          6.5         3.0          5.2         2.0 virginica      10.40
## 149          6.2         3.4          5.4         2.3 virginica      12.42
## 150          5.9         3.0          5.1         1.8 virginica       9.18</code></pre>
<p>Columns can be removed from a data frame by assigning them a value of <code>NULL</code>.</p>
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb186-1"><a href="mod-02.html#cb186-1" tabindex="-1"></a>iris2<span class="sc">$</span>petal.area <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb186-2"><a href="mod-02.html#cb186-2" tabindex="-1"></a><span class="fu">head</span>(iris2)</span></code></pre></div>
<pre><code>##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 1          5.1         3.5          1.4         0.2  setosa
## 2          4.9         3.0          1.4         0.2  setosa
## 3          4.7         3.2          1.3         0.2  setosa
## 4          4.6         3.1          1.5         0.2  setosa
## 5          5.0         3.6          1.4         0.2  setosa
## 6          5.4         3.9          1.7         0.4  setosa</code></pre>
<p>We’ll explore more aspects of data frames in <a href="mod-03.html#mod-03-dataframe">Module 3</a>.</p>
</div>
<div id="matrices-and-arrays" class="section level3 hasAnchor" number="2.6.3">
<h3><span class="header-section-number">2.6.3</span> Matrices and arrays<a href="mod-02.html#matrices-and-arrays" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Sometimes you have a structured set of values that does not need to be stored as a data frame, or should not be stored as a data frame for ease of access. These situations call for matrices and arrays.</p>
<p>An <strong>array</strong> with two dimensions is a <strong>matrix</strong>. The dimensions of a matrix are its <strong>rows</strong> and <strong>columns</strong>. Matrices look a little like data frames, but internally they are very different. Unlike a data frame and like all arrays, a matrix can contain values of <em>only one type</em>. Matrices are necessary for any matrix operations (e.g., Leslie matrices) and for some function inputs. For most routine data analyses, data frames are easier to work with.</p>
<p>Matrices are created with the <code>matrix()</code> function. The matrix function must be provided with the values, the dimensions of the matrix, and the method for filling (by rows or by columns). If you don’t know the values yet, you can just use <code>NA</code> or 0.</p>
<div class="sourceCode" id="cb188"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb188-1"><a href="mod-02.html#cb188-1" tabindex="-1"></a>my.mat1 <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">12</span>,<span class="at">nrow=</span><span class="dv">3</span>,<span class="at">ncol=</span><span class="dv">4</span>)</span>
<span id="cb188-2"><a href="mod-02.html#cb188-2" tabindex="-1"></a>my.mat1</span>
<span id="cb188-3"><a href="mod-02.html#cb188-3" tabindex="-1"></a><span class="do">##      [,1] [,2] [,3] [,4]</span></span>
<span id="cb188-4"><a href="mod-02.html#cb188-4" tabindex="-1"></a><span class="do">## [1,]    1    4    7   10</span></span>
<span id="cb188-5"><a href="mod-02.html#cb188-5" tabindex="-1"></a><span class="do">## [2,]    2    5    8   11</span></span>
<span id="cb188-6"><a href="mod-02.html#cb188-6" tabindex="-1"></a><span class="do">## [3,]    3    6    9   12</span></span>
<span id="cb188-7"><a href="mod-02.html#cb188-7" tabindex="-1"></a></span>
<span id="cb188-8"><a href="mod-02.html#cb188-8" tabindex="-1"></a>my.mat2 <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">12</span>,<span class="at">nrow=</span><span class="dv">3</span>,<span class="at">ncol=</span><span class="dv">4</span>,<span class="at">byrow=</span><span class="cn">TRUE</span>)</span>
<span id="cb188-9"><a href="mod-02.html#cb188-9" tabindex="-1"></a>my.mat2</span>
<span id="cb188-10"><a href="mod-02.html#cb188-10" tabindex="-1"></a><span class="do">##      [,1] [,2] [,3] [,4]</span></span>
<span id="cb188-11"><a href="mod-02.html#cb188-11" tabindex="-1"></a><span class="do">## [1,]    1    2    3    4</span></span>
<span id="cb188-12"><a href="mod-02.html#cb188-12" tabindex="-1"></a><span class="do">## [2,]    5    6    7    8</span></span>
<span id="cb188-13"><a href="mod-02.html#cb188-13" tabindex="-1"></a><span class="do">## [3,]    9   10   11   12</span></span>
<span id="cb188-14"><a href="mod-02.html#cb188-14" tabindex="-1"></a></span>
<span id="cb188-15"><a href="mod-02.html#cb188-15" tabindex="-1"></a>my.mat3 <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="cn">NA</span>,<span class="at">nrow=</span><span class="dv">3</span>,<span class="at">ncol=</span><span class="dv">4</span>)</span>
<span id="cb188-16"><a href="mod-02.html#cb188-16" tabindex="-1"></a>my.mat3</span>
<span id="cb188-17"><a href="mod-02.html#cb188-17" tabindex="-1"></a><span class="do">##      [,1] [,2] [,3] [,4]</span></span>
<span id="cb188-18"><a href="mod-02.html#cb188-18" tabindex="-1"></a><span class="do">## [1,]   NA   NA   NA   NA</span></span>
<span id="cb188-19"><a href="mod-02.html#cb188-19" tabindex="-1"></a><span class="do">## [2,]   NA   NA   NA   NA</span></span>
<span id="cb188-20"><a href="mod-02.html#cb188-20" tabindex="-1"></a><span class="do">## [3,]   NA   NA   NA   NA</span></span></code></pre></div>
<p>Just as with other objects, the values in a matrix can be extracted using bracket notation. Because matrices have two dimensions, you must specify indices for both dimensions. This is done using a comma <code>,</code> to separate the row indices and column indices within the brackets. Just as with data frames, row indices are before the comma; column indices are after the comma. Leaving the row index or the column index blank will select all rows or columns, respectively.</p>
<div class="sourceCode" id="cb189"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb189-1"><a href="mod-02.html#cb189-1" tabindex="-1"></a>my.mat1[<span class="dv">2</span>,<span class="dv">1</span>]<span class="co"># row 2, column 1</span></span></code></pre></div>
<pre><code>## [1] 2</code></pre>
<div class="sourceCode" id="cb191"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb191-1"><a href="mod-02.html#cb191-1" tabindex="-1"></a>my.mat1[<span class="dv">3</span>,<span class="dv">3</span>]<span class="co"># row 3, column 3</span></span></code></pre></div>
<pre><code>## [1] 9</code></pre>
<p>Entire rows or columns of a matrix can be extracted by leaving the other dimension blank. You must still include the comma within the brackets. Notice that both the rows and columns of a matrix are vectors, not a matrix. If you want to get a row or column as a 1 row or 1 column matrix, you must specifically request this.</p>
<div class="sourceCode" id="cb193"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb193-1"><a href="mod-02.html#cb193-1" tabindex="-1"></a><span class="co"># first row</span></span>
<span id="cb193-2"><a href="mod-02.html#cb193-2" tabindex="-1"></a>my.mat1[<span class="dv">1</span>,]</span>
<span id="cb193-3"><a href="mod-02.html#cb193-3" tabindex="-1"></a><span class="do">## [1]  1  4  7 10</span></span>
<span id="cb193-4"><a href="mod-02.html#cb193-4" tabindex="-1"></a><span class="fu">is.matrix</span>(my.mat1[<span class="dv">1</span>,])</span>
<span id="cb193-5"><a href="mod-02.html#cb193-5" tabindex="-1"></a><span class="do">## [1] FALSE</span></span>
<span id="cb193-6"><a href="mod-02.html#cb193-6" tabindex="-1"></a></span>
<span id="cb193-7"><a href="mod-02.html#cb193-7" tabindex="-1"></a><span class="co"># third column</span></span>
<span id="cb193-8"><a href="mod-02.html#cb193-8" tabindex="-1"></a>my.mat1[,<span class="dv">3</span>]</span>
<span id="cb193-9"><a href="mod-02.html#cb193-9" tabindex="-1"></a><span class="do">## [1] 7 8 9</span></span>
<span id="cb193-10"><a href="mod-02.html#cb193-10" tabindex="-1"></a></span>
<span id="cb193-11"><a href="mod-02.html#cb193-11" tabindex="-1"></a><span class="co"># convert extracted part to matrix</span></span>
<span id="cb193-12"><a href="mod-02.html#cb193-12" tabindex="-1"></a>my.mat3 <span class="ot">&lt;-</span> <span class="fu">matrix</span>(my.mat1[<span class="dv">1</span>,], <span class="at">nrow=</span><span class="dv">1</span>)</span>
<span id="cb193-13"><a href="mod-02.html#cb193-13" tabindex="-1"></a><span class="fu">is.matrix</span>(my.mat3)</span>
<span id="cb193-14"><a href="mod-02.html#cb193-14" tabindex="-1"></a><span class="do">## [1] TRUE</span></span></code></pre></div>
<p>An array in R is usually only called an array if it has <span class="math inline">\(\ge\)</span> 3 dimensions. Data stored this way can make programming easier, if there is a clear relationship between the dimensions. Arrays are not often encountered by the average user. The example below has 3 “layers”, each of which has 4 rows and 2 columns.</p>
<div class="sourceCode" id="cb194"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb194-1"><a href="mod-02.html#cb194-1" tabindex="-1"></a>my.array <span class="ot">&lt;-</span> <span class="fu">array</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">24</span>,<span class="at">dim=</span><span class="fu">c</span>(<span class="dv">4</span>,<span class="dv">2</span>,<span class="dv">3</span>))</span>
<span id="cb194-2"><a href="mod-02.html#cb194-2" tabindex="-1"></a>my.array</span></code></pre></div>
<pre><code>## , , 1
## 
##      [,1] [,2]
## [1,]    1    5
## [2,]    2    6
## [3,]    3    7
## [4,]    4    8
## 
## , , 2
## 
##      [,1] [,2]
## [1,]    9   13
## [2,]   10   14
## [3,]   11   15
## [4,]   12   16
## 
## , , 3
## 
##      [,1] [,2]
## [1,]   17   21
## [2,]   18   22
## [3,]   19   23
## [4,]   20   24</code></pre>
<p>One very nifty (and potentially confusing) ability of arrays is the ability to be sliced into differently shaped pieces<a href="#fn35" class="footnote-ref" id="fnref35"><sup>35</sup></a>. Consider the examples below and see if you can work out why the results are what they are.</p>
<div class="sourceCode" id="cb196"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb196-1"><a href="mod-02.html#cb196-1" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb196-2"><a href="mod-02.html#cb196-2" tabindex="-1"></a>my.array[,,<span class="dv">1</span>]</span>
<span id="cb196-3"><a href="mod-02.html#cb196-3" tabindex="-1"></a><span class="do">##      [,1] [,2]</span></span>
<span id="cb196-4"><a href="mod-02.html#cb196-4" tabindex="-1"></a><span class="do">## [1,]    1    5</span></span>
<span id="cb196-5"><a href="mod-02.html#cb196-5" tabindex="-1"></a><span class="do">## [2,]    2    6</span></span>
<span id="cb196-6"><a href="mod-02.html#cb196-6" tabindex="-1"></a><span class="do">## [3,]    3    7</span></span>
<span id="cb196-7"><a href="mod-02.html#cb196-7" tabindex="-1"></a><span class="do">## [4,]    4    8</span></span>
<span id="cb196-8"><a href="mod-02.html#cb196-8" tabindex="-1"></a>my.array[,<span class="dv">1</span>,]</span>
<span id="cb196-9"><a href="mod-02.html#cb196-9" tabindex="-1"></a><span class="do">##      [,1] [,2] [,3]</span></span>
<span id="cb196-10"><a href="mod-02.html#cb196-10" tabindex="-1"></a><span class="do">## [1,]    1    9   17</span></span>
<span id="cb196-11"><a href="mod-02.html#cb196-11" tabindex="-1"></a><span class="do">## [2,]    2   10   18</span></span>
<span id="cb196-12"><a href="mod-02.html#cb196-12" tabindex="-1"></a><span class="do">## [3,]    3   11   19</span></span>
<span id="cb196-13"><a href="mod-02.html#cb196-13" tabindex="-1"></a><span class="do">## [4,]    4   12   20</span></span>
<span id="cb196-14"><a href="mod-02.html#cb196-14" tabindex="-1"></a>my.array[<span class="dv">1</span>,,]</span>
<span id="cb196-15"><a href="mod-02.html#cb196-15" tabindex="-1"></a><span class="do">##      [,1] [,2] [,3]</span></span>
<span id="cb196-16"><a href="mod-02.html#cb196-16" tabindex="-1"></a><span class="do">## [1,]    1    9   17</span></span>
<span id="cb196-17"><a href="mod-02.html#cb196-17" tabindex="-1"></a><span class="do">## [2,]    5   13   21</span></span>
<span id="cb196-18"><a href="mod-02.html#cb196-18" tabindex="-1"></a>my.array[<span class="dv">1</span>,<span class="dv">2</span>,]</span>
<span id="cb196-19"><a href="mod-02.html#cb196-19" tabindex="-1"></a><span class="do">## [1]  5 13 21</span></span>
<span id="cb196-20"><a href="mod-02.html#cb196-20" tabindex="-1"></a>my.array[,<span class="dv">2</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb196-21"><a href="mod-02.html#cb196-21" tabindex="-1"></a><span class="do">##      [,1] [,2]</span></span>
<span id="cb196-22"><a href="mod-02.html#cb196-22" tabindex="-1"></a><span class="do">## [1,]    5   13</span></span>
<span id="cb196-23"><a href="mod-02.html#cb196-23" tabindex="-1"></a><span class="do">## [2,]    6   14</span></span>
<span id="cb196-24"><a href="mod-02.html#cb196-24" tabindex="-1"></a><span class="do">## [3,]    7   15</span></span>
<span id="cb196-25"><a href="mod-02.html#cb196-25" tabindex="-1"></a><span class="do">## [4,]    8   16</span></span></code></pre></div>
</div>
<div id="lists" class="section level3 hasAnchor" number="2.6.4">
<h3><span class="header-section-number">2.6.4</span> Lists<a href="mod-02.html#lists" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>One of the most important kinds of R object is the <strong>list</strong>. A list can be thought of like a <em>series of containers or buckets</em>. Each bucket can contain completely unrelated things, or even other series of buckets. Because of this flexibility, lists are extremely versatile and useful in R programming. Many function outputs are really lists (e.g., the outputs of <code>lm()</code> or <code>t.test()</code>). Data frames are also lists.</p>
<p>Lists can created using the function <code>vector()</code> and then filled later. The following example creates an empty list with 4 elements (i.e., buckets), then fills each element with something different.</p>
<div class="sourceCode" id="cb197"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb197-1"><a href="mod-02.html#cb197-1" tabindex="-1"></a>my.list <span class="ot">&lt;-</span> <span class="fu">vector</span>(<span class="st">&quot;list&quot;</span>,<span class="at">length=</span><span class="dv">4</span>)</span>
<span id="cb197-2"><a href="mod-02.html#cb197-2" tabindex="-1"></a>my.list[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span></span>
<span id="cb197-3"><a href="mod-02.html#cb197-3" tabindex="-1"></a>my.list[[<span class="dv">2</span>]] <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;a&quot;</span>,<span class="st">&quot;b&quot;</span>,<span class="st">&quot;c&quot;</span>)</span>
<span id="cb197-4"><a href="mod-02.html#cb197-4" tabindex="-1"></a>my.list[[<span class="dv">3</span>]] <span class="ot">&lt;-</span> <span class="fu">t.test</span>(iris<span class="sc">$</span>Sepal.Length)</span>
<span id="cb197-5"><a href="mod-02.html#cb197-5" tabindex="-1"></a>my.list[[<span class="dv">4</span>]] <span class="ot">&lt;-</span> iris</span>
<span id="cb197-6"><a href="mod-02.html#cb197-6" tabindex="-1"></a>my.list</span></code></pre></div>
<pre><code>## [[1]]
##  [1]  1  2  3  4  5  6  7  8  9 10
## 
## [[2]]
## [1] &quot;a&quot; &quot;b&quot; &quot;c&quot;
## 
## [[3]]
## 
##  One Sample t-test
## 
## data:  iris$Sepal.Length
## t = 86.425, df = 149, p-value &lt; 2.2e-16
## alternative hypothesis: true mean is not equal to 0
## 95 percent confidence interval:
##  5.709732 5.976934
## sample estimates:
## mean of x 
##  5.843333 
## 
## 
## [[4]]
##     Sepal.Length Sepal.Width Petal.Length Petal.Width    Species
## 1            5.1         3.5          1.4         0.2     setosa
## 2            4.9         3.0          1.4         0.2     setosa
## 3            4.7         3.2          1.3         0.2     setosa
## 4            4.6         3.1          1.5         0.2     setosa
## 5            5.0         3.6          1.4         0.2     setosa
## 6            5.4         3.9          1.7         0.4     setosa
## 7            4.6         3.4          1.4         0.3     setosa
## 8            5.0         3.4          1.5         0.2     setosa
## 9            4.4         2.9          1.4         0.2     setosa
## 10           4.9         3.1          1.5         0.1     setosa
## 11           5.4         3.7          1.5         0.2     setosa
## 12           4.8         3.4          1.6         0.2     setosa
## 13           4.8         3.0          1.4         0.1     setosa
## 14           4.3         3.0          1.1         0.1     setosa
## 15           5.8         4.0          1.2         0.2     setosa
## 16           5.7         4.4          1.5         0.4     setosa
## 17           5.4         3.9          1.3         0.4     setosa
## 18           5.1         3.5          1.4         0.3     setosa
## 19           5.7         3.8          1.7         0.3     setosa
## 20           5.1         3.8          1.5         0.3     setosa
## 21           5.4         3.4          1.7         0.2     setosa
## 22           5.1         3.7          1.5         0.4     setosa
## 23           4.6         3.6          1.0         0.2     setosa
## 24           5.1         3.3          1.7         0.5     setosa
## 25           4.8         3.4          1.9         0.2     setosa
## 26           5.0         3.0          1.6         0.2     setosa
## 27           5.0         3.4          1.6         0.4     setosa
## 28           5.2         3.5          1.5         0.2     setosa
## 29           5.2         3.4          1.4         0.2     setosa
## 30           4.7         3.2          1.6         0.2     setosa
## 31           4.8         3.1          1.6         0.2     setosa
## 32           5.4         3.4          1.5         0.4     setosa
## 33           5.2         4.1          1.5         0.1     setosa
## 34           5.5         4.2          1.4         0.2     setosa
## 35           4.9         3.1          1.5         0.2     setosa
## 36           5.0         3.2          1.2         0.2     setosa
## 37           5.5         3.5          1.3         0.2     setosa
## 38           4.9         3.6          1.4         0.1     setosa
## 39           4.4         3.0          1.3         0.2     setosa
## 40           5.1         3.4          1.5         0.2     setosa
## 41           5.0         3.5          1.3         0.3     setosa
## 42           4.5         2.3          1.3         0.3     setosa
## 43           4.4         3.2          1.3         0.2     setosa
## 44           5.0         3.5          1.6         0.6     setosa
## 45           5.1         3.8          1.9         0.4     setosa
## 46           4.8         3.0          1.4         0.3     setosa
## 47           5.1         3.8          1.6         0.2     setosa
## 48           4.6         3.2          1.4         0.2     setosa
## 49           5.3         3.7          1.5         0.2     setosa
## 50           5.0         3.3          1.4         0.2     setosa
## 51           7.0         3.2          4.7         1.4 versicolor
## 52           6.4         3.2          4.5         1.5 versicolor
## 53           6.9         3.1          4.9         1.5 versicolor
## 54           5.5         2.3          4.0         1.3 versicolor
## 55           6.5         2.8          4.6         1.5 versicolor
## 56           5.7         2.8          4.5         1.3 versicolor
## 57           6.3         3.3          4.7         1.6 versicolor
## 58           4.9         2.4          3.3         1.0 versicolor
## 59           6.6         2.9          4.6         1.3 versicolor
## 60           5.2         2.7          3.9         1.4 versicolor
## 61           5.0         2.0          3.5         1.0 versicolor
## 62           5.9         3.0          4.2         1.5 versicolor
## 63           6.0         2.2          4.0         1.0 versicolor
## 64           6.1         2.9          4.7         1.4 versicolor
## 65           5.6         2.9          3.6         1.3 versicolor
## 66           6.7         3.1          4.4         1.4 versicolor
## 67           5.6         3.0          4.5         1.5 versicolor
## 68           5.8         2.7          4.1         1.0 versicolor
## 69           6.2         2.2          4.5         1.5 versicolor
## 70           5.6         2.5          3.9         1.1 versicolor
## 71           5.9         3.2          4.8         1.8 versicolor
## 72           6.1         2.8          4.0         1.3 versicolor
## 73           6.3         2.5          4.9         1.5 versicolor
## 74           6.1         2.8          4.7         1.2 versicolor
## 75           6.4         2.9          4.3         1.3 versicolor
## 76           6.6         3.0          4.4         1.4 versicolor
## 77           6.8         2.8          4.8         1.4 versicolor
## 78           6.7         3.0          5.0         1.7 versicolor
## 79           6.0         2.9          4.5         1.5 versicolor
## 80           5.7         2.6          3.5         1.0 versicolor
## 81           5.5         2.4          3.8         1.1 versicolor
## 82           5.5         2.4          3.7         1.0 versicolor
## 83           5.8         2.7          3.9         1.2 versicolor
## 84           6.0         2.7          5.1         1.6 versicolor
## 85           5.4         3.0          4.5         1.5 versicolor
## 86           6.0         3.4          4.5         1.6 versicolor
## 87           6.7         3.1          4.7         1.5 versicolor
## 88           6.3         2.3          4.4         1.3 versicolor
## 89           5.6         3.0          4.1         1.3 versicolor
## 90           5.5         2.5          4.0         1.3 versicolor
## 91           5.5         2.6          4.4         1.2 versicolor
## 92           6.1         3.0          4.6         1.4 versicolor
## 93           5.8         2.6          4.0         1.2 versicolor
## 94           5.0         2.3          3.3         1.0 versicolor
## 95           5.6         2.7          4.2         1.3 versicolor
## 96           5.7         3.0          4.2         1.2 versicolor
## 97           5.7         2.9          4.2         1.3 versicolor
## 98           6.2         2.9          4.3         1.3 versicolor
## 99           5.1         2.5          3.0         1.1 versicolor
## 100          5.7         2.8          4.1         1.3 versicolor
## 101          6.3         3.3          6.0         2.5  virginica
## 102          5.8         2.7          5.1         1.9  virginica
## 103          7.1         3.0          5.9         2.1  virginica
## 104          6.3         2.9          5.6         1.8  virginica
## 105          6.5         3.0          5.8         2.2  virginica
## 106          7.6         3.0          6.6         2.1  virginica
## 107          4.9         2.5          4.5         1.7  virginica
## 108          7.3         2.9          6.3         1.8  virginica
## 109          6.7         2.5          5.8         1.8  virginica
## 110          7.2         3.6          6.1         2.5  virginica
## 111          6.5         3.2          5.1         2.0  virginica
## 112          6.4         2.7          5.3         1.9  virginica
## 113          6.8         3.0          5.5         2.1  virginica
## 114          5.7         2.5          5.0         2.0  virginica
## 115          5.8         2.8          5.1         2.4  virginica
## 116          6.4         3.2          5.3         2.3  virginica
## 117          6.5         3.0          5.5         1.8  virginica
## 118          7.7         3.8          6.7         2.2  virginica
## 119          7.7         2.6          6.9         2.3  virginica
## 120          6.0         2.2          5.0         1.5  virginica
## 121          6.9         3.2          5.7         2.3  virginica
## 122          5.6         2.8          4.9         2.0  virginica
## 123          7.7         2.8          6.7         2.0  virginica
## 124          6.3         2.7          4.9         1.8  virginica
## 125          6.7         3.3          5.7         2.1  virginica
## 126          7.2         3.2          6.0         1.8  virginica
## 127          6.2         2.8          4.8         1.8  virginica
## 128          6.1         3.0          4.9         1.8  virginica
## 129          6.4         2.8          5.6         2.1  virginica
## 130          7.2         3.0          5.8         1.6  virginica
## 131          7.4         2.8          6.1         1.9  virginica
## 132          7.9         3.8          6.4         2.0  virginica
## 133          6.4         2.8          5.6         2.2  virginica
## 134          6.3         2.8          5.1         1.5  virginica
## 135          6.1         2.6          5.6         1.4  virginica
## 136          7.7         3.0          6.1         2.3  virginica
## 137          6.3         3.4          5.6         2.4  virginica
## 138          6.4         3.1          5.5         1.8  virginica
## 139          6.0         3.0          4.8         1.8  virginica
## 140          6.9         3.1          5.4         2.1  virginica
## 141          6.7         3.1          5.6         2.4  virginica
## 142          6.9         3.1          5.1         2.3  virginica
## 143          5.8         2.7          5.1         1.9  virginica
## 144          6.8         3.2          5.9         2.3  virginica
## 145          6.7         3.3          5.7         2.5  virginica
## 146          6.7         3.0          5.2         2.3  virginica
## 147          6.3         2.5          5.0         1.9  virginica
## 148          6.5         3.0          5.2         2.0  virginica
## 149          6.2         3.4          5.4         2.3  virginica
## 150          5.9         3.0          5.1         1.8  virginica</code></pre>
<p>Another way to make a list is to make its elements first and combine them into a list, with function <code>list()</code>.</p>
<div class="sourceCode" id="cb199"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb199-1"><a href="mod-02.html#cb199-1" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">runif</span>(<span class="dv">10</span>)</span>
<span id="cb199-2"><a href="mod-02.html#cb199-2" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="st">&quot;zebra&quot;</span></span>
<span id="cb199-3"><a href="mod-02.html#cb199-3" tabindex="-1"></a>z <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">12</span>, <span class="at">nrow=</span><span class="dv">3</span>)</span>
<span id="cb199-4"><a href="mod-02.html#cb199-4" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">list</span>(x, y, z)</span>
<span id="cb199-5"><a href="mod-02.html#cb199-5" tabindex="-1"></a>a</span></code></pre></div>
<pre><code>## [[1]]
##  [1] 0.7439746 0.6191592 0.6262453 0.2171577 0.2165673 0.3889450 0.9424557
##  [8] 0.9626080 0.7398553 0.7332459
## 
## [[2]]
## [1] &quot;zebra&quot;
## 
## [[3]]
##      [,1] [,2] [,3] [,4]
## [1,]    1    4    7   10
## [2,]    2    5    8   11
## [3,]    3    6    9   12</code></pre>
<p>Optionally, you can name each element as you make the list.</p>
<div class="sourceCode" id="cb201"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb201-1"><a href="mod-02.html#cb201-1" tabindex="-1"></a>b <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">x=</span>x, <span class="at">y=</span>y, <span class="at">z=</span>z)</span>
<span id="cb201-2"><a href="mod-02.html#cb201-2" tabindex="-1"></a>b</span></code></pre></div>
<pre><code>## $x
##  [1] 0.7439746 0.6191592 0.6262453 0.2171577 0.2165673 0.3889450 0.9424557
##  [8] 0.9626080 0.7398553 0.7332459
## 
## $y
## [1] &quot;zebra&quot;
## 
## $z
##      [,1] [,2] [,3] [,4]
## [1,]    1    4    7   10
## [2,]    2    5    8   11
## [3,]    3    6    9   12</code></pre>
<p>Much like vectors, lists can be extended by using concatenation:</p>
<div class="sourceCode" id="cb203"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb203-1"><a href="mod-02.html#cb203-1" tabindex="-1"></a>a2 <span class="ot">&lt;-</span> <span class="fu">c</span>(a,b)</span>
<span id="cb203-2"><a href="mod-02.html#cb203-2" tabindex="-1"></a>a2</span></code></pre></div>
<pre><code>## [[1]]
##  [1] 0.7439746 0.6191592 0.6262453 0.2171577 0.2165673 0.3889450 0.9424557
##  [8] 0.9626080 0.7398553 0.7332459
## 
## [[2]]
## [1] &quot;zebra&quot;
## 
## [[3]]
##      [,1] [,2] [,3] [,4]
## [1,]    1    4    7   10
## [2,]    2    5    8   11
## [3,]    3    6    9   12
## 
## $x
##  [1] 0.7439746 0.6191592 0.6262453 0.2171577 0.2165673 0.3889450 0.9424557
##  [8] 0.9626080 0.7398553 0.7332459
## 
## $y
## [1] &quot;zebra&quot;
## 
## $z
##      [,1] [,2] [,3] [,4]
## [1,]    1    4    7   10
## [2,]    2    5    8   11
## [3,]    3    6    9   12</code></pre>
<p>Single elements of a list can be accessed with <strong>double brackets</strong> <code>[[]]</code>. As with other objects, elements can be selected by name or by index. If selecting by name, the <code>$</code> symbol can be used instead of double brackets.</p>
<div class="sourceCode" id="cb205"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb205-1"><a href="mod-02.html#cb205-1" tabindex="-1"></a>a[[<span class="dv">1</span>]]</span>
<span id="cb205-2"><a href="mod-02.html#cb205-2" tabindex="-1"></a><span class="do">##  [1] 0.7439746 0.6191592 0.6262453 0.2171577 0.2165673 0.3889450 0.9424557</span></span>
<span id="cb205-3"><a href="mod-02.html#cb205-3" tabindex="-1"></a><span class="do">##  [8] 0.9626080 0.7398553 0.7332459</span></span>
<span id="cb205-4"><a href="mod-02.html#cb205-4" tabindex="-1"></a>b[[<span class="dv">2</span>]]</span>
<span id="cb205-5"><a href="mod-02.html#cb205-5" tabindex="-1"></a><span class="do">## [1] &quot;zebra&quot;</span></span>
<span id="cb205-6"><a href="mod-02.html#cb205-6" tabindex="-1"></a>b[[<span class="st">&quot;z&quot;</span>]]</span>
<span id="cb205-7"><a href="mod-02.html#cb205-7" tabindex="-1"></a><span class="do">##      [,1] [,2] [,3] [,4]</span></span>
<span id="cb205-8"><a href="mod-02.html#cb205-8" tabindex="-1"></a><span class="do">## [1,]    1    4    7   10</span></span>
<span id="cb205-9"><a href="mod-02.html#cb205-9" tabindex="-1"></a><span class="do">## [2,]    2    5    8   11</span></span>
<span id="cb205-10"><a href="mod-02.html#cb205-10" tabindex="-1"></a><span class="do">## [3,]    3    6    9   12</span></span>
<span id="cb205-11"><a href="mod-02.html#cb205-11" tabindex="-1"></a>b<span class="sc">$</span>z</span>
<span id="cb205-12"><a href="mod-02.html#cb205-12" tabindex="-1"></a><span class="do">##      [,1] [,2] [,3] [,4]</span></span>
<span id="cb205-13"><a href="mod-02.html#cb205-13" tabindex="-1"></a><span class="do">## [1,]    1    4    7   10</span></span>
<span id="cb205-14"><a href="mod-02.html#cb205-14" tabindex="-1"></a><span class="do">## [2,]    2    5    8   11</span></span>
<span id="cb205-15"><a href="mod-02.html#cb205-15" tabindex="-1"></a><span class="do">## [3,]    3    6    9   12</span></span></code></pre></div>
<p>Using <strong>single brackets</strong> to access a list will return a <em>list</em> of the requested elements, not the requested elements themselves. This can be a little confusing. Compare the following two commands and notice that the first returns a vector containing the numbers 1 through 10 (the first element of <code>my.list</code>), while the second returns a list with 1 element, and that element is a vector with the numbers 1 through 10.</p>
<div class="sourceCode" id="cb206"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb206-1"><a href="mod-02.html#cb206-1" tabindex="-1"></a>my.list[[<span class="dv">1</span>]]</span>
<span id="cb206-2"><a href="mod-02.html#cb206-2" tabindex="-1"></a><span class="do">##  [1]  1  2  3  4  5  6  7  8  9 10</span></span>
<span id="cb206-3"><a href="mod-02.html#cb206-3" tabindex="-1"></a>my.list[<span class="dv">1</span>]</span>
<span id="cb206-4"><a href="mod-02.html#cb206-4" tabindex="-1"></a><span class="do">## [[1]]</span></span>
<span id="cb206-5"><a href="mod-02.html#cb206-5" tabindex="-1"></a><span class="do">##  [1]  1  2  3  4  5  6  7  8  9 10</span></span></code></pre></div>
<p>The reason for this confusion is that the single brackets notation allows you to extract multiple elements of the list at once to make a new list.</p>
<div class="sourceCode" id="cb207"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb207-1"><a href="mod-02.html#cb207-1" tabindex="-1"></a>a2 <span class="ot">&lt;-</span> a[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb207-2"><a href="mod-02.html#cb207-2" tabindex="-1"></a>a2</span></code></pre></div>
<pre><code>## [[1]]
##  [1] 0.7439746 0.6191592 0.6262453 0.2171577 0.2165673 0.3889450 0.9424557
##  [8] 0.9626080 0.7398553 0.7332459
## 
## [[2]]
## [1] &quot;zebra&quot;</code></pre>
</div>
<div id="s3-s4-and-other-paradigms" class="section level3 hasAnchor" number="2.6.5">
<h3><span class="header-section-number">2.6.5</span> S3, S4, and other paradigms<a href="mod-02.html#s3-s4-and-other-paradigms" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We’ve seen so far that R is object-oriented. What is little appreciated is that R has at least three object-oriented paradigms: <strong>S3</strong>, <strong>S4</strong>, and <strong>R5</strong>. Most objects in R are S3 objects: data frames, matrices, vectors, etc. The R5 objects are more formally called “Reference Classes” and are still being developed. The S4 objects are only occasionally encountered by the average user, although some R developers believe that S4 objects should replace S3 in general use. Their reasons are esoteric and we won’t get into them here. In a nutshell, S4 classes are structured more formally than S3 classes and are better for validation and code testing.</p>
<p>One key difference in practice between S4 and S3 objects is that S4 objects contain <strong>slots</strong> or <strong>attributes</strong>, which are accessed using <code>@</code> instead of <code>$</code>. Within each slot, an S4 object can contain all sorts of data, often in list form. For example, the <code>SpatialPolygonsDataFrame</code> class in package <code>sp</code>, the R manifestation of an ESRI shapefile, is an S4 object.</p>
</div>
</div>
<div id="r-data-types" class="section level2 hasAnchor" number="2.7">
<h2><span class="header-section-number">2.7</span> R data types<a href="mod-02.html#r-data-types" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In R, data are stored in objects with a well-defined structure. The structures of these objects can facilitate manipulating and retrieving your data. The structure of an R object is determined by its class. The format of data (i.e., the kinds of values contained in an object) is referred to as the data <strong>type</strong>. Most R users only encounter types when trying to understand error messages. This is because many R functions will return errors if their inputs are not of the correct type.</p>
<div id="character-type" class="section level3 hasAnchor" number="2.7.1">
<h3><span class="header-section-number">2.7.1</span> Character type<a href="mod-02.html#character-type" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>Character</strong> data are text strings. They are input and printed using quotation marks (<code>""</code>). Note that straight quotes are used. No other kind of quotation mark is valid in R code. If you paste in text from another program that has “curly quotes” (e.g., from Word), you will get an error.</p>
<div class="sourceCode" id="cb209"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb209-1"><a href="mod-02.html#cb209-1" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;aardvark&quot;</span>, <span class="st">&quot;bunny&quot;</span>, <span class="st">&quot;caterpillar&quot;</span>)</span>
<span id="cb209-2"><a href="mod-02.html#cb209-2" tabindex="-1"></a>a</span></code></pre></div>
<pre><code>## [1] &quot;aardvark&quot;    &quot;bunny&quot;       &quot;caterpillar&quot;</code></pre>
<p>Notice that in the command above the values in <code>a</code> are defined in quotation marks. Without quotation marks, R would interpret those words as the names of objects. For example:</p>
<div class="sourceCode" id="cb211"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb211-1"><a href="mod-02.html#cb211-1" tabindex="-1"></a>caterpillar <span class="ot">&lt;-</span> <span class="st">&quot;lion&quot;</span></span>
<span id="cb211-2"><a href="mod-02.html#cb211-2" tabindex="-1"></a><span class="fu">c</span>(<span class="st">&quot;aardvark&quot;</span>, <span class="st">&quot;bunny&quot;</span>, caterpillar)</span></code></pre></div>
<pre><code>## [1] &quot;aardvark&quot; &quot;bunny&quot;    &quot;lion&quot;</code></pre>
<p>There are some special character vectors pre-built into R that can be useful in programming: <code>letters</code> returns the lower case letters a-z, while <code>LETTERS</code> returns the upper-case letters A-Z.</p>
<div class="sourceCode" id="cb213"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb213-1"><a href="mod-02.html#cb213-1" tabindex="-1"></a>letters</span>
<span id="cb213-2"><a href="mod-02.html#cb213-2" tabindex="-1"></a><span class="do">##  [1] &quot;a&quot; &quot;b&quot; &quot;c&quot; &quot;d&quot; &quot;e&quot; &quot;f&quot; &quot;g&quot; &quot;h&quot; &quot;i&quot; &quot;j&quot; &quot;k&quot; &quot;l&quot; &quot;m&quot; &quot;n&quot; &quot;o&quot; &quot;p&quot; &quot;q&quot; &quot;r&quot; &quot;s&quot;</span></span>
<span id="cb213-3"><a href="mod-02.html#cb213-3" tabindex="-1"></a><span class="do">## [20] &quot;t&quot; &quot;u&quot; &quot;v&quot; &quot;w&quot; &quot;x&quot; &quot;y&quot; &quot;z&quot;</span></span>
<span id="cb213-4"><a href="mod-02.html#cb213-4" tabindex="-1"></a>LETTERS</span>
<span id="cb213-5"><a href="mod-02.html#cb213-5" tabindex="-1"></a><span class="do">##  [1] &quot;A&quot; &quot;B&quot; &quot;C&quot; &quot;D&quot; &quot;E&quot; &quot;F&quot; &quot;G&quot; &quot;H&quot; &quot;I&quot; &quot;J&quot; &quot;K&quot; &quot;L&quot; &quot;M&quot; &quot;N&quot; &quot;O&quot; &quot;P&quot; &quot;Q&quot; &quot;R&quot; &quot;S&quot;</span></span>
<span id="cb213-6"><a href="mod-02.html#cb213-6" tabindex="-1"></a><span class="do">## [20] &quot;T&quot; &quot;U&quot; &quot;V&quot; &quot;W&quot; &quot;X&quot; &quot;Y&quot; &quot;Z&quot;</span></span>
<span id="cb213-7"><a href="mod-02.html#cb213-7" tabindex="-1"></a>letters[<span class="dv">11</span><span class="sc">:</span><span class="dv">20</span>]</span>
<span id="cb213-8"><a href="mod-02.html#cb213-8" tabindex="-1"></a><span class="do">##  [1] &quot;k&quot; &quot;l&quot; &quot;m&quot; &quot;n&quot; &quot;o&quot; &quot;p&quot; &quot;q&quot; &quot;r&quot; &quot;s&quot; &quot;t&quot;</span></span></code></pre></div>
<p>Similarly, <code>month.name</code> and <code>month.abb</code> contain the names of the 12 months and their 3 letter abbreviations, respectively. The latter is very useful for making figure axes.</p>
<div class="sourceCode" id="cb214"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb214-1"><a href="mod-02.html#cb214-1" tabindex="-1"></a><span class="fu">plot</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">12</span>, <span class="fu">runif</span>(<span class="dv">12</span>), <span class="at">xaxt=</span><span class="st">&quot;n&quot;</span>, <span class="at">xlab=</span> <span class="st">&quot;Month&quot;</span>)</span>
<span id="cb214-2"><a href="mod-02.html#cb214-2" tabindex="-1"></a><span class="fu">axis</span>(<span class="at">side=</span><span class="dv">1</span>, <span class="at">at=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">12</span>, <span class="at">labels=</span>month.abb)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-141-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>As an aside, the function <code>substr()</code> extracts parts of character strings (“substring”). One application is getting one letter abbreviations of the months for a figure axis. The example below selects from the month names starting and ending at character 1. The argument <code>xaxt="n"</code> in <code>plot()</code> suppresses the <em>x</em>-axis, so we can add our own with <code>axis()</code>.</p>
<div class="sourceCode" id="cb215"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb215-1"><a href="mod-02.html#cb215-1" tabindex="-1"></a>use.months <span class="ot">&lt;-</span> <span class="fu">substr</span>(month.name, <span class="dv">1</span>, <span class="dv">1</span>)</span>
<span id="cb215-2"><a href="mod-02.html#cb215-2" tabindex="-1"></a><span class="fu">plot</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">12</span>, <span class="fu">runif</span>(<span class="dv">12</span>), <span class="at">xaxt=</span><span class="st">&quot;n&quot;</span>, <span class="at">xlab=</span> <span class="st">&quot;Month&quot;</span>)</span>
<span id="cb215-3"><a href="mod-02.html#cb215-3" tabindex="-1"></a><span class="fu">axis</span>(<span class="at">side=</span><span class="dv">1</span>, <span class="at">at=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">12</span>, <span class="at">labels=</span>use.months)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-142-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="numeric-type" class="section level3 hasAnchor" number="2.7.2">
<h3><span class="header-section-number">2.7.2</span> Numeric type<a href="mod-02.html#numeric-type" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Most data are stored as the <strong>numeric</strong> type. This includes <strong>floating point numbers</strong> (e.g., <code>3.14159</code> or <code>3.14e0</code>). Floating point numbers are referred to as type “double” in R (short for “<a href="https://en.wikipedia.org/wiki/Double-precision_floating-point_format">double-precision floating point</a>”).</p>
<div class="sourceCode" id="cb216"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb216-1"><a href="mod-02.html#cb216-1" tabindex="-1"></a><span class="fu">typeof</span>(<span class="dv">2</span>)</span></code></pre></div>
<pre><code>## [1] &quot;double&quot;</code></pre>
<div class="sourceCode" id="cb218"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb218-1"><a href="mod-02.html#cb218-1" tabindex="-1"></a><span class="fu">typeof</span>(pi)</span></code></pre></div>
<pre><code>## [1] &quot;double&quot;</code></pre>
</div>
<div id="integer-type" class="section level3 hasAnchor" number="2.7.3">
<h3><span class="header-section-number">2.7.3</span> Integer type<a href="mod-02.html#integer-type" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><strong>Integers</strong> are numbers with no fractional components. Some numbers are stored as integers, but integers are not always interchangeable with floating point numbers. For example, some probability distribution functions return integer values rather than floating point values (e.g., binomial or Poisson). This can be problematic if very large numbers are expected because the maximum integer value in R is much smaller than the maximum floating-point value. Note that having the type <code>integer</code> in R and being an integer from a mathematical standpoint are not synonymous.</p>
<div class="sourceCode" id="cb220"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb220-1"><a href="mod-02.html#cb220-1" tabindex="-1"></a><span class="co"># numbers are double by default:</span></span>
<span id="cb220-2"><a href="mod-02.html#cb220-2" tabindex="-1"></a><span class="fu">typeof</span>(<span class="dv">7</span>)</span>
<span id="cb220-3"><a href="mod-02.html#cb220-3" tabindex="-1"></a><span class="do">## [1] &quot;double&quot;</span></span>
<span id="cb220-4"><a href="mod-02.html#cb220-4" tabindex="-1"></a></span>
<span id="cb220-5"><a href="mod-02.html#cb220-5" tabindex="-1"></a><span class="co"># convert to integer with as.integer():</span></span>
<span id="cb220-6"><a href="mod-02.html#cb220-6" tabindex="-1"></a><span class="fu">typeof</span>(<span class="fu">as.integer</span>(<span class="dv">7</span>))</span>
<span id="cb220-7"><a href="mod-02.html#cb220-7" tabindex="-1"></a><span class="do">## [1] &quot;integer&quot;</span></span>
<span id="cb220-8"><a href="mod-02.html#cb220-8" tabindex="-1"></a></span>
<span id="cb220-9"><a href="mod-02.html#cb220-9" tabindex="-1"></a><span class="co"># test whether input is integer type:</span></span>
<span id="cb220-10"><a href="mod-02.html#cb220-10" tabindex="-1"></a><span class="fu">is.integer</span>(<span class="dv">7</span>)</span>
<span id="cb220-11"><a href="mod-02.html#cb220-11" tabindex="-1"></a><span class="do">## [1] FALSE</span></span>
<span id="cb220-12"><a href="mod-02.html#cb220-12" tabindex="-1"></a></span>
<span id="cb220-13"><a href="mod-02.html#cb220-13" tabindex="-1"></a><span class="co"># test whether input is numeric type:</span></span>
<span id="cb220-14"><a href="mod-02.html#cb220-14" tabindex="-1"></a><span class="fu">is.numeric</span>(<span class="dv">7</span>)</span>
<span id="cb220-15"><a href="mod-02.html#cb220-15" tabindex="-1"></a><span class="do">## [1] TRUE</span></span>
<span id="cb220-16"><a href="mod-02.html#cb220-16" tabindex="-1"></a></span>
<span id="cb220-17"><a href="mod-02.html#cb220-17" tabindex="-1"></a><span class="co"># some functions return integer type as output:</span></span>
<span id="cb220-18"><a href="mod-02.html#cb220-18" tabindex="-1"></a><span class="fu">typeof</span>(<span class="fu">rbinom</span>(<span class="dv">1</span>, <span class="dv">10</span>, <span class="fl">0.7</span>))</span>
<span id="cb220-19"><a href="mod-02.html#cb220-19" tabindex="-1"></a><span class="do">## [1] &quot;integer&quot;</span></span>
<span id="cb220-20"><a href="mod-02.html#cb220-20" tabindex="-1"></a></span>
<span id="cb220-21"><a href="mod-02.html#cb220-21" tabindex="-1"></a><span class="co"># 7 is mathematically an integer (7 mod 1 == 0),</span></span>
<span id="cb220-22"><a href="mod-02.html#cb220-22" tabindex="-1"></a><span class="co"># but is not of integer type in R</span></span>
<span id="cb220-23"><a href="mod-02.html#cb220-23" tabindex="-1"></a><span class="dv">7</span> <span class="sc">%%</span> <span class="dv">1</span></span>
<span id="cb220-24"><a href="mod-02.html#cb220-24" tabindex="-1"></a><span class="do">## [1] 0</span></span></code></pre></div>
<p>Integers are used as indices for subsetting and accessing other objects. A sequence of integers can be made using the colon operator as follows:</p>
<div class="sourceCode" id="cb221"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb221-1"><a href="mod-02.html#cb221-1" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span></span>
<span id="cb221-2"><a href="mod-02.html#cb221-2" tabindex="-1"></a>a</span>
<span id="cb221-3"><a href="mod-02.html#cb221-3" tabindex="-1"></a><span class="do">##  [1]  1  2  3  4  5  6  7  8  9 10</span></span>
<span id="cb221-4"><a href="mod-02.html#cb221-4" tabindex="-1"></a><span class="fu">typeof</span>(a)</span>
<span id="cb221-5"><a href="mod-02.html#cb221-5" tabindex="-1"></a><span class="do">## [1] &quot;integer&quot;</span></span></code></pre></div>
<p>An important consequence of using integers as index values is that object dimensions are limited to the <strong>R integer limit</strong>. This is probably not a major concern for most users. Your machine will likely run out of memory before you could make such an object anyway. You can check the integer limit on your machine with this line:</p>
<div class="sourceCode" id="cb222"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb222-1"><a href="mod-02.html#cb222-1" tabindex="-1"></a>.Machine<span class="sc">$</span>integer.max</span>
<span id="cb222-2"><a href="mod-02.html#cb222-2" tabindex="-1"></a><span class="do">## [1] 2147483647</span></span></code></pre></div>
<p>On my machine, this is about 2.14 × 10<sup>9</sup>. If you need data objects with dimensions larger than this, you may need to either use a specialized R package that can handle huge objects, or rethink what you are doing.</p>
</div>
<div id="logical-type" class="section level3 hasAnchor" number="2.7.4">
<h3><span class="header-section-number">2.7.4</span> Logical type<a href="mod-02.html#logical-type" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The <strong>logical</strong> data type stores <code>TRUE</code> and <code>FALSE</code> values. Note that <code>TRUE</code> and <code>FALSE</code> are always written in capital letters, without quotation marks. Logical values are produced by logical tests, such as the test for equality using <code>==</code>:</p>
<div class="sourceCode" id="cb223"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb223-1"><a href="mod-02.html#cb223-1" tabindex="-1"></a><span class="dv">1</span><span class="sc">+</span><span class="dv">1</span> <span class="sc">==</span> <span class="dv">2</span>            </span>
<span id="cb223-2"><a href="mod-02.html#cb223-2" tabindex="-1"></a><span class="do">## [1] TRUE</span></span>
<span id="cb223-3"><a href="mod-02.html#cb223-3" tabindex="-1"></a><span class="fu">typeof</span>(<span class="dv">1</span><span class="sc">+</span><span class="dv">1</span> <span class="sc">==</span> <span class="dv">2</span>)    </span>
<span id="cb223-4"><a href="mod-02.html#cb223-4" tabindex="-1"></a><span class="do">## [1] &quot;logical&quot;</span></span>
<span id="cb223-5"><a href="mod-02.html#cb223-5" tabindex="-1"></a><span class="fu">typeof</span>(<span class="dv">1</span><span class="sc">+</span><span class="dv">1</span> <span class="sc">==</span> <span class="dv">3</span>)    </span>
<span id="cb223-6"><a href="mod-02.html#cb223-6" tabindex="-1"></a><span class="do">## [1] &quot;logical&quot;</span></span></code></pre></div>
<p>Another useful test you can do is to test whether elements of one set are in another:</p>
<div class="sourceCode" id="cb224"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb224-1"><a href="mod-02.html#cb224-1" tabindex="-1"></a><span class="dv">5</span> <span class="sc">%in%</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span></span>
<span id="cb224-2"><a href="mod-02.html#cb224-2" tabindex="-1"></a><span class="do">## [1] TRUE</span></span>
<span id="cb224-3"><a href="mod-02.html#cb224-3" tabindex="-1"></a><span class="dv">5</span> <span class="sc">%in%</span> <span class="dv">6</span><span class="sc">:</span><span class="dv">10</span></span>
<span id="cb224-4"><a href="mod-02.html#cb224-4" tabindex="-1"></a><span class="do">## [1] FALSE</span></span>
<span id="cb224-5"><a href="mod-02.html#cb224-5" tabindex="-1"></a><span class="dv">1</span><span class="sc">:</span><span class="dv">10</span> <span class="sc">%in%</span> <span class="dv">5</span><span class="sc">:</span><span class="dv">15</span></span>
<span id="cb224-6"><a href="mod-02.html#cb224-6" tabindex="-1"></a><span class="do">##  [1] FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE</span></span></code></pre></div>
<p>You can test with inequalities:</p>
<div class="sourceCode" id="cb225"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb225-1"><a href="mod-02.html#cb225-1" tabindex="-1"></a><span class="dv">2</span> <span class="sc">&gt;=</span> <span class="dv">1</span></span>
<span id="cb225-2"><a href="mod-02.html#cb225-2" tabindex="-1"></a><span class="do">## [1] TRUE</span></span>
<span id="cb225-3"><a href="mod-02.html#cb225-3" tabindex="-1"></a><span class="dv">3</span> <span class="sc">&lt;</span> <span class="dv">2</span></span>
<span id="cb225-4"><a href="mod-02.html#cb225-4" tabindex="-1"></a><span class="do">## [1] FALSE</span></span></code></pre></div>
<p>Sometimes it can be useful to store logical values as 1 or 0. You can do this by multiplying a logical by 1 or using function <code>as.numeric</code>.</p>
<div class="sourceCode" id="cb226"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb226-1"><a href="mod-02.html#cb226-1" tabindex="-1"></a><span class="dv">1</span><span class="sc">*</span>(<span class="dv">1</span><span class="sc">+</span><span class="dv">1</span> <span class="sc">==</span> <span class="dv">2</span>)</span>
<span id="cb226-2"><a href="mod-02.html#cb226-2" tabindex="-1"></a><span class="do">## [1] 1</span></span>
<span id="cb226-3"><a href="mod-02.html#cb226-3" tabindex="-1"></a><span class="dv">1</span><span class="sc">*</span>(<span class="dv">1</span><span class="sc">+</span><span class="dv">1</span> <span class="sc">==</span> <span class="dv">3</span>)</span>
<span id="cb226-4"><a href="mod-02.html#cb226-4" tabindex="-1"></a><span class="do">## [1] 0</span></span>
<span id="cb226-5"><a href="mod-02.html#cb226-5" tabindex="-1"></a><span class="fu">typeof</span>(<span class="dv">1</span><span class="sc">+</span><span class="dv">1</span> <span class="sc">==</span> <span class="dv">2</span>)</span>
<span id="cb226-6"><a href="mod-02.html#cb226-6" tabindex="-1"></a><span class="do">## [1] &quot;logical&quot;</span></span>
<span id="cb226-7"><a href="mod-02.html#cb226-7" tabindex="-1"></a><span class="fu">typeof</span>(<span class="dv">1</span><span class="sc">*</span>(<span class="dv">1</span><span class="sc">+</span><span class="dv">1</span> <span class="sc">==</span> <span class="dv">2</span>))</span>
<span id="cb226-8"><a href="mod-02.html#cb226-8" tabindex="-1"></a><span class="do">## [1] &quot;double&quot;</span></span>
<span id="cb226-9"><a href="mod-02.html#cb226-9" tabindex="-1"></a><span class="fu">typeof</span>(<span class="fu">as.numeric</span>(<span class="dv">1</span><span class="sc">+</span><span class="dv">1</span> <span class="sc">==</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>))</span>
<span id="cb226-10"><a href="mod-02.html#cb226-10" tabindex="-1"></a><span class="do">## [1] &quot;double&quot;</span></span></code></pre></div>
<p>The <code>!</code> symbol <strong>negates</strong> or <strong>reverses</strong> logical values. It does not negate logical values stored as numbers.</p>
<div class="sourceCode" id="cb227"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb227-1"><a href="mod-02.html#cb227-1" tabindex="-1"></a><span class="dv">2</span><span class="sc">+</span><span class="dv">2</span> <span class="sc">==</span> <span class="dv">4</span></span>
<span id="cb227-2"><a href="mod-02.html#cb227-2" tabindex="-1"></a><span class="do">## [1] TRUE</span></span>
<span id="cb227-3"><a href="mod-02.html#cb227-3" tabindex="-1"></a><span class="sc">!</span> <span class="dv">2</span><span class="sc">+</span><span class="dv">2</span> <span class="sc">==</span> <span class="dv">4</span></span>
<span id="cb227-4"><a href="mod-02.html#cb227-4" tabindex="-1"></a><span class="do">## [1] FALSE</span></span>
<span id="cb227-5"><a href="mod-02.html#cb227-5" tabindex="-1"></a><span class="dv">1</span><span class="sc">:</span><span class="dv">10</span> <span class="sc">%in%</span> <span class="dv">5</span><span class="sc">:</span><span class="dv">15</span></span>
<span id="cb227-6"><a href="mod-02.html#cb227-6" tabindex="-1"></a><span class="do">##  [1] FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE</span></span>
<span id="cb227-7"><a href="mod-02.html#cb227-7" tabindex="-1"></a><span class="sc">!</span><span class="dv">1</span><span class="sc">:</span><span class="dv">10</span> <span class="sc">%in%</span> <span class="dv">5</span><span class="sc">:</span><span class="dv">15</span></span>
<span id="cb227-8"><a href="mod-02.html#cb227-8" tabindex="-1"></a><span class="do">##  [1]  TRUE  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE</span></span></code></pre></div>
<p>The <code>any()</code> and <code>all()</code> functions test to see if any values in a set of logical values is true (<code>any()</code>) or if all of them are true (<code>all()</code>).</p>
<div class="sourceCode" id="cb228"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb228-1"><a href="mod-02.html#cb228-1" tabindex="-1"></a><span class="dv">1</span><span class="sc">:</span><span class="dv">5</span> <span class="sc">&lt;</span> <span class="dv">3</span></span>
<span id="cb228-2"><a href="mod-02.html#cb228-2" tabindex="-1"></a><span class="do">## [1]  TRUE  TRUE FALSE FALSE FALSE</span></span>
<span id="cb228-3"><a href="mod-02.html#cb228-3" tabindex="-1"></a><span class="fu">any</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span> <span class="sc">&lt;</span> <span class="dv">3</span>)</span>
<span id="cb228-4"><a href="mod-02.html#cb228-4" tabindex="-1"></a><span class="do">## [1] TRUE</span></span>
<span id="cb228-5"><a href="mod-02.html#cb228-5" tabindex="-1"></a><span class="fu">all</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span> <span class="sc">&lt;</span> <span class="dv">3</span>)</span>
<span id="cb228-6"><a href="mod-02.html#cb228-6" tabindex="-1"></a><span class="do">## [1] FALSE</span></span></code></pre></div>
<p><code>TRUE</code> and <code>FALSE</code> can also be abbreviated as uppercase <code>T</code> or <code>F</code>, but not lowercase <code>t</code> or <code>f</code>. This means that you should <strong><em>not</em></strong> create objects with names <code>T</code>, <code>F</code>, <code>TRUE</code>, etc. (don’t name an object <code>t</code> either, because that is the name of the transpose function). The code block below illustrates the mischief that can be created by naming an object <code>T</code> and masking the built-in value for true.</p>
<div class="sourceCode" id="cb229"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb229-1"><a href="mod-02.html#cb229-1" tabindex="-1"></a><span class="co"># (1+1==2) is true.</span></span>
<span id="cb229-2"><a href="mod-02.html#cb229-2" tabindex="-1"></a>T <span class="sc">==</span> (<span class="dv">1</span><span class="sc">+</span><span class="dv">1</span><span class="sc">==</span><span class="dv">2</span>)</span>
<span id="cb229-3"><a href="mod-02.html#cb229-3" tabindex="-1"></a><span class="do">## [1] TRUE</span></span>
<span id="cb229-4"><a href="mod-02.html#cb229-4" tabindex="-1"></a></span>
<span id="cb229-5"><a href="mod-02.html#cb229-5" tabindex="-1"></a><span class="co"># make a new object T with value 3</span></span>
<span id="cb229-6"><a href="mod-02.html#cb229-6" tabindex="-1"></a>T <span class="ot">&lt;-</span> <span class="dv">3</span></span>
<span id="cb229-7"><a href="mod-02.html#cb229-7" tabindex="-1"></a></span>
<span id="cb229-8"><a href="mod-02.html#cb229-8" tabindex="-1"></a><span class="co"># (1+1==2) is now not true, unless it is</span></span>
<span id="cb229-9"><a href="mod-02.html#cb229-9" tabindex="-1"></a>T <span class="sc">==</span> (<span class="dv">1</span><span class="sc">+</span><span class="dv">1</span><span class="sc">==</span><span class="dv">2</span>)</span>
<span id="cb229-10"><a href="mod-02.html#cb229-10" tabindex="-1"></a><span class="do">## [1] FALSE</span></span>
<span id="cb229-11"><a href="mod-02.html#cb229-11" tabindex="-1"></a><span class="cn">TRUE</span> <span class="sc">==</span> (<span class="dv">1</span><span class="sc">+</span><span class="dv">1</span><span class="sc">==</span><span class="dv">2</span>)</span>
<span id="cb229-12"><a href="mod-02.html#cb229-12" tabindex="-1"></a><span class="do">## [1] TRUE</span></span>
<span id="cb229-13"><a href="mod-02.html#cb229-13" tabindex="-1"></a></span>
<span id="cb229-14"><a href="mod-02.html#cb229-14" tabindex="-1"></a><span class="co"># remove the troublesome T</span></span>
<span id="cb229-15"><a href="mod-02.html#cb229-15" tabindex="-1"></a><span class="fu">rm</span>(T)</span>
<span id="cb229-16"><a href="mod-02.html#cb229-16" tabindex="-1"></a></span>
<span id="cb229-17"><a href="mod-02.html#cb229-17" tabindex="-1"></a><span class="co"># things are back to normal</span></span>
<span id="cb229-18"><a href="mod-02.html#cb229-18" tabindex="-1"></a>T <span class="sc">==</span> (<span class="dv">1</span><span class="sc">+</span><span class="dv">1</span><span class="sc">==</span><span class="dv">2</span>)</span>
<span id="cb229-19"><a href="mod-02.html#cb229-19" tabindex="-1"></a><span class="do">## [1] TRUE</span></span></code></pre></div>
</div>
<div id="special-values" class="section level3 hasAnchor" number="2.7.5">
<h3><span class="header-section-number">2.7.5</span> Special values<a href="mod-02.html#special-values" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="missing-values-na" class="section level4 hasAnchor" number="2.7.5.1">
<h4><span class="header-section-number">2.7.5.1</span> Missing values <code>NA</code><a href="mod-02.html#missing-values-na" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Missing data in R are labeled as <code>NA</code>, or “Not Available”. This is R-speak for a blank or missing value. Some functions can ignore <code>NA</code>, some cannot. Some functions produce <code>NA</code> when you supply them with improper inputs. Note that this is a feature, not a bug. R is set up to produce <code>NA</code> by default in many cases to help you track down where the missing values are coming from. Function <code>is.na()</code> tests to see if a value is missing.</p>
<div class="sourceCode" id="cb230"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb230-1"><a href="mod-02.html#cb230-1" tabindex="-1"></a></span>
<span id="cb230-2"><a href="mod-02.html#cb230-2" tabindex="-1"></a>aa <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>,<span class="cn">NA</span>)</span>
<span id="cb230-3"><a href="mod-02.html#cb230-3" tabindex="-1"></a></span>
<span id="cb230-4"><a href="mod-02.html#cb230-4" tabindex="-1"></a>aa <span class="sc">+</span> <span class="dv">2</span></span>
<span id="cb230-5"><a href="mod-02.html#cb230-5" tabindex="-1"></a><span class="do">## [1]  3  4  5  6  7  8  9 10 NA</span></span>
<span id="cb230-6"><a href="mod-02.html#cb230-6" tabindex="-1"></a></span>
<span id="cb230-7"><a href="mod-02.html#cb230-7" tabindex="-1"></a><span class="co"># returns NA</span></span>
<span id="cb230-8"><a href="mod-02.html#cb230-8" tabindex="-1"></a><span class="fu">mean</span>(aa)</span>
<span id="cb230-9"><a href="mod-02.html#cb230-9" tabindex="-1"></a><span class="do">## [1] NA</span></span>
<span id="cb230-10"><a href="mod-02.html#cb230-10" tabindex="-1"></a></span>
<span id="cb230-11"><a href="mod-02.html#cb230-11" tabindex="-1"></a><span class="co"># option to ignore NA values</span></span>
<span id="cb230-12"><a href="mod-02.html#cb230-12" tabindex="-1"></a><span class="fu">mean</span>(aa,<span class="at">na.rm=</span><span class="cn">TRUE</span>)      </span>
<span id="cb230-13"><a href="mod-02.html#cb230-13" tabindex="-1"></a><span class="do">## [1] 4.5</span></span>
<span id="cb230-14"><a href="mod-02.html#cb230-14" tabindex="-1"></a></span>
<span id="cb230-15"><a href="mod-02.html#cb230-15" tabindex="-1"></a><span class="co"># test for NA</span></span>
<span id="cb230-16"><a href="mod-02.html#cb230-16" tabindex="-1"></a><span class="fu">is.na</span>(aa)               </span>
<span id="cb230-17"><a href="mod-02.html#cb230-17" tabindex="-1"></a><span class="do">## [1] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE</span></span>
<span id="cb230-18"><a href="mod-02.html#cb230-18" tabindex="-1"></a></span>
<span id="cb230-19"><a href="mod-02.html#cb230-19" tabindex="-1"></a><span class="co"># which elements are NA?</span></span>
<span id="cb230-20"><a href="mod-02.html#cb230-20" tabindex="-1"></a><span class="fu">which</span>(<span class="fu">is.na</span>(aa))        </span>
<span id="cb230-21"><a href="mod-02.html#cb230-21" tabindex="-1"></a><span class="do">## [1] 9</span></span>
<span id="cb230-22"><a href="mod-02.html#cb230-22" tabindex="-1"></a></span>
<span id="cb230-23"><a href="mod-02.html#cb230-23" tabindex="-1"></a><span class="co"># remove NA elements</span></span>
<span id="cb230-24"><a href="mod-02.html#cb230-24" tabindex="-1"></a>aa[<span class="sc">-</span><span class="fu">which</span>(<span class="fu">is.na</span>(aa))]   </span>
<span id="cb230-25"><a href="mod-02.html#cb230-25" tabindex="-1"></a><span class="do">## [1] 1 2 3 4 5 6 7 8</span></span>
<span id="cb230-26"><a href="mod-02.html#cb230-26" tabindex="-1"></a></span>
<span id="cb230-27"><a href="mod-02.html#cb230-27" tabindex="-1"></a><span class="co"># another way to remove NA elements</span></span>
<span id="cb230-28"><a href="mod-02.html#cb230-28" tabindex="-1"></a>aa[<span class="fu">which</span>(<span class="sc">!</span><span class="fu">is.na</span>(aa))]   </span>
<span id="cb230-29"><a href="mod-02.html#cb230-29" tabindex="-1"></a><span class="do">## [1] 1 2 3 4 5 6 7 8</span></span>
<span id="cb230-30"><a href="mod-02.html#cb230-30" tabindex="-1"></a></span>
<span id="cb230-31"><a href="mod-02.html#cb230-31" tabindex="-1"></a><span class="co"># returns NA: Poisson distribution supported only on [0,+Inf]</span></span>
<span id="cb230-32"><a href="mod-02.html#cb230-32" tabindex="-1"></a><span class="fu">rpois</span>(<span class="dv">10</span>,<span class="sc">-</span><span class="dv">2</span>) </span>
<span id="cb230-33"><a href="mod-02.html#cb230-33" tabindex="-1"></a><span class="do">## Warning in rpois(10, -2): NAs produced</span></span>
<span id="cb230-34"><a href="mod-02.html#cb230-34" tabindex="-1"></a><span class="do">##  [1] NA NA NA NA NA NA NA NA NA NA</span></span></code></pre></div>
</div>
<div id="infinities" class="section level4 hasAnchor" number="2.7.5.2">
<h4><span class="header-section-number">2.7.5.2</span> Infinities<a href="mod-02.html#infinities" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Positive and negative infinity (<span class="math inline">\(+\infty\)</span> and <span class="math inline">\(-\infty\)</span>) are labeled in R as <code>Inf</code> and <code>–Inf</code>. Dividing a non-zero number by 0 produces <code>Inf</code>. Taking the log of 0 produces <code>–Inf</code>. Trying to generate a value greater than R’s maximal value can produce <code>Inf</code>.</p>
<div class="sourceCode" id="cb231"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb231-1"><a href="mod-02.html#cb231-1" tabindex="-1"></a><span class="dv">5</span><span class="sc">/</span><span class="dv">0</span></span>
<span id="cb231-2"><a href="mod-02.html#cb231-2" tabindex="-1"></a><span class="do">## [1] Inf</span></span>
<span id="cb231-3"><a href="mod-02.html#cb231-3" tabindex="-1"></a><span class="fu">log</span>(<span class="dv">0</span>)</span>
<span id="cb231-4"><a href="mod-02.html#cb231-4" tabindex="-1"></a><span class="do">## [1] -Inf</span></span>
<span id="cb231-5"><a href="mod-02.html#cb231-5" tabindex="-1"></a><span class="dv">10</span><span class="sc">^</span><span class="dv">10</span><span class="sc">^</span><span class="dv">10</span></span>
<span id="cb231-6"><a href="mod-02.html#cb231-6" tabindex="-1"></a><span class="do">## [1] Inf</span></span></code></pre></div>
</div>
<div id="undefined-values" class="section level4 hasAnchor" number="2.7.5.3">
<h4><span class="header-section-number">2.7.5.3</span> Undefined values<a href="mod-02.html#undefined-values" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Some values that are undefined are labeled <code>NaN</code> (not a number). This means that a value is undefined, not that it is missing (which would be <code>NA</code>) or infinite (which would be <code>Inf</code> or <code>-Inf</code>). Dividing 0 by 0, or taking the log of a negative number, produces <code>NaN</code>. The function <code>is.nan()</code> tests to see if a value is <code>NaN</code>.</p>
<div class="sourceCode" id="cb232"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb232-1"><a href="mod-02.html#cb232-1" tabindex="-1"></a><span class="dv">0</span><span class="sc">/</span><span class="dv">0</span></span>
<span id="cb232-2"><a href="mod-02.html#cb232-2" tabindex="-1"></a><span class="do">## [1] NaN</span></span>
<span id="cb232-3"><a href="mod-02.html#cb232-3" tabindex="-1"></a><span class="fu">log</span>(<span class="sc">-</span><span class="dv">1</span>)</span>
<span id="cb232-4"><a href="mod-02.html#cb232-4" tabindex="-1"></a><span class="do">## Warning in log(-1): NaNs produced</span></span>
<span id="cb232-5"><a href="mod-02.html#cb232-5" tabindex="-1"></a><span class="do">## [1] NaN</span></span></code></pre></div>
</div>
<div id="special-values-are-useful" class="section level4 hasAnchor" number="2.7.5.4">
<h4><span class="header-section-number">2.7.5.4</span> Special values are useful<a href="mod-02.html#special-values-are-useful" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Why are there so many special values? To help track down where erroneous values are coming from. If every undefined value returned the same symbol (or word, etc.), then it would be harder to know why a program was returning undefined values. Instead, the different results help point the way:</p>
<ul>
<li><code>Inf</code>: Positive non-zero number divided by 0</li>
<li><code>-Inf</code>: Negative non-zero number divided by 0 <em>or</em> <code>log(0)</code>.</li>
<li><code>NaN</code>: <code>0/0</code> or logarithm of a negative number</li>
</ul>
</div>
</div>
</div>
<div id="manage-r-code-as-scripts-.r-files" class="section level2 hasAnchor" number="2.8">
<h2><span class="header-section-number">2.8</span> Manage R code as scripts (.r files)<a href="mod-02.html#manage-r-code-as-scripts-.r-files" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Over your career you will probably write a lot of code. You will download and run even more code. This page describes how to manage code that you write (scripts) and code that you get from R repositories (packages).</p>
<p>Often it is easiest to write your R code in separate files, called <strong>scripts</strong>. An R script has the file extension <strong>.r</strong> or <strong>.R</strong>. Scripts can be used as files that store R code (basically plain text files) or as stand-alone R programs. One advantage of keeping your R code in scripts is that you will avoid the formatting issues that arise when writing R code in word processors like Microsoft Word. Another advantage comes from using specialized code editors like Notepad++, RStudio, or others: syntax highlighting. Consider the image below:</p>
<p><img src="02_12.jpg" width="90%" style="display: block; margin: auto;" /></p>
<p>This shows some typical R code in Notepad++ with comments, a <code>for</code> loop, some functions being called, datasets being subsetted, etc. Notice how code elements are colored differently, so it’s easier to see the different parts. Notice also how matching braces <code>{}</code> are highlighted. This is <strong><em>extremely</em></strong> helpful when programming complex operations.</p>
<p>The final advantage to working with R scripts that I’ll mention is automation and running from source. R can read and run .r files with the <code>source()</code> command. This can be useful when managing a large set of simulations or analyses: code the individual simulations in a stand-alone script, and then repeatedly call that script from a different script or R instance. This can work as a simple (albeit labor-intensive) way to achieve parallelization.</p>
</div>
<div id="manage-and-use-r-packages" class="section level2 hasAnchor" number="2.9">
<h2><span class="header-section-number">2.9</span> Manage and use R packages<a href="mod-02.html#manage-and-use-r-packages" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><strong>Packages</strong> are modules that can be added to R to extend its capabilities. Most packages are written by R users and researchers who need functions that are not provided with the base R installation. These add-ons are one of the greatest strengths of R compared to other software packages. Commercial and proprietary statistics programs (like SAS) are upgraded only by a single group of developers, if at all. In contrast, new functions and capabilities are added to R by the user community every day.</p>
<p>Some packages are widely used and cited, others less so. Do your homework before depending on a package. Is the package included or recommended with the base R installation? Is the package associated with any peer-reviewed publications? Does the package author appear to have the credentials that would qualify them to produce a package? Do other packages depend on this package? All of these are good signs that a package is reliable.</p>
<p>Most packages are stored on The Comprehensive R Archive Network, or <a href="https://cran.r-project.org/">CRAN</a>. Other repositories exist, such as the Geological Survey R Archive Network, or <a href="https://owi.usgs.gov/R/gran.html">GRAN</a>. R-Forge, BioConductor, and GitHub also host some R packages. R-Forge and BioConductor can usually be accessed from the R command line just like CRAN. Installing packages from GitHub may require a different procedure. Most R packages are written in the R language itself, so even if you can’t find a way to install a package you can usually just copy and run its source code.</p>
<div id="your-r-library" class="section level3 hasAnchor" number="2.9.1">
<h3><span class="header-section-number">2.9.1</span> Your R library<a href="mod-02.html#your-r-library" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>When you install R packages, the files are downloaded and stored on your machine for R to access when needed. R will probably create a folder in your R home directory for this purpose. The first time you install a package in R you will be asked if R can create a personal library in which to store the package. Package installations are specific to the version of R that you used for the installation, and specific to your operating system login. This means that if someone else logs in to your machine and wants to use a package, they will have to install the package for themselves. This also means that if you update your R installation to a new version, you may need to install packages again because some packages are version-specific.</p>
</div>
<div id="installing-packages-using-the-r-gui" class="section level3 hasAnchor" number="2.9.2">
<h3><span class="header-section-number">2.9.2</span> Installing packages using the R GUI<a href="mod-02.html#installing-packages-using-the-r-gui" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In the R GUI, go to Packages – Install package(s)…. You will be prompted to select a CRAN mirror (i.e., server from which to download). Pick any mirror you like. Then, scroll to and select your package. If the install fails, try a different mirror by going to Packages – Set CRAN mirror…, and then retrying the installation.</p>
</div>
<div id="installing-packages-in-rstudio" class="section level3 hasAnchor" number="2.9.3">
<h3><span class="header-section-number">2.9.3</span> Installing packages in RStudio<a href="mod-02.html#installing-packages-in-rstudio" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>You can install a package in RStudio by going to Tools–Install Packages… and searching for the name of the package you want.</p>
</div>
<div id="installing-packages-using-the-r-console" class="section level3 hasAnchor" number="2.9.4">
<h3><span class="header-section-number">2.9.4</span> Installing packages using the R console<a href="mod-02.html#installing-packages-using-the-r-console" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In the R console, you can install packages using function <code>install.packages()</code> as shown below. You will still be asked to select your CRAN mirror (most packages are hosted on CRAN).</p>
<div class="sourceCode" id="cb233"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb233-1"><a href="mod-02.html#cb233-1" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;vegan&quot;</span>)</span></code></pre></div>
</div>
<div id="working-with-packages-in-r" class="section level3 hasAnchor" number="2.9.5">
<h3><span class="header-section-number">2.9.5</span> Working with packages in R<a href="mod-02.html#working-with-packages-in-r" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>From R, you can access the packages you have installed by <strong>loading</strong> them. In most situations you must load a package before you use it. The usual way is to use the <code>library()</code> function. Once a package is loaded, you can use its functions and datasets. The function <code>require()</code> does the same thing as <code>library()</code>, but it is designed for use inside other functions. Most of the time you should use <code>library()</code>. The command below will load the package <code>vegan</code> so its functions and datasets become available:</p>
<div class="sourceCode" id="cb234"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb234-1"><a href="mod-02.html#cb234-1" tabindex="-1"></a><span class="fu">library</span>(vegan)</span></code></pre></div>
<p>Notice that the package name does not have to be in quotation marks within the <code>library()</code> function.</p>
<p>You can see what packages are currently attached with <code>search()</code> or <code>sessionInfo()</code>. Packages can be unloaded detaching them from the workspace: <code>detach(package:x)</code>, where <code>x</code> is the name of the package you want to detach. In the example below, the first command loads the <code>MASS</code> package. Then, we can see that is loaded in the workspace with <code>search()</code>. The next command unloads <code>MASS</code>, then verifies that MASS was unloaded from the workspace with another <code>search()</code> command.</p>
<div class="sourceCode" id="cb235"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb235-1"><a href="mod-02.html#cb235-1" tabindex="-1"></a><span class="fu">search</span>()</span>
<span id="cb235-2"><a href="mod-02.html#cb235-2" tabindex="-1"></a><span class="do">## [1] &quot;.GlobalEnv&quot;        &quot;package:stats&quot;     &quot;package:graphics&quot; </span></span>
<span id="cb235-3"><a href="mod-02.html#cb235-3" tabindex="-1"></a><span class="do">## [4] &quot;package:grDevices&quot; &quot;package:utils&quot;     &quot;package:datasets&quot; </span></span>
<span id="cb235-4"><a href="mod-02.html#cb235-4" tabindex="-1"></a><span class="do">## [7] &quot;package:methods&quot;   &quot;Autoloads&quot;         &quot;package:base&quot;</span></span>
<span id="cb235-5"><a href="mod-02.html#cb235-5" tabindex="-1"></a><span class="fu">library</span>(MASS)</span>
<span id="cb235-6"><a href="mod-02.html#cb235-6" tabindex="-1"></a><span class="fu">search</span>()</span>
<span id="cb235-7"><a href="mod-02.html#cb235-7" tabindex="-1"></a><span class="do">##  [1] &quot;.GlobalEnv&quot;        &quot;package:MASS&quot;      &quot;package:stats&quot;    </span></span>
<span id="cb235-8"><a href="mod-02.html#cb235-8" tabindex="-1"></a><span class="do">##  [4] &quot;package:graphics&quot;  &quot;package:grDevices&quot; &quot;package:utils&quot;    </span></span>
<span id="cb235-9"><a href="mod-02.html#cb235-9" tabindex="-1"></a><span class="do">##  [7] &quot;package:datasets&quot;  &quot;package:methods&quot;   &quot;Autoloads&quot;        </span></span>
<span id="cb235-10"><a href="mod-02.html#cb235-10" tabindex="-1"></a><span class="do">## [10] &quot;package:base&quot;</span></span>
<span id="cb235-11"><a href="mod-02.html#cb235-11" tabindex="-1"></a><span class="fu">detach</span>(package<span class="sc">:</span>MASS)</span>
<span id="cb235-12"><a href="mod-02.html#cb235-12" tabindex="-1"></a><span class="fu">search</span>()</span>
<span id="cb235-13"><a href="mod-02.html#cb235-13" tabindex="-1"></a><span class="do">## [1] &quot;.GlobalEnv&quot;        &quot;package:stats&quot;     &quot;package:graphics&quot; </span></span>
<span id="cb235-14"><a href="mod-02.html#cb235-14" tabindex="-1"></a><span class="do">## [4] &quot;package:grDevices&quot; &quot;package:utils&quot;     &quot;package:datasets&quot; </span></span>
<span id="cb235-15"><a href="mod-02.html#cb235-15" tabindex="-1"></a><span class="do">## [7] &quot;package:methods&quot;   &quot;Autoloads&quot;         &quot;package:base&quot;</span></span></code></pre></div>
<p>There is rarely any need to unload a package from the workspace unless there are naming conflicts. The most common situation is when two attached packages have functions with the same name. In this situation, R will use the most recently loaded package’s function. This phenomenon is called <strong>masking</strong>. When there is a name conflict, you can use the double colon operator <code>::</code> to specify which package’s function you want to use.</p>
<p>Using <code>::</code> can also access a function without loading the entire package into the workspace. If both packages are already loaded into your workspace, then <code>::</code> will simply select which package’s function to use. Note that a package must be installed on your machine before you can use its functions.</p>
<p>The example below shows how <code>::</code> can be used to access the <code>rtriangle()</code> and <code>fitdistr()</code> functions without loading their packages.</p>
<div class="sourceCode" id="cb236"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb236-1"><a href="mod-02.html#cb236-1" tabindex="-1"></a>base<span class="sc">::</span><span class="fu">log</span>(<span class="dv">10</span>)</span>
<span id="cb236-2"><a href="mod-02.html#cb236-2" tabindex="-1"></a><span class="fu">log</span>(<span class="dv">10</span>)</span>
<span id="cb236-3"><a href="mod-02.html#cb236-3" tabindex="-1"></a>triangle<span class="sc">::</span><span class="fu">rtriangle</span>(<span class="dv">10</span>, <span class="dv">1</span>, <span class="dv">10</span>, <span class="dv">5</span>)</span>
<span id="cb236-4"><a href="mod-02.html#cb236-4" tabindex="-1"></a>MASS<span class="sc">::</span><span class="fu">fitdistr</span>(<span class="fu">rnorm</span>(<span class="dv">100</span>, <span class="dv">50</span>, <span class="dv">10</span>), <span class="st">&quot;normal&quot;</span>)</span>
<span id="cb236-5"><a href="mod-02.html#cb236-5" tabindex="-1"></a></span>
<span id="cb236-6"><a href="mod-02.html#cb236-6" tabindex="-1"></a><span class="co"># neither package (triangle or MASS) is attached:</span></span>
<span id="cb236-7"><a href="mod-02.html#cb236-7" tabindex="-1"></a><span class="fu">search</span>()</span>
<span id="cb236-8"><a href="mod-02.html#cb236-8" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
</div>
<div id="package-dependencies" class="section level3 hasAnchor" number="2.9.6">
<h3><span class="header-section-number">2.9.6</span> Package dependencies<a href="mod-02.html#package-dependencies" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Many R packages are built taking advantage of other R packages. The packages that a package depends on are called its <strong>dependencies</strong> or <strong>depends</strong>. Some packages have no dependencies (other than R itself), while others have many. Something to think about when deciding whether to use a package are the number and nature of its dependencies. If you use a package that depends on many other packages, then you are relying on the authors of those packages to maintain and update their code to work with new versions of R and new versions of packages. And, relying upon the authors of all of those packages’ dependencies. And so on. You can see the dependencies of a package by checking out its page on CRAN. For example, the <code>triangle</code> package <span class="citation">(<a href="#ref-carnell2019">Carnell 2019</a>)</span> is <a href="https://cran.r-project.org/package=triangle">here</a>. You can see the current version, what packages it depends on (“Depends”), and what packages depend on it (“Reverse depends”).</p>
<p>The point here is that you don’t want your code to break because one of the packages you use hasn’t been updated. This can get very frustrating if, for example, several packages you use all depend on each other, or even different versions of each other! Most R package authors and/or maintainers are pretty good about keeping their packages up to date, but some are not. If you really need a function from a package that is no longer maintained, or incompatible with your current version of R, you have three options:</p>
<ol style="list-style-type: decimal">
<li>Download and use the older version of R that is compatible with the package. This is not recommended because you will lose the benefit of recent bug fixes and modifications to R. This option is also not recommended because by rolling back to an older version of R, you will likely need to roll back to older versions of other packages.</li>
<li>Find another, current package that does what you need to do. This can be a quite a headache if you have a large codebase or a very specialized function.</li>
<li>Access the source code to the package you need, and “borrow” the functions or code you need.</li>
</ol>
<p>Sometimes, if you only need 1 or 2 functions from a package, or you need to modify a function from a package to suit your needs. In these cases, you can “borrow” directly from the package source code. For most packages, the CRAN page will have a link to an online code repository (usually Github) and/or the <strong>package source</strong> as a “tarball” (“.tar.gz” file). A tarball is a compressed version of the package source code that may or may not need to be compiled using special functions in R.</p>
<p>Because most R packages are written in R itself, you can usually find and adapt the code you need. This is almost always permitted by the license of the package because most packages are released under some version of the GNU Public License (GPL). If you do borrow code from a package, it is considered good form to attribute the package authors in some way (e.g., “we adapted code from the package”triangle” version 0.12 (Carnell 2019)“).</p>
</div>
<div id="citing-packages" class="section level3 hasAnchor" number="2.9.7">
<h3><span class="header-section-number">2.9.7</span> Citing packages<a href="mod-02.html#citing-packages" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Just like any software, you need to cite a package if you use it. This is both to give appropriate credit to the package authors, and to allow others to understand what you did. The appropriate citation for a package can be accessed from the console with the <code>citation()</code> command.</p>
<p>Citation for R itself:</p>
<div class="sourceCode" id="cb237"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb237-1"><a href="mod-02.html#cb237-1" tabindex="-1"></a><span class="fu">citation</span>()</span></code></pre></div>
<pre><code>## To cite R in publications use:
## 
##   R Core Team (2023). _R: A Language and Environment for Statistical
##   Computing_. R Foundation for Statistical Computing, Vienna, Austria.
##   &lt;https://www.R-project.org/&gt;.
## 
## A BibTeX entry for LaTeX users is
## 
##   @Manual{,
##     title = {R: A Language and Environment for Statistical Computing},
##     author = {{R Core Team}},
##     organization = {R Foundation for Statistical Computing},
##     address = {Vienna, Austria},
##     year = {2023},
##     url = {https://www.R-project.org/},
##   }
## 
## We have invested a lot of time and effort in creating R, please cite it
## when using it for data analysis. See also &#39;citation(&quot;pkgname&quot;)&#39; for
## citing R packages.</code></pre>
<p>Citation for a package (note the quotation marks):</p>
<div class="sourceCode" id="cb239"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb239-1"><a href="mod-02.html#cb239-1" tabindex="-1"></a><span class="fu">citation</span>(<span class="st">&quot;MASS&quot;</span>)</span></code></pre></div>
<pre><code>## To cite the MASS package in publications use:
## 
##   Venables, W. N. &amp; Ripley, B. D. (2002) Modern Applied Statistics with
##   S. Fourth Edition. Springer, New York. ISBN 0-387-95457-0
## 
## A BibTeX entry for LaTeX users is
## 
##   @Book{,
##     title = {Modern Applied Statistics with S},
##     author = {W. N. Venables and B. D. Ripley},
##     publisher = {Springer},
##     edition = {Fourth},
##     address = {New York},
##     year = {2002},
##     note = {ISBN 0-387-95457-0},
##     url = {https://www.stats.ox.ac.uk/pub/MASS4/},
##   }</code></pre>
<div class="sourceCode" id="cb241"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb241-1"><a href="mod-02.html#cb241-1" tabindex="-1"></a><span class="fu">citation</span>(<span class="st">&quot;triangle&quot;</span>)</span></code></pre></div>
<pre><code>## To cite package &#39;triangle&#39; in publications use:
## 
##   Carnell R (2022). _triangle: Distribution Functions and Parameter
##   Estimates for the Triangle Distribution_. R package version 1.0,
##   &lt;https://CRAN.R-project.org/package=triangle&gt;.
## 
## A BibTeX entry for LaTeX users is
## 
##   @Manual{,
##     title = {triangle: Distribution Functions and Parameter Estimates for the Triangle
## Distribution},
##     author = {Rob Carnell},
##     year = {2022},
##     note = {R package version 1.0},
##     url = {https://CRAN.R-project.org/package=triangle},
##   }</code></pre>
</div>
</div>
<div id="r-documentation" class="section level2 hasAnchor" number="2.10">
<h2><span class="header-section-number">2.10</span> R documentation<a href="mod-02.html#r-documentation" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The most important lesson of this R and biological data analysis course is simple: <strong><em>Don’t panic!</em></strong> Lots of people have trouble with R. The learning curve is steep, and the error messages are inscrutable. But, there are a lot of resources available to help you.</p>
<div id="documentation-help-files" class="section level3 hasAnchor" number="2.10.1">
<h3><span class="header-section-number">2.10.1</span> Documentation (help) files<a href="mod-02.html#documentation-help-files" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Documentation for each function, or help files, come included with the base R installation. You can access them from within R by using the <code>?</code> command. <code>?functionname</code> will open the documentation file for function <code>functionname</code>. <code>?</code> will search all currently attached packages. The command below opens the page for the mean function. Notice that this file is stored on your local machine, so no internet access is required.</p>
<div class="sourceCode" id="cb243"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb243-1"><a href="mod-02.html#cb243-1" tabindex="-1"></a>?mean</span></code></pre></div>
<p><img src="02_13.jpg" width="70%" style="display: block; margin: auto;" /></p>
<p>Some functions belong to a family of related functions that share a common documentation page. For example, all of the functions related to the normal distribution share a common page:</p>
<div class="sourceCode" id="cb244"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb244-1"><a href="mod-02.html#cb244-1" tabindex="-1"></a>?dnorm</span>
<span id="cb244-2"><a href="mod-02.html#cb244-2" tabindex="-1"></a>?pnorm</span>
<span id="cb244-3"><a href="mod-02.html#cb244-3" tabindex="-1"></a>?qnorm</span>
<span id="cb244-4"><a href="mod-02.html#cb244-4" tabindex="-1"></a>?rnorm</span></code></pre></div>
<p><img src="02_14.jpg" width="70%" style="display: block; margin: auto;" /></p>
<p>You can also search for help on functions in packages that are not attached. Or, you might want to know about a function but don’t know what package it belongs to. Either way, <code>??functionname</code> is a more general search option that will search any package in your library. Compare what happens if search for function <code>fitdistr()</code> using <code>?</code> and <code>??</code> (this function is in package <code>MASS</code>, and so not loaded by default). With <code>??</code>, R searches all installed packages and returns a browser window with every result that might match your search term.</p>
<div class="sourceCode" id="cb245"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb245-1"><a href="mod-02.html#cb245-1" tabindex="-1"></a>?fitdistr</span>
<span id="cb245-2"><a href="mod-02.html#cb245-2" tabindex="-1"></a><span class="do">## No documentation for ‘fitdistr’ in specified packages and libraries:</span></span>
<span id="cb245-3"><a href="mod-02.html#cb245-3" tabindex="-1"></a><span class="do">## you could try ‘??fitdistr’</span></span>
<span id="cb245-4"><a href="mod-02.html#cb245-4" tabindex="-1"></a></span>
<span id="cb245-5"><a href="mod-02.html#cb245-5" tabindex="-1"></a>??fitdistr</span></code></pre></div>
<p><img src="02_15.jpg" width="70%" style="display: block; margin: auto;" /></p>
<p>The R documentation pages all tend to follow the same format:</p>
<ul>
<li><strong>Description:</strong> brief description of what the function does.</li>
<li><strong>Usage:</strong> presentation of the function with its arguments in their typical format <em>and expected order</em>.</li>
<li><strong>Arguments:</strong> list of the arguments to the function <em>in order</em>, their required type/format, and allowed values. This section is very important. When my code isn’t working, the answer to why is usually in the arguments section.</li>
<li><strong>Details:</strong> additional information necessary to use the function.</li>
<li><strong>Value:</strong> description of the function’s outputs.</li>
<li><strong>Source:</strong> description of where the original algorithm or method came from.</li>
<li><strong>References:</strong> citations for the methods used.</li>
<li><strong>See also:</strong> list of similar or related R functions.</li>
<li><strong>Examples:</strong> illustration of how to use the function. This section is <em>very</em> important. If your code is not working, compare how you are using the function to the examples. Are you inputs in the right format? Did you name your arguments or put unnamed arguments in the wrong order? Careful examination and running of the examples will usually reveal why your code is not working.</li>
</ul>
<p>R documentation pages tend to be very terse and include very little in the way of background or explanation. They are generally meant to refresh the memory of people already familiar with R and with statistics, not to educate new R users or inexperienced statisticians. If you fall into one of the latter categories, or if you are trying to learn a new function, read on.</p>
</div>
<div id="r-vignettes" class="section level3 hasAnchor" number="2.10.2">
<h3><span class="header-section-number">2.10.2</span> R vignettes<a href="mod-02.html#r-vignettes" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>A <strong>vignette</strong> is a document demonstrating the use of an R package or its functions, with more details and explanation than in the help files. Not all packages offer vignettes. You can see what vignettes are available with the command <code>vignette()</code>. Specific vignettes are accessed by name and package. The examples below show how to get a list of available vignettes, available vignettes by package, and open specific vignettes.</p>
<div class="sourceCode" id="cb246"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb246-1"><a href="mod-02.html#cb246-1" tabindex="-1"></a><span class="co"># list vignettes in all installed packages</span></span>
<span id="cb246-2"><a href="mod-02.html#cb246-2" tabindex="-1"></a><span class="fu">vignette</span>()</span></code></pre></div>
<p><img src="02_16.jpg" width="90%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb247"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb247-1"><a href="mod-02.html#cb247-1" tabindex="-1"></a><span class="co"># list vignettes in package vegan</span></span>
<span id="cb247-2"><a href="mod-02.html#cb247-2" tabindex="-1"></a><span class="fu">vignette</span>(<span class="at">package=</span><span class="st">&quot;vegan&quot;</span>)</span></code></pre></div>
<p><img src="02_17.jpg" width="90%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb248"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb248-1"><a href="mod-02.html#cb248-1" tabindex="-1"></a><span class="co"># open vignette &quot;intro-vegan&quot; in local PDF reader</span></span>
<span id="cb248-2"><a href="mod-02.html#cb248-2" tabindex="-1"></a><span class="fu">vignette</span>(<span class="st">&quot;intro-vegan&quot;</span>, <span class="at">package=</span><span class="st">&quot;vegan&quot;</span>)</span></code></pre></div>
<p><img src="02_18.jpg" width="90%" style="display: block; margin: auto;" /></p>
</div>
<div id="official-r-project-resources" class="section level3 hasAnchor" number="2.10.3">
<h3><span class="header-section-number">2.10.3</span> Official R Project resources<a href="mod-02.html#official-r-project-resources" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="r-project-resources" class="section level4 hasAnchor" number="2.10.3.1">
<h4><span class="header-section-number">2.10.3.1</span> R Project resources<a href="mod-02.html#r-project-resources" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The R project maintains an <a href="https://www.r-project.org/help.html">official help page</a> with lists of places to get help. They also maintain a <a href="https://cran.r-project.org/manuals.html">list of manuals and documentation</a>. If you want a handy desk reference, you could do worse than the <a href="https://cran.r-project.org/doc/contrib/Short-refcard.pdf">R reference card</a> (accessed 2021-12-21).</p>
<p>The <a href="https://www.jstatsoft.org/index">Journal of Statistical Software</a> publishes papers about new statistical algorithms and software packages. Many of its papers are about new R packages and applications. The papers about new R packages and methods often include extensive code examples.</p>
</div>
<div id="r-task-views" class="section level4 hasAnchor" number="2.10.3.2">
<h4><span class="header-section-number">2.10.3.2</span> R task views<a href="mod-02.html#r-task-views" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>The <a href="https://cran.r-project.org/web/views/">Task Views</a> on CRAN contain frequently updated lists of packages that are useful for specific domains. For example, ecologists might check the <a href="https://cran.r-project.org/web/views/Environmetrics.html">Environmetrics</a>, <a href="https://cran.r-project.org/web/views/Hydrology.html">Hydrology</a> or <a href="https://cran.r-project.org/web/views/Spatial.html">Spatial</a> task views. Geneticists might be more interested in <a href="https://cran.r-project.org/web/views/Genetics.html">Genetics</a> or <a href="https://cran.r-project.org/web/views/Phylogenetics.html">Phylogenetics</a>. Physiologists and biochemists might be interested in <a href="https://cran.r-project.org/web/views/Pharmacokinetics.html">Pharmacokinetics</a>, <a href="https://cran.r-project.org/web/views/ClinicalTrials.html">Clinical Trials</a>, or <a href="https://cran.r-project.org/web/views/Survival.html">Survival</a>. I’ve spent a lot of time in the <a href="https://cran.r-project.org/web/views/Bayesian.html">Bayesian</a>, <a href="https://cran.r-project.org/web/views/NumericalMathematics.html">Numerical Mathematics</a>, and <a href="https://cran.r-project.org/web/views/MachineLearning.html">Machine Learning</a> task views. There’s something for everyone…even a <a href="https://cran.r-project.org/web/views/TeachingStatistics.html">Teaching Statistics</a> task view with resources to help you learn R and stats at the same time.</p>
<p>(All links accessed 2021-12-21).</p>
</div>
<div id="r-mailing-lists" class="section level4 hasAnchor" number="2.10.3.3">
<h4><span class="header-section-number">2.10.3.3</span> R mailing lists<a href="mod-02.html#r-mailing-lists" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>If your question has been asked before—and this is very likely—there may be an answer waiting for you on the <a href="https://www.r-project.org/mail.html">R Help mailing list</a>. Results from the R help mailing list will usually show up in Google results. You can submit a question to the R mailing list, but be very cautious about doing so. You will need to provide a short, clear, reproducible example in your question so that other people can see the issue you are having<a href="#fn36" class="footnote-ref" id="fnref36"><sup>36</sup></a>. You also need to be sure that your question hasn’t been asked before. Some members of the mailing list will respond with great hostility and condescension if you fail to follow those rules.</p>
<p>There are several <a href="https://stat.ethz.ch/mailman/listinfo">special interest groups (SIG)</a>, e-mail mailing lists for R users who work in specific fields. For example, <a href="https://stat.ethz.ch/mailman/listinfo/r-sig-ecology">R-sig-ecology</a> is an e-mail list for ecologists who use R. Other SIG of interest to biologists include <a href="https://stat.ethz.ch/mailman/listinfo/r-sig-genetics">R-sig-genetics</a> for genetics, <a href="https://stat.ethz.ch/mailman/listinfo/r-sig-mixed-models">R-sig-mixed-models</a> for mixed models, <a href="https://stat.ethz.ch/mailman/listinfo/r-sig-phylo">R-sig-phylo</a> for phylogenetics and related methods, and <a href="https://stat.ethz.ch/mailman/listinfo/r-sig-geo">R-sig-Geo</a> for GIS and spatial analysis.</p>
</div>
</div>
<div id="unofficial-online-resources" class="section level3 hasAnchor" number="2.10.4">
<h3><span class="header-section-number">2.10.4</span> Unofficial online resources<a href="mod-02.html#unofficial-online-resources" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Unofficial resources are becoming more and more abundant and useful. One of the most important sources for R help is <a href="https://stackoverflow.com">StackOverflow</a>. StackOverflow is a hugely popular forum site where programmers ask and answer questions, and has an active <a href="https://stackoverflow.com/questions/tagged/r">R section</a>. When (not if) you find a solution on StackOverflow, include a link to the answer in your code as a comment. This can save you some headache down the line when you need to remember how you solved a problem. Other useful resources include <a href="https://rseek.org/">RSeek.org</a>, a search engine for R resources; and <a href="https://www.r-bloggers.com/">R-Bloggers</a>, which aggregates many blogs about R and statistics.</p>
<p>Many R users maintain blogs or websites that may or may not be aggregated on R-Bloggers. You can usually find these via a Google search. I have used the pages on <a href="https://www.statmethods.net/">Quick-R</a> many times. The UCLA Statistical Consulting Group has an <a href="https://stats.oarc.ucla.edu/r/">excellent collection of tutorials</a>.</p>
</div>
<div id="r-books" class="section level3 hasAnchor" number="2.10.5">
<h3><span class="header-section-number">2.10.5</span> R books<a href="mod-02.html#r-books" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The last 10 years have seen an explosion of books published specifically about the R language and its uses. Some of these books are general introductions to R and statistics; others are very domain specific. Inclusion of a book here does not constitute specific endorsement. I own some of these, but others are included because they are common R-related references that you might find helpful.</p>
<ul>
<li>The <a href="https://www.springer.com/series/6991">Use R! series</a> contains many high-quality general and domain-specific books, although some of the older titles are a bit dated. For example:
<ul>
<li>A Beginner’s Guide to R by Alain Zuur et al.</li>
<li>Data Manipulation with R by Phil Spector</li>
<li>Introductory time series with R by P.S.P. Cowperthwaite and A.V. Metcalfe.</li>
<li>Bayesian computation with R by Jim Albert</li>
<li>Introductory Statistics with R by Peter Dalgaard</li>
<li>Morphometrics with R by Julien Claude</li>
</ul></li>
<li>The R Book by Michael Crawley</li>
<li>Advanced R by Hadley Wickham</li>
<li>R for Dummies by Andrie DeVries</li>
</ul>
<p>When I was first starting out with R I found several books invaluable. Having code examples printed where I could highlight, annotate, etc., was really helpful. Over time I’ve found myself relying less on textbooks and more on online sources (especially StackOverflow). Still, a there are a few books that I use frequently, for their statistical exposition more than their code examples. At least one of <span class="citation">Bolker (<a href="#ref-bolker2008ecological">2008</a>)</span>, <span class="citation">Kéry (<a href="#ref-kery2010">2010</a>)</span>, <span class="citation">Legendre and Legendre (<a href="#ref-legendre2012numerical">2012</a>)</span>, <span class="citation">Zuur et al. (<a href="#ref-zuur2007analysing">2007</a>)</span>, and <span class="citation">Zuur et al. (<a href="#ref-zuur2009mixed">2009a</a>)</span> is usually on my desk at any given time.</p>
</div>
<div id="two-reminders" class="section level3 hasAnchor" number="2.10.6">
<h3><span class="header-section-number">2.10.6</span> Two reminders<a href="mod-02.html#two-reminders" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>I said it at the beginning of this section, and it bears repeating: <strong><em>Don’t panic!</em></strong> R has a steep learning curve, but the pay-off is worth it. There are plenty of resources and people out there who can help you.</p>
<p>Finally, be sure to <strong><em>read the manual</em></strong> when you are stuck. The R help pages and code examples usually have a solution. Make sure your inputs are in the right format. Check for syntax and spelling error–including capitalization! As a funny meme on I saw on Reddit said, <em>6 hours of debugging can save you 5 minutes of reading documentation</em><a href="#fn37" class="footnote-ref" id="fnref37"><sup>37</sup></a>.</p>

</div>
</div>
</div>
<h3>Literature Cited<a href="literature-cited.html#literature-cited" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0" line-spacing="2">
<div id="ref-bolker2008ecological" class="csl-entry">
Bolker, B. M. 2008. Ecological models and data in <span>R</span>. Princeton University Press, Princeton, New Jersey, USA.
</div>
<div id="ref-carnell2019" class="csl-entry">
Carnell, R. 2019. <a href="https://CRAN.R-project.org/package=triangle">Triangle: Provides the standard distribution functions for the triangle distribution</a>.
</div>
<div id="ref-gould1974" class="csl-entry">
Gould, S. J. 1974. <a href="https://doi.org/10.2307/2407322">The origin and function of ’bizarre’ structures: Antler size and skull size in the ’irish elk,’ <em><span>M</span>egaloceros giganteus</em></a>. Evolution 28:191–220.
</div>
<div id="ref-kery2010" class="csl-entry">
Kéry, M. 2010. Introduction to <span>W</span>in<span>BUGS</span> for ecologists: <span>B</span>ayesian approach to regression, <span>ANOVA</span>, mixed models and related analyses. Academic Press.
</div>
<div id="ref-legendre2012numerical" class="csl-entry">
Legendre, P., and L. Legendre. 2012. Numerical ecology. Elsevier, Amsterdam.
</div>
<div id="ref-spector2008" class="csl-entry">
Spector, P. 2008. Data manipulation with <span>R</span>. Springer Science+Business Media, New York.
</div>
<div id="ref-tufte2001visual" class="csl-entry">
Tufte, E. 2001. The visual display of quantitative information. Cheshire: Graphic Press.–2001.–213 p.
</div>
<div id="ref-wickham2019" class="csl-entry">
Wickham, H., M. Averick, J. Bryan, W. Chang, L. D. McGowan, R. François, G. Grolemund, A. Hayes, L. Henry, J. Hester, M. Kuhn, T. L. Pedersen, E. Miller, S. M. Bache, K. Müller, J. Ooms, D. Robinson, D. P. Seidel, V. Spinu, K. Takahashi, D. Vaughan, C. Wilke, K. Woo, and H. Yutani. 2019. <a href="https://doi.org/10.21105/joss.01686">Welcome to the <span class="nocase">tidyverse</span></a>. Journal of Open Source Software 4:1686.
</div>
<div id="ref-zuur2007analysing" class="csl-entry">
Zuur, A. F., E. N. Ieno, and G. M. Smith. 2007. Analysing ecological data. Springer Science+Business Media, New York.
</div>
<div id="ref-zuur2009mixed" class="csl-entry">
Zuur, A. F., E. N. Ieno, N. J. Walker, A. A. Saveliev, G. M. Smith, and others. 2009a. Mixed effects models and extensions in ecology with <span>R</span>. Springer, New York.
</div>
</div>
<div class="footnotes">
<hr />
<ol start="22">
<li id="fn22"><p>This essay says a lot of things I’ve thought for years, but have never been able to articulate.<a href="mod-02.html#fnref22" class="footnote-back">↩︎</a></p></li>
<li id="fn23"><p>In my humble opinion, this aspect of <code>tidyverse</code> is just insane.<a href="mod-02.html#fnref23" class="footnote-back">↩︎</a></p></li>
<li id="fn24"><p>Spoiler alert: it doesn’t.<a href="mod-02.html#fnref24" class="footnote-back">↩︎</a></p></li>
<li id="fn25"><p>The program RStudio is a product of a company that used to call itself RStudio, but which recently changed its name to Posit. They did this to reflect their shift away from being mainly R-focused, to embracing other data science tools like Python<a href="mod-02.html#fnref25" class="footnote-back">↩︎</a></p></li>
<li id="fn26"><p>Yours might look different than mine, because I set some graphics options.<a href="mod-02.html#fnref26" class="footnote-back">↩︎</a></p></li>
<li id="fn27"><p>Your RStudio might look different at first. You can change the color scheme by going to Tools - Global Options - Appearance and selecting a theme. Here I’m using the “Classic” theme with size 12 Courier New font and 110% zoom.<a href="mod-02.html#fnref27" class="footnote-back">↩︎</a></p></li>
<li id="fn28"><p>Image: F. Atirador. URL: <a href="https://upload.wikimedia.org/wikipedia/commons/5/5a/Irish_Elk_Side_%28white_background%29.jpg" class="uri">https://upload.wikimedia.org/wikipedia/commons/5/5a/Irish_Elk_Side_%28white_background%29.jpg</a> (accessed 2021-12-20).<a href="mod-02.html#fnref28" class="footnote-back">↩︎</a></p></li>
<li id="fn29"><p>This issue seems to confuse a lot students who are Mac users, so I’m belaboring the point here.<a href="mod-02.html#fnref29" class="footnote-back">↩︎</a></p></li>
<li id="fn30"><p>As we’ll discuss in the <a href="mod-03.html#mod-03-inout">data manipulation module</a>, naming R objects is a bit of an art. Names should be informative yet brief. I use a standard list of names like “<code>dat</code>”, “<code>met</code>”, and “<code>res</code>”, etc. in all of my projects because they are short but easy to identify (“data”, “metadata”, and “results”, respectively).<a href="mod-02.html#fnref30" class="footnote-back">↩︎</a></p></li>
<li id="fn31"><p>I don’t recommend doing this, as the desktop can become easily cluttered.<a href="mod-02.html#fnref31" class="footnote-back">↩︎</a></p></li>
<li id="fn32"><p>This is true if there are no random elements in your code. If you draw any random numbers, you will need to set the random number seed in your script to ensure exact reproducibility.<a href="mod-02.html#fnref32" class="footnote-back">↩︎</a></p></li>
<li id="fn33"><p>Recently I had to revisit an analysis I’d done 7 years ago and update it with new parameters. If 2015 me had not written copious comments in the code, 2022 me might have had to rewrite a lot of code.<a href="mod-02.html#fnref33" class="footnote-back">↩︎</a></p></li>
<li id="fn34"><p>It never even occurred to me to try this until I was prepping these course notes.<a href="mod-02.html#fnref34" class="footnote-back">↩︎</a></p></li>
<li id="fn35"><p>My personal record is a 7-dimensional array, which I used in a community simulation model. The model included many species-, age-, and sex-specific demographic parameters that had to be extracted and summarized, and because of the size of the model an array-based solution was more computationally-efficient than a list- or data frame-based solution. I don’t recommend building arrays that complex…it was a nightmare to work with.<a href="mod-02.html#fnref35" class="footnote-back">↩︎</a></p></li>
<li id="fn36"><p>This means a <em>very</em> minimal code example that people can use to get the same problem that you are having. Don’t post your entire dataset or include many lines of pre-processing steps or other code that led up to your problem. Use base R functions or built-in datasets to create a small dataset or set of values that will immediately provoke the same error code. Often the process of doing this will help you solve your own problem!<a href="mod-02.html#fnref36" class="footnote-back">↩︎</a></p></li>
<li id="fn37"><p>I think this was originally a tweet that went viral, but the joke spread through the internet and now you can buy it on a t-shirt. Regardless, this phenomenon is much older than Twitter.<a href="mod-02.html#fnref37" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="mod-01.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="mod-03.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/02-intro.rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["applied_biol_data_analysis_20220816.pdf", "applied_biol_data_analysis_20220816.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
