<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Module 6 Exploratory data analysis 3: Multivariate data exploration | Applied Biological Data Analysis: Statistics and R for Biologists</title>
  <meta name="description" content="Helping biologists to become more informed users of R and statistical methods." />
  <meta name="generator" content="bookdown 0.28 and GitBook 2.6.7" />

  <meta property="og:title" content="Module 6 Exploratory data analysis 3: Multivariate data exploration | Applied Biological Data Analysis: Statistics and R for Biologists" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://greenquanteco.github.io/book/lab_logo_02.jpg" />
  <meta property="og:description" content="Helping biologists to become more informed users of R and statistical methods." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Module 6 Exploratory data analysis 3: Multivariate data exploration | Applied Biological Data Analysis: Statistics and R for Biologists" />
  
  <meta name="twitter:description" content="Helping biologists to become more informed users of R and statistical methods." />
  <meta name="twitter:image" content="https://greenquanteco.github.io/book/lab_logo_02.jpg" />

<meta name="author" content="Nick Green, Kennesaw State University" />


<meta name="date" content="2022-08-16" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="fav.ico" type="image/x-icon" />
<link rel="prev" href="mod-05.html"/>
<link rel="next" href="mod-07.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Applied biological data analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#license-and-permissions"><i class="fa fa-check"></i>License and permissions</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-description"><i class="fa fa-check"></i>Course description</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-objectives"><i class="fa fa-check"></i>Course objectives</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-requirements"><i class="fa fa-check"></i>Course requirements</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#recommended-reading"><i class="fa fa-check"></i>Recommended reading</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-organization"><i class="fa fa-check"></i>Course organization</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#about-the-author"><i class="fa fa-check"></i>About the author</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="mod-01.html"><a href="mod-01.html"><i class="fa fa-check"></i><b>1</b> Statistics in modern biology</a>
<ul>
<li class="chapter" data-level="1.1" data-path="mod-01.html"><a href="mod-01.html#overview"><i class="fa fa-check"></i><b>1.1</b> Overview</a></li>
<li class="chapter" data-level="1.2" data-path="mod-01.html"><a href="mod-01.html#statistics-in-modern-biology"><i class="fa fa-check"></i><b>1.2</b> Statistics in modern biology</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="mod-01.html"><a href="mod-01.html#the-scientific-method"><i class="fa fa-check"></i><b>1.2.1</b> The scientific method</a></li>
<li class="chapter" data-level="1.2.2" data-path="mod-01.html"><a href="mod-01.html#example-data-analysis"><i class="fa fa-check"></i><b>1.2.2</b> Example data analysis</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="mod-01.html"><a href="mod-01.html#misuses-of-statistics"><i class="fa fa-check"></i><b>1.3</b> Misuses of statistics</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="mod-01.html"><a href="mod-01.html#proving-the-trivial-and-meaningless-hypotheses"><i class="fa fa-check"></i><b>1.3.1</b> “Proving” the trivial and meaningless hypotheses</a></li>
<li class="chapter" data-level="1.3.2" data-path="mod-01.html"><a href="mod-01.html#inappropriate-methods"><i class="fa fa-check"></i><b>1.3.2</b> Inappropriate methods</a></li>
<li class="chapter" data-level="1.3.3" data-path="mod-01.html"><a href="mod-01.html#p-hacking-and-data-dredging"><i class="fa fa-check"></i><b>1.3.3</b> <em>P</em>-hacking and data dredging</a></li>
<li class="chapter" data-level="1.3.4" data-path="mod-01.html"><a href="mod-01.html#inadequate-sample-sizes-and-pseudoreplication"><i class="fa fa-check"></i><b>1.3.4</b> Inadequate sample sizes and pseudoreplication</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="mod-01.html"><a href="mod-01.html#mod-01-nhst"><i class="fa fa-check"></i><b>1.4</b> <em>P</em>-values and null hypothesis significance testing (NHST)</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="mod-01.html"><a href="mod-01.html#definition"><i class="fa fa-check"></i><b>1.4.1</b> Definition</a></li>
<li class="chapter" data-level="1.4.2" data-path="mod-01.html"><a href="mod-01.html#aside-another-way-to-frame-null-hypotheses"><i class="fa fa-check"></i><b>1.4.2</b> Aside: Another way to frame null hypotheses</a></li>
<li class="chapter" data-level="1.4.3" data-path="mod-01.html"><a href="mod-01.html#history-and-status-of-p-values"><i class="fa fa-check"></i><b>1.4.3</b> History and status of <em>P</em>-values</a></li>
<li class="chapter" data-level="1.4.4" data-path="mod-01.html"><a href="mod-01.html#where-p-values-come-from"><i class="fa fa-check"></i><b>1.4.4</b> Where <em>P</em>-values come from</a></li>
<li class="chapter" data-level="1.4.5" data-path="mod-01.html"><a href="mod-01.html#what-p-values-mean-and-do-not-mean"><i class="fa fa-check"></i><b>1.4.5</b> What <em>P</em>-values mean and do not mean</a></li>
<li class="chapter" data-level="1.4.6" data-path="mod-01.html"><a href="mod-01.html#do-you-need-a-p-value"><i class="fa fa-check"></i><b>1.4.6</b> Do you need a <em>P</em>-value?</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="mod-01.html"><a href="mod-01.html#alternatives-to-nhst"><i class="fa fa-check"></i><b>1.5</b> Alternatives to NHST</a>
<ul>
<li class="chapter" data-level="1.5.1" data-path="mod-01.html"><a href="mod-01.html#bayesian-inference"><i class="fa fa-check"></i><b>1.5.1</b> Bayesian inference</a></li>
<li class="chapter" data-level="1.5.2" data-path="mod-01.html"><a href="mod-01.html#information-theoretic-methods"><i class="fa fa-check"></i><b>1.5.2</b> Information-theoretic methods</a></li>
<li class="chapter" data-level="1.5.3" data-path="mod-01.html"><a href="mod-01.html#machine-learning"><i class="fa fa-check"></i><b>1.5.3</b> Machine learning</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="mod-02.html"><a href="mod-02.html"><i class="fa fa-check"></i><b>2</b> Introduction to R</a>
<ul>
<li class="chapter" data-level="2.1" data-path="mod-02.html"><a href="mod-02.html#getting-started-with-r"><i class="fa fa-check"></i><b>2.1</b> Getting started with R</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="mod-02.html"><a href="mod-02.html#what-is-r"><i class="fa fa-check"></i><b>2.1.1</b> What is R?</a></li>
<li class="chapter" data-level="2.1.2" data-path="mod-02.html"><a href="mod-02.html#advantages-of-r"><i class="fa fa-check"></i><b>2.1.2</b> Advantages of R</a></li>
<li class="chapter" data-level="2.1.3" data-path="mod-02.html"><a href="mod-02.html#disadvantages-of-r"><i class="fa fa-check"></i><b>2.1.3</b> Disadvantages of R</a></li>
<li class="chapter" data-level="2.1.4" data-path="mod-02.html"><a href="mod-02.html#base-r-and-vs.-tidyverse"><i class="fa fa-check"></i><b>2.1.4</b> Base R and (vs.?) <code>tidyverse</code></a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="mod-02.html"><a href="mod-02.html#download-and-install-r-and-rstudio"><i class="fa fa-check"></i><b>2.2</b> Download and install R (and RStudio)</a></li>
<li class="chapter" data-level="2.3" data-path="mod-02.html"><a href="mod-02.html#using-r"><i class="fa fa-check"></i><b>2.3</b> Using R</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="mod-02.html"><a href="mod-02.html#using-the-base-r-gui"><i class="fa fa-check"></i><b>2.3.1</b> Using the base R GUI</a></li>
<li class="chapter" data-level="2.3.2" data-path="mod-02.html"><a href="mod-02.html#using-r-in-rstudio"><i class="fa fa-check"></i><b>2.3.2</b> Using R in RStudio</a></li>
<li class="chapter" data-level="2.3.3" data-path="mod-02.html"><a href="mod-02.html#using-r-with-other-programs"><i class="fa fa-check"></i><b>2.3.3</b> Using R with other programs</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="mod-02.html"><a href="mod-02.html#mod-02-first"><i class="fa fa-check"></i><b>2.4</b> A first R session</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="mod-02.html"><a href="mod-02.html#import-data"><i class="fa fa-check"></i><b>2.4.1</b> Import data</a></li>
<li class="chapter" data-level="2.4.2" data-path="mod-02.html"><a href="mod-02.html#explore-and-visualize-data"><i class="fa fa-check"></i><b>2.4.2</b> Explore and visualize data</a></li>
<li class="chapter" data-level="2.4.3" data-path="mod-02.html"><a href="mod-02.html#transform-data"><i class="fa fa-check"></i><b>2.4.3</b> Transform data</a></li>
<li class="chapter" data-level="2.4.4" data-path="mod-02.html"><a href="mod-02.html#analyze-data"><i class="fa fa-check"></i><b>2.4.4</b> Analyze data</a></li>
<li class="chapter" data-level="2.4.5" data-path="mod-02.html"><a href="mod-02.html#write-out-results"><i class="fa fa-check"></i><b>2.4.5</b> Write out results</a></li>
<li class="chapter" data-level="2.4.6" data-path="mod-02.html"><a href="mod-02.html#save-your-work"><i class="fa fa-check"></i><b>2.4.6</b> Save your work?</a></li>
<li class="chapter" data-level="2.4.7" data-path="mod-02.html"><a href="mod-02.html#whats-next"><i class="fa fa-check"></i><b>2.4.7</b> What’s next?</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="mod-02.html"><a href="mod-02.html#write-and-execute-commands-in-the-r-console"><i class="fa fa-check"></i><b>2.5</b> Write and execute commands in the R console</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="mod-02.html"><a href="mod-02.html#r-commandsbasics"><i class="fa fa-check"></i><b>2.5.1</b> R commands–basics</a></li>
<li class="chapter" data-level="2.5.2" data-path="mod-02.html"><a href="mod-02.html#elements-of-r-code"><i class="fa fa-check"></i><b>2.5.2</b> Elements of R code</a></li>
<li class="chapter" data-level="2.5.3" data-path="mod-02.html"><a href="mod-02.html#the-r-workspace"><i class="fa fa-check"></i><b>2.5.3</b> The R workspace</a></li>
<li class="chapter" data-level="2.5.4" data-path="mod-02.html"><a href="mod-02.html#r-code-basics-assignment-and-operators"><i class="fa fa-check"></i><b>2.5.4</b> R code basics: assignment and operators</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="mod-02.html"><a href="mod-02.html#mod-02-struct"><i class="fa fa-check"></i><b>2.6</b> Basic R data structures</a>
<ul>
<li class="chapter" data-level="2.6.1" data-path="mod-02.html"><a href="mod-02.html#vectors"><i class="fa fa-check"></i><b>2.6.1</b> Vectors</a></li>
<li class="chapter" data-level="2.6.2" data-path="mod-02.html"><a href="mod-02.html#data-frames"><i class="fa fa-check"></i><b>2.6.2</b> Data frames</a></li>
<li class="chapter" data-level="2.6.3" data-path="mod-02.html"><a href="mod-02.html#matrices-and-arrays"><i class="fa fa-check"></i><b>2.6.3</b> Matrices and arrays</a></li>
<li class="chapter" data-level="2.6.4" data-path="mod-02.html"><a href="mod-02.html#lists"><i class="fa fa-check"></i><b>2.6.4</b> Lists</a></li>
<li class="chapter" data-level="2.6.5" data-path="mod-02.html"><a href="mod-02.html#s3-s4-and-other-paradigms"><i class="fa fa-check"></i><b>2.6.5</b> S3, S4, and other paradigms</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="mod-02.html"><a href="mod-02.html#r-data-types"><i class="fa fa-check"></i><b>2.7</b> R data types</a>
<ul>
<li class="chapter" data-level="2.7.1" data-path="mod-02.html"><a href="mod-02.html#character-type"><i class="fa fa-check"></i><b>2.7.1</b> Character type</a></li>
<li class="chapter" data-level="2.7.2" data-path="mod-02.html"><a href="mod-02.html#numeric-type"><i class="fa fa-check"></i><b>2.7.2</b> Numeric type</a></li>
<li class="chapter" data-level="2.7.3" data-path="mod-02.html"><a href="mod-02.html#integer-type"><i class="fa fa-check"></i><b>2.7.3</b> Integer type</a></li>
<li class="chapter" data-level="2.7.4" data-path="mod-02.html"><a href="mod-02.html#logical-type"><i class="fa fa-check"></i><b>2.7.4</b> Logical type</a></li>
<li class="chapter" data-level="2.7.5" data-path="mod-02.html"><a href="mod-02.html#special-values"><i class="fa fa-check"></i><b>2.7.5</b> Special values</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="mod-02.html"><a href="mod-02.html#manage-r-code-as-scripts-.r-files"><i class="fa fa-check"></i><b>2.8</b> Manage R code as scripts (.r files)</a></li>
<li class="chapter" data-level="2.9" data-path="mod-02.html"><a href="mod-02.html#manage-and-use-r-packages"><i class="fa fa-check"></i><b>2.9</b> Manage and use R packages</a>
<ul>
<li class="chapter" data-level="2.9.1" data-path="mod-02.html"><a href="mod-02.html#your-r-library"><i class="fa fa-check"></i><b>2.9.1</b> Your R library</a></li>
<li class="chapter" data-level="2.9.2" data-path="mod-02.html"><a href="mod-02.html#installing-packages-using-the-r-gui"><i class="fa fa-check"></i><b>2.9.2</b> Installing packages using the R GUI</a></li>
<li class="chapter" data-level="2.9.3" data-path="mod-02.html"><a href="mod-02.html#installing-packages-in-rstudio"><i class="fa fa-check"></i><b>2.9.3</b> Installing packages in RStudio</a></li>
<li class="chapter" data-level="2.9.4" data-path="mod-02.html"><a href="mod-02.html#installing-packages-using-the-r-console"><i class="fa fa-check"></i><b>2.9.4</b> Installing packages using the R console</a></li>
<li class="chapter" data-level="2.9.5" data-path="mod-02.html"><a href="mod-02.html#working-with-packages-in-r"><i class="fa fa-check"></i><b>2.9.5</b> Working with packages in R</a></li>
<li class="chapter" data-level="2.9.6" data-path="mod-02.html"><a href="mod-02.html#package-dependencies"><i class="fa fa-check"></i><b>2.9.6</b> Package dependencies</a></li>
<li class="chapter" data-level="2.9.7" data-path="mod-02.html"><a href="mod-02.html#citing-packages"><i class="fa fa-check"></i><b>2.9.7</b> Citing packages</a></li>
</ul></li>
<li class="chapter" data-level="2.10" data-path="mod-02.html"><a href="mod-02.html#r-documentation"><i class="fa fa-check"></i><b>2.10</b> R documentation</a>
<ul>
<li class="chapter" data-level="2.10.1" data-path="mod-02.html"><a href="mod-02.html#documentation-help-files"><i class="fa fa-check"></i><b>2.10.1</b> Documentation (help) files</a></li>
<li class="chapter" data-level="2.10.2" data-path="mod-02.html"><a href="mod-02.html#r-vignettes"><i class="fa fa-check"></i><b>2.10.2</b> R vignettes</a></li>
<li class="chapter" data-level="2.10.3" data-path="mod-02.html"><a href="mod-02.html#official-r-project-resources"><i class="fa fa-check"></i><b>2.10.3</b> Official R Project resources</a></li>
<li class="chapter" data-level="2.10.4" data-path="mod-02.html"><a href="mod-02.html#unofficial-online-resources"><i class="fa fa-check"></i><b>2.10.4</b> Unofficial online resources</a></li>
<li class="chapter" data-level="2.10.5" data-path="mod-02.html"><a href="mod-02.html#r-books"><i class="fa fa-check"></i><b>2.10.5</b> R books</a></li>
<li class="chapter" data-level="2.10.6" data-path="mod-02.html"><a href="mod-02.html#two-reminders"><i class="fa fa-check"></i><b>2.10.6</b> Two reminders</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="mod-03.html"><a href="mod-03.html"><i class="fa fa-check"></i><b>3</b> Data manipulation with R</a>
<ul>
<li class="chapter" data-level="3.1" data-path="mod-03.html"><a href="mod-03.html#mod-03-inout"><i class="fa fa-check"></i><b>3.1</b> Data import and export</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="mod-03.html"><a href="mod-03.html#importing-data-preliminaries"><i class="fa fa-check"></i><b>3.1.1</b> Importing data: preliminaries</a></li>
<li class="chapter" data-level="3.1.2" data-path="mod-03.html"><a href="mod-03.html#importing-data-from-text-files-with-read.csv-and-read.table"><i class="fa fa-check"></i><b>3.1.2</b> Importing data from text files with <code>read.csv()</code> and <code>read.table()</code></a></li>
<li class="chapter" data-level="3.1.3" data-path="mod-03.html"><a href="mod-03.html#importing-data-from-saved-workspaces"><i class="fa fa-check"></i><b>3.1.3</b> Importing data from saved workspaces</a></li>
<li class="chapter" data-level="3.1.4" data-path="mod-03.html"><a href="mod-03.html#importing-data-special-cases"><i class="fa fa-check"></i><b>3.1.4</b> Importing data: special cases:</a></li>
<li class="chapter" data-level="3.1.5" data-path="mod-03.html"><a href="mod-03.html#export-data-from-r"><i class="fa fa-check"></i><b>3.1.5</b> Export data from R</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="mod-03.html"><a href="mod-03.html#making-values-in-r"><i class="fa fa-check"></i><b>3.2</b> Making values in R</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="mod-03.html"><a href="mod-03.html#producing-arbitrary-values-with-c"><i class="fa fa-check"></i><b>3.2.1</b> Producing arbitrary values with <code>c()</code></a></li>
<li class="chapter" data-level="3.2.2" data-path="mod-03.html"><a href="mod-03.html#generating-regular-values"><i class="fa fa-check"></i><b>3.2.2</b> Generating regular values</a></li>
<li class="chapter" data-level="3.2.3" data-path="mod-03.html"><a href="mod-03.html#generating-random-values"><i class="fa fa-check"></i><b>3.2.3</b> Generating random values</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="mod-03.html"><a href="mod-03.html#selecting-data-with"><i class="fa fa-check"></i><b>3.3</b> Selecting data with <code>[]</code></a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="mod-03.html"><a href="mod-03.html#mod-03-brackets"><i class="fa fa-check"></i><b>3.3.1</b> Basics of brackets</a></li>
<li class="chapter" data-level="3.3.2" data-path="mod-03.html"><a href="mod-03.html#extracting-and-selecting-data-with-logical-tests"><i class="fa fa-check"></i><b>3.3.2</b> Extracting and selecting data with logical tests</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="mod-03.html"><a href="mod-03.html#managing-dates-and-characters"><i class="fa fa-check"></i><b>3.4</b> Managing dates and characters</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="mod-03.html"><a href="mod-03.html#temporal-data-and-dates"><i class="fa fa-check"></i><b>3.4.1</b> Temporal data and dates</a></li>
<li class="chapter" data-level="3.4.2" data-path="mod-03.html"><a href="mod-03.html#character-data-text"><i class="fa fa-check"></i><b>3.4.2</b> Character data (text)</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="mod-03.html"><a href="mod-03.html#mod-03-dataframe"><i class="fa fa-check"></i><b>3.5</b> Data frame management</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="mod-03.html"><a href="mod-03.html#data-frame-structure"><i class="fa fa-check"></i><b>3.5.1</b> Data frame structure</a></li>
<li class="chapter" data-level="3.5.2" data-path="mod-03.html"><a href="mod-03.html#common-data-frame-operations"><i class="fa fa-check"></i><b>3.5.2</b> Common data frame operations</a></li>
<li class="chapter" data-level="3.5.3" data-path="mod-03.html"><a href="mod-03.html#other-data-frame-operations"><i class="fa fa-check"></i><b>3.5.3</b> Other data frame operations</a></li>
<li class="chapter" data-level="3.5.4" data-path="mod-03.html"><a href="mod-03.html#mod-03-reshape"><i class="fa fa-check"></i><b>3.5.4</b> Reshaping data frames</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="mod-04.html"><a href="mod-04.html"><i class="fa fa-check"></i><b>4</b> Exploratory data analysis 1: Data description and summarization</a>
<ul>
<li class="chapter" data-level="4.1" data-path="mod-04.html"><a href="mod-04.html#motivation"><i class="fa fa-check"></i><b>4.1</b> Motivation</a></li>
<li class="chapter" data-level="4.2" data-path="mod-04.html"><a href="mod-04.html#descriptive-and-summary-statistics"><i class="fa fa-check"></i><b>4.2</b> Descriptive and summary statistics</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="mod-04.html"><a href="mod-04.html#basic-summary-statistics"><i class="fa fa-check"></i><b>4.2.1</b> Basic summary statistics</a></li>
<li class="chapter" data-level="4.2.2" data-path="mod-04.html"><a href="mod-04.html#mod-04-agg"><i class="fa fa-check"></i><b>4.2.2</b> Summarizing by groups (AKA: pivot tables)</a></li>
<li class="chapter" data-level="4.2.3" data-path="mod-04.html"><a href="mod-04.html#summarizing-data-with-the-apply-family"><i class="fa fa-check"></i><b>4.2.3</b> Summarizing data with the <code>apply()</code> family</a></li>
<li class="chapter" data-level="4.2.4" data-path="mod-04.html"><a href="mod-04.html#mod-04-tab"><i class="fa fa-check"></i><b>4.2.4</b> Tabulation (frequency tables)</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="mod-04.html"><a href="mod-04.html#mod-04-vis"><i class="fa fa-check"></i><b>4.3</b> Visualizing single variables</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="mod-04.html"><a href="mod-04.html#boxplots-aka-box-and-whisker-plots"><i class="fa fa-check"></i><b>4.3.1</b> Boxplots (aka: box-and-whisker plots)</a></li>
<li class="chapter" data-level="4.3.2" data-path="mod-04.html"><a href="mod-04.html#histograms"><i class="fa fa-check"></i><b>4.3.2</b> Histograms</a></li>
<li class="chapter" data-level="4.3.3" data-path="mod-04.html"><a href="mod-04.html#mod04-pdfplots"><i class="fa fa-check"></i><b>4.3.3</b> Kernel density plots</a></li>
<li class="chapter" data-level="4.3.4" data-path="mod-04.html"><a href="mod-04.html#empirical-cumulative-density-function-ecdf-plots"><i class="fa fa-check"></i><b>4.3.4</b> Empirical cumulative density function (ECDF) plots</a></li>
<li class="chapter" data-level="4.3.5" data-path="mod-04.html"><a href="mod-04.html#quantile-quantile-qq-plots"><i class="fa fa-check"></i><b>4.3.5</b> Quantile-quantile (QQ) plots</a></li>
<li class="chapter" data-level="4.3.6" data-path="mod-04.html"><a href="mod-04.html#how-should-i-plot-my-data"><i class="fa fa-check"></i><b>4.3.6</b> How should I plot my data?</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="mod-05.html"><a href="mod-05.html"><i class="fa fa-check"></i><b>5</b> Exploratory data analysis 2: Probability distributions for biology</a>
<ul>
<li class="chapter" data-level="5.1" data-path="mod-05.html"><a href="mod-05.html#mod-05-dists1"><i class="fa fa-check"></i><b>5.1</b> Statistical distributions</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="mod-05.html"><a href="mod-05.html#probability-distributions-in-r"><i class="fa fa-check"></i><b>5.1.1</b> Probability distributions in R</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="mod-05.html"><a href="mod-05.html#mod-05-discrete"><i class="fa fa-check"></i><b>5.2</b> Discrete distributions</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="mod-05.html"><a href="mod-05.html#bernoulli-distribution"><i class="fa fa-check"></i><b>5.2.1</b> Bernoulli distribution</a></li>
<li class="chapter" data-level="5.2.2" data-path="mod-05.html"><a href="mod-05.html#binomial-distribution"><i class="fa fa-check"></i><b>5.2.2</b> Binomial distribution</a></li>
<li class="chapter" data-level="5.2.3" data-path="mod-05.html"><a href="mod-05.html#poisson-distribution"><i class="fa fa-check"></i><b>5.2.3</b> Poisson distribution</a></li>
<li class="chapter" data-level="5.2.4" data-path="mod-05.html"><a href="mod-05.html#negative-binomial-distribution"><i class="fa fa-check"></i><b>5.2.4</b> Negative binomial distribution</a></li>
<li class="chapter" data-level="5.2.5" data-path="mod-05.html"><a href="mod-05.html#geometric-distribution"><i class="fa fa-check"></i><b>5.2.5</b> Geometric distribution</a></li>
<li class="chapter" data-level="5.2.6" data-path="mod-05.html"><a href="mod-05.html#beta-binomial-distribution"><i class="fa fa-check"></i><b>5.2.6</b> Beta-binomial distribution</a></li>
<li class="chapter" data-level="5.2.7" data-path="mod-05.html"><a href="mod-05.html#multinomial-distribution"><i class="fa fa-check"></i><b>5.2.7</b> Multinomial distribution</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="mod-05.html"><a href="mod-05.html#mod-05-condists"><i class="fa fa-check"></i><b>5.3</b> Continuous distributions</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="mod-05.html"><a href="mod-05.html#uniform-distribution"><i class="fa fa-check"></i><b>5.3.1</b> Uniform distribution</a></li>
<li class="chapter" data-level="5.3.2" data-path="mod-05.html"><a href="mod-05.html#mod-05-norm"><i class="fa fa-check"></i><b>5.3.2</b> Normal distribution</a></li>
<li class="chapter" data-level="5.3.3" data-path="mod-05.html"><a href="mod-05.html#lognormal-distribution"><i class="fa fa-check"></i><b>5.3.3</b> Lognormal distribution</a></li>
<li class="chapter" data-level="5.3.4" data-path="mod-05.html"><a href="mod-05.html#gamma-distribution"><i class="fa fa-check"></i><b>5.3.4</b> Gamma distribution</a></li>
<li class="chapter" data-level="5.3.5" data-path="mod-05.html"><a href="mod-05.html#beta-distribution"><i class="fa fa-check"></i><b>5.3.5</b> Beta distribution</a></li>
<li class="chapter" data-level="5.3.6" data-path="mod-05.html"><a href="mod-05.html#dirichlet-distribution"><i class="fa fa-check"></i><b>5.3.6</b> Dirichlet distribution</a></li>
<li class="chapter" data-level="5.3.7" data-path="mod-05.html"><a href="mod-05.html#exponential-distribution"><i class="fa fa-check"></i><b>5.3.7</b> Exponential distribution</a></li>
<li class="chapter" data-level="5.3.8" data-path="mod-05.html"><a href="mod-05.html#triangular-distribution"><i class="fa fa-check"></i><b>5.3.8</b> Triangular distribution</a></li>
<li class="chapter" data-level="5.3.9" data-path="mod-05.html"><a href="mod-05.html#distributions-summary"><i class="fa fa-check"></i><b>5.3.9</b> Distributions summary</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="mod-05.html"><a href="mod-05.html#mod-05-dists2"><i class="fa fa-check"></i><b>5.4</b> Fitting and testing distributions</a>
<ul>
<li class="chapter" data-level="5.4.1" data-path="mod-05.html"><a href="mod-05.html#estimating-distributional-parameters"><i class="fa fa-check"></i><b>5.4.1</b> Estimating distributional parameters</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="mod-05.html"><a href="mod-05.html#graphical-methods-for-examining-distributions"><i class="fa fa-check"></i><b>5.5</b> Graphical methods for examining distributions</a>
<ul>
<li class="chapter" data-level="5.5.1" data-path="mod-05.html"><a href="mod-05.html#comparing-distributions-with-pdf-or-pmf-plots"><i class="fa fa-check"></i><b>5.5.1</b> Comparing distributions with PDF or PMF plots</a></li>
<li class="chapter" data-level="5.5.2" data-path="mod-05.html"><a href="mod-05.html#comparing-distributions-using-cdf-and-ecdf"><i class="fa fa-check"></i><b>5.5.2</b> Comparing distributions using CDF and ECDF</a></li>
<li class="chapter" data-level="5.5.3" data-path="mod-05.html"><a href="mod-05.html#comparing-distributions-using-qq-plots"><i class="fa fa-check"></i><b>5.5.3</b> Comparing distributions using QQ plots</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="mod-05.html"><a href="mod-05.html#formal-tests-for-distributions"><i class="fa fa-check"></i><b>5.6</b> Formal tests for distributions</a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="mod-05.html"><a href="mod-05.html#tests-for-normality"><i class="fa fa-check"></i><b>5.6.1</b> Tests for normality</a></li>
<li class="chapter" data-level="5.6.2" data-path="mod-05.html"><a href="mod-05.html#kolmogorov-smirnov-tests"><i class="fa fa-check"></i><b>5.6.2</b> Kolmogorov-Smirnov tests</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="mod-05.html"><a href="mod-05.html#mod-05-trans"><i class="fa fa-check"></i><b>5.7</b> Data transformations</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="mod-05.html"><a href="mod-05.html#why-transform"><i class="fa fa-check"></i><b>5.7.1</b> Why transform?</a></li>
<li class="chapter" data-level="5.7.2" data-path="mod-05.html"><a href="mod-05.html#transforms-vs.-link-functions"><i class="fa fa-check"></i><b>5.7.2</b> Transforms vs. link functions</a></li>
<li class="chapter" data-level="5.7.3" data-path="mod-05.html"><a href="mod-05.html#common-transformations"><i class="fa fa-check"></i><b>5.7.3</b> Common transformations</a></li>
<li class="chapter" data-level="5.7.4" data-path="mod-05.html"><a href="mod-05.html#uncommon-transformations"><i class="fa fa-check"></i><b>5.7.4</b> Uncommon transformations</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="mod-06.html"><a href="mod-06.html"><i class="fa fa-check"></i><b>6</b> Exploratory data analysis 3: Multivariate data exploration</a>
<ul>
<li class="chapter" data-level="6.1" data-path="mod-06.html"><a href="mod-06.html#scatterplots-for-2-continuous-variables"><i class="fa fa-check"></i><b>6.1</b> Scatterplots for 2 continuous variables</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="mod-06.html"><a href="mod-06.html#r-plotting-basics"><i class="fa fa-check"></i><b>6.1.1</b> R plotting basics</a></li>
<li class="chapter" data-level="6.1.2" data-path="mod-06.html"><a href="mod-06.html#graphical-parameters-and-advanced-plots-with-par"><i class="fa fa-check"></i><b>6.1.2</b> Graphical parameters and advanced plots with <code>par()</code></a></li>
<li class="chapter" data-level="6.1.3" data-path="mod-06.html"><a href="mod-06.html#really-advanced-plotting"><i class="fa fa-check"></i><b>6.1.3</b> Really advanced plotting</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="mod-06.html"><a href="mod-06.html#mod-06-smat"><i class="fa fa-check"></i><b>6.2</b> Scatterplot matrices for many variables</a></li>
<li class="chapter" data-level="6.3" data-path="mod-06.html"><a href="mod-06.html#lattice-plots-for-hierarchical-data"><i class="fa fa-check"></i><b>6.3</b> Lattice plots for hierarchical data</a></li>
<li class="chapter" data-level="6.4" data-path="mod-06.html"><a href="mod-06.html#continuous-variables-by-group"><i class="fa fa-check"></i><b>6.4</b> Continuous variables by group</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="mod-06.html"><a href="mod-06.html#boxplots"><i class="fa fa-check"></i><b>6.4.1</b> Boxplots</a></li>
<li class="chapter" data-level="6.4.2" data-path="mod-06.html"><a href="mod-06.html#barplots"><i class="fa fa-check"></i><b>6.4.2</b> Barplots</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="mod-06.html"><a href="mod-06.html#ordination-brief-introduction"><i class="fa fa-check"></i><b>6.5</b> Ordination (brief introduction)</a>
<ul>
<li class="chapter" data-level="6.5.1" data-path="mod-06.html"><a href="mod-06.html#principal-components-analysis-pca"><i class="fa fa-check"></i><b>6.5.1</b> Principal components analysis (PCA)</a></li>
<li class="chapter" data-level="6.5.2" data-path="mod-06.html"><a href="mod-06.html#nonmetric-multidimensional-scaling-nmds"><i class="fa fa-check"></i><b>6.5.2</b> Nonmetric multidimensional scaling (NMDS)</a></li>
<li class="chapter" data-level="6.5.3" data-path="mod-06.html"><a href="mod-06.html#plotting-ordinations"><i class="fa fa-check"></i><b>6.5.3</b> Plotting ordinations</a></li>
<li class="chapter" data-level="6.5.4" data-path="mod-06.html"><a href="mod-06.html#ordination-wrap-up-for-now"><i class="fa fa-check"></i><b>6.5.4</b> Ordination wrap-up (for now)</a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="mod-06.html"><a href="mod-06.html#principles-for-good-data-graphics"><i class="fa fa-check"></i><b>6.6</b> Principles for good data graphics</a>
<ul>
<li class="chapter" data-level="6.6.1" data-path="mod-06.html"><a href="mod-06.html#maximize-the-data-to-ink-ratio"><i class="fa fa-check"></i><b>6.6.1</b> Maximize the data-to-ink ratio</a></li>
<li class="chapter" data-level="6.6.2" data-path="mod-06.html"><a href="mod-06.html#use-labels-not-symbols"><i class="fa fa-check"></i><b>6.6.2</b> Use labels, not symbols</a></li>
<li class="chapter" data-level="6.6.3" data-path="mod-06.html"><a href="mod-06.html#be-consistent"><i class="fa fa-check"></i><b>6.6.3</b> Be consistent</a></li>
<li class="chapter" data-level="6.6.4" data-path="mod-06.html"><a href="mod-06.html#consider-legibility"><i class="fa fa-check"></i><b>6.6.4</b> Consider legibility</a></li>
<li class="chapter" data-level="6.6.5" data-path="mod-06.html"><a href="mod-06.html#set-yourself-up-for-success"><i class="fa fa-check"></i><b>6.6.5</b> Set yourself up for success</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="mod-07.html"><a href="mod-07.html"><i class="fa fa-check"></i><b>7</b> Common statistical problems</a>
<ul>
<li class="chapter" data-level="7.1" data-path="mod-07.html"><a href="mod-07.html#outliers-and-erroneous-values"><i class="fa fa-check"></i><b>7.1</b> Outliers and erroneous values</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="mod-07.html"><a href="mod-07.html#detecting-outliers"><i class="fa fa-check"></i><b>7.1.1</b> Detecting outliers</a></li>
<li class="chapter" data-level="7.1.2" data-path="mod-07.html"><a href="mod-07.html#dealing-with-outliers"><i class="fa fa-check"></i><b>7.1.2</b> Dealing with outliers</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="mod-07.html"><a href="mod-07.html#autocorrelation"><i class="fa fa-check"></i><b>7.2</b> Autocorrelation</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="mod-07.html"><a href="mod-07.html#graphical-methods-to-identify-autocorrelation"><i class="fa fa-check"></i><b>7.2.1</b> Graphical methods to identify autocorrelation</a></li>
<li class="chapter" data-level="7.2.2" data-path="mod-07.html"><a href="mod-07.html#mantel-tests-for-autocorrelation"><i class="fa fa-check"></i><b>7.2.2</b> Mantel tests for autocorrelation</a></li>
<li class="chapter" data-level="7.2.3" data-path="mod-07.html"><a href="mod-07.html#dealing-with-autocorrelation"><i class="fa fa-check"></i><b>7.2.3</b> Dealing with autocorrelation</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="mod-07.html"><a href="mod-07.html#mod-07-multicol"><i class="fa fa-check"></i><b>7.3</b> Collinearity</a></li>
<li class="chapter" data-level="7.4" data-path="mod-07.html"><a href="mod-07.html#missing-data"><i class="fa fa-check"></i><b>7.4</b> Missing data</a>
<ul>
<li class="chapter" data-level="7.4.1" data-path="mod-07.html"><a href="mod-07.html#option-1-ignore-missing-data"><i class="fa fa-check"></i><b>7.4.1</b> Option 1: Ignore missing data</a></li>
<li class="chapter" data-level="7.4.2" data-path="mod-07.html"><a href="mod-07.html#option-2-interpolate-or-impute-missing-data"><i class="fa fa-check"></i><b>7.4.2</b> Option 2: Interpolate or impute missing data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="mod-08.html"><a href="mod-08.html"><i class="fa fa-check"></i><b>8</b> Planning your analysis (what test?)</a>
<ul>
<li class="chapter" data-level="8.1" data-path="mod-08.html"><a href="mod-08.html#how-to-use-this-guide"><i class="fa fa-check"></i><b>8.1</b> How to use this guide</a></li>
<li class="chapter" data-level="8.2" data-path="mod-08.html"><a href="mod-08.html#what-question-are-you-trying-to-answer"><i class="fa fa-check"></i><b>8.2</b> What question are you trying to answer?</a></li>
<li class="chapter" data-level="8.3" data-path="mod-08.html"><a href="mod-08.html#testing-for-a-difference-in-mean-or-location"><i class="fa fa-check"></i><b>8.3</b> Testing for a difference in mean or location</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="mod-08.html"><a href="mod-08.html#additional-considerations"><i class="fa fa-check"></i><b>8.3.1</b> Additional considerations</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="mod-08.html"><a href="mod-08.html#testing-for-a-continuous-relationship-between-two-or-more-variables"><i class="fa fa-check"></i><b>8.4</b> Testing for a continuous relationship between two or more variables</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="mod-08.html"><a href="mod-08.html#additional-considerations-1"><i class="fa fa-check"></i><b>8.4.1</b> Additional considerations</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="mod-08.html"><a href="mod-08.html#independence-between-categories"><i class="fa fa-check"></i><b>8.5</b> Independence between categories</a></li>
<li class="chapter" data-level="8.6" data-path="mod-08.html"><a href="mod-08.html#classifying-observations"><i class="fa fa-check"></i><b>8.6</b> Classifying observations</a></li>
<li class="chapter" data-level="8.7" data-path="mod-08.html"><a href="mod-08.html#conclusions"><i class="fa fa-check"></i><b>8.7</b> Conclusions</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="mod-09.html"><a href="mod-09.html"><i class="fa fa-check"></i><b>9</b> Generalized linear models (GLM)</a>
<ul>
<li class="chapter" data-level="9.1" data-path="mod-09.html"><a href="mod-09.html#mod-09-lm"><i class="fa fa-check"></i><b>9.1</b> Prelude with linear models</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="mod-09.html"><a href="mod-09.html#assumptions-of-linear-models"><i class="fa fa-check"></i><b>9.1.1</b> Assumptions of linear models</a></li>
<li class="chapter" data-level="9.1.2" data-path="mod-09.html"><a href="mod-09.html#linear-regression-in-r"><i class="fa fa-check"></i><b>9.1.2</b> Linear regression in R</a></li>
<li class="chapter" data-level="9.1.3" data-path="mod-09.html"><a href="mod-09.html#multiple-linear-regression"><i class="fa fa-check"></i><b>9.1.3</b> Multiple linear regression</a></li>
<li class="chapter" data-level="9.1.4" data-path="mod-09.html"><a href="mod-09.html#mod-09-anova"><i class="fa fa-check"></i><b>9.1.4</b> ANOVA and ANCOVA with <code>lm()</code></a></li>
<li class="chapter" data-level="9.1.5" data-path="mod-09.html"><a href="mod-09.html#variations-on-linear-models"><i class="fa fa-check"></i><b>9.1.5</b> Variations on linear models</a></li>
<li class="chapter" data-level="9.1.6" data-path="mod-09.html"><a href="mod-09.html#example-linear-regression-workflow"><i class="fa fa-check"></i><b>9.1.6</b> Example linear regression workflow</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="mod-09.html"><a href="mod-09.html#mod-09-basic"><i class="fa fa-check"></i><b>9.2</b> GLM basics</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="mod-09.html"><a href="mod-09.html#example-glms"><i class="fa fa-check"></i><b>9.2.1</b> Example GLMS</a></li>
<li class="chapter" data-level="9.2.2" data-path="mod-09.html"><a href="mod-09.html#glm-families"><i class="fa fa-check"></i><b>9.2.2</b> GLM families</a></li>
<li class="chapter" data-level="9.2.3" data-path="mod-09.html"><a href="mod-09.html#glm-link-functions"><i class="fa fa-check"></i><b>9.2.3</b> GLM link functions</a></li>
<li class="chapter" data-level="9.2.4" data-path="mod-09.html"><a href="mod-09.html#deviance-and-other-glm-diagnostics"><i class="fa fa-check"></i><b>9.2.4</b> Deviance and other GLM diagnostics</a></li>
<li class="chapter" data-level="9.2.5" data-path="mod-09.html"><a href="mod-09.html#to-pseudo-r2-or-not-to-pseudo-r2"><i class="fa fa-check"></i><b>9.2.5</b> To pseudo-<em>R</em><sup>2</sup> or not to pseudo-<em>R</em><sup>2</sup>?</a></li>
<li class="chapter" data-level="9.2.6" data-path="mod-09.html"><a href="mod-09.html#common-glms"><i class="fa fa-check"></i><b>9.2.6</b> Common GLMs</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="mod-09.html"><a href="mod-09.html#mod-09-loglin"><i class="fa fa-check"></i><b>9.3</b> Log-linear models</a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="mod-09.html"><a href="mod-09.html#mod-09-loglin-examp"><i class="fa fa-check"></i><b>9.3.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="9.3.2" data-path="mod-09.html"><a href="mod-09.html#example-with-real-data"><i class="fa fa-check"></i><b>9.3.2</b> Example with real data</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="mod-09.html"><a href="mod-09.html#mod-09-poisson"><i class="fa fa-check"></i><b>9.4</b> Poisson GLM for counts</a>
<ul>
<li class="chapter" data-level="9.4.1" data-path="mod-09.html"><a href="mod-09.html#example-with-simulated-data"><i class="fa fa-check"></i><b>9.4.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="9.4.2" data-path="mod-09.html"><a href="mod-09.html#example-with-real-data-1"><i class="fa fa-check"></i><b>9.4.2</b> Example with real data</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="mod-09.html"><a href="mod-09.html#mod-09-quasi"><i class="fa fa-check"></i><b>9.5</b> Quasi-Poisson and negative binomial GLM</a>
<ul>
<li class="chapter" data-level="9.5.1" data-path="mod-09.html"><a href="mod-09.html#example-with-simulated-data-1"><i class="fa fa-check"></i><b>9.5.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="9.5.2" data-path="mod-09.html"><a href="mod-09.html#example-with-real-data-2"><i class="fa fa-check"></i><b>9.5.2</b> Example with real data</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="mod-09.html"><a href="mod-09.html#mod-09-logistic"><i class="fa fa-check"></i><b>9.6</b> Logistic regression for binary outcomes</a>
<ul>
<li class="chapter" data-level="9.6.1" data-path="mod-09.html"><a href="mod-09.html#example-with-simulated-data-2"><i class="fa fa-check"></i><b>9.6.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="9.6.2" data-path="mod-09.html"><a href="mod-09.html#example-with-real-data-3"><i class="fa fa-check"></i><b>9.6.2</b> Example with real data</a></li>
<li class="chapter" data-level="9.6.3" data-path="mod-09.html"><a href="mod-09.html#mod-09-auc"><i class="fa fa-check"></i><b>9.6.3</b> Logistic GLM diagnostics: AUC and ROC</a></li>
</ul></li>
<li class="chapter" data-level="9.7" data-path="mod-09.html"><a href="mod-09.html#mod-09-binom"><i class="fa fa-check"></i><b>9.7</b> Binomial GLM for proportional data</a>
<ul>
<li class="chapter" data-level="9.7.1" data-path="mod-09.html"><a href="mod-09.html#binomial-glm"><i class="fa fa-check"></i><b>9.7.1</b> Binomial GLM</a></li>
<li class="chapter" data-level="9.7.2" data-path="mod-09.html"><a href="mod-09.html#example-with-simulated-data-3"><i class="fa fa-check"></i><b>9.7.2</b> Example with simulated data</a></li>
</ul></li>
<li class="chapter" data-level="9.8" data-path="mod-09.html"><a href="mod-09.html#mod-09-gamma"><i class="fa fa-check"></i><b>9.8</b> Gamma models for overdispersed data</a>
<ul>
<li class="chapter" data-level="9.8.1" data-path="mod-09.html"><a href="mod-09.html#example-with-simulated-data-4"><i class="fa fa-check"></i><b>9.8.1</b> Example with simulated data</a></li>
</ul></li>
<li class="chapter" data-level="9.9" data-path="mod-09.html"><a href="mod-09.html#mod-09-beyond"><i class="fa fa-check"></i><b>9.9</b> Beyond GLM: Overview of GAM and GEE</a>
<ul>
<li class="chapter" data-level="9.9.1" data-path="mod-09.html"><a href="mod-09.html#mod-09-gam"><i class="fa fa-check"></i><b>9.9.1</b> Generalized additive models (GAM)</a></li>
<li class="chapter" data-level="9.9.2" data-path="mod-09.html"><a href="mod-09.html#mod-09-gee"><i class="fa fa-check"></i><b>9.9.2</b> Generalized estimating equations (GEE)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="mod-09.html"><a href="mod-09.html#mod-10"><i class="fa fa-check"></i><b>10</b> Nonlinear models</a>
<ul>
<li class="chapter" data-level="10.1" data-path="mod-10.html"><a href="mod-10.html"><i class="fa fa-check"></i><b>10.1</b> Background</a></li>
<li class="chapter" data-level="10.2" data-path="mod-10.html"><a href="mod-10.html#mod-10-intro"><i class="fa fa-check"></i><b>10.2</b> Nonlinear least squares (NLS)</a></li>
<li class="chapter" data-level="10.3" data-path="mod-10.html"><a href="mod-10.html#mod-10-micmen"><i class="fa fa-check"></i><b>10.3</b> Michaelis-Menten curves</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="mod-10.html"><a href="mod-10.html#mod-10-micmen-sim"><i class="fa fa-check"></i><b>10.3.1</b> Example with simulated data</a></li>
<li class="chapter" data-level="10.3.2" data-path="mod-10.html"><a href="mod-10.html#example-with-real-data-4"><i class="fa fa-check"></i><b>10.3.2</b> Example with real data</a></li>
<li class="chapter" data-level="10.3.3" data-path="mod-10.html"><a href="mod-10.html#alternative-strategies-for-the-analysis"><i class="fa fa-check"></i><b>10.3.3</b> Alternative strategies for the analysis</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="mod-10.html"><a href="mod-10.html#mod-10-grow"><i class="fa fa-check"></i><b>10.4</b> Biological growth curves</a>
<ul>
<li class="chapter" data-level="10.4.1" data-path="mod-10.html"><a href="mod-10.html#gompertz-and-von-bertalanffy-curves"><i class="fa fa-check"></i><b>10.4.1</b> Gompertz and von Bertalanffy curves</a></li>
<li class="chapter" data-level="10.4.2" data-path="mod-10.html"><a href="mod-10.html#example-with-real-data-5"><i class="fa fa-check"></i><b>10.4.2</b> Example with real data</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="mod-10.html"><a href="mod-10.html#dose-response-curves"><i class="fa fa-check"></i><b>10.5</b> Dose response curves</a></li>
<li class="chapter" data-level="10.6" data-path="mod-10.html"><a href="mod-10.html#alternatives-to-nls"><i class="fa fa-check"></i><b>10.6</b> Alternatives to NLS</a>
<ul>
<li class="chapter" data-level="10.6.1" data-path="mod-10.html"><a href="mod-10.html#generalized-nonlinear-models"><i class="fa fa-check"></i><b>10.6.1</b> Generalized nonlinear models</a></li>
<li class="chapter" data-level="10.6.2" data-path="mod-10.html"><a href="mod-10.html#quantile-regression"><i class="fa fa-check"></i><b>10.6.2</b> Quantile regression</a></li>
<li class="chapter" data-level="10.6.3" data-path="mod-10.html"><a href="mod-10.html#mod-10-gam"><i class="fa fa-check"></i><b>10.6.3</b> Generalized additive models (GAM)</a></li>
<li class="chapter" data-level="10.6.4" data-path="mod-10.html"><a href="mod-10.html#classification-and-regression-trees-cart"><i class="fa fa-check"></i><b>10.6.4</b> Classification and regression trees (CART)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="mod-11.html"><a href="mod-11.html"><i class="fa fa-check"></i><b>11</b> Mixed models</a>
<ul>
<li class="chapter" data-level="11.1" data-path="mod-11.html"><a href="mod-11.html#prelude-glm"><i class="fa fa-check"></i><b>11.1</b> Prelude (GLM)</a></li>
<li class="chapter" data-level="11.2" data-path="mod-11.html"><a href="mod-11.html#mod-11-lmm"><i class="fa fa-check"></i><b>11.2</b> Linear mixed models (LMM)</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="mod-11.html"><a href="mod-11.html#formal-definition-and-example"><i class="fa fa-check"></i><b>11.2.1</b> Formal definition and example</a></li>
<li class="chapter" data-level="11.2.2" data-path="mod-11.html"><a href="mod-11.html#example-with-simulated-data-5"><i class="fa fa-check"></i><b>11.2.2</b> Example with simulated data</a></li>
<li class="chapter" data-level="11.2.3" data-path="mod-11.html"><a href="mod-11.html#p-values-in-lmm"><i class="fa fa-check"></i><b>11.2.3</b> <em>P</em>-values in LMM</a></li>
<li class="chapter" data-level="11.2.4" data-path="mod-11.html"><a href="mod-11.html#specifying-random-effects"><i class="fa fa-check"></i><b>11.2.4</b> Specifying random effects</a></li>
<li class="chapter" data-level="11.2.5" data-path="mod-11.html"><a href="mod-11.html#lmm-example-with-real-data"><i class="fa fa-check"></i><b>11.2.5</b> LMM example with real data</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="mod-11.html"><a href="mod-11.html#mod-11-glmm"><i class="fa fa-check"></i><b>11.3</b> Generalized linear mixed models (GLMM)</a>
<ul>
<li class="chapter" data-level="11.3.1" data-path="mod-11.html"><a href="mod-11.html#definition-1"><i class="fa fa-check"></i><b>11.3.1</b> Definition</a></li>
<li class="chapter" data-level="11.3.2" data-path="mod-11.html"><a href="mod-11.html#glmm-on-simulated-data"><i class="fa fa-check"></i><b>11.3.2</b> GLMM on simulated data</a></li>
<li class="chapter" data-level="11.3.3" data-path="mod-11.html"><a href="mod-11.html#glmm-on-real-data"><i class="fa fa-check"></i><b>11.3.3</b> GLMM on real data</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="mod-11.html"><a href="mod-11.html#mod-11-nlme"><i class="fa fa-check"></i><b>11.4</b> Nonlinear mixed models (NLME)</a>
<ul>
<li class="chapter" data-level="11.4.1" data-path="mod-11.html"><a href="mod-11.html#definition-and-background"><i class="fa fa-check"></i><b>11.4.1</b> Definition and background</a></li>
<li class="chapter" data-level="11.4.2" data-path="mod-11.html"><a href="mod-11.html#nlme-on-simulated-data"><i class="fa fa-check"></i><b>11.4.2</b> NLME on simulated data</a></li>
<li class="chapter" data-level="11.4.3" data-path="mod-11.html"><a href="mod-11.html#nlme-on-real-data"><i class="fa fa-check"></i><b>11.4.3</b> NLME on real data</a></li>
</ul></li>
<li class="chapter" data-level="11.5" data-path="mod-11.html"><a href="mod-11.html#mod-11-gamm"><i class="fa fa-check"></i><b>11.5</b> (Generalized) additive mixed models (AMM/GAMM)</a>
<ul>
<li class="chapter" data-level="11.5.1" data-path="mod-11.html"><a href="mod-11.html#example-gamm-with-real-data"><i class="fa fa-check"></i><b>11.5.1</b> Example GAMM with real data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="12" data-path="mod-12.html"><a href="mod-12.html"><i class="fa fa-check"></i><b>12</b> Multivariate data analysis</a>
<ul>
<li class="chapter" data-level="12.1" data-path="mod-12.html"><a href="mod-12.html#multivariate-data"><i class="fa fa-check"></i><b>12.1</b> Multivariate data</a>
<ul>
<li class="chapter" data-level="12.1.1" data-path="mod-12.html"><a href="mod-12.html#univariate-vs.-multivariate-data"><i class="fa fa-check"></i><b>12.1.1</b> Univariate vs. multivariate data</a></li>
<li class="chapter" data-level="12.1.2" data-path="mod-12.html"><a href="mod-12.html#components-of-multivariate-data"><i class="fa fa-check"></i><b>12.1.2</b> Components of multivariate data</a></li>
</ul></li>
<li class="chapter" data-level="12.2" data-path="mod-12.html"><a href="mod-12.html#mod-12-dist"><i class="fa fa-check"></i><b>12.2</b> Distance metrics: biological (dis)similarity</a>
<ul>
<li class="chapter" data-level="12.2.1" data-path="mod-12.html"><a href="mod-12.html#euclidean-distance"><i class="fa fa-check"></i><b>12.2.1</b> Euclidean distance</a></li>
<li class="chapter" data-level="12.2.2" data-path="mod-12.html"><a href="mod-12.html#bray-curtis-and-other-distance-metrics"><i class="fa fa-check"></i><b>12.2.2</b> Bray-Curtis and other distance metrics</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="mod-12.html"><a href="mod-12.html#clustering"><i class="fa fa-check"></i><b>12.3</b> Clustering</a>
<ul>
<li class="chapter" data-level="12.3.1" data-path="mod-12.html"><a href="mod-12.html#k-means-clustering"><i class="fa fa-check"></i><b>12.3.1</b> <em>K</em>-means clustering</a></li>
<li class="chapter" data-level="12.3.2" data-path="mod-12.html"><a href="mod-12.html#hierarchical-agglomerative-clustering"><i class="fa fa-check"></i><b>12.3.2</b> Hierarchical agglomerative clustering</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="mod-12.html"><a href="mod-12.html#mod-12-sims"><i class="fa fa-check"></i><b>12.4</b> Analyzing dissimilarity</a>
<ul>
<li class="chapter" data-level="12.4.1" data-path="mod-12.html"><a href="mod-12.html#mantel-tests-distance-vs.-distance"><i class="fa fa-check"></i><b>12.4.1</b> Mantel tests: distance vs. distance</a></li>
<li class="chapter" data-level="12.4.2" data-path="mod-12.html"><a href="mod-12.html#comparing-dissimilarity-between-groups"><i class="fa fa-check"></i><b>12.4.2</b> Comparing dissimilarity between groups</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="mod-12.html"><a href="mod-12.html#mod-12-ord"><i class="fa fa-check"></i><b>12.5</b> Ordination</a>
<ul>
<li class="chapter" data-level="12.5.1" data-path="mod-12.html"><a href="mod-12.html#principal-components-analysis-pca-1"><i class="fa fa-check"></i><b>12.5.1</b> Principal components analysis (PCA)</a></li>
<li class="chapter" data-level="12.5.2" data-path="mod-12.html"><a href="mod-12.html#nmds-and-other-ordination-methods"><i class="fa fa-check"></i><b>12.5.2</b> NMDS and other ordination methods</a></li>
<li class="chapter" data-level="12.5.3" data-path="mod-12.html"><a href="mod-12.html#other-ordination-techniques"><i class="fa fa-check"></i><b>12.5.3</b> Other ordination techniques</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="literature-cited.html"><a href="literature-cited.html"><i class="fa fa-check"></i>Literature Cited</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Applied Biological Data Analysis: Statistics and R for Biologists</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="mod-06" class="section level1 hasAnchor" number="6">
<h1><span class="header-section-number">Module 6</span> Exploratory data analysis 3: Multivariate data exploration<a href="mod-06.html#mod-06" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>So far in this course we have explored ways to <a href="mod-04.html#mod-04-vis">explore distributions of single variables</a>: histograms, ECDF plots, probability density plots, and so on. On this page we will explore ways to explore relationships <em>between</em> variables. This is a vital preliminary step in studying how two or more variables might be correlated with each other, or how one might cause the other. The focus on this page is on exploratory, correlative methods. Actual inference about relationships between two variables is better handled by <a href="mod-09.html#mod-09-lm">linear models (LM)</a>, <a href="mod-09.html#mod-09-basic">generalized linear models (GLM)</a>, or <a href="mod-09.html#mod-10">other kinds of models</a>. This page also contain a brief introduction to ordination, which can help identify patterns in datasets with many variables. A more complete presentation of ordination can be found in <a href="mod-12.html#mod-12">Module 12</a>.</p>
<div id="scatterplots-for-2-continuous-variables" class="section level2 hasAnchor" number="6.1">
<h2><span class="header-section-number">6.1</span> Scatterplots for 2 continuous variables<a href="mod-06.html#scatterplots-for-2-continuous-variables" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The <strong>scatterplot</strong> is one of the most important tools in the biologist’s data toolbox. It is probably the simplest type of figure in routine use: simply plot the values in one variable on one axis, and the values of another variable on another (perpendicular) axis. The role of the scatterplot in exploratory data analysis is to help visualize the form of the relationship between two variables <span class="citation">(<a href="#ref-bolker2008ecological" role="doc-biblioref">Bolker 2008</a>)</span>. The figure below shows just 6 of the possible curve types that you might discover.</p>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-429-1.png" width="768" style="display: block; margin: auto;" /></p>
<div id="r-plotting-basics" class="section level3 hasAnchor" number="6.1.1">
<h3><span class="header-section-number">6.1.1</span> R plotting basics<a href="mod-06.html#r-plotting-basics" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We have already made some scatterplots, but it’s worth taking a closer look at the R methods for scatterplots. The basic function for scatterplots is <code>plot()</code>. The first argument is taken to be the x coordinates, and the second argument to be the y coordinates. You can also use the formula interface (<code>y ~ x</code>), but the coordinates method (x, y) usually makes for cleaner code.</p>
<p>Calling <code>plot()</code> does two things: first, it <em>creates a new plot</em>; second, it <em>plots the data</em> on the new plot area. Other components can be added to the plot with subsequent commands: <code>points()</code> for points, <code>text()</code> for text, <code>lines()</code> for lines, <code>polygon()</code> for polygons, <code>segments()</code> for line segments, and so on. These functions will <strong><em>add</em></strong> to an existing plot, but will not create a new plot area. So, call <code>plot()</code> first, then <code>points()</code> or one of the others if needed. Calling <code>points()</code> or one of the “adding” functions if there is not already an active plot area will return an error.</p>
<p>Here is a basic scatterplot:</p>
<div class="sourceCode" id="cb753"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb753-1"><a href="mod-06.html#cb753-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(iris<span class="sc">$</span>Petal.Width,iris<span class="sc">$</span>Petal.Length)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/chunkx470-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>The default scatterplot is rather unattractive, but works just fine for exploring data. Fortunately, almost every aspect of a plot can be customized using different plot arguments and graphical parameters. Here are some commonly used plot options:</p>
<div class="sourceCode" id="cb754"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb754-1"><a href="mod-06.html#cb754-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(iris<span class="sc">$</span>Petal.Width,iris<span class="sc">$</span>Petal.Length,</span>
<span id="cb754-2"><a href="mod-06.html#cb754-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">xlab=</span><span class="st">&quot;Petal width (cm)&quot;</span>,             <span class="co"># x label</span></span>
<span id="cb754-3"><a href="mod-06.html#cb754-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">ylab=</span><span class="st">&quot;Petal Length (cm)&quot;</span>,            <span class="co"># y label</span></span>
<span id="cb754-4"><a href="mod-06.html#cb754-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">xlim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">3</span>),                         <span class="co"># x axis limits</span></span>
<span id="cb754-5"><a href="mod-06.html#cb754-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">8</span>),                         <span class="co"># y axis limits</span></span>
<span id="cb754-6"><a href="mod-06.html#cb754-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">main=</span><span class="st">&quot;Petal length vs. petal width&quot;</span>  <span class="co"># plot title</span></span>
<span id="cb754-7"><a href="mod-06.html#cb754-7" aria-hidden="true" tabindex="-1"></a>)<span class="co">#plot</span></span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-430-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>Sometimes it helps to change the color or style of points to show categories in the data. This can be done using some arguments to <code>plot()</code>:</p>
<ul>
<li><code>pch</code> changes the symbol used for points. Think “<strong>p</strong>oint <strong>ch</strong>aracter”. See <code>?points</code> for a list of available symbols.</li>
<li><code>cex</code> changes the size of symbols. Think “<strong>c</strong>haracter <strong>ex</strong>pansion”. The default <code>cex</code> is 1; other values scale the points relative to this. E.g., <code>cex=2</code> makes points twice as large.</li>
<li><code>col</code> changes the color. Think “<strong>col</strong>or”.R can produce many colors; run the command <code>colors()</code> to see a named list<a href="#fn56" class="footnote-ref" id="fnref56"><sup>56</sup></a>.</li>
</ul>
<p>Each these arguments can take vector of values, so you can assign colors or shape to each point. The values will be used in the same order as the observations used to draw the points (e.g., rows of a data frame). I find it convenient to use rules like <code>ifelse()</code> commands to define symbology for plots, and store the symbology information in the same data frame as the data.</p>
<p>The examples below illustrate several ways of assigning symbology to observations within a data frame. In the first, <code>ifelse()</code> is used to make vectors of colors (<code>use.col</code>) defined by the variable <code>species</code>. Notice that a legend is provided to tell the reader which symbols mean what.</p>
<div class="sourceCode" id="cb755"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb755-1"><a href="mod-06.html#cb755-1" aria-hidden="true" tabindex="-1"></a><span class="co"># method 1: ifelse()</span></span>
<span id="cb755-2"><a href="mod-06.html#cb755-2" aria-hidden="true" tabindex="-1"></a>use.col <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(iris<span class="sc">$</span>Species <span class="sc">==</span> <span class="st">&quot;setosa&quot;</span>, <span class="st">&quot;black&quot;</span>, </span>
<span id="cb755-3"><a href="mod-06.html#cb755-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ifelse</span>(iris<span class="sc">$</span>Species <span class="sc">==</span> <span class="st">&quot;versicolor&quot;</span>, <span class="st">&quot;red&quot;</span>, <span class="st">&quot;blue&quot;</span>))</span>
<span id="cb755-4"><a href="mod-06.html#cb755-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb755-5"><a href="mod-06.html#cb755-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(iris<span class="sc">$</span>Petal.Width,iris<span class="sc">$</span>Petal.Length,</span>
<span id="cb755-6"><a href="mod-06.html#cb755-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">col=</span>use.col, <span class="co"># color by species</span></span>
<span id="cb755-7"><a href="mod-06.html#cb755-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">pch=</span><span class="dv">16</span>       <span class="co"># solid dots   </span></span>
<span id="cb755-8"><a href="mod-06.html#cb755-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb755-9"><a href="mod-06.html#cb755-9" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;bottomright&quot;</span>, </span>
<span id="cb755-10"><a href="mod-06.html#cb755-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend=</span><span class="fu">c</span>(<span class="st">&quot;Setosa&quot;</span>, <span class="st">&quot;Versicolor&quot;</span>, <span class="st">&quot;Virginica&quot;</span>),</span>
<span id="cb755-11"><a href="mod-06.html#cb755-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">pch=</span><span class="dv">16</span>, <span class="at">col=</span><span class="fu">c</span>(<span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>, <span class="st">&quot;blue&quot;</span>))</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-431-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>The next example uses <code>match()</code> instead of <code>ifelse()</code> to assign colors by <code>species</code>.</p>
<div class="sourceCode" id="cb756"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb756-1"><a href="mod-06.html#cb756-1" aria-hidden="true" tabindex="-1"></a><span class="co"># method 2: match()</span></span>
<span id="cb756-2"><a href="mod-06.html#cb756-2" aria-hidden="true" tabindex="-1"></a>spps <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">unique</span>(iris<span class="sc">$</span>Species))</span>
<span id="cb756-3"><a href="mod-06.html#cb756-3" aria-hidden="true" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;yellow&quot;</span>, <span class="st">&quot;purple&quot;</span>)</span>
<span id="cb756-4"><a href="mod-06.html#cb756-4" aria-hidden="true" tabindex="-1"></a>use.col <span class="ot">&lt;-</span> cols[<span class="fu">match</span>(iris<span class="sc">$</span>Species, spps)]</span>
<span id="cb756-5"><a href="mod-06.html#cb756-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb756-6"><a href="mod-06.html#cb756-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(iris<span class="sc">$</span>Petal.Width,iris<span class="sc">$</span>Petal.Length,</span>
<span id="cb756-7"><a href="mod-06.html#cb756-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">col=</span>use.col, <span class="co"># color by species</span></span>
<span id="cb756-8"><a href="mod-06.html#cb756-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">pch=</span><span class="dv">16</span>      <span class="co"># solid dots    </span></span>
<span id="cb756-9"><a href="mod-06.html#cb756-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb756-10"><a href="mod-06.html#cb756-10" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;bottomright&quot;</span>, </span>
<span id="cb756-11"><a href="mod-06.html#cb756-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend=</span><span class="fu">c</span>(<span class="st">&quot;Setosa&quot;</span>, <span class="st">&quot;Versicolor&quot;</span>, <span class="st">&quot;Virginica&quot;</span>),</span>
<span id="cb756-12"><a href="mod-06.html#cb756-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">pch=</span><span class="dv">16</span>, <span class="at">col=</span>cols)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-432-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>Below is an example of using shapes (<code>use.pch</code>) to define groups. See the help page for <code>points()</code> (<code>?points</code>) to see the available plot symbols.</p>
<div class="sourceCode" id="cb757"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb757-1"><a href="mod-06.html#cb757-1" aria-hidden="true" tabindex="-1"></a><span class="co"># method 3: match(), but with shape</span></span>
<span id="cb757-2"><a href="mod-06.html#cb757-2" aria-hidden="true" tabindex="-1"></a>spps <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">unique</span>(iris<span class="sc">$</span>Species))</span>
<span id="cb757-3"><a href="mod-06.html#cb757-3" aria-hidden="true" tabindex="-1"></a>shps <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">15</span>, <span class="dv">16</span>, <span class="dv">17</span>)</span>
<span id="cb757-4"><a href="mod-06.html#cb757-4" aria-hidden="true" tabindex="-1"></a>use.shp <span class="ot">&lt;-</span> shps[<span class="fu">match</span>(iris<span class="sc">$</span>Species, spps)]</span>
<span id="cb757-5"><a href="mod-06.html#cb757-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb757-6"><a href="mod-06.html#cb757-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(iris<span class="sc">$</span>Petal.Width,iris<span class="sc">$</span>Petal.Length,</span>
<span id="cb757-7"><a href="mod-06.html#cb757-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">pch=</span>use.shp <span class="co"># shape by species</span></span>
<span id="cb757-8"><a href="mod-06.html#cb757-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb757-9"><a href="mod-06.html#cb757-9" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;bottomright&quot;</span>, </span>
<span id="cb757-10"><a href="mod-06.html#cb757-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend=</span><span class="fu">c</span>(<span class="st">&quot;Setosa&quot;</span>, <span class="st">&quot;Versicolor&quot;</span>, <span class="st">&quot;Virginica&quot;</span>),</span>
<span id="cb757-11"><a href="mod-06.html#cb757-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">pch=</span>shps)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-433-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>You can think of the plot generated by <code>plot()</code> like a canvas. Once the canvas is in place, things can be “painted” on, but like paint cannot be removed. If you want to remove something, you’ll need to remake the plot without it. You can also just comment out the command you don’t want.</p>
<div class="sourceCode" id="cb758"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb758-1"><a href="mod-06.html#cb758-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(iris<span class="sc">$</span>Petal.Width,iris<span class="sc">$</span>Petal.Length,</span>
<span id="cb758-2"><a href="mod-06.html#cb758-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">col=</span>use.col, <span class="at">pch=</span><span class="dv">16</span></span>
<span id="cb758-3"><a href="mod-06.html#cb758-3" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb758-4"><a href="mod-06.html#cb758-4" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="fl">1.5</span>, <span class="dv">5</span>, <span class="st">&quot;Flowers!&quot;</span>, <span class="at">cex=</span><span class="dv">4</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>)</span>
<span id="cb758-5"><a href="mod-06.html#cb758-5" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;bottomright&quot;</span>, <span class="at">legend=</span>spps, <span class="at">col=</span>cols, <span class="at">pch=</span><span class="dv">16</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-434-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>There’s no way to remove the offending text: R doesn’t have an “Undo” button! Our only option is just to remake the plot without the text.</p>
<div class="sourceCode" id="cb759"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb759-1"><a href="mod-06.html#cb759-1" aria-hidden="true" tabindex="-1"></a><span class="co"># remake without the offending text:</span></span>
<span id="cb759-2"><a href="mod-06.html#cb759-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(iris<span class="sc">$</span>Petal.Width,iris<span class="sc">$</span>Petal.Length,</span>
<span id="cb759-3"><a href="mod-06.html#cb759-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">col=</span>use.col, <span class="at">pch=</span><span class="dv">16</span></span>
<span id="cb759-4"><a href="mod-06.html#cb759-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb759-5"><a href="mod-06.html#cb759-5" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;bottomright&quot;</span>, <span class="at">legend=</span>spps, <span class="at">col=</span>cols, <span class="at">pch=</span><span class="dv">16</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-435-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>Sometimes I’ll just comment out the plot component that needs deleting; this way it is easier to add back in later. I do this a lot when I’m trying to iteratively build a figure with lots of components.</p>
<div class="sourceCode" id="cb760"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb760-1"><a href="mod-06.html#cb760-1" aria-hidden="true" tabindex="-1"></a><span class="co"># not run:</span></span>
<span id="cb760-2"><a href="mod-06.html#cb760-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb760-3"><a href="mod-06.html#cb760-3" aria-hidden="true" tabindex="-1"></a><span class="co"># remake without the offending text (commented out):</span></span>
<span id="cb760-4"><a href="mod-06.html#cb760-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(iris<span class="sc">$</span>Petal.Width,iris<span class="sc">$</span>Petal.Length,</span>
<span id="cb760-5"><a href="mod-06.html#cb760-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">col=</span>use.col, <span class="at">pch=</span><span class="dv">16</span></span>
<span id="cb760-6"><a href="mod-06.html#cb760-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb760-7"><a href="mod-06.html#cb760-7" aria-hidden="true" tabindex="-1"></a><span class="co">#text(1.5, 5, &quot;Flowers!&quot;, cex=4, col=&quot;red&quot;)</span></span>
<span id="cb760-8"><a href="mod-06.html#cb760-8" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;bottomright&quot;</span>, <span class="at">legend=</span>spps, <span class="at">col=</span>cols, <span class="at">pch=</span><span class="dv">16</span>)</span></code></pre></div>
</div>
<div id="graphical-parameters-and-advanced-plots-with-par" class="section level3 hasAnchor" number="6.1.2">
<h3><span class="header-section-number">6.1.2</span> Graphical parameters and advanced plots with <code>par()</code><a href="mod-06.html#graphical-parameters-and-advanced-plots-with-par" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Many options that affect plots can only be set using the <code>par()</code> function. Take a look at the <code>par()</code> help page to get a sense of the variety of options available (<code>?par</code>). It is important to keep in mind is that once <code>par()</code> options are set in an R session, they will stay that way until you change them using <code>par()</code> again. So, if you are making multiple figures within the same R workspace you will need to pay attention to what you have done. You can always see all current settings by running the command <code>par()</code>.</p>
<p>Below is an illustration of using <code>par()</code> to change graphics options:</p>
<p>Without <code>par()</code>:</p>
<div class="sourceCode" id="cb761"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb761-1"><a href="mod-06.html#cb761-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(iris<span class="sc">$</span>Petal.Length<span class="sc">~</span>iris<span class="sc">$</span>Sepal.Length,</span>
<span id="cb761-2"><a href="mod-06.html#cb761-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">xlab=</span><span class="st">&quot;Sepal length (cm)&quot;</span>,</span>
<span id="cb761-3"><a href="mod-06.html#cb761-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">ylab=</span><span class="st">&quot;Petal length (cm)&quot;</span>,</span>
<span id="cb761-4"><a href="mod-06.html#cb761-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">8</span>),<span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">8</span>))</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-437-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>With <code>par()</code>, we can do the following:</p>
<ul>
<li>Narrow the margins (<code>mar</code>) so the plot takes up more of the figure space</li>
<li>Remove the box (<code>bty</code>) because it is unnecessary</li>
<li>Turn <em>y</em>-axis numbers to horizontal (<code>las</code>) so they are easy to read</li>
<li>Bring axis labels in closer (<code>mgp</code>) to save space</li>
<li>Set size of axis text (<code>cex.axis</code>) and labels (<code>cex.lab</code>) for legibility</li>
</ul>
<div class="sourceCode" id="cb762"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb762-1"><a href="mod-06.html#cb762-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar=</span><span class="fu">c</span>(<span class="fl">4.1</span>, <span class="fl">4.1</span>, <span class="fl">1.1</span>, <span class="fl">1.1</span>),  <span class="co"># margin sizes</span></span>
<span id="cb762-2"><a href="mod-06.html#cb762-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">bty=</span><span class="st">&quot;n&quot;</span>,                    <span class="co"># no box around plot</span></span>
<span id="cb762-3"><a href="mod-06.html#cb762-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">las=</span><span class="dv">1</span>,                      <span class="co"># axis labels in reading direction</span></span>
<span id="cb762-4"><a href="mod-06.html#cb762-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">mgp=</span><span class="fu">c</span>(<span class="fl">2.25</span>, <span class="dv">1</span>,<span class="dv">0</span>),           <span class="co"># position of axis components</span></span>
<span id="cb762-5"><a href="mod-06.html#cb762-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">cex.axis=</span><span class="fl">1.2</span>,               <span class="co"># size of axis numbers</span></span>
<span id="cb762-6"><a href="mod-06.html#cb762-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">cex.lab=</span><span class="fl">1.2</span>)                <span class="co"># size of axis titles</span></span>
<span id="cb762-7"><a href="mod-06.html#cb762-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(iris<span class="sc">$</span>Petal.Length<span class="sc">~</span>iris<span class="sc">$</span>Sepal.Length,</span>
<span id="cb762-8"><a href="mod-06.html#cb762-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">xlab=</span><span class="st">&quot;Sepal length (cm)&quot;</span>,</span>
<span id="cb762-9"><a href="mod-06.html#cb762-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">ylab=</span><span class="st">&quot;Petal length (cm)&quot;</span>,</span>
<span id="cb762-10"><a href="mod-06.html#cb762-10" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">8</span>),<span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">8</span>))</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-438-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>We will use <code>par()</code> often in this course to clean up figures and make them more attractive. Most of the time when you are exploring data you won’t need to mess with <code>par()</code> (except for making multi-panel figures). But, you should definitely use <code>par()</code> when preparing figures for presentations or publications. Proper use of <code>par()</code> is the key to making clean, informative, and professional-looking figures.</p>
<div id="multi-panel-figures-with-par" class="section level4 hasAnchor" number="6.1.2.1">
<h4><span class="header-section-number">6.1.2.1</span> Multi-panel figures with <code>par()</code><a href="mod-06.html#multi-panel-figures-with-par" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>One of the most common ways to use <code>par()</code> is to make multi-panel figures. The panels are specified in terms of the number of rows and columns in the figure. The arguments <code>mfrow</code> and <code>mfcol</code> define the panel layout.</p>
<ul>
<li><code>mfrow</code>: you supply the number of rows and number of columns, in that order.</li>
<li><code>mfcol</code>: works the same way, but in reverse: supply the number of columns and number of rows, in that order.</li>
</ul>
<p>Once you set <code>mfrow</code> or <code>mfcol</code> (but never both), a new plot area within the graphics window will be produced each time you call <code>plot()</code>. The graphics window will be divided evenly according to the number of panels you requested (e.g., <code>mfrow=c(2,3)</code> will yield 6 panels). Panels are drawn by row and then by column (with <code>mfrow</code>) or by column and then by row (with <code>mfcol</code>).</p>
<p>If you produce more plots than you have “slots” specified by <code>mfrow</code> or <code>mfcol</code>, the graphics device will be cleared and the plot panels will be filled again, in the same order as before. So, if you set <code>mfrow=c(2,2)</code>, and then make 5 plots, you will end up with a graphics window that has 1 plot (the fifth) in the upper left corner.</p>
<p>The example below shows the use of <code>par()$mfrow</code> to make a 2 <span class="math inline">\(\times\)</span> 2 figure.</p>
<div class="sourceCode" id="cb763"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb763-1"><a href="mod-06.html#cb763-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>),           <span class="co"># layout</span></span>
<span id="cb763-2"><a href="mod-06.html#cb763-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">mar=</span><span class="fu">c</span>(<span class="fl">4.1</span>,<span class="fl">4.1</span>,<span class="fl">1.1</span>,<span class="fl">1.1</span>), <span class="co"># margin sizes</span></span>
<span id="cb763-3"><a href="mod-06.html#cb763-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">bty=</span><span class="st">&quot;n&quot;</span>,                <span class="co"># no box around plot</span></span>
<span id="cb763-4"><a href="mod-06.html#cb763-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">las=</span><span class="dv">1</span>,                  <span class="co"># rotate axis text</span></span>
<span id="cb763-5"><a href="mod-06.html#cb763-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">cex.axis=</span><span class="fl">1.2</span>,           <span class="co"># axis text size</span></span>
<span id="cb763-6"><a href="mod-06.html#cb763-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">cex.lab=</span><span class="fl">1.2</span>)            <span class="co"># label text size</span></span>
<span id="cb763-7"><a href="mod-06.html#cb763-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(iris<span class="sc">$</span>Petal.Length, <span class="at">main=</span><span class="st">&quot;Plot 1&quot;</span>)</span>
<span id="cb763-8"><a href="mod-06.html#cb763-8" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(iris<span class="sc">$</span>Petal.Width, <span class="at">main=</span><span class="st">&quot;Plot 2&quot;</span>)</span>
<span id="cb763-9"><a href="mod-06.html#cb763-9" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(iris<span class="sc">$</span>Sepal.Length, <span class="at">main=</span><span class="st">&quot;Plot 3&quot;</span>)</span>
<span id="cb763-10"><a href="mod-06.html#cb763-10" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(iris<span class="sc">$</span>Sepal.Width, <span class="at">main=</span><span class="st">&quot;Plot 4&quot;</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/chunkx480-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>When making multi-panel plots, it helps to line up axes that correspond to each other. The commands below show two separate figures with aligned axes. This can be helpful when showing different responses to the same explanatory variable (the 2 <span class="math inline">\(\times\)</span> 1 figure):</p>
<div class="sourceCode" id="cb764"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb764-1"><a href="mod-06.html#cb764-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">1</span>),           <span class="co"># layout</span></span>
<span id="cb764-2"><a href="mod-06.html#cb764-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">mar=</span><span class="fu">c</span>(<span class="fl">4.1</span>,<span class="fl">4.1</span>,<span class="fl">1.1</span>,<span class="fl">1.1</span>), <span class="co"># margin sizes</span></span>
<span id="cb764-3"><a href="mod-06.html#cb764-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">bty=</span><span class="st">&quot;n&quot;</span>,                <span class="co"># no box around plot</span></span>
<span id="cb764-4"><a href="mod-06.html#cb764-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">las=</span><span class="dv">1</span>,                  <span class="co"># rotate axis text</span></span>
<span id="cb764-5"><a href="mod-06.html#cb764-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">cex.axis=</span><span class="fl">1.2</span>,           <span class="co"># axis text size</span></span>
<span id="cb764-6"><a href="mod-06.html#cb764-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">cex.lab=</span><span class="fl">1.2</span>)            <span class="co"># label text size</span></span>
<span id="cb764-7"><a href="mod-06.html#cb764-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(iris<span class="sc">$</span>Petal.Length, iris<span class="sc">$</span>Sepal.Length)</span>
<span id="cb764-8"><a href="mod-06.html#cb764-8" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(iris<span class="sc">$</span>Petal.Length, iris<span class="sc">$</span>Sepal.Width)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-439-1.png" width="384" style="display: block; margin: auto;" /></p>
<p>The figure below is a 1 <span class="math inline">\(\times\)</span> 2 figure that shows how a single response variable relates to two different predictors.</p>
<div class="sourceCode" id="cb765"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb765-1"><a href="mod-06.html#cb765-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>),           <span class="co"># layout</span></span>
<span id="cb765-2"><a href="mod-06.html#cb765-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">mar=</span><span class="fu">c</span>(<span class="fl">4.1</span>,<span class="fl">4.1</span>,<span class="fl">1.1</span>,<span class="fl">1.1</span>), <span class="co"># margin sizes</span></span>
<span id="cb765-3"><a href="mod-06.html#cb765-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">bty=</span><span class="st">&quot;n&quot;</span>,                <span class="co"># no box around plot</span></span>
<span id="cb765-4"><a href="mod-06.html#cb765-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">las=</span><span class="dv">1</span>,                  <span class="co"># rotate axis text</span></span>
<span id="cb765-5"><a href="mod-06.html#cb765-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">cex.axis=</span><span class="dv">1</span>,           <span class="co"># axis text size</span></span>
<span id="cb765-6"><a href="mod-06.html#cb765-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">cex.lab=</span><span class="dv">1</span>)            <span class="co"># label text size</span></span>
<span id="cb765-7"><a href="mod-06.html#cb765-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(iris<span class="sc">$</span>Petal.Width, iris<span class="sc">$</span>Petal.Length)</span>
<span id="cb765-8"><a href="mod-06.html#cb765-8" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(iris<span class="sc">$</span>Petal.Length<span class="sc">~</span>iris<span class="sc">$</span>Species)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-440-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Sometimes you may need to manually set axis limits to make sure that the panels line up exactly.</p>
<div class="sourceCode" id="cb766"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb766-1"><a href="mod-06.html#cb766-1" aria-hidden="true" tabindex="-1"></a><span class="co"># example with X and Y axes not aligned:</span></span>
<span id="cb766-2"><a href="mod-06.html#cb766-2" aria-hidden="true" tabindex="-1"></a>d1 <span class="ot">&lt;-</span> iris[<span class="fu">which</span>(iris<span class="sc">$</span>Species <span class="sc">==</span> <span class="st">&quot;setosa&quot;</span>),]</span>
<span id="cb766-3"><a href="mod-06.html#cb766-3" aria-hidden="true" tabindex="-1"></a>d2 <span class="ot">&lt;-</span> iris[<span class="fu">which</span>(iris<span class="sc">$</span>Species <span class="sc">==</span> <span class="st">&quot;versicolor&quot;</span>),]</span>
<span id="cb766-4"><a href="mod-06.html#cb766-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb766-5"><a href="mod-06.html#cb766-5" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">1</span>))</span>
<span id="cb766-6"><a href="mod-06.html#cb766-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(d1<span class="sc">$</span>Petal.Length, d1<span class="sc">$</span>Sepal.Length)</span>
<span id="cb766-7"><a href="mod-06.html#cb766-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(d2<span class="sc">$</span>Petal.Length, d2<span class="sc">$</span>Sepal.Length)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-441-1.png" width="384" style="display: block; margin: auto;" /></p>
<p>Here’s the same plot, but with the axes lined up by setting the axis limits with <code>xlim</code> and <code>ylim</code>.</p>
<div class="sourceCode" id="cb767"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb767-1"><a href="mod-06.html#cb767-1" aria-hidden="true" tabindex="-1"></a><span class="co"># same plot but with axes lined up:</span></span>
<span id="cb767-2"><a href="mod-06.html#cb767-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">1</span>))</span>
<span id="cb767-3"><a href="mod-06.html#cb767-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(d1<span class="sc">$</span>Petal.Length, d1<span class="sc">$</span>Sepal.Length, <span class="at">xlim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">6</span>), <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">8</span>))</span>
<span id="cb767-4"><a href="mod-06.html#cb767-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(d2<span class="sc">$</span>Petal.Length, d2<span class="sc">$</span>Sepal.Length, <span class="at">xlim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">6</span>), <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">8</span>))</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-442-1.png" width="384" style="display: block; margin: auto;" /></p>
<p>The argument <code>mfcol</code> to <code>par()</code> works similarly to <code>mfrow</code>, but on columns instead of rows. For <code>mfcol</code> you supply the number of columns, then the number of rows. Likewise, panels are filled by column first instead of row first. The panel layout <code>mfrow=c(2,3)</code> is the same as <code>mfcol=c(3,2)</code>, but the panels will be filled in a different order:</p>
<p><img src="06_01.jpg" width="100%" style="display: block; margin: auto;" /></p>
<p>Whether to use <code>mfrow</code> or <code>mfcol</code> is usually a matter of preference. Clever use of one or the other can allow you to automatically make multi-panel plots in a preferred layout from data stored in a list or indexed by a vector.</p>
<p>One last thing about multi-panel plots: the <code>plot()</code> function both creates new panels and creates plots within each panel. Part of this process is defining the coordinate system for a panel. When you make a plot, subsequent commands that add elements to plots such as <code>points()</code>, <code>abline()</code>, <code>legend()</code>, etc., will use the coordinate system of the most recent panel. Think about this when designing complicated figures. The example below demonstrates how a legend is placed according to the coordinate system of the most recently-created plot.</p>
<div class="sourceCode" id="cb768"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb768-1"><a href="mod-06.html#cb768-1" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fl">1e3</span></span>
<span id="cb768-2"><a href="mod-06.html#cb768-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb768-3"><a href="mod-06.html#cb768-3" aria-hidden="true" tabindex="-1"></a><span class="co"># legend in first panel:</span></span>
<span id="cb768-4"><a href="mod-06.html#cb768-4" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>))</span>
<span id="cb768-5"><a href="mod-06.html#cb768-5" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">rnorm</span>(n))</span>
<span id="cb768-6"><a href="mod-06.html#cb768-6" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;topleft&quot;</span>, <span class="at">legend=</span><span class="st">&quot;Data&quot;</span>, <span class="at">fill=</span><span class="st">&quot;lightgrey&quot;</span>)</span>
<span id="cb768-7"><a href="mod-06.html#cb768-7" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">rnorm</span>(n))</span>
<span id="cb768-8"><a href="mod-06.html#cb768-8" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">rnorm</span>(n))</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-444-1.png" width="864" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb769"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb769-1"><a href="mod-06.html#cb769-1" aria-hidden="true" tabindex="-1"></a><span class="co"># legend in second panel (note where legend() is):</span></span>
<span id="cb769-2"><a href="mod-06.html#cb769-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>))</span>
<span id="cb769-3"><a href="mod-06.html#cb769-3" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">rnorm</span>(n))</span>
<span id="cb769-4"><a href="mod-06.html#cb769-4" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">rnorm</span>(n))</span>
<span id="cb769-5"><a href="mod-06.html#cb769-5" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;topright&quot;</span>, <span class="at">legend=</span><span class="st">&quot;Data&quot;</span>, <span class="at">fill=</span><span class="st">&quot;lightgrey&quot;</span>)</span>
<span id="cb769-6"><a href="mod-06.html#cb769-6" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">rnorm</span>(n))</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-445-1.png" width="864" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="really-advanced-plotting" class="section level3 hasAnchor" number="6.1.3">
<h3><span class="header-section-number">6.1.3</span> Really advanced plotting<a href="mod-06.html#really-advanced-plotting" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="plotmath-and-expressions" class="section level4 hasAnchor" number="6.1.3.1">
<h4><span class="header-section-number">6.1.3.1</span> <code>plotmath()</code> and expressions<a href="mod-06.html#plotmath-and-expressions" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Very often we need to present special symbols on our figures, including mathematical symbols and Greek letters. The R function that does this is <code>plotmath()</code>. This function allows very precise specification of special symbols with a relatively straighforward syntax. See the help page (<code>?plotmath</code>) to see some of the symbols that are available. The downside of using <code>plotmath()</code> is that it usually works within <code>expression()</code>, which can be a horrendous pain in the neck.</p>
<p>The examples below demonstrate some common <code>plotmath()</code> uses.</p>
<div id="symbols-in-axis-labels" class="section level5 hasAnchor" number="6.1.3.1.1">
<h5><span class="header-section-number">6.1.3.1.1</span> Symbols in axis labels<a href="mod-06.html#symbols-in-axis-labels" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>The degree symbol is very common: it is used for temperature and geographic coordinates. As such, biologists should be able to use it in their figures. The example below shows how. The axis labels are provided as “expressions”, which R will parse into text. Inside <code>expression()</code>, the <code>~</code> symbol produces a space, and the <code>*</code> symbol will put two text things together with no space between them.</p>
<div class="sourceCode" id="cb770"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb770-1"><a href="mod-06.html#cb770-1" aria-hidden="true" tabindex="-1"></a><span class="co"># made up data</span></span>
<span id="cb770-2"><a href="mod-06.html#cb770-2" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">runif</span>(<span class="dv">100</span>, <span class="dv">0</span>, <span class="dv">89</span>)</span>
<span id="cb770-3"><a href="mod-06.html#cb770-3" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="sc">-</span><span class="fl">0.4</span><span class="sc">*</span>x<span class="sc">+</span><span class="dv">39</span> <span class="sc">+</span> <span class="fu">rnorm</span>(<span class="dv">100</span>)</span>
<span id="cb770-4"><a href="mod-06.html#cb770-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x, y,</span>
<span id="cb770-5"><a href="mod-06.html#cb770-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">xlab=</span><span class="fu">expression</span>(Latitude<span class="sc">~</span>(degree<span class="sc">*</span>N)),</span>
<span id="cb770-6"><a href="mod-06.html#cb770-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">ylab=</span><span class="fu">expression</span>(Daily<span class="sc">~</span>temperature<span class="sc">~</span>(degree<span class="sc">*</span>C)))</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-446-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>Biologists also sometimes have to deal with chemistry, and that means subscripts (and maybe superscripts). These can also be made with <code>plotmath()</code> and <code>expression()</code>. In the <em>y</em>-axis label below, the left superscript for the isotope is made as an exponent to an invisible or “phantom” character with <code>phantom(0)</code><a href="#fn57" class="footnote-ref" id="fnref57"><sup>57</sup></a>. The equation printed in the plot area shows that <code>plotmath()</code> expressions work equally well in any function that puts text on a plot–like <code>text()</code>–and how to add italics.</p>
<div class="sourceCode" id="cb771"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb771-1"><a href="mod-06.html#cb771-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x, y,</span>
<span id="cb771-2"><a href="mod-06.html#cb771-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">xlab=</span><span class="fu">expression</span>(CO[<span class="dv">2</span>]<span class="sc">~</span>concentration<span class="sc">~</span>(ppm)),</span>
<span id="cb771-3"><a href="mod-06.html#cb771-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">ylab=</span><span class="fu">expression</span>(<span class="fu">phantom</span>(<span class="dv">0</span>)<span class="sc">^</span><span class="dv">14</span><span class="sc">*</span>C<span class="sc">~</span>concentration<span class="sc">~</span>(ppm)))</span>
<span id="cb771-4"><a href="mod-06.html#cb771-4" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="dv">20</span>, <span class="dv">10</span>, <span class="fu">expression</span>(<span class="fu">italic</span>(E)<span class="sc">==</span><span class="fu">italic</span>(MC)<span class="sc">^</span><span class="dv">2</span>), <span class="at">cex=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-447-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="adding-pictures-to-plots" class="section level4 hasAnchor" number="6.1.3.2">
<h4><span class="header-section-number">6.1.3.2</span> Adding pictures to plots<a href="mod-06.html#adding-pictures-to-plots" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>Sometimes it’s nice to supplement a figure with an external image such as your focal organism. There are several ways to do this. The easiest way is to use an image stored locally. The example below uses an image of a flower (<em>Caladenia gracilis</em>) from <a href="https://beta.phylopic.org/">Phylopic</a>. This website hosts many 1000s of free-to-use images of organisms, organized by taxonomy and phylogeny. This particular image was created by Michael Scroggie and is in the public domain. If you ever use an image, make sure to acknowledge the source!</p>
<p>First, download the image and put it in your R home directory. Then, import the image using the appropriate package (usually <code>jpg</code> or <code>png</code>). Then, use the <code>rasterImage()</code> function to position the image on your plot. <code>rasterImage()</code> takes four coordinates: <code>xleft</code>, <code>ybottom</code>, <code>xright</code>, <code>ytop</code>, <em>in that order</em>.</p>
<div class="sourceCode" id="cb772"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb772-1"><a href="mod-06.html#cb772-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(png)</span>
<span id="cb772-2"><a href="mod-06.html#cb772-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb772-3"><a href="mod-06.html#cb772-3" aria-hidden="true" tabindex="-1"></a>im.name <span class="ot">&lt;-</span> <span class="st">&quot;iris_image.png&quot;</span></span>
<span id="cb772-4"><a href="mod-06.html#cb772-4" aria-hidden="true" tabindex="-1"></a>im <span class="ot">&lt;-</span> <span class="fu">readPNG</span>(im.name, <span class="at">native=</span><span class="cn">TRUE</span>)</span>
<span id="cb772-5"><a href="mod-06.html#cb772-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb772-6"><a href="mod-06.html#cb772-6" aria-hidden="true" tabindex="-1"></a><span class="co"># make a plot</span></span>
<span id="cb772-7"><a href="mod-06.html#cb772-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(iris<span class="sc">$</span>Petal.Length, iris<span class="sc">$</span>Sepal.Length)</span>
<span id="cb772-8"><a href="mod-06.html#cb772-8" aria-hidden="true" tabindex="-1"></a><span class="co"># add image</span></span>
<span id="cb772-9"><a href="mod-06.html#cb772-9" aria-hidden="true" tabindex="-1"></a><span class="fu">rasterImage</span>(im, <span class="dv">2</span>,<span class="fl">6.5</span>,<span class="dv">3</span>,<span class="dv">8</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/chunkx490-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>The uses of this are endless. One interesting option is to use an image as an icon (instead of say, points). The code below is crude but it works.</p>
<div class="sourceCode" id="cb773"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb773-1"><a href="mod-06.html#cb773-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(png)</span>
<span id="cb773-2"><a href="mod-06.html#cb773-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb773-3"><a href="mod-06.html#cb773-3" aria-hidden="true" tabindex="-1"></a>im.name <span class="ot">&lt;-</span> <span class="st">&quot;lizard_image.png&quot;</span></span>
<span id="cb773-4"><a href="mod-06.html#cb773-4" aria-hidden="true" tabindex="-1"></a>im <span class="ot">&lt;-</span> <span class="fu">readPNG</span>(im.name, <span class="at">native=</span><span class="cn">TRUE</span>)</span>
<span id="cb773-5"><a href="mod-06.html#cb773-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb773-6"><a href="mod-06.html#cb773-6" aria-hidden="true" tabindex="-1"></a><span class="co"># make a plot</span></span>
<span id="cb773-7"><a href="mod-06.html#cb773-7" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">20</span></span>
<span id="cb773-8"><a href="mod-06.html#cb773-8" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">runif</span>(n, <span class="dv">1</span>, <span class="dv">20</span>)</span>
<span id="cb773-9"><a href="mod-06.html#cb773-9" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">runif</span>(n, <span class="dv">1</span>, <span class="dv">10</span>)</span>
<span id="cb773-10"><a href="mod-06.html#cb773-10" aria-hidden="true" tabindex="-1"></a><span class="co"># make a blank plot:</span></span>
<span id="cb773-11"><a href="mod-06.html#cb773-11" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x, y, <span class="at">type=</span><span class="st">&quot;n&quot;</span>)</span>
<span id="cb773-12"><a href="mod-06.html#cb773-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb773-13"><a href="mod-06.html#cb773-13" aria-hidden="true" tabindex="-1"></a><span class="co"># add image to each point in a loop</span></span>
<span id="cb773-14"><a href="mod-06.html#cb773-14" aria-hidden="true" tabindex="-1"></a><span class="co"># may need to adjust dimensions and multiplier</span></span>
<span id="cb773-15"><a href="mod-06.html#cb773-15" aria-hidden="true" tabindex="-1"></a><span class="co"># for aspect ratio</span></span>
<span id="cb773-16"><a href="mod-06.html#cb773-16" aria-hidden="true" tabindex="-1"></a>dims <span class="ot">&lt;-</span> <span class="fu">dim</span>(im)</span>
<span id="cb773-17"><a href="mod-06.html#cb773-17" aria-hidden="true" tabindex="-1"></a>im.size <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="co">#(width in units of x axis)</span></span>
<span id="cb773-18"><a href="mod-06.html#cb773-18" aria-hidden="true" tabindex="-1"></a>xoff <span class="ot">&lt;-</span> (dims[<span class="dv">1</span>]<span class="sc">/</span>dims[<span class="dv">2</span>])<span class="sc">*</span>im.size<span class="sc">/</span><span class="dv">2</span></span>
<span id="cb773-19"><a href="mod-06.html#cb773-19" aria-hidden="true" tabindex="-1"></a>yoff <span class="ot">&lt;-</span> (dims[<span class="dv">1</span>]<span class="sc">/</span>dims[<span class="dv">2</span>])<span class="sc">*</span>im.size<span class="sc">/</span><span class="dv">2</span></span>
<span id="cb773-20"><a href="mod-06.html#cb773-20" aria-hidden="true" tabindex="-1"></a><span class="co"># add the images in a loop with one step</span></span>
<span id="cb773-21"><a href="mod-06.html#cb773-21" aria-hidden="true" tabindex="-1"></a><span class="co"># for each point</span></span>
<span id="cb773-22"><a href="mod-06.html#cb773-22" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>n){</span>
<span id="cb773-23"><a href="mod-06.html#cb773-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rasterImage</span>(im, x[i]<span class="sc">-</span>xoff, y[i]<span class="sc">-</span>yoff, x[i]<span class="sc">+</span>xoff, y[i]<span class="sc">+</span>yoff)</span>
<span id="cb773-24"><a href="mod-06.html#cb773-24" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-448-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>I don’t know if a journal editor will let you get away with using lizards as points, but it’s nice to know that you can.</p>
</div>
<div id="fun-with-axes" class="section level4 hasAnchor" number="6.1.3.3">
<h4><span class="header-section-number">6.1.3.3</span> Fun with axes<a href="mod-06.html#fun-with-axes" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<div id="custom-axes" class="section level5 hasAnchor" number="6.1.3.3.1">
<h5><span class="header-section-number">6.1.3.3.1</span> Custom axes<a href="mod-06.html#custom-axes" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>The function <code>axis()</code> adds an axis to a plot. The main arguments are, in order, the side on which to put the axis; the values at which to put tick marks; and optionally the labels to put on the tick marks. Like most graphics functions, <code>axis()</code> adds to an existing plot region <em>in the current coordinate system</em>. This means that if you want to put a custom axis on a plot, it’s best to suppress the axis that is created automatically by <code>plot()</code>. This is done with argument (to <code>plot()</code>) <code>xaxt="n"</code> for the <em>x</em>-axis or <code>yaxt="n"</code> for the <em>y</em>-axis. Think of these as setting “X AXis Type” or “Y AXis Type” to “none”.</p>
<p>The <code>side</code> on which to put an axis is one of 1, 2, 3, or 4, counting clockwise from the bottom. I.e., side 1 is the bottom (<em>x</em>-axis), side 2 is the left (<em>y</em>-axis), and so on.</p>
<p>The locations for the tick marks (<code>at</code>) are in the coordinate system of the current plot. This is important to keep in mind if you are dealing with data on a transformed scale such as the log scale.</p>
<div class="sourceCode" id="cb774"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb774-1"><a href="mod-06.html#cb774-1" aria-hidden="true" tabindex="-1"></a><span class="co"># step 1: make a plot and suppress the axis you</span></span>
<span id="cb774-2"><a href="mod-06.html#cb774-2" aria-hidden="true" tabindex="-1"></a><span class="co">#         want to customize</span></span>
<span id="cb774-3"><a href="mod-06.html#cb774-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(iris<span class="sc">$</span>Petal.Length, iris<span class="sc">$</span>Sepal.Length,</span>
<span id="cb774-4"><a href="mod-06.html#cb774-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">xaxt=</span><span class="st">&quot;n&quot;</span>)</span>
<span id="cb774-5"><a href="mod-06.html#cb774-5" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb774-6"><a href="mod-06.html#cb774-6" aria-hidden="true" tabindex="-1"></a><span class="co"># step 2: make the custom axis</span></span>
<span id="cb774-7"><a href="mod-06.html#cb774-7" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="at">side=</span><span class="dv">1</span>, <span class="at">at=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>)    </span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-449-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>If you don’t specify axis labels, then R will use the values specified for the tick marks. These are usually okay, unless you want a specific number format or number of decimal places. In such cases you can add labels as text strings or numbers. The number of labels must match the number of tick marks (or, their lengths must be compatible via the recycling rule).</p>
<div class="sourceCode" id="cb775"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb775-1"><a href="mod-06.html#cb775-1" aria-hidden="true" tabindex="-1"></a><span class="co"># step 1: make a plot and suppress the axis you</span></span>
<span id="cb775-2"><a href="mod-06.html#cb775-2" aria-hidden="true" tabindex="-1"></a><span class="co">#         want to customize</span></span>
<span id="cb775-3"><a href="mod-06.html#cb775-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(iris<span class="sc">$</span>Petal.Length, iris<span class="sc">$</span>Sepal.Length,</span>
<span id="cb775-4"><a href="mod-06.html#cb775-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">xaxt=</span><span class="st">&quot;n&quot;</span>)</span>
<span id="cb775-5"><a href="mod-06.html#cb775-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb775-6"><a href="mod-06.html#cb775-6" aria-hidden="true" tabindex="-1"></a><span class="co"># step 2: make the custom axis with custom labels</span></span>
<span id="cb775-7"><a href="mod-06.html#cb775-7" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="at">side=</span><span class="dv">1</span>, <span class="at">at=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>, <span class="at">labels=</span>LETTERS[<span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>])</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-450-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<div id="logarithmic-axes" class="section level5 hasAnchor" number="6.1.3.3.2">
<h5><span class="header-section-number">6.1.3.3.2</span> Logarithmic axes<a href="mod-06.html#logarithmic-axes" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>In biology we often plot values on a logarithmic scale. There are a couple of ways to do this in R.</p>
<p><strong>Method 1</strong></p>
<p>Use argument <code>log=</code> to <code>plot()</code>. You can specify <code>"x"</code>, <code>"y"</code>, or <code>"xy"</code>.</p>
<div class="sourceCode" id="cb776"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb776-1"><a href="mod-06.html#cb776-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(iris<span class="sc">$</span>Petal.Length, iris<span class="sc">$</span>Sepal.Length, <span class="at">log=</span><span class="st">&quot;xy&quot;</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-451-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>Notice how the tick marks are unevenly spaced? This is because they are on the log scale.</p>
<p><strong>Method 2</strong></p>
<p>The second method is to plot log-transformed values, and add axes manually. This method is a little more work than method 1, but I find it works better for making publication-quality figures. This method shows the degree of granularity and precision possible with R base graphics.</p>
<div class="sourceCode" id="cb777"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb777-1"><a href="mod-06.html#cb777-1" aria-hidden="true" tabindex="-1"></a><span class="co"># toy data on log10 scale</span></span>
<span id="cb777-2"><a href="mod-06.html#cb777-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb777-3"><a href="mod-06.html#cb777-3" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">20</span>, <span class="dv">0</span>, <span class="dv">1</span>)</span>
<span id="cb777-4"><a href="mod-06.html#cb777-4" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="dv">1</span><span class="fl">+1.02</span><span class="sc">*</span>x<span class="sc">+</span><span class="fu">rnorm</span>(<span class="dv">20</span>)</span>
<span id="cb777-5"><a href="mod-06.html#cb777-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb777-6"><a href="mod-06.html#cb777-6" aria-hidden="true" tabindex="-1"></a><span class="co"># make </span></span>
<span id="cb777-7"><a href="mod-06.html#cb777-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x,y, <span class="at">xaxt=</span><span class="st">&quot;n&quot;</span>, <span class="at">yaxt=</span><span class="st">&quot;n&quot;</span>, <span class="at">xlim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb777-8"><a href="mod-06.html#cb777-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb777-9"><a href="mod-06.html#cb777-9" aria-hidden="true" tabindex="-1"></a><span class="co"># make sequence of x axis values</span></span>
<span id="cb777-10"><a href="mod-06.html#cb777-10" aria-hidden="true" tabindex="-1"></a>xtick <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="sc">-</span><span class="dv">2</span>, <span class="dv">2</span>, <span class="at">by=</span><span class="dv">1</span>)</span>
<span id="cb777-11"><a href="mod-06.html#cb777-11" aria-hidden="true" tabindex="-1"></a>ytick <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="dv">3</span>, <span class="at">by=</span><span class="dv">1</span>)</span>
<span id="cb777-12"><a href="mod-06.html#cb777-12" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="at">side=</span><span class="dv">1</span>, <span class="at">at=</span>xtick, <span class="at">labels=</span><span class="dv">10</span><span class="sc">^</span>xtick)</span>
<span id="cb777-13"><a href="mod-06.html#cb777-13" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="at">side=</span><span class="dv">2</span>, <span class="at">at=</span>ytick, <span class="at">labels=</span><span class="dv">10</span><span class="sc">^</span>ytick)</span>
<span id="cb777-14"><a href="mod-06.html#cb777-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb777-15"><a href="mod-06.html#cb777-15" aria-hidden="true" tabindex="-1"></a><span class="co"># add little ticks for values between powers of 10</span></span>
<span id="cb777-16"><a href="mod-06.html#cb777-16" aria-hidden="true" tabindex="-1"></a><span class="co"># argument tcl controls length of tick mark (-0.5 is default)</span></span>
<span id="cb777-17"><a href="mod-06.html#cb777-17" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="at">side=</span><span class="dv">1</span>, <span class="at">at=</span><span class="fu">log10</span>(<span class="fu">c</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">9</span><span class="sc">/</span><span class="dv">100</span>, <span class="dv">2</span><span class="sc">:</span><span class="dv">9</span><span class="sc">/</span><span class="dv">10</span>, <span class="dv">2</span><span class="sc">:</span><span class="dv">9</span>, <span class="dv">2</span><span class="sc">:</span><span class="dv">9</span><span class="sc">*</span><span class="dv">10</span>)),</span>
<span id="cb777-18"><a href="mod-06.html#cb777-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">labels=</span><span class="cn">NA</span>, <span class="at">tcl=</span><span class="sc">-</span><span class="fl">0.3</span>)</span>
<span id="cb777-19"><a href="mod-06.html#cb777-19" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="at">side=</span><span class="dv">2</span>, <span class="at">at=</span><span class="fu">log10</span>(<span class="fu">c</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">9</span><span class="sc">/</span><span class="dv">10</span>, <span class="dv">2</span><span class="sc">:</span><span class="dv">9</span>, <span class="dv">2</span><span class="sc">:</span><span class="dv">9</span><span class="sc">*</span><span class="dv">10</span>, <span class="dv">2</span><span class="sc">:</span><span class="dv">9</span><span class="sc">*</span><span class="dv">100</span>)),</span>
<span id="cb777-20"><a href="mod-06.html#cb777-20" aria-hidden="true" tabindex="-1"></a>     <span class="at">labels=</span><span class="cn">NA</span>, <span class="at">tcl=</span><span class="sc">-</span><span class="fl">0.3</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-452-1.png" width="576" style="display: block; margin: auto;" />
One thing to watch out for is the base of the logarithm used for plotting. By default, R uses the natural log (base <em>e</em>). But, we usually want to show base 10 because it’s more interpretable. This means that you need to keep track of which values are on the original scale, which values are on the natural log scale, and what the corresponding <span class="math inline">\(log_10\)</span> values are.</p>
</div>
<div id="extra-axes" class="section level5 hasAnchor" number="6.1.3.3.3">
<h5><span class="header-section-number">6.1.3.3.3</span> Extra axes<a href="mod-06.html#extra-axes" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>It is possible, but rarely advisable, to put additional axes on a plot. The reason that additional axes are generally not a good idea is that it is <em>very</em> easy to mislead your reader (and yourself!). But, if you insist, here is how to add a second <em>y</em>-axis to a plot. The secret is to create a new, invisible plot on top of the first plot using <code>par(new=TRUE)</code>. This new plot will use the already existing plot area, but will have its own coordinate system. In the example below, we expand the right margin to 5.1 (the fourth value in <code>mar</code>) to make room for the right axis label. The label itself is added with <code>mtext()</code> (short for “margin text”).</p>
<div class="sourceCode" id="cb778"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb778-1"><a href="mod-06.html#cb778-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar=</span><span class="fu">c</span>(<span class="fl">5.1</span>, <span class="fl">5.1</span>, <span class="fl">1.1</span>, <span class="fl">5.1</span>))</span>
<span id="cb778-2"><a href="mod-06.html#cb778-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(iris<span class="sc">$</span>Petal.Length, iris<span class="sc">$</span>Sepal.Length)</span>
<span id="cb778-3"><a href="mod-06.html#cb778-3" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">new=</span><span class="cn">TRUE</span>)</span>
<span id="cb778-4"><a href="mod-06.html#cb778-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="fu">runif</span>(<span class="dv">10</span>), <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">pch=</span><span class="dv">15</span>,</span>
<span id="cb778-5"><a href="mod-06.html#cb778-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">xaxt=</span><span class="st">&quot;n&quot;</span>, <span class="at">yaxt=</span><span class="st">&quot;n&quot;</span>,  <span class="co"># suppress axes</span></span>
<span id="cb778-6"><a href="mod-06.html#cb778-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">xlab=</span><span class="st">&quot;&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;&quot;</span>,    <span class="co"># suppress axis titles</span></span>
<span id="cb778-7"><a href="mod-06.html#cb778-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>))</span>
<span id="cb778-8"><a href="mod-06.html#cb778-8" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="at">side=</span><span class="dv">4</span>, <span class="at">at=</span><span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">by=</span><span class="fl">0.25</span>), <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">col.axis=</span><span class="st">&quot;red&quot;</span>)</span>
<span id="cb778-9"><a href="mod-06.html#cb778-9" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="at">side=</span><span class="dv">4</span>, <span class="st">&quot;Price of tea in China&quot;</span>, <span class="at">line=</span><span class="dv">3</span>, <span class="at">col=</span><span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-453-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<div id="finer-points-of-axis-labels" class="section level5 hasAnchor" number="6.1.3.3.4">
<h5><span class="header-section-number">6.1.3.3.4</span> Finer points of axis labels<a href="mod-06.html#finer-points-of-axis-labels" class="anchor-section" aria-label="Anchor link to header"></a></h5>
<p>Coming soon:
- Useful <code>par()</code> settings: mgp, las, and others
- <code>title()</code> vs. <code>mtext()</code></p>
</div>
</div>
</div>
</div>
<div id="mod-06-smat" class="section level2 hasAnchor" number="6.2">
<h2><span class="header-section-number">6.2</span> Scatterplot matrices for many variables<a href="mod-06.html#mod-06-smat" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>A scatterplot plots one variable against another, and is probably the best way to see the relationship (if any) between two variables. But what if you have many variables? Making dozens of scatterplots can be tedious and time consuming. A scatterplot matrix makes many scatterplots at once, allowing relationships between many variables to be visualized at once. The base function to do this is <code>pairs()</code>.</p>
<div class="sourceCode" id="cb779"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb779-1"><a href="mod-06.html#cb779-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load some data</span></span>
<span id="cb779-2"><a href="mod-06.html#cb779-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(crabs, <span class="at">package=</span><span class="st">&quot;MASS&quot;</span>)</span>
<span id="cb779-3"><a href="mod-06.html#cb779-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb779-4"><a href="mod-06.html#cb779-4" aria-hidden="true" tabindex="-1"></a><span class="co"># make a spare copy</span></span>
<span id="cb779-5"><a href="mod-06.html#cb779-5" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> crabs</span>
<span id="cb779-6"><a href="mod-06.html#cb779-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb779-7"><a href="mod-06.html#cb779-7" aria-hidden="true" tabindex="-1"></a><span class="co"># define columns for scatterplot matrix</span></span>
<span id="cb779-8"><a href="mod-06.html#cb779-8" aria-hidden="true" tabindex="-1"></a>dat.cols <span class="ot">&lt;-</span> <span class="dv">4</span><span class="sc">:</span><span class="dv">8</span></span>
<span id="cb779-9"><a href="mod-06.html#cb779-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb779-10"><a href="mod-06.html#cb779-10" aria-hidden="true" tabindex="-1"></a><span class="co"># make scatterplot matrix</span></span>
<span id="cb779-11"><a href="mod-06.html#cb779-11" aria-hidden="true" tabindex="-1"></a><span class="fu">pairs</span>(x[,dat.cols])</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-454-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>In the scatterplot matrix, every variable is plotted against every other variable. Variables are labeled on the diagonal. E.g., the plots in the first row have “FL” as their <em>y</em>-axis, and “RW”, “CL”, “CW”, and “BD” as their <em>x</em>-axes. Likewise, plots in the first column have “FL” as their <em>x</em>-axis, and “RW”, “CL”, “CW”, and “BD” as their <em>y</em>-axes.</p>
<p>One common modification to the default <code>pairs()</code> plot is to replace scatterplots above the diagonal with correlation coefficients. Another modification is to add linear regression or LOESS<a href="#fn58" class="footnote-ref" id="fnref58"><sup>58</sup></a> lines to the lower plots to help highlight the relationships. The function below is adapted from the help page for <code>pairs()</code>. Notice that the size of the text for each coefficient is scaled to the magnitude of the coefficient. Can you figure out what piece of the code is doing that?</p>
<div class="sourceCode" id="cb780"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb780-1"><a href="mod-06.html#cb780-1" aria-hidden="true" tabindex="-1"></a><span class="co"># define a function to add correlation coefficients</span></span>
<span id="cb780-2"><a href="mod-06.html#cb780-2" aria-hidden="true" tabindex="-1"></a>panel.cor <span class="ot">&lt;-</span> <span class="cf">function</span>(x, y, <span class="at">digits =</span> <span class="dv">2</span>, <span class="at">prefix =</span> <span class="st">&quot;&quot;</span>, cex.cor, ...)</span>
<span id="cb780-3"><a href="mod-06.html#cb780-3" aria-hidden="true" tabindex="-1"></a>{</span>
<span id="cb780-4"><a href="mod-06.html#cb780-4" aria-hidden="true" tabindex="-1"></a>    <span class="co">#usr &lt;- par(&quot;usr&quot;); on.exit(par(usr))</span></span>
<span id="cb780-5"><a href="mod-06.html#cb780-5" aria-hidden="true" tabindex="-1"></a>    <span class="co">#par(usr = c(0, 1, 0, 1))</span></span>
<span id="cb780-6"><a href="mod-06.html#cb780-6" aria-hidden="true" tabindex="-1"></a>    r <span class="ot">&lt;-</span> <span class="fu">cor</span>(x, y)</span>
<span id="cb780-7"><a href="mod-06.html#cb780-7" aria-hidden="true" tabindex="-1"></a>    txt <span class="ot">&lt;-</span> <span class="fu">format</span>(<span class="fu">c</span>(r, <span class="fl">0.123456789</span>), <span class="at">digits =</span> digits)[<span class="dv">1</span>]</span>
<span id="cb780-8"><a href="mod-06.html#cb780-8" aria-hidden="true" tabindex="-1"></a>    txt <span class="ot">&lt;-</span> <span class="fu">paste0</span>(prefix, txt)</span>
<span id="cb780-9"><a href="mod-06.html#cb780-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="fu">missing</span>(cex.cor)) cex.cor <span class="ot">&lt;-</span> <span class="fl">0.8</span><span class="sc">/</span><span class="fu">strwidth</span>(txt)</span>
<span id="cb780-10"><a href="mod-06.html#cb780-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">text</span>(<span class="fl">0.5</span>, <span class="fl">0.5</span>, txt, <span class="at">cex =</span> <span class="dv">2</span><span class="sc">*</span><span class="fu">abs</span>(r))</span>
<span id="cb780-11"><a href="mod-06.html#cb780-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb780-12"><a href="mod-06.html#cb780-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb780-13"><a href="mod-06.html#cb780-13" aria-hidden="true" tabindex="-1"></a><span class="co"># example of use:</span></span>
<span id="cb780-14"><a href="mod-06.html#cb780-14" aria-hidden="true" tabindex="-1"></a><span class="fu">pairs</span>(x[,<span class="fu">c</span>(<span class="dv">2</span>, dat.cols)], </span>
<span id="cb780-15"><a href="mod-06.html#cb780-15" aria-hidden="true" tabindex="-1"></a>      <span class="at">lower.panel=</span>panel.smooth,</span>
<span id="cb780-16"><a href="mod-06.html#cb780-16" aria-hidden="true" tabindex="-1"></a>      <span class="at">upper.panel=</span>panel.cor, <span class="at">gap=</span><span class="dv">0</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-455-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Sometimes the Spearman rank correlation coefficient <span class="math inline">\(\rho\)</span> is more informative than the Pearson linear correlation coefficient <em>r</em>. The function <code>panel.cor()</code> can be modified to use <span class="math inline">\(\rho\)</span> instead of <em>r</em>:</p>
<div class="sourceCode" id="cb781"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb781-1"><a href="mod-06.html#cb781-1" aria-hidden="true" tabindex="-1"></a>panel.cor2 <span class="ot">&lt;-</span> <span class="cf">function</span>(x, y, <span class="at">digits =</span> <span class="dv">2</span>, <span class="at">prefix =</span> <span class="st">&quot;&quot;</span>, cex.cor, ...)</span>
<span id="cb781-2"><a href="mod-06.html#cb781-2" aria-hidden="true" tabindex="-1"></a>{</span>
<span id="cb781-3"><a href="mod-06.html#cb781-3" aria-hidden="true" tabindex="-1"></a>    <span class="co">#usr &lt;- par(&quot;usr&quot;); on.exit(par(usr))</span></span>
<span id="cb781-4"><a href="mod-06.html#cb781-4" aria-hidden="true" tabindex="-1"></a>    <span class="co">#par(usr = c(0, 1, 0, 1))</span></span>
<span id="cb781-5"><a href="mod-06.html#cb781-5" aria-hidden="true" tabindex="-1"></a>    r <span class="ot">&lt;-</span> <span class="fu">cor.test</span>(x, y, <span class="at">method=</span><span class="st">&quot;spearman&quot;</span>)<span class="sc">$</span>estimate</span>
<span id="cb781-6"><a href="mod-06.html#cb781-6" aria-hidden="true" tabindex="-1"></a>    txt <span class="ot">&lt;-</span> <span class="fu">format</span>(<span class="fu">c</span>(r, <span class="fl">0.123456789</span>), <span class="at">digits =</span> digits)[<span class="dv">1</span>]</span>
<span id="cb781-7"><a href="mod-06.html#cb781-7" aria-hidden="true" tabindex="-1"></a>    txt <span class="ot">&lt;-</span> <span class="fu">paste0</span>(prefix, txt)</span>
<span id="cb781-8"><a href="mod-06.html#cb781-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="fu">missing</span>(cex.cor)) cex.cor <span class="ot">&lt;-</span> <span class="fl">0.8</span><span class="sc">/</span><span class="fu">strwidth</span>(txt)</span>
<span id="cb781-9"><a href="mod-06.html#cb781-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">text</span>(<span class="fl">0.5</span>, <span class="fl">0.5</span>, txt, <span class="at">cex =</span> <span class="dv">2</span><span class="sc">*</span><span class="fu">abs</span>(r))</span>
<span id="cb781-10"><a href="mod-06.html#cb781-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb781-11"><a href="mod-06.html#cb781-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb781-12"><a href="mod-06.html#cb781-12" aria-hidden="true" tabindex="-1"></a><span class="co"># example of use:</span></span>
<span id="cb781-13"><a href="mod-06.html#cb781-13" aria-hidden="true" tabindex="-1"></a><span class="co"># (might return warnings but these are ok)</span></span>
<span id="cb781-14"><a href="mod-06.html#cb781-14" aria-hidden="true" tabindex="-1"></a><span class="fu">pairs</span>(x[,dat.cols], </span>
<span id="cb781-15"><a href="mod-06.html#cb781-15" aria-hidden="true" tabindex="-1"></a>      <span class="at">lower.panel=</span>panel.smooth,</span>
<span id="cb781-16"><a href="mod-06.html#cb781-16" aria-hidden="true" tabindex="-1"></a>      <span class="at">upper.panel=</span>panel.cor2, <span class="at">gap=</span><span class="dv">0</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-456-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="lattice-plots-for-hierarchical-data" class="section level2 hasAnchor" number="6.3">
<h2><span class="header-section-number">6.3</span> Lattice plots for hierarchical data<a href="mod-06.html#lattice-plots-for-hierarchical-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In another module we will explore mixed models, also known as hierarchical models, which can account for relationships that differ between groups of data. These are similar in some ways to analysis of covariance (ANCOVA) models with interactions. An example of what ANCOVA can look like is shown below:</p>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/chunkx500-1.png" width="864" style="display: block; margin: auto;" /></p>
<p>In the left panel, the <em>Y</em> variable increases as a function of the <em>X</em> variable. There is also an effect of the group variable (i.e., a factor): <em>Y</em> values in group 2 are on average greater than values in group 1. This is called “main effects” ANCOVA. In the right panel, <em>Y</em> increases with <em>X</em>, and the slope of <em>Y</em> with respect to <em>X</em> varies between groups. In other words, <em>X</em> has a greater effect on <em>Y</em> in group 2 than in group 1. This is called an “interaction” between <em>X</em> and the grouping variable. Analyses that include such interactions are sometimes called “interaction effects” ANCOVA because the effects of one variable “interact with” the effects of another variable. Both types of ANCOVA are extremely common in biology.</p>
<p>One key difference between an ANCOVA model and a mixed model is that while in ANCOVA the slope and intercept are fitted specifically to each level of the grouping variable, consuming degrees of freedom, in a mixed model these slopes and intercepts are considered to be drawn from a random distribution. For now, all you need to understand is that the relationship between response and predictor variables can differ between groups. A <strong>lattice</strong> or <strong>trellis</strong> plot is a way to visualize such differences. In R these are named for package <code>lattice</code> which is an older package for producing hierarchical plots. The newer <code>ggplot2</code> package can also do the job.</p>
<div class="sourceCode" id="cb782"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb782-1"><a href="mod-06.html#cb782-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(lattice)</span>
<span id="cb782-2"><a href="mod-06.html#cb782-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb782-3"><a href="mod-06.html#cb782-3" aria-hidden="true" tabindex="-1"></a><span class="co"># scatterplot of 2 continuous variables</span></span>
<span id="cb782-4"><a href="mod-06.html#cb782-4" aria-hidden="true" tabindex="-1"></a><span class="co"># by species</span></span>
<span id="cb782-5"><a href="mod-06.html#cb782-5" aria-hidden="true" tabindex="-1"></a><span class="fu">xyplot</span>(Petal.Length <span class="sc">~</span> Sepal.Length <span class="sc">|</span> Species, <span class="at">data=</span>iris)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-457-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Lattice plots can group observations by more than one variable, as shown below.</p>
<div class="sourceCode" id="cb783"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb783-1"><a href="mod-06.html#cb783-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add another (pretend) variable</span></span>
<span id="cb783-2"><a href="mod-06.html#cb783-2" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> iris</span>
<span id="cb783-3"><a href="mod-06.html#cb783-3" aria-hidden="true" tabindex="-1"></a>x<span class="sc">$</span>color <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;purple&quot;</span>, <span class="st">&quot;white&quot;</span>)</span>
<span id="cb783-4"><a href="mod-06.html#cb783-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb783-5"><a href="mod-06.html#cb783-5" aria-hidden="true" tabindex="-1"></a><span class="fu">xyplot</span>(Petal.Length <span class="sc">~</span> Sepal.Length <span class="sc">|</span> Species<span class="sc">+</span>color, <span class="at">data=</span>x)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-458-1.png" width="768" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb784"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb784-1"><a href="mod-06.html#cb784-1" aria-hidden="true" tabindex="-1"></a><span class="co"># same plot, different layout:</span></span>
<span id="cb784-2"><a href="mod-06.html#cb784-2" aria-hidden="true" tabindex="-1"></a><span class="fu">xyplot</span>(Petal.Length <span class="sc">~</span> Sepal.Length <span class="sc">|</span> color<span class="sc">+</span>Species, <span class="at">data=</span>x)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-458-2.png" width="768" style="display: block; margin: auto;" /></p>
<p>Functions in <code>lattice</code> can also produce hierarchical boxplots. This is the graphical equivalent of <a href="mod-04.html#mod-04-agg">aggregating by more than one variable</a>.</p>
<div class="sourceCode" id="cb785"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb785-1"><a href="mod-06.html#cb785-1" aria-hidden="true" tabindex="-1"></a><span class="co"># box-and-whisker plots by color within species</span></span>
<span id="cb785-2"><a href="mod-06.html#cb785-2" aria-hidden="true" tabindex="-1"></a><span class="fu">bwplot</span>(Petal.Length<span class="sc">~</span>color<span class="sc">|</span>Species, <span class="at">data=</span>x)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-459-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>The <code>tidyverse</code> equivalent to the first <code>lattice</code> plot above is:</p>
<div class="sourceCode" id="cb786"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb786-1"><a href="mod-06.html#cb786-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb786-2"><a href="mod-06.html#cb786-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>x) <span class="sc">+</span> </span>
<span id="cb786-3"><a href="mod-06.html#cb786-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>Sepal.Length, <span class="at">y=</span>Petal.Length)) <span class="sc">+</span> </span>
<span id="cb786-4"><a href="mod-06.html#cb786-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>Species, <span class="at">nrow=</span><span class="dv">1</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-460-1.png" width="768" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb787"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb787-1"><a href="mod-06.html#cb787-1" aria-hidden="true" tabindex="-1"></a><span class="co"># same plot, different layout:</span></span>
<span id="cb787-2"><a href="mod-06.html#cb787-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>x) <span class="sc">+</span> </span>
<span id="cb787-3"><a href="mod-06.html#cb787-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">mapping=</span><span class="fu">aes</span>(<span class="at">x=</span>Sepal.Length, <span class="at">y=</span>Petal.Length)) <span class="sc">+</span> </span>
<span id="cb787-4"><a href="mod-06.html#cb787-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>Species, <span class="at">nrow=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-460-2.png" width="768" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb788"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb788-1"><a href="mod-06.html#cb788-1" aria-hidden="true" tabindex="-1"></a><span class="co"># same plot, another layout:</span></span>
<span id="cb788-2"><a href="mod-06.html#cb788-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data=</span>x) <span class="sc">+</span> </span>
<span id="cb788-3"><a href="mod-06.html#cb788-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">y=</span>Petal.Length, <span class="at">x=</span>color)) <span class="sc">+</span> </span>
<span id="cb788-4"><a href="mod-06.html#cb788-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>Species,<span class="at">nrow=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-460-3.png" width="768" style="display: block; margin: auto;" /></p>
<p>Whether you use <code>lattice</code> or <code>ggplot2</code> to make your hierarchical graphs is largely a matter of personal preference. Both packages work just fine for exploratory data analysis. Both share the advantage of being very fast, with the ability to represent a lot of data with very little “ink” and code. Both share the disadvantage of being tricky to customize because many of the graphical options are “hidden” inside layers of other functions.</p>
<p>If you need to present a hierarchical plot in a presentation or manuscript, and need to meet very specific formatting requirements, you may need to make your figure using <code>base</code> graphics. The commands below reproduce the figures above in <code>base</code> graphics. Note that while using <code>base</code> graphics requires more coding to get the same result, it can be <strong><em>much</em></strong> easier to customize the plots because the plotting options are not buried inside other functions.</p>
<div class="sourceCode" id="cb789"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb789-1"><a href="mod-06.html#cb789-1" aria-hidden="true" tabindex="-1"></a>flag1 <span class="ot">&lt;-</span> <span class="fu">which</span>(x<span class="sc">$</span>Species <span class="sc">==</span> <span class="st">&quot;setosa&quot;</span>)</span>
<span id="cb789-2"><a href="mod-06.html#cb789-2" aria-hidden="true" tabindex="-1"></a>flag2 <span class="ot">&lt;-</span> <span class="fu">which</span>(x<span class="sc">$</span>Species <span class="sc">==</span> <span class="st">&quot;versicolor&quot;</span>)</span>
<span id="cb789-3"><a href="mod-06.html#cb789-3" aria-hidden="true" tabindex="-1"></a>flag3 <span class="ot">&lt;-</span> <span class="fu">which</span>(x<span class="sc">$</span>Species <span class="sc">==</span> <span class="st">&quot;virginica&quot;</span>)</span>
<span id="cb789-4"><a href="mod-06.html#cb789-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb789-5"><a href="mod-06.html#cb789-5" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>), <span class="at">mar=</span><span class="fu">c</span>(<span class="fl">5.1</span>, <span class="fl">5.1</span>, <span class="fl">1.1</span>, <span class="fl">1.1</span>),</span>
<span id="cb789-6"><a href="mod-06.html#cb789-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">las=</span><span class="dv">1</span>, <span class="at">lend=</span><span class="dv">1</span>, <span class="at">bty=</span><span class="st">&quot;n&quot;</span>,</span>
<span id="cb789-7"><a href="mod-06.html#cb789-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">cex.lab=</span><span class="fl">1.3</span>, <span class="at">cex.axis=</span><span class="fl">1.3</span>)</span>
<span id="cb789-8"><a href="mod-06.html#cb789-8" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x<span class="sc">$</span>Sepal.Length[flag1], x<span class="sc">$</span>Petal.Length[flag1], </span>
<span id="cb789-9"><a href="mod-06.html#cb789-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlim=</span><span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">8</span>), <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">8</span>), </span>
<span id="cb789-10"><a href="mod-06.html#cb789-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab=</span><span class="st">&quot;Sepal length&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;Petal length&quot;</span>) </span>
<span id="cb789-11"><a href="mod-06.html#cb789-11" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span>(<span class="at">main=</span><span class="fu">expression</span>(<span class="fu">italic</span>(Iris)<span class="sc">~</span><span class="fu">italic</span>(setosa)))</span>
<span id="cb789-12"><a href="mod-06.html#cb789-12" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x<span class="sc">$</span>Sepal.Length[flag2], x<span class="sc">$</span>Petal.Length[flag2], </span>
<span id="cb789-13"><a href="mod-06.html#cb789-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlim=</span><span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">8</span>), <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">8</span>), </span>
<span id="cb789-14"><a href="mod-06.html#cb789-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab=</span><span class="st">&quot;Sepal length&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;Petal length&quot;</span>) </span>
<span id="cb789-15"><a href="mod-06.html#cb789-15" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span>(<span class="at">main=</span><span class="fu">expression</span>(<span class="fu">italic</span>(Iris)<span class="sc">~</span><span class="fu">italic</span>(versicolor)))</span>
<span id="cb789-16"><a href="mod-06.html#cb789-16" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x<span class="sc">$</span>Sepal.Length[flag3], x<span class="sc">$</span>Petal.Length[flag3], </span>
<span id="cb789-17"><a href="mod-06.html#cb789-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlim=</span><span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">8</span>), <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">8</span>), </span>
<span id="cb789-18"><a href="mod-06.html#cb789-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab=</span><span class="st">&quot;Sepal length&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;Petal length&quot;</span>) </span>
<span id="cb789-19"><a href="mod-06.html#cb789-19" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span>(<span class="at">main=</span><span class="fu">expression</span>(<span class="fu">italic</span>(Iris)<span class="sc">~</span><span class="fu">italic</span>(virginica)))</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-461-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>Here is the hierarchical boxplot:</p>
<div class="sourceCode" id="cb790"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb790-1"><a href="mod-06.html#cb790-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>), <span class="at">mar=</span><span class="fu">c</span>(<span class="fl">5.1</span>, <span class="fl">5.1</span>, <span class="fl">1.1</span>, <span class="fl">1.1</span>), </span>
<span id="cb790-2"><a href="mod-06.html#cb790-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">las=</span><span class="dv">1</span>, <span class="at">lend=</span><span class="dv">1</span>, <span class="at">bty=</span><span class="st">&quot;n&quot;</span>, </span>
<span id="cb790-3"><a href="mod-06.html#cb790-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">cex.lab=</span><span class="fl">1.3</span>, <span class="at">cex.axis=</span><span class="fl">1.3</span>)</span>
<span id="cb790-4"><a href="mod-06.html#cb790-4" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(Petal.Length<span class="sc">~</span>color, <span class="at">data=</span>x[flag1,], </span>
<span id="cb790-5"><a href="mod-06.html#cb790-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">xlab=</span><span class="st">&quot;Color&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;Petal length&quot;</span>,</span>
<span id="cb790-6"><a href="mod-06.html#cb790-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">8</span>))</span>
<span id="cb790-7"><a href="mod-06.html#cb790-7" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span>(<span class="at">main=</span><span class="fu">expression</span>(<span class="fu">italic</span>(Iris)<span class="sc">~</span><span class="fu">italic</span>(setosa)))</span>
<span id="cb790-8"><a href="mod-06.html#cb790-8" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(Petal.Length<span class="sc">~</span>color, <span class="at">data=</span>x[flag2,],</span>
<span id="cb790-9"><a href="mod-06.html#cb790-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">xlab=</span><span class="st">&quot;Color&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;Petal length&quot;</span>,</span>
<span id="cb790-10"><a href="mod-06.html#cb790-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">8</span>))</span>
<span id="cb790-11"><a href="mod-06.html#cb790-11" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span>(<span class="at">main=</span><span class="fu">expression</span>(<span class="fu">italic</span>(Iris)<span class="sc">~</span><span class="fu">italic</span>(versicolor)))</span>
<span id="cb790-12"><a href="mod-06.html#cb790-12" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(Petal.Length<span class="sc">~</span>color, <span class="at">data=</span>x[flag3,], </span>
<span id="cb790-13"><a href="mod-06.html#cb790-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">xlab=</span><span class="st">&quot;Color&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;Petal length&quot;</span>,</span>
<span id="cb790-14"><a href="mod-06.html#cb790-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">8</span>))</span>
<span id="cb790-15"><a href="mod-06.html#cb790-15" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span>(<span class="at">main=</span><span class="fu">expression</span>(<span class="fu">italic</span>(Iris)<span class="sc">~</span><span class="fu">italic</span>(virginica)))</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-462-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="continuous-variables-by-group" class="section level2 hasAnchor" number="6.4">
<h2><span class="header-section-number">6.4</span> Continuous variables by group<a href="mod-06.html#continuous-variables-by-group" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>One of the most common data exploration tasks is visualizing how some variable differs between groups. The most common tools for this are <strong>boxplots</strong> (also known as <strong>box-and-whiskers plots</strong>) and <strong>barplots</strong>. The key difference between them is that barplots show a single summary statistic, usually the group mean; whereas boxplots give more details about how values are distributed. Because of this I usually prefer boxplots.</p>
<div id="boxplots" class="section level3 hasAnchor" number="6.4.1">
<h3><span class="header-section-number">6.4.1</span> Boxplots<a href="mod-06.html#boxplots" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>A boxplot of one continuous variable across several groups can be obtained with the <code>boxplot()</code> function. The plot is specified as <code>values~group</code>, much like a statistical model.</p>
<div class="sourceCode" id="cb791"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb791-1"><a href="mod-06.html#cb791-1" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(Petal.Length<span class="sc">~</span>Species, <span class="at">data=</span>iris)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-463-1.png" width="576" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb792"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb792-1"><a href="mod-06.html#cb792-1" aria-hidden="true" tabindex="-1"></a><span class="co"># equivalent (not run):</span></span>
<span id="cb792-2"><a href="mod-06.html#cb792-2" aria-hidden="true" tabindex="-1"></a><span class="co">#boxplot(iris$Petal.Length~iris$Species)</span></span></code></pre></div>
<p>The boxplot shows the median (thick line), 25<sup>th</sup> and 75<sup>th</sup> percentiles (limits of box), and outliers (whiskers and sometimes points). If you use an ANOVA or similar method to test for group differences, a boxplot is a great way to visualize the test.</p>
<p>Sometimes I prefer to make a more minimalist version of a boxplot with just points and lines.</p>
<div class="sourceCode" id="cb793"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb793-1"><a href="mod-06.html#cb793-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get means and +- 1 SD</span></span>
<span id="cb793-2"><a href="mod-06.html#cb793-2" aria-hidden="true" tabindex="-1"></a>agg <span class="ot">&lt;-</span> <span class="fu">aggregate</span>(Petal.Length<span class="sc">~</span>Species, <span class="at">data=</span>iris, mean)</span>
<span id="cb793-3"><a href="mod-06.html#cb793-3" aria-hidden="true" tabindex="-1"></a>agg<span class="sc">$</span>sd <span class="ot">&lt;-</span> <span class="fu">aggregate</span>(Petal.Length<span class="sc">~</span>Species, <span class="at">data=</span>iris, sd)<span class="sc">$</span>Petal.Length</span>
<span id="cb793-4"><a href="mod-06.html#cb793-4" aria-hidden="true" tabindex="-1"></a>agg<span class="sc">$</span>lo <span class="ot">&lt;-</span> agg<span class="sc">$</span>Petal.Length<span class="sc">-</span>agg<span class="sc">$</span>sd</span>
<span id="cb793-5"><a href="mod-06.html#cb793-5" aria-hidden="true" tabindex="-1"></a>agg<span class="sc">$</span>up <span class="ot">&lt;-</span> agg<span class="sc">$</span>Petal.Length<span class="sc">+</span>agg<span class="sc">$</span>sd</span>
<span id="cb793-6"><a href="mod-06.html#cb793-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb793-7"><a href="mod-06.html#cb793-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, agg<span class="sc">$</span>Petal.Length, <span class="at">type=</span><span class="st">&quot;n&quot;</span>, <span class="at">xaxt=</span><span class="st">&quot;n&quot;</span>,</span>
<span id="cb793-8"><a href="mod-06.html#cb793-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">xlim=</span><span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">3.5</span>), <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">7</span>),</span>
<span id="cb793-9"><a href="mod-06.html#cb793-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">xlab=</span><span class="st">&quot;&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;Petal length (cm)&quot;</span>)</span>
<span id="cb793-10"><a href="mod-06.html#cb793-10" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, agg<span class="sc">$</span>lo, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, agg<span class="sc">$</span>up, <span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb793-11"><a href="mod-06.html#cb793-11" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, agg<span class="sc">$</span>Petal.Length, <span class="at">pch=</span><span class="dv">21</span>, <span class="at">bg=</span><span class="st">&quot;white&quot;</span>, <span class="at">cex=</span><span class="dv">2</span>, <span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb793-12"><a href="mod-06.html#cb793-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb793-13"><a href="mod-06.html#cb793-13" aria-hidden="true" tabindex="-1"></a><span class="co"># custom axis with species names in italics</span></span>
<span id="cb793-14"><a href="mod-06.html#cb793-14" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="at">side=</span><span class="dv">1</span>, <span class="at">at=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="at">labels=</span><span class="fu">paste</span>(<span class="st">&quot;I.&quot;</span>, agg<span class="sc">$</span>Species), <span class="at">font.axis=</span><span class="dv">3</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-464-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>There are lots of ways to customize such a plot. Maybe use median and IQR, or median and quartiles, or whatever summaries help you make sense of your data. Some people like the “whiskers” to have little perpendicular bits at the end, like the base <code>boxplot()</code> function. In R this is done with the <code>arrows()</code> function.</p>
<div class="sourceCode" id="cb794"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb794-1"><a href="mod-06.html#cb794-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get means and +- 1 SD</span></span>
<span id="cb794-2"><a href="mod-06.html#cb794-2" aria-hidden="true" tabindex="-1"></a>agg <span class="ot">&lt;-</span> <span class="fu">aggregate</span>(Petal.Length<span class="sc">~</span>Species, <span class="at">data=</span>iris, mean)</span>
<span id="cb794-3"><a href="mod-06.html#cb794-3" aria-hidden="true" tabindex="-1"></a>agg<span class="sc">$</span>sd <span class="ot">&lt;-</span> <span class="fu">aggregate</span>(Petal.Length<span class="sc">~</span>Species, <span class="at">data=</span>iris, sd)<span class="sc">$</span>Petal.Length</span>
<span id="cb794-4"><a href="mod-06.html#cb794-4" aria-hidden="true" tabindex="-1"></a>agg<span class="sc">$</span>lo <span class="ot">&lt;-</span> agg<span class="sc">$</span>Petal.Length<span class="sc">-</span>agg<span class="sc">$</span>sd</span>
<span id="cb794-5"><a href="mod-06.html#cb794-5" aria-hidden="true" tabindex="-1"></a>agg<span class="sc">$</span>up <span class="ot">&lt;-</span> agg<span class="sc">$</span>Petal.Length<span class="sc">+</span>agg<span class="sc">$</span>sd</span>
<span id="cb794-6"><a href="mod-06.html#cb794-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb794-7"><a href="mod-06.html#cb794-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, agg<span class="sc">$</span>Petal.Length, <span class="at">type=</span><span class="st">&quot;n&quot;</span>, <span class="at">xaxt=</span><span class="st">&quot;n&quot;</span>,</span>
<span id="cb794-8"><a href="mod-06.html#cb794-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">xlim=</span><span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">3.5</span>), <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">7</span>),</span>
<span id="cb794-9"><a href="mod-06.html#cb794-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">xlab=</span><span class="st">&quot;&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;Petal length (cm)&quot;</span>)</span>
<span id="cb794-10"><a href="mod-06.html#cb794-10" aria-hidden="true" tabindex="-1"></a><span class="fu">arrows</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, agg<span class="sc">$</span>lo, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, agg<span class="sc">$</span>up, <span class="at">lwd=</span><span class="dv">2</span>,</span>
<span id="cb794-11"><a href="mod-06.html#cb794-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">angle=</span><span class="dv">90</span>,     <span class="co"># arrow head perpendicular</span></span>
<span id="cb794-12"><a href="mod-06.html#cb794-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">length=</span><span class="fl">0.05</span>,  <span class="co"># length of arrow head (trial and error)</span></span>
<span id="cb794-13"><a href="mod-06.html#cb794-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">code=</span><span class="dv">3</span>)       <span class="co"># 3 = head at both ends</span></span>
<span id="cb794-14"><a href="mod-06.html#cb794-14" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, agg<span class="sc">$</span>Petal.Length, <span class="at">pch=</span><span class="dv">21</span>, <span class="at">bg=</span><span class="st">&quot;white&quot;</span>, <span class="at">cex=</span><span class="dv">2</span>, <span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb794-15"><a href="mod-06.html#cb794-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb794-16"><a href="mod-06.html#cb794-16" aria-hidden="true" tabindex="-1"></a><span class="co"># custom axis with species names in italics</span></span>
<span id="cb794-17"><a href="mod-06.html#cb794-17" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="at">side=</span><span class="dv">1</span>, <span class="at">at=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="at">labels=</span><span class="fu">paste</span>(<span class="st">&quot;I.&quot;</span>, agg<span class="sc">$</span>Species), <span class="at">font.axis=</span><span class="dv">3</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-465-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>Boxplots can be made using multiple variables, but this gets tricky. Best practice is to color or shade the boxes to help readers tell which box is which.</p>
<div class="sourceCode" id="cb795"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb795-1"><a href="mod-06.html#cb795-1" aria-hidden="true" tabindex="-1"></a><span class="co"># spare copy of iris</span></span>
<span id="cb795-2"><a href="mod-06.html#cb795-2" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> iris</span>
<span id="cb795-3"><a href="mod-06.html#cb795-3" aria-hidden="true" tabindex="-1"></a>a<span class="sc">$</span>color <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;pink&quot;</span>, <span class="st">&quot;white&quot;</span>)</span>
<span id="cb795-4"><a href="mod-06.html#cb795-4" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(Petal.Length<span class="sc">~</span>Species<span class="sc">+</span>color, <span class="at">data=</span>a,</span>
<span id="cb795-5"><a href="mod-06.html#cb795-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">col=</span><span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">&quot;pink&quot;</span>, <span class="st">&quot;white&quot;</span>), <span class="at">each=</span><span class="dv">3</span>))</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/chunkx510-1.png" width="768" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb796"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb796-1"><a href="mod-06.html#cb796-1" aria-hidden="true" tabindex="-1"></a><span class="co"># organized differently:</span></span>
<span id="cb796-2"><a href="mod-06.html#cb796-2" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(Petal.Length<span class="sc">~</span>color<span class="sc">+</span>Species, <span class="at">data=</span>a,</span>
<span id="cb796-3"><a href="mod-06.html#cb796-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">col=</span><span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">&quot;pink&quot;</span>, <span class="st">&quot;white&quot;</span>), <span class="at">times=</span><span class="dv">3</span>))</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/chunkx510-2.png" width="768" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb797"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb797-1"><a href="mod-06.html#cb797-1" aria-hidden="true" tabindex="-1"></a><span class="co"># equivalent (not run):</span></span>
<span id="cb797-2"><a href="mod-06.html#cb797-2" aria-hidden="true" tabindex="-1"></a><span class="co">#boxplot(Petal.Length~color+Species, data=a,</span></span>
<span id="cb797-3"><a href="mod-06.html#cb797-3" aria-hidden="true" tabindex="-1"></a><span class="co">#    col=c(&quot;pink&quot;, &quot;white&quot;))</span></span></code></pre></div>
</div>
<div id="barplots" class="section level3 hasAnchor" number="6.4.2">
<h3><span class="header-section-number">6.4.2</span> Barplots<a href="mod-06.html#barplots" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Barplots are usually used to show the <em>mean</em> within each group. But, that’s all they show. Barplots might be more appropriate than barplots if you have so many groups that a boxplot is too cluttered to read, or if the distributions within each group are known to be symmetrical and about the same width (and thus, showing them is not necessary).</p>
<p>Unlike <code>boxplot()</code>, <code>barplot()</code> does not calculate its own values. The values supplied to <code>barplot()</code> will be used as the bar heights. This means you need to calculate the means (or whatever) ahead of time using a function like <code>aggregate()</code>. In the example, below, SD is also calculated so the plot will include some measure of variability.</p>
<p>Another difference is that unlike <code>boxplot()</code>, the <em>x</em> coordinates in a barplot are calculated by the <code>barplot()</code> function. What this means is that you often need to run it twice: once to get the <em>x</em> coordinates, and again to make the plot. Those <em>x</em> coordinates are needed to place the line segments that show the SD.</p>
<div class="sourceCode" id="cb798"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb798-1"><a href="mod-06.html#cb798-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get means and SD</span></span>
<span id="cb798-2"><a href="mod-06.html#cb798-2" aria-hidden="true" tabindex="-1"></a>agg <span class="ot">&lt;-</span> <span class="fu">aggregate</span>(Petal.Length<span class="sc">~</span>Species, <span class="at">data=</span>iris, mean)</span>
<span id="cb798-3"><a href="mod-06.html#cb798-3" aria-hidden="true" tabindex="-1"></a>agg<span class="sc">$</span>sd <span class="ot">&lt;-</span> <span class="fu">aggregate</span>(Petal.Length<span class="sc">~</span>Species, <span class="at">data=</span>iris, sd)<span class="sc">$</span>Petal.Length</span>
<span id="cb798-4"><a href="mod-06.html#cb798-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb798-5"><a href="mod-06.html#cb798-5" aria-hidden="true" tabindex="-1"></a><span class="co"># get x axis values</span></span>
<span id="cb798-6"><a href="mod-06.html#cb798-6" aria-hidden="true" tabindex="-1"></a>bar <span class="ot">&lt;-</span> <span class="fu">barplot</span>(Petal.Length<span class="sc">~</span>Species, <span class="at">data=</span>agg)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-466-1.png" width="576" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb799"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb799-1"><a href="mod-06.html#cb799-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make plot with second call</span></span>
<span id="cb799-2"><a href="mod-06.html#cb799-2" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(Petal.Length<span class="sc">~</span>Species, <span class="at">data=</span>agg, <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">7</span>))</span>
<span id="cb799-3"><a href="mod-06.html#cb799-3" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(bar[,<span class="dv">1</span>], agg<span class="sc">$</span>Petal.Length, bar[,<span class="dv">1</span>], agg<span class="sc">$</span>Petal.Length<span class="sc">+</span>agg<span class="sc">$</span>sd)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-466-2.png" width="576" style="display: block; margin: auto;" /></p>
<p>The above plot is called a “dynamite plot” because the bar and error bar look like an old stick of dynamite. Because they show less than a boxplot, and are harder to make than a boxplot, I almost never make dynamite plots. Instead, I suggest you use boxplots or something like the “point-and-whisker” plot that I showed above.</p>
</div>
</div>
<div id="ordination-brief-introduction" class="section level2 hasAnchor" number="6.5">
<h2><span class="header-section-number">6.5</span> Ordination (brief introduction)<a href="mod-06.html#ordination-brief-introduction" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The methods above are a way to examine single variables, or single pairs of variables. But what about many variables at once? In other words, how can we examine variation between samples by taking into account multiple variables at once? This is the realm of <strong>multivariate statistics</strong>.</p>
<p><strong>Ordination</strong> is literally plotting or ordering observations along two or more axes. How that ordering is done varies wildly among techniques. We will get deeper into ordination later in the course, but for now let’s look at just two ordination techniques, <strong>principal components analysis (PCA)</strong> and <strong>nonmetric multidimensional scaling (NMDS or NMS)</strong>, with a focus on interpretation rather than calculation. Some of the goals of all ordination techniques are:</p>
<ul>
<li><strong>Cluster identification:</strong> observations that are closer to each other in the ordination space are more similar to each other</li>
<li><strong>Dimension reduction:</strong> the axes of an ordination are estimates of synthetic variables that combine information about many variables at once</li>
</ul>
<div id="principal-components-analysis-pca" class="section level3 hasAnchor" number="6.5.1">
<h3><span class="header-section-number">6.5.1</span> Principal components analysis (PCA)<a href="mod-06.html#principal-components-analysis-pca" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Principal components analysis (PCA) is a method for extracting gradients from a multivariate dataset that capture most of the variation in that dataset. These gradients are calculated by finding linear combinations of the variables that minimize sums of squared deviations from the gradient. This means that PCA has a lot in common with linear regression, and many of the same assumptions apply.</p>
<p>If you’ve never heard of PCA or ordination, it might be worth watching a video that explains and shows the basic ideas. Here is <a href="https://www.youtube.com/watch?v=HMOI_lkzW08"><strong>one that only takes 5 minutes and has a nice theme song</strong></a> (accessed 2021-08-10). For a more in-depth introduction to PCA, see <a href="mod-12.html#mod-12-ord">this page</a>. Then come back here for some notes on how to use PCA for data exploration.</p>
<p>PCA is available in base R using the functions <code>prcomp()</code> and <code>princomp()</code>. Both of these functions produce similar outputs, but there are some subtle differences between them. For this course we are going to use the functions in package <code>vegan</code>. Although designed for community ecology, <code>vegan</code> is widely used for ordination and multivariate analysis in many fields. Importantly, <code>vegan</code> is actively maintained and updated with new techniques as they are developed. Also importantly, <code>vegan</code> offers a common interface for many kinds of ordination.</p>
<p>The example below uses the <code>varespec</code> dataset from <code>vegan</code>. Each column contains % cover by different species of plants in lichen pastures in northern Finland and Scandinavia. For the sake of making a simple example, we will ordinate only the 15 species with the most variance (many species in the dataset are rare or sparse, so including them makes the plot harder to interpret…in “real life”, don’t remove variables from your analysis without good reason!).</p>
<div class="sourceCode" id="cb800"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb800-1"><a href="mod-06.html#cb800-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vegan)</span></code></pre></div>
<pre><code>## Loading required package: permute</code></pre>
<pre><code>## This is vegan 2.6-2</code></pre>
<div class="sourceCode" id="cb803"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb803-1"><a href="mod-06.html#cb803-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get data and make a spare copy</span></span>
<span id="cb803-2"><a href="mod-06.html#cb803-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(varespec)</span>
<span id="cb803-3"><a href="mod-06.html#cb803-3" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> varespec</span>
<span id="cb803-4"><a href="mod-06.html#cb803-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb803-5"><a href="mod-06.html#cb803-5" aria-hidden="true" tabindex="-1"></a><span class="co"># identify the variables with most variance</span></span>
<span id="cb803-6"><a href="mod-06.html#cb803-6" aria-hidden="true" tabindex="-1"></a>vx <span class="ot">&lt;-</span> <span class="fu">rev</span>(<span class="fu">sort</span>(<span class="fu">apply</span>(x, <span class="dv">2</span>, var)))</span>
<span id="cb803-7"><a href="mod-06.html#cb803-7" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> x[,<span class="fu">names</span>(vx)[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]]</span>
<span id="cb803-8"><a href="mod-06.html#cb803-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb803-9"><a href="mod-06.html#cb803-9" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate PCA and plot it</span></span>
<span id="cb803-10"><a href="mod-06.html#cb803-10" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">rda</span>(x, <span class="at">scale=</span><span class="cn">TRUE</span>)</span>
<span id="cb803-11"><a href="mod-06.html#cb803-11" aria-hidden="true" tabindex="-1"></a><span class="fu">biplot</span>(p1)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-467-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>The figure above is called a <strong>PCA biplot</strong>. Each point is a sample (site) where vegetation was sampled. By default, the samples are identified by their row number. The red arrows show sites where different species were more abundant (or, more generally, where the variables are increasing). For example, sites towards the top right had greater cover by species “Cladstel” (short for <em>Cladonia stellaris</em>). Sites towards the lower left had less cover by Cladstel. Sites towards the top left had more cover by “Pleuschr” (<em>Pleurozium schreberi</em>).</p>
<p>The <code>summary()</code> command will print a lot of information about the ordination. It prints the eigenvalues of each PC, then scales them as “Proportion explained”. I.e., the proportion of variation explained by each PC is the same as its eigenvalue divided by the sum of all eigenvalues.</p>
<div class="sourceCode" id="cb804"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb804-1"><a href="mod-06.html#cb804-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(p1)</span></code></pre></div>
<pre><code>## 
## Call:
## rda(X = x, scale = TRUE) 
## 
## Partitioning of correlations:
##               Inertia Proportion
## Total              10          1
## Unconstrained      10          1
## 
## Eigenvalues, and their contribution to the correlations 
## 
## Importance of components:
##                          PC1    PC2    PC3    PC4     PC5    PC6     PC7
## Eigenvalue            2.7808 1.9382 1.3148 1.0987 0.83622 0.7150 0.69134
## Proportion Explained  0.2781 0.1938 0.1315 0.1099 0.08362 0.0715 0.06913
## Cumulative Proportion 0.2781 0.4719 0.6034 0.7132 0.79687 0.8684 0.93750
##                           PC8     PC9    PC10
## Eigenvalue            0.33428 0.17151 0.11919
## Proportion Explained  0.03343 0.01715 0.01192
## Cumulative Proportion 0.97093 0.98808 1.00000
## 
## Scaling 2 for species and site scores
## * Species are scaled proportional to eigenvalues
## * Sites are unscaled: weighted dispersion equal on all dimensions
## * General scaling constant of scores:  3.894323 
## 
## 
## Species scores
## 
##               PC1        PC2     PC3     PC4      PC5      PC6
## Cladstel  0.22980 -0.9638221  0.2063 -0.5766  0.22839 -0.13546
## Pleuschr -0.94726  0.3626752 -0.2495  0.2859  0.31322 -0.08035
## Cladrang  1.03450 -0.0478938 -0.2618  0.4311  0.14433 -0.14610
## Cladarbu  0.92222  0.5003246 -0.1221  0.3915 -0.31092  0.13843
## Dicrfusc -0.33989  0.7013881 -0.1721 -0.5513 -0.19752 -0.32352
## Vaccviti -0.57531 -0.5738560  0.1310  0.1739 -0.66855  0.52296
## Dicrsp   -0.38606  0.2903390  0.8144  0.3716  0.52606  0.26321
## Callvulg  0.34621  0.5471556 -0.3068 -0.6120  0.25486  0.73026
## Cladunci -0.02526  0.6491055  0.7160 -0.1866 -0.41338 -0.19511
## Vaccmyrt -0.79400 -0.0005864 -0.6992  0.1767 -0.04322 -0.01670
## 
## 
## Site scores (weighted sums of species scores)
## 
##         PC1      PC2       PC3      PC4      PC5        PC6
## 18  0.39657 -0.16342 -0.079895  0.68235 -0.98085  0.6198150
## 15 -0.40025  0.48524 -0.119881  0.10251 -0.25276 -0.2907006
## 24 -0.88573  0.67138  2.334307  1.08698  1.26348  0.8403431
## 27 -1.10452 -0.03240 -0.660890  0.55886  0.14165 -0.0003119
## 23 -0.40297 -0.35825  0.084662  0.42892 -1.19313  0.8452582
## 19 -0.26161 -0.30040 -0.123290  0.01983  0.17886 -0.1684937
## 22 -0.85012  1.13056 -0.797382 -1.33729 -0.65017 -0.5866657
## 16 -0.15994  0.98635 -0.571647 -0.91477  0.37822 -0.8158903
## 28 -1.65484  0.25754 -1.629606  0.80391  1.00293 -0.5951382
## 13  1.11019  1.09130 -0.918028 -1.32805  1.01767  2.7409133
## 14  0.14308  1.65235  1.451371 -1.01710 -1.53004 -0.9476702
## 20 -0.14436  0.02087  0.385155  0.19971 -0.55881  0.0248203
## 25 -0.70668  0.52484  0.951767  0.41019  0.99674  0.1995443
## 7   1.27601  0.28464 -0.392360  1.25885 -0.39985 -0.1596378
## 5   1.33049  0.15587 -0.500076  1.08816  0.58014 -0.9228755
## 6   1.24463  0.31814 -0.096950  1.06221 -0.85622 -0.0198950
## 3   0.70748 -0.62339  0.020031 -0.26749  0.77095 -0.8805476
## 4   0.73772  0.02904 -0.134862 -0.22867  0.70704 -0.3411419
## 2   0.39724 -1.10654  0.219538 -0.67452  0.72211 -0.6843507
## 9  -0.23911 -1.53329  0.521191 -0.91391 -0.36176  0.3560478
## 12  0.05053 -1.04256  0.379673 -0.50633 -0.09530  0.0281206
## 10  0.09930 -1.27173  0.418480 -1.00789  0.35197 -0.2898518
## 11  0.32926 -0.31288 -0.003708  0.12732  0.07082  0.0582269
## 21 -1.01237 -0.86325 -0.737599  0.36622 -1.30368  0.9900814</code></pre>
<p>The “Importance of components” part tells us that the first principal component (PC1) explains about 55% of the variation in the dataset. PC2 explains an additional 26%. The rule of thumb is to present enough PCs to account for <span class="math inline">\(\ge\)</span> 80% of the variation. In this case, PC1 and PC2 are sufficient. If it takes many PCs to reach 80% of variation, that may be a sign that the PCs are not capturing much meaningful structure in your dataset (or, there simply isn’t much structure to capture).</p>
<p>Finally, we can see the variable “loadings”, which tell us how much each variable is correlated with each principal component<a href="#fn59" class="footnote-ref" id="fnref59"><sup>59</sup></a>. These values are interpreted the same way as linear correlation coefficients. In the example below the loadings for PCs 1-4 are requested using argument choices = 1:4.</p>
<div class="sourceCode" id="cb806"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb806-1"><a href="mod-06.html#cb806-1" aria-hidden="true" tabindex="-1"></a><span class="fu">scores</span>(p1, <span class="at">choices =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="at">display =</span> <span class="st">&quot;species&quot;</span>, <span class="at">scaling =</span> <span class="dv">0</span>)</span></code></pre></div>
<pre><code>##                 PC1           PC2         PC3        PC4
## Cladstel  0.1118986 -0.5621722123  0.14606363 -0.4467186
## Pleuschr -0.4612677  0.2115389675 -0.17670529  0.2214946
## Cladrang  0.5037497 -0.0279351716 -0.18537465  0.3339431
## Cladarbu  0.4490762  0.2918262462 -0.08645467  0.3032787
## Dicrfusc -0.1655069  0.4091013432 -0.12191016 -0.4270946
## Vaccviti -0.2801468 -0.3347151730  0.09280300  0.1347303
## Dicrsp   -0.1879928  0.1693471498  0.57673161  0.2878989
## Callvulg  0.1685866  0.3191415425 -0.21724150 -0.4741263
## Cladunci -0.0123003  0.3786062372  0.50705444 -0.1445745
## Vaccmyrt -0.3866379 -0.0003420561 -0.49518742  0.1368990
## attr(,&quot;const&quot;)
## [1] 3.894323</code></pre>
<p>The loadings tell us that PC1 is strongly influenced by species <code>Cladstel</code> and less so by <code>Pleuschr</code>. Compare this to the orientation and lengths of the arrows for these species in the biplot and see if you can find a connection.</p>
</div>
<div id="nonmetric-multidimensional-scaling-nmds" class="section level3 hasAnchor" number="6.5.2">
<h3><span class="header-section-number">6.5.2</span> Nonmetric multidimensional scaling (NMDS)<a href="mod-06.html#nonmetric-multidimensional-scaling-nmds" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>PCA and many related techniques are based on linear algebra and eigenvalues. This is fine for datasets where variables are mostly linearly related to each other, or can be transformed to be linearly related. Most of the eigenvalue-based techniques also require data to be mostly normally distributed.</p>
<p>If relationships between variables are nonlinear, or if the many other assumptions of eigenvalue-based ordination cannot be met, then the next best option is a non-parametric ordination technique called nonmetric multidimensional scaling (NMDS or NMS). Unlike PCA, which solves linear algebra problems to extract synthetic gradients (“principal components”), NMDS works by trying iteratively to arrange the samples into a reduced dimensional space that preserves the rank order of the distance matrix.</p>
<p>What? Let’s break that down:</p>
<p><strong>Distance matrix:</strong> a matrix containing a measure of dissimilarity (aka: “distance”) between each pair of samples. This is exactly analogous to a road mileage chart in an old paper road atlas.</p>
<p><strong>Distance (dissimilarity) measure:</strong> a quantity that measures how different two samples are in terms of several variables. Greater values indicate that two samples are more different from each other; smaller values indicate that two samples are more similar to each other. See <a href="mod-12.html#mod-12-dist">this page</a> for a more in-depth explanation of distance measures.</p>
<p>The most famous distance metric is the Euclidean distance metric, shown here as the distance between two points on the <em>XY</em> plane:</p>
<p><span class="math display">\[D_{Euc.}=\sqrt{\left(x_1-x_2\right)^2+\left(y_1-y_2\right)^2}\]</span></p>
<p>The metric can be generalized to any number of dimensions <em>k</em>.</p>
<p><span class="math display">\[D_{Euc.}=\sqrt{\sum_{k=1}^{k}\left(x_1-x_2\right)^2}\]</span></p>
<p>The Euclidean distance is rarely used for NMDS. Instead, the default distance metric is the <strong>Bray-Curtis distance</strong>. This metric is called <strong>Sørensen distance</strong> when used for binary values (e.g., presence-absence), and Bray-Curtis when used for quantitative values.</p>
<p>The distance matrix contains the dissimilarities between samples, in terms of all of the variables in the dataset. If there are <em>k</em> variables, then these are distances through <em>k</em>-dimensional space. NMDS tries to represent the relative dissimilarities between samples in fewer than <em>k</em> dimensions, while preserving the <strong>rank order</strong> of those dissimilarities.</p>
<p>The R function <code>metaMDS()</code> performs NMDS. Getting a biplot of the ordination is a little more involved than for PCA. Note that NMDS involves some random number sampling in its algorithms, so you should set the random number seed to make sure your analysis is reproducible.</p>
<div class="sourceCode" id="cb808"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb808-1"><a href="mod-06.html#cb808-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb808-2"><a href="mod-06.html#cb808-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb808-3"><a href="mod-06.html#cb808-3" aria-hidden="true" tabindex="-1"></a><span class="co"># fit NMDS</span></span>
<span id="cb808-4"><a href="mod-06.html#cb808-4" aria-hidden="true" tabindex="-1"></a>n1 <span class="ot">&lt;-</span> <span class="fu">metaMDS</span>(x)</span></code></pre></div>
<pre><code>## Square root transformation
## Wisconsin double standardization
## Run 0 stress 0.1258291 
## Run 1 stress 0.128478 
## Run 2 stress 0.128478 
## Run 3 stress 0.1273261 
## Run 4 stress 0.2014993 
## Run 5 stress 0.1991787 
## Run 6 stress 0.128478 
## Run 7 stress 0.128478 
## Run 8 stress 0.128478 
## Run 9 stress 0.1258291 
## ... Procrustes: rmse 2.910913e-06  max resid 9.961616e-06 
## ... Similar to previous best
## Run 10 stress 0.1258291 
## ... Procrustes: rmse 1.699748e-06  max resid 4.320923e-06 
## ... Similar to previous best
## Run 11 stress 0.128478 
## Run 12 stress 0.128478 
## Run 13 stress 0.128478 
## Run 14 stress 0.1258291 
## ... Procrustes: rmse 1.791907e-06  max resid 5.731764e-06 
## ... Similar to previous best
## Run 15 stress 0.128478 
## Run 16 stress 0.1976983 
## Run 17 stress 0.1258291 
## ... Procrustes: rmse 3.90069e-06  max resid 1.369466e-05 
## ... Similar to previous best
## Run 18 stress 0.1273261 
## Run 19 stress 0.1258291 
## ... Procrustes: rmse 2.251913e-06  max resid 7.461718e-06 
## ... Similar to previous best
## Run 20 stress 0.235875 
## *** Solution reached</code></pre>
<div class="sourceCode" id="cb810"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb810-1"><a href="mod-06.html#cb810-1" aria-hidden="true" tabindex="-1"></a><span class="co"># extract &quot;scores&quot; (coordinates in NMDS space)</span></span>
<span id="cb810-2"><a href="mod-06.html#cb810-2" aria-hidden="true" tabindex="-1"></a><span class="co"># these are matrices: column 1 = x, column 2 = y</span></span>
<span id="cb810-3"><a href="mod-06.html#cb810-3" aria-hidden="true" tabindex="-1"></a>n1s <span class="ot">&lt;-</span> <span class="fu">scores</span>(n1, <span class="st">&quot;sites&quot;</span>)</span>
<span id="cb810-4"><a href="mod-06.html#cb810-4" aria-hidden="true" tabindex="-1"></a>n1p <span class="ot">&lt;-</span> <span class="fu">scores</span>(n1, <span class="st">&quot;species&quot;</span>)</span>
<span id="cb810-5"><a href="mod-06.html#cb810-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb810-6"><a href="mod-06.html#cb810-6" aria-hidden="true" tabindex="-1"></a><span class="co"># basic NMDS plot</span></span>
<span id="cb810-7"><a href="mod-06.html#cb810-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(n1s)</span>
<span id="cb810-8"><a href="mod-06.html#cb810-8" aria-hidden="true" tabindex="-1"></a><span class="co"># add biplot arrows and labels</span></span>
<span id="cb810-9"><a href="mod-06.html#cb810-9" aria-hidden="true" tabindex="-1"></a><span class="fu">segments</span>(<span class="dv">0</span>, <span class="dv">0</span>, n1p[,<span class="dv">1</span>], n1p[,<span class="dv">2</span>], <span class="at">col=</span><span class="st">&quot;red&quot;</span>)</span>
<span id="cb810-10"><a href="mod-06.html#cb810-10" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(n1p[,<span class="dv">1</span>], n1p[,<span class="dv">2</span>], <span class="fu">rownames</span>(n1p), <span class="at">col=</span><span class="st">&quot;red&quot;</span>, <span class="at">xpd=</span><span class="cn">NA</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-470-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Because NMDS solutions are reached through random sampling, they are not guaranteed to reach a stable condition every time. You will need to set the random number seed to get reproducible results. There is also no guarantee that the solution reached by NMDS will be the optimal one.</p>
<p>Another consequence of the way that NMDS works is that the axes themselves are arbitrary, unlike the axes in PCA. This means that we can’t calculate a percentage of variation explained by each axis the way we did for PCA. By default, R will rotate NMDS solutions so that the greatest amount of variation is associated with axis 1, but there is no guarantee that this axis captures a meaningful biological gradient.</p>
<p>The most important diagnostic parameter in an NMDS is the <strong>stress</strong>, which is a measure of how much the distance matrix is distorted by compression into low dimensional space. Lower stress value indicate better fit. Most people consider a stress <span class="math inline">\(\ge\)</span> 0.2 to indicate an unreliable fit. Some authors recommend lower thresholds like 0.15 or 0.1. Package <code>vegan</code> and most sources use stress values &lt;1 <span class="citation">(e.g., <a href="#ref-legendre2012numerical" role="doc-biblioref">Legendre and Legendre 2012</a>)</span>; some authors scale stress by a factor of 100 relative to the values in <code>vegan</code>. For example, the program PC-ORD <span class="citation">(<a href="#ref-mccune2002analysis" role="doc-biblioref">McCune et al. 2002</a>)</span> uses stress values on the 100<span class="math inline">\(\times\)</span> scale.</p>
</div>
<div id="plotting-ordinations" class="section level3 hasAnchor" number="6.5.3">
<h3><span class="header-section-number">6.5.3</span> Plotting ordinations<a href="mod-06.html#plotting-ordinations" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Extracting the scores from an ordination object can let you plot points with all of the <code>base</code> graphics formatting tricks we learned above, such as using color or shape to indicate group membership. In my experience this is more flexible than using the built-in graphics capabilities in <code>vegan</code><a href="#fn60" class="footnote-ref" id="fnref60"><sup>60</sup></a>, but your mileage may vary.</p>
<div class="sourceCode" id="cb811"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb811-1"><a href="mod-06.html#cb811-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> crabs</span>
<span id="cb811-2"><a href="mod-06.html#cb811-2" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">rda</span>(x[,<span class="dv">4</span><span class="sc">:</span><span class="dv">8</span>])</span>
<span id="cb811-3"><a href="mod-06.html#cb811-3" aria-hidden="true" tabindex="-1"></a>n1 <span class="ot">&lt;-</span> <span class="fu">metaMDS</span>(x[,<span class="dv">4</span><span class="sc">:</span><span class="dv">8</span>])</span></code></pre></div>
<pre><code>## Square root transformation
## Wisconsin double standardization
## Run 0 stress 0.08013593 
## Run 1 stress 0.08013594 
## ... Procrustes: rmse 2.47761e-06  max resid 3.004629e-05 
## ... Similar to previous best
## Run 2 stress 0.08013594 
## ... Procrustes: rmse 5.318632e-06  max resid 5.42044e-05 
## ... Similar to previous best
## Run 3 stress 0.08013593 
## ... Procrustes: rmse 1.789737e-05  max resid 0.0001945269 
## ... Similar to previous best
## Run 4 stress 0.08013596 
## ... Procrustes: rmse 1.021893e-05  max resid 0.0001280391 
## ... Similar to previous best
## Run 5 stress 0.08013593 
## ... New best solution
## ... Procrustes: rmse 3.631654e-06  max resid 4.608495e-05 
## ... Similar to previous best
## Run 6 stress 0.08013593 
## ... Procrustes: rmse 4.019493e-06  max resid 5.419834e-05 
## ... Similar to previous best
## Run 7 stress 0.08013592 
## ... New best solution
## ... Procrustes: rmse 6.01397e-06  max resid 8.261179e-05 
## ... Similar to previous best
## Run 8 stress 0.08013594 
## ... Procrustes: rmse 1.059936e-05  max resid 0.0001315101 
## ... Similar to previous best
## Run 9 stress 0.08013592 
## ... New best solution
## ... Procrustes: rmse 2.632514e-06  max resid 3.234954e-05 
## ... Similar to previous best
## Run 10 stress 0.08013593 
## ... Procrustes: rmse 1.374691e-05  max resid 0.0001352853 
## ... Similar to previous best
## Run 11 stress 0.08013593 
## ... Procrustes: rmse 9.848983e-06  max resid 9.509947e-05 
## ... Similar to previous best
## Run 12 stress 0.08013593 
## ... Procrustes: rmse 9.871024e-06  max resid 0.0001008813 
## ... Similar to previous best
## Run 13 stress 0.08013593 
## ... Procrustes: rmse 6.457749e-06  max resid 6.335005e-05 
## ... Similar to previous best
## Run 14 stress 0.08013595 
## ... Procrustes: rmse 1.626131e-05  max resid 0.0001518457 
## ... Similar to previous best
## Run 15 stress 0.08013595 
## ... Procrustes: rmse 1.446168e-05  max resid 0.0001679656 
## ... Similar to previous best
## Run 16 stress 0.08013593 
## ... Procrustes: rmse 1.148406e-05  max resid 0.0001448735 
## ... Similar to previous best
## Run 17 stress 0.08013593 
## ... Procrustes: rmse 9.122826e-06  max resid 0.000112683 
## ... Similar to previous best
## Run 18 stress 0.08013594 
## ... Procrustes: rmse 1.22868e-05  max resid 0.0001430928 
## ... Similar to previous best
## Run 19 stress 0.08013595 
## ... Procrustes: rmse 1.588575e-05  max resid 0.000217428 
## ... Similar to previous best
## Run 20 stress 0.08013592 
## ... Procrustes: rmse 2.291841e-06  max resid 1.545649e-05 
## ... Similar to previous best
## *** Solution reached</code></pre>
<div class="sourceCode" id="cb813"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb813-1"><a href="mod-06.html#cb813-1" aria-hidden="true" tabindex="-1"></a>p1s <span class="ot">&lt;-</span> <span class="fu">scores</span>(p1)<span class="sc">$</span>sites</span>
<span id="cb813-2"><a href="mod-06.html#cb813-2" aria-hidden="true" tabindex="-1"></a>p1p <span class="ot">&lt;-</span> <span class="fu">scores</span>(p1)<span class="sc">$</span>species</span>
<span id="cb813-3"><a href="mod-06.html#cb813-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb813-4"><a href="mod-06.html#cb813-4" aria-hidden="true" tabindex="-1"></a>n1s <span class="ot">&lt;-</span> <span class="fu">scores</span>(n1, <span class="st">&quot;sites&quot;</span>)</span>
<span id="cb813-5"><a href="mod-06.html#cb813-5" aria-hidden="true" tabindex="-1"></a>n1p <span class="ot">&lt;-</span> <span class="fu">scores</span>(n1, <span class="st">&quot;species&quot;</span>)</span>
<span id="cb813-6"><a href="mod-06.html#cb813-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb813-7"><a href="mod-06.html#cb813-7" aria-hidden="true" tabindex="-1"></a>cols1 <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(x<span class="sc">$</span>sex <span class="sc">==</span> <span class="st">&quot;F&quot;</span>, <span class="st">&quot;red&quot;</span>, <span class="st">&quot;blue&quot;</span>)</span>
<span id="cb813-8"><a href="mod-06.html#cb813-8" aria-hidden="true" tabindex="-1"></a>pch1 <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(x<span class="sc">$</span>sp <span class="sc">==</span> <span class="st">&quot;B&quot;</span>, <span class="dv">2</span>, <span class="dv">1</span>)</span>
<span id="cb813-9"><a href="mod-06.html#cb813-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb813-10"><a href="mod-06.html#cb813-10" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb813-11"><a href="mod-06.html#cb813-11" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(p1s, <span class="at">col=</span>cols1, <span class="at">pch=</span>pch1, <span class="at">main=</span><span class="st">&quot;PCA fit&quot;</span>)</span>
<span id="cb813-12"><a href="mod-06.html#cb813-12" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;topleft&quot;</span>,</span>
<span id="cb813-13"><a href="mod-06.html#cb813-13" aria-hidden="true" tabindex="-1"></a>       <span class="at">legend=</span><span class="fu">c</span>(<span class="st">&quot;B female&quot;</span>, <span class="st">&quot;B male&quot;</span>, <span class="st">&quot;O female&quot;</span>, <span class="st">&quot;O male&quot;</span>),</span>
<span id="cb813-14"><a href="mod-06.html#cb813-14" aria-hidden="true" tabindex="-1"></a>       <span class="at">pch=</span><span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">1</span>),</span>
<span id="cb813-15"><a href="mod-06.html#cb813-15" aria-hidden="true" tabindex="-1"></a>       <span class="at">col=</span><span class="fu">c</span>(<span class="st">&quot;red&quot;</span>, <span class="st">&quot;blue&quot;</span>, <span class="st">&quot;red&quot;</span>, <span class="st">&quot;blue&quot;</span>))</span>
<span id="cb813-16"><a href="mod-06.html#cb813-16" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(n1s, <span class="at">col=</span>cols1, <span class="at">pch=</span>pch1, <span class="at">main=</span><span class="st">&quot;NMDS fit&quot;</span>)</span></code></pre></div>
<p><img src="applied_biol_data_analysis_20220816_files/figure-html/unnamed-chunk-471-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="ordination-wrap-up-for-now" class="section level3 hasAnchor" number="6.5.4">
<h3><span class="header-section-number">6.5.4</span> Ordination wrap-up (for now)<a href="mod-06.html#ordination-wrap-up-for-now" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Ordination is a collection of methods for reducing the dimensionality of data. Biological datasets often have many variables, but most variation in those datasets can often be captured using only a few of them (or few combinations of the variables). The positions of the samples in the ordination space can thus reveal patterns that are too complicated to see in the full data space. Interpreting ordination axes in biological terms can be very fruitful. When an axis captures important information about the samples, it can be used as a variable in another method (e.g., as a predictor variable in a linear model). Analysis of clustering or positional differences between groups in ordination space (especially NMDS space) can allow for researchers to test for differences in many variables at once.</p>
</div>
</div>
<div id="principles-for-good-data-graphics" class="section level2 hasAnchor" number="6.6">
<h2><span class="header-section-number">6.6</span> Principles for good data graphics<a href="mod-06.html#principles-for-good-data-graphics" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><em>Section under construction!</em></p>
<div id="maximize-the-data-to-ink-ratio" class="section level3 hasAnchor" number="6.6.1">
<h3><span class="header-section-number">6.6.1</span> Maximize the data-to-ink ratio<a href="mod-06.html#maximize-the-data-to-ink-ratio" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
<div id="use-labels-not-symbols" class="section level3 hasAnchor" number="6.6.2">
<h3><span class="header-section-number">6.6.2</span> Use labels, not symbols<a href="mod-06.html#use-labels-not-symbols" class="anchor-section" aria-label="Anchor link to header"></a></h3>
</div>
<div id="be-consistent" class="section level3 hasAnchor" number="6.6.3">
<h3><span class="header-section-number">6.6.3</span> Be consistent<a href="mod-06.html#be-consistent" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Often you will have the same variable or kind of measurement in multiple figures in a thesis or manuscript. It is <strong>extremely</strong> helpful to your reader if the symbology stays the same across those figures.</p>
</div>
<div id="consider-legibility" class="section level3 hasAnchor" number="6.6.4">
<h3><span class="header-section-number">6.6.4</span> Consider legibility<a href="mod-06.html#consider-legibility" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Color</p>
<p>Text size</p>
<p>Screens and copiers</p>
</div>
<div id="set-yourself-up-for-success" class="section level3 hasAnchor" number="6.6.5">
<h3><span class="header-section-number">6.6.5</span> Set yourself up for success<a href="mod-06.html#set-yourself-up-for-success" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Keep everything in the same data frame</p>
<p>Avoid hard coding</p>
<p><code>par()</code> is your friend</p>
<p>Save your scripts!</p>

</div>
</div>
</div>
<h3>Literature Cited<a href="literature-cited.html#literature-cited" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent" line-spacing="2">
<div id="ref-bolker2008ecological" class="csl-entry">
Bolker, B. M. 2008. Ecological models and data in <span>R</span>. Princeton University Press, Princeton, New Jersey, USA.
</div>
<div id="ref-legendre2012numerical" class="csl-entry">
Legendre, P., and L. Legendre. 2012. Numerical ecology. Elsevier, Amsterdam.
</div>
<div id="ref-mccune2002analysis" class="csl-entry">
McCune, B., J. B. Grace, and D. L. Urban. 2002. Analysis of ecological communities. MjM software design Gleneden Beach, Oregon, USA.
</div>
</div>
<div class="footnotes">
<hr />
<ol start="56">
<li id="fn56"><p>You can also specify colors by their RGB or hex codes. A graphical reference is <a href="https://rstudio-pubs-static.s3.amazonaws.com/3486_79191ad32cf74955b4502b8530aad627.html">here</a> (accessed 2021-12-17). The <a href="https://colorbrewer2.org/">ColorBrewer</a> website (accessed 2021-12-17) and R package can help picking appropriate color schemes.<a href="mod-06.html#fnref56" class="footnote-back">↩︎</a></p></li>
<li id="fn57"><p>Sometimes you have to be clever with <code>plotmath()</code>. If it’s stupid and it works, then it’s not stupid.<a href="mod-06.html#fnref57" class="footnote-back">↩︎</a></p></li>
<li id="fn58"><p>Locally estimated sums of squares, a common smoothing curve algorithm.<a href="mod-06.html#fnref58" class="footnote-back">↩︎</a></p></li>
<li id="fn59"><p>In the base R PCA functions, these values are obtained from the <code>rotation</code> part of the output in <code>prcomp()</code> or from the <code>loadings</code> component in <code>princomp()</code>.<a href="mod-06.html#fnref59" class="footnote-back">↩︎</a></p></li>
<li id="fn60"><p>An introduction by the package’s lead author can be found here: <a href="https://cran.r-project.org/web/packages/vegan/vignettes/intro-vegan.pdf" class="uri">https://cran.r-project.org/web/packages/vegan/vignettes/intro-vegan.pdf</a> (accessed 2021-09-17)<a href="mod-06.html#fnref60" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="mod-05.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="mod-07.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/06-explore3.rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["applied_biol_data_analysis_20220816.pdf", "applied_biol_data_analysis_20220816.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
